// anchor-server-common/utilities/episode/constants.js
const startAndEndTimeToolArgumentsRegex = /startTime=(\d*\.?\d*)&endTime=(\d*\.?\d*)/;

const CLIENT_TYPE_IOS = 'ios';
const CLIENT_TYPE_ANDROID = 'android';

const mtTalkRatioRuleMinimumClientVersionByClientType = {};
mtTalkRatioRuleMinimumClientVersionByClientType[CLIENT_TYPE_IOS] = '3.73.0';
mtTalkRatioRuleMinimumClientVersionByClientType[CLIENT_TYPE_ANDROID] = '3.84.0';

module.exports = {
    COLORS: {
        AD: '#1cee4e',
        CALL_IN: '#2dcfb3',
        DRAFT: '#8940fA',
        BACKGROUND: '#292f36',
        INTERLUDE: '#fface7',
        SOUND_EFFECT: '#00b3ce',
        MUSIC: '#428eff',
        SPOTIFY_MUSIC: '#509BF5',
        RECORD: '#fd6767',
        RECORD_WITH_FRIENDS: '#d34cef',
        UPLOAD_FROM_WEB: '#8940fa',
    },
    startAndEndTimeToolArgumentsRegex,
    CALLS_TO_ACTION: {
        NONE: 'none',
        PUBLISH: 'publish',
        DISTRIBUTE: 'distributeEverywhere',
        SPONSORSHIP: 'sponsorDetails',
    },
    STATUSES: {
        SCHEDULED: 'scheduled',
        DRAFT: 'draft',
        EVERYWHERE: 'everywhere',
    },
    MAX_EPISODE_TAGS: 5,
    TAG_SEARCH_PERMISSIBLE_LANGUAGES: ['en'],
    MAX_MUSIC_CONTENT_PERCENTAGE_MT_EPISODE: 90.0,
    MUSIC_AND_TALK_PUBLISH_ERRORS: {
        MUSIC_NOT_ALLOWED: 'musicNotAllowed',
        NOT_DISTRIBUTED_TO_SPOTIFY: 'notDistributedToSpotify',
        NOT_TAXI_DISTRIBUTED: 'notTaxiDistributed',
        EPISODE_CONTAINS_ONLY_MUSIC_AND_ADS: 'episodeContainsOnlyMusicAndAds',
        EPISODE_CONTAINS_MONETIZATION_VIOLATIONS: 'episodeContainsMonetizationViolations',
        EPISODE_IS_Q_AND_A_ENABLED: 'episodeQandAEnabled',
        EPISODE_IS_PAYWALL_CONTENT: 'episodePaywallContent',
        EPISODE_REQUIRES_MORE_TALK_CONTENT: 'episodeRequiresMoreTalkContent',
    },
    CLIENT_TYPES: {
        IOS: CLIENT_TYPE_IOS,
        ANDROID: CLIENT_TYPE_ANDROID,
    },
    MT_TALK_RATIO_RULE_MINIMUM_VERSIONS_BY_CLIENT: mtTalkRatioRuleMinimumClientVersionByClientType,
};
