// @atlaskit/media-ui/dist/esm/EmbedCard/components/styled.js
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function _templateObject17() {
  var data = _taggedTemplateLiteral(["\n  position: absolute;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  overflow: hidden;\n  pointer-events: none;\n  /* z-index has to be 1 higher than the number of avatars in the avatar stack */\n  z-index: ", ";\n"]);

  _templateObject17 = function _templateObject17() {
    return data;
  };

  return data;
}

function _templateObject16() {
  var data = _taggedTemplateLiteral(["\n  margin-top: 8px;\n  text-align: right;\n\n  > * {\n    margin-top: 4px;\n  }\n\n  > * + * {\n    margin-left: 4px;\n  }\n"]);

  _templateObject16 = function _templateObject16() {
    return data;
  };

  return data;
}

function _templateObject15() {
  var data = _taggedTemplateLiteral(["\n  margin-top: 8px;\n"]);

  _templateObject15 = function _templateObject15() {
    return data;
  };

  return data;
}

function _templateObject14() {
  var data = _taggedTemplateLiteral(["\n  ", "\n  ", "\n  float: right;\n  margin: 4px 0 12px 12px;\n  background-color: ", ";\n  background-image: url(", ");\n  background-size: cover;\n"]);

  _templateObject14 = function _templateObject14() {
    return data;
  };

  return data;
}

function _templateObject13() {
  var data = _taggedTemplateLiteral(["\n  margin-top: 4px;\n"]);

  _templateObject13 = function _templateObject13() {
    return data;
  };

  return data;
}

function _templateObject12() {
  var data = _taggedTemplateLiteral(["\n  margin-top: 7px;\n  color: ", ";\n  font-size: 12px;\n  font-weight: normal;\n  line-height: ", ";\n  max-height: ", "px;\n  overflow: hidden;\n"]);

  _templateObject12 = function _templateObject12() {
    return data;
  };

  return data;
}

function _templateObject11() {
  var data = _taggedTemplateLiteral(["\n  margin-top: 4px;\n  color: ", ";\n  font-size: 12px;\n  font-weight: normal;\n  line-height: ", ";\n  ", ";\n"]);

  _templateObject11 = function _templateObject11() {
    return data;
  };

  return data;
}

function _templateObject10() {
  var data = _taggedTemplateLiteral(["\n  color: ", ";\n  font-size: 16px;\n  font-weight: 500;\n  line-height: ", ";\n  max-height: ", "px;\n  overflow: hidden;\n"]);

  _templateObject10 = function _templateObject10() {
    return data;
  };

  return data;
}

function _templateObject9() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  flex-direction: row;\n  box-sizing: border-box;\n  padding: 8px 12px 12px 12px;\n"]);

  _templateObject9 = function _templateObject9() {
    return data;
  };

  return data;
}

function _templateObject8() {
  var data = _taggedTemplateLiteral(["\n  ", "\n  ", "\n\n  /* hide the alt text when the image cannot be found */\n  overflow: hidden;\n"]);

  _templateObject8 = function _templateObject8() {
    return data;
  };

  return data;
}

function _templateObject7() {
  var data = _taggedTemplateLiteral(["\n  position: relative;\n\n  ", "\n  ", "\n  background-color: white;\n  position: absolute;\n  z-index: 1;\n  width: 100%;\n  top: ", ";\n  overflow: hidden;\n  height: calc(100% - ", ");\n  transition: box-shadow 0.3s;\n\n  > .calc-height > div > div {\n    left: unset !important;\n    width: unset !important;\n    height: unset !important;\n    position: initial !important;\n    padding-bottom: unset !important;\n  }\n  > .embed-preview > div {\n    margin: 0 auto;\n  }\n\n  ", ";\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = _taggedTemplateLiteral(["\n  ", " color: ", ";\n  font-size: 12px;\n  line-height: 16px;\n  ", ";\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n  ", "\n  ", "\n  ", " margin-right: 4px;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  height: ", "px;\n  position: absolute;\n  z-index: 1;\n  width: 100%;\n  display: flex;\n  align-items: center;\n  color: ", ";\n  opacity: 0;\n  transition: 300ms opacity cubic-bezier(0.15, 1, 0.3, 1);\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  ", ";\n  margin-top: 10px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  ", " &:hover {\n    text-decoration: none;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  ", "\n  ", "\n  ", "\n  ", "\n  ", "\n  display: inline-flex;\n  flex-direction: column;\n  box-sizing: border-box;\n  font-family: ", ";\n  width: 100%;\n  user-select: none;\n  line-height: initial;\n  transition: background 0.3s;\n  position: relative;\n  height: 100%;\n  ", "\n  ", "\n\n  &:after {\n    content: '';\n    background: transparent;\n    transition: background 0.3s, box-shadow 0.3s;\n    position: absolute;\n    width: calc(100% + ", ");\n    height: calc(100% + ", ");\n    left: -", ";\n    ", "\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import styled, { css } from 'styled-components';
import { size as csssize } from '../../mixins';
import { fontFamily } from '@atlaskit/theme/constants';
import * as colors from '@atlaskit/theme/colors';
import { borderRadius, size, ellipsis } from '../../mixins';
import { gs as gridSize } from '../../BlockCard/utils';
export var className = 'media-card-frame';
export var cardShadow = "";

function minWidth(_ref) {
  var minWidth = _ref.minWidth;

  if (minWidth) {
    return "min-width: ".concat(minWidth, "px;");
  } else {
    return '';
  }
}

function maxWidth(_ref2) {
  var maxWidth = _ref2.maxWidth;

  if (maxWidth) {
    return "max-width: ".concat(maxWidth, "px; margin: 0 auto;");
  } else {
    return 'margin: 0 auto;';
  }
}

function getinteractiveStyles(_ref3) {
  var isInteractive = _ref3.isInteractive,
      isFrameVisible = _ref3.isFrameVisible;
  var visibleFrameStyles = isFrameVisible ? visibleStyles : '';
  var interactiveFrameStyles = isInteractive ? "\n  &:hover {\n    ".concat(visibleStyles, "\n  }\n  &:active {\n    background-color: ").concat(colors.B50, ";\n  }\n") : '';
  return "".concat(visibleFrameStyles).concat(interactiveFrameStyles);
}

function selected(_ref4) {
  var isSelected = _ref4.isSelected;
  return isSelected ? "\n    ".concat(visibleStyles, "\n    &::after {\n      cursor: pointer;\n      box-shadow: 0 0 0 3px ").concat(colors.B100, ";\n      content: '';\n      outline: none;\n      position: absolute;\n      height: 100%;\n      width: 100%;\n      left: 0;\n      ").concat(borderRadius, "\n    }\n    ") : '';
}

var wrapperStyles = css(_templateObject(), borderRadius, minWidth, maxWidth, getinteractiveStyles, visible, fontFamily(), function (_ref5) {
  var inheritDimensions = _ref5.inheritDimensions;

  if (inheritDimensions) {
    return "\n      height: 100%;\n    ";
  } else {
    return "height: ".concat(gridSize(54));
  }
}, selected, gridSize(2), gridSize(1), gridSize(1), borderRadius);
var visibleStyles = "\n  background-color: ".concat(colors.N30, ";\n\n  &:after {\n    background: ").concat(colors.N30, " !important;\n  }\n  .embed-header {\n    opacity: 1;\n  }");

function visible(_ref6) {
  var isVisible = _ref6.isVisible;
  return isVisible ? visibleStyles : '';
}

export var LinkWrapper = styled.div(_templateObject2(), wrapperStyles);
export var Wrapper = styled.div(_templateObject3(), wrapperStyles);
export var embedHeaderHeight = 32;
export var Header = styled.div(_templateObject4(), embedHeaderHeight, colors.N300);
export var IconWrapper = styled.div(_templateObject5(), borderRadius, size(16), function (_ref7) {
  var isPlaceholder = _ref7.isPlaceholder;

  if (isPlaceholder) {
    return "\n      background-color: ".concat(colors.N30, ";\n    ");
  } else {
    return '';
  }
});
export var TextWrapper = styled.div(_templateObject6(), function (_ref8) {
  var isPlaceholder = _ref8.isPlaceholder;

  if (isPlaceholder) {
    return "\n        ".concat(borderRadius, "\n        width: 125px;\n        height: 12px;\n        background-color: ").concat(colors.N30, ";\n      ");
  } else {
    return '';
  }
}, colors.N300, ellipsis('none'));
// NB: `overflow` is kept as `hidden` since
// the internal contents of the `iframe` should
// manage scrolling behaviour.
export var Content = styled.div(_templateObject7(), borderRadius, cardShadow, gridSize(4), gridSize(4), function (_ref9) {
  var isInteractive = _ref9.isInteractive;

  if (isInteractive) {
    return "\n          .".concat(className, ":hover & {\n            box-shadow: 0 4px 8px -2px rgba(23, 43, 77, 0.32),\n              0 0 1px rgba(23, 43, 77, 0.25);\n          }\n        ");
  } else {
    return '';
  }
});
export var Image = styled.img(_templateObject8(), function (_ref10) {
  var size = _ref10.size;
  return csssize(size);
}, borderRadius);
export var maxAvatarCount = 6;
export var ContentWrapper = styled.div(_templateObject9());
export var Title = styled.div(_templateObject10(), colors.N900, 20 / 16, 20 * 4);
export var Byline = styled.div(_templateObject11(), colors.N300, 16 / 12, ellipsis('100%'));
export var Description = styled.div(_templateObject12(), colors.N800, 18 / 12, 18 * 3);
export var ResolvedViewIconWrapper = styled.div(_templateObject13());
export var Thumbnail = styled.div(_templateObject14(), borderRadius, size(48), colors.N30, function (_ref11) {
  var src = _ref11.src;
  return src;
});
export var UsersWrapper = styled.div(_templateObject15());
export var ActionsWrapper = styled.div(_templateObject16());
export var AlertWrapper = styled.div(_templateObject17(), maxAvatarCount + 1);
