// @atlaskit/atlassian-switcher/dist/esm/ui/components/loaders.js
import Loadable from 'react-loadable';
export var loadAtlassianSwitcher = function loadAtlassianSwitcher() {
  return import(
  /* webpackChunkName: "@ak-switcher-chunk-atlassian-switcher" */
  './atlassian-switcher');
};
export var loadJiraSwitcher = function loadJiraSwitcher() {
  return import(
  /* webpackChunkName: "@ak-switcher-chunk-jira-switcher" */
  './jira-switcher');
};
export var loadConfluenceSwitcher = function loadConfluenceSwitcher() {
  return import(
  /* webpackChunkName: "@ak-switcher-chunk-confluence-switcher" */
  './confluence-switcher');
};
export var loadGenericSwitcher = function loadGenericSwitcher() {
  return import(
  /* webpackChunkName: "@ak-switcher-chunk-generic-switcher" */
  './generic-switcher');
};
export var loadTrelloSwitcher = function loadTrelloSwitcher() {
  return import(
  /* webpackChunkName: "@ak-switcher-chunk-trello-switcher" */
  './trello-switcher');
};
export var AtlassianSwitcherLoader = Loadable({
  loader: loadAtlassianSwitcher,
  loading: function loading() {
    return null;
  }
});
export var JiraSwitcherLoader = Loadable({
  loader: loadJiraSwitcher,
  loading: function loading() {
    return null;
  }
});
export var ConfluenceSwitcherLoader = Loadable({
  loader: loadConfluenceSwitcher,
  loading: function loading() {
    return null;
  }
});
export var GenericSwitcherLoader = Loadable({
  loader: loadGenericSwitcher,
  loading: function loading() {
    return null;
  }
});
export var TrelloSwitcherLoader = Loadable({
  loader: loadTrelloSwitcher,
  loading: function loading() {
    return null;
  }
});
