// @udacity/ureact-components/dist/range-input/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = undefined;

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

var _dec, _class, _class2, _temp;

require('!style-loader!css-loader!react-input-range/lib/css/index.css');

var _reactInputRange = require('react-input-range');

var _reactInputRange2 = _interopRequireDefault(_reactInputRange);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactCssModules = require('react-css-modules');

var _reactCssModules2 = _interopRequireDefault(_reactCssModules);

var _noop2 = require('lodash/noop');

var _noop3 = _interopRequireDefault(_noop2);

var _rangeHelper = require('./_range-helper');

var _indexModule = require('./index.module.scss');

var _indexModule2 = _interopRequireDefault(_indexModule);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var RangeInput = (_dec = (0, _reactCssModules2.default)(_indexModule2.default), _dec(_class = (_temp = _class2 = function(_React$Component) {
    _inherits(RangeInput, _React$Component);

    function RangeInput() {
        _classCallCheck(this, RangeInput);

        return _possibleConstructorReturn(this, (RangeInput.__proto__ || Object.getPrototypeOf(RangeInput)).apply(this, arguments));
    }

    _createClass(RangeInput, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                min = _props.min,
                max = _props.max,
                step = _props.step,
                value = _props.value,
                _onChange = _props.onChange,
                disabled = _props.disabled,
                _onChangeComplete = _props.onChangeComplete,
                benchmarks = _props.benchmarks;


            var dots = void 0;
            if (benchmarks) {
                dots = (0, _rangeHelper.determineBenchmarks)(min, max, step);
            }

            return _react2.default.createElement(
                'div', {
                    styleName: 'range-container'
                },
                _react2.default.createElement(_reactInputRange2.default, {
                    maxValue: max,
                    minValue: min,
                    step: step,
                    value: value,
                    disabled: disabled,
                    onChange: function onChange(value) {
                        return _onChange(value);
                    },
                    onChangeComplete: function onChangeComplete(value) {
                        return _onChangeComplete(value);
                    }
                }),
                dots ? dots.map(function(d) {
                    return _react2.default.createElement('div', {
                        key: d.value,
                        title: d.value,
                        styleName: 'benchmark',
                        style: {
                            left: d.percent + '%'
                        },
                        onClick: function onClick() {
                            return _onChangeComplete(d.value);
                        }
                    });
                }) : null
            );
        }
    }]);

    return RangeInput;
}(_react2.default.Component), _class2.displayName = 'ureact-range-input', _class2.propTypes = {
    onChange: _propTypes2.default.func.isRequired,
    onChangeComplete: _propTypes2.default.func,
    value: _propTypes2.default.number.isRequired,
    max: _propTypes2.default.number.isRequired,
    min: _propTypes2.default.number,
    step: _propTypes2.default.number
}, _class2.defaultProps = {
    min: 0,
    max: 100,
    step: 1,
    value: 0,
    onChange: (0, _noop3.default)()
}, _temp)) || _class);
exports.default = RangeInput;
