// @splitsoftware/splitio/lib/services/request/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.noCacheExtraHeader = exports.default = void 0;

var _objectAssign = _interopRequireDefault(require("object-assign"));

var _options = _interopRequireDefault(require("./options"));

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
function RequestFactory(settings, relativeUrl, params, extraHeaders) {
  var token = settings.core.authorizationKey;
  var version = settings.version;
  var _settings$runtime = settings.runtime,
      ip = _settings$runtime.ip,
      hostname = _settings$runtime.hostname;
  var headers = {};
  var baseline = (0, _options.default)();
  headers['Accept'] = 'application/json';
  headers['Content-Type'] = 'application/json';
  headers['Authorization'] = "Bearer ".concat(token);
  headers['SplitSDKVersion'] = version;
  if (ip) headers['SplitSDKMachineIP'] = ip;
  if (hostname) headers['SplitSDKMachineName'] = hostname;
  if (extraHeaders) (0, _objectAssign.default)(headers, extraHeaders);
  return (0, _objectAssign.default)({
    headers: headers,
    url: settings.url(relativeUrl)
  }, baseline, params);
}

var _default = RequestFactory;
exports.default = _default;
var noCacheExtraHeader = {
  'Cache-Control': 'no-cache'
};
exports.noCacheExtraHeader = noCacheExtraHeader;
