// feather/lib/components/alert/index.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/es.weak-map.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.symbol.iterator.js");Object.defineProperty(exports, "__esModule", { value: true });exports.Alert = void 0;var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _animation = require("../../../lib/animation");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _types = require("../../../lib/types");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}

var getTheme = function getTheme(status) {
  switch (status) {
    case _types.AlertStatus.PRIMARY:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('purple-7'),
      (0, _cssVariables.default)('purple-2'));


    case _types.AlertStatus.SECONDARY:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('neutral-3'),
      (0, _cssVariables.default)('neutral-2'));


    case _types.AlertStatus.SUCCESS:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('green-5'),
      (0, _cssVariables.default)('green-2'));


    case _types.AlertStatus.DANGER:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('red-5'),
      (0, _cssVariables.default)('red-2'));


    case _types.AlertStatus.LIGHT:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('neutral-5'),
      (0, _cssVariables.default)('neutral-1'));


    case _types.AlertStatus.DARK:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('neutral-8'),
      (0, _cssVariables.default)('neutral-5'));


    default:
      return (0, _styledComponents.css)(["color:", ";background:", ";"],
      (0, _cssVariables.default)('purple-7'),
      (0, _cssVariables.default)('purple-2'));}


};

var Container = _styledComponents.default.div.withConfig({ displayName: "alert__Container", componentId: "zcxv7o-0" })(["padding:", "px;", ";opacity:", ";transform:", ";text-align:", ";border-radius:2px;transition:height 0.2s ", ",opacity 0.2s ", ",transform 0.2s ", ";will-change:height,opacity,transform;", " ", ""],
function (props) {return props.show ? 20 : 0;},
function (props) {return props.show ? '' : 'height: 0 !important;';},
function (props) {return props.show ? 1 : 0;},
function (props) {
  if (props.slideFrom === _types.AlertSlideFrom.TOP || props.slideFrom === _types.AlertSlideFrom.BOTTOM) {
    if (props.show) {
      return 'translateY(0)';
    }
    return "translateY(".concat(props.slideFrom === _types.AlertSlideFrom.TOP ? -50 : 50, "px)");
  }
  if (props.show) {
    return 'translateX(0)';
  }
  return "translateX(".concat(props.slideFrom === _types.AlertSlideFrom.LEFT ? -50 : 50, "px)");
},
function (props) {return props.align ? props.align : _types.AlertMessageAlign.LEFT;},

_animation.transitionDefault, _animation.transitionDefault, _animation.transitionDefault,


function (props) {return getTheme(props.status);},
function (props) {return props.styles;});


var Message = _styledComponents.default.p.withConfig({ displayName: "alert__Message", componentId: "zcxv7o-1" })(["margin:0;padding:0;font-size:14px;font-weight:500;line-height:1.4;", ""],






function (props) {return props.styles;});


var Alert = function Alert(_ref)







{var _ref$show = _ref.show,show = _ref$show === void 0 ? false : _ref$show,_ref$status = _ref.status,status = _ref$status === void 0 ? _types.AlertStatus.PRIMARY : _ref$status,_ref$align = _ref.align,align = _ref$align === void 0 ? _types.AlertMessageAlign.LEFT : _ref$align,_ref$slideFrom = _ref.slideFrom,slideFrom = _ref$slideFrom === void 0 ? _types.AlertSlideFrom.TOP : _ref$slideFrom,message = _ref.message,innerHTML = _ref.innerHTML,_ref$styles = _ref.styles,styles = _ref$styles === void 0 ? {} : _ref$styles;
  var containerRef = (0, _react.useRef)(null);
  var messageRef = (0, _react.useRef)(null);

  (0, _react.useEffect)(function () {
    var currentHeight = messageRef.current ? messageRef.current.offsetHeight + 40 : 0;

    if (containerRef.current) {
      containerRef.current.style.height = "".concat(currentHeight, "px");
    }

    // make height auto after height transition is done
    if (show) {
      setTimeout(function () {
        if (containerRef.current) {
          containerRef.current.style.height = 'auto';
        }
      }, 200);
    }
  }, [show, message]);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(Container, {
      ref: containerRef,
      show: show,
      status: status,
      align: align,
      slideFrom: slideFrom,
      styles: styles.CONTAINER, children: /*#__PURE__*/

      (0, _jsxRuntime.jsx)(Message, _objectSpread(_objectSpread({
        ref: messageRef,
        styles: styles.MESSAGE },
      innerHTML && message == null // message takes precedence over innerHTML
      ? { dangerouslySetInnerHTML: { __html: innerHTML } } :
      null), {}, { children:

        message })) }));



};exports.Alert = Alert;
