// @udacity/ureact-app-layout/dist/index.js
! function(e, t) {
    if ("object" == typeof exports && "object" == typeof module) module.exports = t(require("react"));
    else if ("function" == typeof define && define.amd) define(["react"], t);
    else {
        var n = "object" == typeof exports ? t(require("react")) : t(e.react);
        for (var a in n)("object" == typeof exports ? exports : e)[a] = n[a]
    }
}(window, (function(e) {
    return function(e) {
        var t = {};

        function n(a) {
            if (t[a]) return t[a].exports;
            var r = t[a] = {
                i: a,
                l: !1,
                exports: {}
            };
            return e[a].call(r.exports, r, r.exports, n), r.l = !0, r.exports
        }
        return n.m = e, n.c = t, n.d = function(e, t, a) {
            n.o(e, t) || Object.defineProperty(e, t, {
                enumerable: !0,
                get: a
            })
        }, n.r = function(e) {
            "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(e, Symbol.toStringTag, {
                value: "Module"
            }), Object.defineProperty(e, "__esModule", {
                value: !0
            })
        }, n.t = function(e, t) {
            if (1 & t && (e = n(e)), 8 & t) return e;
            if (4 & t && "object" == typeof e && e && e.__esModule) return e;
            var a = Object.create(null);
            if (n.r(a), Object.defineProperty(a, "default", {
                    enumerable: !0,
                    value: e
                }), 2 & t && "string" != typeof e)
                for (var r in e) n.d(a, r, function(t) {
                    return e[t]
                }.bind(null, r));
            return a
        }, n.n = function(e) {
            var t = e && e.__esModule ? function() {
                return e.default
            } : function() {
                return e
            };
            return n.d(t, "a", t), t
        }, n.o = function(e, t) {
            return Object.prototype.hasOwnProperty.call(e, t)
        }, n.p = "", n(n.s = 25)
    }([function(t, n) {
        t.exports = e
    }, function(e, t, n) {
        "use strict";
        var a, r = function() {
                return void 0 === a && (a = Boolean(window && document && document.all && !window.atob)), a
            },
            i = function() {
                var e = {};
                return function(t) {
                    if (void 0 === e[t]) {
                        var n = document.querySelector(t);
                        if (window.HTMLIFrameElement && n instanceof window.HTMLIFrameElement) try {
                            n = n.contentDocument.head
                        } catch (e) {
                            n = null
                        }
                        e[t] = n
                    }
                    return e[t]
                }
            }(),
            o = [];

        function c(e) {
            for (var t = -1, n = 0; n < o.length; n++)
                if (o[n].identifier === e) {
                    t = n;
                    break
                }
            return t
        }

        function l(e, t) {
            for (var n = {}, a = [], r = 0; r < e.length; r++) {
                var i = e[r],
                    l = t.base ? i[0] + t.base : i[0],
                    u = n[l] || 0,
                    s = "".concat(l, " ").concat(u);
                n[l] = u + 1;
                var p = c(s),
                    d = {
                        css: i[1],
                        media: i[2],
                        sourceMap: i[3]
                    }; - 1 !== p ? (o[p].references++, o[p].updater(d)) : o.push({
                    identifier: s,
                    updater: _(d, t),
                    references: 1
                }), a.push(s)
            }
            return a
        }

        function u(e) {
            var t = document.createElement("style"),
                a = e.attributes || {};
            if (void 0 === a.nonce) {
                var r = n.nc;
                r && (a.nonce = r)
            }
            if (Object.keys(a).forEach((function(e) {
                    t.setAttribute(e, a[e])
                })), "function" == typeof e.insert) e.insert(t);
            else {
                var o = i(e.insert || "head");
                if (!o) throw new Error("Couldn't find a style target. This probably means that the value for the 'insert' parameter is invalid.");
                o.appendChild(t)
            }
            return t
        }
        var s, p = (s = [], function(e, t) {
            return s[e] = t, s.filter(Boolean).join("\n")
        });

        function d(e, t, n, a) {
            var r = n ? "" : a.media ? "@media ".concat(a.media, " {").concat(a.css, "}") : a.css;
            if (e.styleSheet) e.styleSheet.cssText = p(t, r);
            else {
                var i = document.createTextNode(r),
                    o = e.childNodes;
                o[t] && e.removeChild(o[t]), o.length ? e.insertBefore(i, o[t]) : e.appendChild(i)
            }
        }

        function m(e, t, n) {
            var a = n.css,
                r = n.media,
                i = n.sourceMap;
            if (r ? e.setAttribute("media", r) : e.removeAttribute("media"), i && btoa && (a += "\n/*# sourceMappingURL=data:application/json;base64,".concat(btoa(unescape(encodeURIComponent(JSON.stringify(i)))), " */")), e.styleSheet) e.styleSheet.cssText = a;
            else {
                for (; e.firstChild;) e.removeChild(e.firstChild);
                e.appendChild(document.createTextNode(a))
            }
        }
        var f = null,
            v = 0;

        function _(e, t) {
            var n, a, r;
            if (t.singleton) {
                var i = v++;
                n = f || (f = u(t)), a = d.bind(null, n, i, !1), r = d.bind(null, n, i, !0)
            } else n = u(t), a = m.bind(null, n, t), r = function() {
                ! function(e) {
                    if (null === e.parentNode) return !1;
                    e.parentNode.removeChild(e)
                }(n)
            };
            return a(e),
                function(t) {
                    if (t) {
                        if (t.css === e.css && t.media === e.media && t.sourceMap === e.sourceMap) return;
                        a(e = t)
                    } else r()
                }
        }
        e.exports = function(e, t) {
            (t = t || {}).singleton || "boolean" == typeof t.singleton || (t.singleton = r());
            var n = l(e = e || [], t);
            return function(e) {
                if (e = e || [], "[object Array]" === Object.prototype.toString.call(e)) {
                    for (var a = 0; a < n.length; a++) {
                        var r = c(n[a]);
                        o[r].references--
                    }
                    for (var i = l(e, t), u = 0; u < n.length; u++) {
                        var s = c(n[u]);
                        0 === o[s].references && (o[s].updater(), o.splice(s, 1))
                    }
                    n = i
                }
            }
        }
    }, function(e, t, n) {
        "use strict";
        e.exports = function(e) {
            var t = [];
            return t.toString = function() {
                return this.map((function(t) {
                    var n = function(e, t) {
                        var n = e[1] || "",
                            a = e[3];
                        if (!a) return n;
                        if (t && "function" == typeof btoa) {
                            var r = (o = a, c = btoa(unescape(encodeURIComponent(JSON.stringify(o)))), l = "sourceMappingURL=data:application/json;charset=utf-8;base64,".concat(c), "/*# ".concat(l, " */")),
                                i = a.sources.map((function(e) {
                                    return "/*# sourceURL=".concat(a.sourceRoot || "").concat(e, " */")
                                }));
                            return [n].concat(i).concat([r]).join("\n")
                        }
                        var o, c, l;
                        return [n].join("\n")
                    }(t, e);
                    return t[2] ? "@media ".concat(t[2], " {").concat(n, "}") : n
                })).join("")
            }, t.i = function(e, n, a) {
                "string" == typeof e && (e = [
                    [null, e, ""]
                ]);
                var r = {};
                if (a)
                    for (var i = 0; i < this.length; i++) {
                        var o = this[i][0];
                        null != o && (r[o] = !0)
                    }
                for (var c = 0; c < e.length; c++) {
                    var l = [].concat(e[c]);
                    a && r[l[0]] || (n && (l[2] ? l[2] = "".concat(n, " and ").concat(l[2]) : l[2] = n), t.push(l))
                }
            }, t
        }
    }, function(e, t) {
        e.exports = function(e) {
            var t = typeof e;
            return null != e && ("object" == t || "function" == t)
        }
    }, function(e, t, n) {
        var a = n(1),
            r = n(24);
        "string" == typeof(r = r.__esModule ? r.default : r) && (r = [
            [e.i, r, ""]
        ]);
        var i = {
            insert: "head",
            singleton: !1
        };
        a(r, i);
        e.exports = r.locals || {}
    }, function(e, t, n) {
        var a = n(8),
            r = n(3);
        e.exports = function(e, t, n) {
            var i = !0,
                o = !0;
            if ("function" != typeof e) throw new TypeError("Expected a function");
            return r(n) && (i = "leading" in n ? !!n.leading : i, o = "trailing" in n ? !!n.trailing : o), a(e, t, {
                leading: i,
                maxWait: t,
                trailing: o
            })
        }
    }, function(e, t, n) {
        var a = n(10),
            r = "object" == typeof self && self && self.Object === Object && self,
            i = a || r || Function("return this")();
        e.exports = i
    }, function(e, t, n) {
        var a = n(6).Symbol;
        e.exports = a
    }, function(e, t, n) {
        var a = n(3),
            r = n(9),
            i = n(12),
            o = Math.max,
            c = Math.min;
        e.exports = function(e, t, n) {
            var l, u, s, p, d, m, f = 0,
                v = !1,
                _ = !1,
                h = !0;
            if ("function" != typeof e) throw new TypeError("Expected a function");

            function y(t) {
                var n = l,
                    a = u;
                return l = u = void 0, f = t, p = e.apply(a, n)
            }

            function g(e) {
                return f = e, d = setTimeout(x, t), v ? y(e) : p
            }

            function b(e) {
                var n = e - m;
                return void 0 === m || n >= t || n < 0 || _ && e - f >= s
            }

            function x() {
                var e = r();
                if (b(e)) return w(e);
                d = setTimeout(x, function(e) {
                    var n = t - (e - m);
                    return _ ? c(n, s - (e - f)) : n
                }(e))
            }

            function w(e) {
                return d = void 0, h && l ? y(e) : (l = u = void 0, p)
            }

            function E() {
                var e = r(),
                    n = b(e);
                if (l = arguments, u = this, m = e, n) {
                    if (void 0 === d) return g(m);
                    if (_) return clearTimeout(d), d = setTimeout(x, t), y(m)
                }
                return void 0 === d && (d = setTimeout(x, t)), p
            }
            return t = i(t) || 0, a(n) && (v = !!n.leading, s = (_ = "maxWait" in n) ? o(i(n.maxWait) || 0, t) : s, h = "trailing" in n ? !!n.trailing : h), E.cancel = function() {
                void 0 !== d && clearTimeout(d), f = 0, l = m = u = d = void 0
            }, E.flush = function() {
                return void 0 === d ? p : w(r())
            }, E
        }
    }, function(e, t, n) {
        var a = n(6);
        e.exports = function() {
            return a.Date.now()
        }
    }, function(e, t, n) {
        (function(t) {
            var n = "object" == typeof t && t && t.Object === Object && t;
            e.exports = n
        }).call(this, n(11))
    }, function(e, t) {
        var n;
        n = function() {
            return this
        }();
        try {
            n = n || new Function("return this")()
        } catch (e) {
            "object" == typeof window && (n = window)
        }
        e.exports = n
    }, function(e, t, n) {
        var a = n(3),
            r = n(13),
            i = /^\s+|\s+$/g,
            o = /^[-+]0x[0-9a-f]+$/i,
            c = /^0b[01]+$/i,
            l = /^0o[0-7]+$/i,
            u = parseInt;
        e.exports = function(e) {
            if ("number" == typeof e) return e;
            if (r(e)) return NaN;
            if (a(e)) {
                var t = "function" == typeof e.valueOf ? e.valueOf() : e;
                e = a(t) ? t + "" : t
            }
            if ("string" != typeof e) return 0 === e ? e : +e;
            e = e.replace(i, "");
            var n = c.test(e);
            return n || l.test(e) ? u(e.slice(2), n ? 2 : 8) : o.test(e) ? NaN : +e
        }
    }, function(e, t, n) {
        var a = n(14),
            r = n(17);
        e.exports = function(e) {
            return "symbol" == typeof e || r(e) && "[object Symbol]" == a(e)
        }
    }, function(e, t, n) {
        var a = n(7),
            r = n(15),
            i = n(16),
            o = a ? a.toStringTag : void 0;
        e.exports = function(e) {
            return null == e ? void 0 === e ? "[object Undefined]" : "[object Null]" : o && o in Object(e) ? r(e) : i(e)
        }
    }, function(e, t, n) {
        var a = n(7),
            r = Object.prototype,
            i = r.hasOwnProperty,
            o = r.toString,
            c = a ? a.toStringTag : void 0;
        e.exports = function(e) {
            var t = i.call(e, c),
                n = e[c];
            try {
                e[c] = void 0;
                var a = !0
            } catch (e) {}
            var r = o.call(e);
            return a && (t ? e[c] = n : delete e[c]), r
        }
    }, function(e, t) {
        var n = Object.prototype.toString;
        e.exports = function(e) {
            return n.call(e)
        }
    }, function(e, t) {
        e.exports = function(e) {
            return null != e && "object" == typeof e
        }
    }, function(e, t, n) {
        var a = n(1),
            r = n(19);
        "string" == typeof(r = r.__esModule ? r.default : r) && (r = [
            [e.i, r, ""]
        ]);
        var i = {
            insert: "head",
            singleton: !1
        };
        a(r, i);
        e.exports = r.locals || {}
    }, function(e, t, n) {
        (t = n(2)(!1)).push([e.i, ".ureact-app__header .page-title{color:#2e3d49;font-size:1.125rem;font-weight:normal;padding:.875rem}", ""]), e.exports = t
    }, function(e, t, n) {
        var a = n(1),
            r = n(21);
        "string" == typeof(r = r.__esModule ? r.default : r) && (r = [
            [e.i, r, ""]
        ]);
        var i = {
            insert: "head",
            singleton: !1
        };
        a(r, i);
        e.exports = r.locals || {}
    }, function(e, t, n) {
        (t = n(2)(!1)).push([e.i, ".ureact-app__body{display:flex;flex-direction:column;flex:1 0 auto;transition:.3s height ease-in-out}", ""]), e.exports = t
    }, function(e, t, n) {
        var a = n(1),
            r = n(23);
        "string" == typeof(r = r.__esModule ? r.default : r) && (r = [
            [e.i, r, ""]
        ]);
        var i = {
            insert: "head",
            singleton: !1
        };
        a(r, i);
        e.exports = r.locals || {}
    }, function(e, t, n) {
        (t = n(2)(!1)).push([e.i, ".ureact-app__container{bottom:0;display:flex;font-size:1rem;left:0;position:absolute;right:0;top:0}.ureact-app__nav{background:#1f2932;bottom:0;height:100%;left:0;position:absolute;top:0;transform:translateX(0);transition:.2s all ease-in-out;z-index:10}.ureact-app__nav.large-primary{width:16rem}.ureact-app__nav.large-primary .primary-nav{width:100%}.ureact-app__nav.primary-and-secondary{width:20rem}.ureact-app__nav.secondary-only{width:20rem}.ureact-app__nav.small-primary{width:4rem}.ureact-app__nav.small-primary .primary-nav,.ureact-app__nav.primary-and-secondary .primary-nav{width:4rem}.ureact-app__nav .nav__a11y-link{display:block;height:0;left:-999em;overflow:hidden;position:absolute;text-indent:-999em;width:0}.ureact-app__nav .nav__contents{display:flex;height:100%}.ureact-app__nav .nav__contents .primary-nav,.ureact-app__nav .nav__contents .secondary-nav{height:100%}.ureact-app__nav .nav__contents .primary-nav{flex-grow:0;flex-shrink:0;position:relative;overflow:hidden}.ureact-app__nav .nav__contents .secondary-nav{background:#2e3d49;overflow-y:auto;width:100%}.ureact-app__nav.hidden{transform:translateX(-100%)}.ureact-app__nav.hidden .nav__contents{animation-name:hideFromKeyboard;animation-delay:.2s;animation-fill-mode:forwards;animation-duration:.1s;flex-basis:0;height:100vh;overflow:hidden}.ureact-app__main{box-sizing:border-box;height:100%;min-width:300px;position:relative;transition:.2s padding-top ease-in-out;width:100%}.ureact-app__main .title-area{background-color:#fff;box-shadow:0 0 40px rgba(0,0,0,.2);left:0;position:absolute;right:0;top:0;transition:.2s margin-left ease-in-out,.2s margin-right ease-in-out;z-index:9}.ureact-app__main .content-area{display:flex;flex-direction:column;height:100%;overflow-y:auto;transition:.2s padding-top ease-in-out,.2s margin-left ease-in-out,.2s margin-right ease-in-out}.ureact-app__main .hidden-header{visibility:hidden}.secondary-only+.ureact-app__main .title-area,.secondary-only+.ureact-app__main .content-area{margin-left:20rem}.primary-and-secondary+.ureact-app__main .title-area,.primary-and-secondary+.ureact-app__main .content-area{margin-left:20rem}.small-primary+.ureact-app__main .title-area,.small-primary+.ureact-app__main .content-area{margin-left:4rem}.large-primary+.ureact-app__main .title-area,.large-primary+.ureact-app__main .content-area{margin-left:16rem}.secondary-only+.ureact-app__main .title-area,.secondary-only+.ureact-app__main .content-area{margin-left:20rem}.ureact-app__main .title-area,.ureact-app__main .content-area,.hidden+.ureact-app__main .title-area,.hidden+.ureact-app__main .content-area,.none+.ureact-app__main .title-area,.none+.ureact-app__main .content-area{margin-left:0}body[dir=rtl] .ureact-app__nav{left:auto;right:0}body[dir=rtl] .ureact-app__nav.hidden{transform:translateX(100%)}body[dir=rtl] .secondary-only+.ureact-app__main .title-area,body[dir=rtl] .secondary-only+.ureact-app__main .content-area{margin-left:0;margin-right:20rem}body[dir=rtl] .primary-and-secondary+.ureact-app__main .title-area,body[dir=rtl] .primary-and-secondary+.ureact-app__main .content-area{margin-left:0;margin-right:20rem}body[dir=rtl] .small-primary+.ureact-app__main .title-area,body[dir=rtl] .small-primary+.ureact-app__main .content-area{margin-left:0;margin-right:4rem}body[dir=rtl] .large-primary+.ureact-app__main .title-area,body[dir=rtl] .large-primary+.ureact-app__main .content-area{margin-left:0;margin-right:16rem}body[dir=rtl] .ureact-app__main .title-area,body[dir=rtl] .ureact-app__main .content-area,body[dir=rtl] .hidden+.ureact-app__main .title-area,body[dir=rtl] .hidden+.ureact-app__main .content-area,body[dir=rtl] .none+.ureact-app__main .title-area,body[dir=rtl] .none+.ureact-app__main .content-area{margin-left:0;margin-right:0}@media(max-width: 46.24em){.ureact-app__main .title-area,.ureact-app__main .content-area{width:100%}}@keyframes hideFromKeyboard{0%{visibility:visible}100%{visibility:hidden}}", ""]), e.exports = t
    }, function(e, t, n) {
        (t = n(2)(!1)).push([e.i, "._nav,.ureact-app__nav .nav-large,.ureact-app__nav .nav-small{display:flex;flex-direction:column;font-size:2rem;height:100%;overflow-x:hidden;overflow-y:auto;padding:0}._nav .logo,.ureact-app__nav .nav-large .logo,.ureact-app__nav .nav-small .logo{margin-left:1rem;margin-right:1rem}.ureact-app__nav .nav-small .logo{width:2rem}.ureact-app__nav .nav-small .nav-item-title{opacity:0;overflow:hidden;position:absolute;text-indent:-999em}.ureact-app__nav .nav-large{width:14rem}.ureact-app__nav .nav-large .nav-item-icon-container{margin-right:1rem}.ureact-app__nav .logo{flex:0 0 auto;margin:1rem 0}.ureact-app__nav .nav-groups{display:flex;flex-direction:column;flex:1 1 100%;justify-content:space-between}.ureact-app__nav .nav-group{list-style:none;margin:0;padding:0}.ureact-app__nav .nav-item{line-height:1;margin-bottom:.25rem}.ureact-app__nav .nav-item a{align-items:center;color:#dbe2e8;display:flex;padding:.5rem .75rem;text-decoration:none}.ureact-app__nav .nav-item-icon-container{display:inline-block;position:relative;text-align:center}.ureact-app__nav .nav-item-icon{position:relative;width:2rem}.ureact-app__nav .nav-item-banner{background:#9f4cdf;border-radius:2px;font-size:.75rem;font-weight:600;left:50%;padding:1.5px 4px;position:absolute;text-transform:uppercase;bottom:-0.5rem;transform:translateX(-50%)}.ureact-app__nav .nav-item-title{font-size:1rem}.ureact-app__nav body[dir=rtl] ._nav .logo,.ureact-app__nav body[dir=rtl] .nav-small .logo,.ureact-app__nav body[dir=rtl] .nav-large .logo{margin-left:1rem;margin-right:1.25rem}.ureact-app__nav body[dir=rtl] .nav-large .nav-item-icon{margin-left:1rem;margin-right:0}.ureact-app__nav body[dir=rtl] .nav-small .nav-item-title{text-indent:999em}.ureact-app__nav body[dir=rtl] .nav-item-banner{left:auto;right:50%;transform:translateX(50%)}", ""]), e.exports = t
    }, function(e, t, n) {
        "use strict";
        n.r(t), n.d(t, "Provider", (function() {
            return d
        })), n.d(t, "Layout", (function() {
            return g
        })), n.d(t, "Nav", (function() {
            return w
        })), n.d(t, "NavGroup", (function() {
            return E
        })), n.d(t, "NavItem", (function() {
            return j
        })), n.d(t, "LayoutConsumer", (function() {
            return s
        }));
        var a = n(0),
            r = n.n(a),
            i = n(5),
            o = n.n(i),
            c = function() {
                return window.innerWidth <= 900
            },
            l = r.a.createContext({
                isContentScrolled: !1,
                isMobile: !1,
                isShortBrowser: !1,
                isSidebarOpen: !1,
                toggleSidebar: function() {},
                _setScrollContainer: function() {}
            }),
            u = l.Provider,
            s = l.Consumer,
            p = function() {
                return (p = Object.assign || function(e) {
                    for (var t, n = 1, a = arguments.length; n < a; n++)
                        for (var r in t = arguments[n]) Object.prototype.hasOwnProperty.call(t, r) && (e[r] = t[r]);
                    return e
                }).apply(this, arguments)
            },
            d = function(e) {
                var t = e.children,
                    n = document,
                    i = Object(a.useState)({
                        isContentScrolled: !1,
                        isMobile: !1,
                        isShortBrowser: !1,
                        isSidebarOpen: !c()
                    }),
                    l = i[0],
                    s = i[1],
                    d = function() {
                        s((function(e) {
                            var t, a = c();
                            return {
                                isMobile: c(),
                                isShortBrowser: window.innerHeight <= 600,
                                isSidebarOpen: e.isMobile === a ? e.isSidebarOpen : !a,
                                isContentScrolled: !!n && (t = n, t === document ? (document.documentElement || document.body).scrollTop > 150 : !!t && t.scrollTop > 150)
                            }
                        }))
                    },
                    m = o()(d, 50),
                    f = o()(d, 50);
                return Object(a.useEffect)((function() {
                    return d(), window.addEventListener("resize", m),
                        function() {
                            window.removeEventListener("resize", m), n = null
                        }
                }), []), r.a.createElement(u, {
                    value: p(p({}, l), {
                        toggleSidebar: function(e) {
                            s((function(t) {
                                return p(p({}, t), {
                                    isSidebarOpen: e || !t.isSidebarOpen
                                })
                            }))
                        },
                        _setScrollContainer: function(e) {
                            n && n.removeEventListener("scroll", f), (n = e) && (d(), n.addEventListener("scroll", f))
                        }
                    })
                }, t)
            },
            m = (n(18), function() {
                return (m = Object.assign || function(e) {
                    for (var t, n = 1, a = arguments.length; n < a; n++)
                        for (var r in t = arguments[n]) Object.prototype.hasOwnProperty.call(t, r) && (e[r] = t[r]);
                    return e
                }).apply(this, arguments)
            }),
            f = function(e) {
                var t = e.children,
                    n = e.title;
                return r.a.createElement("div", {
                    className: "ureact-app__header"
                }, n && r.a.createElement("h2", {
                    className: "page-title"
                }, n), t && r.a.createElement("div", null, t))
            },
            v = function(e) {
                var t = e.children,
                    n = e.title;
                return r.a.createElement(s, null, (function(e) {
                    return r.a.createElement(u, {
                        value: m(m({}, e), {
                            isContentScrolled: !1
                        })
                    }, r.a.createElement(f, {
                        title: n
                    }, t))
                }))
            },
            _ = (n(20), function(e) {
                var t = e.children;
                return r.a.createElement("div", {
                    id: "content",
                    className: "ureact-app__body"
                }, t)
            }),
            h = (n(22), function() {
                return (h = Object.assign || function(e) {
                    for (var t, n = 1, a = arguments.length; n < a; n++)
                        for (var r in t = arguments[n]) Object.prototype.hasOwnProperty.call(t, r) && (e[r] = t[r]);
                    return e
                }).apply(this, arguments)
            }),
            y = function(e) {
                var t = e.backgroundColor,
                    n = void 0 === t ? "white" : t,
                    i = e.children,
                    o = e.header,
                    c = e.isSidebarOpen,
                    l = e.nav,
                    u = e.sidebar,
                    s = e._setScrollContainer,
                    p = Object(a.useRef)(null);
                Object(a.useEffect)((function() {
                    s(p.current)
                }), []);
                var d = Object(a.useMemo)((function() {
                    switch (!0) {
                        case !u && !l:
                            return "none";
                        case !!u && !!l:
                            return "primary-and-secondary";
                        case !!u && !l:
                            return "secondary-only";
                        case !u && !!l && "large" === l.props.variant:
                            return "large-primary";
                        case !u && !!l:
                            return "small-primary"
                    }
                    return ""
                }), [u, l]);
                return r.a.createElement("div", {
                    className: "ureact-app__container"
                }, l || u ? r.a.createElement("nav", {
                    className: "ureact-app__nav " + d + " " + (c ? "" : "hidden")
                }, r.a.createElement("a", {
                    className: "nav__a11y-link",
                    href: "#content"
                }, "Jump to content"), r.a.createElement("a", {
                    className: "nav__a11y-link",
                    href: "mailto:a11y@udacity.com?subject=Accessibility%20Support",
                    rel: "noreferrer",
                    target: "_blank"
                }, "Email for accessibility support"), r.a.createElement("div", {
                    className: "nav__contents"
                }, !!l && r.a.createElement("div", {
                    className: "primary-nav"
                }, l), u && r.a.createElement("div", {
                    id: "main-layout-sidebar",
                    className: "secondary-nav"
                }, u))) : null, r.a.createElement("main", {
                    className: "ureact-app__main"
                }, !!o && r.a.createElement("div", {
                    className: "title-area"
                }, r.a.createElement(f, null, o)), r.a.createElement("div", {
                    className: "content-area",
                    style: {
                        backgroundColor: n
                    },
                    ref: p
                }, !!o && r.a.createElement("div", {
                    "aria-hidden": !0,
                    className: "hidden-header"
                }, r.a.createElement(v, null, r.a.cloneElement(o))), r.a.createElement(_, null, i))))
            },
            g = function(e) {
                return r.a.createElement(s, null, (function(t) {
                    var n = t.isSidebarOpen,
                        a = t._setScrollContainer;
                    return r.a.createElement(y, h({}, e, {
                        isSidebarOpen: n,
                        _setScrollContainer: a
                    }))
                }))
            },
            b = (n(4), function() {
                return (b = Object.assign || function(e) {
                    for (var t, n = 1, a = arguments.length; n < a; n++)
                        for (var r in t = arguments[n]) Object.prototype.hasOwnProperty.call(t, r) && (e[r] = t[r]);
                    return e
                }).apply(this, arguments)
            }),
            x = function(e) {
                return r.a.createElement("svg", b({
                    width: 177,
                    height: 31
                }, e), r.a.createElement("title", null, "Udacity Logo"), r.a.createElement("g", {
                    fill: "none",
                    fillRule: "evenodd"
                }, r.a.createElement("path", {
                    d: "M58 17.6c0 2.6-2.1 4.5-5 4.5s-5-1.9-5-4.5V8h-2v9.7c0 3.4 2.9 6.3 7 6.3s7-3 7-6.3V8h-2v9.6zM74.4 8H69v16h5.4c4.5 0 7.6-3.1 7.6-8 0-4.8-3.1-8-7.6-8zm-.2 14H71V10h3.2c3.4 0 5.8 2.3 5.8 6 0 4-2.5 5.9-5.8 6zm43.8.3c-3.6 0-6-2.7-6-6.3 0-3.6 2.5-6.2 6-6.2 2.8 0 4.5 1.6 4.5 1.6l.8-1.4s-1.8-1.9-5.5-1.9c-4.8 0-7.8 3.6-7.8 7.9 0 4.4 3.1 8.1 8 8.1 3.8 0 5.7-2.6 5.7-2.6l-1.1-1.2c-.1-.1-1.6 2-4.6 2zM133 8h2v16h-2zm11 2h5v14h2V10h5V8h-12v2zm30.8-2l-4.8 7.3-4.9-7.3H163l6 8.9V24h2v-7.1l5.9-8.9h-2.1zM94.9 8l-6.2 16h2l1.7-4.6 6.5-1.4 2.2 6h2L97 8h-2.1zM93 17.5l3-7.4h.1l2.5 6.3-5.6 1.1z",
                    fill: "#FFF"
                }), r.a.createElement("path", {
                    d: "M29 .1l-9 4.7v14.9c0 .6 0 1.3-.2 1.9-.1.6-.3 1.2-.5 1.7-.1.1-.1.3-.2.4-.3 0-2.5-.7-3.5-1.4-.4-.3-.8-.6-1.2-1-.4-.4-.7-.8-1-1.2-.3-.4-.6-.9-.8-1.4-.2-.5-.4-1-.5-1.6-.1-.5-.1-1.1-.1-1.6V1.2L10 0 0 5.9v13.7C0 25.9 5.1 31 11.3 31c2.1 0 4-.5 5.7-1.5l8.4-4.8c3.3-1.7 5.6-5.2 5.6-9.1V1.3L29 .1zM17.9 26.2c-.9.9-1.6 1.4-2.6 1.9-.6.2-1.3.5-1.9.7-.6.1-1.2.2-1.8.2-.6 0-1.6-.1-2.2-.2-.6-.1-1.2-.3-1.7-.5s-1.1-.5-1.6-.9c-1.3-.9-1.9-1.7-2.5-2.5-.3-.5-.6-1-.9-1.6-.2-.6-.4-1.1-.5-1.7-.2-.7-.2-1.4-.2-2V7.1l8-4.7v13.1c0 5 3.7 9.1 8.4 10.1-.2.2-.3.4-.5.6zm11-9c-.1.5-.3 1-.5 1.5s-.6 1.2-1 1.7c-.6.7-1.2 1.2-2.7 2.1-1.3.7-2.4.9-3.1 1.1.5-1.2.5-2.5.5-3.9L22 6l7-3.7v13.3c0 .5 0 1.1-.1 1.6z",
                    fill: "#02B3E4"
                })))
            },
            w = function(e) {
                var t = e.children,
                    n = e.isEnterpriseOnly,
                    a = e.variant,
                    i = void 0 === a ? "small" : a;
                return r.a.createElement("div", {
                    className: "ureact-app__nav nav-" + i
                }, n ? r.a.createElement(x, {
                    className: "logo"
                }) : r.a.createElement("a", {
                    href: "https://udacity.com",
                    target: "_blank",
                    rel: "noreferrer"
                }, r.a.createElement(x, {
                    className: "logo"
                })), r.a.createElement("div", {
                    className: "nav-groups"
                }, t))
            },
            E = function(e) {
                var t = e.children,
                    n = e.className,
                    a = "nav-group";
                return n && (a += " " + n), r.a.createElement("ul", {
                    className: a
                }, t)
            },
            j = function(e) {
                var t = e.banner,
                    n = e.children,
                    a = e.className,
                    i = e.href,
                    o = void 0 === i ? "#" : i,
                    c = e.onClick,
                    l = e.title,
                    u = e.target,
                    s = "nav-item";
                a && (s += " " + a);
                return r.a.createElement("li", {
                    className: s
                }, r.a.createElement("a", {
                    href: o,
                    target: u,
                    onClick: function(e) {
                        c && (e.preventDefault(), c(e))
                    },
                    title: l
                }, r.a.createElement("span", {
                    className: "nav-item-icon-container"
                }, r.a.createElement("span", {
                    className: "nav-item-icon"
                }, n), t && r.a.createElement("span", {
                    className: "nav-item-banner"
                }, t)), r.a.createElement("span", {
                    className: "nav-item-title"
                }, l)))
            }
    }])
}));
