// feather/lib/components/tooltip/Tooltip.js
"use strict";require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");Object.defineProperty(exports, "__esModule", { value: true });exports.Tooltip = exports.TooltipContent = void 0;var _react = require("react");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ContextualHelp = require("../../../lib/components/ContextualHelp");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _themes = require("../../../lib/themes");
var _typography = require("../../../lib/typography");
var _zIndexes = require("../../../lib/zIndexes");

var _TooltipWrapper = require("./TooltipWrapper");
var _types = require("./types");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}






var TooltipContentWrapper = _styledComponents.default.div.withConfig({ displayName: "Tooltip__TooltipContentWrapper", componentId: "sc-3ua8e5-0" })(["z-index:", ";border-radius:4px;background-color:", ";padding:6px 8px;max-width:256px;line-height:1;white-space:pre-wrap;word-break:keep-all;color:", ";font-family:", ";font-size:12px;font-weight:500;", ""],
_zIndexes.ZIndexes.tooltip,

(0, _cssVariables.default)('neutral-9'),





(0, _cssVariables.default)('bg-1'),
_typography.typeface.system,


function (props) {return props.$tooltipContentStyle;});


var Arrow = _styledComponents.default.div.withConfig({ displayName: "Tooltip__Arrow", componentId: "sc-3ua8e5-1" })(["position:absolute;pointer-events:none;&[data-placement*='bottom']{top:0;margin-top:-6px;width:12px;height:6px;&::before{border-width:0 6px 6px 6px;border-color:transparent transparent ", " transparent;}}&[data-placement*='top']{bottom:0;margin-bottom:-6px;width:12px;height:6px;&::before{border-width:6px 6px 0 6px;border-color:", " transparent transparent transparent;}}&[data-placement*='right']{left:0;margin-left:-6px;height:12px;width:6px;&::before{border-width:6px 6px 6px 0;border-color:transparent ", " transparent transparent;}}&[data-placement*='left']{right:0;margin-right:-6px;height:12px;width:6px;&::before{border-width:6px 0 6px 6px;border-color:transparent transparent transparent ", ";}}&::before{content:'';margin:auto;display:block;width:0;height:0;border-style:solid;pointer-events:none;}"],










(0, _cssVariables.default)('neutral-9'),










(0, _cssVariables.default)('neutral-9'),










(0, _cssVariables.default)('neutral-9'),










(0, _cssVariables.default)('neutral-9'));














var TooltipContent = /*#__PURE__*/(0, _react.forwardRef)(
function (_ref, ref) {var variant = _ref.variant,content = _ref.content,style = _ref.style,placement = _ref.placement,arrowProps = _ref.arrowProps,onTooltipReferenceMouseLeave = _ref.onTooltipReferenceMouseLeave,tooltipContentStyle = _ref.tooltipContentStyle;return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(_themes.FeatherThemeContext.Provider, { value: _themes.FeatherThemeEnum.Neutral, children: /*#__PURE__*/
      (0, _jsxRuntime.jsxs)(TooltipContentWrapper, {
        ref: ref,
        role: "tooltip",
        onMouseLeave: onTooltipReferenceMouseLeave,
        style: style,
        variant: variant,
        $tooltipContentStyle: tooltipContentStyle, children: [/*#__PURE__*/





        (0, _jsxRuntime.jsx)(Arrow, { ref: arrowProps.ref, style: arrowProps.style, "data-placement": placement, "data-is-arrow": "true" }),
        content] }) }));});





/**
 * Tooltips display a text label identifying an element, such as a description of its function.
 */exports.TooltipContent = TooltipContent;
var Tooltip = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  /**
   * Default variant of Tooltip is TooltipVariant.Dark
   */
  var _props$variant = props.variant,variant = _props$variant === void 0 ? _types.TooltipVariant.Dark : _props$variant;

  if (variant === _types.TooltipVariant.Light) {
    /**
     * ContextualHelp component is for Tooltips using TooltipVariant.Light
     */
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_ContextualHelp.ContextualHelp, _objectSpread({ ref: ref }, props));
  }

  var children = props.children,tooltipContentStyle = props.tooltipContentStyle,wrapperProps = _objectWithoutProperties(props, ["children", "tooltipContentStyle"]);
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(_TooltipWrapper.TooltipWrapper, _objectSpread(_objectSpread({}, wrapperProps), {}, { tooltipRef: ref, target: children, children:
      function children(popperChildrenProps) {return /*#__PURE__*/(
          (0, _jsxRuntime.jsx)(TooltipContent, _objectSpread(_objectSpread({},
          popperChildrenProps), {}, {
            variant: variant,
            tooltipContentStyle: tooltipContentStyle,
            content: props.content })));} })));




});exports.Tooltip = Tooltip;
