// @postman/aether/esmLib/components/Heading/Heading.js
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { globals, aliases } from '../../design-tokens/tokens';
import { isTextTruncatable, wrapWithTooltip } from '../../utils/common/index';

function getTextStyles(type, theme) {
  switch (type) {
    case 'h1':
      return "\n        font-size: ".concat(theme['text-size-xxxl'], ";\n        line-height: 1.2;\n        color: ").concat(theme['content-color-primary'], ";\n        letter-spacing: -0.72px;\n      ");

    case 'h2':
      return "\n        font-size: ".concat(theme['text-size-xxl'], ";\n        line-height: 1.2;\n        color: ").concat(theme['content-color-primary'], ";\n        letter-spacing: -0.56px;\n      ");

    case 'h3':
      return "\n        font-size: ".concat(theme['text-size-xl'], ";\n        line-height: 1.28;\n        color: ").concat(theme['content-color-primary'], ";\n        letter-spacing: -0.12px;\n      ");

    case 'h4':
      return "\n        font-size: ".concat(theme['text-size-l'], ";\n        line-height: 1.44;\n        color: ").concat(theme['content-color-primary'], ";\n        letter-spacing: -0.12px;\n      ");

    case 'h5':
      return "\n        font-size: ".concat(theme['text-size-m'], ";\n        line-height: 1.4;\n        color: ").concat(theme['content-color-primary'], ";\n      ");

    case 'h6':
      return "\n        font-size: ".concat(theme['text-size-m'], ";\n        line-height: 1.4;\n        color: ").concat(theme['content-color-secondary'], ";\n      ");

    default:
      return null;
  }
}
/**
 * Function to get truncation related styles for the typography components
 * @param {boolean} isTruncated - determines whether single line truncation is enabled
 * @param {number} maxLines - enables multi-line truncation
 * @returns a string of css styles.
 */


function getTextTruncationStyles(isTruncated, maxLines) {
  // single line truncation styles
  // also adding styles on top of truncation styles to prevent tooltip coming on truncated elements on Safari.
  var isTruncatedStyles = "\n      text-overflow: ellipsis;\n      overflow: hidden;\n      white-space: nowrap;\n\n      &::after {\n        content: ' ';\n        display: block;\n        width: 0;\n        height: 0;\n        line-height: 0;\n      }\n    "; // multi-line truncation styles

  var maxLinesStyles = "\n    display: -webkit-box;\n    -webkit-line-clamp: ".concat(maxLines, ";\n    -webkit-box-orient: vertical;  \n    overflow: hidden;\n    ");
  return "\n    ".concat(isTruncated ? isTruncatedStyles : '', "\n    ").concat(maxLines ? maxLinesStyles : '', "\n    ");
}
/**
 * Function to apply bottom margin to the Heading according to its type
 * It is added since the codebase extensively uses margins to align Headings. And therefore we do not want to
 * enforce the use of classNames for that.
 * @param {String} type - the type of heading (h1 - h6)
 */


function getMarginStyles(type, theme) {
  switch (type) {
    case 'h1':
      return "margin-bottom: ".concat(theme['spacing-l']);

    case 'h2':
      return "margin-bottom: ".concat(theme['spacing-m']);

    case 'h3':
      return "margin-bottom: ".concat(theme['spacing-s']);

    case 'h4':
      return "margin-bottom: ".concat(theme['spacing-xs']);

    case 'h5':
      return "margin-bottom: ".concat(theme['spacing-xs']);

    case 'h6':
      return "margin-bottom: ".concat(theme['spacing-xs']);

    default:
      return "margin: ".concat(theme['spacing-zero']);
  }
}

var StyledHeading = styled.h1.withConfig({
  displayName: "Heading__StyledHeading",
  componentId: "sc-wz2vta-0"
})(["margin:", ";font-family:", ";font-weight:", ";word-break:", ";", ";", ";", ";", ""], function (props) {
  return props.theme['spacing-zero'];
}, function (props) {
  return props.theme['text-family-default'];
}, function (props) {
  return props.theme['text-weight-medium'];
}, function (props) {
  return props.wordBreak;
}, function (props) {
  return props.styleAs ? getTextStyles(props.styleAs, props.theme) : getTextStyles(props.as, props.theme);
}, function (props) {
  return props.color && "color: ".concat(props.theme[props.color], ";");
}, function (props) {
  return props.hasBottomSpacing && getMarginStyles(props.as, props.theme);
}, function (props) {
  return getTextTruncationStyles(props.isTruncated, props.maxLines);
});
export default function Heading(props) {
  var text = props.text,
      type = props.type,
      styleAs = props.styleAs,
      color = props.color,
      _wordBreak = props._wordBreak,
      className = props.className,
      hasBottomSpacing = props.hasBottomSpacing,
      isTruncated = props.isTruncated,
      maxLines = props.maxLines,
      tooltip = props.tooltip;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isTooltipVisible = _useState2[0],
      setTooltipVisible = _useState2[1];
  /**
   * Function to detect if the text content is truncated when the text is hovered to show a tooltip
   * @param {object} e - the event object for onMouseEnter listener
   */


  function handleTooltipVisible(e) {
    if (e.target && e.target.scrollWidth > e.target.offsetWidth) {
      setTooltipVisible(true);
    } else {
      setTooltipVisible(false);
    }
  }

  var jsx = /*#__PURE__*/React.createElement(StyledHeading, {
    as: type,
    styleAs: styleAs,
    color: color,
    className: className,
    hasBottomSpacing: hasBottomSpacing,
    isTruncated: isTruncated,
    maxLines: maxLines,
    onMouseEnter: handleTooltipVisible,
    wordBreak: _wordBreak
  }, text); // If single line truncation enabled, wrap inside tooltip else render the plain old component.

  return isTooltipVisible && isTextTruncatable(props) ? wrapWithTooltip(jsx, tooltip || text) : jsx;
}
Heading.defaultProps = {
  color: '',
  type: 'h1',
  styleAs: null,
  className: '',
  hasBottomSpacing: false,
  isTruncated: false,
  maxLines: undefined,
  tooltip: '',
  _wordBreak: 'normal'
};
Heading.propTypes = {
  type: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  text: PropTypes.string.isRequired,
  color: PropTypes.oneOf([''].concat(_toConsumableArray(Object.keys(globals.colors)), _toConsumableArray(Object.keys(aliases.light.content)))),
  styleAs: PropTypes.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  className: PropTypes.string,
  hasBottomSpacing: PropTypes.bool,
  isTruncated: function isTruncated(_ref) {
    var _isTruncated = _ref.isTruncated,
        maxLines = _ref.maxLines;

    if (typeof _isTruncated !== 'boolean') {
      return new Error('Heading: expected a boolean value for isTruncated prop');
    }

    if (_isTruncated && maxLines) {
      return new Error('Heading: maxLines should not be used together with isTruncated prop');
    }

    return null;
  },
  maxLines: PropTypes.number,
  tooltip: function tooltip(props) {
    var tooltip = props.tooltip,
        isTruncated = props.isTruncated,
        maxLines = props.maxLines;

    if (typeof tooltip !== 'string') {
      return new Error('Heading: tooltip prop should only have a string value.');
    }

    if (tooltip && (!isTruncated || maxLines)) {
      return new Error('Heading: tooltip prop should only be used when using single-line truncation by using isTruncated prop.');
    }

    return null;
  },
  _wordBreak: PropTypes.oneOf(['normal', 'break-all', 'keep-all', 'break-word', 'inherit', 'initial', 'revert', 'unset'])
};
