// feather/lib/components/avatar/Avatar.js
"use strict";require("core-js/modules/es.symbol.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.symbol.iterator.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.array.slice.js");require("core-js/modules/es.function.name.js");require("core-js/modules/es.array.from.js");Object.defineProperty(exports, "__esModule", { value: true });exports.Avatar = void 0;require("core-js/modules/es.array.includes.js");require("core-js/modules/es.string.includes.js");var _react = require("react");
var _styledComponents = _interopRequireDefault(require("styled-components"));

var _spinner = require("../../../lib/components/spinner");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));

var _getInterpolatedYValue = require("../../utils/getInterpolatedYValue");
var _AvatarType = require("./AvatarType");
var _getHashcode = require("./getHashcode");
var _RandomAvatar = require("./RandomAvatar");
var _StatusDot = require("./StatusDot");var _jsxRuntime = require("react/jsx-runtime");var _shapeMap;function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _slicedToArray(arr, i) {return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();}function _nonIterableRest() {throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");}function _unsupportedIterableToArray(o, minLen) {if (!o) return;if (typeof o === "string") return _arrayLikeToArray(o, minLen);var n = Object.prototype.toString.call(o).slice(8, -1);if (n === "Object" && o.constructor) n = o.constructor.name;if (n === "Map" || n === "Set") return Array.from(o);if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);}function _arrayLikeToArray(arr, len) {if (len == null || len > arr.length) len = arr.length;for (var i = 0, arr2 = new Array(len); i < len; i++) {arr2[i] = arr[i];}return arr2;}function _iterableToArrayLimit(arr, i) {var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]);if (_i == null) return;var _arr = [];var _n = true;var _d = false;var _s, _e;try {for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {_arr.push(_s.value);if (i && _arr.length === i) break;}} catch (err) {_d = true;_e = err;} finally {try {if (!_n && _i["return"] != null) _i["return"]();} finally {if (_d) throw _e;}}return _arr;}function _arrayWithHoles(arr) {if (Array.isArray(arr)) return arr;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}var















AvatarShape;(function (AvatarShape) {AvatarShape["RoundedSquare"] = "square";AvatarShape["Circle"] = "circle";})(AvatarShape || (AvatarShape = {}));




var sizeMap = {
  xsmall: 16,
  small: 20,
  medium: 32,
  xmedium: 40,
  large: 60 };


var spinnerSizeMap = [
[60, 24],
[40, 20],
[32, 16],
[20, 10],
[16, 8]];


var squareAvatarBorderRadiusMap = [
[60, 12],
[40, 8],
[32, 4],
[20, 4],
[16, 2]];


var shapeMap = (_shapeMap = {}, _defineProperty(_shapeMap,
_AvatarType.AvatarType.Application, AvatarShape.RoundedSquare), _defineProperty(_shapeMap,
_AvatarType.AvatarType.Channel, AvatarShape.RoundedSquare), _defineProperty(_shapeMap,
_AvatarType.AvatarType.Member, AvatarShape.Circle), _defineProperty(_shapeMap,
_AvatarType.AvatarType.Organization, AvatarShape.RoundedSquare), _defineProperty(_shapeMap,
_AvatarType.AvatarType.User, AvatarShape.Circle), _defineProperty(_shapeMap,
_AvatarType.AvatarType.Bot, AvatarShape.Circle), _shapeMap);


var statusColorMap = {
  online: (0, _cssVariables.default)('bg-positive'),
  away: (0, _cssVariables.default)('bg-attention'),
  offline: (0, _cssVariables.default)('neutral-5') };


var Container = _styledComponents.default.div.withConfig({ displayName: "Avatar__Container", componentId: "sc-5mfacb-0" })(["position:relative;flex:none;width:", "px;height:", "px;"],


function (_ref) {var $size = _ref.$size;return $size;},
function (_ref2) {var $size = _ref2.$size;return $size;});


var AvatarWrapper = _styledComponents.default.div.withConfig({ displayName: "Avatar__AvatarWrapper", componentId: "sc-5mfacb-1" })(["position:relative;display:flex;justify-content:center;align-items:center;width:", "px;height:", "px;border-radius:", ";overflow:hidden;user-select:none;"],








function (_ref3) {var $size = _ref3.$size;return $size;},
function (_ref4) {var $size = _ref4.$size;return $size;},
function (_ref5) {var $shape = _ref5.$shape,$squareBorderRadius = _ref5.$squareBorderRadius;return $shape === 'circle' ? '50%' : "".concat($squareBorderRadius, "px");});




var SpinnerWrapper = (0, _styledComponents.default)(AvatarWrapper).withConfig({ displayName: "Avatar__SpinnerWrapper", componentId: "sc-5mfacb-2" })(["background-color:", ";"],
(0, _cssVariables.default)('bg-3'));


var Image = _styledComponents.default.img.withConfig({ displayName: "Avatar__Image", componentId: "sc-5mfacb-3" })(["position:absolute;top:0;left:0;width:100%;height:100%;object-fit:cover;"]);








var NotFoundAvatar = function NotFoundAvatar(_ref6) {var size = _ref6.size;return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)("svg", { xmlns: "http://www.w3.org/2000/svg", width: size, height: size, viewBox: "0 0 72 72", children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)("g", { transform: "translate(16 16)", children: /*#__PURE__*/
        (0, _jsxRuntime.jsx)("path", {
          fill: (0, _cssVariables.default)('content-3'),
          d: "M20 38.333C9.875 38.333 1.667 30.125 1.667 20 1.667 9.875 9.875 1.667 20 1.667c10.125 0 18.333 8.208 18.333 18.333 0 10.125-8.208 18.333-18.333 18.333zM20 35c8.284 0 15-6.716 15-15 0-8.284-6.716-15-15-15-8.284 0-15 6.716-15 15 0 8.284 6.716 15 15 15zm1.383-11.644c0 .92-.746 1.667-1.666 1.667-.92 0-1.667-.746-1.667-1.667v-1.681c0-.718.46-1.355 1.14-1.582.108-.036.332-.121.631-.254.505-.224 1.01-.494 1.471-.801 1.147-.764 1.758-1.58 1.758-2.366.002-1.625-1.168-3.015-2.77-3.29-1.602-.275-3.168.646-3.708 2.179-.305.868-1.257 1.325-2.125 1.02-.868-.306-1.325-1.258-1.02-2.126 1.08-3.067 4.213-4.908 7.417-4.359 3.204.55 5.544 3.33 5.54 6.579 0 2.132-1.264 3.817-3.243 5.136-.61.407-1.212.734-1.758.987v.558zm-3.333 4.702c0-.92.746-1.667 1.667-1.667.92 0 1.666.747 1.666 1.667v.275c0 .92-.746 1.667-1.666 1.667-.92 0-1.667-.746-1.667-1.667v-.193-.082z" }) }) }));};





var Avatar = function Avatar(_ref7) {var _ref7$profileID = _ref7.profileID,profileID = _ref7$profileID === void 0 ? 0 : _ref7$profileID,imageUrl = _ref7.imageUrl,className = _ref7.className,type = _ref7.type,size = _ref7.size,status = _ref7.status,isLoading = _ref7.isLoading;
  var shape = shapeMap[type];

  var _useState = (0, _react.useState)(false),_useState2 = _slicedToArray(_useState, 2),didFailToLoadImage = _useState2[0],setDidFailToLoadImage = _useState2[1];
  var _useState3 = (0, _react.useState)(true),_useState4 = _slicedToArray(_useState3, 2),isImageLoadCompleted = _useState4[0],setIsImageLoadCompleted = _useState4[1];

  (0, _react.useEffect)(function () {
    // If imageUrl is empty, set isImageLoadCompleted true.
    setIsImageLoadCompleted(!imageUrl);

    setDidFailToLoadImage(!imageUrl);
  }, [imageUrl]);

  var onImgLoad = (0, _react.useCallback)(function () {return setIsImageLoadCompleted(true);}, []);

  var profileHash = typeof profileID === 'string' ? (0, _getHashcode.getHashcode)(profileID) : profileID;
  var sizeInNumber = typeof size === 'number' ? size : sizeMap[size];
  var squareAvatarBorderRadius = (0, _getInterpolatedYValue.getInterpolatedYValue)(squareAvatarBorderRadiusMap, sizeInNumber);
  var spinnerSize = (0, _getInterpolatedYValue.getInterpolatedYValue)(spinnerSizeMap, sizeInNumber);

  var shapedSpinner = (0, _react.useMemo)(function () {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsx)(SpinnerWrapper, { $size: sizeInNumber, $shape: shape, $squareBorderRadius: squareAvatarBorderRadius, children: /*#__PURE__*/
        (0, _jsxRuntime.jsx)(_spinner.Spinner, { size: spinnerSize, stroke: (0, _cssVariables.default)('content-3') }) }));


  }, [shape, sizeInNumber, spinnerSize, squareAvatarBorderRadius]);

  var statusDot =
  status && [_AvatarType.AvatarType.User, _AvatarType.AvatarType.Member, _AvatarType.AvatarType.Bot].includes(type) ? /*#__PURE__*/
  (0, _jsxRuntime.jsx)(_StatusDot.StatusDot, { color: statusColorMap[status], avatarSize: sizeInNumber, type: type }) :
  null;

  var wrapAvatar = (0, _react.useCallback)(
  function (children) {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsx)(AvatarWrapper, {
        $size: sizeInNumber,
        $shape: shape,
        $squareBorderRadius: squareAvatarBorderRadius,
        style: isImageLoadCompleted ? undefined : { backgroundColor: (0, _cssVariables.default)('neutral-3') }, children:

        children }));


  },
  [isImageLoadCompleted, shape, sizeInNumber, squareAvatarBorderRadius]);


  if (isLoading === true) {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsxs)(Container, { className: className, $size: sizeInNumber, children: [
        shapedSpinner,
        statusDot] }));


  }

  if (!imageUrl) {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsxs)(Container, { className: className, $size: sizeInNumber, children: [
        wrapAvatar( /*#__PURE__*/(0, _jsxRuntime.jsx)(_RandomAvatar.RandomAvatar, { type: type, size: sizeInNumber, hashcode: profileHash })),
        statusDot] }));


  }

  if (didFailToLoadImage) {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsxs)(Container, { className: className, $size: sizeInNumber, children: [
        wrapAvatar( /*#__PURE__*/(0, _jsxRuntime.jsx)(NotFoundAvatar, { size: sizeInNumber })),
        statusDot] }));


  }

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(Container, { className: className, $size: sizeInNumber, children: [/*#__PURE__*/
      (0, _jsxRuntime.jsx)(AvatarWrapper, {
        $size: sizeInNumber,
        $shape: shape,
        $squareBorderRadius: squareAvatarBorderRadius,
        style: isImageLoadCompleted ? undefined : { backgroundColor: (0, _cssVariables.default)('neutral-3') }, children:

        wrapAvatar( /*#__PURE__*/
        (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, { children: [/*#__PURE__*/
          (0, _jsxRuntime.jsx)(Image, {
            ref: function ref(imgElement) {
              var isImageLoaded = imgElement ? imgElement.complete && imgElement.naturalHeight !== 0 : false;
              setIsImageLoadCompleted(isImageLoaded);
            },
            src: imageUrl,
            onError: function onError() {
              setDidFailToLoadImage(true);
            },
            onLoad: onImgLoad,
            style: { opacity: isImageLoadCompleted ? 1 : 0 } }),

          !isImageLoadCompleted && /*#__PURE__*/(0, _jsxRuntime.jsx)(_spinner.Spinner, { size: spinnerSize, stroke: (0, _cssVariables.default)('neutral-6') })] })) }),



      statusDot] }));


};exports.Avatar = Avatar;
