// @postman/aether/esmLib/components/CheckBox/Checkbox.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useState } from 'react';
import styled from 'styled-components';
import propTypes from 'prop-types';
import Text from '../Text/Text';
import Label from '../Label/Label';
import IconCheckboxTick from '../../design-tokens/design-system-icons/icon-checkbox-tick';
import IconCheckboxDash from '../../design-tokens/design-system-icons/icon-checkbox-dash';
/**
 * @description This function dynamically decides the border style of checkbox
 * @param {boolean} isChecked holds the checked status of checkbox
 * @param {boolean} isDisabled holds the disabled status of checkbox
 * @param {boolean} isIndeterminate holds the indeterminate property of checkbox
 * @param {object} theme holds current theme value
 */

function getBorderStyle(isChecked, isDisabled, isIndeterminate, theme) {
  var borderStyle;
  borderStyle = isChecked || isIndeterminate ? "".concat(theme['border-width-default'], " ").concat(theme['border-style-solid'], " transparent") : "".concat(theme['border-width-default'], " ").concat(theme['border-style-solid'], " ").concat(theme['checkbox-border-color']);
  if (isDisabled) borderStyle = "".concat(theme['border-width-default'], " ").concat(theme['border-style-solid'], " ").concat(theme['checkbox-disabled-border-color']);
  return borderStyle;
}
/**
 * @description This function dynamically decides which icon to place
 *  * @param {boolean} isChecked holds the checked status of checkbox
 * @param {boolean} isIndeterminate holds the indeterminate property of checkbox
 */


function getCurrentIcon(isChecked, isIndeterminate) {
  // indeterminate icon suppresses the presence of checked icon
  if (isIndeterminate) {
    return /*#__PURE__*/React.createElement(IconCheckboxDash, null);
  }

  if (isChecked) {
    return /*#__PURE__*/React.createElement(IconCheckboxTick, null);
  }

  return '';
}
/**
 * @description This function decides background color of Checkbox
 * @param {boolean} isChecked holds the checked status of checkbox
 * @param {boolean} isIndeterminate holds the indeterminate property of checkbox
 * @param {boolean} isDisabled specifies whether the checkbox is disabled or not
 * @param {string} type holds type of checkbox
 * @param {object} theme holds the current theme values
 */


function defineCheckBoxColor(isChecked, isIndeterminate, isDisabled, type, theme) {
  if (isChecked || isIndeterminate) {
    if (isDisabled) {
      return theme['checkbox-disabled-background-color'];
    }

    if (type === 'primary') {
      return theme['checkbox-background-color'];
    }

    if (type === 'secondary') {
      return theme['content-color-primary'];
    }
  }

  return 'transparent';
} // styled component for main checkbox container


var StyledCheckBoxActionableContainer = styled.span.withConfig({
  displayName: "Checkbox__StyledCheckBoxActionableContainer",
  componentId: "sc-u5fnvh-0"
})(["position:relative;cursor:pointer;user-select:none;display:flex;width:max-content;align-items:center;& *{", "}"], function (props) {
  return props.isDisabled ? 'cursor: not-allowed;' : 'cursor: pointer';
}); // styled component for label and helper text container

var StyledIconContainer = styled.span.withConfig({
  displayName: "Checkbox__StyledIconContainer",
  componentId: "sc-u5fnvh-1"
})(["min-height:14px;min-width:14px;height:14px;width:14px;border:", ";border-radius:", ";background-color:", ";box-sizing:border-box;text-align:center;z-index:1;display:flex;justify-content:center;align-items:center;margin-right:", ";svg{path{fill:", ";}}"], function (props) {
  return getBorderStyle(props.isChecked, props.isDisabled, props.isIndeterminate, props.theme);
}, function (props) {
  return props.theme[['border-radius-s']];
}, function (props) {
  return defineCheckBoxColor(props.isChecked, props.isIndeterminate, props.isDisabled, props.type, props.theme);
}, function (props) {
  return props.label && props.theme['spacing-s'];
}, function (props) {
  return props.theme['checkbox-marker-color'];
}); // Container for Input CheckBoxes

var StyledInputCheckBox = styled.input.attrs({
  type: 'checkbox'
}).withConfig({
  displayName: "Checkbox__StyledInputCheckBox",
  componentId: "sc-u5fnvh-2"
})(["margin:0;border:0;height:14px;width:14px;position:absolute;appearance:none;&:focus{outline-color:", ";}"], function (props) {
  return props.theme['checkbox-outline-color'];
}); // container for helper Texts

var StyledHelperTextContainer = styled.span.withConfig({
  displayName: "Checkbox__StyledHelperTextContainer",
  componentId: "sc-u5fnvh-3"
})(["margin-left:22px;"]);
var StyledCheckBoxContainer = styled.div.withConfig({
  displayName: "Checkbox__StyledCheckBoxContainer",
  componentId: "sc-u5fnvh-4"
})(["width:max-content;"]);
export default function Checkbox(props) {
  var label = props.label,
      isChecked = props.isChecked,
      id = props.id,
      name = props.name,
      value = props.value,
      className = props.className,
      helperText = props.helperText,
      isIndeterminate = props.isIndeterminate,
      isDisabled = props.isDisabled,
      onChange = props.onChange,
      onMouseDown = props.onMouseDown,
      onMouseUp = props.onMouseUp,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      onMouseEnter = props.onMouseEnter,
      onMouseLeave = props.onMouseLeave,
      type = props.type; // state for maintaining user interaction which is the source of generating all styles

  var _useState = useState('default'),
      _useState2 = _slicedToArray(_useState, 2),
      interactionState = _useState2[0],
      setInteractionState = _useState2[1];

  var ariaChecked = 'false';

  if (isIndeterminate) {
    ariaChecked = 'mixed';
  } else if (isChecked) {
    ariaChecked = 'true';
  }

  var handleOnChange = function handleOnChange(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      if (onChange) onChange(!isChecked, value, event);
    }
  };

  var handleOnMouseEnter = function handleOnMouseEnter(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      setInteractionState('hover');
      if (onMouseEnter) onMouseEnter(event);
    }
  };

  var handleOnMouseLeave = function handleOnMouseLeave(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      setInteractionState('default');
      if (onMouseLeave) onMouseLeave(event);
    }
  };

  var handleOnMouseDown = function handleOnMouseDown(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      setInteractionState('active');
      if (onMouseDown) onMouseDown(event);
    }
  };

  var handleOnMouseUp = function handleOnMouseUp(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      if (onMouseUp) onMouseUp(event);
    }
  };

  var handleOnFocus = function handleOnFocus(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      setInteractionState('focus');
      if (onFocus) onFocus(event);
    }
  };

  var handleOnBlur = function handleOnBlur(event) {
    // stopping unnecessary Fn calls when checkbox is disabled
    if (!isDisabled) {
      if (onBlur) onBlur(event);
      setInteractionState('default');
    }
  };

  return /*#__PURE__*/React.createElement(StyledCheckBoxContainer, {
    className: className
  }, /*#__PURE__*/React.createElement(StyledCheckBoxActionableContainer, {
    isDisabled: isDisabled,
    onClick: handleOnChange,
    onMouseEnter: handleOnMouseEnter,
    onMouseLeave: handleOnMouseLeave,
    onMouseDown: handleOnMouseDown,
    onMouseUp: handleOnMouseUp,
    onBlur: handleOnBlur,
    onFocus: handleOnFocus,
    "aria-checked": ariaChecked,
    role: "checkbox"
  }, /*#__PURE__*/React.createElement(StyledIconContainer, {
    label: label,
    isChecked: isChecked,
    isIndeterminate: isIndeterminate,
    isDisabled: isDisabled,
    interactionState: interactionState,
    type: type
  }, getCurrentIcon(isChecked, isIndeterminate)), /*#__PURE__*/React.createElement(StyledInputCheckBox, {
    id: id,
    name: name,
    value: value,
    checked: isChecked,
    onChange: function onChange() {}
  }),
  /* label component */
  label ? /*#__PURE__*/React.createElement(Label, {
    htmlFor: id,
    isDisabled: isDisabled,
    type: "optionSelector",
    text: label
  }) : ''),
  /* helper text component */
  label && helperText ? /*#__PURE__*/React.createElement(StyledHelperTextContainer, null, /*#__PURE__*/React.createElement(Text, {
    type: "body-small",
    color: "content-color-secondary"
  }, helperText)) : '');
}
Checkbox.defaultProps = {
  isChecked: false,
  id: null,
  name: null,
  value: '',
  className: '',
  helperText: '',
  isIndeterminate: false,
  isDisabled: false,
  label: '',
  onMouseDown: null,
  onMouseUp: null,
  onFocus: null,
  onBlur: null,
  onMouseEnter: null,
  onMouseLeave: null,
  type: 'primary'
};
Checkbox.propTypes = {
  /** Use this prop to set checked state of CheckBox component */
  isChecked: propTypes.bool,

  /** Use this prop to set id prop of CheckBox component */
  id: propTypes.string,

  /** Use this prop to set name prop of CheckBox component */
  name: propTypes.string,

  /** Use this prop to set value prop of CheckBox component */
  value: propTypes.string,

  /** Use this prop to provide a label to the CheckBox component */
  label: propTypes.string,

  /** Defines the actionable nature of the Checkbox */
  type: propTypes.oneOf(['primary', 'secondary']),

  /** Use this prop to provide a helper text to the CheckBox component */
  helperText: propTypes.string,

  /** Use this prop to set indeterminate of checkbox. When a checkbox is indeterminate, it overrides the selection state. */
  isIndeterminate: propTypes.bool,

  /** Use this prop to disable the checkbox */
  isDisabled: propTypes.bool,

  /** This prop accepts a function which would be triggered when the checkbox is clicked */
  onChange: propTypes.func.isRequired,

  /** This prop accepts a function which would be triggered when the mouse enters within its box area */
  onMouseEnter: propTypes.func,

  /** This prop accepts a function which would be triggered when the mouse leaves within its box area */
  onMouseLeave: propTypes.func,

  /** This prop accepts a function which would be triggered when the mouse buttons are clicked down */
  onMouseDown: propTypes.func,

  /** This prop accepts a function which would be triggered when the mouse buttons are released after clicking */
  onMouseUp: propTypes.func,

  /** This prop accepts a function which would be triggered when the checkbox is focused */
  onFocus: propTypes.func,

  /** This prop accepts a function which would be triggered when the checkbox is focused out */
  onBlur: propTypes.func,

  /** You can provide classNames to the component. Adding styles using class name is not recommended. */
  className: propTypes.string
};
