// @elastic/charts/dist/chart_types/heatmap/layout/types/viewmodel_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nullShapeViewModel = exports.nullHeatmapViewModel = exports.isPickedCells = void 0;
var __1 = require("../../..");
var config_1 = require("../config/config");
function isPickedCells(v) {
    return Array.isArray(v);
}
exports.isPickedCells = isPickedCells;
exports.nullHeatmapViewModel = {
    gridOrigin: {
        x: 0,
        y: 0,
    },
    gridLines: {
        x: [],
        y: [],
        stroke: { width: 0, color: { r: 0, g: 0, b: 0, opacity: 0 } },
    },
    cells: [],
    xValues: [],
    yValues: [],
    pageSize: 0,
};
var nullShapeViewModel = function (specifiedConfig) { return ({
    config: specifiedConfig || config_1.config,
    heatmapViewModel: exports.nullHeatmapViewModel,
    pickQuads: function () { return []; },
    pickDragArea: function () { return ({ cells: [], x: [], y: [], chartType: __1.ChartType.Heatmap }); },
    pickDragShape: function () { return ({ x: 0, y: 0, width: 0, height: 0 }); },
    pickHighlightedArea: function () { return ({ x: 0, y: 0, width: 0, height: 0 }); },
}); };
exports.nullShapeViewModel = nullShapeViewModel;
