// @splitsoftware/splitio/lib/utils/context/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IMPRESSIONS_COUNTER = exports.STORAGE = exports.STATUS_MANAGER = exports.SETTINGS = exports.READY_FROM_CACHE = exports.READY = exports.READINESS = exports.PRODUCER = exports.MY_SEGMENTS_CHANGE_WORKER = exports.INTEGRATIONS_MANAGER = exports.HAS_TIMEDOUT = exports.EVENTS = exports.DESTROYED = exports.COLLECTORS = void 0;
var COLLECTORS = 'metrics_collectors';
exports.COLLECTORS = COLLECTORS;
var DESTROYED = 'is_destroyed';
exports.DESTROYED = DESTROYED;
var EVENTS = 'events_publisher';
exports.EVENTS = EVENTS;
var HAS_TIMEDOUT = 'has_timedout';
exports.HAS_TIMEDOUT = HAS_TIMEDOUT;
var INTEGRATIONS_MANAGER = 'integrations_manager';
exports.INTEGRATIONS_MANAGER = INTEGRATIONS_MANAGER;
var MY_SEGMENTS_CHANGE_WORKER = 'my_segments_change_worker';
exports.MY_SEGMENTS_CHANGE_WORKER = MY_SEGMENTS_CHANGE_WORKER;
var PRODUCER = 'producer';
exports.PRODUCER = PRODUCER;
var READINESS = 'readiness_gate';
exports.READINESS = READINESS;
var READY = 'is_ready';
exports.READY = READY;
var READY_FROM_CACHE = 'is_ready_from_cache';
exports.READY_FROM_CACHE = READY_FROM_CACHE;
var SETTINGS = 'settings';
exports.SETTINGS = SETTINGS;
var STATUS_MANAGER = 'status_manager';
exports.STATUS_MANAGER = STATUS_MANAGER;
var STORAGE = 'storage';
exports.STORAGE = STORAGE;
var IMPRESSIONS_COUNTER = 'impressions_counter';
exports.IMPRESSIONS_COUNTER = IMPRESSIONS_COUNTER;
var _default = {
  COLLECTORS: COLLECTORS,
  DESTROYED: DESTROYED,
  EVENTS: EVENTS,
  HAS_TIMEDOUT: HAS_TIMEDOUT,
  INTEGRATIONS_MANAGER: INTEGRATIONS_MANAGER,
  MY_SEGMENTS_CHANGE_WORKER: MY_SEGMENTS_CHANGE_WORKER,
  PRODUCER: PRODUCER,
  READINESS: READINESS,
  READY: READY,
  READY_FROM_CACHE: READY_FROM_CACHE,
  SETTINGS: SETTINGS,
  STATUS_MANAGER: STATUS_MANAGER,
  STORAGE: STORAGE,
  IMPRESSIONS_COUNTER: IMPRESSIONS_COUNTER
};
exports.default = _default;
