// @pipedrive/leads-sidebar/dist/Organization/__generated__/Organization_organization.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ReaderFragment } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type Organization_organization = {
    readonly relatedOrganizationId: string;
    readonly isLinkedToLead: boolean;
    readonly isDeleted: boolean | null;
    readonly name: string | null;
    readonly address: string | null;
    readonly customFields: ReadonlyArray<{
        readonly id: string;
        readonly " $fragmentRefs": FragmentRefs<"CustomFieldRenderer_customField" | "CustomFieldRenderer_customFieldInfo">;
    } | null> | null;
    readonly " $refType": "Organization_organization";
};
export type Organization_organization$data = Organization_organization;
export type Organization_organization$key = {
    readonly " $data"?: Organization_organization$data;
    readonly " $fragmentRefs": FragmentRefs<"Organization_organization">;
};



const node: ReaderFragment = {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "Organization_organization",
    "selections": [
        {
            "alias": "relatedOrganizationId",
            "args": [
                {
                    "kind": "Literal",
                    "name": "opaque",
                    "value": false
                }
            ],
            "kind": "ScalarField",
            "name": "id",
            "storageKey": "id(opaque:false)"
        },
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "isLinkedToLead",
            "storageKey": null
        },
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "isDeleted",
            "storageKey": null
        },
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "name",
            "storageKey": null
        },
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "address",
            "storageKey": null
        },
        {
            "alias": null,
            "args": null,
            "concreteType": null,
            "kind": "LinkedField",
            "name": "customFields",
            "plural": true,
            "selections": [
                {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "id",
                    "storageKey": null
                },
                {
                    "args": null,
                    "kind": "FragmentSpread",
                    "name": "CustomFieldRenderer_customField"
                },
                {
                    "args": null,
                    "kind": "FragmentSpread",
                    "name": "CustomFieldRenderer_customFieldInfo"
                }
            ],
            "storageKey": null
        }
    ],
    "type": "Organization",
    "abstractKey": null
} as any;
(node as any).hash = 'd15e2bf6b491ffd3ca4569c501d93ec9';
export default node;
