// @bbyca/account-components/dist/providers/CustomerAccountStatusProvider/ApiCustomerAccountStatusProvider.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as url from "url";
import { ApiCsrfTokenProvider } from "../";
import { StatusCode } from "../../errors";
import { cieUtilities } from "../../utils/cieUtilities";
import ApiBaseProvider from "../BaseProvider/ApiBaseProvider";
export class ApiCustomerAccountStatusProvider extends ApiBaseProvider {
    constructor(cieServiceUrl) {
        super(cieServiceUrl);
        this.getCustomerAccountStatus = (captchaResponse, email, locale, tid) => __awaiter(this, void 0, void 0, function* () {
            const customHeaders = {
                "Accept-Language": locale,
            };
            const lookUpCustomerUrl = url.parse(`${this.apiUrl}/lookup`);
            const formattedUrl = url.format(lookUpCustomerUrl);
            const headers = yield this.generateHttpHeaders(customHeaders);
            const response = yield fetch(formattedUrl, {
                body: JSON.stringify({
                    captchaResponse,
                    email,
                    tid,
                }),
                credentials: "include",
                headers,
                method: "POST",
            });
            if (response.status !== StatusCode.OK) {
                return this.handleError(response.status, formattedUrl);
            }
        });
        this.generateHttpHeaders = (customHeaders = {}) => __awaiter(this, void 0, void 0, function* () {
            const headers = Object.assign(Object.assign({}, customHeaders), { "Accept": "application/vnd.bestbuy+json", "Content-Type": "application/vnd.bestbuy+json" });
            yield this.addCsrfToken(headers);
            return headers;
        });
        this.addCsrfToken = (headers) => __awaiter(this, void 0, void 0, function* () {
            let csrfToken = cieUtilities.getCsrfToken();
            const csrfTokenProvider = new ApiCsrfTokenProvider(this.cieServiceUrl);
            if (!csrfToken) {
                yield csrfTokenProvider.getCsrfToken();
                csrfToken = cieUtilities.getCsrfToken();
            }
            if (csrfToken) {
                headers["X-TX"] = csrfToken;
            }
        });
        this.cieServiceUrl = cieServiceUrl;
    }
}
export default ApiCustomerAccountStatusProvider;
