// @udacity/ureact-rate-nanodegree/build/components/modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = undefined;

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

var _dec, _class, _class2, _temp2;

var _modal = require('@udacity/ureact-components/dist/modal');

var _modal2 = _interopRequireDefault(_modal);

var _veritasComponents = require('@udacity/veritas-components');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _stars = require('./stars');

var _stars2 = _interopRequireDefault(_stars);

var _i18n = require('../i18n');

var _modal3 = require('./modal.scss');

var _modal4 = _interopRequireDefault(_modal3);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var RatingModal = (_dec = cssModule(_modal4.default), _dec(_class = (_temp2 = _class2 = function(_React$Component) {
    _inherits(RatingModal, _React$Component);

    function RatingModal() {
        var _ref;

        var _temp, _this, _ret;

        _classCallCheck(this, RatingModal);

        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RatingModal.__proto__ || Object.getPrototypeOf(RatingModal)).call.apply(_ref, [this].concat(args))), _this), _this.handleClickStar = function(rating) {
            var _this$props = _this.props,
                review = _this$props.review,
                handleChangeReview = _this$props.handleChangeReview;

            handleChangeReview(rating, review);
        }, _this.handleChangeReview = function(evt) {
            var _this$props2 = _this.props,
                rating = _this$props2.rating,
                handleChangeReview = _this$props2.handleChangeReview;

            handleChangeReview(rating, evt.target.value);
        }, _this.handleSubmit = function() {
            if (_this._isRatingValid) {
                _this.props.handleSubmit();
                _this.props.handleRequestClose();
            }
        }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(RatingModal, [{
        key: '_isRatingValid',
        value: function _isRatingValid() {
            var _props = this.props,
                rating = _props.rating,
                review = _props.review;

            return rating > 3 || _.get(review, 'length');
        }
    }, {
        key: 'render',
        value: function render() {
            var _props2 = this.props,
                rating = _props2.rating,
                review = _props2.review,
                isModalOpen = _props2.isModalOpen,
                handleRequestClose = _props2.handleRequestClose;


            var modalStyle = {
                content: {
                    width: '610px',
                    top: '7%'
                }
            };

            return _react2.default.createElement(
                _modal2.default, {
                    isOpen: isModalOpen,
                    onRequestClose: handleRequestClose,
                    style: modalStyle
                },
                _react2.default.createElement(
                    'div', {
                        styleName: this._isRatingValid() ? 'container' : 'warning-container'
                    },
                    _react2.default.createElement(
                        'h3', {
                            styleName: 'title'
                        },
                        (0, _i18n.__)('Share Your Experience')
                    ),
                    _react2.default.createElement(
                        'p', {
                            styleName: 'summary'
                        },
                        (0, _i18n.__)('Help other students out by rating this Nanodegree and leaving an honest review about your learning experience so far.')
                    ),
                    _react2.default.createElement(
                        'h4', {
                            styleName: 'cta-rating'
                        },
                        (0, _i18n.__)('How would you rate this Nanodegree program?')
                    ),
                    _react2.default.createElement(_stars2.default, {
                        rating: rating,
                        syleName: 'stars',
                        onClick: this.handleClickStar
                    }),
                    _react2.default.createElement(
                        'h4', {
                            styleName: 'cta-review'
                        },
                        (0, _i18n.__)('How\'s the program going? How has it matched your needs and expectations?'),
                        rating > 3 ? null : (0, _i18n.__)(' (required)')
                    ),
                    _react2.default.createElement('textarea', {
                        styleName: 'form-text-area',
                        placeholder: (0, _i18n.__)('Write about your experience here'),
                        onChange: this.handleChangeReview,
                        value: review,
                        cols: '30',
                        rows: '10'
                    }),
                    _react2.default.createElement(
                        'p', {
                            styleName: 'disclaimer'
                        },
                        (0, _i18n.__)('Your review and first name will be shared publicly')
                    ),
                    this._isRatingValid() ? null : _react2.default.createElement(
                        'h5', {
                            styleName: 'rating-warning'
                        },
                        (0, _i18n.__)('Please tell us why to submit your review so that we can improve!')
                    ),
                    _react2.default.createElement(_veritasComponents.Button, {
                        label: 'Submit Review',
                        onClick: this.handleSubmit,
                        variant: 'primary',
                        disabled: !rating || !this._isRatingValid()
                    }),
                    _react2.default.createElement('div', {
                        styleName: 'icon'
                    })
                )
            );
        }
    }]);

    return RatingModal;
}(_react2.default.Component), _class2.displayName = 'rate-nanodegree/modal', _class2.propTypes = {
    isModalOpen: _propTypes2.default.bool.isRequired,
    handleSubmit: _propTypes2.default.func.isRequired,
    handleRequestClose: _propTypes2.default.func.isRequired,
    handleChangeReview: _propTypes2.default.func.isRequired,
    rating: _propTypes2.default.number,
    review: _propTypes2.default.string
}, _temp2)) || _class);
exports.default = RatingModal;
