// @udacity/ureact-rate-nanodegree/build/components/rating.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = undefined;

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

var _dec, _class, _class2, _temp;

var _modal = require('./modal');

var _modal2 = _interopRequireDefault(_modal);

var _stars = require('./stars');

var _stars2 = _interopRequireDefault(_stars);

var _i18n = require('../i18n');

var _veritasComponents = require('@udacity/veritas-components');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _propTypes3 = require('./prop-types');

var _propTypes4 = _interopRequireDefault(_propTypes3);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _rating = require('./rating.scss');

var _rating2 = _interopRequireDefault(_rating);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var RateNanodegree = (_dec = cssModule(_rating2.default, {
    allowMultiple: true
}), _dec(_class = (_temp = _class2 = function(_React$Component) {
    _inherits(RateNanodegree, _React$Component);

    function RateNanodegree(props) {
        _classCallCheck(this, RateNanodegree);

        var _this = _possibleConstructorReturn(this, (RateNanodegree.__proto__ || Object.getPrototypeOf(RateNanodegree)).call(this, props));

        _this.onChange = function(rating, review) {
            _this.setState({
                rating: rating,
                review: review
            });
        };

        _this.handleClickOpenModal = function() {
            _this.setState({
                isModalOpen: true
            });
        };

        _this.onClickOpenModal = function(rating) {
            _this.setState({
                isModalOpen: true,
                rating: rating
            });
        };

        _this.onClickCloseModal = function() {
            _this.setState({
                isModalOpen: false
            });
        };

        _this.onSubmit = function() {
            var _this$state = _this.state,
                rating = _this$state.rating,
                review = _this$state.review;
            var _this$props = _this.props,
                ndKey = _this$props.ndKey,
                userKey = _this$props.userKey,
                userName = _this$props.userName,
                onSubmit = _this$props.onSubmit,
                locale = _this$props.locale;


            return onSubmit({
                ndKey: ndKey,
                userKey: userKey,
                userName: userName,
                rating: rating,
                review: review,
                locale: locale
            });
        };

        _this.state = {
            isModalOpen: false
        };
        return _this;
    }

    _createClass(RateNanodegree, [{
        key: '_renderRating',
        value: function _renderRating() {
            var _state = this.state,
                isModalOpen = _state.isModalOpen,
                rating = _state.rating,
                review = _state.review;
            var _props = this.props,
                styleName = _props.styleName,
                ndRatingResponse = _props.ndRatingResponse;


            return _react2.default.createElement(
                'div', {
                    styleName: styleName
                }, !_.get(ndRatingResponse, 'rating') ? _react2.default.createElement(
                    'div', {
                        styleName: 'content'
                    },
                    _react2.default.createElement(
                        'h5', {
                            styleName: 'cta',
                            onClick: this.handleClickOpenModal
                        },
                        (0, _i18n.__)('Rate this program')
                    ),
                    _react2.default.createElement(_stars2.default, {
                        color: _stars.RatingColors.WHITE,
                        'static': true,
                        rating: rating,
                        onClick: this.onClickOpenModal.bind(this)
                    })
                ) : null,
                _react2.default.createElement(_modal2.default, {
                    rating: rating,
                    review: review,
                    isModalOpen: isModalOpen,
                    handleChangeReview: this.onChange,
                    handleSubmit: this.onSubmit,
                    handleRequestClose: this.onClickCloseModal
                })
            );
        }
    }, {
        key: 'render',
        value: function render() {
            var isFetching = this.props.isFetching;


            if (isFetching) {
                return _react2.default.createElement(
                    'div', {
                        styleName: 'content'
                    },
                    _react2.default.createElement(_veritasComponents.Loading, null)
                );
            } else {
                return this._renderRating();
            }
        }
    }]);

    return RateNanodegree;
}(_react2.default.Component), _class2.displayName = 'rate-nanodegree', _class2.propTypes = {
    ndKey: _propTypes2.default.string.isRequired,
    userKey: _propTypes2.default.string.isRequired,
    userName: _propTypes2.default.string,
    locale: _propTypes2.default.string,
    isFetching: _propTypes2.default.bool,
    ndRatingResponse: _propTypes4.default.ndRatingResponse,
    onSubmit: _propTypes2.default.func.isRequired
}, _class2.defaultProps = {
    userName: null,
    locale: null,
    isFetching: false,
    ndRatingResponse: null
}, _temp)) || _class);
exports.default = RateNanodegree;
