// @udacity/ureact-hoth/build/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _values = require('babel-runtime/core-js/object/values');

var _values2 = _interopRequireDefault(_values);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _jsCookie = require('js-cookie');

var _jsCookie2 = _interopRequireDefault(_jsCookie);

var _jwtDecode = require('jwt-decode');

var _jwtDecode2 = _interopRequireDefault(_jwtDecode);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

var JWT_COOKIE_NAME = '_jwt';
var JWT_COOKIE_DOMAIN = '.udacity.com';

var AUTH_URLS = {
    staging: 'https://auth-staging.udacity.com',
    production: 'https://auth.udacity.com'
};

var AUTH_TYPES = {
    signIn: '/sign-in',
    signUp: '/sign-up',
    default: '/'
};

var AuthService = function() {
    function AuthService() {
        (0, _classCallCheck3.default)(this, AuthService);
    }

    (0, _createClass3.default)(AuthService, [{
        key: '_redirect',
        value: function _redirect(href) {
            window.location.href = href;
        }
    }, {
        key: 'getJWT',
        value: function getJWT() {
            return _jsCookie2.default.get(JWT_COOKIE_NAME);
        }
    }, {
        key: 'isAuthenticated',
        value: function isAuthenticated() {
            return !!this.getJWT();
        }
    }, {
        key: 'getCurrentUser',
        value: function getCurrentUser() {
            var jwt = this.getJWT();
            return jwt ? (0, _jwtDecode2.default)(jwt) : {};
        }
    }, {
        key: 'getCurrentUserId',
        value: function getCurrentUserId() {
            var user = this.getCurrentUser();
            return user.uid;
        }
    }, {
        key: 'signIn',
        value: function signIn(_ref) {
            var env = _ref.env,
                returnUrl = _ref.returnUrl;

            this.authenticate({
                env: env,
                returnUrl: returnUrl,
                authType: AUTH_TYPES.signIn
            });
        }
    }, {
        key: 'signUp',
        value: function signUp(_ref2) {
            var env = _ref2.env,
                returnUrl = _ref2.returnUrl;

            this.authenticate({
                env: env,
                returnUrl: returnUrl,
                authType: AUTH_TYPES.signUp
            });
        }
    }, {
        key: 'authenticate',
        value: function authenticate(_ref3) {
            var env = _ref3.env,
                returnUrl = _ref3.returnUrl,
                _ref3$authType = _ref3.authType,
                authType = _ref3$authType === undefined ? AUTH_TYPES.default : _ref3$authType;

            var authUrl = AUTH_URLS[env];
            (0, _invariant2.default)(authUrl, '\'' + env + '\' is not a valid \'env\'');
            (0, _invariant2.default)((0, _values2.default)(AUTH_TYPES).includes(authType), '\'' + authType + '\' must be one of ' + (0, _values2.default)(AUTH_TYPES));
            (0, _invariant2.default)(returnUrl, 'Must specify a `returnUrl`');

            this._redirect('' + authUrl + authType + '?next=' + encodeURIComponent(returnUrl));
        }
    }, {
        key: 'signOut',
        value: function signOut() {
            this._redirect('https://www.udacity.com/account/logout');
        }
    }]);
    return AuthService;
}();

exports.default = new AuthService();
