// @postman/aether/esmLib/components/Illustration/spot-illustrations/Authentication.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function Authentication(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M56.7001 17.31V12.71H42.4301V17.31",
    fill: "white"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M56.7001 17.31V12.71H42.4301V17.31",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M72.95 80.93H26.25C24.48 80.93 23.04 79.49 23.04 77.72V20.7C23.04 18.93 24.48 17.49 26.25 17.49H72.95C74.72 17.49 76.16 18.93 76.16 20.7V77.72C76.16 79.49 74.72 80.93 72.95 80.93Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.15 66.6501H11.35C8.88 66.6501 6.88 64.6501 6.88 62.1801V47.0401C6.88 44.5701 8.88 42.5701 11.35 42.5701H27.14C29.61 42.5701 31.61 44.5701 31.61 47.0401V62.1901C31.61 64.6501 29.61 66.6501 27.15 66.6501Z",
    fill: "white",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M58.51 46.6899C57.21 43.6999 54.23 41.6099 50.76 41.6099C46.9 41.6099 43.65 44.1899 42.64 47.7199C42.64 47.7199 48.1 52.2399 54.79 49.3099L58.51 46.6899Z",
    fill: "#FFA165"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M58.51 46.6899C57.21 43.6999 54.23 41.6099 50.76 41.6099C46.9 41.6099 43.65 44.1899 42.64 47.7199",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M50.28 50.2799C56.935 50.2799 62.33 44.885 62.33 38.2299C62.33 31.5749 56.935 26.1799 50.28 26.1799C43.6249 26.1799 38.23 31.5749 38.23 38.2299C38.23 44.885 43.6249 50.2799 50.28 50.2799Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M50.1199 38.12C52.2904 38.12 54.0499 36.3605 54.0499 34.19C54.0499 32.0195 52.2904 30.26 50.1199 30.26C47.9495 30.26 46.1899 32.0195 46.1899 34.19C46.1899 36.3605 47.9495 38.12 50.1199 38.12Z",
    fill: "#FFA466",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M76.53 45.95C82.56 45.95 89.04 50.8399 89.04 56.8699V58.54C94.97 58.54 99.77 63.34 99.77 69.27V69.82",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M17.17 50.6301C20.94 50.6401 24.15 53.4201 24.76 57.2201C24.84 57.7501 24.85 58.2901 24.86 58.8301C24.86 59.0101 24.79 59.2201 24.71 59.3801C24.56 59.6701 24.37 59.9301 24.2 60.2101C24.11 60.3501 23.99 60.4301 23.83 60.3601C23.66 60.2901 23.63 60.1501 23.64 59.9701C23.7 59.1201 23.75 58.2701 23.62 57.4201C23.25 54.8501 21.25 52.6101 18.73 52.0301C16.25 51.4601 14.13 52.1401 12.36 54.0001C12.29 54.0801 12.19 54.1601 12.09 54.1801C11.93 54.2301 11.76 54.0901 11.77 53.9301C11.79 53.6101 11.8 53.2901 11.89 52.9801C11.94 52.7901 12.08 52.6101 12.23 52.4801C13.47 51.4101 14.91 50.8101 16.53 50.6701C16.74 50.6501 16.95 50.6401 17.17 50.6301Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M22.53 59.0299C22.43 59.8399 22.32 60.7399 22.19 61.6299C22.09 62.2899 21.67 62.5199 21.15 62.6999C20.94 62.7699 20.77 62.5599 20.82 62.3199C20.94 61.7299 21.1 61.1299 21.18 60.5299C21.31 59.5899 21.4 58.6299 21.24 57.6799C20.94 55.8999 19.44 54.4799 17.67 54.2799C15.42 54.0199 13.46 55.5199 13.06 57.7899C13.04 57.9199 13.03 58.0499 13.01 58.1799C12.99 58.3299 12.91 58.4199 12.76 58.4499C12.62 58.4699 12.5 58.4099 12.45 58.2699C12.32 57.8999 12.18 57.5399 12.1 57.1599C12.06 56.9699 12.1 56.7399 12.17 56.5599C12.99 54.5299 14.47 53.3199 16.62 53.0899C19.41 52.7899 21.86 54.6899 22.38 57.4999C22.49 57.9699 22.49 58.4599 22.53 59.0299Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M17.1799 48.22C20.6999 48.33 23.4899 49.82 25.5199 52.77C25.7999 53.17 25.9399 53.62 25.9399 54.11C25.9399 54.34 25.9399 54.57 25.9399 54.81C25.9399 55.05 25.8599 55.17 25.6999 55.19C25.4999 55.22 25.4199 55.09 25.3499 54.93C24.9499 53.99 24.4199 53.13 23.7399 52.38C22.1599 50.65 20.2199 49.67 17.9099 49.46C16.2499 49.31 14.6799 49.63 13.1999 50.38C13.0499 50.46 12.9099 50.5 12.7799 50.36C12.6599 50.23 12.6899 50.09 12.7599 49.94C13.0699 49.3 13.5499 48.88 14.2199 48.67C15.1699 48.38 16.1699 48.23 17.1799 48.22Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M20.14 58.9701C20.07 60.2301 19.88 61.4701 19.45 62.6601C19.29 63.0901 19.15 63.1901 18.69 63.1901C18.6 63.1901 18.52 63.1901 18.43 63.1801C18.18 63.1601 18.06 62.9701 18.16 62.7301C18.73 61.3801 18.94 59.9601 18.94 58.5101C18.94 57.4801 18.15 56.6701 17.16 56.6701C16.17 56.6701 15.39 57.4801 15.38 58.5001C15.38 59.3801 15.29 60.2501 14.97 61.0901C14.83 61.4701 14.63 61.5101 14.34 61.2201C14.25 61.1301 14.16 61.0401 14.08 60.9501C13.95 60.7901 13.9 60.6101 13.96 60.4101C14.14 59.7801 14.19 59.1401 14.2 58.4901C14.2 56.9101 15.37 55.6301 16.93 55.4801C18.41 55.3401 19.8 56.4301 20.07 57.9501C20.13 58.2701 20.12 58.6201 20.14 58.9701Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M17.1699 47C16.7899 47.02 16.4099 47.05 16.0299 47.08C15.8599 47.09 15.7299 47.04 15.6799 46.88C15.6199 46.71 15.7099 46.59 15.8499 46.51C16.7099 46.01 17.5999 45.74 18.6099 45.9C19.7999 46.09 20.9499 46.37 22.0599 46.84C23.0699 47.27 23.7899 48.03 24.3799 48.93C24.4699 49.07 24.4599 49.21 24.3399 49.32C24.2099 49.44 24.0699 49.42 23.9399 49.31C23.2299 48.77 22.4599 48.31 21.6399 47.95C20.1999 47.32 18.7199 47.02 17.1699 47Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M17.7599 59.0102C17.7099 60.2002 17.4899 61.3502 16.9899 62.4302C16.8399 62.7602 16.7299 62.8002 16.4099 62.6502C16.2799 62.5902 16.1499 62.5302 16.0299 62.4602C15.8399 62.3602 15.7899 62.2102 15.8799 62.0202C16.3899 60.9602 16.5699 59.8302 16.5799 58.6602C16.5799 58.5802 16.5799 58.5102 16.5799 58.4302C16.5999 58.1002 16.8599 57.8502 17.1799 57.8602C17.4899 57.8702 17.7399 58.1102 17.7599 58.4302C17.7599 58.6302 17.7599 58.8202 17.7599 59.0102Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M62.96 56.49H38.63V60.71H62.96V56.49Z",
    fill: "white",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M60.9 65.49H40.7V69.71H60.9V65.49Z",
    fill: "white",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M6.89 30.0702L8.26 30.8402L6.89 31.6102L6.13 33.0202L5.38 31.6102L4 30.8402L5.38 30.0702L6.13 28.6602L6.89 30.0702Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M15.5701 75.38L16.9501 76.15L15.5701 76.92L14.8201 78.33L14.0601 76.92L12.6801 76.15L14.0601 75.38L14.8201 73.97L15.5701 75.38Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M92.1299 11.78L96.5199 13.75L92.0399 15.49L90.0699 19.88L88.3299 15.4L83.9399 13.43L88.4199 11.69L90.3899 7.30005L92.1299 11.78Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "96.27",
    height: "75.43",
    fill: "white",
    transform: "translate(4 6)"
  }))));
}

export default withTheme(Authentication);
