// @postman/aether/esmLib/components/Illustration/spot-illustrations/AutoFlexPay.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function AutoFlexPay(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M85.12 69.54C91.83 69.54 97.26 74.98 97.26 81.68",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M58.45 39.6099C57.31 36.2799 54.16 33.8899 50.45 33.8899C46.48 33.8899 43.16 36.6199 42.25 40.3099C42.25 40.3099 49.47 44.5599 56.02 40.8899",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M69.36 80.5899C69.36 80.5899 82.59 79.9799 83.75 63.6899C83.75 63.6899 84.95 66.8099 84.79 70.4399C84.61 74.3799 82.89 78.9299 77.99 80.7199C75.77 81.5299 73.33 81.1199 69.36 80.5899Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M31.64 74.55C31.64 74.55 33.97 78.92 28.29 81.17C28.29 81.17 34.53 80.93 35.04 77.36L31.64 74.55Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M42.24 40.3099C43.15 36.6299 46.48 33.8899 50.44 33.8899C54.15 33.8899 57.3 36.2799 58.44 39.6099",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M49.97 42.57C56.625 42.57 62.02 37.175 62.02 30.52C62.02 23.8649 56.625 18.47 49.97 18.47C43.315 18.47 37.92 23.8649 37.92 30.52C37.92 37.175 43.315 42.57 49.97 42.57Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M49.97 42.57C56.625 42.57 62.02 37.175 62.02 30.52C62.02 23.8649 56.625 18.47 49.97 18.47C43.315 18.47 37.92 23.8649 37.92 30.52C37.92 37.175 43.315 42.57 49.97 42.57Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M50.08 30.91C52.2504 30.91 54.01 29.1505 54.01 26.98C54.01 24.8096 52.2504 23.05 50.08 23.05C47.9095 23.05 46.15 24.8096 46.15 26.98C46.15 29.1505 47.9095 30.91 50.08 30.91Z",
    fill: "#FDA968",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M19.35 79.3099C20.27 75.6399 23.59 72.9099 27.55 72.9099C31.2 72.9099 34.3 75.2199 35.49 78.4499",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.07 81.59C33.725 81.59 39.12 76.195 39.12 69.54C39.12 62.885 33.725 57.49 27.07 57.49C20.415 57.49 15.02 62.885 15.02 69.54C15.02 76.195 20.415 81.59 27.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.07 81.59C33.725 81.59 39.12 76.195 39.12 69.54C39.12 62.885 33.725 57.49 27.07 57.49C20.415 57.49 15.02 62.885 15.02 69.54C15.02 76.195 20.415 81.59 27.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.18 69.9301C29.3505 69.9301 31.11 68.1705 31.11 66.0001C31.11 63.8296 29.3505 62.0701 27.18 62.0701C25.0095 62.0701 23.25 63.8296 23.25 66.0001C23.25 68.1705 25.0095 69.9301 27.18 69.9301Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.07 81.59C79.725 81.59 85.12 76.195 85.12 69.54C85.12 62.885 79.725 57.49 73.07 57.49C66.415 57.49 61.02 62.885 61.02 69.54C61.02 76.195 66.415 81.59 73.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.07 81.59C79.725 81.59 85.12 76.195 85.12 69.54C85.12 62.885 79.725 57.49 73.07 57.49C66.415 57.49 61.02 62.885 61.02 69.54C61.02 76.195 66.415 81.59 73.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M33.41 59.46L44.81 41.76",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M67.02 58.9499L55.17 41.4199",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M38.93 71.1599H61.17",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M11.3001 32.7099L12.3101 33.2699L11.3001 33.8299L10.7501 34.8699L10.1901 33.8299L9.18005 33.2699L10.1901 32.7099L10.7501 31.6699L11.3001 32.7099Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M37.41 29.51H36.04C26.46 29.51 18.69 37.28 18.69 46.86L18.66 46.89C12.24 46.9 7.5 52.07 7.5 58.49",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M86.7 48.3401L87.71 48.9101L86.7 49.4701L86.15 50.5001L85.6 49.4701L84.59 48.9101L85.6 48.3401L86.15 47.3101L86.7 48.3401Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M84.1853 16.3947C85.9423 13.8497 85.355 10.3977 82.8736 8.68466C80.3921 6.97157 76.9562 7.64601 75.1992 10.1911C73.4422 12.7361 74.0294 16.1881 76.5109 17.9011C78.9923 19.6142 82.4282 18.9398 84.1853 16.3947Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M85.31 14.2201C87.73 16.3701 89.08 18.2101 88.6 18.9201C87.88 19.9701 83.36 18.1001 78.51 14.7501C73.66 11.4001 70.31 7.84011 71.03 6.79011C71.52 6.09011 73.73 6.70011 76.6 8.21011",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.42 75.2601C73.31 75.2601 73.22 75.2301 73.15 75.1701C73.08 75.1101 73.04 75.0401 73.04 74.9601V74.4801C72.43 74.4701 71.9 74.3701 71.46 74.1801C71.02 73.9901 70.68 73.7701 70.44 73.5201C70.2 73.2701 70.08 73.0401 70.08 72.8401C70.08 72.7401 70.11 72.6401 70.16 72.5401C70.21 72.4401 70.28 72.3501 70.36 72.2801C70.44 72.2101 70.53 72.1701 70.61 72.1701C70.72 72.1701 70.83 72.2301 70.94 72.3401C71.05 72.4501 71.19 72.5801 71.36 72.7201C71.52 72.8701 71.74 73.0001 72.01 73.1201C72.28 73.2401 72.63 73.3101 73.04 73.3301V69.4601C72.71 69.3501 72.39 69.2301 72.07 69.1001C71.75 68.9701 71.46 68.8001 71.2 68.6001C70.93 68.4001 70.72 68.1501 70.57 67.8401C70.41 67.5301 70.34 67.1401 70.34 66.6801C70.34 66.2201 70.42 65.8401 70.57 65.5201C70.72 65.2101 70.93 64.9501 71.19 64.7501C71.45 64.5501 71.74 64.4001 72.06 64.2901C72.38 64.1801 72.7 64.1201 73.03 64.0901V63.6501C73.03 63.5601 73.07 63.4901 73.14 63.4301C73.21 63.3701 73.3 63.3401 73.41 63.3401C73.5 63.3401 73.58 63.3701 73.66 63.4301C73.74 63.4901 73.78 63.5601 73.78 63.6501V64.0601C74.03 64.0701 74.29 64.1001 74.56 64.1501C74.83 64.2001 75.07 64.2601 75.3 64.3401C75.53 64.4201 75.71 64.5101 75.85 64.6301C75.99 64.7501 76.06 64.8801 76.06 65.0201C76.06 65.1001 76.04 65.1901 76 65.2901C75.96 65.3901 75.9 65.4801 75.83 65.5701C75.75 65.6501 75.66 65.6901 75.54 65.6901C75.44 65.6901 75.32 65.6501 75.18 65.5701C75.04 65.4901 74.86 65.4001 74.64 65.3201C74.42 65.2301 74.13 65.1801 73.78 65.1601V68.4701C74.08 68.5801 74.38 68.7101 74.69 68.8501C74.99 69.0001 75.27 69.1901 75.52 69.4101C75.77 69.6301 75.96 69.9201 76.11 70.2601C76.26 70.6001 76.33 71.0201 76.33 71.5201C76.33 72.4001 76.09 73.0801 75.62 73.5601C75.15 74.0401 74.54 74.3301 73.79 74.4401V74.9601C73.79 75.0401 73.75 75.1101 73.67 75.1701C73.59 75.2301 73.51 75.2601 73.42 75.2601ZM73.1 68.2101V65.1601C72.81 65.1901 72.54 65.2501 72.3 65.3601C72.06 65.4601 71.86 65.6101 71.72 65.8001C71.57 65.9901 71.5 66.2501 71.5 66.5601C71.5 66.8901 71.57 67.1601 71.72 67.3701C71.86 67.5801 72.06 67.7501 72.3 67.8701C72.54 67.9901 72.81 68.1101 73.1 68.2101ZM73.74 73.3001C74.2 73.2201 74.55 73.0501 74.8 72.7701C75.05 72.4901 75.18 72.1101 75.18 71.6301C75.18 71.2601 75.11 70.9601 74.98 70.7201C74.85 70.4801 74.68 70.2901 74.46 70.1301C74.24 69.9801 74 69.8401 73.74 69.7401V73.3001Z",
    fill: illustrationStrokeColor
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "42.2432",
    y1: "38.1372",
    x2: "58.4472",
    y2: "38.1372",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "90.77",
    height: "76.19",
    fill: "white",
    transform: "translate(7 6)"
  }))));
}

export default withTheme(AutoFlexPay);
