// @postman/aether/esmLib/components/Illustration/spot-illustrations/Categories.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function Categories(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M21.68 7.83009L26.07 9.79009L21.59 11.5301L19.62 15.9301L17.88 11.4401L13.49 9.48009L17.97 7.73009L19.94 3.34009L21.68 7.83009Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M84.57 23.3601H24.44C22.01 23.3601 20.04 25.3301 20.04 27.7601V69.2301C20.04 71.6602 22.01 73.6301 24.44 73.6301H84.57C87.0001 73.6301 88.97 71.6602 88.97 69.2301V27.7601C88.97 25.3301 87.0001 23.3601 84.57 23.3601Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M51.17 35.02H26.84V39.24H51.17V35.02Z",
    fill: "white",
    stroke: illustrationStrokeColor,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M81.01 44.02H26.84V48.24H81.01V44.02Z",
    fill: "white",
    stroke: illustrationStrokeColor,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M70.43 56.02H29.67C28.107 56.02 26.84 57.2871 26.84 58.85V63.36C26.84 64.923 28.107 66.19 29.67 66.19H70.43C71.9929 66.19 73.26 64.923 73.26 63.36V58.85C73.26 57.2871 71.9929 56.02 70.43 56.02Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M70.43 56.02H29.67C28.107 56.02 26.84 57.2871 26.84 58.85V63.36C26.84 64.923 28.107 66.19 29.67 66.19H70.43C71.9929 66.19 73.26 64.923 73.26 63.36V58.85C73.26 57.2871 71.9929 56.02 70.43 56.02Z",
    stroke: illustrationStrokeColor,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27 23.3101V20.7601C27 19.5932 27.4636 18.474 28.2887 17.6488C29.1139 16.8237 30.233 16.3601 31.4 16.3601H91.56C92.727 16.3601 93.8461 16.8237 94.6713 17.6488C95.4964 18.474 95.96 19.5932 95.96 20.7601V62.2401C95.96 63.4071 95.4964 64.5262 94.6713 65.3514C93.8461 66.1765 92.727 66.6401 91.56 66.6401H89.16",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M1.32998 74.4101C1.14998 69.2801 5.11998 63.6201 10.24 63.4101L11.66 63.3601C11.51 59.1201 14.86 54.4401 20 54.3601",
    stroke: illustrationStrokeColor,
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M12.1 44.0599L13.47 44.8299L12.1 45.5999L11.34 47.0099L10.59 45.5999L9.20996 44.8299L10.59 44.0599L11.34 42.6499L12.1 44.0599Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M97.1 9.0599L98.47 9.8299L97.1 10.5999L96.34 12.0099L95.59 10.5999L94.21 9.8299L95.59 9.0599L96.34 7.6499L97.1 9.0599Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "26.84",
    y1: "61.11",
    x2: "73.26",
    y2: "61.11",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  }))));
}

export default withTheme(Categories);
