// @postman/aether/esmLib/components/Illustration/spot-illustrations/Collaborate.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function Collaborate(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M57.63 39.5901C60.11 37.1701 53.94 33.8201 50.27 33.8201C46.36 33.8201 43.08 36.4701 42.11 40.0701C42.11 40.0801 51.98 45.1001 57.63 39.5901Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M77.78 74.2698C77.78 74.2698 79.97 78.9198 74.29 81.1698C74.29 81.1698 80.56 81.1098 81.07 77.5398L77.78 74.2698Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M31.81 74.21C31.81 74.21 33.97 78.91 28.29 81.17C28.29 81.17 34.57 81.03 35.08 77.47L31.81 74.21Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M42.29 40.1499C43.26 36.5499 46.55 33.8999 50.45 33.8999C54.11 33.8999 57.23 36.2299 58.41 39.4899",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M49.97 42.57C56.6251 42.57 62.02 37.175 62.02 30.52C62.02 23.8649 56.6251 18.47 49.97 18.47C43.315 18.47 37.92 23.8649 37.92 30.52C37.92 37.175 43.315 42.57 49.97 42.57Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M49.97 42.57C56.6251 42.57 62.02 37.175 62.02 30.52C62.02 23.8649 56.6251 18.47 49.97 18.47C43.315 18.47 37.92 23.8649 37.92 30.52C37.92 37.175 43.315 42.57 49.97 42.57Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M50.08 30.91C52.2505 30.91 54.01 29.1505 54.01 26.98C54.01 24.8096 52.2505 23.05 50.08 23.05C47.9095 23.05 46.15 24.8096 46.15 26.98C46.15 29.1505 47.9095 30.91 50.08 30.91Z",
    fill: "#FDA767",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M19.41 79.0899C20.4 75.5299 23.67 72.9199 27.55 72.9199C31.24 72.9199 34.38 75.2899 35.53 78.5899",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.07 81.59C33.725 81.59 39.12 76.195 39.12 69.54C39.12 62.885 33.725 57.49 27.07 57.49C20.415 57.49 15.02 62.885 15.02 69.54C15.02 76.195 20.415 81.59 27.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.07 81.59C33.725 81.59 39.12 76.195 39.12 69.54C39.12 62.885 33.725 57.49 27.07 57.49C20.415 57.49 15.02 62.885 15.02 69.54C15.02 76.195 20.415 81.59 27.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M27.18 69.9298C29.3505 69.9298 31.11 68.1703 31.11 65.9998C31.11 63.8293 29.3505 62.0698 27.18 62.0698C25.0095 62.0698 23.25 63.8293 23.25 65.9998C23.25 68.1703 25.0095 69.9298 27.18 69.9298Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M65.39 79.1399C66.37 75.5499 69.65 72.9199 73.54 72.9199C77.22 72.9199 80.35 75.2699 81.51 78.5499",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.07 81.59C79.725 81.59 85.12 76.195 85.12 69.54C85.12 62.885 79.725 57.49 73.07 57.49C66.415 57.49 61.02 62.885 61.02 69.54C61.02 76.195 66.415 81.59 73.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.07 81.59C79.725 81.59 85.12 76.195 85.12 69.54C85.12 62.885 79.725 57.49 73.07 57.49C66.415 57.49 61.02 62.885 61.02 69.54C61.02 76.195 66.415 81.59 73.07 81.59Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.18 69.9298C75.3505 69.9298 77.11 68.1703 77.11 65.9998C77.11 63.8293 75.3505 62.0698 73.18 62.0698C71.0095 62.0698 69.25 63.8293 69.25 65.9998C69.25 68.1703 71.0095 69.9298 73.18 69.9298Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M33.41 59.4598L44.81 41.7598",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M67.02 58.9498L55.09 41.3298",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M38.9301 71.1599H61.1701",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M79.16 43.2299L80.17 43.79L79.16 44.36L78.6 45.3899L78.05 44.36L77.04 43.79L78.05 43.2299L78.6 42.2L79.16 43.2299Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M11.3001 32.7099L12.3101 33.2699L11.3001 33.8299L10.7501 34.8699L10.1901 33.8299L9.18005 33.2699L10.1901 32.7099L10.7501 31.6699L11.3001 32.7099Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M37.41 29.5098H36.04C26.46 29.5098 18.69 37.2798 18.69 46.8598L18.66 46.8898C12.24 46.8998 7.5 52.0698 7.5 58.4898",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M85.12 69.5398C91.83 69.5398 97.26 74.9798 97.26 81.6798",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M84.1853 15.3947C85.9423 12.8497 85.355 9.39775 82.8736 7.68466C80.3921 5.97157 76.9562 6.64601 75.1992 9.19107C73.4422 11.7361 74.0294 15.1881 76.5109 16.9011C78.9923 18.6142 82.4282 17.9398 84.1853 15.3947Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M85.31 13.2199C87.73 15.3699 89.08 17.2099 88.6 17.9199C87.88 18.9699 83.36 17.0999 78.51 13.7499C73.66 10.3999 70.31 6.83987 71.03 5.78987C71.52 5.08987 73.73 5.69987 76.6 7.20987",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "42.11",
    y1: "37.9998",
    x2: "58.2027",
    y2: "37.9998",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "90.77",
    height: "77.19",
    fill: "white",
    transform: "translate(7 5)"
  }))));
}

export default withTheme(Collaborate);
