// @postman/aether/esmLib/components/Illustration/spot-illustrations/EasyManagement.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function EasyManagement(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M15.39 39.07L10.44 39.43L11.12 38.97C12.51 38.76 13.3 38.26 14.12 37.57C15.08 36.75 15.07 36.16 15.32 34.53L15.39 32.22V39.07Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M64.19 39.48C64.19 39.48 66.67 37.63 66.91 32.96C66.91 32.96 71.34 35.97 65.64 40.16C65.64 40.16 65.19 40.82 64.19 39.48Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M52.62 57.1301C51.55 53.6801 48.68 51.8801 44.89 51.8801C40.88 51.8801 37.71 53.9501 36.84 57.6901L43.57 60.1401L48.9 59.2201L52.62 57.1301Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M67.64 39.2801C68.49 41.7601 68.95 44.4201 68.95 47.1901C68.95 60.6401 58.05 71.5401 44.6 71.5401C31.15 71.5401 20.25 60.6401 20.25 47.1901C20.25 33.7401 31.15 22.8401 44.6 22.8401C52.56 22.8401 59.63 26.6601 64.07 32.5601",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M13.31 31.2001C13.4 31.0301 13.48 30.8601 13.57 30.7001C19.48 19.6101 31.16 12.05 44.6 12.05C64.01 12.05 79.74 27.78 79.74 47.19C79.74 64.15 67.72 78.3 51.74 81.6C51.57 81.64 51.39 81.67 51.22 81.7",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M10.27 39.71C10.22 39.95 10.17 40.2 10.12 40.45C9.69997 42.63 9.46997 44.89 9.46997 47.19C9.46997 65.89 24.08 81.18 42.51 82.26L43.04 82.37",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M36.65 58.01C37.52 54.26 40.87 51.47 44.88 51.47C48.68 51.47 51.89 53.97 52.96 57.42",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M44.41 60.1501C51.065 60.1501 56.46 54.7551 56.46 48.1C56.46 41.445 51.065 36.05 44.41 36.05C37.755 36.05 32.36 41.445 32.36 48.1C32.36 54.7551 37.755 60.1501 44.41 60.1501Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M44.41 60.1501C51.065 60.1501 56.46 54.7551 56.46 48.1C56.46 41.445 51.065 36.05 44.41 36.05C37.755 36.05 32.36 41.445 32.36 48.1C32.36 54.7551 37.755 60.1501 44.41 60.1501Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M44.52 48.4901C46.6905 48.4901 48.45 46.7306 48.45 44.5601C48.45 42.3896 46.6905 40.6301 44.52 40.6301C42.3495 40.6301 40.59 42.3896 40.59 44.5601C40.59 46.7306 42.3495 48.4901 44.52 48.4901Z",
    fill: "#FFA266",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M64.84 39.9901C67.0105 39.9901 68.77 38.2306 68.77 36.0601C68.77 33.8897 67.0105 32.1301 64.84 32.1301C62.6695 32.1301 60.91 33.8897 60.91 36.0601C60.91 38.2306 62.6695 39.9901 64.84 39.9901Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M14.23 75.15L15.6 75.92L14.23 76.69L13.47 78.1L12.72 76.69L11.34 75.92L12.72 75.15L13.47 73.74L14.23 75.15Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M90.53 45.36L91.91 46.13L90.53 46.9L89.78 48.31L89.03 46.9L87.65 46.13L89.03 45.36L89.78 43.95L90.53 45.36Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M14.67 39.5501H8.51C7.95 39.5501 7.5 39.1001 7.5 38.5401V32.3901C7.5 31.8301 7.95 31.3801 8.51 31.3801H14.66C15.22 31.3801 15.67 31.8301 15.67 32.3901V38.5401C15.68 39.0901 15.23 39.5501 14.67 39.5501Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M79.65 48.26C81.66 50.65 82.87 53.73 82.87 57.1V59.19C90.33 59.19 96.37 65.23 96.37 72.69V73.38",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M84.86 7.78005L89.25 9.75005L84.77 11.49L82.8 15.88L81.06 11.4L76.67 9.43005L81.15 7.69005L83.12 3.30005L84.86 7.78005Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M46.41 85.46C46.41 85.46 49.11 83.39 49.35 78.72C49.35 78.72 53.78 81.73 48.08 85.92C48.08 85.93 47.23 86.04 46.41 85.46Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M47 85.7001C49.1705 85.7001 50.93 83.9406 50.93 81.7701C50.93 79.5996 49.1705 77.8401 47 77.8401C44.8295 77.8401 43.07 79.5996 43.07 81.7701C43.07 83.9406 44.8295 85.7001 47 85.7001Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "36.838",
    y1: "56.0142",
    x2: "52.6196",
    y2: "56.0142",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "89.88",
    height: "84.2",
    fill: "white",
    transform: "translate(7 2)"
  }))));
}

export default withTheme(EasyManagement);
