// @postman/aether/esmLib/components/Illustration/spot-illustrations/FreeForever.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function FreeForever(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M56.74 47.1801C56.74 44.2501 54.44 41.8701 51.6 41.8701C50.17 41.8701 48.88 42.4701 47.95 43.4401C47.02 42.4701 45.73 41.8701 44.3 41.8701C41.46 41.8701 39.16 44.2501 39.16 47.1801C39.16 48.7901 39.87 50.2201 40.97 51.2001L40.95 51.2201L48.21 58.3001L55.52 50.6801L55.47 50.6401C56.26 49.6901 56.74 48.5001 56.74 47.1801Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M56.74 46.5699C56.74 43.7299 54.44 41.4299 51.6 41.4299C50.17 41.4299 48.88 42.0099 47.95 42.9499C47.02 42.0099 45.73 41.4299 44.3 41.4299C41.46 41.4299 39.16 43.7299 39.16 46.5699C39.16 48.1299 39.87 49.5199 40.97 50.4599L40.95 50.4799L48.21 57.3399L55.47 49.9099C56.26 48.9999 56.74 47.8499 56.74 46.5699Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M15.37 37.82C20.09 24.52 32.78 15 47.7 15C51.44 15 55.03 15.6 58.4 16.7",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M65.13 19.75C75.22 25.72 81.99 36.71 81.99 49.29C81.99 68.23 66.64 83.58 47.7 83.58C28.76 83.58 13.41 68.23 13.41 49.29C13.41 48.47 13.43 47.43 13.49 46.62",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M42.5401 69.4299C33.3101 67.0799 26.4801 58.7199 26.4801 48.7599C26.4801 36.9799 36.0301 27.4299 47.8101 27.4299C59.5901 27.4299 69.1401 36.9799 69.1401 48.7599C69.1401 59.3399 61.4301 68.1299 51.3301 69.7999C51.1901 69.8199 51.0501 69.8499 50.9001 69.8699",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M82.2001 53.76C84.2101 56.15 85.4201 59.23 85.4201 62.6V64.69C92.8801 64.69 98.9201 70.73 98.9201 78.19V78.88",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M11.14 73.8601L12.15 74.4301L11.14 74.9901L10.59 76.0201L10.04 74.9901L9.03003 74.4301L10.04 73.8601L10.59 72.8301L11.14 73.8601Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M88.0701 33.8901L89.0801 34.4601L88.0701 35.0201L87.5201 36.0501L86.9601 35.0201L85.9501 34.4601L86.9601 33.8901L87.5201 32.8601L88.0701 33.8901Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M14.49 9.78005L18.88 11.75L14.4 13.49L12.43 17.88L10.69 13.4L6.30005 11.43L10.78 9.69005L12.75 5.30005L14.49 9.78005Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M60.9601 21.2201C60.9601 21.2201 63.6001 19.3201 63.6401 14.6501C63.6401 14.6501 68.2001 17.4601 62.6801 21.9001C62.6801 21.9001 62.1601 22.1801 60.9601 21.2201Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M45.63 73.3001C45.63 73.3001 48.14 71.2501 48.38 66.5801C48.38 66.5801 53.84 69.2001 48.14 73.4001C48.13 73.4001 46.73 74.3101 45.63 73.3001Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M46.6201 73.7601C48.7905 73.7601 50.5501 72.0006 50.5501 69.8301C50.5501 67.6597 48.7905 65.9001 46.6201 65.9001C44.4496 65.9001 42.6901 67.6597 42.6901 69.8301C42.6901 72.0006 44.4496 73.7601 46.6201 73.7601Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M61.81 21.7899C63.9805 21.7899 65.74 20.0304 65.74 17.8599C65.74 15.6895 63.9805 13.9299 61.81 13.9299C59.6395 13.9299 57.88 15.6895 57.88 17.8599C57.88 20.0304 59.6395 21.7899 61.81 21.7899Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M17.9101 45.81L12.9601 46.1699L13.1901 46.13C14.5801 45.92 15.7701 45.36 16.5901 44.66C17.5501 43.84 17.5901 42.8899 17.8401 41.2599L17.9101 38.95V45.81Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M17.33 46.3101H11.18C10.62 46.3101 10.17 45.8601 10.17 45.3001V39.1501C10.17 38.5901 10.62 38.1401 11.18 38.1401H17.33C17.89 38.1401 18.34 38.5901 18.34 39.1501V45.3001C18.34 45.8601 17.89 46.3101 17.33 46.3101Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "39.1606",
    y1: "50.083",
    x2: "56.7444",
    y2: "50.083",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "94.43",
    height: "80.09",
    fill: "white",
    transform: "translate(5 4)"
  }))));
}

export default withTheme(FreeForever);
