// @postman/aether/esmLib/components/Illustration/spot-illustrations/Invite.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function Invite(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M71.55 73.83L71.59 62.11L89.99 61.98L74.62 63.44L74.73 73.6L71.55 73.83Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M11.05 32.6701L12.06 33.2401L11.05 33.8001L10.5 34.8301L9.95 33.8001L8.94 33.2401L9.95 32.6701L10.5 31.6401L11.05 32.6701Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M98.41 45.57L99.42 46.14L98.41 46.7L97.86 47.73L97.31 46.7L96.3 46.14L97.31 45.57L97.86 44.54L98.41 45.57Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M31.96 39.6501L32.39 39.6801C24.81 39.6801 19.21 45.7701 19.21 53.3501V55.4401C11.75 55.4401 5.71002 61.4801 5.71002 68.9401V69.6301",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M16.41 5.78005L20.8 7.75005L16.32 9.49005L14.35 13.88L12.61 9.40005L8.22003 7.43005L12.7 5.69005L14.67 1.30005L16.41 5.78005Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M50.42 67.9599C49.31 64.5799 46.14 62.1399 42.39 62.1399C38.39 62.1399 35.04 64.9199 34.16 68.6499C34.16 68.6499 38.4 72.2199 45.83 70.2899L50.42 67.9599Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M34.21 68.8999C35.09 65.1699 38.44 62.3899 42.44 62.3899C46.19 62.3899 49.37 64.8299 50.47 68.2099",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M41.96 71.07C48.6151 71.07 54.01 65.675 54.01 59.02C54.01 52.3649 48.6151 46.97 41.96 46.97C35.305 46.97 29.91 52.3649 29.91 59.02C29.91 65.675 35.305 71.07 41.96 71.07Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M41.96 71.07C48.6151 71.07 54.01 65.675 54.01 59.02C54.01 52.3649 48.6151 46.97 41.96 46.97C35.305 46.97 29.91 52.3649 29.91 59.02C29.91 65.675 35.305 71.07 41.96 71.07Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M42.07 59.41C44.2405 59.41 46 57.6505 46 55.48C46 53.3096 44.2405 51.55 42.07 51.55C39.8995 51.55 38.14 53.3096 38.14 55.48C38.14 57.6505 39.8995 59.41 42.07 59.41Z",
    fill: "#FFA466",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M52.8 39.57H32.08C30.19 39.57 28.66 38.04 28.66 36.15V15.43C28.66 13.54 30.19 12.01 32.08 12.01H52.8C54.69 12.01 56.22 13.54 56.22 15.43V36.15C56.22 38.04 54.69 39.57 52.8 39.57Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M90.37 62.1702V49.5802C90.37 47.6902 88.84 46.1602 86.95 46.1602H66.23C64.34 46.1602 62.81 47.6902 62.81 49.5802V70.3002C62.81 72.1902 64.34 73.7202 66.23 73.7202H74.42",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M86.96 39.1901H66.23C64.34 39.1901 62.81 37.6601 62.81 35.7701V15.0501C62.81 13.1601 64.34 11.6301 66.23 11.6301H86.95C88.84 11.6301 90.37 13.1601 90.37 15.0501V35.7701C90.37 37.6601 88.84 39.1901 86.96 39.1901Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M100.39 62.6702H74.88V81.4802H100.39V62.6702Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M75.08 64.2903L88.04 72.7903L100.59 64.2903",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "34.1625",
    y1: "66.4962",
    x2: "50.4215",
    y2: "66.4962",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "96.89",
    height: "87.65",
    fill: "white",
    transform: "translate(4)"
  }))));
}

export default withTheme(Invite);
