// @postman/aether/esmLib/components/Illustration/spot-illustrations/MonitorAndMocks.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function MonitorAndMocks(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M21.26 49.8599C21.26 49.8599 23.98 50.1299 24.66 46.1899C24.66 46.1899 26.77 48.1599 24.8 50.0599C22.83 51.9599 21.26 49.8599 21.26 49.8599Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M30.35 59.69C30.35 59.69 33.07 59.96 33.75 56.02C33.75 56.02 35.86 57.99 33.89 59.89C31.92 61.79 30.35 59.69 30.35 59.69Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M56.11 56.5802C56.11 56.5802 58.83 56.8502 59.51 52.9102C59.51 52.9102 61.62 54.8802 59.65 56.7802C57.67 58.6802 56.11 56.5802 56.11 56.5802Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M44.45 46.6199C44.45 46.6199 47.17 46.89 47.85 42.95C47.85 42.95 49.96 44.92 47.99 46.82C46.01 48.72 44.45 46.6199 44.45 46.6199Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M57.58 68.8299H19.15C16.9 68.8299 15.08 67.0099 15.08 64.7599V30.9999C15.08 28.7499 16.9 26.9299 19.15 26.9299H54.76",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M71.22 40.45V63.0999",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M60.38 53.7099L71.16 43.4399",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M47.92 47.1702L56.11 53.8502",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M34.54 56.04L44.3 46.99",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M24.85 50.79L30.05 56.65",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M15.01 56.5901L21.76 50.6001",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M23.29 50.9401C24.621 50.9401 25.7 49.8611 25.7 48.5301C25.7 47.1991 24.621 46.1201 23.29 46.1201C21.959 46.1201 20.88 47.1991 20.88 48.5301C20.88 49.8611 21.959 50.9401 23.29 50.9401Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M32.37 60.4401C33.701 60.4401 34.78 59.3611 34.78 58.0301C34.78 56.6991 33.701 55.6201 32.37 55.6201C31.039 55.6201 29.96 56.6991 29.96 58.0301C29.96 59.3611 31.039 60.4401 32.37 60.4401Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M46.16 47.4501C47.491 47.4501 48.57 46.3711 48.57 45.0401C48.57 43.7091 47.491 42.6301 46.16 42.6301C44.829 42.6301 43.75 43.7091 43.75 45.0401C43.75 46.3711 44.829 47.4501 46.16 47.4501Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M58.2 57.4701C59.531 57.4701 60.61 56.3912 60.61 55.0601C60.61 53.7291 59.531 52.6501 58.2 52.6501C56.869 52.6501 55.79 53.7291 55.79 55.0601C55.79 56.3912 56.869 57.4701 58.2 57.4701Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M15.49 11.78L19.88 13.75L15.4 15.49L13.43 19.88L11.69 15.4L7.29999 13.43L11.78 11.69L13.75 7.30005L15.49 11.78Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M77.5 34.9599C76.39 31.5799 73.22 29.1399 69.47 29.1399C65.47 29.1399 62.12 31.9199 61.24 35.6499C61.24 35.6499 65.48 39.2199 72.91 37.2899L77.5 34.9599Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M61.29 35.8999C62.17 32.1699 65.52 29.3899 69.52 29.3899C73.27 29.3899 76.45 31.8299 77.55 35.2099",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M69.04 38.07C75.695 38.07 81.09 32.675 81.09 26.02C81.09 19.3649 75.695 13.97 69.04 13.97C62.385 13.97 56.99 19.3649 56.99 26.02C56.99 32.675 62.385 38.07 69.04 38.07Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M69.04 38.07C75.695 38.07 81.09 32.675 81.09 26.02C81.09 19.3649 75.695 13.97 69.04 13.97C62.385 13.97 56.99 19.3649 56.99 26.02C56.99 32.675 62.385 38.07 69.04 38.07Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M69.15 26.41C71.3205 26.41 73.08 24.6505 73.08 22.48C73.08 20.3096 71.3205 18.55 69.15 18.55C66.9795 18.55 65.22 20.3096 65.22 22.48C65.22 24.6505 66.9795 26.41 69.15 26.41Z",
    fill: "#FFA466",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M88.29 63.1799H57.89V82.0299H88.29V63.1799Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M73.8 81.9001V72.1101H77.41V81.9001",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M80.72 81.9001V72.1101H84.33V81.9001",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M76.36 35.8501H76.65C84.55 35.8501 90.95 42.2501 90.95 50.1501C94.98 50.1501 98.25 53.4201 98.25 57.4501",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M63.43 75.5201C63.43 75.5201 66.07 74.0501 66.31 70.0701C66.31 70.0701 70.06 72.6801 65.16 76.2001C65.16 76.1901 64.4 76.3201 63.43 75.5201Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M64.28 75.99C66.1301 75.99 67.63 74.4902 67.63 72.64C67.63 70.7899 66.1301 69.29 64.28 69.29C62.4298 69.29 60.93 70.7899 60.93 72.64C60.93 74.4902 62.4298 75.99 64.28 75.99Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M14.61 76.4599L15.62 77.0299L14.61 77.5899L14.06 78.6199L13.5 77.5899L12.5 77.0299L13.5 76.4599L14.06 75.4299L14.61 76.4599Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M92.93 30.6899L93.94 31.2599L92.93 31.8199L92.38 32.8499L91.82 31.8199L90.81 31.2599L91.82 30.6899L92.38 29.6599L92.93 30.6899Z",
    fill: illustrationStrokeColor
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "61.2432",
    y1: "33.4962",
    x2: "77.5023",
    y2: "33.4962",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "92.75",
    height: "76.53",
    fill: "white",
    transform: "translate(6 6)"
  }))));
}

export default withTheme(MonitorAndMocks);
