// @postman/aether/esmLib/components/Illustration/spot-illustrations/PayLater.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function PayLater(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M93.3 49.7599L94.68 50.5299L93.3 51.2999L92.55 52.7099L91.8 51.2999L90.42 50.5299L91.8 49.7599L92.55 48.3499L93.3 49.7599Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M100.59 81.64V80.95C100.59 73.49 94.55 67.45 87.09 67.45V65.36C87.09 61.99 85.88 58.91 83.87 56.52C83.87 56.52 83.03 55.44 82.38 55.54",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M90.54 11.18L94.93 13.15L90.45 14.89L88.48 19.28L86.74 14.8L82.35 12.83L86.83 11.09L88.8 6.69995L90.54 11.18Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M75.48 22.73H77.66C80.1 22.73 82.07 24.7 82.07 27.14V72.94C82.07 75.38 80.1 77.35 77.66 77.35H21.15C18.71 77.35 16.74 75.38 16.74 72.94V27.14C16.74 24.7 18.71 22.73 21.15 22.73H22.75",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M25.57 22.73H72.66",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M35.61 52.5799H27.71C26.32 52.5799 25.19 51.4499 25.19 50.0599V42.1599C25.19 40.7699 26.32 39.6399 27.71 39.6399H35.61C37 39.6399 38.13 40.7699 38.13 42.1599V50.0599C38.12 51.4599 37 52.5799 35.61 52.5799Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M53.94 52.5799H46.04C44.65 52.5799 43.52 51.4499 43.52 50.0599V42.1599C43.52 40.7699 44.65 39.6399 46.04 39.6399H53.94C55.33 39.6399 56.46 40.7699 56.46 42.1599V50.0599C56.46 51.4599 55.33 52.5799 53.94 52.5799Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M72.28 52.5799H64.38C62.99 52.5799 61.86 51.4499 61.86 50.0599V42.1599C61.86 40.7699 62.99 39.6399 64.38 39.6399H72.28C73.67 39.6399 74.8 40.7699 74.8 42.1599V50.0599C74.8 51.4599 73.67 52.5799 72.28 52.5799Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M35.61 70.67H27.71C26.32 70.67 25.19 69.54 25.19 68.15V60.25C25.19 58.86 26.32 57.73 27.71 57.73H35.61C37 57.73 38.13 58.86 38.13 60.25V68.15C38.12 69.55 37 70.67 35.61 70.67Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M53.94 70.67H46.04C44.65 70.67 43.52 69.54 43.52 68.15V60.25C43.52 58.86 44.65 57.73 46.04 57.73H53.94C55.33 57.73 56.46 58.86 56.46 60.25V68.15C56.46 69.55 55.33 70.67 53.94 70.67Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M72.28 70.67H64.38C62.99 70.67 61.86 69.54 61.86 68.15V60.25C61.86 58.86 62.99 57.73 64.38 57.73H72.28C73.67 57.73 74.8 58.86 74.8 60.25V68.15C74.8 69.55 73.67 70.67 72.28 70.67Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M72.28 57.74H64.38C62.99 57.74 61.86 58.87 61.86 60.26V68.16C61.86 69.55 62.99 70.68 64.38 70.68H72.28C73.67 70.68 74.8 69.55 74.8 68.16V60.26C74.8 58.87 73.67 57.74 72.28 57.74ZM70.04 67.45C69.69 67.8 69.25 68.01 68.71 68.09V68.47C68.71 68.53 68.68 68.58 68.62 68.62C68.56 68.66 68.5 68.69 68.43 68.69C68.35 68.69 68.29 68.67 68.23 68.62C68.18 68.57 68.15 68.52 68.15 68.47V68.12C67.71 68.11 67.32 68.04 67 67.9C66.68 67.76 66.43 67.6 66.25 67.42C66.07 67.24 65.99 67.07 65.99 66.92C65.99 66.85 66.01 66.78 66.05 66.7C66.09 66.62 66.14 66.56 66.2 66.51C66.26 66.46 66.32 66.43 66.39 66.43C66.47 66.43 66.55 66.47 66.63 66.55C66.71 66.63 66.81 66.72 66.93 66.83C67.05 66.94 67.21 67.03 67.41 67.12C67.61 67.21 67.86 67.26 68.16 67.27V64.45C67.92 64.37 67.69 64.28 67.46 64.18C67.23 64.08 67.02 63.96 66.83 63.81C66.63 63.67 66.48 63.48 66.37 63.25C66.26 63.02 66.2 62.74 66.2 62.4C66.2 62.07 66.26 61.79 66.37 61.56C66.48 61.33 66.63 61.14 66.83 61C67.02 60.86 67.23 60.75 67.46 60.67C67.69 60.59 67.93 60.54 68.17 60.52V60.2C68.17 60.14 68.2 60.08 68.25 60.04C68.3 60 68.37 59.98 68.45 59.98C68.51 59.98 68.58 60 68.64 60.04C68.7 60.08 68.73 60.14 68.73 60.2V60.5C68.91 60.51 69.1 60.53 69.3 60.56C69.49 60.6 69.67 60.64 69.84 60.7C70.01 60.76 70.14 60.83 70.24 60.91C70.34 60.99 70.39 61.09 70.39 61.2C70.39 61.26 70.38 61.32 70.35 61.4C70.32 61.47 70.28 61.54 70.22 61.6C70.16 61.66 70.09 61.69 70.01 61.69C69.94 61.69 69.85 61.66 69.75 61.6C69.65 61.54 69.51 61.48 69.35 61.41C69.19 61.35 68.98 61.31 68.73 61.29V63.7C68.95 63.78 69.17 63.87 69.39 63.98C69.61 64.09 69.81 64.23 69.99 64.39C70.17 64.55 70.31 64.76 70.42 65.01C70.53 65.26 70.58 65.57 70.58 65.93C70.55 66.61 70.38 67.1 70.04 67.45Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M72.28 70.67H64.38C62.99 70.67 61.86 69.54 61.86 68.15V60.25C61.86 58.86 62.99 57.73 64.38 57.73H72.28C73.67 57.73 74.8 58.86 74.8 60.25V68.15C74.8 69.55 73.67 70.67 72.28 70.67Z",
    stroke: illustrationStrokeColor,
    strokeWidth: "0.9676",
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M16.73 33.3999H81.93",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M24.14 26.06C23.48 26.06 22.94 25.52 22.94 24.86V20.91C22.94 20.25 23.48 19.71 24.14 19.71C24.8 19.71 25.34 20.25 25.34 20.91V24.86C25.34 25.52 24.8 26.06 24.14 26.06Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M74.14 26.06C73.48 26.06 72.94 25.52 72.94 24.86V20.91C72.94 20.25 73.48 19.71 74.14 19.71C74.8 19.71 75.34 20.25 75.34 20.91V24.86C75.34 25.52 74.8 26.06 74.14 26.06Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M67.19 61.7998C67.08 61.9398 67.03 62.1198 67.03 62.3498C67.03 62.5898 67.08 62.7898 67.19 62.9398C67.3 63.0898 67.44 63.2098 67.61 63.3098C67.79 63.3998 67.98 63.4898 68.19 63.5598V61.3398C67.98 61.3598 67.78 61.4098 67.61 61.4798C67.43 61.5498 67.29 61.6598 67.19 61.7998Z",
    fill: "url(#paint1_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M69.18 64.9599C69.02 64.8499 68.85 64.7499 68.66 64.6699V67.2599C68.99 67.1999 69.25 67.0699 69.43 66.8699C69.61 66.6699 69.71 66.3899 69.71 66.0399C69.71 65.7699 69.66 65.5499 69.57 65.3799C69.48 65.2099 69.34 65.0699 69.18 64.9599Z",
    fill: "url(#paint2_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M6.89 68.3499L8.26 69.1199L6.89 69.8899L6.13 71.2999L5.38 69.8899L4 69.1199L5.38 68.3499L6.13 66.9399L6.89 68.3499Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "61.8633",
    y1: "64.2064",
    x2: "74.7968",
    y2: "64.2064",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint1_linear",
    x1: "67.0275",
    y1: "62.4488",
    x2: "68.1916",
    y2: "62.4488",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint2_linear",
    x1: "68.6584",
    y1: "65.9703",
    x2: "69.7072",
    y2: "65.9703",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  }))));
}

export default withTheme(PayLater);
