// @postman/aether/esmLib/components/Illustration/spot-illustrations/PopularWorkspaces.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function PopularWorkspaces(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M41.87 56.1201C41.87 56.1201 52.52 61.0201 50.81 73.0301L48.14 72.8401C48.14 72.8401 48.74 61.7001 41.87 56.1201Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M54.5799 38.4799H33.8599C31.9699 38.4799 30.4399 36.9499 30.4399 35.0599V14.3399C30.4399 12.4499 31.9699 10.9199 33.8599 10.9199H54.5799C56.4699 10.9199 57.9999 12.4499 57.9999 14.3399V35.0599C57.9999 36.9499 56.4699 38.4799 54.5799 38.4799Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M30.4399 52.7201V48.4901C30.4399 46.6001 31.9699 45.0701 33.8599 45.0701H54.5799C56.4699 45.0701 57.9999 46.6001 57.9999 48.4901V69.2101C57.9999 71.1001 56.4699 72.6301 54.5799 72.6301H48.1399",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M88.74 72.6301H68.01C66.12 72.6301 64.59 71.1001 64.59 69.2101V48.4901C64.59 46.6001 66.12 45.0701 68.01 45.0701H88.73C90.62 45.0701 92.15 46.6001 92.15 48.4901V69.2101C92.15 71.1001 90.62 72.6301 88.74 72.6301Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M88.74 38.1H68.01C66.12 38.1 64.59 36.57 64.59 34.68V13.96C64.59 12.07 66.12 10.54 68.01 10.54H88.73C90.62 10.54 92.15 12.07 92.15 13.96V34.68C92.15 36.57 90.62 38.1 88.74 38.1Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M31.17 86.5699C40.3876 86.5699 47.86 79.0976 47.86 69.8799C47.86 60.6623 40.3876 53.1899 31.17 53.1899C21.9523 53.1899 14.48 60.6623 14.48 69.8799C14.48 79.0976 21.9523 86.5699 31.17 86.5699Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M39.85 68.79C39.85 65.86 37.55 63.48 34.71 63.48C33.28 63.48 31.99 64.08 31.06 65.05C30.13 64.08 28.84 63.48 27.41 63.48C24.57 63.48 22.27 65.86 22.27 68.79C22.27 70.4 22.98 71.83 24.08 72.81L24.06 72.83L31.32 79.91L38.63 72.29L38.58 72.25C39.37 71.3 39.85 70.1 39.85 68.79Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M39.85 68.17C39.85 65.33 37.55 63.03 34.71 63.03C33.28 63.03 31.99 63.61 31.06 64.55C30.13 63.61 28.84 63.03 27.41 63.03C24.57 63.03 22.27 65.33 22.27 68.17C22.27 69.73 22.98 71.12 24.08 72.06L24.06 72.08L31.32 78.94L38.58 71.51C39.37 70.61 39.85 69.45 39.85 68.17Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M18.73 6.78005L23.13 8.75005L18.64 10.49L16.68 14.88L14.93 10.4L10.54 8.43005L15.03 6.69005L16.99 2.30005L18.73 6.78005Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M8.71999 29.4799L9.72999 30.04L8.71999 30.61L8.16999 31.6399L7.61999 30.61L6.60999 30.04L7.61999 29.4799L8.16999 28.45L8.71999 29.4799Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M100.87 49.7199L101.88 50.2899L100.87 50.8499L100.32 51.8799L99.76 50.8499L98.75 50.2899L99.76 49.7199L100.32 48.6899L100.87 49.7199Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M30.1 35.53C22.52 35.53 17 41.67 17 49.26V51.35C9.54 51.35 3.5 57.39 3.5 64.85V65.54",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "22.2656",
    y1: "71.6859",
    x2: "39.8495",
    y2: "71.6859",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "98.88",
    height: "86.06",
    fill: "white",
    transform: "translate(3 1)"
  }))));
}

export default withTheme(PopularWorkspaces);
