// @postman/aether/esmLib/components/Illustration/spot-illustrations/Security.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function Security(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M44.49 27.5302H37.92L38.3 25.4202H44.87L44.49 27.5302Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M71.49 27.53H64.92L64.51 25.45H71.08L71.49 27.53Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M70.24 85.37H38.26C34.14 85.37 30.8 82.03 30.8 77.91C30.8 73.79 34.14 70.45 38.26 70.45H70.24C74.36 70.45 77.7 73.79 77.7 77.91C77.7 82.03 74.36 85.37 70.24 85.37Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M74.12 62.3201H35.37C34.12 62.3201 33.1 61.3101 33.1 60.0501V30.1701C33.1 28.9201 34.11 27.9001 35.37 27.9001H74.12C75.37 27.9001 76.39 28.9101 76.39 30.1701V60.0401C76.39 61.3001 75.37 62.3201 74.12 62.3201Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M37.95 27.27V23.12C37.95 13.83 45.4799 6.29004 54.7799 6.29004C64.0699 6.29004 71.61 13.82 71.61 23.12V27.27",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M44.7999 27.2699V23.4799C44.7999 17.9299 49.2999 13.4299 54.8499 13.4299C60.3999 13.4299 64.8999 17.9299 64.8999 23.4799V27.2699",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M59.69 44.49C59.69 41.75 57.47 39.53 54.73 39.53C51.99 39.53 49.77 41.75 49.77 44.49C49.77 46.18 50.62 47.67 51.91 48.56L51.78 55.58H57.63L57.37 48.68C58.76 47.8 59.69 46.26 59.69 44.49Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M76.5601 45.5C84.1401 45.5 89.1801 51.5 89.1801 59.08V61.17C96.6401 61.17 102.12 66.21 102.12 73.67V74.36",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M21.52 51.62C22.19 47.74 25.6 44.53 29.67 44.53H32.86",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M5.33996 69.78C3.84996 68.56 9.00996 63.5 13.04 63.5C16.76 63.5 19.92 65.91 21.05 69.25C21.04 69.25 11.71 74.99 5.33996 69.78Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M21.0399 69.25C19.9099 65.91 16.7499 63.5 13.0299 63.5C8.99994 63.5 5.78994 66.08 4.93994 69.85",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M12.55 72.1701C19.205 72.1701 24.6 66.7751 24.6 60.1201C24.6 53.465 19.205 48.0701 12.55 48.0701C5.89497 48.0701 0.5 53.465 0.5 60.1201C0.5 66.7751 5.89497 72.1701 12.55 72.1701Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M12.55 72.1701C19.205 72.1701 24.6 66.7751 24.6 60.1201C24.6 53.465 19.205 48.0701 12.55 48.0701C5.89497 48.0701 0.5 53.465 0.5 60.1201C0.5 66.7751 5.89497 72.1701 12.55 72.1701Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M12.66 60.5101C14.8305 60.5101 16.59 58.7506 16.59 56.5801C16.59 54.4097 14.8305 52.6501 12.66 52.6501C10.4895 52.6501 8.72998 54.4097 8.72998 56.5801C8.72998 58.7506 10.4895 60.5101 12.66 60.5101Z",
    fill: "#FDA968",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M23.98 24.7601L25.36 25.5301L23.98 26.3001L23.22 27.7101L22.47 26.3001L21.09 25.5301L22.47 24.7601L23.22 23.3601L23.98 24.7601Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M97.46 46.8001L98.84 47.5801L97.46 48.3501L96.71 49.7501L95.96 48.3501L94.58 47.5801L95.96 46.8001L96.71 45.4001L97.46 46.8001Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M96.2509 8.49974C97.6155 5.72427 96.5279 2.39578 93.8218 1.06535C91.1157 -0.265078 87.8159 0.906363 86.4513 3.68184C85.0868 6.45731 86.1743 9.7858 88.8804 11.1162C91.5865 12.4467 94.8864 11.2752 96.2509 8.49974Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M97.04 6.19017C99.75 7.96017 101.36 9.59017 100.98 10.3502C100.42 11.4902 95.68 10.3102 90.39 7.71017C85.1 5.11017 81.27 2.08017 81.83 0.940167C82.21 0.170167 84.49 0.460167 87.55 1.52017",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M44.42 75.3V80.46",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M42.1899 76.55L46.5399 79.34",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M42.3 79.33L46.62 76.51",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M54.42 75.3V80.46",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M52.1899 76.55L56.5399 79.34",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M52.3 79.33L56.62 76.51",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M64.42 75.3V80.46",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M62.1899 76.55L66.5399 79.34",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M62.3 79.33L66.6201 76.51",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "5.07256",
    y1: "67.7296",
    x2: "21.0415",
    y2: "67.7296",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "102.62",
    height: "85.87",
    fill: "white"
  }))));
}

export default withTheme(Security);
