// @postman/aether/esmLib/components/Illustration/spot-illustrations/SharedRequests.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function SharedRequests(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M21.71 9.78005L26.1 11.75L21.62 13.49L19.65 17.88L17.91 13.4L13.51 11.43L18 9.69005L19.96 5.30005L21.71 9.78005Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M54.93 46.1H34.21C32.32 46.1 30.79 44.57 30.79 42.68V21.96C30.79 20.07 32.32 18.54 34.21 18.54H54.93C56.82 18.54 58.35 20.07 58.35 21.96V42.68C58.35 44.57 56.82 46.1 54.93 46.1Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M92.77 29.6899V42.6799C92.77 44.5699 91.24 46.0999 89.35 46.0999H68.63",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M65.21 42.68V21.96C65.21 20.07 66.74 18.54 68.63 18.54H79.32",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M58.15 58.24L58.35 76.68C58.35 78.57 56.82 80.1 54.93 80.1H34.21C32.32 80.1 30.79 78.57 30.79 76.68V59.15V55.96C30.79 54.07 32.32 52.54 34.21 52.54L53.5 52.64",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M89.0899 79.7202H68.3699C66.4799 79.7202 64.95 78.1902 64.95 76.3002V55.5802C64.95 53.6902 66.4799 52.1602 68.3699 52.1602H89.0899C90.9799 52.1602 92.51 53.6902 92.51 55.5802V76.3002C92.51 78.1902 90.9799 79.7202 89.0899 79.7202Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("g", {
    opacity: "0.7"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M86.6799 31.02C86.6799 31.02 89.8599 36.04 81.8799 41.28L81.2899 36.47",
    fill: "url(#paint0_linear)"
  })), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M79.23 23.67C79.23 23.67 73.92 20.88 69.21 29.19L74.05 29.46",
    fill: "url(#paint1_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M86.6799 31.02C86.6799 31.02 89.8599 36.04 81.8799 41.28L81.2899 36.47",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M77.99 33.79L41.01 79.42",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M87.4516 25.075C87.6016 23.9528 86.8179 22.9221 85.7012 22.7728C84.5844 22.6235 83.5575 23.4122 83.4075 24.5344C83.2575 25.6567 84.0412 26.6874 85.1579 26.8367C86.2747 26.986 87.3016 26.1973 87.4516 25.075Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M92.32 19.34C92.24 18.87 91.99 17.88 91.99 17.88C91.75 17.82 91.52 17.77 91.3 17.72C87.93 17.05 84.47 18.3 82.11 20.81L79.09 24.02L73.85 29.59L77.58 33.21L81.27 36.87L86.78 31.01L89.64 27.97C91.8 25.66 92.85 22.47 92.32 19.34ZM86.94 26.18C86.18 27.02 84.9 27.08 84.06 26.32C83.23 25.56 83.16 24.27 83.92 23.43C84.68 22.59 85.96 22.53 86.8 23.29C87.64 24.05 87.7 25.34 86.94 26.18Z",
    fill: "url(#paint2_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M91.99 17.88C91.75 17.82 91.52 17.77 91.3 17.72C87.93 17.05 84.47 18.3 82.11 20.81L79.09 24.02L73.85 29.59L77.58 33.21L81.27 36.87L86.78 31.01L89.64 27.97C91.81 25.66 92.87 22.47 92.34 19.34C92.25 18.87 91.99 17.88 91.99 17.88Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M88.1454 25.1678C88.3466 23.6624 87.2938 22.2794 85.7939 22.0789C84.2939 21.8784 82.9149 22.9363 82.7136 24.4417C82.5124 25.9471 83.5652 27.33 85.0651 27.5305C86.5651 27.731 87.9442 26.6732 88.1454 25.1678Z",
    fill: "white"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M88.1454 25.1678C88.3466 23.6624 87.2938 22.2794 85.7939 22.0789C84.2939 21.8784 82.9149 22.9363 82.7136 24.4417C82.5124 25.9471 83.5652 27.33 85.0651 27.5305C86.5651 27.731 87.9442 26.6732 88.1454 25.1678Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M79.23 23.67C79.23 23.67 73.92 20.88 69.21 29.19L74.05 29.46",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M76.41 32.74L31.45 72.24",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M66.0599 14.0901L67.0699 14.6501L66.0599 15.2201L65.4999 16.2501L64.9499 15.2201L63.9399 14.6501L64.9499 14.0901L65.4999 13.0601L66.0599 14.0901Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M19.89 42.25L20.9 42.82L19.89 43.38L19.34 44.41L18.79 43.38L17.78 42.82L18.79 42.25L19.34 41.22L19.89 42.25Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M30.55 58.6301C26 60.5101 23.67 64.6701 23.67 69.2801V71.0401C16.4 71.0401 10.5 76.1301 10.5 82.4101V83.0001",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "81.2871",
    y1: "36.1559",
    x2: "87.3065",
    y2: "36.1559",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint1_linear",
    x1: "69.2102",
    y1: "26.3452",
    x2: "79.2315",
    y2: "26.3452",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint2_linear",
    x1: "73.8352",
    y1: "27.2024",
    x2: "92.4582",
    y2: "27.2024",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "83.27",
    height: "79.5",
    fill: "white",
    transform: "translate(10 4)"
  }))));
}

export default withTheme(SharedRequests);
