// @postman/aether/esmLib/components/Illustration/spot-illustrations/Winners.js
/* eslint-disable react/prop-types */
import React from 'react';
import { withTheme } from 'styled-components';

function Winners(props) {
  var theme = props.theme;
  var illustrationStrokeColor = theme['illustration-stroke-color'] || '#686868';
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 104 88",
    height: "88px",
    width: "104px",
    fill: "none"
  }, /*#__PURE__*/React.createElement("g", {
    clipPath: "url(#clip0)"
  }, /*#__PURE__*/React.createElement("path", {
    opacity: "0.2",
    d: "M35.6301 47.79C35.6301 47.79 44.5801 52.26 52.8001 43.73C52.8001 43.73 41.1501 50.08 31.6801 41.02C31.6701 41.03 35.7301 44.46 35.6301 47.79Z",
    fill: "#262626"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M23.67 5.92993C22.59 9.39993 22.76 13.2199 22.76 17.2199C22.76 33.2399 32.35 46.2299 44.19 46.2299C56.03 46.2299 65.62 33.2399 65.62 17.2199C65.62 13.2099 65.79 9.39993 64.71 5.92993H25.76H23.67Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M33.2401 42.4199C33.2401 42.4199 41.7701 55.1599 31.3301 68.1699",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M54.6601 42.4199C54.6601 42.4199 46.1301 55.1599 56.5701 68.1699",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M57.0401 68.3401H29.0201C23.6701 68.3401 19.3301 72.6801 19.3301 78.0301H56.8401",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M65.21 10.4101C72.86 10.1601 79.26 16.1601 79.51 23.8101C79.76 31.4601 73.76 37.8601 66.11 38.1101C63.94 38.1801 61.88 37.7501 60.02 36.9201",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M65.41 16.5902C69.65 16.4502 73.1901 19.7802 73.3301 24.0102C73.4701 28.2402 70.14 31.7902 65.91 31.9302C65.08 31.9602 64.27 31.8502 63.51 31.6302L63.05 31.4202",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M23.0301 10.4101C15.3801 10.1601 8.98007 16.1601 8.73007 23.8101C8.48007 31.4601 14.4801 37.8601 22.1301 38.1101C24.3001 38.1801 26.3601 37.7501 28.2201 36.9201",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M22.83 16.5902C18.59 16.4502 15.05 19.7802 14.91 24.0102C14.77 28.2502 18.1 31.7902 22.33 31.9302C23.16 31.9602 23.97 31.8502 24.73 31.6302L25.17 31.5502",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    opacity: "0.7",
    d: "M44.9 15.65L46.79 19.48C46.95 19.8 47.25 20.02 47.6 20.07L51.83 20.68C52.71 20.81 53.07 21.89 52.43 22.52L49.37 25.5C49.12 25.75 49 26.1 49.06 26.4501L49.78 30.66C49.93 31.54 49.01 32.21 48.22 31.8L44.44 29.81C44.13 29.65 43.75 29.65 43.44 29.81L39.66 31.8C38.87 32.22 37.95 31.54 38.1 30.66L38.82 26.4501C38.88 26.1 38.76 25.74 38.51 25.5L35.45 22.52C34.81 21.9 35.16 20.81 36.05 20.68L40.28 20.07C40.63 20.02 40.93 19.8 41.09 19.48L42.98 15.65C43.36 14.85 44.5 14.85 44.9 15.65Z",
    fill: "url(#paint0_linear)"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M44.9 15.65L46.79 19.48C46.95 19.8 47.25 20.02 47.6 20.07L51.83 20.68C52.71 20.81 53.07 21.89 52.43 22.52L49.37 25.5C49.12 25.75 49 26.1001 49.06 26.4501L49.78 30.66C49.93 31.54 49.01 32.21 48.22 31.8L44.44 29.81C44.13 29.65 43.75 29.65 43.44 29.81L39.66 31.8C38.87 32.22 37.95 31.54 38.1 30.66L38.82 26.4501C38.88 26.1001 38.76 25.74 38.51 25.5L35.45 22.52C34.81 21.9 35.16 20.81 36.05 20.68L40.28 20.07C40.63 20.02 40.93 19.8 41.09 19.48L42.98 15.65C43.36 14.85 44.5 14.85 44.9 15.65Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M97.5084 10.4956C98.8729 7.72012 97.7854 4.39163 95.0793 3.0612C92.3732 1.73077 89.0733 2.90221 87.7088 5.67769C86.3442 8.45316 87.4318 11.7816 90.1378 13.1121C92.8439 14.4425 96.1438 13.2711 97.5084 10.4956Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M98.97 84.1501L59.95 84.7501C58.34 84.7701 57.02 83.4901 56.99 81.8701L56.56 53.9101C56.54 52.3001 57.82 50.9801 59.44 50.9501L98.46 50.3501C100.07 50.3301 101.39 51.6101 101.42 53.2301L101.85 81.1901C101.87 82.8001 100.58 84.1301 98.97 84.1501Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M81.06 71.1401C83.6337 71.1401 85.72 69.0537 85.72 66.4801C85.72 63.9064 83.6337 61.8201 81.06 61.8201C78.4864 61.8201 76.4 63.9064 76.4 66.4801C76.4 69.0537 78.4864 71.1401 81.06 71.1401Z",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M98.3 8.18992C101.01 9.95992 102.62 11.5899 102.24 12.3499C101.68 13.4899 96.94 12.3099 91.65 9.70992C86.36 7.10992 82.53 4.07992 83.09 2.93992C83.47 2.16992 85.75 2.45992 88.81 3.51992",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M94.14 30.6301L95.14 31.1901L94.14 31.7501L93.6 32.7601L93.05 31.7501L92.05 31.1901L93.05 30.6301L93.6 29.6101L94.14 30.6301Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M4.14005 40.6301L5.14005 41.1901L4.14005 41.7501L3.60005 42.7601L3.05005 41.7501L2.05005 41.1901L3.05005 40.6301L3.60005 39.6101L4.14005 40.6301Z",
    fill: illustrationStrokeColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M28.73 37.6001C21.15 37.6001 15 43.7501 15 51.3301V53.4201C7.54 53.4201 2.5 59.4601 2.5 66.9201V67.6101",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10",
    strokeLinecap: "round",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M57.75 83.8001L47.78 78.0601",
    stroke: illustrationStrokeColor,
    strokeMiterlimit: "10"
  })), /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "paint0_linear",
    x1: "35.1101",
    y1: "23.4917",
    x2: "52.7504",
    y2: "23.4917",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    stopColor: "#FF7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.2957",
    stopColor: "#FE7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.4996",
    stopColor: "#FC7C0A"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.6759",
    stopColor: "#F88B15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.8364",
    stopColor: "#F3A125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.9849",
    stopColor: "#ECBD39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#EBC03B"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip0"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "100.8",
    height: "83.25",
    fill: "white",
    transform: "translate(2 2)"
  }))));
}

export default withTheme(Winners);
