// @pipedrive/report-bad-lead/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var conventionUiReact = require('@pipedrive/convention-ui-react');
var React = require('react');
var React__default = _interopDefault(React);
var styled = _interopDefault(require('styled-components'));
var fetch$1 = require('@pipedrive/fetch');

function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var runtime_1 = createCommonjsModule(function (module) {
/**
 * Copyright (c) 2014-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

var runtime = (function (exports) {

  var Op = Object.prototype;
  var hasOwn = Op.hasOwnProperty;
  var undefined$1; // More compressible than void 0.
  var $Symbol = typeof Symbol === "function" ? Symbol : {};
  var iteratorSymbol = $Symbol.iterator || "@@iterator";
  var asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator";
  var toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag";

  function wrap(innerFn, outerFn, self, tryLocsList) {
    // If outerFn provided and outerFn.prototype is a Generator, then outerFn.prototype instanceof Generator.
    var protoGenerator = outerFn && outerFn.prototype instanceof Generator ? outerFn : Generator;
    var generator = Object.create(protoGenerator.prototype);
    var context = new Context(tryLocsList || []);

    // The ._invoke method unifies the implementations of the .next,
    // .throw, and .return methods.
    generator._invoke = makeInvokeMethod(innerFn, self, context);

    return generator;
  }
  exports.wrap = wrap;

  // Try/catch helper to minimize deoptimizations. Returns a completion
  // record like context.tryEntries[i].completion. This interface could
  // have been (and was previously) designed to take a closure to be
  // invoked without arguments, but in all the cases we care about we
  // already have an existing method we want to call, so there's no need
  // to create a new function object. We can even get away with assuming
  // the method takes exactly one argument, since that happens to be true
  // in every case, so we don't have to touch the arguments object. The
  // only additional allocation required is the completion record, which
  // has a stable shape and so hopefully should be cheap to allocate.
  function tryCatch(fn, obj, arg) {
    try {
      return { type: "normal", arg: fn.call(obj, arg) };
    } catch (err) {
      return { type: "throw", arg: err };
    }
  }

  var GenStateSuspendedStart = "suspendedStart";
  var GenStateSuspendedYield = "suspendedYield";
  var GenStateExecuting = "executing";
  var GenStateCompleted = "completed";

  // Returning this object from the innerFn has the same effect as
  // breaking out of the dispatch switch statement.
  var ContinueSentinel = {};

  // Dummy constructor functions that we use as the .constructor and
  // .constructor.prototype properties for functions that return Generator
  // objects. For full spec compliance, you may wish to configure your
  // minifier not to mangle the names of these two functions.
  function Generator() {}
  function GeneratorFunction() {}
  function GeneratorFunctionPrototype() {}

  // This is a polyfill for %IteratorPrototype% for environments that
  // don't natively support it.
  var IteratorPrototype = {};
  IteratorPrototype[iteratorSymbol] = function () {
    return this;
  };

  var getProto = Object.getPrototypeOf;
  var NativeIteratorPrototype = getProto && getProto(getProto(values([])));
  if (NativeIteratorPrototype &&
      NativeIteratorPrototype !== Op &&
      hasOwn.call(NativeIteratorPrototype, iteratorSymbol)) {
    // This environment has a native %IteratorPrototype%; use it instead
    // of the polyfill.
    IteratorPrototype = NativeIteratorPrototype;
  }

  var Gp = GeneratorFunctionPrototype.prototype =
    Generator.prototype = Object.create(IteratorPrototype);
  GeneratorFunction.prototype = Gp.constructor = GeneratorFunctionPrototype;
  GeneratorFunctionPrototype.constructor = GeneratorFunction;
  GeneratorFunctionPrototype[toStringTagSymbol] =
    GeneratorFunction.displayName = "GeneratorFunction";

  // Helper for defining the .next, .throw, and .return methods of the
  // Iterator interface in terms of a single ._invoke method.
  function defineIteratorMethods(prototype) {
    ["next", "throw", "return"].forEach(function(method) {
      prototype[method] = function(arg) {
        return this._invoke(method, arg);
      };
    });
  }

  exports.isGeneratorFunction = function(genFun) {
    var ctor = typeof genFun === "function" && genFun.constructor;
    return ctor
      ? ctor === GeneratorFunction ||
        // For the native GeneratorFunction constructor, the best we can
        // do is to check its .name property.
        (ctor.displayName || ctor.name) === "GeneratorFunction"
      : false;
  };

  exports.mark = function(genFun) {
    if (Object.setPrototypeOf) {
      Object.setPrototypeOf(genFun, GeneratorFunctionPrototype);
    } else {
      genFun.__proto__ = GeneratorFunctionPrototype;
      if (!(toStringTagSymbol in genFun)) {
        genFun[toStringTagSymbol] = "GeneratorFunction";
      }
    }
    genFun.prototype = Object.create(Gp);
    return genFun;
  };

  // Within the body of any async function, `await x` is transformed to
  // `yield regeneratorRuntime.awrap(x)`, so that the runtime can test
  // `hasOwn.call(value, "__await")` to determine if the yielded value is
  // meant to be awaited.
  exports.awrap = function(arg) {
    return { __await: arg };
  };

  function AsyncIterator(generator, PromiseImpl) {
    function invoke(method, arg, resolve, reject) {
      var record = tryCatch(generator[method], generator, arg);
      if (record.type === "throw") {
        reject(record.arg);
      } else {
        var result = record.arg;
        var value = result.value;
        if (value &&
            typeof value === "object" &&
            hasOwn.call(value, "__await")) {
          return PromiseImpl.resolve(value.__await).then(function(value) {
            invoke("next", value, resolve, reject);
          }, function(err) {
            invoke("throw", err, resolve, reject);
          });
        }

        return PromiseImpl.resolve(value).then(function(unwrapped) {
          // When a yielded Promise is resolved, its final value becomes
          // the .value of the Promise<{value,done}> result for the
          // current iteration.
          result.value = unwrapped;
          resolve(result);
        }, function(error) {
          // If a rejected Promise was yielded, throw the rejection back
          // into the async generator function so it can be handled there.
          return invoke("throw", error, resolve, reject);
        });
      }
    }

    var previousPromise;

    function enqueue(method, arg) {
      function callInvokeWithMethodAndArg() {
        return new PromiseImpl(function(resolve, reject) {
          invoke(method, arg, resolve, reject);
        });
      }

      return previousPromise =
        // If enqueue has been called before, then we want to wait until
        // all previous Promises have been resolved before calling invoke,
        // so that results are always delivered in the correct order. If
        // enqueue has not been called before, then it is important to
        // call invoke immediately, without waiting on a callback to fire,
        // so that the async generator function has the opportunity to do
        // any necessary setup in a predictable way. This predictability
        // is why the Promise constructor synchronously invokes its
        // executor callback, and why async functions synchronously
        // execute code before the first await. Since we implement simple
        // async functions in terms of async generators, it is especially
        // important to get this right, even though it requires care.
        previousPromise ? previousPromise.then(
          callInvokeWithMethodAndArg,
          // Avoid propagating failures to Promises returned by later
          // invocations of the iterator.
          callInvokeWithMethodAndArg
        ) : callInvokeWithMethodAndArg();
    }

    // Define the unified helper method that is used to implement .next,
    // .throw, and .return (see defineIteratorMethods).
    this._invoke = enqueue;
  }

  defineIteratorMethods(AsyncIterator.prototype);
  AsyncIterator.prototype[asyncIteratorSymbol] = function () {
    return this;
  };
  exports.AsyncIterator = AsyncIterator;

  // Note that simple async functions are implemented on top of
  // AsyncIterator objects; they just return a Promise for the value of
  // the final result produced by the iterator.
  exports.async = function(innerFn, outerFn, self, tryLocsList, PromiseImpl) {
    if (PromiseImpl === void 0) PromiseImpl = Promise;

    var iter = new AsyncIterator(
      wrap(innerFn, outerFn, self, tryLocsList),
      PromiseImpl
    );

    return exports.isGeneratorFunction(outerFn)
      ? iter // If outerFn is a generator, return the full iterator.
      : iter.next().then(function(result) {
          return result.done ? result.value : iter.next();
        });
  };

  function makeInvokeMethod(innerFn, self, context) {
    var state = GenStateSuspendedStart;

    return function invoke(method, arg) {
      if (state === GenStateExecuting) {
        throw new Error("Generator is already running");
      }

      if (state === GenStateCompleted) {
        if (method === "throw") {
          throw arg;
        }

        // Be forgiving, per 25.3.3.3.3 of the spec:
        // https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generatorresume
        return doneResult();
      }

      context.method = method;
      context.arg = arg;

      while (true) {
        var delegate = context.delegate;
        if (delegate) {
          var delegateResult = maybeInvokeDelegate(delegate, context);
          if (delegateResult) {
            if (delegateResult === ContinueSentinel) continue;
            return delegateResult;
          }
        }

        if (context.method === "next") {
          // Setting context._sent for legacy support of Babel's
          // function.sent implementation.
          context.sent = context._sent = context.arg;

        } else if (context.method === "throw") {
          if (state === GenStateSuspendedStart) {
            state = GenStateCompleted;
            throw context.arg;
          }

          context.dispatchException(context.arg);

        } else if (context.method === "return") {
          context.abrupt("return", context.arg);
        }

        state = GenStateExecuting;

        var record = tryCatch(innerFn, self, context);
        if (record.type === "normal") {
          // If an exception is thrown from innerFn, we leave state ===
          // GenStateExecuting and loop back for another invocation.
          state = context.done
            ? GenStateCompleted
            : GenStateSuspendedYield;

          if (record.arg === ContinueSentinel) {
            continue;
          }

          return {
            value: record.arg,
            done: context.done
          };

        } else if (record.type === "throw") {
          state = GenStateCompleted;
          // Dispatch the exception by looping back around to the
          // context.dispatchException(context.arg) call above.
          context.method = "throw";
          context.arg = record.arg;
        }
      }
    };
  }

  // Call delegate.iterator[context.method](context.arg) and handle the
  // result, either by returning a { value, done } result from the
  // delegate iterator, or by modifying context.method and context.arg,
  // setting context.delegate to null, and returning the ContinueSentinel.
  function maybeInvokeDelegate(delegate, context) {
    var method = delegate.iterator[context.method];
    if (method === undefined$1) {
      // A .throw or .return when the delegate iterator has no .throw
      // method always terminates the yield* loop.
      context.delegate = null;

      if (context.method === "throw") {
        // Note: ["return"] must be used for ES3 parsing compatibility.
        if (delegate.iterator["return"]) {
          // If the delegate iterator has a return method, give it a
          // chance to clean up.
          context.method = "return";
          context.arg = undefined$1;
          maybeInvokeDelegate(delegate, context);

          if (context.method === "throw") {
            // If maybeInvokeDelegate(context) changed context.method from
            // "return" to "throw", let that override the TypeError below.
            return ContinueSentinel;
          }
        }

        context.method = "throw";
        context.arg = new TypeError(
          "The iterator does not provide a 'throw' method");
      }

      return ContinueSentinel;
    }

    var record = tryCatch(method, delegate.iterator, context.arg);

    if (record.type === "throw") {
      context.method = "throw";
      context.arg = record.arg;
      context.delegate = null;
      return ContinueSentinel;
    }

    var info = record.arg;

    if (! info) {
      context.method = "throw";
      context.arg = new TypeError("iterator result is not an object");
      context.delegate = null;
      return ContinueSentinel;
    }

    if (info.done) {
      // Assign the result of the finished delegate to the temporary
      // variable specified by delegate.resultName (see delegateYield).
      context[delegate.resultName] = info.value;

      // Resume execution at the desired location (see delegateYield).
      context.next = delegate.nextLoc;

      // If context.method was "throw" but the delegate handled the
      // exception, let the outer generator proceed normally. If
      // context.method was "next", forget context.arg since it has been
      // "consumed" by the delegate iterator. If context.method was
      // "return", allow the original .return call to continue in the
      // outer generator.
      if (context.method !== "return") {
        context.method = "next";
        context.arg = undefined$1;
      }

    } else {
      // Re-yield the result returned by the delegate method.
      return info;
    }

    // The delegate iterator is finished, so forget it and continue with
    // the outer generator.
    context.delegate = null;
    return ContinueSentinel;
  }

  // Define Generator.prototype.{next,throw,return} in terms of the
  // unified ._invoke helper method.
  defineIteratorMethods(Gp);

  Gp[toStringTagSymbol] = "Generator";

  // A Generator should always return itself as the iterator object when the
  // @@iterator function is called on it. Some browsers' implementations of the
  // iterator prototype chain incorrectly implement this, causing the Generator
  // object to not be returned from this call. This ensures that doesn't happen.
  // See https://github.com/facebook/regenerator/issues/274 for more details.
  Gp[iteratorSymbol] = function() {
    return this;
  };

  Gp.toString = function() {
    return "[object Generator]";
  };

  function pushTryEntry(locs) {
    var entry = { tryLoc: locs[0] };

    if (1 in locs) {
      entry.catchLoc = locs[1];
    }

    if (2 in locs) {
      entry.finallyLoc = locs[2];
      entry.afterLoc = locs[3];
    }

    this.tryEntries.push(entry);
  }

  function resetTryEntry(entry) {
    var record = entry.completion || {};
    record.type = "normal";
    delete record.arg;
    entry.completion = record;
  }

  function Context(tryLocsList) {
    // The root entry object (effectively a try statement without a catch
    // or a finally block) gives us a place to store values thrown from
    // locations where there is no enclosing try statement.
    this.tryEntries = [{ tryLoc: "root" }];
    tryLocsList.forEach(pushTryEntry, this);
    this.reset(true);
  }

  exports.keys = function(object) {
    var keys = [];
    for (var key in object) {
      keys.push(key);
    }
    keys.reverse();

    // Rather than returning an object with a next method, we keep
    // things simple and return the next function itself.
    return function next() {
      while (keys.length) {
        var key = keys.pop();
        if (key in object) {
          next.value = key;
          next.done = false;
          return next;
        }
      }

      // To avoid creating an additional object, we just hang the .value
      // and .done properties off the next function object itself. This
      // also ensures that the minifier will not anonymize the function.
      next.done = true;
      return next;
    };
  };

  function values(iterable) {
    if (iterable) {
      var iteratorMethod = iterable[iteratorSymbol];
      if (iteratorMethod) {
        return iteratorMethod.call(iterable);
      }

      if (typeof iterable.next === "function") {
        return iterable;
      }

      if (!isNaN(iterable.length)) {
        var i = -1, next = function next() {
          while (++i < iterable.length) {
            if (hasOwn.call(iterable, i)) {
              next.value = iterable[i];
              next.done = false;
              return next;
            }
          }

          next.value = undefined$1;
          next.done = true;

          return next;
        };

        return next.next = next;
      }
    }

    // Return an iterator with no values.
    return { next: doneResult };
  }
  exports.values = values;

  function doneResult() {
    return { value: undefined$1, done: true };
  }

  Context.prototype = {
    constructor: Context,

    reset: function(skipTempReset) {
      this.prev = 0;
      this.next = 0;
      // Resetting context._sent for legacy support of Babel's
      // function.sent implementation.
      this.sent = this._sent = undefined$1;
      this.done = false;
      this.delegate = null;

      this.method = "next";
      this.arg = undefined$1;

      this.tryEntries.forEach(resetTryEntry);

      if (!skipTempReset) {
        for (var name in this) {
          // Not sure about the optimal order of these conditions:
          if (name.charAt(0) === "t" &&
              hasOwn.call(this, name) &&
              !isNaN(+name.slice(1))) {
            this[name] = undefined$1;
          }
        }
      }
    },

    stop: function() {
      this.done = true;

      var rootEntry = this.tryEntries[0];
      var rootRecord = rootEntry.completion;
      if (rootRecord.type === "throw") {
        throw rootRecord.arg;
      }

      return this.rval;
    },

    dispatchException: function(exception) {
      if (this.done) {
        throw exception;
      }

      var context = this;
      function handle(loc, caught) {
        record.type = "throw";
        record.arg = exception;
        context.next = loc;

        if (caught) {
          // If the dispatched exception was caught by a catch block,
          // then let that catch block handle the exception normally.
          context.method = "next";
          context.arg = undefined$1;
        }

        return !! caught;
      }

      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        var record = entry.completion;

        if (entry.tryLoc === "root") {
          // Exception thrown outside of any try block that could handle
          // it, so set the completion value of the entire function to
          // throw the exception.
          return handle("end");
        }

        if (entry.tryLoc <= this.prev) {
          var hasCatch = hasOwn.call(entry, "catchLoc");
          var hasFinally = hasOwn.call(entry, "finallyLoc");

          if (hasCatch && hasFinally) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            } else if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else if (hasCatch) {
            if (this.prev < entry.catchLoc) {
              return handle(entry.catchLoc, true);
            }

          } else if (hasFinally) {
            if (this.prev < entry.finallyLoc) {
              return handle(entry.finallyLoc);
            }

          } else {
            throw new Error("try statement without catch or finally");
          }
        }
      }
    },

    abrupt: function(type, arg) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc <= this.prev &&
            hasOwn.call(entry, "finallyLoc") &&
            this.prev < entry.finallyLoc) {
          var finallyEntry = entry;
          break;
        }
      }

      if (finallyEntry &&
          (type === "break" ||
           type === "continue") &&
          finallyEntry.tryLoc <= arg &&
          arg <= finallyEntry.finallyLoc) {
        // Ignore the finally entry if control is not jumping to a
        // location outside the try/catch block.
        finallyEntry = null;
      }

      var record = finallyEntry ? finallyEntry.completion : {};
      record.type = type;
      record.arg = arg;

      if (finallyEntry) {
        this.method = "next";
        this.next = finallyEntry.finallyLoc;
        return ContinueSentinel;
      }

      return this.complete(record);
    },

    complete: function(record, afterLoc) {
      if (record.type === "throw") {
        throw record.arg;
      }

      if (record.type === "break" ||
          record.type === "continue") {
        this.next = record.arg;
      } else if (record.type === "return") {
        this.rval = this.arg = record.arg;
        this.method = "return";
        this.next = "end";
      } else if (record.type === "normal" && afterLoc) {
        this.next = afterLoc;
      }

      return ContinueSentinel;
    },

    finish: function(finallyLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.finallyLoc === finallyLoc) {
          this.complete(entry.completion, entry.afterLoc);
          resetTryEntry(entry);
          return ContinueSentinel;
        }
      }
    },

    "catch": function(tryLoc) {
      for (var i = this.tryEntries.length - 1; i >= 0; --i) {
        var entry = this.tryEntries[i];
        if (entry.tryLoc === tryLoc) {
          var record = entry.completion;
          if (record.type === "throw") {
            var thrown = record.arg;
            resetTryEntry(entry);
          }
          return thrown;
        }
      }

      // The context.catch method must only be called with a location
      // argument that corresponds to a known catch block.
      throw new Error("illegal catch attempt");
    },

    delegateYield: function(iterable, resultName, nextLoc) {
      this.delegate = {
        iterator: values(iterable),
        resultName: resultName,
        nextLoc: nextLoc
      };

      if (this.method === "next") {
        // Deliberately forget the last sent value so that we don't
        // accidentally pass it on to the delegate.
        this.arg = undefined$1;
      }

      return ContinueSentinel;
    }
  };

  // Regardless of whether this script is executing as a CommonJS module
  // or not, return the runtime object so that we can declare the variable
  // regeneratorRuntime in the outer scope, which allows this module to be
  // injected easily by `bin/regenerator --include-runtime script.js`.
  return exports;

}(
  // If this script is executing as a CommonJS module, use module.exports
  // as the regeneratorRuntime namespace. Otherwise create a new empty
  // object. Either way, the resulting object will be used to initialize
  // the regeneratorRuntime variable at the top of this file.
   module.exports 
));

try {
  regeneratorRuntime = runtime;
} catch (accidentalStrictMode) {
  // This module should not be running in strict mode, so the above
  // assignment should always work unless something is misconfigured. Just
  // in case runtime.js accidentally runs in strict mode, we can escape
  // strict mode using a global Function call. This could conceivably fail
  // if a Content Security Policy forbids using Function, but in that case
  // the proper solution is to fix the accidental strict mode problem. If
  // you've misconfigured your bundler to force strict mode and applied a
  // CSP to forbid Function, and you're not willing to fix either of those
  // problems, please detail your unique predicament in a GitHub issue.
  Function("r", "regeneratorRuntime = r")(runtime);
}
});

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);
    if (enumerableOnly) symbols = symbols.filter(function (sym) {
      return Object.getOwnPropertyDescriptor(object, sym).enumerable;
    });
    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _isNativeReflectConstruct() {
  if (typeof Reflect === "undefined" || !Reflect.construct) return false;
  if (Reflect.construct.sham) return false;
  if (typeof Proxy === "function") return true;

  try {
    Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
    return true;
  } catch (e) {
    return false;
  }
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _createSuper(Derived) {
  var hasNativeReflectConstruct = _isNativeReflectConstruct();

  return function _createSuperInternal() {
    var Super = _getPrototypeOf(Derived),
        result;

    if (hasNativeReflectConstruct) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

var LogLevel;

(function (LogLevel) {
  LogLevel["Debug"] = "debug";
  LogLevel["Info"] = "info";
  LogLevel["Warn"] = "warn";
  LogLevel["Error"] = "error";
})(LogLevel || (LogLevel = {}));

var ErrorBoundary = /*#__PURE__*/function (_React$Component) {
  _inherits(ErrorBoundary, _React$Component);

  var _super = _createSuper(ErrorBoundary);

  function ErrorBoundary(props) {
    var _this;

    _classCallCheck(this, ErrorBoundary);

    _this = _super.call(this, props);
    _this.state = {
      hasError: false
    };
    return _this;
  }

  _createClass(ErrorBoundary, [{
    key: "componentDidCatch",
    value: function componentDidCatch(error, info) {
      var _this$props = this.props,
          logger = _this$props.logger,
          _this$props$loggingDa = _this$props.loggingData,
          _this$props$loggingDa2 = _this$props$loggingDa.message,
          message = _this$props$loggingDa2 === void 0 ? error.message : _this$props$loggingDa2,
          facility = _this$props$loggingDa.facility,
          _this$props$loggingDa3 = _this$props$loggingDa.level,
          level = _this$props$loggingDa3 === void 0 ? 'error' : _this$props$loggingDa3;
      this.setState({
        hasError: true
      });

      if (!logger) {
        return;
      }

      return logger.remote(level, message, _objectSpread2({
        errorMessage: error.message,
        stack: error.stack
      }, info), facility);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props2 = this.props,
          children = _this$props2.children,
          error = _this$props2.error;
      var hasError = this.state.hasError;

      if (hasError) {
        return error;
      }

      return children;
    }
  }]);

  return ErrorBoundary;
}(React__default.Component);

_defineProperty(ErrorBoundary, "defaultProps", {
  logger: null,
  loggingData: {
    message: 'Unknown Error',
    level: 'error',
    facility: 'unknown-facility'
  }
});

var fakeTranslator = {
  languageCode: 'en',
  gettext: function gettext(text) {
    return text;
  },
  dgettext: function dgettext(domain, text) {
    return text;
  },
  dcgettext: function dcgettext(domain, text, category) {
    return text;
  },
  ngettext: function ngettext(singularText, pluralText, count) {
    return singularText;
  },
  dngettext: function dngettext(domain, singularText, pluralText, count) {
    return singularText;
  },
  dcngettext: function dcngettext(domain, singularText, pluralText, count, category) {
    return singularText;
  },
  pgettext: function pgettext(context, text) {
    return text;
  },
  dpgettext: function dpgettext(domain, context, text) {
    return text;
  },
  npgettext: function npgettext(context, singularText, pluralText, count) {
    return singularText;
  },
  dnpgettext: function dnpgettext(domain, context, singularText, pluralText, count) {
    return singularText;
  },
  dcnpgettext: function dcnpgettext(domain, context, singularText, pluralText, count, category) {
    return singularText;
  },
  sprintf: function sprintf(text) {
    return '';
  }
};
var TranslatorContext = /*#__PURE__*/React__default.createContext( // HACK: This will allow the user to enter only `TranslatorClient`
// value. And we can still check if the context was inserted to
// the page properly.
null);
var useTranslator = function useTranslator() {
  var translatorContext = React.useContext(TranslatorContext); // To keep this hook backward compatible, we still have to return the fake translator

  if (translatorContext == null) {
    console.error('useTranslator must be used within a TranslatorContext');
    return fakeTranslator;
  }

  return translatorContext;
};

function _defineProperty$1(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

var defineProperty = _defineProperty$1;

function ownKeys$1(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys$1(Object(source), true).forEach(function (key) { defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys$1(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ActionTypes;

(function (ActionTypes) {
  ActionTypes["SetBadLeadFormValue"] = "SET_BAD_LEAD_FORM_VALUE_ACTION";
  ActionTypes["SetBadLeadFormHasSuccess"] = "SET_BAD_LEAD_FORM_HAS_SUCCESS_ACTION";
  ActionTypes["SetBadLeadFormHasFailure"] = "SET_BAD_LEAD_FORM_HAS_FAILURE_ACTION";
  ActionTypes["SetBadLeadFormInit"] = "SET_BAD_LEAD_FORM_INIT";
})(ActionTypes || (ActionTypes = {}));

var BadLeadFormReducer = function BadLeadFormReducer(state, action) {
  switch (action.type) {
    case ActionTypes.SetBadLeadFormValue:
      {
        var inputName = action.inputName,
            inputValue = action.inputValue;
        return _objectSpread(_objectSpread({}, state), {}, defineProperty({}, inputName, inputValue));
      }

    case ActionTypes.SetBadLeadFormHasSuccess:
      return _objectSpread(_objectSpread({}, state), {}, {
        hasSuccess: true
      });

    case ActionTypes.SetBadLeadFormHasFailure:
      return _objectSpread(_objectSpread({}, state), {}, {
        hasError: true
      });

    case ActionTypes.SetBadLeadFormInit:
      return _objectSpread(_objectSpread({}, action.defaults), {}, {
        hasError: false,
        hasSuccess: false,
        inProgress: false
      });

    default:
      return state;
  }
};

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

var arrayWithHoles = _arrayWithHoles;

function _iterableToArrayLimit(arr, i) {
  if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return;
  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

var iterableToArrayLimit = _iterableToArrayLimit;

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) {
    arr2[i] = arr[i];
  }

  return arr2;
}

var arrayLikeToArray = _arrayLikeToArray;

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return arrayLikeToArray(o, minLen);
}

var unsupportedIterableToArray = _unsupportedIterableToArray;

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}

var nonIterableRest = _nonIterableRest;

function _slicedToArray(arr, i) {
  return arrayWithHoles(arr) || iterableToArrayLimit(arr, i) || unsupportedIterableToArray(arr, i) || nonIterableRest();
}

var slicedToArray = _slicedToArray;

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
  try {
    var info = gen[key](arg);
    var value = info.value;
  } catch (error) {
    reject(error);
    return;
  }

  if (info.done) {
    resolve(value);
  } else {
    Promise.resolve(value).then(_next, _throw);
  }
}

function _asyncToGenerator(fn) {
  return function () {
    var self = this,
        args = arguments;
    return new Promise(function (resolve, reject) {
      var gen = fn.apply(self, args);

      function _next(value) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
      }

      function _throw(err) {
        asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
      }

      _next(undefined);
    });
  };
}

var asyncToGenerator = _asyncToGenerator;

var fe = createCommonjsModule(function (module, exports) {
!function(t,e){module.exports=e();}(window,(function(){return function(t){var e={};function n(r){if(e[r])return e[r].exports;var o=e[r]={i:r,l:!1,exports:{}};return t[r].call(o.exports,o,o.exports,n),o.l=!0,o.exports}return n.m=t,n.c=e,n.d=function(t,e,r){n.o(t,e)||Object.defineProperty(t,e,{enumerable:!0,get:r});},n.r=function(t){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(t,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(t,"__esModule",{value:!0});},n.t=function(t,e){if(1&e&&(t=n(t)),8&e)return t;if(4&e&&"object"==typeof t&&t&&t.__esModule)return t;var r=Object.create(null);if(n.r(r),Object.defineProperty(r,"default",{enumerable:!0,value:t}),2&e&&"string"!=typeof t)for(var o in t)n.d(r,o,function(e){return t[e]}.bind(null,o));return r},n.n=function(t){var e=t&&t.__esModule?function(){return t.default}:function(){return t};return n.d(e,"a",e),e},n.o=function(t,e){return Object.prototype.hasOwnProperty.call(t,e)},n.p="/",n(n.s=85)}([function(t,e,n){(function(e){var n=function(t){return t&&t.Math==Math&&t};t.exports=n("object"==typeof globalThis&&globalThis)||n("object"==typeof window&&window)||n("object"==typeof self&&self)||n("object"==typeof e&&e)||Function("return this")();}).call(this,n(86));},function(t,e){t.exports=function(t){try{return !!t()}catch(t){return !0}};},function(t,e,n){var r=n(0),o=n(36),i=n(4),c=n(28),u=n(40),a=n(64),s=o("wks"),f=r.Symbol,l=a?f:f&&f.withoutSetter||c;t.exports=function(t){return i(s,t)||(u&&i(f,t)?s[t]=f[t]:s[t]=l("Symbol."+t)),s[t]};},function(t,e){t.exports=function(t){return "object"==typeof t?null!==t:"function"==typeof t};},function(t,e){var n={}.hasOwnProperty;t.exports=function(t,e){return n.call(t,e)};},function(t,e,n){var r=n(3);t.exports=function(t){if(!r(t))throw TypeError(String(t)+" is not an object");return t};},function(t,e,n){var r=n(7),o=n(57),i=n(5),c=n(26),u=Object.defineProperty;e.f=r?u:function(t,e,n){if(i(t),e=c(e,!0),i(n),o)try{return u(t,e,n)}catch(t){}if("get"in n||"set"in n)throw TypeError("Accessors not supported");return "value"in n&&(t[e]=n.value),t};},function(t,e,n){var r=n(1);t.exports=!r((function(){return 7!=Object.defineProperty({},1,{get:function(){return 7}})[1]}));},function(t,e,n){var r=n(0),o=n(24).f,i=n(9),c=n(10),u=n(34),a=n(59),s=n(39);t.exports=function(t,e){var n,f,l,p,h,v=t.target,y=t.global,d=t.stat;if(n=y?r:d?r[v]||u(v,{}):(r[v]||{}).prototype)for(f in e){if(p=e[f],l=t.noTargetGet?(h=o(n,f))&&h.value:n[f],!s(y?f:v+(d?".":"#")+f,t.forced)&&void 0!==l){if(typeof p==typeof l)continue;a(p,l);}(t.sham||l&&l.sham)&&i(p,"sham",!0),c(n,f,p,t);}};},function(t,e,n){var r=n(7),o=n(6),i=n(17);t.exports=r?function(t,e,n){return o.f(t,e,i(1,n))}:function(t,e,n){return t[e]=n,t};},function(t,e,n){var r=n(0),o=n(9),i=n(4),c=n(34),u=n(35),a=n(14),s=a.get,f=a.enforce,l=String(String).split("String");(t.exports=function(t,e,n,u){var a=!!u&&!!u.unsafe,s=!!u&&!!u.enumerable,p=!!u&&!!u.noTargetGet;"function"==typeof n&&("string"!=typeof e||i(n,"name")||o(n,"name",e),f(n).source=l.join("string"==typeof e?e:"")),t!==r?(a?!p&&t[e]&&(s=!0):delete t[e],s?t[e]=n:o(t,e,n)):s?t[e]=n:c(e,n);})(Function.prototype,"toString",(function(){return "function"==typeof this&&s(this).source||u(this)}));},function(t,e){var n={}.toString;t.exports=function(t){return n.call(t).slice(8,-1)};},function(t,e,n){var r=n(60),o=n(0),i=function(t){return "function"==typeof t?t:void 0};t.exports=function(t,e){return arguments.length<2?i(r[t])||i(o[t]):r[t]&&r[t][e]||o[t]&&o[t][e]};},function(t,e,n){var r=n(56),o=n(25);t.exports=function(t){return r(o(t))};},function(t,e,n){var r,o,i,c=n(87),u=n(0),a=n(3),s=n(9),f=n(4),l=n(27),p=n(19),h=u.WeakMap;if(c){var v=new h,y=v.get,d=v.has,g=v.set;r=function(t,e){return g.call(v,t,e),e},o=function(t){return y.call(v,t)||{}},i=function(t){return d.call(v,t)};}else {var x=l("state");p[x]=!0,r=function(t,e){return s(t,x,e),e},o=function(t){return f(t,x)?t[x]:{}},i=function(t){return f(t,x)};}t.exports={set:r,get:o,has:i,enforce:function(t){return i(t)?o(t):r(t,{})},getterFor:function(t){return function(e){var n;if(!a(e)||(n=o(e)).type!==t)throw TypeError("Incompatible receiver, "+t+" required");return n}}};},function(t,e,n){var r=n(25);t.exports=function(t){return Object(r(t))};},function(t,e){t.exports=function(t){if("function"!=typeof t)throw TypeError(String(t)+" is not a function");return t};},function(t,e){t.exports=function(t,e){return {enumerable:!(1&t),configurable:!(2&t),writable:!(4&t),value:e}};},function(t,e){t.exports=!1;},function(t,e){t.exports={};},function(t,e,n){var r=n(29),o=Math.min;t.exports=function(t){return t>0?o(r(t),9007199254740991):0};},function(t,e,n){var r,o=n(5),i=n(90),c=n(38),u=n(19),a=n(66),s=n(33),f=n(27),l=f("IE_PROTO"),p=function(){},h=function(t){return "<script>"+t+"<\/script>"},v=function(){try{r=document.domain&&new ActiveXObject("htmlfile");}catch(t){}var t,e;v=r?function(t){t.write(h("")),t.close();var e=t.parentWindow.Object;return t=null,e}(r):((e=s("iframe")).style.display="none",a.appendChild(e),e.src=String("javascript:"),(t=e.contentWindow.document).open(),t.write(h("document.F=Object")),t.close(),t.F);for(var n=c.length;n--;)delete v.prototype[c[n]];return v()};u[l]=!0,t.exports=Object.create||function(t,e){var n;return null!==t?(p.prototype=o(t),n=new p,p.prototype=null,n[l]=t):n=v(),void 0===e?n:i(n,e)};},function(t,e,n){var r=n(6).f,o=n(4),i=n(2)("toStringTag");t.exports=function(t,e,n){t&&!o(t=n?t:t.prototype,i)&&r(t,i,{configurable:!0,value:e});};},function(t,e){t.exports={};},function(t,e,n){var r=n(7),o=n(55),i=n(17),c=n(13),u=n(26),a=n(4),s=n(57),f=Object.getOwnPropertyDescriptor;e.f=r?f:function(t,e){if(t=c(t),e=u(e,!0),s)try{return f(t,e)}catch(t){}if(a(t,e))return i(!o.f.call(t,e),t[e])};},function(t,e){t.exports=function(t){if(null==t)throw TypeError("Can't call method on "+t);return t};},function(t,e,n){var r=n(3);t.exports=function(t,e){if(!r(t))return t;var n,o;if(e&&"function"==typeof(n=t.toString)&&!r(o=n.call(t)))return o;if("function"==typeof(n=t.valueOf)&&!r(o=n.call(t)))return o;if(!e&&"function"==typeof(n=t.toString)&&!r(o=n.call(t)))return o;throw TypeError("Can't convert object to primitive value")};},function(t,e,n){var r=n(36),o=n(28),i=r("keys");t.exports=function(t){return i[t]||(i[t]=o(t))};},function(t,e){var n=0,r=Math.random();t.exports=function(t){return "Symbol("+String(void 0===t?"":t)+")_"+(++n+r).toString(36)};},function(t,e){var n=Math.ceil,r=Math.floor;t.exports=function(t){return isNaN(t=+t)?0:(t>0?r:n)(t)};},function(t,e,n){var r=n(16);t.exports=function(t,e,n){if(r(t),void 0===e)return t;switch(n){case 0:return function(){return t.call(e)};case 1:return function(n){return t.call(e,n)};case 2:return function(n,r){return t.call(e,n,r)};case 3:return function(n,r,o){return t.call(e,n,r,o)}}return function(){return t.apply(e,arguments)}};},function(t,e,n){var r=n(13),o=n(95),i=n(23),c=n(14),u=n(46),a=c.set,s=c.getterFor("Array Iterator");t.exports=u(Array,"Array",(function(t,e){a(this,{type:"Array Iterator",target:r(t),index:0,kind:e});}),(function(){var t=s(this),e=t.target,n=t.kind,r=t.index++;return !e||r>=e.length?(t.target=void 0,{value:void 0,done:!0}):"keys"==n?{value:r,done:!1}:"values"==n?{value:e[r],done:!1}:{value:[r,e[r]],done:!1}}),"values"),i.Arguments=i.Array,o("keys"),o("values"),o("entries");},function(t,e,n){var r=n(8),o=n(0),i=n(12),c=n(18),u=n(7),a=n(40),s=n(64),f=n(1),l=n(4),p=n(41),h=n(3),v=n(5),y=n(15),d=n(13),g=n(26),x=n(17),m=n(21),b=n(65),w=n(37),S=n(91),E=n(63),_=n(24),O=n(6),j=n(55),P=n(9),k=n(10),A=n(36),T=n(27),F=n(19),R=n(28),L=n(2),I=n(67),N=n(68),C=n(22),M=n(14),$=n(92).forEach,D=T("hidden"),U=L("toPrimitive"),G=M.set,z=M.getterFor("Symbol"),B=Object.prototype,V=o.Symbol,Y=i("JSON","stringify"),K=_.f,Q=O.f,W=S.f,H=j.f,X=A("symbols"),q=A("op-symbols"),J=A("string-to-symbol-registry"),Z=A("symbol-to-string-registry"),tt=A("wks"),et=o.QObject,nt=!et||!et.prototype||!et.prototype.findChild,rt=u&&f((function(){return 7!=m(Q({},"a",{get:function(){return Q(this,"a",{value:7}).a}})).a}))?function(t,e,n){var r=K(B,e);r&&delete B[e],Q(t,e,n),r&&t!==B&&Q(B,e,r);}:Q,ot=function(t,e){var n=X[t]=m(V.prototype);return G(n,{type:"Symbol",tag:t,description:e}),u||(n.description=e),n},it=s?function(t){return "symbol"==typeof t}:function(t){return Object(t)instanceof V},ct=function(t,e,n){t===B&&ct(q,e,n),v(t);var r=g(e,!0);return v(n),l(X,r)?(n.enumerable?(l(t,D)&&t[D][r]&&(t[D][r]=!1),n=m(n,{enumerable:x(0,!1)})):(l(t,D)||Q(t,D,x(1,{})),t[D][r]=!0),rt(t,r,n)):Q(t,r,n)},ut=function(t,e){v(t);var n=d(e),r=b(n).concat(lt(n));return $(r,(function(e){u&&!at.call(n,e)||ct(t,e,n[e]);})),t},at=function(t){var e=g(t,!0),n=H.call(this,e);return !(this===B&&l(X,e)&&!l(q,e))&&(!(n||!l(this,e)||!l(X,e)||l(this,D)&&this[D][e])||n)},st=function(t,e){var n=d(t),r=g(e,!0);if(n!==B||!l(X,r)||l(q,r)){var o=K(n,r);return !o||!l(X,r)||l(n,D)&&n[D][r]||(o.enumerable=!0),o}},ft=function(t){var e=W(d(t)),n=[];return $(e,(function(t){l(X,t)||l(F,t)||n.push(t);})),n},lt=function(t){var e=t===B,n=W(e?q:d(t)),r=[];return $(n,(function(t){!l(X,t)||e&&!l(B,t)||r.push(X[t]);})),r};(a||(k((V=function(){if(this instanceof V)throw TypeError("Symbol is not a constructor");var t=arguments.length&&void 0!==arguments[0]?String(arguments[0]):void 0,e=R(t),n=function(t){this===B&&n.call(q,t),l(this,D)&&l(this[D],e)&&(this[D][e]=!1),rt(this,e,x(1,t));};return u&&nt&&rt(B,e,{configurable:!0,set:n}),ot(e,t)}).prototype,"toString",(function(){return z(this).tag})),k(V,"withoutSetter",(function(t){return ot(R(t),t)})),j.f=at,O.f=ct,_.f=st,w.f=S.f=ft,E.f=lt,I.f=function(t){return ot(L(t),t)},u&&(Q(V.prototype,"description",{configurable:!0,get:function(){return z(this).description}}),c||k(B,"propertyIsEnumerable",at,{unsafe:!0}))),r({global:!0,wrap:!0,forced:!a,sham:!a},{Symbol:V}),$(b(tt),(function(t){N(t);})),r({target:"Symbol",stat:!0,forced:!a},{for:function(t){var e=String(t);if(l(J,e))return J[e];var n=V(e);return J[e]=n,Z[n]=e,n},keyFor:function(t){if(!it(t))throw TypeError(t+" is not a symbol");if(l(Z,t))return Z[t]},useSetter:function(){nt=!0;},useSimple:function(){nt=!1;}}),r({target:"Object",stat:!0,forced:!a,sham:!u},{create:function(t,e){return void 0===e?m(t):ut(m(t),e)},defineProperty:ct,defineProperties:ut,getOwnPropertyDescriptor:st}),r({target:"Object",stat:!0,forced:!a},{getOwnPropertyNames:ft,getOwnPropertySymbols:lt}),r({target:"Object",stat:!0,forced:f((function(){E.f(1);}))},{getOwnPropertySymbols:function(t){return E.f(y(t))}}),Y)&&r({target:"JSON",stat:!0,forced:!a||f((function(){var t=V();return "[null]"!=Y([t])||"{}"!=Y({a:t})||"{}"!=Y(Object(t))}))},{stringify:function(t,e,n){for(var r,o=[t],i=1;arguments.length>i;)o.push(arguments[i++]);if(r=e,(h(e)||void 0!==t)&&!it(t))return p(e)||(e=function(t,e){if("function"==typeof r&&(e=r.call(this,t,e)),!it(e))return e}),o[1]=e,Y.apply(null,o)}});V.prototype[U]||P(V.prototype,U,V.prototype.valueOf),C(V,"Symbol"),F[D]=!0;},function(t,e,n){var r=n(0),o=n(3),i=r.document,c=o(i)&&o(i.createElement);t.exports=function(t){return c?i.createElement(t):{}};},function(t,e,n){var r=n(0),o=n(9);t.exports=function(t,e){try{o(r,t,e);}catch(n){r[t]=e;}return e};},function(t,e,n){var r=n(58),o=Function.toString;"function"!=typeof r.inspectSource&&(r.inspectSource=function(t){return o.call(t)}),t.exports=r.inspectSource;},function(t,e,n){var r=n(18),o=n(58);(t.exports=function(t,e){return o[t]||(o[t]=void 0!==e?e:{})})("versions",[]).push({version:"3.6.4",mode:r?"pure":"global",copyright:"© 2020 Denis Pushkarev (zloirock.ru)"});},function(t,e,n){var r=n(61),o=n(38).concat("length","prototype");e.f=Object.getOwnPropertyNames||function(t){return r(t,o)};},function(t,e){t.exports=["constructor","hasOwnProperty","isPrototypeOf","propertyIsEnumerable","toLocaleString","toString","valueOf"];},function(t,e,n){var r=n(1),o=/#|\.prototype\./,i=function(t,e){var n=u[c(t)];return n==s||n!=a&&("function"==typeof e?r(e):!!e)},c=i.normalize=function(t){return String(t).replace(o,".").toLowerCase()},u=i.data={},a=i.NATIVE="N",s=i.POLYFILL="P";t.exports=i;},function(t,e,n){var r=n(1);t.exports=!!Object.getOwnPropertySymbols&&!r((function(){return !String(Symbol())}));},function(t,e,n){var r=n(11);t.exports=Array.isArray||function(t){return "Array"==r(t)};},function(t,e,n){var r=n(8),o=n(7),i=n(0),c=n(4),u=n(3),a=n(6).f,s=n(59),f=i.Symbol;if(o&&"function"==typeof f&&(!("description"in f.prototype)||void 0!==f().description)){var l={},p=function(){var t=arguments.length<1||void 0===arguments[0]?void 0:String(arguments[0]),e=this instanceof p?new f(t):void 0===t?f():f(t);return ""===t&&(l[e]=!0),e};s(p,f);var h=p.prototype=f.prototype;h.constructor=p;var v=h.toString,y="Symbol(test)"==String(f("test")),d=/^Symbol\((.*)\)[^)]+$/;a(h,"description",{configurable:!0,get:function(){var t=u(this)?this.valueOf():this,e=v.call(t);if(c(l,t))return "";var n=y?e.slice(7,-1):e.replace(d,"$1");return ""===n?void 0:n}}),r({global:!0,forced:!0},{Symbol:p});}},function(t,e,n){n(68)("iterator");},function(t,e,n){var r=n(8),o=n(1),i=n(41),c=n(3),u=n(15),a=n(20),s=n(93),f=n(69),l=n(94),p=n(2),h=n(45),v=p("isConcatSpreadable"),y=h>=51||!o((function(){var t=[];return t[v]=!1,t.concat()[0]!==t})),d=l("concat"),g=function(t){if(!c(t))return !1;var e=t[v];return void 0!==e?!!e:i(t)};r({target:"Array",proto:!0,forced:!y||!d},{concat:function(t){var e,n,r,o,i,c=u(this),l=f(c,0),p=0;for(e=-1,r=arguments.length;e<r;e++)if(i=-1===e?c:arguments[e],g(i)){if(p+(o=a(i.length))>9007199254740991)throw TypeError("Maximum allowed index exceeded");for(n=0;n<o;n++,p++)n in i&&s(l,p,i[n]);}else {if(p>=9007199254740991)throw TypeError("Maximum allowed index exceeded");s(l,p++,i);}return l.length=p,l}});},function(t,e,n){var r,o,i=n(0),c=n(70),u=i.process,a=u&&u.versions,s=a&&a.v8;s?o=(r=s.split("."))[0]+r[1]:c&&(!(r=c.match(/Edge\/(\d+)/))||r[1]>=74)&&(r=c.match(/Chrome\/(\d+)/))&&(o=r[1]),t.exports=o&&+o;},function(t,e,n){var r=n(8),o=n(96),i=n(47),c=n(73),u=n(22),a=n(9),s=n(10),f=n(2),l=n(18),p=n(23),h=n(71),v=h.IteratorPrototype,y=h.BUGGY_SAFARI_ITERATORS,d=f("iterator"),g=function(){return this};t.exports=function(t,e,n,f,h,x,m){o(n,e,f);var b,w,S,E=function(t){if(t===h&&k)return k;if(!y&&t in j)return j[t];switch(t){case"keys":case"values":case"entries":return function(){return new n(this,t)}}return function(){return new n(this)}},_=e+" Iterator",O=!1,j=t.prototype,P=j[d]||j["@@iterator"]||h&&j[h],k=!y&&P||E(h),A="Array"==e&&j.entries||P;if(A&&(b=i(A.call(new t)),v!==Object.prototype&&b.next&&(l||i(b)===v||(c?c(b,v):"function"!=typeof b[d]&&a(b,d,g)),u(b,_,!0,!0),l&&(p[_]=g))),"values"==h&&P&&"values"!==P.name&&(O=!0,k=function(){return P.call(this)}),l&&!m||j[d]===k||a(j,d,k),p[e]=k,h)if(w={values:E("values"),keys:x?k:E("keys"),entries:E("entries")},m)for(S in w)!y&&!O&&S in j||s(j,S,w[S]);else r({target:e,proto:!0,forced:y||O},w);return w};},function(t,e,n){var r=n(4),o=n(15),i=n(27),c=n(72),u=i("IE_PROTO"),a=Object.prototype;t.exports=c?Object.getPrototypeOf:function(t){return t=o(t),r(t,u)?t[u]:"function"==typeof t.constructor&&t instanceof t.constructor?t.constructor.prototype:t instanceof Object?a:null};},function(t,e,n){var r=n(49),o=n(10),i=n(98);r||o(Object.prototype,"toString",i,{unsafe:!0});},function(t,e,n){var r={};r[n(2)("toStringTag")]="z",t.exports="[object z]"===String(r);},function(t,e){t.exports=function(t,e,n){if(!(t instanceof e))throw TypeError("Incorrect "+(n?n+" ":"")+"invocation");return t};},function(t,e,n){var r=n(5),o=n(101),i=n(20),c=n(30),u=n(102),a=n(103),s=function(t,e){this.stopped=t,this.result=e;};(t.exports=function(t,e,n,f,l){var p,h,v,y,d,g,x,m=c(e,n,f?2:1);if(l)p=t;else {if("function"!=typeof(h=u(t)))throw TypeError("Target is not iterable");if(o(h)){for(v=0,y=i(t.length);y>v;v++)if((d=f?m(r(x=t[v])[0],x[1]):m(t[v]))&&d instanceof s)return d;return new s(!1)}p=h.call(t);}for(g=p.next;!(x=g.call(p)).done;)if("object"==typeof(d=a(p,m,x.value,f))&&d&&d instanceof s)return d;return new s(!1)}).stop=function(t){return new s(!0,t)};},function(t,e,n){var r=n(81).charAt,o=n(14),i=n(46),c=o.set,u=o.getterFor("String Iterator");i(String,"String",(function(t){c(this,{type:"String Iterator",string:String(t),index:0});}),(function(){var t,e=u(this),n=e.string,o=e.index;return o>=n.length?{value:void 0,done:!0}:(t=r(n,o),e.index+=t.length,{value:t,done:!1})}));},function(t,e,n){var r=n(0),o=n(109),i=n(31),c=n(9),u=n(2),a=u("iterator"),s=u("toStringTag"),f=i.values;for(var l in o){var p=r[l],h=p&&p.prototype;if(h){if(h[a]!==f)try{c(h,a,f);}catch(t){h[a]=f;}if(h[s]||c(h,s,l),o[l])for(var v in i)if(h[v]!==i[v])try{c(h,v,i[v]);}catch(t){h[v]=i[v];}}}},function(t,e,n){var r,o,i=n(83),c=n(112),u=RegExp.prototype.exec,a=String.prototype.replace,s=u,f=(r=/a/,o=/b*/g,u.call(r,"a"),u.call(o,"a"),0!==r.lastIndex||0!==o.lastIndex),l=c.UNSUPPORTED_Y||c.BROKEN_CARET,p=void 0!==/()??/.exec("")[1];(f||p||l)&&(s=function(t){var e,n,r,o,c=this,s=l&&c.sticky,h=i.call(c),v=c.source,y=0,d=t;return s&&(-1===(h=h.replace("y","")).indexOf("g")&&(h+="g"),d=String(t).slice(c.lastIndex),c.lastIndex>0&&(!c.multiline||c.multiline&&"\n"!==t[c.lastIndex-1])&&(v="(?: "+v+")",d=" "+d,y++),n=new RegExp("^(?:"+v+")",h)),p&&(n=new RegExp("^"+v+"$(?!\\s)",h)),f&&(e=c.lastIndex),r=u.call(s?n:c,d),s?r?(r.input=r.input.slice(y),r[0]=r[0].slice(y),r.index=c.lastIndex,c.lastIndex+=r[0].length):c.lastIndex=0:f&&r&&(c.lastIndex=c.global?r.index+r[0].length:e),p&&r&&r.length>1&&a.call(r[0],n,(function(){for(o=1;o<arguments.length-2;o++)void 0===arguments[o]&&(r[o]=void 0);})),r}),t.exports=s;},function(t,e,n){var r={}.propertyIsEnumerable,o=Object.getOwnPropertyDescriptor,i=o&&!r.call({1:2},1);e.f=i?function(t){var e=o(this,t);return !!e&&e.enumerable}:r;},function(t,e,n){var r=n(1),o=n(11),i="".split;t.exports=r((function(){return !Object("z").propertyIsEnumerable(0)}))?function(t){return "String"==o(t)?i.call(t,""):Object(t)}:Object;},function(t,e,n){var r=n(7),o=n(1),i=n(33);t.exports=!r&&!o((function(){return 7!=Object.defineProperty(i("div"),"a",{get:function(){return 7}}).a}));},function(t,e,n){var r=n(0),o=n(34),i=r["__core-js_shared__"]||o("__core-js_shared__",{});t.exports=i;},function(t,e,n){var r=n(4),o=n(88),i=n(24),c=n(6);t.exports=function(t,e){for(var n=o(e),u=c.f,a=i.f,s=0;s<n.length;s++){var f=n[s];r(t,f)||u(t,f,a(e,f));}};},function(t,e,n){var r=n(0);t.exports=r;},function(t,e,n){var r=n(4),o=n(13),i=n(62).indexOf,c=n(19);t.exports=function(t,e){var n,u=o(t),a=0,s=[];for(n in u)!r(c,n)&&r(u,n)&&s.push(n);for(;e.length>a;)r(u,n=e[a++])&&(~i(s,n)||s.push(n));return s};},function(t,e,n){var r=n(13),o=n(20),i=n(89),c=function(t){return function(e,n,c){var u,a=r(e),s=o(a.length),f=i(c,s);if(t&&n!=n){for(;s>f;)if((u=a[f++])!=u)return !0}else for(;s>f;f++)if((t||f in a)&&a[f]===n)return t||f||0;return !t&&-1}};t.exports={includes:c(!0),indexOf:c(!1)};},function(t,e){e.f=Object.getOwnPropertySymbols;},function(t,e,n){var r=n(40);t.exports=r&&!Symbol.sham&&"symbol"==typeof Symbol.iterator;},function(t,e,n){var r=n(61),o=n(38);t.exports=Object.keys||function(t){return r(t,o)};},function(t,e,n){var r=n(12);t.exports=r("document","documentElement");},function(t,e,n){var r=n(2);e.f=r;},function(t,e,n){var r=n(60),o=n(4),i=n(67),c=n(6).f;t.exports=function(t){var e=r.Symbol||(r.Symbol={});o(e,t)||c(e,t,{value:i.f(t)});};},function(t,e,n){var r=n(3),o=n(41),i=n(2)("species");t.exports=function(t,e){var n;return o(t)&&("function"!=typeof(n=t.constructor)||n!==Array&&!o(n.prototype)?r(n)&&null===(n=n[i])&&(n=void 0):n=void 0),new(void 0===n?Array:n)(0===e?0:e)};},function(t,e,n){var r=n(12);t.exports=r("navigator","userAgent")||"";},function(t,e,n){var r,o,i,c=n(47),u=n(9),a=n(4),s=n(2),f=n(18),l=s("iterator"),p=!1;[].keys&&("next"in(i=[].keys())?(o=c(c(i)))!==Object.prototype&&(r=o):p=!0),null==r&&(r={}),f||a(r,l)||u(r,l,(function(){return this})),t.exports={IteratorPrototype:r,BUGGY_SAFARI_ITERATORS:p};},function(t,e,n){var r=n(1);t.exports=!r((function(){function t(){}return t.prototype.constructor=null,Object.getPrototypeOf(new t)!==t.prototype}));},function(t,e,n){var r=n(5),o=n(97);t.exports=Object.setPrototypeOf||("__proto__"in{}?function(){var t,e=!1,n={};try{(t=Object.getOwnPropertyDescriptor(Object.prototype,"__proto__").set).call(n,[]),e=n instanceof Array;}catch(t){}return function(n,i){return r(n),o(i),e?t.call(n,i):n.__proto__=i,n}}():void 0);},function(t,e,n){var r=n(49),o=n(11),i=n(2)("toStringTag"),c="Arguments"==o(function(){return arguments}());t.exports=r?o:function(t){var e,n,r;return void 0===t?"Undefined":null===t?"Null":"string"==typeof(n=function(t,e){try{return t[e]}catch(t){}}(e=Object(t),i))?n:c?o(e):"Object"==(r=o(e))&&"function"==typeof e.callee?"Arguments":r};},function(t,e,n){var r=n(10);t.exports=function(t,e,n){for(var o in e)r(t,o,e[o],n);return t};},function(t,e,n){var r=n(12),o=n(6),i=n(2),c=n(7),u=i("species");t.exports=function(t){var e=r(t),n=o.f;c&&e&&!e[u]&&n(e,u,{configurable:!0,get:function(){return this}});};},function(t,e,n){var r=n(2)("iterator"),o=!1;try{var i=0,c={next:function(){return {done:!!i++}},return:function(){o=!0;}};c[r]=function(){return this},Array.from(c,(function(){throw 2}));}catch(t){}t.exports=function(t,e){if(!e&&!o)return !1;var n=!1;try{var i={};i[r]=function(){return {next:function(){return {done:n=!0}}}},t(i);}catch(t){}return n};},function(t,e,n){var r,o,i,c=n(0),u=n(1),a=n(11),s=n(30),f=n(66),l=n(33),p=n(79),h=c.location,v=c.setImmediate,y=c.clearImmediate,d=c.process,g=c.MessageChannel,x=c.Dispatch,m=0,b={},w=function(t){if(b.hasOwnProperty(t)){var e=b[t];delete b[t],e();}},S=function(t){return function(){w(t);}},E=function(t){w(t.data);},_=function(t){c.postMessage(t+"",h.protocol+"//"+h.host);};v&&y||(v=function(t){for(var e=[],n=1;arguments.length>n;)e.push(arguments[n++]);return b[++m]=function(){("function"==typeof t?t:Function(t)).apply(void 0,e);},r(m),m},y=function(t){delete b[t];},"process"==a(d)?r=function(t){d.nextTick(S(t));}:x&&x.now?r=function(t){x.now(S(t));}:g&&!p?(i=(o=new g).port2,o.port1.onmessage=E,r=s(i.postMessage,i,1)):!c.addEventListener||"function"!=typeof postMessage||c.importScripts||u(_)?r="onreadystatechange"in l("script")?function(t){f.appendChild(l("script")).onreadystatechange=function(){f.removeChild(this),w(t);};}:function(t){setTimeout(S(t),0);}:(r=_,c.addEventListener("message",E,!1))),t.exports={set:v,clear:y};},function(t,e,n){var r=n(70);t.exports=/(iphone|ipod|ipad).*applewebkit/i.test(r);},function(t,e,n){var r=n(16),o=function(t){var e,n;this.promise=new t((function(t,r){if(void 0!==e||void 0!==n)throw TypeError("Bad Promise constructor");e=t,n=r;})),this.resolve=r(e),this.reject=r(n);};t.exports.f=function(t){return new o(t)};},function(t,e,n){var r=n(29),o=n(25),i=function(t){return function(e,n){var i,c,u=String(o(e)),a=r(n),s=u.length;return a<0||a>=s?t?"":void 0:(i=u.charCodeAt(a))<55296||i>56319||a+1===s||(c=u.charCodeAt(a+1))<56320||c>57343?t?u.charAt(a):i:t?u.slice(a,a+2):c-56320+(i-55296<<10)+65536}};t.exports={codeAt:i(!1),charAt:i(!0)};},function(t,e,n){var r=n(8),o=n(54);r({target:"RegExp",proto:!0,forced:/./.exec!==o},{exec:o});},function(t,e,n){var r=n(5);t.exports=function(){var t=r(this),e="";return t.global&&(e+="g"),t.ignoreCase&&(e+="i"),t.multiline&&(e+="m"),t.dotAll&&(e+="s"),t.unicode&&(e+="u"),t.sticky&&(e+="y"),e};},function(t,e,n){var r=n(19),o=n(3),i=n(4),c=n(6).f,u=n(28),a=n(125),s=u("meta"),f=0,l=Object.isExtensible||function(){return !0},p=function(t){c(t,s,{value:{objectID:"O"+ ++f,weakData:{}}});},h=t.exports={REQUIRED:!1,fastKey:function(t,e){if(!o(t))return "symbol"==typeof t?t:("string"==typeof t?"S":"P")+t;if(!i(t,s)){if(!l(t))return "F";if(!e)return "E";p(t);}return t[s].objectID},getWeakData:function(t,e){if(!i(t,s)){if(!l(t))return !0;if(!e)return !1;p(t);}return t[s].weakData},onFreeze:function(t){return a&&h.REQUIRED&&l(t)&&!i(t,s)&&p(t),t}};r[s]=!0;},function(t,e,n){function r(t){return (r="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(t){return typeof t}:function(t){return t&&"function"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?"symbol":typeof t})(t)}function o(t,e,n,r,o,i,c){try{var u=t[i](c),a=u.value;}catch(t){return void n(t)}u.done?e(a):Promise.resolve(a).then(r,o);}function i(t){return function(){var e=this,n=arguments;return new Promise((function(r,i){var c=t.apply(e,n);function u(t){o(c,r,i,u,a,"next",t);}function a(t){o(c,r,i,u,a,"throw",t);}u(void 0);}))}}n(32),n(42),n(43),n(44),n(31),n(48),n(99),n(52),n(53),n(110);var c=n(111),u=n(118),a=function(t,e){var n=arguments.length>2&&void 0!==arguments[2]?arguments[2]:{},r=n.assetsDomain,o=void 0===r?"pipedriveassets.com":r,i=n.cacheTag,c=void 0===i?"latest":i,u=n.services,a=(u=void 0===u?{}:u)[t],s=(a=void 0===a?{}:a).cacheSum,f=void 0===s?"latest":s;return "https://cdn.".concat(o,"/translator-hub/").concat(t,"/").concat(e,"?v=").concat(c,"&langv=").concat(f)},s=function(t){return "en"===t||"en-US"===t||"function"!=typeof fetch},f=function(){var t=i(regeneratorRuntime.mark((function t(e){var n;return regeneratorRuntime.wrap((function(t){for(;;)switch(t.prev=t.next){case 0:return t.prev=0,t.next=3,fetch(e);case 3:n=t.sent,t.next=11;break;case 6:return t.prev=6,t.t0=t.catch(0),t.next=10,fetch(e);case 10:n=t.sent;case 11:if(n.ok){t.next=18;break}return t.t1=u,t.next=15,n.text();case 15:throw t.t2=t.sent,t.t3=n,new t.t1(t.t2,t.t3);case 18:return t.abrupt("return",n.json());case 19:case"end":return t.stop()}}),t,null,[[0,6]])})));return function(e){return t.apply(this,arguments)}}();t.exports=function(){var t=i(regeneratorRuntime.mark((function t(e,n,o){var i,u,l;return regeneratorRuntime.wrap((function(t){for(;;)switch(t.prev=t.next){case 0:if("object"===("undefined"==typeof window?"undefined":r(window))){t.next=2;break}throw Error("Cannot create a translator instance unless we are in the browser");case 2:if("object"===r(e)&&(e=(i=e).serviceName,n=i.language,o=i.logger),u="",t.prev=4,!s(n)){t.next=7;break}return t.abrupt("return",new c);case 7:return "function"==typeof o&&(o=o(e)),u=a(e,n,window.pdTranslatorConfig),t.next=11,f(u);case 11:return l=t.sent,t.abrupt("return",new c(l));case 15:return t.prev=15,t.t0=t.catch(4),"object"===r(o)&&o.logError&&o.logError(t.t0,"Failed building translations client","error",{serviceName:e,hubUrl:u,requestedLanguage:n,errMessage:t.t0&&"object"===r(t.t0)?t.t0.message:'*translator-client unknown error* "'.concat(r(t.t0),'"')}),t.abrupt("return",new c);case 19:case"end":return t.stop()}}),t,null,[[4,15]])})));return function(e,n,r){return t.apply(this,arguments)}}();},function(t,e){var n;n=function(){return this}();try{n=n||new Function("return this")();}catch(t){"object"==typeof window&&(n=window);}t.exports=n;},function(t,e,n){var r=n(0),o=n(35),i=r.WeakMap;t.exports="function"==typeof i&&/native code/.test(o(i));},function(t,e,n){var r=n(12),o=n(37),i=n(63),c=n(5);t.exports=r("Reflect","ownKeys")||function(t){var e=o.f(c(t)),n=i.f;return n?e.concat(n(t)):e};},function(t,e,n){var r=n(29),o=Math.max,i=Math.min;t.exports=function(t,e){var n=r(t);return n<0?o(n+e,0):i(n,e)};},function(t,e,n){var r=n(7),o=n(6),i=n(5),c=n(65);t.exports=r?Object.defineProperties:function(t,e){i(t);for(var n,r=c(e),u=r.length,a=0;u>a;)o.f(t,n=r[a++],e[n]);return t};},function(t,e,n){var r=n(13),o=n(37).f,i={}.toString,c="object"==typeof window&&window&&Object.getOwnPropertyNames?Object.getOwnPropertyNames(window):[];t.exports.f=function(t){return c&&"[object Window]"==i.call(t)?function(t){try{return o(t)}catch(t){return c.slice()}}(t):o(r(t))};},function(t,e,n){var r=n(30),o=n(56),i=n(15),c=n(20),u=n(69),a=[].push,s=function(t){var e=1==t,n=2==t,s=3==t,f=4==t,l=6==t,p=5==t||l;return function(h,v,y,d){for(var g,x,m=i(h),b=o(m),w=r(v,y,3),S=c(b.length),E=0,_=d||u,O=e?_(h,S):n?_(h,0):void 0;S>E;E++)if((p||E in b)&&(x=w(g=b[E],E,m),t))if(e)O[E]=x;else if(x)switch(t){case 3:return !0;case 5:return g;case 6:return E;case 2:a.call(O,g);}else if(f)return !1;return l?-1:s||f?f:O}};t.exports={forEach:s(0),map:s(1),filter:s(2),some:s(3),every:s(4),find:s(5),findIndex:s(6)};},function(t,e,n){var r=n(26),o=n(6),i=n(17);t.exports=function(t,e,n){var c=r(e);c in t?o.f(t,c,i(0,n)):t[c]=n;};},function(t,e,n){var r=n(1),o=n(2),i=n(45),c=o("species");t.exports=function(t){return i>=51||!r((function(){var e=[];return (e.constructor={})[c]=function(){return {foo:1}},1!==e[t](Boolean).foo}))};},function(t,e,n){var r=n(2),o=n(21),i=n(6),c=r("unscopables"),u=Array.prototype;null==u[c]&&i.f(u,c,{configurable:!0,value:o(null)}),t.exports=function(t){u[c][t]=!0;};},function(t,e,n){var r=n(71).IteratorPrototype,o=n(21),i=n(17),c=n(22),u=n(23),a=function(){return this};t.exports=function(t,e,n){var s=e+" Iterator";return t.prototype=o(r,{next:i(1,n)}),c(t,s,!1,!0),u[s]=a,t};},function(t,e,n){var r=n(3);t.exports=function(t){if(!r(t)&&null!==t)throw TypeError("Can't set "+String(t)+" as a prototype");return t};},function(t,e,n){var r=n(49),o=n(74);t.exports=r?{}.toString:function(){return "[object "+o(this)+"]"};},function(t,e,n){var r,o,i,c,u=n(8),a=n(18),s=n(0),f=n(12),l=n(100),p=n(10),h=n(75),v=n(22),y=n(76),d=n(3),g=n(16),x=n(50),m=n(11),b=n(35),w=n(51),S=n(77),E=n(104),_=n(78).set,O=n(105),j=n(106),P=n(107),k=n(80),A=n(108),T=n(14),F=n(39),R=n(2),L=n(45),I=R("species"),N="Promise",C=T.get,M=T.set,$=T.getterFor(N),D=l,U=s.TypeError,G=s.document,z=s.process,B=f("fetch"),V=k.f,Y=V,K="process"==m(z),Q=!!(G&&G.createEvent&&s.dispatchEvent),W=F(N,(function(){if(!(b(D)!==String(D))){if(66===L)return !0;if(!K&&"function"!=typeof PromiseRejectionEvent)return !0}if(a&&!D.prototype.finally)return !0;if(L>=51&&/native code/.test(D))return !1;var t=D.resolve(1),e=function(t){t((function(){}),(function(){}));};return (t.constructor={})[I]=e,!(t.then((function(){}))instanceof e)})),H=W||!S((function(t){D.all(t).catch((function(){}));})),X=function(t){var e;return !(!d(t)||"function"!=typeof(e=t.then))&&e},q=function(t,e,n){if(!e.notified){e.notified=!0;var r=e.reactions;O((function(){for(var o=e.value,i=1==e.state,c=0;r.length>c;){var u,a,s,f=r[c++],l=i?f.ok:f.fail,p=f.resolve,h=f.reject,v=f.domain;try{l?(i||(2===e.rejection&&et(t,e),e.rejection=1),!0===l?u=o:(v&&v.enter(),u=l(o),v&&(v.exit(),s=!0)),u===f.promise?h(U("Promise-chain cycle")):(a=X(u))?a.call(u,p,h):p(u)):h(o);}catch(t){v&&!s&&v.exit(),h(t);}}e.reactions=[],e.notified=!1,n&&!e.rejection&&Z(t,e);}));}},J=function(t,e,n){var r,o;Q?((r=G.createEvent("Event")).promise=e,r.reason=n,r.initEvent(t,!1,!0),s.dispatchEvent(r)):r={promise:e,reason:n},(o=s["on"+t])?o(r):"unhandledrejection"===t&&P("Unhandled promise rejection",n);},Z=function(t,e){_.call(s,(function(){var n,r=e.value;if(tt(e)&&(n=A((function(){K?z.emit("unhandledRejection",r,t):J("unhandledrejection",t,r);})),e.rejection=K||tt(e)?2:1,n.error))throw n.value}));},tt=function(t){return 1!==t.rejection&&!t.parent},et=function(t,e){_.call(s,(function(){K?z.emit("rejectionHandled",t):J("rejectionhandled",t,e.value);}));},nt=function(t,e,n,r){return function(o){t(e,n,o,r);}},rt=function(t,e,n,r){e.done||(e.done=!0,r&&(e=r),e.value=n,e.state=2,q(t,e,!0));},ot=function(t,e,n,r){if(!e.done){e.done=!0,r&&(e=r);try{if(t===n)throw U("Promise can't be resolved itself");var o=X(n);o?O((function(){var r={done:!1};try{o.call(n,nt(ot,t,r,e),nt(rt,t,r,e));}catch(n){rt(t,r,n,e);}})):(e.value=n,e.state=1,q(t,e,!1));}catch(n){rt(t,{done:!1},n,e);}}};W&&(D=function(t){x(this,D,N),g(t),r.call(this);var e=C(this);try{t(nt(ot,this,e),nt(rt,this,e));}catch(t){rt(this,e,t);}},(r=function(t){M(this,{type:N,done:!1,notified:!1,parent:!1,reactions:[],rejection:!1,state:0,value:void 0});}).prototype=h(D.prototype,{then:function(t,e){var n=$(this),r=V(E(this,D));return r.ok="function"!=typeof t||t,r.fail="function"==typeof e&&e,r.domain=K?z.domain:void 0,n.parent=!0,n.reactions.push(r),0!=n.state&&q(this,n,!1),r.promise},catch:function(t){return this.then(void 0,t)}}),o=function(){var t=new r,e=C(t);this.promise=t,this.resolve=nt(ot,t,e),this.reject=nt(rt,t,e);},k.f=V=function(t){return t===D||t===i?new o(t):Y(t)},a||"function"!=typeof l||(c=l.prototype.then,p(l.prototype,"then",(function(t,e){var n=this;return new D((function(t,e){c.call(n,t,e);})).then(t,e)}),{unsafe:!0}),"function"==typeof B&&u({global:!0,enumerable:!0,forced:!0},{fetch:function(t){return j(D,B.apply(s,arguments))}}))),u({global:!0,wrap:!0,forced:W},{Promise:D}),v(D,N,!1,!0),y(N),i=f(N),u({target:N,stat:!0,forced:W},{reject:function(t){var e=V(this);return e.reject.call(void 0,t),e.promise}}),u({target:N,stat:!0,forced:a||W},{resolve:function(t){return j(a&&this===i?D:this,t)}}),u({target:N,stat:!0,forced:H},{all:function(t){var e=this,n=V(e),r=n.resolve,o=n.reject,i=A((function(){var n=g(e.resolve),i=[],c=0,u=1;w(t,(function(t){var a=c++,s=!1;i.push(void 0),u++,n.call(e,t).then((function(t){s||(s=!0,i[a]=t,--u||r(i));}),o);})),--u||r(i);}));return i.error&&o(i.value),n.promise},race:function(t){var e=this,n=V(e),r=n.reject,o=A((function(){var o=g(e.resolve);w(t,(function(t){o.call(e,t).then(n.resolve,r);}));}));return o.error&&r(o.value),n.promise}});},function(t,e,n){var r=n(0);t.exports=r.Promise;},function(t,e,n){var r=n(2),o=n(23),i=r("iterator"),c=Array.prototype;t.exports=function(t){return void 0!==t&&(o.Array===t||c[i]===t)};},function(t,e,n){var r=n(74),o=n(23),i=n(2)("iterator");t.exports=function(t){if(null!=t)return t[i]||t["@@iterator"]||o[r(t)]};},function(t,e,n){var r=n(5);t.exports=function(t,e,n,o){try{return o?e(r(n)[0],n[1]):e(n)}catch(e){var i=t.return;throw void 0!==i&&r(i.call(t)),e}};},function(t,e,n){var r=n(5),o=n(16),i=n(2)("species");t.exports=function(t,e){var n,c=r(t).constructor;return void 0===c||null==(n=r(c)[i])?e:o(n)};},function(t,e,n){var r,o,i,c,u,a,s,f,l=n(0),p=n(24).f,h=n(11),v=n(78).set,y=n(79),d=l.MutationObserver||l.WebKitMutationObserver,g=l.process,x=l.Promise,m="process"==h(g),b=p(l,"queueMicrotask"),w=b&&b.value;w||(r=function(){var t,e;for(m&&(t=g.domain)&&t.exit();o;){e=o.fn,o=o.next;try{e();}catch(t){throw o?c():i=void 0,t}}i=void 0,t&&t.enter();},m?c=function(){g.nextTick(r);}:d&&!y?(u=!0,a=document.createTextNode(""),new d(r).observe(a,{characterData:!0}),c=function(){a.data=u=!u;}):x&&x.resolve?(s=x.resolve(void 0),f=s.then,c=function(){f.call(s,r);}):c=function(){v.call(l,r);}),t.exports=w||function(t){var e={fn:t,next:void 0};i&&(i.next=e),o||(o=e,c()),i=e;};},function(t,e,n){var r=n(5),o=n(3),i=n(80);t.exports=function(t,e){if(r(t),o(e)&&e.constructor===t)return e;var n=i.f(t);return (0, n.resolve)(e),n.promise};},function(t,e,n){var r=n(0);t.exports=function(t,e){var n=r.console;n&&n.error&&(1===arguments.length?n.error(t):n.error(t,e));};},function(t,e){t.exports=function(t){try{return {error:!1,value:t()}}catch(t){return {error:!0,value:t}}};},function(t,e){t.exports={CSSRuleList:0,CSSStyleDeclaration:0,CSSValueList:0,ClientRectList:0,DOMRectList:0,DOMStringList:0,DOMTokenList:1,DataTransferItemList:0,FileList:0,HTMLAllCollection:0,HTMLCollection:0,HTMLFormElement:0,HTMLSelectElement:0,MediaList:0,MimeTypeArray:0,NamedNodeMap:0,NodeList:1,PaintRequestList:0,Plugin:0,PluginArray:0,SVGLengthList:0,SVGNumberList:0,SVGPathSegList:0,SVGPointList:0,SVGStringList:0,SVGTransformList:0,SourceBufferList:0,StyleSheetList:0,TextTrackCueList:0,TextTrackList:0,TouchList:0};},function(t,e,n){var r=function(t){var e=Object.prototype,n=e.hasOwnProperty,r="function"==typeof Symbol?Symbol:{},o=r.iterator||"@@iterator",i=r.asyncIterator||"@@asyncIterator",c=r.toStringTag||"@@toStringTag";function u(t,e,n,r){var o=e&&e.prototype instanceof f?e:f,i=Object.create(o.prototype),c=new S(r||[]);return i._invoke=function(t,e,n){var r="suspendedStart";return function(o,i){if("executing"===r)throw new Error("Generator is already running");if("completed"===r){if("throw"===o)throw i;return _()}for(n.method=o,n.arg=i;;){var c=n.delegate;if(c){var u=m(c,n);if(u){if(u===s)continue;return u}}if("next"===n.method)n.sent=n._sent=n.arg;else if("throw"===n.method){if("suspendedStart"===r)throw r="completed",n.arg;n.dispatchException(n.arg);}else "return"===n.method&&n.abrupt("return",n.arg);r="executing";var f=a(t,e,n);if("normal"===f.type){if(r=n.done?"completed":"suspendedYield",f.arg===s)continue;return {value:f.arg,done:n.done}}"throw"===f.type&&(r="completed",n.method="throw",n.arg=f.arg);}}}(t,n,c),i}function a(t,e,n){try{return {type:"normal",arg:t.call(e,n)}}catch(t){return {type:"throw",arg:t}}}t.wrap=u;var s={};function f(){}function l(){}function p(){}var h={};h[o]=function(){return this};var v=Object.getPrototypeOf,y=v&&v(v(E([])));y&&y!==e&&n.call(y,o)&&(h=y);var d=p.prototype=f.prototype=Object.create(h);function g(t){["next","throw","return"].forEach((function(e){t[e]=function(t){return this._invoke(e,t)};}));}function x(t){var e;this._invoke=function(r,o){function i(){return new Promise((function(e,i){!function e(r,o,i,c){var u=a(t[r],t,o);if("throw"!==u.type){var s=u.arg,f=s.value;return f&&"object"==typeof f&&n.call(f,"__await")?Promise.resolve(f.__await).then((function(t){e("next",t,i,c);}),(function(t){e("throw",t,i,c);})):Promise.resolve(f).then((function(t){s.value=t,i(s);}),(function(t){return e("throw",t,i,c)}))}c(u.arg);}(r,o,e,i);}))}return e=e?e.then(i,i):i()};}function m(t,e){var n=t.iterator[e.method];if(void 0===n){if(e.delegate=null,"throw"===e.method){if(t.iterator.return&&(e.method="return",e.arg=void 0,m(t,e),"throw"===e.method))return s;e.method="throw",e.arg=new TypeError("The iterator does not provide a 'throw' method");}return s}var r=a(n,t.iterator,e.arg);if("throw"===r.type)return e.method="throw",e.arg=r.arg,e.delegate=null,s;var o=r.arg;return o?o.done?(e[t.resultName]=o.value,e.next=t.nextLoc,"return"!==e.method&&(e.method="next",e.arg=void 0),e.delegate=null,s):o:(e.method="throw",e.arg=new TypeError("iterator result is not an object"),e.delegate=null,s)}function b(t){var e={tryLoc:t[0]};1 in t&&(e.catchLoc=t[1]),2 in t&&(e.finallyLoc=t[2],e.afterLoc=t[3]),this.tryEntries.push(e);}function w(t){var e=t.completion||{};e.type="normal",delete e.arg,t.completion=e;}function S(t){this.tryEntries=[{tryLoc:"root"}],t.forEach(b,this),this.reset(!0);}function E(t){if(t){var e=t[o];if(e)return e.call(t);if("function"==typeof t.next)return t;if(!isNaN(t.length)){var r=-1,i=function e(){for(;++r<t.length;)if(n.call(t,r))return e.value=t[r],e.done=!1,e;return e.value=void 0,e.done=!0,e};return i.next=i}}return {next:_}}function _(){return {value:void 0,done:!0}}return l.prototype=d.constructor=p,p.constructor=l,p[c]=l.displayName="GeneratorFunction",t.isGeneratorFunction=function(t){var e="function"==typeof t&&t.constructor;return !!e&&(e===l||"GeneratorFunction"===(e.displayName||e.name))},t.mark=function(t){return Object.setPrototypeOf?Object.setPrototypeOf(t,p):(t.__proto__=p,c in t||(t[c]="GeneratorFunction")),t.prototype=Object.create(d),t},t.awrap=function(t){return {__await:t}},g(x.prototype),x.prototype[i]=function(){return this},t.AsyncIterator=x,t.async=function(e,n,r,o){var i=new x(u(e,n,r,o));return t.isGeneratorFunction(n)?i:i.next().then((function(t){return t.done?t.value:i.next()}))},g(d),d[c]="Generator",d[o]=function(){return this},d.toString=function(){return "[object Generator]"},t.keys=function(t){var e=[];for(var n in t)e.push(n);return e.reverse(),function n(){for(;e.length;){var r=e.pop();if(r in t)return n.value=r,n.done=!1,n}return n.done=!0,n}},t.values=E,S.prototype={constructor:S,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=void 0,this.done=!1,this.delegate=null,this.method="next",this.arg=void 0,this.tryEntries.forEach(w),!t)for(var e in this)"t"===e.charAt(0)&&n.call(this,e)&&!isNaN(+e.slice(1))&&(this[e]=void 0);},stop:function(){this.done=!0;var t=this.tryEntries[0].completion;if("throw"===t.type)throw t.arg;return this.rval},dispatchException:function(t){if(this.done)throw t;var e=this;function r(n,r){return c.type="throw",c.arg=t,e.next=n,r&&(e.method="next",e.arg=void 0),!!r}for(var o=this.tryEntries.length-1;o>=0;--o){var i=this.tryEntries[o],c=i.completion;if("root"===i.tryLoc)return r("end");if(i.tryLoc<=this.prev){var u=n.call(i,"catchLoc"),a=n.call(i,"finallyLoc");if(u&&a){if(this.prev<i.catchLoc)return r(i.catchLoc,!0);if(this.prev<i.finallyLoc)return r(i.finallyLoc)}else if(u){if(this.prev<i.catchLoc)return r(i.catchLoc,!0)}else {if(!a)throw new Error("try statement without catch or finally");if(this.prev<i.finallyLoc)return r(i.finallyLoc)}}}},abrupt:function(t,e){for(var r=this.tryEntries.length-1;r>=0;--r){var o=this.tryEntries[r];if(o.tryLoc<=this.prev&&n.call(o,"finallyLoc")&&this.prev<o.finallyLoc){var i=o;break}}i&&("break"===t||"continue"===t)&&i.tryLoc<=e&&e<=i.finallyLoc&&(i=null);var c=i?i.completion:{};return c.type=t,c.arg=e,i?(this.method="next",this.next=i.finallyLoc,s):this.complete(c)},complete:function(t,e){if("throw"===t.type)throw t.arg;return "break"===t.type||"continue"===t.type?this.next=t.arg:"return"===t.type?(this.rval=this.arg=t.arg,this.method="return",this.next="end"):"normal"===t.type&&e&&(this.next=e),s},finish:function(t){for(var e=this.tryEntries.length-1;e>=0;--e){var n=this.tryEntries[e];if(n.finallyLoc===t)return this.complete(n.completion,n.afterLoc),w(n),s}},catch:function(t){for(var e=this.tryEntries.length-1;e>=0;--e){var n=this.tryEntries[e];if(n.tryLoc===t){var r=n.completion;if("throw"===r.type){var o=r.arg;w(n);}return o}}throw new Error("illegal catch attempt")},delegateYield:function(t,e,n){return this.delegate={iterator:E(t),resultName:e,nextLoc:n},"next"===this.method&&(this.arg=void 0),s}},t}(t.exports);try{regeneratorRuntime=r;}catch(t){Function("r","regeneratorRuntime = r")(r);}},function(t,e,n){function r(t){return (r="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(t){return typeof t}:function(t){return t&&"function"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?"symbol":typeof t})(t)}n(32),n(42),n(43),n(44),n(31),n(48),n(82),n(52),n(113),n(53);var o=n(117);function i(){var t=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};this.options=t,this.jed=new o(this.options);}i.prototype.setOptions=function(t){"object"===r(t)&&(this.options=t,this.jed=new o(this.options));},i.prototype.getOptions=function(){return this.options},i.prototype.getLanguage=function(){var t=(((this.options||{}).locale_data||{}).messages||{})[""]||{};return t.lang&&"string"==typeof t.lang?t.lang:null},i.prototype.getPdLanguage=function(){var t=this.getLanguage();return t?t.replace("_","-"):t},i.prototype.getJed=function(){return this.jed},i.prototype.translate=function(t,e){for(var n=arguments.length,r=new Array(n>2?n-2:0),o=2;o<n;o++)r[o-2]=arguments[o];var i=r[e];return void 0===i?this.jed[t].apply(this.jed,r):this.sprintf(this.jed[t].apply(this.jed,r),i)},i.prototype.gettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["gettext",1].concat(e))},i.prototype.dgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dgettext",2].concat(e))},i.prototype.dcgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dcgettext",3].concat(e))},i.prototype.ngettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["ngettext",3].concat(e))},i.prototype.dngettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dngettext",4].concat(e))},i.prototype.dcngettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dcngettext",5].concat(e))},i.prototype.pgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["pgettext",2].concat(e))},i.prototype.dpgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dpgettext",3].concat(e))},i.prototype.npgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["npgettext",4].concat(e))},i.prototype.dnpgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dnpgettext",5].concat(e))},i.prototype.dcnpgettext=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];return this.translate.apply(this,["dcnpgettext",5].concat(e))},i.prototype.sprintf=function(){for(var t=arguments.length,e=new Array(t),n=0;n<t;n++)e[n]=arguments[n];try{return this.jed.sprintf.apply(this.jed,e)}catch(t){return "function"==typeof this.options.onError&&this.options.onError(t,e),e[0]}},t.exports=i;},function(t,e,n){var r=n(1);function o(t,e){return RegExp(t,e)}e.UNSUPPORTED_Y=r((function(){var t=o("a","y");return t.lastIndex=2,null!=t.exec("abcd")})),e.BROKEN_CARET=r((function(){var t=o("^r","gy");return t.lastIndex=2,null!=t.exec("str")}));},function(t,e,n){var r=n(114),o=n(5),i=n(15),c=n(20),u=n(29),a=n(25),s=n(115),f=n(116),l=Math.max,p=Math.min,h=Math.floor,v=/\$([$&'`]|\d\d?|<[^>]*>)/g,y=/\$([$&'`]|\d\d?)/g;r("replace",2,(function(t,e,n,r){var d=r.REGEXP_REPLACE_SUBSTITUTES_UNDEFINED_CAPTURE,g=r.REPLACE_KEEPS_$0,x=d?"$":"$0";return [function(n,r){var o=a(this),i=null==n?void 0:n[t];return void 0!==i?i.call(n,o,r):e.call(String(o),n,r)},function(t,r){if(!d&&g||"string"==typeof r&&-1===r.indexOf(x)){var i=n(e,t,this,r);if(i.done)return i.value}var a=o(t),h=String(this),v="function"==typeof r;v||(r=String(r));var y=a.global;if(y){var b=a.unicode;a.lastIndex=0;}for(var w=[];;){var S=f(a,h);if(null===S)break;if(w.push(S),!y)break;""===String(S[0])&&(a.lastIndex=s(h,c(a.lastIndex),b));}for(var E,_="",O=0,j=0;j<w.length;j++){S=w[j];for(var P=String(S[0]),k=l(p(u(S.index),h.length),0),A=[],T=1;T<S.length;T++)A.push(void 0===(E=S[T])?E:String(E));var F=S.groups;if(v){var R=[P].concat(A,k,h);void 0!==F&&R.push(F);var L=String(r.apply(void 0,R));}else L=m(P,h,k,A,F,r);k>=O&&(_+=h.slice(O,k)+L,O=k+P.length);}return _+h.slice(O)}];function m(t,n,r,o,c,u){var a=r+t.length,s=o.length,f=y;return void 0!==c&&(c=i(c),f=v),e.call(u,f,(function(e,i){var u;switch(i.charAt(0)){case"$":return "$";case"&":return t;case"`":return n.slice(0,r);case"'":return n.slice(a);case"<":u=c[i.slice(1,-1)];break;default:var f=+i;if(0===f)return e;if(f>s){var l=h(f/10);return 0===l?e:l<=s?void 0===o[l-1]?i.charAt(1):o[l-1]+i.charAt(1):e}u=o[f-1];}return void 0===u?"":u}))}}));},function(t,e,n){n(82);var r=n(10),o=n(1),i=n(2),c=n(54),u=n(9),a=i("species"),s=!o((function(){var t=/./;return t.exec=function(){var t=[];return t.groups={a:"7"},t},"7"!=="".replace(t,"$<a>")})),f="$0"==="a".replace(/./,"$0"),l=i("replace"),p=!!/./[l]&&""===/./[l]("a","$0"),h=!o((function(){var t=/(?:)/,e=t.exec;t.exec=function(){return e.apply(this,arguments)};var n="ab".split(t);return 2!==n.length||"a"!==n[0]||"b"!==n[1]}));t.exports=function(t,e,n,l){var v=i(t),y=!o((function(){var e={};return e[v]=function(){return 7},7!=""[t](e)})),d=y&&!o((function(){var e=!1,n=/a/;return "split"===t&&((n={}).constructor={},n.constructor[a]=function(){return n},n.flags="",n[v]=/./[v]),n.exec=function(){return e=!0,null},n[v](""),!e}));if(!y||!d||"replace"===t&&(!s||!f||p)||"split"===t&&!h){var g=/./[v],x=n(v,""[t],(function(t,e,n,r,o){return e.exec===c?y&&!o?{done:!0,value:g.call(e,n,r)}:{done:!0,value:t.call(n,e,r)}:{done:!1}}),{REPLACE_KEEPS_$0:f,REGEXP_REPLACE_SUBSTITUTES_UNDEFINED_CAPTURE:p}),m=x[0],b=x[1];r(String.prototype,t,m),r(RegExp.prototype,v,2==e?function(t,e){return b.call(t,this,e)}:function(t){return b.call(t,this)});}l&&u(RegExp.prototype[v],"sham",!0);};},function(t,e,n){var r=n(81).charAt;t.exports=function(t,e,n){return e+(n?r(t,e).length:1)};},function(t,e,n){var r=n(11),o=n(54);t.exports=function(t,e){var n=t.exec;if("function"==typeof n){var i=n.call(t,e);if("object"!=typeof i)throw TypeError("RegExp exec method returned something other than an Object or null");return i}if("RegExp"!==r(t))throw TypeError("RegExp#exec called on incompatible receiver");return o.call(t,e)};},function(t,e,n){
/**
 * @preserve jed.js v0.5.0beta https://github.com/SlexAxton/Jed
 */
!function(n,r){var o=Array.prototype,i=Object.prototype,c=o.slice,u=i.hasOwnProperty,a=o.forEach,s={},f={forEach:function(t,e,n){var r,o,i;if(null!==t)if(a&&t.forEach===a)t.forEach(e,n);else if(t.length===+t.length){for(r=0,o=t.length;r<o;r++)if(r in t&&e.call(n,t[r],r,t)===s)return}else for(i in t)if(u.call(t,i)&&e.call(n,t[i],i,t)===s)return},extend:function(t){return this.forEach(c.call(arguments,1),(function(e){for(var n in e)t[n]=e[n];})),t}},l=function(t){if(this.defaults={locale_data:{messages:{"":{domain:"messages",lang:"en",plural_forms:"nplurals=2; plural=(n != 1);"}}},domain:"messages",debug:!1},this.options=f.extend({},this.defaults,t),this.textdomain(this.options.domain),t.domain&&!this.options.locale_data[this.options.domain])throw new Error("Text domain set to non-existent domain: `"+t.domain+"`")};function p(t){return l.PF.compile(t||"nplurals=2; plural=(n != 1);")}function h(t,e){this._key=t,this._i18n=e;}l.context_delimiter=String.fromCharCode(4),f.extend(h.prototype,{onDomain:function(t){return this._domain=t,this},withContext:function(t){return this._context=t,this},ifPlural:function(t,e){return this._val=t,this._pkey=e,this},fetch:function(t){return "[object Array]"!={}.toString.call(t)&&(t=[].slice.call(arguments,0)),(t&&t.length?l.sprintf:function(t){return t})(this._i18n.dcnpgettext(this._domain,this._context,this._key,this._pkey,this._val),t)}}),f.extend(l.prototype,{translate:function(t){return new h(t,this)},textdomain:function(t){if(!t)return this._textdomain;this._textdomain=t;},gettext:function(t){return this.dcnpgettext.call(this,void 0,void 0,t)},dgettext:function(t,e){return this.dcnpgettext.call(this,t,void 0,e)},dcgettext:function(t,e){return this.dcnpgettext.call(this,t,void 0,e)},ngettext:function(t,e,n){return this.dcnpgettext.call(this,void 0,void 0,t,e,n)},dngettext:function(t,e,n,r){return this.dcnpgettext.call(this,t,void 0,e,n,r)},dcngettext:function(t,e,n,r){return this.dcnpgettext.call(this,t,void 0,e,n,r)},pgettext:function(t,e){return this.dcnpgettext.call(this,void 0,t,e)},dpgettext:function(t,e,n){return this.dcnpgettext.call(this,t,e,n)},dcpgettext:function(t,e,n){return this.dcnpgettext.call(this,t,e,n)},npgettext:function(t,e,n,r){return this.dcnpgettext.call(this,void 0,t,e,n,r)},dnpgettext:function(t,e,n,r,o){return this.dcnpgettext.call(this,t,e,n,r,o)},dcnpgettext:function(t,e,n,r,o){var i;if(r=r||n,t=t||this._textdomain,!this.options)return (i=new l).dcnpgettext.call(i,void 0,void 0,n,r,o);if(!this.options.locale_data)throw new Error("No locale data provided.");if(!this.options.locale_data[t])throw new Error("Domain `"+t+"` was not found.");if(!this.options.locale_data[t][""])throw new Error("No locale meta information provided.");if(!n)throw new Error("No translation key found.");var c,u,a,s=e?e+l.context_delimiter+n:n,f=this.options.locale_data,h=f[t],v=(f.messages||this.defaults.locale_data.messages)[""],y=h[""].plural_forms||h[""]["Plural-Forms"]||h[""]["plural-forms"]||v.plural_forms||v["Plural-Forms"]||v["plural-forms"];if(void 0===o)a=1;else {if("number"!=typeof o&&(o=parseInt(o,10),isNaN(o)))throw new Error("The number that was passed in is not a number.");a=p(y)(o)+1;}if(!h)throw new Error("No domain named `"+t+"` could be found.");return !(c=h[s])||a>=c.length?(this.options.missing_key_callback&&this.options.missing_key_callback(s,t),u=[null,n,r],!0===this.options.debug&&console.log(u[p(y)(o)+1]),u[p()(o)+1]):(u=c[a])||(u=[null,n,r])[p()(o)+1]}});var v,y,d=function(){function t(t){return Object.prototype.toString.call(t).slice(8,-1).toLowerCase()}function e(t,e){for(var n=[];e>0;n[--e]=t);return n.join("")}var n=function(){return n.cache.hasOwnProperty(arguments[0])||(n.cache[arguments[0]]=n.parse(arguments[0])),n.format.call(null,n.cache[arguments[0]],arguments)};return n.format=function(n,r){var o,i,c,u,a,s,f,l=1,p=n.length,h="",v=[];for(i=0;i<p;i++)if("string"===(h=t(n[i])))v.push(n[i]);else if("array"===h){if((u=n[i])[2])for(o=r[l],c=0;c<u[2].length;c++){if(!o.hasOwnProperty(u[2][c]))throw d('[sprintf] property "%s" does not exist',u[2][c]);o=o[u[2][c]];}else o=u[1]?r[u[1]]:r[l++];if(/[^s]/.test(u[8])&&"number"!=t(o))throw d("[sprintf] expecting number but found %s",t(o));switch(null==o&&(o=""),u[8]){case"b":o=o.toString(2);break;case"c":o=String.fromCharCode(o);break;case"d":o=parseInt(o,10);break;case"e":o=u[7]?o.toExponential(u[7]):o.toExponential();break;case"f":o=u[7]?parseFloat(o).toFixed(u[7]):parseFloat(o);break;case"o":o=o.toString(8);break;case"s":o=(o=String(o))&&u[7]?o.substring(0,u[7]):o;break;case"u":o=Math.abs(o);break;case"x":o=o.toString(16);break;case"X":o=o.toString(16).toUpperCase();}o=/[def]/.test(u[8])&&u[3]&&o>=0?"+"+o:o,s=u[4]?"0"==u[4]?"0":u[4].charAt(1):" ",f=u[6]-String(o).length,a=u[6]?e(s,f):"",v.push(u[5]?o+a:a+o);}return v.join("")},n.cache={},n.parse=function(t){for(var e=t,n=[],r=[],o=0;e;){if(null!==(n=/^[^\x25]+/.exec(e)))r.push(n[0]);else if(null!==(n=/^\x25{2}/.exec(e)))r.push("%");else {if(null===(n=/^\x25(?:([1-9]\d*)\$|\(([^\)]+)\))?(\+)?(0|'[^$])?(-)?(\d+)?(?:\.(\d+))?([b-fosuxX])/.exec(e)))throw "[sprintf] huh?";if(n[2]){o|=1;var i=[],c=n[2],u=[];if(null===(u=/^([a-z_][a-z_\d]*)/i.exec(c)))throw "[sprintf] huh?";for(i.push(u[1]);""!==(c=c.substring(u[0].length));)if(null!==(u=/^\.([a-z_][a-z_\d]*)/i.exec(c)))i.push(u[1]);else {if(null===(u=/^\[(\d+)\]/.exec(c)))throw "[sprintf] huh?";i.push(u[1]);}n[2]=i;}else o|=2;if(3===o)throw "[sprintf] mixing positional and named placeholders is not (yet) supported";r.push(n);}e=e.substring(n[0].length);}return r},n}(),g=function(t,e){return e.unshift(t),d.apply(null,e)};l.parse_plural=function(t,e){return t=t.replace(/n/g,e),l.parse_expression(t)},l.sprintf=function(t,e){return "[object Array]"=={}.toString.call(e)?g(t,[].slice.call(e)):d.apply(this,[].slice.call(arguments))},l.prototype.sprintf=function(){return l.sprintf.apply(this,arguments)},(l.PF={}).parse=function(t){var e=l.PF.extractPluralExpr(t);return l.PF.parser.parse.call(l.PF.parser,e)},l.PF.compile=function(t){var e=l.PF.parse(t);return function(t){return !0===(n=l.PF.interpreter(e)(t))?1:n||0;var n;}},l.PF.interpreter=function(t){return function(e){switch(t.type){case"GROUP":return l.PF.interpreter(t.expr)(e);case"TERNARY":return l.PF.interpreter(t.expr)(e)?l.PF.interpreter(t.truthy)(e):l.PF.interpreter(t.falsey)(e);case"OR":return l.PF.interpreter(t.left)(e)||l.PF.interpreter(t.right)(e);case"AND":return l.PF.interpreter(t.left)(e)&&l.PF.interpreter(t.right)(e);case"LT":return l.PF.interpreter(t.left)(e)<l.PF.interpreter(t.right)(e);case"GT":return l.PF.interpreter(t.left)(e)>l.PF.interpreter(t.right)(e);case"LTE":return l.PF.interpreter(t.left)(e)<=l.PF.interpreter(t.right)(e);case"GTE":return l.PF.interpreter(t.left)(e)>=l.PF.interpreter(t.right)(e);case"EQ":return l.PF.interpreter(t.left)(e)==l.PF.interpreter(t.right)(e);case"NEQ":return l.PF.interpreter(t.left)(e)!=l.PF.interpreter(t.right)(e);case"MOD":return l.PF.interpreter(t.left)(e)%l.PF.interpreter(t.right)(e);case"VAR":return e;case"NUM":return t.val;default:throw new Error("Invalid Token found.")}}},l.PF.extractPluralExpr=function(t){t=t.replace(/^\s\s*/,"").replace(/\s\s*$/,""),/;\s*$/.test(t)||(t=t.concat(";"));var e,n=/nplurals\=(\d+);/,r=t.match(n);if(!(r.length>1))throw new Error("nplurals not found in plural_forms string: "+t);if(r[1],!((e=(t=t.replace(n,"")).match(/plural\=(.*);/))&&e.length>1))throw new Error("`plural` expression not found: "+t);return e[1]},l.PF.parser=(v={trace:function(){},yy:{},symbols_:{error:2,expressions:3,e:4,EOF:5,"?":6,":":7,"||":8,"&&":9,"<":10,"<=":11,">":12,">=":13,"!=":14,"==":15,"%":16,"(":17,")":18,n:19,NUMBER:20,$accept:0,$end:1},terminals_:{2:"error",5:"EOF",6:"?",7:":",8:"||",9:"&&",10:"<",11:"<=",12:">",13:">=",14:"!=",15:"==",16:"%",17:"(",18:")",19:"n",20:"NUMBER"},productions_:[0,[3,2],[4,5],[4,3],[4,3],[4,3],[4,3],[4,3],[4,3],[4,3],[4,3],[4,3],[4,3],[4,1],[4,1]],performAction:function(t,e,n,r,o,i,c){var u=i.length-1;switch(o){case 1:return {type:"GROUP",expr:i[u-1]};case 2:this.$={type:"TERNARY",expr:i[u-4],truthy:i[u-2],falsey:i[u]};break;case 3:this.$={type:"OR",left:i[u-2],right:i[u]};break;case 4:this.$={type:"AND",left:i[u-2],right:i[u]};break;case 5:this.$={type:"LT",left:i[u-2],right:i[u]};break;case 6:this.$={type:"LTE",left:i[u-2],right:i[u]};break;case 7:this.$={type:"GT",left:i[u-2],right:i[u]};break;case 8:this.$={type:"GTE",left:i[u-2],right:i[u]};break;case 9:this.$={type:"NEQ",left:i[u-2],right:i[u]};break;case 10:this.$={type:"EQ",left:i[u-2],right:i[u]};break;case 11:this.$={type:"MOD",left:i[u-2],right:i[u]};break;case 12:this.$={type:"GROUP",expr:i[u-1]};break;case 13:this.$={type:"VAR"};break;case 14:this.$={type:"NUM",val:Number(t)};}},table:[{3:1,4:2,17:[1,3],19:[1,4],20:[1,5]},{1:[3]},{5:[1,6],6:[1,7],8:[1,8],9:[1,9],10:[1,10],11:[1,11],12:[1,12],13:[1,13],14:[1,14],15:[1,15],16:[1,16]},{4:17,17:[1,3],19:[1,4],20:[1,5]},{5:[2,13],6:[2,13],7:[2,13],8:[2,13],9:[2,13],10:[2,13],11:[2,13],12:[2,13],13:[2,13],14:[2,13],15:[2,13],16:[2,13],18:[2,13]},{5:[2,14],6:[2,14],7:[2,14],8:[2,14],9:[2,14],10:[2,14],11:[2,14],12:[2,14],13:[2,14],14:[2,14],15:[2,14],16:[2,14],18:[2,14]},{1:[2,1]},{4:18,17:[1,3],19:[1,4],20:[1,5]},{4:19,17:[1,3],19:[1,4],20:[1,5]},{4:20,17:[1,3],19:[1,4],20:[1,5]},{4:21,17:[1,3],19:[1,4],20:[1,5]},{4:22,17:[1,3],19:[1,4],20:[1,5]},{4:23,17:[1,3],19:[1,4],20:[1,5]},{4:24,17:[1,3],19:[1,4],20:[1,5]},{4:25,17:[1,3],19:[1,4],20:[1,5]},{4:26,17:[1,3],19:[1,4],20:[1,5]},{4:27,17:[1,3],19:[1,4],20:[1,5]},{6:[1,7],8:[1,8],9:[1,9],10:[1,10],11:[1,11],12:[1,12],13:[1,13],14:[1,14],15:[1,15],16:[1,16],18:[1,28]},{6:[1,7],7:[1,29],8:[1,8],9:[1,9],10:[1,10],11:[1,11],12:[1,12],13:[1,13],14:[1,14],15:[1,15],16:[1,16]},{5:[2,3],6:[2,3],7:[2,3],8:[2,3],9:[1,9],10:[1,10],11:[1,11],12:[1,12],13:[1,13],14:[1,14],15:[1,15],16:[1,16],18:[2,3]},{5:[2,4],6:[2,4],7:[2,4],8:[2,4],9:[2,4],10:[1,10],11:[1,11],12:[1,12],13:[1,13],14:[1,14],15:[1,15],16:[1,16],18:[2,4]},{5:[2,5],6:[2,5],7:[2,5],8:[2,5],9:[2,5],10:[2,5],11:[2,5],12:[2,5],13:[2,5],14:[2,5],15:[2,5],16:[1,16],18:[2,5]},{5:[2,6],6:[2,6],7:[2,6],8:[2,6],9:[2,6],10:[2,6],11:[2,6],12:[2,6],13:[2,6],14:[2,6],15:[2,6],16:[1,16],18:[2,6]},{5:[2,7],6:[2,7],7:[2,7],8:[2,7],9:[2,7],10:[2,7],11:[2,7],12:[2,7],13:[2,7],14:[2,7],15:[2,7],16:[1,16],18:[2,7]},{5:[2,8],6:[2,8],7:[2,8],8:[2,8],9:[2,8],10:[2,8],11:[2,8],12:[2,8],13:[2,8],14:[2,8],15:[2,8],16:[1,16],18:[2,8]},{5:[2,9],6:[2,9],7:[2,9],8:[2,9],9:[2,9],10:[2,9],11:[2,9],12:[2,9],13:[2,9],14:[2,9],15:[2,9],16:[1,16],18:[2,9]},{5:[2,10],6:[2,10],7:[2,10],8:[2,10],9:[2,10],10:[2,10],11:[2,10],12:[2,10],13:[2,10],14:[2,10],15:[2,10],16:[1,16],18:[2,10]},{5:[2,11],6:[2,11],7:[2,11],8:[2,11],9:[2,11],10:[2,11],11:[2,11],12:[2,11],13:[2,11],14:[2,11],15:[2,11],16:[2,11],18:[2,11]},{5:[2,12],6:[2,12],7:[2,12],8:[2,12],9:[2,12],10:[2,12],11:[2,12],12:[2,12],13:[2,12],14:[2,12],15:[2,12],16:[2,12],18:[2,12]},{4:30,17:[1,3],19:[1,4],20:[1,5]},{5:[2,2],6:[1,7],7:[2,2],8:[1,8],9:[1,9],10:[1,10],11:[1,11],12:[1,12],13:[1,13],14:[1,14],15:[1,15],16:[1,16],18:[2,2]}],defaultActions:{6:[2,1]},parseError:function(t,e){throw new Error(t)},parse:function(t){var e=this,n=[0],r=[null],o=[],i=this.table,c="",u=0,a=0,s=0;this.lexer.setInput(t),this.lexer.yy=this.yy,this.yy.lexer=this.lexer,void 0===this.lexer.yylloc&&(this.lexer.yylloc={});var f=this.lexer.yylloc;function l(){var t;return "number"!=typeof(t=e.lexer.lex()||1)&&(t=e.symbols_[t]||t),t}o.push(f),"function"==typeof this.yy.parseError&&(this.parseError=this.yy.parseError);for(var p,h,v,y,d,g,x,m,b,w,S={};;){if(v=n[n.length-1],this.defaultActions[v]?y=this.defaultActions[v]:(null==p&&(p=l()),y=i[v]&&i[v][p]),void 0===y||!y.length||!y[0]){if(!s){for(g in b=[],i[v])this.terminals_[g]&&g>2&&b.push("'"+this.terminals_[g]+"'");var E="";E=this.lexer.showPosition?"Parse error on line "+(u+1)+":\n"+this.lexer.showPosition()+"\nExpecting "+b.join(", ")+", got '"+this.terminals_[p]+"'":"Parse error on line "+(u+1)+": Unexpected "+(1==p?"end of input":"'"+(this.terminals_[p]||p)+"'"),this.parseError(E,{text:this.lexer.match,token:this.terminals_[p]||p,line:this.lexer.yylineno,loc:f,expected:b});}if(3==s){if(1==p)throw new Error(E||"Parsing halted.");a=this.lexer.yyleng,c=this.lexer.yytext,u=this.lexer.yylineno,f=this.lexer.yylloc,p=l();}for(;!(2..toString()in i[v]);){if(0==v)throw new Error(E||"Parsing halted.");w=1,n.length=n.length-2*w,r.length=r.length-w,o.length=o.length-w,v=n[n.length-1];}h=p,p=2,y=i[v=n[n.length-1]]&&i[v][2],s=3;}if(y[0]instanceof Array&&y.length>1)throw new Error("Parse Error: multiple actions possible at state: "+v+", token: "+p);switch(y[0]){case 1:n.push(p),r.push(this.lexer.yytext),o.push(this.lexer.yylloc),n.push(y[1]),p=null,h?(p=h,h=null):(a=this.lexer.yyleng,c=this.lexer.yytext,u=this.lexer.yylineno,f=this.lexer.yylloc,s>0&&s--);break;case 2:if(x=this.productions_[y[1]][1],S.$=r[r.length-x],S._$={first_line:o[o.length-(x||1)].first_line,last_line:o[o.length-1].last_line,first_column:o[o.length-(x||1)].first_column,last_column:o[o.length-1].last_column},void 0!==(d=this.performAction.call(S,c,a,u,this.yy,y[1],r,o)))return d;x&&(n=n.slice(0,-1*x*2),r=r.slice(0,-1*x),o=o.slice(0,-1*x)),n.push(this.productions_[y[1]][0]),r.push(S.$),o.push(S._$),m=i[n[n.length-2]][n[n.length-1]],n.push(m);break;case 3:return !0}}return !0}},y=function(){var t={EOF:1,parseError:function(t,e){if(!this.yy.parseError)throw new Error(t);this.yy.parseError(t,e);},setInput:function(t){return this._input=t,this._more=this._less=this.done=!1,this.yylineno=this.yyleng=0,this.yytext=this.matched=this.match="",this.conditionStack=["INITIAL"],this.yylloc={first_line:1,first_column:0,last_line:1,last_column:0},this},input:function(){var t=this._input[0];return this.yytext+=t,this.yyleng++,this.match+=t,this.matched+=t,t.match(/\n/)&&this.yylineno++,this._input=this._input.slice(1),t},unput:function(t){return this._input=t+this._input,this},more:function(){return this._more=!0,this},pastInput:function(){var t=this.matched.substr(0,this.matched.length-this.match.length);return (t.length>20?"...":"")+t.substr(-20).replace(/\n/g,"")},upcomingInput:function(){var t=this.match;return t.length<20&&(t+=this._input.substr(0,20-t.length)),(t.substr(0,20)+(t.length>20?"...":"")).replace(/\n/g,"")},showPosition:function(){var t=this.pastInput(),e=new Array(t.length+1).join("-");return t+this.upcomingInput()+"\n"+e+"^"},next:function(){if(this.done)return this.EOF;var t,e;this._input||(this.done=!0),this._more||(this.yytext="",this.match="");for(var n=this._currentRules(),r=0;r<n.length;r++)if(t=this._input.match(this.rules[n[r]]))return (e=t[0].match(/\n.*/g))&&(this.yylineno+=e.length),this.yylloc={first_line:this.yylloc.last_line,last_line:this.yylineno+1,first_column:this.yylloc.last_column,last_column:e?e[e.length-1].length-1:this.yylloc.last_column+t[0].length},this.yytext+=t[0],this.match+=t[0],this.matches=t,this.yyleng=this.yytext.length,this._more=!1,this._input=this._input.slice(t[0].length),this.matched+=t[0],this.performAction.call(this,this.yy,this,n[r],this.conditionStack[this.conditionStack.length-1])||void 0;if(""===this._input)return this.EOF;this.parseError("Lexical error on line "+(this.yylineno+1)+". Unrecognized text.\n"+this.showPosition(),{text:"",token:null,line:this.yylineno});},lex:function(){var t=this.next();return void 0!==t?t:this.lex()},begin:function(t){this.conditionStack.push(t);},popState:function(){return this.conditionStack.pop()},_currentRules:function(){return this.conditions[this.conditionStack[this.conditionStack.length-1]].rules},topState:function(){return this.conditionStack[this.conditionStack.length-2]},pushState:function(t){this.begin(t);},performAction:function(t,e,n,r){switch(n){case 0:break;case 1:return 20;case 2:return 19;case 3:return 8;case 4:return 9;case 5:return 6;case 6:return 7;case 7:return 11;case 8:return 13;case 9:return 10;case 10:return 12;case 11:return 14;case 12:return 15;case 13:return 16;case 14:return 17;case 15:return 18;case 16:return 5;case 17:return "INVALID"}},rules:[/^\s+/,/^[0-9]+(\.[0-9]+)?\b/,/^n\b/,/^\|\|/,/^&&/,/^\?/,/^:/,/^<=/,/^>=/,/^</,/^>/,/^!=/,/^==/,/^%/,/^\(/,/^\)/,/^$/,/^./],conditions:{INITIAL:{rules:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17],inclusive:!0}}};return t}(),v.lexer=y,v),t.exports&&(e=t.exports=l),e.Jed=l;}();},function(t,e,n){function r(t){return (r="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(t){return typeof t}:function(t){return t&&"function"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?"symbol":typeof t})(t)}function o(t,e){if(!(t instanceof e))throw new TypeError("Cannot call a class as a function")}function i(t,e){return !e||"object"!==r(e)&&"function"!=typeof e?function(t){if(void 0===t)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return t}(t):e}function c(t){var e="function"==typeof Map?new Map:void 0;return (c=function(t){if(null===t||(n=t,-1===Function.toString.call(n).indexOf("[native code]")))return t;var n;if("function"!=typeof t)throw new TypeError("Super expression must either be null or a function");if(void 0!==e){if(e.has(t))return e.get(t);e.set(t,r);}function r(){return a(t,arguments,f(this).constructor)}return r.prototype=Object.create(t.prototype,{constructor:{value:r,enumerable:!1,writable:!0,configurable:!0}}),s(r,t)})(t)}function u(){if("undefined"==typeof Reflect||!Reflect.construct)return !1;if(Reflect.construct.sham)return !1;if("function"==typeof Proxy)return !0;try{return Date.prototype.toString.call(Reflect.construct(Date,[],(function(){}))),!0}catch(t){return !1}}function a(t,e,n){return (a=u()?Reflect.construct:function(t,e,n){var r=[null];r.push.apply(r,e);var o=new(Function.bind.apply(t,r));return n&&s(o,n.prototype),o}).apply(null,arguments)}function s(t,e){return (s=Object.setPrototypeOf||function(t,e){return t.__proto__=e,t})(t,e)}function f(t){return (f=Object.setPrototypeOf?Object.getPrototypeOf:function(t){return t.__proto__||Object.getPrototypeOf(t)})(t)}n(32),n(42),n(43),n(44),n(119),n(31),n(122),n(123),n(128),n(48),n(129),n(131),n(52),n(53),t.exports=function(t){function e(){var t,n=arguments.length>0&&void 0!==arguments[0]?arguments[0]:"FetchError",r=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},c=r.status,u=r.statusText;return o(this,e),(t=i(this,f(e).call(this,"".concat(n,". Status: ").concat(c," (").concat(u,")")))).name="FetchError",t.status=c,t.statusText=u,t}return function(t,e){if("function"!=typeof e&&null!==e)throw new TypeError("Super expression must either be null or a function");t.prototype=Object.create(e&&e.prototype,{constructor:{value:t,writable:!0,configurable:!0}}),e&&s(t,e);}(e,t),e}(c(Error));},function(t,e,n){var r=n(8),o=n(62).indexOf,i=n(120),c=n(121),u=[].indexOf,a=!!u&&1/[1].indexOf(1,-0)<0,s=i("indexOf"),f=c("indexOf",{ACCESSORS:!0,1:0});r({target:"Array",proto:!0,forced:a||!s||!f},{indexOf:function(t){return a?u.apply(this,arguments)||0:o(this,t,arguments.length>1?arguments[1]:void 0)}});},function(t,e,n){var r=n(1);t.exports=function(t,e){var n=[][t];return !!n&&r((function(){n.call(null,e||function(){throw 1},1);}))};},function(t,e,n){var r=n(7),o=n(1),i=n(4),c=Object.defineProperty,u={},a=function(t){throw t};t.exports=function(t,e){if(i(u,t))return u[t];e||(e={});var n=[][t],s=!!i(e,"ACCESSORS")&&e.ACCESSORS,f=i(e,0)?e[0]:a,l=i(e,1)?e[1]:void 0;return u[t]=!!n&&!o((function(){if(s&&!r)return !0;var t={length:-1};s?c(t,1,{enumerable:!0,get:a}):t[1]=1,n.call(t,f,l);}))};},function(t,e,n){var r=n(7),o=n(6).f,i=Function.prototype,c=i.toString,u=/^\s*function ([^ (]*)/;!r||"name"in i||o(i,"name",{configurable:!0,get:function(){try{return c.call(this).match(u)[1]}catch(t){return ""}}});},function(t,e,n){var r=n(124),o=n(127);t.exports=r("Map",(function(t){return function(){return t(this,arguments.length?arguments[0]:void 0)}}),o);},function(t,e,n){var r=n(8),o=n(0),i=n(39),c=n(10),u=n(84),a=n(51),s=n(50),f=n(3),l=n(1),p=n(77),h=n(22),v=n(126);t.exports=function(t,e,n){var y=-1!==t.indexOf("Map"),d=-1!==t.indexOf("Weak"),g=y?"set":"add",x=o[t],m=x&&x.prototype,b=x,w={},S=function(t){var e=m[t];c(m,t,"add"==t?function(t){return e.call(this,0===t?0:t),this}:"delete"==t?function(t){return !(d&&!f(t))&&e.call(this,0===t?0:t)}:"get"==t?function(t){return d&&!f(t)?void 0:e.call(this,0===t?0:t)}:"has"==t?function(t){return !(d&&!f(t))&&e.call(this,0===t?0:t)}:function(t,n){return e.call(this,0===t?0:t,n),this});};if(i(t,"function"!=typeof x||!(d||m.forEach&&!l((function(){(new x).entries().next();})))))b=n.getConstructor(e,t,y,g),u.REQUIRED=!0;else if(i(t,!0)){var E=new b,_=E[g](d?{}:-0,1)!=E,O=l((function(){E.has(1);})),j=p((function(t){new x(t);})),P=!d&&l((function(){for(var t=new x,e=5;e--;)t[g](e,e);return !t.has(-0)}));j||((b=e((function(e,n){s(e,b,t);var r=v(new x,e,b);return null!=n&&a(n,r[g],r,y),r}))).prototype=m,m.constructor=b),(O||P)&&(S("delete"),S("has"),y&&S("get")),(P||_)&&S(g),d&&m.clear&&delete m.clear;}return w[t]=b,r({global:!0,forced:b!=x},w),h(b,t),d||n.setStrong(b,t,y),b};},function(t,e,n){var r=n(1);t.exports=!r((function(){return Object.isExtensible(Object.preventExtensions({}))}));},function(t,e,n){var r=n(3),o=n(73);t.exports=function(t,e,n){var i,c;return o&&"function"==typeof(i=e.constructor)&&i!==n&&r(c=i.prototype)&&c!==n.prototype&&o(t,c),t};},function(t,e,n){var r=n(6).f,o=n(21),i=n(75),c=n(30),u=n(50),a=n(51),s=n(46),f=n(76),l=n(7),p=n(84).fastKey,h=n(14),v=h.set,y=h.getterFor;t.exports={getConstructor:function(t,e,n,s){var f=t((function(t,r){u(t,f,e),v(t,{type:e,index:o(null),first:void 0,last:void 0,size:0}),l||(t.size=0),null!=r&&a(r,t[s],t,n);})),h=y(e),d=function(t,e,n){var r,o,i=h(t),c=g(t,e);return c?c.value=n:(i.last=c={index:o=p(e,!0),key:e,value:n,previous:r=i.last,next:void 0,removed:!1},i.first||(i.first=c),r&&(r.next=c),l?i.size++:t.size++,"F"!==o&&(i.index[o]=c)),t},g=function(t,e){var n,r=h(t),o=p(e);if("F"!==o)return r.index[o];for(n=r.first;n;n=n.next)if(n.key==e)return n};return i(f.prototype,{clear:function(){for(var t=h(this),e=t.index,n=t.first;n;)n.removed=!0,n.previous&&(n.previous=n.previous.next=void 0),delete e[n.index],n=n.next;t.first=t.last=void 0,l?t.size=0:this.size=0;},delete:function(t){var e=h(this),n=g(this,t);if(n){var r=n.next,o=n.previous;delete e.index[n.index],n.removed=!0,o&&(o.next=r),r&&(r.previous=o),e.first==n&&(e.first=r),e.last==n&&(e.last=o),l?e.size--:this.size--;}return !!n},forEach:function(t){for(var e,n=h(this),r=c(t,arguments.length>1?arguments[1]:void 0,3);e=e?e.next:n.first;)for(r(e.value,e.key,this);e&&e.removed;)e=e.previous;},has:function(t){return !!g(this,t)}}),i(f.prototype,n?{get:function(t){var e=g(this,t);return e&&e.value},set:function(t,e){return d(this,0===t?0:t,e)}}:{add:function(t){return d(this,t=0===t?0:t,t)}}),l&&r(f.prototype,"size",{get:function(){return h(this).size}}),f},setStrong:function(t,e,n){var r=e+" Iterator",o=y(e),i=y(r);s(t,e,(function(t,e){v(this,{type:r,target:t,state:o(t),kind:e,last:void 0});}),(function(){for(var t=i(this),e=t.kind,n=t.last;n&&n.removed;)n=n.previous;return t.target&&(t.last=n=n?n.next:t.state.first)?"keys"==e?{value:n.key,done:!1}:"values"==e?{value:n.value,done:!1}:{value:[n.key,n.value],done:!1}:(t.target=void 0,{value:void 0,done:!0})}),n?"entries":"values",!n,!0),f(e);}};},function(t,e,n){var r=n(8),o=n(1),i=n(15),c=n(47),u=n(72);r({target:"Object",stat:!0,forced:o((function(){c(1);})),sham:!u},{getPrototypeOf:function(t){return c(i(t))}});},function(t,e,n){var r=n(8),o=n(12),i=n(16),c=n(5),u=n(3),a=n(21),s=n(130),f=n(1),l=o("Reflect","construct"),p=f((function(){function t(){}return !(l((function(){}),[],t)instanceof t)})),h=!f((function(){l((function(){}));})),v=p||h;r({target:"Reflect",stat:!0,forced:v,sham:v},{construct:function(t,e){i(t),c(e);var n=arguments.length<3?t:i(arguments[2]);if(h&&!p)return l(t,e,n);if(t==n){switch(e.length){case 0:return new t;case 1:return new t(e[0]);case 2:return new t(e[0],e[1]);case 3:return new t(e[0],e[1],e[2]);case 4:return new t(e[0],e[1],e[2],e[3])}var r=[null];return r.push.apply(r,e),new(s.apply(t,r))}var o=n.prototype,f=a(u(o)?o:Object.prototype),v=Function.apply.call(t,f,e);return u(v)?v:f}});},function(t,e,n){var r=n(16),o=n(3),i=[].slice,c={},u=function(t,e,n){if(!(e in c)){for(var r=[],o=0;o<e;o++)r[o]="a["+o+"]";c[e]=Function("C,a","return new C("+r.join(",")+")");}return c[e](t,n)};t.exports=Function.bind||function(t){var e=r(this),n=i.call(arguments,1),c=function(){var r=n.concat(i.call(arguments));return this instanceof c?u(e,r.length,r):e.apply(t,r)};return o(e.prototype)&&(c.prototype=e.prototype),c};},function(t,e,n){var r=n(10),o=n(5),i=n(1),c=n(83),u=RegExp.prototype,a=u.toString,s=i((function(){return "/a/b"!=a.call({source:"a",flags:"b"})})),f="toString"!=a.name;(s||f)&&r(RegExp.prototype,"toString",(function(){var t=o(this),e=String(t.source),n=t.flags;return "/"+e+"/"+String(void 0===n&&t instanceof RegExp&&!("flags"in u)?c.call(t):n)}),{unsafe:!0});}])}));
});

var getTranslator = unwrapExports(fe);

var logError = function logError(error, extraData) {
}; // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types

var PROJECT_NAME = 'report-bad-lead';
function loadTranslations(_x) {
  return _loadTranslations.apply(this, arguments);
}

function _loadTranslations() {
  _loadTranslations = asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(locale) {
    var translatorClient;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.prev = 0;
            _context.next = 3;
            return getTranslator(PROJECT_NAME, locale);

          case 3:
            translatorClient = _context.sent;
            return _context.abrupt("return", translatorClient);

          case 7:
            _context.prev = 7;
            _context.t0 = _context["catch"](0);
            logError(_context.t0);
            throw _context.t0;

          case 11:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[0, 7]]);
  }));
  return _loadTranslations.apply(this, arguments);
}

var useTranslatorClient = function useTranslatorClient(locale) {
  var _useState = React.useState(React.useContext(TranslatorContext)),
      _useState2 = slicedToArray(_useState, 2),
      translatorClient = _useState2[0],
      setTranslatorClient = _useState2[1];

  React.useEffect(function () {
    var didCancel = false;

    if (translatorClient) {
      return;
    }

    var fetchData = /*#__PURE__*/function () {
      var _ref = asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var response;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return loadTranslations(locale);

              case 2:
                response = _context.sent;

                if (!didCancel) {
                  setTranslatorClient(response);
                }

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      return function fetchData() {
        return _ref.apply(this, arguments);
      };
    }();

    fetchData();
    return function () {
      didCancel = true;
    };
  }, []);
  return translatorClient;
};

var TranslatorProvider = function TranslatorProvider(_ref2) {
  var children = _ref2.children,
      _ref2$locale = _ref2.locale,
      locale = _ref2$locale === void 0 ? 'en-US' : _ref2$locale;
  var translator = useTranslatorClient(locale);

  if (!translator) {
    return null;
  }

  return /*#__PURE__*/React__default.createElement(TranslatorContext.Provider, {
    value: translator
  }, children);
};

var initialState = {
  hasError: false,
  hasSuccess: false,
  inProgress: false
};
var BadLeadFormContext = /*#__PURE__*/React.createContext({
  state: initialState,
  dispatch: function dispatch() {
    return {};
  }
}); // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types

var useBadLeadFormContext = function useBadLeadFormContext() {
  return React.useContext(BadLeadFormContext);
};

var BadLeadFormContextProvider = function BadLeadFormContextProvider(_ref) {
  var children = _ref.children,
      locale = _ref.locale;

  var _useReducer = React.useReducer(BadLeadFormReducer, initialState),
      _useReducer2 = slicedToArray(_useReducer, 2),
      state = _useReducer2[0],
      dispatch = _useReducer2[1];

  var badLeadFormContext = React.useMemo(function () {
    return {
      state: state,
      dispatch: dispatch
    };
  }, [state, dispatch]);
  return /*#__PURE__*/React__default.createElement(TranslatorProvider, {
    locale: locale
  }, /*#__PURE__*/React__default.createElement(BadLeadFormContext.Provider, {
    value: badLeadFormContext
  }, children));
};

function _taggedTemplateLiteral(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }

  return Object.freeze(Object.defineProperties(strings, {
    raw: {
      value: Object.freeze(raw)
    }
  }));
}

var taggedTemplateLiteral = _taggedTemplateLiteral;

const colors = {
	"white": "rgb(255,255,255)",
	"white64": "rgba(255,255,255,.64)",
	"white32": "rgba(255,255,255,.32)",
	"white16": "rgba(255,255,255,.16)",
	"black": "rgb(38,41,44)",
	"blackShade16Opaque": "#202225",
	"blackShade16": "rgb(32,34,37)",
	"black88Opaque": "#404346",
	"black88": "rgba(38,41,44,.88)",
	"black64Opaque": "#747678",
	"black64": "rgba(38,41,44,.64)",
	"black32Opaque": "#b9babb",
	"black32": "rgba(38,41,44,.32)",
	"black24Opaque": "#cbcccd",
	"black24": "rgba(38,41,44,.24)",
	"black16Opaque": "#dcdddd",
	"black16": "rgba(38,41,44,.16)",
	"black12Opaque": "#e5e5e5",
	"black12": "rgba(38,41,44,.12)",
	"black8Opaque": "#eeeeee",
	"black8": "rgba(38,41,44,.08)",
	"black5Opaque": "#f4f4f4",
	"black5": "rgba(38,41,44,.05)",
	"black4Opaque": "#f7f7f7",
	"black4": "rgba(38,41,44,.04)",
	"green": "rgb(8,167,66)",
	"greenShade16Opaque": "#0c923e",
	"greenShade16": "rgb(12,146,62)",
	"green88Opaque": "#26b259",
	"green88": "rgba(8,167,66,.88)",
	"green64Opaque": "#61c786",
	"green64": "rgba(8,167,66,.64)",
	"green32Opaque": "#b0e2c2",
	"green32": "rgba(8,167,66,.32)",
	"green24Opaque": "#c4ead2",
	"green24": "rgba(8,167,66,.24)",
	"green16Opaque": "#d7f1e1",
	"green16": "rgba(8,167,66,.16)",
	"green12Opaque": "#e1f4e8",
	"green12": "rgba(8,167,66,.12)",
	"green8Opaque": "#ecf8f0",
	"green8": "rgba(8,167,66,.08)",
	"blue": "rgb(49,122,226)",
	"blueShade16Opaque": "#2f6cc5",
	"blueShade16": "rgb(47,108,197)",
	"blue88Opaque": "#4a8ae6",
	"blue88": "rgba(49,122,226,.88)",
	"blue64Opaque": "#7baaed",
	"blue64": "rgba(49,122,226,.64)",
	"blue32Opaque": "#bdd4f5",
	"blue32": "rgba(49,122,226,.32)",
	"blue24Opaque": "#cedff8",
	"blue24": "rgba(49,122,226,.24)",
	"blue16Opaque": "#deeafa",
	"blue16": "rgba(49,122,226,.16)",
	"blue12Opaque": "#e6effb",
	"blue12": "rgba(49,122,226,.12)",
	"blue8Opaque": "#eff5fd",
	"blue8": "rgba(49,122,226,.08)",
	"red": "rgb(249,72,57)",
	"redShade16Opaque": "#d64336",
	"redShade16": "rgb(214,67,54)",
	"red88Opaque": "#fa5e51",
	"red88": "rgba(249,72,57,.88)",
	"red64Opaque": "#fb8a80",
	"red64": "rgba(249,72,57,.64)",
	"red32Opaque": "#fdc4bf",
	"red32": "rgba(249,72,57,.32)",
	"red24Opaque": "#fed3d0",
	"red24": "rgba(249,72,57,.24)",
	"red16Opaque": "#fee2df",
	"red16": "rgba(249,72,57,.16)",
	"red12Opaque": "#fee9e7",
	"red12": "rgba(249,72,57,.12)",
	"red8Opaque": "#fff1f0",
	"red8": "rgba(249,72,57,.08)",
	"yellow": "rgb(255,204,0)",
	"yellowShade16Opaque": "#ddb207",
	"yellowShade16": "rgb(221,178,7)",
	"yellow64Opaque": "#ffdf5c",
	"yellow64": "rgba(255,204,0,.64)",
	"yellow16Opaque": "#fff7d6",
	"yellow16": "rgba(255,204,0,.16)",
	"purple": "rgb(114,30,169)",
	"purpleShade16Opaque": "#661f95",
	"purpleShade16": "rgb(102,31,149)",
	"purple64Opaque": "#a56fc8",
	"purple64": "rgba(114,30,169,.64)",
	"purple16Opaque": "#e8dbf1",
	"purple16": "rgba(114,30,169,.16)",
	"highlightYellow": "rgb(255,245,112)",
	"noteYellow": "rgb(255,252,220)",
	"orchid": "#c762ce",
	"steel": "#8ba5af",
	"sky": "#42b7fc",
	"turquoise": "#00d3e4",
	"lime": "#69c97d",
	"kiwi": "#a3d776",
	"lemon": "#ffed27",
	"apricot": "#ffb32b",
	"salmon": "#ff827b",
	"flamingo": "#ff5592"
}; 

const fonts = {
	"fontFamily": "'Source Sans Pro',sans-serif",
	"fontFamilyMonospace": "'Menlo','Consolas','Courier',monospace",
	"fontWeightNormal": "400",
	"fontWeightSemi": "600",
	"fontWeightBold": "600",
	"fontSizeXxl": "27px",
	"fontSizeXl": "23px",
	"fontSizeL": "18px",
	"fontSizeM": "15px",
	"fontSizeS": "13px",
	"fontSizeXs": "11px",
	"lineHeightXxl": "32px",
	"lineHeightXl": "28px",
	"lineHeightL": "24px",
	"lineHeightM": "20px",
	"lineHeightS": "16px",
	"lineHeightXs": "16px",
	"letterSpacingL": ".04em",
	"letterSpacingM": ".03em",
	"fontTitleXxl": "400 27px/32px 'Source Sans Pro',sans-serif",
	"fontTitleXl": "400 23px/28px 'Source Sans Pro',sans-serif",
	"fontTitleL": "600 18px/24px 'Source Sans Pro',sans-serif",
	"fontTitleM": "600 15px/20px 'Source Sans Pro',sans-serif",
	"fontBodyL": "400 18px/24px 'Source Sans Pro',sans-serif",
	"fontBody": "400 15px/20px 'Source Sans Pro',sans-serif",
	"fontBodyS": "400 13px/16px 'Source Sans Pro',sans-serif",
	"fontButton": "600 15px/20px 'Source Sans Pro',sans-serif",
	"fontButtonS": "600 13px/16px 'Source Sans Pro',sans-serif",
	"fontCaption": "600 15px/20px 'Source Sans Pro',sans-serif",
	"fontCaptionS": "600 13px/16px 'Source Sans Pro',sans-serif",
	"fontBadge": "600 11px/16px 'Source Sans Pro',sans-serif",
	"fontMonospace": "400 13px/16px 'Menlo','Consolas','Courier',monospace"
};

function _templateObject2() {
  var data = taggedTemplateLiteral(["\n\tmargin-left: 8px;\n\tmax-width: 392px;\n\n\t& > p {\n\t\tmargin-bottom: 4px;\n\t}\n\n\t& > p:last-child {\n\t\tmargin-bottom: 0;\n\t}\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = taggedTemplateLiteral(["\n\tbox-shadow: inset 0px -1px 0px ", ";\n\n\t& > div {\n\t\tdisplay: flex;\n\t}\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var Wrapper = styled(conventionUiReact.Panel)(_templateObject(), colors.black12);
var TextWrapper = styled(conventionUiReact.Text)(_templateObject2());

var BadLeadError = function BadLeadError() {
  var translator = useTranslator();
  return /*#__PURE__*/React__default.createElement(Wrapper, {
    color: "red",
    noBorder: true,
    noMargin: true,
    spacing: "m"
  }, /*#__PURE__*/React__default.createElement(conventionUiReact.Icon, {
    icon: "warning",
    color: "red"
  }), /*#__PURE__*/React__default.createElement(TextWrapper, null, /*#__PURE__*/React__default.createElement("p", null, /*#__PURE__*/React__default.createElement("strong", null, translator.gettext('Something went wrong…'))), /*#__PURE__*/React__default.createElement("p", {
    dangerouslySetInnerHTML: {
      __html: translator.gettext('Please try again. If the problem persists, %scontact our customer support%s', ['<a href="https://support.pipedrive.com/" target="blank">', '</a>'])
    }
  })));
};

function _templateObject3() {
  var data = taggedTemplateLiteral(["\n\twidth: 100%;\n\tmin-height: 72px;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$1() {
  var data = taggedTemplateLiteral(["\n\tcolor: ", ";\n\tfont-size: ", ";\n\tmargin-bottom: 4px;\n"]);

  _templateObject2$1 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$1() {
  var data = taggedTemplateLiteral(["\n\twidth: 448px;\n"]);

  _templateObject$1 = function _templateObject() {
    return data;
  };

  return data;
}
var Wrapper$1 = styled(conventionUiReact.Spacing)(_templateObject$1());
var Text = styled.p(_templateObject2$1(), colors.black, fonts.fontSizeM);
var TextArea = styled(conventionUiReact.Textarea)(_templateObject3());

var BadLeadForm = function BadLeadForm() {
  var translator = useTranslator();

  var _useBadLeadFormContex = useBadLeadFormContext(),
      dispatch = _useBadLeadFormContex.dispatch;

  var onChange = React.useCallback(function (e) {
    dispatch({
      type: ActionTypes.SetBadLeadFormValue,
      inputValue: e.target.value,
      inputName: e.target.name
    });
  }, []);
  return /*#__PURE__*/React__default.createElement(Wrapper$1, {
    all: "m",
    bottom: "none"
  }, /*#__PURE__*/React__default.createElement(Text, null, translator.gettext('Add email, phone number or job position that is inaccurate')), /*#__PURE__*/React__default.createElement(TextArea, {
    onChange: onChange,
    name: "contacts"
  }), /*#__PURE__*/React__default.createElement(Text, null, translator.gettext('More details about the issue')), /*#__PURE__*/React__default.createElement(TextArea, {
    onChange: onChange,
    name: "details"
  }));
};

function _templateObject6() {
  var data = taggedTemplateLiteral(["\n\tmargin-bottom: 16px;\n\tcolor: ", ";\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = taggedTemplateLiteral(["\n\tcolor: ", ";\n\tfont-weight: ", ";\n\tmargin: 16px 0;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = taggedTemplateLiteral(["\n\tcolor: ", ";\n\tfont-weight: ", ";\n\tcursor: pointer;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3$1() {
  var data = taggedTemplateLiteral(["\n\tcolor: ", ";\n"]);

  _templateObject3$1 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$2() {
  var data = taggedTemplateLiteral(["\n\tmax-width: 392px;\n"]);

  _templateObject2$2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$2() {
  var data = taggedTemplateLiteral(["\n\tdisplay: flex;\n\tpadding: 16px;\n\tpadding-top: 0;\n"]);

  _templateObject$2 = function _templateObject() {
    return data;
  };

  return data;
}
var Wrapper$2 = styled.div(_templateObject$2());
var TextWrapper$1 = styled.div(_templateObject2$2());
var GrayText = styled(conventionUiReact.Text)(_templateObject3$1(), colors.black64Opaque);
var Toggle = styled.p(_templateObject4(), colors.blue, fonts.fontWeightBold);
var SubTitle = styled.p(_templateObject5(), colors.black64Opaque, fonts.fontWeightBold);
var ListItem = styled.p(_templateObject6(), colors.black64Opaque);

var Info = function Info() {
  var translator = useTranslator();

  var _useState = React.useState(false),
      _useState2 = slicedToArray(_useState, 2),
      isExpanded = _useState2[0],
      setIsExpanded = _useState2[1];

  return /*#__PURE__*/React__default.createElement(Wrapper$2, null, /*#__PURE__*/React__default.createElement(TextWrapper$1, null, /*#__PURE__*/React__default.createElement(GrayText, null, translator.gettext('It is possible to request back Prospector lead credit for a lead that had inaccurate information within 10 days of adding it to Leads Inbox.')), isExpanded && /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(SubTitle, null, translator.gettext('Credit for a bad lead can be refunded in the following cases:')), /*#__PURE__*/React__default.createElement(ListItem, null, translator.gettext('1. Messages sent to any of the email addresses provided by Prospector, can not be delivered')), /*#__PURE__*/React__default.createElement(ListItem, null, translator.gettext('2. Any of the provided phone numbers are incorrect')), /*#__PURE__*/React__default.createElement(ListItem, null, translator.gettext('3. Current job position is not up to date'))), /*#__PURE__*/React__default.createElement(Toggle, {
    onClick: function onClick() {
      return setIsExpanded(!isExpanded);
    }
  }, isExpanded ? translator.gettext('Show less') : translator.gettext('Learn more'))));
};

var refundOutboundPerson = /*#__PURE__*/function () {
  var _ref2 = asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(_ref) {
    var personId, contacts, details, result;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            personId = _ref.personId, contacts = _ref.contacts, details = _ref.details;
            _context.prev = 1;
            _context.next = 4;
            return fetch$1.post('/leads-graphql/', {
              query: "mutation {\n\t\t\t\t\trefundOutboundPerson(\n\t\t\t\t\t\tpersonId: \"".concat(personId, "\",\n\t\t\t\t\t\tcontacts: \"").concat(contacts, "\",\n\t\t\t\t\t\tdetails: \"").concat(details, "\"\n\t\t\t\t\t) { state }\n\t\t\t\t}")
            });

          case 4:
            result = _context.sent;
            return _context.abrupt("return", result.data);

          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](1);

            if (!(_context.t0 instanceof Error)) {
              _context.next = 12;
              break;
            }

            throw _context.t0;

          case 12:
            return _context.abrupt("return", _context.t0.data);

          case 13:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[1, 8]]);
  }));

  return function refundOutboundPerson(_x) {
    return _ref2.apply(this, arguments);
  };
}();

function _templateObject$3() {
  var data = taggedTemplateLiteral(["\n\tmargin-left: 8px;\n\tmin-width: 64px;\n"]);

  _templateObject$3 = function _templateObject() {
    return data;
  };

  return data;
}
var CtaButton = styled(conventionUiReact.Button)(_templateObject$3());

var ModalFooter = function ModalFooter(_ref) {
  var closeModal = _ref.closeModal,
      onRefundTicketCreated = _ref.onRefundTicketCreated;
  var translator = useTranslator();

  var _useBadLeadFormContex = useBadLeadFormContext(),
      state = _useBadLeadFormContex.state,
      dispatch = _useBadLeadFormContex.dispatch;

  var details = state.details,
      contacts = state.contacts,
      inProgress = state.inProgress;

  var _useState = React.useState(false),
      _useState2 = slicedToArray(_useState, 2),
      isBusy = _useState2[0],
      setIsBusy = _useState2[1];

  var sendReport = React.useCallback( /*#__PURE__*/asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            setIsBusy(true);
            _context.prev = 1;
            _context.next = 4;
            return refundOutboundPerson({
              name: state.name,
              personId: state.profile_id,
              contacts: state.contacts,
              details: state.details,
              translator: translator
            });

          case 4:
            if (onRefundTicketCreated) {
              onRefundTicketCreated();
            }

            dispatch({
              type: ActionTypes.SetBadLeadFormHasSuccess
            });
            _context.next = 12;
            break;

          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](1);
            dispatch({
              type: ActionTypes.SetBadLeadFormHasFailure
            });
            setIsBusy(false);

          case 12:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[1, 8]]);
  })), [details, contacts]);
  var isDisabled = React.useMemo(function () {
    return [details, contacts].some(function (inputValue) {
      return !inputValue || inputValue.trim() === '';
    }) || inProgress;
  }, [details, contacts, inProgress]);
  return /*#__PURE__*/React__default.createElement("div", null, /*#__PURE__*/React__default.createElement(conventionUiReact.Button, {
    onClick: closeModal
  }, translator.gettext('Cancel')), /*#__PURE__*/React__default.createElement(CtaButton, {
    disabled: isDisabled || isBusy,
    color: "green",
    onClick: sendReport
  }, isBusy ? /*#__PURE__*/React__default.createElement(conventionUiReact.Spinner, {
    size: "s"
  }) : translator.gettext('Send report')));
};

var ReportBadLeadModal = function ReportBadLeadModal(_ref) {
  var sourcePersonId = _ref.sourcePersonId,
      leadName = _ref.leadName,
      isVisible = _ref.isVisible,
      setModalVisible = _ref.setModalVisible,
      onRefundTicketCreated = _ref.onRefundTicketCreated;
  var translator = useTranslator();

  var _useBadLeadFormContex = useBadLeadFormContext(),
      state = _useBadLeadFormContex.state,
      dispatch = _useBadLeadFormContex.dispatch;

  var hasSuccess = state.hasSuccess,
      hasError = state.hasError;
  React.useEffect(function () {
    if (hasSuccess) {
      setModalVisible(false);
    }

    dispatch({
      type: ActionTypes.SetBadLeadFormInit,
      defaults: {
        name: leadName || '',
        profile_id: sourcePersonId || ''
      }
    });
  }, [leadName, hasSuccess]);
  var closeModal = React.useCallback(function () {
    setModalVisible(false);
  }, []);

  if (!isVisible) {
    return null;
  }

  return /*#__PURE__*/React__default.createElement(conventionUiReact.Modal, {
    onClose: closeModal,
    closeOnEsc: true,
    spacing: "none",
    header: translator.gettext('Report inaccurate lead'),
    visible: true,
    footer: /*#__PURE__*/React__default.createElement(ModalFooter, {
      closeModal: closeModal,
      onRefundTicketCreated: onRefundTicketCreated
    })
  }, /*#__PURE__*/React__default.createElement(React__default.Fragment, null, hasError && /*#__PURE__*/React__default.createElement(BadLeadError, null), /*#__PURE__*/React__default.createElement(BadLeadForm, null), /*#__PURE__*/React__default.createElement(Info, null)));
};

function _templateObject3$2() {
  var data = taggedTemplateLiteral(["\n\tdisplay: flex;\n\tjustify-content: flex-end;\n\talign-items: center;\n\twidth: 20px;\n\tmargin-left: 4px;\n"]);

  _templateObject3$2 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$3() {
  var data = taggedTemplateLiteral(["\n\tz-index: 99999;\n"]);

  _templateObject2$3 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$4() {
  var data = taggedTemplateLiteral(["\n\tpointer-events: none;\n\twidth: 100%;\n"]);

  _templateObject$4 = function _templateObject() {
    return data;
  };

  return data;
}
var Icon = styled(conventionUiReact.Icon)(_templateObject$4());
var Tooltip = styled(conventionUiReact.Tooltip)(_templateObject2$3());
var IconWrapper = styled.div(_templateObject3$2());

var InfoIcon = function InfoIcon(_ref) {
  var refundState = _ref.refundState,
      isDisabled = _ref.isDisabled;
  var translator = useTranslator();
  var getTooltipText = React.useMemo(function () {
    var reportTooltipTextMap = {
      ACTIVE: translator.gettext('You have already reported this lead. Your request will be reviewed within 3 business days.'),
      CLOSED_VALID: translator.gettext('This lead has already been reported and one extra lead credit was added to your account.'),
      CLOSED_INVALID: translator.gettext('This lead has already been reported, but was considered valid.')
    };

    if (refundState) {
      return reportTooltipTextMap[refundState];
    } else if (isDisabled) {
      return translator.gettext('This action is only available within 10 days of adding this lead to Leads Inbox.');
    } else {
      return translator.gettext('This action is available for leads added from Prospector within 10 days of adding it to Leads Inbox.');
    }
  }, [refundState, isDisabled]);
  return /*#__PURE__*/React__default.createElement(Tooltip, {
    content: getTooltipText,
    placement: "top-end",
    portalTo: document.body
  }, /*#__PURE__*/React__default.createElement("div", null, /*#__PURE__*/React__default.createElement(IconWrapper, null, /*#__PURE__*/React__default.createElement(Icon, {
    icon: "info-outline"
  }))));
};

function _templateObject3$3() {
  var data = taggedTemplateLiteral(["\n\tdisplay: flex;\n\tjustify-content: flex-end;\n\talign-items: center;\n\twidth: 20px;\n"]);

  _templateObject3$3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2$4() {
  var data = taggedTemplateLiteral(["\n\twhite-space: nowrap;\n"]);

  _templateObject2$4 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject$5() {
  var data = taggedTemplateLiteral(["\n\tdisplay: flex;\n\tjustify-content: space-between;\n\talign-items: center;\n\tmin-width: 223px;\n"]);

  _templateObject$5 = function _templateObject() {
    return data;
  };

  return data;
}
var ReportBadLeadOption = styled(conventionUiReact.Option)(_templateObject$5());
var Text$1 = styled.p(_templateObject2$4());
var IconWrapper$1 = styled.div(_templateObject3$3());

var ReportBadLeadOption$1 = function ReportBadLeadOption$1(_ref) {
  var setIsModalVisible = _ref.setIsModalVisible,
      _ref$infoIconVisible = _ref.infoIconVisible,
      infoIconVisible = _ref$infoIconVisible === void 0 ? true : _ref$infoIconVisible,
      refundState = _ref.refundState,
      addedAt = _ref.addedAt,
      pdMetrics = _ref.pdMetrics,
      leadId = _ref.leadId;
  var translator = useTranslator();
  var isDisabled = React.useMemo(function () {
    if (!addedAt) {
      return false;
    }

    if (refundState && refundState !== 'NONE') {
      return true;
    }

    var current = new Date();
    var tenDaysAgo = new Date(current.setDate(current.getDate() - 10));
    return addedAt < new Date(tenDaysAgo);
  }, []);

  var handleClick = function handleClick() {
    pdMetrics.trackUsage('', 'prospector_inaccurate_lead_report', 'started', {
      lead_id: leadId
    });
    setIsModalVisible(true);
  };

  return /*#__PURE__*/React__default.createElement(ReportBadLeadOption, {
    onClick: handleClick,
    disabled: isDisabled
  }, /*#__PURE__*/React__default.createElement(Text$1, null, translator.gettext('Report inaccurate lead')), infoIconVisible && refundState !== 'NONE' && /*#__PURE__*/React__default.createElement(InfoIcon, {
    isDisabled: isDisabled,
    refundState: refundState
  }));
};

exports.ReportBadLeadContext = BadLeadFormContextProvider;
exports.ReportBadLeadModal = ReportBadLeadModal;
exports.ReportBadLeadOption = ReportBadLeadOption$1;
