// @splitsoftware/splitio/lib/sync/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ControlTypes = exports.OCCUPANCY = exports.CONTROL = exports.SPLIT_UPDATE = exports.SPLIT_KILL = exports.SEGMENT_UPDATE = exports.MY_SEGMENTS_UPDATE = exports.PUSH_SUBSYSTEM_DOWN = exports.PUSH_SUBSYSTEM_UP = exports.PUSH_RETRYABLE_ERROR = exports.PUSH_NONRETRYABLE_ERROR = exports.SECONDS_BEFORE_EXPIRATION = void 0;
// time for refresh token
var SECONDS_BEFORE_EXPIRATION = 600; // Internal SDK events, subscribed by SyncManager and PushManager

/**
 * emitted on SSE and Authenticate non-recoverable errors, STREAMING_DISABLED control notification and authentication with pushEnabled false
 * triggers `handleNonRetryableError` call
 */

exports.SECONDS_BEFORE_EXPIRATION = SECONDS_BEFORE_EXPIRATION;
var PUSH_NONRETRYABLE_ERROR = 'PUSH_NONRETRYABLE_ERROR';
/**
 * emitted on SSE and Authenticate recoverable errors
 * triggers `handleRetryableError` call
 */

exports.PUSH_NONRETRYABLE_ERROR = PUSH_NONRETRYABLE_ERROR;
var PUSH_RETRYABLE_ERROR = 'PUSH_RETRYABLE_ERROR';
/**
 * emitted on STREAMING_RESUMED control notification, OCCUPANCY different than 0, and SSE onopen event
 * triggers `stopPollingAndSyncAll` call
 */

exports.PUSH_RETRYABLE_ERROR = PUSH_RETRYABLE_ERROR;
var PUSH_SUBSYSTEM_UP = 'PUSH_SUBSYSTEM_UP';
/**
 * emitted on STREAMING_PAUSED control notification, OCCUPANCY equal to 0, PUSH_NONRETRYABLE_ERROR and PUSH_RETRYABLE_ERROR events.
 * triggers `startPolling` and `stopWorkers` calls
 */

exports.PUSH_SUBSYSTEM_UP = PUSH_SUBSYSTEM_UP;
var PUSH_SUBSYSTEM_DOWN = 'PUSH_SUBSYSTEM_DOWN'; // Update-type push notifications, handled by NotificationProcessor

exports.PUSH_SUBSYSTEM_DOWN = PUSH_SUBSYSTEM_DOWN;
var MY_SEGMENTS_UPDATE = 'MY_SEGMENTS_UPDATE';
exports.MY_SEGMENTS_UPDATE = MY_SEGMENTS_UPDATE;
var SEGMENT_UPDATE = 'SEGMENT_UPDATE';
exports.SEGMENT_UPDATE = SEGMENT_UPDATE;
var SPLIT_KILL = 'SPLIT_KILL';
exports.SPLIT_KILL = SPLIT_KILL;
var SPLIT_UPDATE = 'SPLIT_UPDATE'; // Control-type push notifications, handled by NotificationKeeper

exports.SPLIT_UPDATE = SPLIT_UPDATE;
var CONTROL = 'CONTROL';
exports.CONTROL = CONTROL;
var OCCUPANCY = 'OCCUPANCY';
exports.OCCUPANCY = OCCUPANCY;
var ControlTypes = {
  STREAMING_DISABLED: 'STREAMING_DISABLED',
  STREAMING_PAUSED: 'STREAMING_PAUSED',
  STREAMING_RESUMED: 'STREAMING_RESUMED'
};
exports.ControlTypes = ControlTypes;
