// react-relay/lib/relay-hooks/useMutation.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+relay
 * 
 * @format
 */
// flowlint ambiguous-object-type:error
'use strict';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

var React = require('react');

var useRelayEnvironment = require('./useRelayEnvironment');

var _require = require('relay-runtime'),
    defaultCommitMutation = _require.commitMutation;

var useState = React.useState,
    useEffect = React.useEffect,
    useRef = React.useRef,
    useCallback = React.useCallback;

var useIsMountedRef = require('./useIsMountedRef');

function useMutation(mutation) {
  var commitMutationFn = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultCommitMutation;
  var environment = useRelayEnvironment();
  var isMountedRef = useIsMountedRef();
  var environmentRef = useRef(environment);
  var mutationRef = useRef(mutation);
  var inFlightMutationsRef = useRef(new Set());

  var _useState = useState(false),
      isMutationInFlight = _useState[0],
      setMutationInFlight = _useState[1];

  var cleanup = useCallback(function (disposable) {
    if (environmentRef.current === environment && mutationRef.current === mutation) {
      inFlightMutationsRef.current["delete"](disposable);

      if (isMountedRef.current) {
        setMutationInFlight(inFlightMutationsRef.current.size > 0);
      }
    }
  }, [environment, isMountedRef, mutation]);
  useEffect(function () {
    if (environmentRef.current !== environment || mutationRef.current !== mutation) {
      inFlightMutationsRef.current = new Set();

      if (isMountedRef.current) {
        setMutationInFlight(false);
      }

      environmentRef.current = environment;
      mutationRef.current = mutation;
    }
  }, [environment, isMountedRef, mutation]);
  var commit = useCallback(function (config) {
    var disposable = commitMutationFn(environment, (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, config), {}, {
      mutation: mutation,
      onCompleted: function onCompleted(response, errors) {
        cleanup(disposable);
        config.onCompleted && config.onCompleted(response, errors);
      },
      onError: function onError(error) {
        cleanup(disposable);
        config.onError && config.onError(error);
      },
      onUnsubscribe: function onUnsubscribe() {
        cleanup(disposable);
        config.onUnsubscribe && config.onUnsubscribe();
      }
    }));
    inFlightMutationsRef.current.add(disposable);

    if (isMountedRef.current) {
      setMutationInFlight(true);
    }

    return disposable;
  }, [cleanup, commitMutationFn, environment, isMountedRef, mutation]);
  return [commit, isMutationInFlight];
}

module.exports = useMutation;
