// @splitsoftware/splitio/lib/metrics/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _logger = _interopRequireDefault(require("../utils/logger"));

var _timeTracker = _interopRequireDefault(require("../utils/timeTracker"));

var _repeat = _interopRequireDefault(require("../utils/fn/repeat"));

var _metrics = _interopRequireDefault(require("../services/metrics"));

var _times = _interopRequireDefault(require("../services/metrics/times"));

var _counters = _interopRequireDefault(require("../services/metrics/counters"));

var _dto = require("../services/metrics/dto");

var _impressions = _interopRequireDefault(require("../services/impressions"));

var _bulk = _interopRequireDefault(require("../services/impressions/bulk"));

var _count = _interopRequireDefault(require("../services/impressions/count"));

var _dto2 = require("../services/impressions/dto");

var _Collectors = require("./Collectors");

var _constants = require("../utils/constants");

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var log = (0, _logger.default)('splitio-metrics');
var IMPRESSIONS_COUNT_RATE = 1800000; // 30 minutes

var MetricsFactory = function MetricsFactory(context) {
  var impressionsRetries = 0;
  var settings = context.get(context.constants.SETTINGS);
  var storage = context.get(context.constants.STORAGE);
  var impressionsCounter = context.get(context.constants.IMPRESSIONS_COUNTER);
  var shouldPushImpressionsCount = settings.sync.impressionsMode === _constants.OPTIMIZED;

  var pushMetrics = function pushMetrics() {
    if (storage.metrics.isEmpty() && storage.count.isEmpty()) return Promise.resolve();
    log.info('Pushing metrics');

    var latencyTrackerStop = _timeTracker.default.start(_timeTracker.default.TaskNames.METRICS_PUSH); // POST latencies


    var latenciesPromise = storage.metrics.isEmpty() ? null : (0, _metrics.default)((0, _times.default)(settings, {
      body: JSON.stringify((0, _dto.fromLatenciesCollector)(storage.metrics))
    })).then(function () {
      return storage.metrics.clear();
    }).catch(function () {
      return storage.metrics.clear();
    }); // POST counters

    var countersPromise = storage.count.isEmpty() ? null : (0, _metrics.default)((0, _counters.default)(settings, {
      body: JSON.stringify((0, _dto.fromCountersCollector)(storage.count))
    })).then(function () {
      return storage.count.clear();
    }).catch(function () {
      return storage.count.clear();
    });
    return Promise.all([latenciesPromise, countersPromise]).then(function (resp) {
      // After both finishes, track the end and return the results
      latencyTrackerStop();
      return resp;
    });
  };

  var pushImpressions = function pushImpressions() {
    if (storage.impressions.isEmpty()) return Promise.resolve();
    var imprCount = storage.impressions.queue.length;
    log.info("Pushing ".concat(imprCount, " impressions"));

    var latencyTrackerStop = _timeTracker.default.start(_timeTracker.default.TaskNames.IMPRESSIONS_PUSH);

    return (0, _impressions.default)((0, _bulk.default)(settings, {
      body: JSON.stringify((0, _dto2.fromImpressionsCollector)(storage.impressions, settings))
    })).then(function () {
      impressionsRetries = 0;
      storage.impressions.clear();
    }).catch(function (err) {
      if (impressionsRetries) {
        // For now we retry only once.
        log.warn("Droping ".concat(imprCount, " impressions after retry. Reason ").concat(err, "."));
        impressionsRetries = 0;
        storage.impressions.clear();
      } else {
        impressionsRetries++;
        log.warn("Failed to push ".concat(imprCount, " impressions, keeping data to retry on next iteration. Reason ").concat(err, "."));
      }
    }).then(function () {
      return latencyTrackerStop();
    });
  };

  var pushImpressionsCount = function pushImpressionsCount() {
    var pf = (0, _dto2.fromImpressionsCountCollector)(impressionsCounter);
    var imprCounts = pf.length;
    if (imprCounts === 0) return Promise.resolve();
    log.info("Pushing count of impressions for ".concat(imprCounts, " features"));
    return (0, _impressions.default)((0, _count.default)(settings, {
      body: JSON.stringify({
        pf: pf
      })
    })).then(function () {
      impressionsRetries = 0;
    }).catch(function (err) {
      if (impressionsRetries) {
        // For now we retry only once.
        log.warn("Droping count of impressions for ".concat(imprCounts, " features after retry. Reason ").concat(err, "."));
        impressionsRetries = 0;
      } else {
        impressionsRetries++;
        log.warn("Failed to push impressions count for ".concat(imprCounts, " features, keeping data to retry on next iteration. Reason ").concat(err, "."));
      }
    });
  };

  var stopImpressionsPublisher = false;
  var stopPerformancePublisher = false;
  var stopImpressionsCountPublisher = false;
  return {
    start: function start() {
      stopImpressionsPublisher = (0, _repeat.default)(function (schedulePublisher) {
        return pushImpressions().then(function () {
          return schedulePublisher();
        });
      }, settings.scheduler.impressionsRefreshRate);
      stopPerformancePublisher = (0, _repeat.default)(function (schedulePublisher) {
        return pushMetrics().then(function () {
          return schedulePublisher();
        });
      }, settings.scheduler.metricsRefreshRate);

      if (shouldPushImpressionsCount) {
        stopImpressionsCountPublisher = (0, _repeat.default)(function (schedulePublisher) {
          return pushImpressionsCount().then(function () {
            return schedulePublisher();
          });
        }, IMPRESSIONS_COUNT_RATE);
      }
    },
    flush: function flush() {
      if (shouldPushImpressionsCount) pushImpressionsCount();
      return pushImpressions();
    },
    stop: function stop() {
      stopImpressionsPublisher && stopImpressionsPublisher();
      stopPerformancePublisher && stopPerformancePublisher();
      stopImpressionsCountPublisher && stopImpressionsCountPublisher();
    },
    // Metrics collectors
    collectors: {
      segmentChanges: new _Collectors.SegmentChangesCollector(storage),
      splitChanges: new _Collectors.SplitChangesCollector(storage),
      mySegments: new _Collectors.MySegmentsCollector(storage),
      client: new _Collectors.ClientCollector(storage)
    }
  };
};

var _default = MetricsFactory;
exports.default = _default;
