// @postman/react-click-outside/dist/index.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var hoistNonReactStatic = require('hoist-non-react-statics');
var React = require('react');
var ReactDOM = require('react-dom');

function enhanceWithClickOutside(arg) {
  var options = { eventCapturingPhase: true };
  function decoratedComponent(Component) {
    var componentName = Component.displayName || Component.name,
        _options = options,
        eventCapturingPhase = _options.eventCapturingPhase;

    var EnhancedComponent = function (_React$Component) {
      _inherits(EnhancedComponent, _React$Component);

      function EnhancedComponent(props) {
        _classCallCheck(this, EnhancedComponent);

        var _this = _possibleConstructorReturn(this, (EnhancedComponent.__proto__ || Object.getPrototypeOf(EnhancedComponent)).call(this, props));

        _this.handleClickOutside = _this.handleClickOutside.bind(_this);
        return _this;
      }

      _createClass(EnhancedComponent, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
          document.addEventListener('click', this.handleClickOutside, eventCapturingPhase);
        }
      }, {
        key: 'componentWillUnmount',
        value: function componentWillUnmount() {
          document.removeEventListener('click', this.handleClickOutside, eventCapturingPhase);
        }
      }, {
        key: 'handleClickOutside',
        value: function handleClickOutside(e) {
          var domNode = this.__domNode;
          if ((!domNode || !domNode.contains(e.target)) && this.__wrappedInstance && typeof this.__wrappedInstance.handleClickOutside === 'function') {
            this.__wrappedInstance.handleClickOutside(e);
          }
        }
      }, {
        key: 'render',
        value: function render() {
          var _this2 = this;

          var _props = this.props,
              wrappedRef = _props.wrappedRef,
              rest = _objectWithoutProperties(_props, ['wrappedRef']);

          return React.createElement(Component, _extends({}, rest, {
            ref: function ref(c) {
              _this2.__wrappedInstance = c;
              _this2.__domNode = ReactDOM.findDOMNode(c);
              wrappedRef && wrappedRef(c);
            }
          }));
        }
      }]);

      return EnhancedComponent;
    }(React.Component);

    EnhancedComponent.displayName = 'clickOutside(' + componentName + ')';

    return hoistNonReactStatic(EnhancedComponent, Component);
  }

  if ((typeof arg === 'undefined' ? 'undefined' : _typeof(arg)) === 'object') {
    options = arg;
    return decoratedComponent;
  } else {
    return decoratedComponent(arg);
  }
}

module.exports = enhanceWithClickOutside;
