// @postman/aether/esmLib/design-tokens/tokens.js
/* eslint-disable no-underscore-dangle */
export var globals = {
  colors: {
    // grey
    'grey-00': '#FFFFFF',
    'grey-05': '#F9F9F9',
    'grey-10': '#F2F2F2',
    'grey-20': '#EDEDED',
    'grey-30': '#E6E6E6',
    'grey-40': '#A6A6A6',
    'grey-50': '#6B6B6B',
    'grey-60': '#3B3B3B',
    'grey-70': '#303030',
    'grey-80': '#2B2B2B',
    'grey-85': '#262626',
    'grey-90': '#212121',
    // orange
    'orange-10': '#FFF1EB',
    'orange-20': '#FFD1BE',
    'orange-30': '#FFB091',
    'orange-40': '#FF8E64',
    'orange-50': '#FF6C37',
    'orange-60': '#E05320',
    'orange-70': '#D23F0E',
    'orange-80': '#A12700',
    'orange-90': '#553326',
    // blue
    'blue-10': '#E7F0FF',
    'blue-20': '#ADCDFB',
    'blue-30': '#74AEF6',
    'blue-40': '#3E92F2',
    'blue-50': '#097BED',
    'blue-60': '#0265D2',
    'blue-70': '#0053B8',
    'blue-80': '#00439D',
    'blue-90': '#002D70',
    // red
    'red-10': '#FFEBE7',
    'red-20': '#FBC3BA',
    'red-30': '#F79A8E',
    'red-40': '#F37264',
    'red-50': '#EB2013',
    'red-60': '#D10D03',
    'red-70': '#B70700',
    'red-80': '#9C0400',
    'red-90': '#591B08',
    // green
    'green-10': '#E5FFF1',
    'green-20': '#A4EEC4',
    'green-30': '#6BDD9A',
    'green-40': '#38CC74',
    'green-50': '#0CBB52',
    'green-60': '#049D40',
    'green-70': '#007F31',
    'green-80': '#006024',
    'green-90': '#013614',
    // yellow
    'yellow-10': '#FFF9E0',
    'yellow-20': '#FFF4BE',
    'yellow-30': '#FFE47E',
    'yellow-40': '#FFCF3F',
    'yellow-50': '#FFB400',
    'yellow-60': '#DE9C00',
    'yellow-70': '#AD7A03',
    'yellow-80': '#886000',
    'yellow-90': '#523A03',
    // pink
    'pink-10': '#FFF2F9',
    'pink-20': '#FFD8ED',
    'pink-30': '#FDA4D4',
    'pink-40': '#F15EB0',
    'pink-50': '#CA2583',
    'pink-60': '#A61468',
    'pink-70': '#7B0A4B',
    'pink-80': '#63073B',
    'pink-90': '#47042A',
    // purple
    'purple-10': '#F7F2FF',
    'purple-20': '#E4D8F6',
    'purple-30': '#C0A8E1',
    'purple-40': '#926CC2',
    'purple-50': '#784FA9',
    'purple-60': '#623497',
    'purple-70': '#4C207D',
    'purple-80': '#35115C',
    'purple-90': '#250447',
    // teal
    'teal-10': '#F2FEFF',
    'teal-20': '#D0F9FD',
    'teal-30': '#8EEEF7',
    'teal-40': '#3FD6E6',
    'teal-50': '#0CA4B6',
    'teal-60': '#048192',
    'teal-70': '#015965',
    'teal-80': '#02444D',
    'teal-90': '#032F36'
  },
  sizing: {
    'size-xs': '16px',
    'size-s': '24px',
    'size-m': '32px'
  },
  spacing: {
    'spacing-zero': '0px',
    'spacing-xs': '4px',
    'spacing-s': '8px',
    'spacing-m': '12px',
    'spacing-l': '16px',
    'spacing-xl': '24px',
    'spacing-xxl': '32px',
    'spacing-xxxl': '48px'
  },
  shadow: {
    'shadow-focus': "0 0 0 1px #FFFFFF, 0 0 0 3px #ADCDFB"
  },
  typography: {
    // font
    'text-family-default': "'Inter', system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Fira Sans', 'Droid Sans', Helvetica, Arial, sans-serif",
    'text-family-code': "'IBMPlexMono', SFMono-Regular, Menlo, Monaco, Consolas, 'Liberation Mono', 'Courier New', monospace",
    // font-size
    'text-size-xs': '10px',
    'text-size-s': '11px',
    'text-size-m': '12px',
    'text-size-l': '14px',
    'text-size-xl': '16px',
    'text-size-xxl': '18px',
    'text-size-xxxl': '24px',
    // line-height
    'line-height-xs': '12px',
    'line-height-s': '16px',
    'line-height-m': '20px',
    'line-height-l': '24px',
    'line-height-xl': '28px',
    'line-height-xxl': '32px',
    'line-height-xxxl': '40px',
    // font-weight
    'text-weight-regular': 400,
    'text-weight-medium': 600,
    'text-weight-bold': 600
  },
  border: {
    // border-width
    'border-width-default': '1px',
    'border-width-l': '2px',
    'border-width-xl': '3px',
    // border-style
    'border-style-solid': 'solid',
    // border-radius
    'border-radius-s': '2px',
    'border-radius-default': '4px',
    'border-radius-max': '100%'
  },
  _border: {
    'border-style-dashed': 'dashed'
  }
};
export var aliases = {
  light: {
    colors: {
      'base-color-brand': globals.colors['orange-50'],
      'base-color-info': globals.colors['blue-50'],
      'base-color-warning': globals.colors['yellow-50'],
      'base-color-error': globals.colors['red-50'],
      'base-color-success': globals.colors['green-50']
    },
    content: {
      'content-color-primary': globals.colors['grey-90'],
      'content-color-secondary': globals.colors['grey-50'],
      'content-color-tertiary': globals.colors['grey-40'],
      'content-color-brand': globals.colors['orange-70'],
      'content-color-info': globals.colors['blue-70'],
      'content-color-error': globals.colors['red-70'],
      'content-color-warning': globals.colors['yellow-70'],
      'content-color-success': globals.colors['green-70'],
      'content-color-constant': globals.colors['grey-00'],
      'content-color-link': globals.colors['blue-60'],
      'illustration-stroke-color': '#686868'
    },
    background: {
      'background-color-primary': globals.colors['grey-00'],
      'background-color-secondary': globals.colors['grey-05'],
      'background-color-tertiary': globals.colors['grey-10'],
      'background-color-brand': globals.colors['orange-10'],
      'background-color-info': globals.colors['blue-10'],
      'background-color-error': globals.colors['red-10'],
      'background-color-warning': globals.colors['yellow-10'],
      'background-color-success': globals.colors['green-10'],
      'background-modal-backdrop': 'rgba(26, 26, 26, 0.6)'
    },
    border: {
      'border-color-default': globals.colors['grey-20'],
      'border-color-strong': globals.colors['grey-30']
    },
    _border: {
      'border-color-active': 'rgba(26, 26, 26, 0.3)'
    },
    header: {
      'header-background-color-primary': globals.colors['grey-90'],
      'header-background-color-secondary': globals.colors['grey-85'],
      'header-color-content': globals.colors['grey-40'],
      'header-background-highlight-color-primary': globals.colors['grey-70'],
      'header-background-highlight-color-secondary': globals.colors['grey-70']
    },
    controls: {
      'controls-size-default': globals.sizing['size-m'],
      'controls-size-small': globals.sizing['size-s']
    },
    highlight: {
      'highlight-background-color-primary': globals.colors['grey-20'],
      'highlight-background-color-secondary': globals.colors['grey-20'],
      'highlight-background-color-tertiary': globals.colors['grey-30'],
      'highlight-background-color-brand': globals.colors['orange-60'],
      'highlight-background-color-info': globals.colors['blue-60'],
      'highlight-background-color-warning': globals.colors['yellow-60'],
      'highlight-background-color-error': globals.colors['red-60'],
      'highlight-background-color-success': globals.colors['green-60'],
      'highlight-background-color-transparent': 'rgba(26, 26, 26, 0.1)'
    },
    shadow: {
      'shadow-default': '0px 2px 8px 0px rgba(0, 0, 0, 0.2)'
    }
  },
  dark: {
    colors: {
      'base-color-brand': globals.colors['orange-50'],
      'base-color-info': globals.colors['blue-50'],
      'base-color-warning': globals.colors['yellow-50'],
      'base-color-error': globals.colors['red-50'],
      'base-color-success': globals.colors['green-50']
    },
    content: {
      'content-color-primary': globals.colors['grey-00'],
      'content-color-secondary': globals.colors['grey-40'],
      'content-color-tertiary': globals.colors['grey-50'],
      'content-color-brand': globals.colors['orange-40'],
      'content-color-info': globals.colors['blue-30'],
      'content-color-error': globals.colors['red-30'],
      'content-color-warning': globals.colors['yellow-30'],
      'content-color-success': globals.colors['green-30'],
      'content-color-constant': globals.colors['grey-00'],
      'content-color-link': globals.colors['blue-40'],
      'illustration-stroke-color': '#8B8B8B'
    },
    background: {
      'background-color-primary': globals.colors['grey-90'],
      'background-color-secondary': globals.colors['grey-85'],
      'background-color-tertiary': globals.colors['grey-80'],
      'background-color-brand': globals.colors['orange-90'],
      'background-color-info': globals.colors['blue-90'],
      'background-color-error': globals.colors['red-90'],
      'background-color-warning': globals.colors['yellow-90'],
      'background-color-success': globals.colors['green-90'],
      'background-modal-backdrop': 'rgba(26, 26, 26, 0.6)'
    },
    border: {
      'border-color-default': globals.colors['grey-70'],
      'border-color-strong': globals.colors['grey-60']
    },
    _border: {
      'border-color-active': 'rgba(26, 26, 26, 0.3)'
    },
    header: {
      'header-background-color-primary': globals.colors['grey-90'],
      'header-background-color-secondary': globals.colors['grey-85'],
      'header-color-content': globals.colors['grey-40']
    },
    _header: {
      'header-background-highlight-color-primary': globals.colors['grey-70'],
      'header-background-highlight-color-secondary': globals.colors['grey-90']
    },
    controls: {
      'controls-size-default': globals.sizing['size-m'],
      'controls-size-small': globals.sizing['size-s']
    },
    highlight: {
      'highlight-background-color-primary': globals.colors['grey-70'],
      'highlight-background-color-secondary': globals.colors['grey-70'],
      'highlight-background-color-tertiary': globals.colors['grey-60'],
      'highlight-background-color-brand': globals.colors['orange-60'],
      'highlight-background-color-info': globals.colors['blue-60'],
      'highlight-background-color-warning': globals.colors['yellow-60'],
      'highlight-background-color-error': globals.colors['red-60'],
      'highlight-background-color-success': globals.colors['green-60'],
      'highlight-background-color-transparent': 'rgba(255, 255, 255, 0.1)'
    },
    shadow: {
      'shadow-focus': "0 0 0 1px ".concat(globals.colors['grey-90'], ", 0 0 0 3px ").concat(globals.colors['blue-20']),
      'shadow-default': '0px 2px 8px 0px rgba(0, 0, 0, 0.65)'
    }
  }
};
