// react-tabs/lib/components/Tab.js
'use strict';

var _extends = Object.assign || function(target) {
    for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];
        for (var key in source) {
            if (Object.prototype.hasOwnProperty.call(source, key)) {
                target[key] = source[key];
            }
        }
    }
    return target;
};

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _reactDom = require('react-dom');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}

function _objectWithoutProperties(obj, keys) {
    var target = {};
    for (var i in obj) {
        if (keys.indexOf(i) >= 0) continue;
        if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
        target[i] = obj[i];
    }
    return target;
}

module.exports = (0, _createReactClass2.default)({
    displayName: 'Tab',

    propTypes: {
        className: _propTypes2.default.string,
        id: _propTypes2.default.string,
        focus: _propTypes2.default.bool,
        selected: _propTypes2.default.bool,
        disabled: _propTypes2.default.bool,
        activeTabClassName: _propTypes2.default.string,
        disabledTabClassName: _propTypes2.default.string,
        panelId: _propTypes2.default.string,
        children: _propTypes2.default.oneOfType([_propTypes2.default.array, _propTypes2.default.object, _propTypes2.default.string])
    },

    getDefaultProps: function getDefaultProps() {
        return {
            focus: false,
            selected: false,
            id: null,
            panelId: null,
            activeTabClassName: 'ReactTabs__Tab--selected',
            disabledTabClassName: 'ReactTabs__Tab--disabled'
        };
    },
    componentDidMount: function componentDidMount() {
        this.checkFocus();
    },
    componentDidUpdate: function componentDidUpdate() {
        this.checkFocus();
    },
    checkFocus: function checkFocus() {
        if (this.props.selected && this.props.focus) {
            (0, _reactDom.findDOMNode)(this).focus();
        }
    },
    render: function render() {
        var _cx;

        var _props = this.props,
            selected = _props.selected,
            disabled = _props.disabled,
            panelId = _props.panelId,
            activeTabClassName = _props.activeTabClassName,
            disabledTabClassName = _props.disabledTabClassName,
            className = _props.className,
            children = _props.children,
            id = _props.id,
            attributes = _objectWithoutProperties(_props, ['selected', 'disabled', 'panelId', 'activeTabClassName', 'disabledTabClassName', 'className', 'children', 'id']);

        delete attributes.focus;

        return _react2.default.createElement(
            'li',
            _extends({}, attributes, {
                className: (0, _classnames2.default)('ReactTabs__Tab', className, (_cx = {}, _defineProperty(_cx, activeTabClassName, selected), _defineProperty(_cx, disabledTabClassName, disabled), _cx)),
                role: 'tab',
                id: id,
                'aria-selected': selected ? 'true' : 'false',
                'aria-disabled': disabled ? 'true' : 'false',
                'aria-controls': panelId,
                tabIndex: selected ? '0' : null
            }),
            children
        );
    }
});
