// feather/lib/components/link/index.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/es.weak-map.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.symbol.iterator.js");Object.defineProperty(exports, "__esModule", { value: true });exports.Link = void 0;require("core-js/modules/es.array.concat.js");var _react = require("react");
var _reactRouterDom = require("react-router-dom");
var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _animation = require("../../../lib/animation");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _types = require("../../../lib/types");

var _icon = require("../icon");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}

var anchorCSS = function anchorCSS(_ref) {var variant = _ref.variant;
  var _ref2 =
  function () {
    switch (variant) {
      case _types.LinkVariant.Neutral:
        return {
          color: (0, _cssVariables.default)('neutral-10'),
          hoverColor: (0, _cssVariables.default)('purple-7'),
          focusColor: (0, _cssVariables.default)('purple-7'),
          activeColor: (0, _cssVariables.default)('neutral-9'),
          disabledColor: (0, _cssVariables.default)('neutral-5') };

      case _types.LinkVariant.Default:
      case _types.LinkVariant.Inline:
        return {
          color: (0, _cssVariables.default)('purple-7'),
          hoverColor: (0, _cssVariables.default)('purple-8'),
          focusColor: (0, _cssVariables.default)('purple-7'),
          activeColor: (0, _cssVariables.default)('neutral-9'),
          disabledColor: (0, _cssVariables.default)('neutral-5') };

      case _types.LinkVariant.Mono:
        return {
          color: 'inherit',
          hoverColor: 'inherit',
          focusColor: 'inherit',
          activeColor: 'inherit',
          disabledColor: 'inherit',
          hoverOpacity: 0.48,
          activeOpacity: 0.72,
          disabledOpacity: 0.32 };

      default:
        return {};}

  }(),color = _ref2.color,hoverColor = _ref2.hoverColor,activeColor = _ref2.activeColor,focusColor = _ref2.focusColor,disabledColor = _ref2.disabledColor,hoverOpacity = _ref2.hoverOpacity,activeOpacity = _ref2.activeOpacity,disabledOpacity = _ref2.disabledOpacity;

  var textDecoration = function () {
    switch (variant) {
      case _types.LinkVariant.Inline:
      case _types.LinkVariant.Mono:
        return "text-decoration: underline;";
      default:
        return "text-decoration: none;";}

  }();

  return "\n  ".concat(

  variant === _types.LinkVariant.Inline ? (0,
  _styledComponents.css)(["display:inline;"]) : (0,


  _styledComponents.css)(["display:inline-flex;flex-direction:row;align-items:center;"]), "\n    color: ").concat(





  color, ";\n    font-weight: 500;\n    ").concat(

  textDecoration, "\n\n    transition: 0.2s ").concat(

  _animation.transitionDefault, ";\n    transition-property: color, outline, opacity, box-shadow;\n\n    svg { \n      fill: ").concat(



  color, "; \n      transition: fill 0.2s ").concat(
  _animation.transitionDefault, ";\n    }\n\n    :hover {\n      color: ").concat(



  hoverColor, ";\n      text-decoration: underline;\n      opacity: ").concat(

  hoverOpacity, ";\n\n      svg {\n        fill: ").concat(


  hoverColor, ";\n      }\n    }\n\n    :not(:active):focus {\n      color: ").concat(




  focusColor, ";\n      box-shadow: 0 0 0 2px ").concat(
  (0, _cssVariables.default)('purple-7'), ";\n    }\n\n    :active {\n      color: ").concat(



  activeColor, ";\n      opacity: ").concat(
  activeOpacity, ";\n\n      svg {\n        fill: ").concat(


  (0, _cssVariables.default)('neutral-9'), ";\n      }\n    }   \n\n    &[aria-disabled=\"true\"] {\n      pointer-events: none;\n      color: ").concat(





  disabledColor, ";\n      opacity: ").concat(
  disabledOpacity, ";\n\n      svg {\n        fill: ").concat(


  (0, _cssVariables.default)('neutral-5'), ";\n      }\n    }\n  ");



};

var Anchor = _styledComponents.default.a.withConfig({ displayName: "link__Anchor", componentId: "vp9jy3-0" })(["", ""],
anchorCSS);


var StyledReactRouterDomLink = (0, _styledComponents.default)(_reactRouterDom.Link).withConfig({ displayName: "link__StyledReactRouterDomLink", componentId: "vp9jy3-1" })(["", ""],
anchorCSS);


var DecorativeIcon = (0, _styledComponents.default)(_icon.Icon).withConfig({ displayName: "link__DecorativeIcon", componentId: "vp9jy3-2" })(["margin-left:4px;"]);



var Link = /*#__PURE__*/(0, _react.forwardRef)(
function (_ref3,

ref)
{var children = _ref3.children,iconProps = _ref3.iconProps,_ref3$variant = _ref3.variant,variant = _ref3$variant === void 0 ? _types.LinkVariant.Default : _ref3$variant,_ref3$disabled = _ref3.disabled,disabled = _ref3$disabled === void 0 ? false : _ref3$disabled,useReactRouter = _ref3.useReactRouter,href = _ref3.href,anchorAttributes = _objectWithoutProperties(_ref3, ["children", "iconProps", "variant", "disabled", "useReactRouter", "href"]);
  if (useReactRouter && href) {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsxs)(StyledReactRouterDomLink, _objectSpread(_objectSpread(_objectSpread({
        to: href,
        variant: variant,
        innerRef: ref },
      anchorAttributes),
      disabled ? { tabIndex: -1, role: 'button', 'aria-disabled': true } : null), {}, { children: [

        children,
        iconProps && /*#__PURE__*/(0, _jsxRuntime.jsx)(DecorativeIcon, _objectSpread({}, iconProps))] })));


  }

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(Anchor, _objectSpread(_objectSpread(_objectSpread({
      href: href,
      variant: variant,
      ref: ref },
    anchorAttributes),
    disabled ? { tabIndex: -1, role: 'button', 'aria-disabled': true } : null), {}, { children: [

      children,
      iconProps && /*#__PURE__*/(0, _jsxRuntime.jsx)(DecorativeIcon, _objectSpread({}, iconProps))] })));


});exports.Link = Link;
