// @pipedrive/insights-core/lib/builders/nestedQuery/reports/deals/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRevenueForecastQueries = exports.getStatsQueries = exports.getRevenueMovementQuery = exports.getRevenueQuery = exports.getProgressQueries = exports.getOverallConversionQueries = exports.getFunnelConversionQuery = exports.getDurationQueries = void 0;
const duration_1 = __importDefault(require("./duration"));
exports.getDurationQueries = duration_1.default;
const funnelConversion_1 = __importDefault(require("./funnelConversion"));
exports.getFunnelConversionQuery = funnelConversion_1.default;
const overallConversion_1 = __importDefault(require("./overallConversion"));
exports.getOverallConversionQueries = overallConversion_1.default;
const progress_1 = __importDefault(require("./progress"));
exports.getProgressQueries = progress_1.default;
const revenue_1 = __importDefault(require("./revenue"));
exports.getRevenueQuery = revenue_1.default;
const revenueMovement_1 = __importDefault(require("./revenueMovement"));
exports.getRevenueMovementQuery = revenueMovement_1.default;
const stats_1 = __importDefault(require("./stats"));
exports.getStatsQueries = stats_1.default;
const revenueForecast_1 = __importDefault(require("./revenueForecast"));
exports.getRevenueForecastQueries = revenueForecast_1.default;
