// feather/lib/components/codeSnippet/index.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.symbol.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.symbol.iterator.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.array.slice.js");require("core-js/modules/es.function.name.js");require("core-js/modules/es.weak-map.js");require("core-js/modules/es.object.get-own-property-descriptor.js");Object.defineProperty(exports, "__esModule", { value: true });exports.CodeSnippet = exports.Language = exports.CodeSnippetTheme = void 0;require("core-js/modules/es.array.concat.js");require("core-js/modules/es.array.map.js");require("core-js/modules/es.array.from.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/es.regexp.exec.js");require("core-js/modules/es.string.split.js");require("core-js/modules/es.string.trim.js");var _react = require("react");
var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _copyToClipboard = _interopRequireDefault(require("copy-to-clipboard"));
var _prismjs = _interopRequireDefault(require("prismjs"));
require("prismjs/components/prism-bash");
require("prismjs/components/prism-c");
require("prismjs/components/prism-csharp");
require("prismjs/components/prism-http");
require("prismjs/components/prism-java");
require("prismjs/components/prism-json");
require("prismjs/components/prism-markup");
require("prismjs/components/prism-objectivec");
require("prismjs/components/prism-python");
require("prismjs/components/prism-ruby");
require("prismjs/components/prism-swift");

var _animation = require("../../../lib/animation");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _themes = require("../../../lib/themes");

var _button = require("../button");
var _ScrollBar = require("../ScrollBar");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function _slicedToArray(arr, i) {return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();}function _nonIterableRest() {throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");}function _unsupportedIterableToArray(o, minLen) {if (!o) return;if (typeof o === "string") return _arrayLikeToArray(o, minLen);var n = Object.prototype.toString.call(o).slice(8, -1);if (n === "Object" && o.constructor) n = o.constructor.name;if (n === "Map" || n === "Set") return Array.from(o);if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);}function _arrayLikeToArray(arr, len) {if (len == null || len > arr.length) len = arr.length;for (var i = 0, arr2 = new Array(len); i < len; i++) {arr2[i] = arr[i];}return arr2;}function _iterableToArrayLimit(arr, i) {var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]);if (_i == null) return;var _arr = [];var _n = true;var _d = false;var _s, _e;try {for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {_arr.push(_s.value);if (i && _arr.length === i) break;}} catch (err) {_d = true;_e = err;} finally {try {if (!_n && _i["return"] != null) _i["return"]();} finally {if (_d) throw _e;}}return _arr;}function _arrayWithHoles(arr) {if (Array.isArray(arr)) return arr;}var

CodeSnippetTheme;exports.CodeSnippetTheme = CodeSnippetTheme;(function (CodeSnippetTheme) {CodeSnippetTheme["Light"] = "light";CodeSnippetTheme["Dark"] = "dark";})(CodeSnippetTheme || (exports.CodeSnippetTheme = CodeSnippetTheme = {}));var




Language;exports.Language = Language;(function (Language) {Language["Bash"] = "bash";Language["C"] = "c";Language["CSharp"] = "csharp";Language["Http"] = "http";Language["Java"] = "java";Language["JavaScript"] = "javascript";Language["Json"] = "json";Language["Kotlin"] = "kotlin";Language["Markup"] = "markup";Language["Objectivec"] = "objectivec";Language["Python"] = "python";Language["Ruby"] = "ruby";Language["Swift"] = "swift";Language["TypeScript"] = "typescript";})(Language || (exports.Language = Language = {}));



































var defaultTheme = CodeSnippetTheme.Light;
var fontStyle = { fontSize: 13, lineHeight: 20 };
var codeBlockTopPadding = 16;

var CodeTheme = {
  light: (0, _styledComponents.css)(["color:", ";.token.selector,.token.attr-name,.token.string,.token.char,.token.inserted{color:", ";}.token.atrule,.token.attr-value,.token.keyword{color:", ";}.token.function{color:", ";}.token.class-name{color:", ";}.token.property{color:", ";}.token.boolean{color:", ";}.token.punctuation{color:", ";&.annotation{color:", ";}}.token.operator{color:", ";}.token.comment,.token.prolog,.token.doctype,.token.cdata{color:", ";}.token.tag,.token.number,.token.constant,.token.symbol,.token.deleted{color:", ";}.token.builtin{color:", ";}.token.entity,.token.url,.language-css .token.string,.style .token.string{color:white;}.token.regex,.token.important,.token.variable{color:", ";}"],
  (0, _cssVariables.default)('neutral-10'),





  (0, _cssVariables.default)('purple-7'),





  (0, _cssVariables.default)('purple-9'),



  (0, _cssVariables.default)('blue-7'),



  (0, _cssVariables.default)('blue-5'),



  (0, _cssVariables.default)('neutral-10'),



  (0, _cssVariables.default)('green-5'),



  (0, _cssVariables.default)('red-4'),

  (0, _cssVariables.default)('green-4'),




  (0, _cssVariables.default)('red-6'),






  (0, _cssVariables.default)('neutral-6'),







  (0, _cssVariables.default)('orange-5'),



  (0, _cssVariables.default)('neutral-9'),












  (0, _cssVariables.default)('neutral-6')),


  dark: (0, _styledComponents.css)(["color:white;.token.selector,.token.attr-name,.token.string,.token.char,.token.inserted{color:", ";}.token.atrule,.token.attr-value,.token.keyword{color:#b4a4ff;}.token.function{color:", ";}.token.class-name{color:", ";}.token.property{color:", ";}.token.boolean{color:", ";}.token.punctuation{color:", ";&.annotation{color:", ";}}.token.operator{color:", ";}.token.comment,.token.prolog,.token.doctype,.token.cdata{color:", ";}.token.tag,.token.number,.token.constant,.token.symbol,.token.deleted{color:", ";}.token.builtin{color:", ";}.token.entity,.token.url,.language-css .token.string,.style .token.string{color:white;}.token.regex,.token.important,.token.variable{color:", ";}"],






  (0, _cssVariables.default)('purple-3'),









  (0, _cssVariables.default)('blue-4'),



  (0, _cssVariables.default)('blue-3'),



  (0, _cssVariables.default)('violet-3'),



  (0, _cssVariables.default)('green-5'),



  (0, _cssVariables.default)('red-4'),

  (0, _cssVariables.default)('green-3'),




  (0, _cssVariables.default)('red-5'),






  (0, _cssVariables.default)('neutral-6'),







  (0, _cssVariables.default)('orange-4'),



  (0, _cssVariables.default)('neutral-6'),












  (0, _cssVariables.default)('neutral-6')) };




var themeMap = {
  dark: _themes.FeatherThemeEnum.Neutral,
  light: _themes.FeatherThemeEnum.Gray };


var transitionWithProperties = function transitionWithProperties(properties) {return (0, _styledComponents.css)(["transition-property:", ";transition-duration:0.2s;transition-timing-function:", ";"],
  properties,

  _animation.transitionDefault);};


var CodeContent = _styledComponents.default.pre.withConfig({ displayName: "codeSnippet__CodeContent", componentId: "sc-1b5p2dt-0" })(["font-family:'Roboto Mono',monospace;display:inline-block;text-align:left;white-space:pre;width:100%;tab-size:4;hyphens:none;> span{vertical-align:baseline;}", " .namespace{opacity:0.7;}.token.important,.token.bold{font-weight:bold;}.token.italic{font-style:italic;}.token.entity{cursor:help;}", ";", ";"],













function (_ref) {var lineWrap = _ref.lineWrap;return (
    lineWrap && (0,
    _styledComponents.css)(["white-space:pre-wrap;span{word-break:break-all;}"]));},
























function (_ref2) {var theme = _ref2.theme;return CodeTheme[theme];},
transitionWithProperties('color'));


var IconContent = _styledComponents.default.div.withConfig({ displayName: "codeSnippet__IconContent", componentId: "sc-1b5p2dt-1" })(["display:grid;grid-auto-flow:column;grid-gap:4px;margin-left:auto;"]);






var ScrollbarWrapper = _styledComponents.default.div.withConfig({ displayName: "codeSnippet__ScrollbarWrapper", componentId: "sc-1b5p2dt-2" })(["@import url('https://fonts.googleapis.com/css?family=Roboto+Mono&display=swap');position:relative;display:flex;justify-content:space-between;padding:", "px;padding-bottom:0;", " font-size:", "px;letter-spacing:-0.3px;line-height:", "px;", " -webkit-overflow-scrolling:touch;", ";"],




codeBlockTopPadding,

function (_ref3) {var hasLineNumber = _ref3.hasLineNumber;return hasLineNumber && 'padding-left: 0;';},

fontStyle.fontSize,

fontStyle.lineHeight,

function (_ref4) {var rows = _ref4.rows;
  if (rows) {
    if (rows === 1) {
      return (0, _styledComponents.css)(["height:40px;padding:0 16px;align-items:center;", "{display:flex;align-items:center;height:40px;flex:1;overflow-x:auto;}", "{margin-left:36px;}"],



      CodeContent,






      IconContent);



    }
    return (0, _styledComponents.css)(["height:", "px;"],
    fontStyle.lineHeight * rows + codeBlockTopPadding);

  }
},



transitionWithProperties('background'));


ScrollbarWrapper.displayName = 'ScrollbarWrapper';

var LineNumber = _styledComponents.default.span.withConfig({ displayName: "codeSnippet__LineNumber", componentId: "sc-1b5p2dt-3" })(["position:sticky;pointer-events:none;display:flex;flex-direction:column;align-items:flex-start;top:0;left:0;min-width:36px;padding-left:8px;margin-right:1rem;border-right:1px solid ", ";font-family:'Roboto Mono',monospace;letter-spacing:-1px;", ";user-select:none;"],











function (props) {return props.$theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('neutral-3') : (0, _cssVariables.default)('neutral-8');},



transitionWithProperties('background, border-right'));




LineNumber.displayName = 'LineNumber';

var LineNumberRow = _styledComponents.default.span.withConfig({ displayName: "codeSnippet__LineNumberRow", componentId: "sc-1b5p2dt-4" })(["pointer-events:none;display:block;color:", ";display:block;padding-right:0.8em;text-align:right;"],


(0, _cssVariables.default)('neutral-6'));





LineNumberRow.displayName = 'LineNumberRow';

var ScrollContainer = _styledComponents.default.div.withConfig({ displayName: "codeSnippet__ScrollContainer", componentId: "sc-1b5p2dt-5" })(["display:flex;"]);



var Tabs = _styledComponents.default.div.withConfig({ displayName: "codeSnippet__Tabs", componentId: "sc-1b5p2dt-6" })(["display:flex;align-items:center;height:40px;border-bottom:1px solid ", ";padding:0 16px;padding-right:8px;border-radius:4px 4px 0 0;", ";"],




function (props) {return (
    props.$theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('border-3') : (0, _cssVariables.default)('border-inverse-2'));},




transitionWithProperties('border-bottom'));


Tabs.displayName = 'Tabs';

var Tab = _styledComponents.default.span.withConfig({ displayName: "codeSnippet__Tab", componentId: "sc-1b5p2dt-7" })(["display:inline-block;position:relative;padding:10px 0;font-size:14px;font-weight:600;line-height:1.43;cursor:pointer;color:", ";&::before{content:'';display:block;position:absolute;bottom:0;left:0;width:100%;height:2px;background:transparent;transition:background 0.2s ", ";}", " ", ";& + &{margin-left:24px;}"],








function (props) {return (
    props.$theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('content-3') : (0, _cssVariables.default)('content-inverse-2'));},










_animation.transitionDefault,


function (_ref5) {var isActive = _ref5.isActive,$theme = _ref5.$theme;return (
    isActive && "\n      color: ".concat(

    $theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('content-primary') : (0, _cssVariables.default)('purple-3'), ";\n      &::before {\n        \n        background: ").concat(


    $theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('bg-primary') : (0, _cssVariables.default)('purple-3'), ";\n      }\n    "));},



transitionWithProperties('background'));






Tab.displayName = 'Tab';

var Container = _styledComponents.default.div.withConfig({ displayName: "codeSnippet__Container", componentId: "sc-1b5p2dt-8" })(["", " background:", ";border-radius:4px;", ";"],
function (_ref6) {var $theme = _ref6.$theme;
  var backgroundColor = $theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('neutral-2') : (0, _cssVariables.default)('neutral-9');

  return (0, _styledComponents.css)(["background:", ";", "{background:", ";}"],
  backgroundColor,
  LineNumber,
  backgroundColor);


},
function (_ref7) {var $theme = _ref7.$theme;return (
    $theme === CodeSnippetTheme.Light ? (0, _cssVariables.default)('bg-disabled') : (0, _cssVariables.default)('bg-black'));},

transitionWithProperties('background'));


var CodeSnippet = function CodeSnippet(_ref8)










{var _ref8$data = _ref8.data,dataProp = _ref8$data === void 0 ? [] : _ref8$data,onClickTab = _ref8.onClickTab,onChangeTheme = _ref8.onChangeTheme,_ref8$activeIndex = _ref8.activeIndex,activeIndex = _ref8$activeIndex === void 0 ? 0 : _ref8$activeIndex,_ref8$startLineNumber = _ref8.startLineNumber,startLineNumber = _ref8$startLineNumber === void 0 ? 1 : _ref8$startLineNumber,_ref8$hasLineNumber = _ref8.hasLineNumber,hasLineNumber = _ref8$hasLineNumber === void 0 ? false : _ref8$hasLineNumber,className = _ref8.className,_ref8$theme = _ref8.theme,theme = _ref8$theme === void 0 ? defaultTheme : _ref8$theme,rows = _ref8.rows,_ref8$lineWrap = _ref8.lineWrap,lineWrap = _ref8$lineWrap === void 0 ? false : _ref8$lineWrap;
  var codeRef = (0, _react.useRef)(null);
  var _useState = (0, _react.useState)(false),_useState2 = _slicedToArray(_useState, 2),isCopied = _useState2[0],setIsCopied = _useState2[1];

  var data = (0, _react.useMemo)(function () {
    if (typeof dataProp === 'string') {
      return [{ code: dataProp }];
    }
    return dataProp;
  }, [dataProp]);

  var highlightCode = (0, _react.useRef)(function () {});

  (0, _react.useEffect)(function () {
    highlightCode.current = function () {
      if (codeRef.current && data[activeIndex].code && _prismjs.default) {
        _prismjs.default.highlightElement(codeRef.current);
      }
    };
  }, [data, activeIndex]);

  (0, _react.useEffect)(function () {
    highlightCode.current();
  }, [theme, activeIndex]);

  (0, _react.useEffect)(function () {
    // For Next.js apps, we have to re-highlight the code right after the document has finished its loading.
    var documentReadyStateChangeListener = function documentReadyStateChangeListener() {
      if (document.readyState === 'complete') {
        highlightCode.current();
      }
    };

    document.addEventListener('readystatechange', documentReadyStateChangeListener);
    return function () {
      document.removeEventListener('readystatechange', documentReadyStateChangeListener);
    };
  }, []);

  var handleClickTab = (0, _react.useCallback)(
  function (index) {return function () {
      onClickTab && onClickTab(index);
    };},
  [onClickTab]);


  var handleClickToggleTheme = (0, _react.useCallback)(function () {
    onChangeTheme === null || onChangeTheme === void 0 ? void 0 : onChangeTheme(theme, theme === CodeSnippetTheme.Light ? CodeSnippetTheme.Dark : CodeSnippetTheme.Light);
  }, [theme, onChangeTheme]);

  var handleClickCopy = (0, _react.useCallback)(function () {
    (0, _copyToClipboard.default)(data[activeIndex].code);
    setIsCopied(true);
    setTimeout(function () {return setIsCopied(false);}, 2500);
  }, [data, activeIndex]);

  var renderIconContent = (0, _react.useMemo)(function () {
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsxs)(IconContent, { children: [
        onChangeTheme && /*#__PURE__*/
        (0, _jsxRuntime.jsx)(_button.IconButton, {
          icon: theme === CodeSnippetTheme.Light ? 'theme-dark' : 'theme-light',
          buttonType: "secondary",
          size: "xsmall",
          onClick: handleClickToggleTheme,
          title: theme === CodeSnippetTheme.Light ? 'Dark theme skin' : 'Light theme skin',
          "data-test-id": "ThemeButton" }), /*#__PURE__*/


        (0, _jsxRuntime.jsx)(_button.IconButton, {
          icon: "copy",
          buttonType: "secondary",
          size: "xsmall",
          onClick: handleClickCopy,
          title: isCopied ? 'Copied!' : 'Copy' })] }));



  }, [onChangeTheme, theme, handleClickToggleTheme, handleClickCopy, isCopied]);

  var activeCode = data[activeIndex].code;

  var lineNumbers = (0, _react.useMemo)(
  function () {return (
      hasLineNumber && /*#__PURE__*/
      (0, _jsxRuntime.jsx)(LineNumber, { "aria-hidden": "true", className: "line-numbers-rows", $theme: theme, children:
        Array.from(Array(activeCode.trim().split('\n').length)).map(function (_, index) {return /*#__PURE__*/(
            (0, _jsxRuntime.jsx)(LineNumberRow, { children: startLineNumber + index }, "line-number-".concat(index)));}) }));},



  [activeCode, hasLineNumber, startLineNumber, theme]);


  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(_themes.FeatherThemeContext.Provider, { value: themeMap[theme], children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)(Container, { className: className, $theme: theme, children:
        rows === 1 ? /*#__PURE__*/
        (0, _jsxRuntime.jsxs)(ScrollbarWrapper, { hasLineNumber: false, rows: rows, children: [/*#__PURE__*/
          (0, _jsxRuntime.jsx)(CodeContent, {
            ref: codeRef,
            className: data[activeIndex].language && "language-".concat(data[activeIndex].language),
            lineWrap: lineWrap,
            theme: theme, children:

            activeCode }),

          renderIconContent] }) : /*#__PURE__*/


        (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, { children: [/*#__PURE__*/
          (0, _jsxRuntime.jsxs)(Tabs, { $theme: theme, children: [
            data.length > 1 &&
            data.map(function (item, index) {return /*#__PURE__*/(
                (0, _jsxRuntime.jsx)(Tab, {

                  isActive: index === activeIndex,
                  $theme: theme,
                  onClick: handleClickTab(index), children:

                  item.label }, "".concat(item.label, "-").concat(index)));}),


            renderIconContent] }), /*#__PURE__*/

          (0, _jsxRuntime.jsx)(ScrollbarWrapper, { className: hasLineNumber ? 'line-numbers' : '', hasLineNumber: hasLineNumber, rows: rows, children: /*#__PURE__*/
            (0, _jsxRuntime.jsx)(_ScrollBar.ScrollBar, { style: { paddingBottom: 16 }, children: /*#__PURE__*/
              (0, _jsxRuntime.jsxs)(ScrollContainer, { lineWrap: lineWrap, children: [
                lineNumbers, /*#__PURE__*/
                (0, _jsxRuntime.jsx)(CodeContent, {
                  ref: codeRef,
                  className: data[activeIndex].language && "language-".concat(data[activeIndex].language),
                  lineWrap: lineWrap,
                  theme: theme, children:

                  activeCode })] }) }) })] }) }) }));









};exports.CodeSnippet = CodeSnippet;
