// @elastic/charts/dist/chart_types/xy_chart/utils/axis_type_utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHorizontalGrid = exports.isVerticalGrid = exports.isHorizontalAxis = exports.isVerticalAxis = exports.isBounded = exports.isCompleteBound = exports.isUpperBound = exports.isLowerBound = void 0;
var common_1 = require("../../../utils/common");
function isLowerBound(domain) {
    return domain.min != null;
}
exports.isLowerBound = isLowerBound;
function isUpperBound(domain) {
    return domain.max != null;
}
exports.isUpperBound = isUpperBound;
function isCompleteBound(domain) {
    return domain.max != null && domain.min != null;
}
exports.isCompleteBound = isCompleteBound;
function isBounded(domain) {
    return domain.max != null || domain.min != null;
}
exports.isBounded = isBounded;
function isVerticalAxis(axisPosition) {
    return axisPosition === common_1.Position.Left || axisPosition === common_1.Position.Right;
}
exports.isVerticalAxis = isVerticalAxis;
function isHorizontalAxis(axisPosition) {
    return axisPosition === common_1.Position.Top || axisPosition === common_1.Position.Bottom;
}
exports.isHorizontalAxis = isHorizontalAxis;
function isVerticalGrid(axisPosition) {
    return isHorizontalAxis(axisPosition);
}
exports.isVerticalGrid = isVerticalGrid;
function isHorizontalGrid(axisPosition) {
    return isVerticalAxis(axisPosition);
}
exports.isHorizontalGrid = isHorizontalGrid;
