// @elastic/charts/dist/chart_types/xy_chart/utils/grid_lines.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGridLineForHorizontalAxisAt = exports.getGridLineForVerticalAxisAt = exports.getGridLinesForSpec = exports.getGridLines = void 0;
var color_library_wrappers_1 = require("../../../common/color_library_wrappers");
var common_1 = require("../../../utils/common");
var line_1 = require("../renderer/canvas/primitives/line");
var axis_type_utils_1 = require("./axis_type_utils");
var panel_1 = require("./panel");
var panel_utils_1 = require("./panel_utils");
function getGridLines(axesSpecs, axesGeoms, themeAxisStyle, scales) {
    var panelSize = panel_1.getPanelSize(scales);
    return panel_utils_1.getPerPanelMap(scales, function () {
        var lines = axesGeoms.reduce(function (linesAcc, _a) {
            var axis = _a.axis, visibleTicks = _a.visibleTicks;
            var axisSpec = axesSpecs.find(function (_a) {
                var id = _a.id;
                return id === axis.id;
            });
            if (!axisSpec) {
                return linesAcc;
            }
            var linesForSpec = getGridLinesForSpec(axisSpec, visibleTicks, themeAxisStyle, panelSize);
            if (!linesForSpec) {
                return linesAcc;
            }
            return __spread(linesAcc, [linesForSpec]);
        }, []);
        return { lineGroups: lines };
    });
}
exports.getGridLines = getGridLines;
function getGridLinesForSpec(axisSpec, visibleTicks, themeAxisStyle, panelSize) {
    var _a;
    var isVertical = axis_type_utils_1.isVerticalAxis(axisSpec.position);
    var axisStyle = common_1.mergePartial(themeAxisStyle, axisSpec.style, {
        mergeOptionalPartialValues: true,
    });
    var gridLineThemeStyle = isVertical ? axisStyle.gridLine.vertical : axisStyle.gridLine.horizontal;
    var gridLineStyles = axisSpec.gridLine ? common_1.mergePartial(gridLineThemeStyle, axisSpec.gridLine) : gridLineThemeStyle;
    var showGridLines = (_a = axisSpec.showGridLines) !== null && _a !== void 0 ? _a : gridLineStyles.visible;
    if (!showGridLines) {
        return null;
    }
    var lines = visibleTicks.map(function (tick) {
        return isVertical
            ? getGridLineForVerticalAxisAt(tick.position, panelSize)
            : getGridLineForHorizontalAxisAt(tick.position, panelSize);
    });
    if (!gridLineStyles.stroke || !gridLineStyles.strokeWidth || gridLineStyles.strokeWidth < line_1.MIN_STROKE_WIDTH) {
        return null;
    }
    var strokeColor = color_library_wrappers_1.stringToRGB(gridLineStyles.stroke);
    strokeColor.opacity =
        gridLineStyles.opacity !== undefined ? strokeColor.opacity * gridLineStyles.opacity : strokeColor.opacity;
    var stroke = {
        color: strokeColor,
        width: gridLineStyles.strokeWidth,
        dash: gridLineStyles.dash,
    };
    return {
        lines: lines,
        stroke: stroke,
        axisId: axisSpec.id,
    };
}
exports.getGridLinesForSpec = getGridLinesForSpec;
function getGridLineForVerticalAxisAt(tickPosition, panelSize) {
    return { x1: 0, y1: tickPosition, x2: panelSize.width, y2: tickPosition };
}
exports.getGridLineForVerticalAxisAt = getGridLineForVerticalAxisAt;
function getGridLineForHorizontalAxisAt(tickPosition, panelSize) {
    return { x1: tickPosition, y1: 0, x2: tickPosition, y2: panelSize.height };
}
exports.getGridLineForHorizontalAxisAt = getGridLineForHorizontalAxisAt;
