// anchor-server-common/utilities/user/constants.js
module.exports = {
    EXTERNAL_ID_TYPES: {
        FACEBOOK: 'facebook',
        GOOGLE: 'google',
        PODCAST_NETWORK: 'podcastnetwork',
        PODCAST_NETWORK_ROLE: 'podcast_network_role',
        STRIPE: 'stripe',
        STRIPE_LINK: 'stripe_link',
        CREATION_PLATFORM: 'creation_platform',
        ACQUISITION_SOURCE: 'acquisition_source',
        TWITTER: 'twitter',
        IP_ADDRESS: 'ipaddress',
        WORDPRESS_SITE_ID: 'wordpress_site_id',
    },
    PODCAST_NETWORK_ROLE: {
        ADMIN: 'admin',
    },
    APPLE_DEMO_TEST_ACCOUNT_ID: {
        PRODUCTION: 658668,
        DEVELOPMENT: 2571,
    },
    VALID_USER_EMAIL_REGEX: /^(?!\.)((?!.*\.{2})[a-zA-Z0-9\u00E0-\u00FC.!#$%&'"*+-/=?^_`{|}~\-\d]+)@(?!\.)([a-zA-Z0-9\u00E0-\u00FC\-\.\d]+)((\.([a-zA-Z]){2,63})+)$/i,
    FUTURE_FORMATS: {
        // This list should be deprecated after we fully roll out language/country expansion
        ALLOWED_MT_COUNTRIES: [
            'United States',
            'Puerto Rico',
            'United Kingdom',
            'Canada',
            'Australia',
            'New Zealand',
            'Ireland',
        ],
        ALLOWED_MT_LANGUAGE_COUNTRY_MAP: {
            en: [
                'United States',
                'Puerto Rico',
                'United Kingdom',
                'Canada',
                'Australia',
                'New Zealand',
                'Ireland',
                'Brazil',
                'Mexico',
                'Colombia',
                'Indonesia',
                'Philippines',
                'Chile',
                'Argentina',
                'Japan',
                'India',
                'France',
                'Germany',
                'Spain',
                'Italy',
                'Netherlands',
                'Sweden',
            ],
            es: ['United States', 'Puerto Rico', 'Mexico', 'Colombia', 'Chile', 'Argentina', 'Spain',],
            in: ['United States', 'Indonesia'], // the frontend incorrectly sets `in` as the language code for indonesia in station metadata
            id: ['United States', 'Indonesia'],
            pt: ['United States', 'Brazil'],
            ja: ['United States', 'Japan'],
            fr: ['United States', 'France'],
            de: ['United States', 'Germany'],
            it: ['United States', 'Italy'],
            nl: ['United States', 'Netherlands'],
            sv: ['United States', 'Sweden'],
            fi: ['United States', 'Philippines'],
            hi: ['United States', 'India'],
        },
        STATES: {
            ALLOWED_TO_CREATE: 'isAllowedToCreate',
            NOT_ALLOWED_TO_CREATE: 'isNotAllowedToCreate',
        },
    },
};
