// @atlassian/butler-command-parser/lib/parse/stringMatchers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAtMentionVariableValue = exports.isVariableNameValue = exports.isAtMentionValue = exports.isStringValue = exports.isHiddenSymbol = exports.isSymbol = exports.getOptional = exports.isOptional = exports.isTerminal = exports.isAtMention = exports.isVariableName = exports.isStringVariable = exports.isInlineEnd = exports.isEnd = void 0;
function isEnd(string) {
    return string === '<END>';
}
exports.isEnd = isEnd;
function isInlineEnd(string) {
    return string === '<INLINE_END>';
}
exports.isInlineEnd = isInlineEnd;
function isStringVariable(string) {
    return string.match(/^\$[A-Z_]+\??$/) != null;
}
exports.isStringVariable = isStringVariable;
function isVariableName(string) {
    return string.match(/^{[A-Z_]+}$/) != null;
}
exports.isVariableName = isVariableName;
function isAtMention(string) {
    return string.match(/^@[A-Z_]+\??$/) != null;
}
exports.isAtMention = isAtMention;
function isConditional(string) {
    return string.match(/^[A-Z_][A-Z_?:]*\??$/) != null;
}
function isRewrite(string) {
    const split = string.split('=>');
    return split.length === 2 && isSymbol(split[0]) && isSymbol(split[1]);
}
function isTerminal(string) {
    return (!isSymbol(string) &&
        !isConditional(string) &&
        !isRewrite(string) &&
        !isStringVariable(string) &&
        !isVariableName(string) &&
        !isAtMention(string));
}
exports.isTerminal = isTerminal;
function isOptional(string) {
    return !isTerminal(string) && string.slice(-1) === '?';
}
exports.isOptional = isOptional;
function getOptional(string) {
    return string.slice(0, -1);
}
exports.getOptional = getOptional;
function isSymbol(string) {
    return string.match(/^[A-Z_]+\??$/) != null;
}
exports.isSymbol = isSymbol;
function isHiddenSymbol(string) {
    return isSymbol(string) && string[0] === '_';
}
exports.isHiddenSymbol = isHiddenSymbol;
function isStringValue(string) {
    return string[0] === '"' && string[string.length - 1] === '"';
}
exports.isStringValue = isStringValue;
function isAtMentionValue(string) {
    return string.match(/^@.*$/) != null;
}
exports.isAtMentionValue = isAtMentionValue;
function isVariableNameValue(string) {
    return string.match(/^{{[#%]?[a-zA-Z0-9_+-]+}}$|^{\$?[a-zA-Z0-9_+-]+}$/) != null;
}
exports.isVariableNameValue = isVariableNameValue;
function isAtMentionVariableValue(string) {
    return string.match(/^@{.*}$/) != null;
}
exports.isAtMentionVariableValue = isAtMentionVariableValue;
