// @coursera/apollo-link-naptime/dist/apollo-link-naptime/getNaptimeResourceMetadata.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _constants = require("./constants");

// Check if query node has @naptime directive and parse it to return resource name,
// version and action.
var getNaptimeResourceMetadata = function getNaptimeResourceMetadata(node) {
  if (node.directives && node.directives.find(function (d) {
    return d.name.value === 'naptime';
  })) {
    var match = node.name.value.match(/^(\w+)V(\d+)$/);

    if (match) {
      var actionNode = node.selectionSet.selections[0];
      return {
        actionNode: actionNode,
        resourceTypename: "".concat(_constants.TYPENAME_PREFIX).concat(match[1], "V").concat(match[2]),
        resource: match[1].charAt(0).toLowerCase() + match[1].slice(1),
        version: match[2],
        action: actionNode.name.value
      };
    } else {
      throw new Error("Invalid resource name '".concat(node.name.value, "'"));
    }
  }

  return {};
};

var _default = getNaptimeResourceMetadata;
exports.default = _default;
