// @pipedrive/leads-sidebar/dist/NoteEditor/NoteEditor.styles.ts
import styled, { css } from 'styled-components';
import { ReactPdWysiwyg } from '@pipedrive/pd-wysiwyg';
import { colors } from '@pipedrive/convention-ui-css/dist/js/variables';
import { Text } from '@pipedrive/convention-ui-react';

interface StylesRules {
	displayBorder?: boolean;
}

interface PlaceholderProps extends StylesRules {
	withContent: boolean;
}

export const Placeholder = styled.div<PlaceholderProps>`
	display: flex;
	align-items: center;
	width: 100%;
	padding: 20px;
	border: ${(props) => props.displayBorder && `solid 1px ${colors.black12}`};
	box-sizing: border-box;
	color: ${colors.black32Opaque};
	cursor: pointer;
	&:hover {
		background-color: ${colors.black4};
		border-color: ${colors.black12};
	}

	${({ withContent }) =>
		withContent &&
		css`
			background-color: ${colors.noteYellow};
			&:hover {
				background-color: ${colors.noteYellow};
			}
		`}
`;

interface WrapperProps extends StylesRules {
	error: boolean;
}

export const Wrapper = styled.div<WrapperProps>`
	display: flex;
	flex-direction: column;
	background-color: ${colors.noteYellow};
	border: ${(props) => props.displayBorder && `solid 1px ${colors.black12}`};
	word-break: break-word;
	position: relative;

	.editorToolbar {
		border: none;
		background-color: ${colors.noteYellow};
		flex-wrap: wrap;
		justify-content: space-between;
		height: unset;
		margin-top: 8px;
		> * {
			margin-bottom: 8px;
		}
	}

	${({ error }) =>
		error &&
		css`
			border: 1px solid ${colors.red};
		`}
`;

export const Wysiwyg = styled(ReactPdWysiwyg)`
	background-color: ${colors.noteYellow} !important;
`;

export const FoldableNoteContent = styled(Text)`
	/* styles copy pasted from @pipedrive/pd-wysiwyg/src/react/wysiwyg.scss */
	word-break: break-word;
	a {
		color: ${colors.blue};
	}

	b,
	strong {
		font-weight: bold;
	}

	em,
	i {
		font-style: italic;
	}

	u {
		text-decoration: underline;
	}

	ul {
		list-style: disc;
	}

	ol {
		list-style: decimal;
	}

	ul,
	ol {
		margin-top: 1em;
		margin-bottom: 1em;
		padding: 0 2.5em 0 2.5em;
		padding-inline-start: 2.5em;
		padding-inline-end: 0;

		ul,
		ol {
			margin-top: 0;
			margin-bottom: 0;
		}
	}
`;
