// @atlaskit/atlassian-switcher/dist/esm/types.js
export var RecentContainerType;

(function (RecentContainerType) {
  RecentContainerType["JIRA_PROJECT"] = "jira-project";
  RecentContainerType["CONFLUENCE_SPACE"] = "confluence-space";
})(RecentContainerType || (RecentContainerType = {}));

export var CollaborationGraphRecentContainerType;

(function (CollaborationGraphRecentContainerType) {
  CollaborationGraphRecentContainerType["JIRA_PROJECT"] = "jiraProject";
  CollaborationGraphRecentContainerType["CONFLUENCE_SPACE"] = "confluenceSpace";
})(CollaborationGraphRecentContainerType || (CollaborationGraphRecentContainerType = {}));

export var Permissions;

(function (Permissions) {
  Permissions["MANAGE"] = "manage";
  Permissions["CAN_INVITE_USERS"] = "invite-users";
  Permissions["ADD_PRODUCTS"] = "add-products";
})(Permissions || (Permissions = {}));

export var Product;

(function (Product) {
  Product["BITBUCKET"] = "bitbucket";
  Product["CONFLUENCE"] = "confluence";
  Product["HOME"] = "home";
  Product["JIRA"] = "jira";
  Product["SITE_ADMIN"] = "site-admin";
  Product["TRUSTED_ADMIN"] = "trusted-admin";
  Product["TRELLO"] = "trello";
  Product["START"] = "start";
})(Product || (Product = {}));

export var Feature;

(function (Feature) {
  Feature["disableCustomLinks"] = "disableCustomLinks";
  Feature["enableRecentContainers"] = "enableRecentContainers";
  Feature["disableSwitchToHeading"] = "disableSwitchToHeading";
  Feature["xflow"] = "xflow";
  Feature["isProductStoreInTrelloJSWFirstEnabled"] = "isProductStoreInTrelloJSWFirstEnabled";
  Feature["isProductStoreInTrelloConfluenceFirstEnabled"] = "isProductStoreInTrelloConfluenceFirstEnabled";
  Feature["isSlackDiscoveryEnabled"] = "isSlackDiscoveryEnabled";
  Feature["isTrustedAdminUIDeprecationEnabled"] = "isTrustedAdminUIDeprecationEnabled";
})(Feature || (Feature = {}));

// Coincides with the product types in the Available Products Service
export var SwitcherProductType;

(function (SwitcherProductType) {
  SwitcherProductType["JIRA_BUSINESS"] = "JIRA_BUSINESS";
  SwitcherProductType["JIRA_WORK_MANAGEMENT"] = "JIRA_WORK_MANAGEMENT";
  SwitcherProductType["JIRA_SERVICE_DESK"] = "JIRA_SERVICE_DESK";
  SwitcherProductType["JIRA_SOFTWARE"] = "JIRA_SOFTWARE";
  SwitcherProductType["CONFLUENCE"] = "CONFLUENCE";
  SwitcherProductType["OPSGENIE"] = "OPSGENIE";
  SwitcherProductType["BITBUCKET"] = "BITBUCKET";
  SwitcherProductType["STATUSPAGE"] = "STATUSPAGE";
  SwitcherProductType["TRELLO"] = "TRELLO";
  SwitcherProductType["DRAGONFRUIT"] = "DRAGONFRUIT";
  SwitcherProductType["TEAM_CENTRAL"] = "TEAM_CENTRAL";
  SwitcherProductType["AVOCADO"] = "AVOCADO";
})(SwitcherProductType || (SwitcherProductType = {}));

export var AnalyticsItemType;

(function (AnalyticsItemType) {
  AnalyticsItemType["PRODUCT"] = "product";
  AnalyticsItemType["ADMIN"] = "admin";
  AnalyticsItemType["TRY"] = "try";
  AnalyticsItemType["JOIN"] = "join";
  AnalyticsItemType["CUSTOM_LINK"] = "customLink";
  AnalyticsItemType["DISCOVER_FIXED_LINKS"] = "discover-fixed-links";
  AnalyticsItemType["RECENT"] = "recent";
})(AnalyticsItemType || (AnalyticsItemType = {}));

export var ProductKey;

(function (ProductKey) {
  ProductKey["CONFLUENCE"] = "confluence.ondemand";
  ProductKey["JIRA_CORE"] = "jira-core.ondemand";
  ProductKey["JIRA_SOFTWARE"] = "jira-software.ondemand";
  ProductKey["JIRA_SERVICE_DESK"] = "jira-servicedesk.ondemand";
  ProductKey["BITBUCKET"] = "bitbucket";
  ProductKey["OPSGENIE"] = "opsgenie";
  ProductKey["STATUSPAGE"] = "statuspage";
  ProductKey["TRELLO"] = "trello";
  ProductKey["DRAGONFRUIT"] = "dragonfruit";
  ProductKey["TEAM_CENTRAL"] = "townsquare";
  ProductKey["AVOCADO"] = "avocado";
})(ProductKey || (ProductKey = {}));

export var DiscoverLinkItemKeys;

(function (DiscoverLinkItemKeys) {
  DiscoverLinkItemKeys["DISCOVER_MORE"] = "discover-more";
  DiscoverLinkItemKeys["GIT_TOOLS"] = "appswitcher.git.tools";
  DiscoverLinkItemKeys["SLACK_INTEGRATION"] = "slack-integration";
})(DiscoverLinkItemKeys || (DiscoverLinkItemKeys = {}));
