// feather/lib/components/Tag/index.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.symbol.iterator.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.array.slice.js");require("core-js/modules/es.function.name.js");require("core-js/modules/es.array.from.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");require("core-js/modules/es.weak-map.js");Object.defineProperty(exports, "__esModule", { value: true });exports.AvatarTag = exports.Tag = exports.TagVariant = void 0;var _react = require("react");
var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _avatar = require("../../../lib/components/avatar");
var _tooltip = require("../../../lib/components/tooltip");

var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _typography = require("../../../lib/typography");

var _icon = require("../icon");var _jsxRuntime = require("react/jsx-runtime");var _themes;function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _slicedToArray(arr, i) {return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();}function _nonIterableRest() {throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");}function _unsupportedIterableToArray(o, minLen) {if (!o) return;if (typeof o === "string") return _arrayLikeToArray(o, minLen);var n = Object.prototype.toString.call(o).slice(8, -1);if (n === "Object" && o.constructor) n = o.constructor.name;if (n === "Map" || n === "Set") return Array.from(o);if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);}function _arrayLikeToArray(arr, len) {if (len == null || len > arr.length) len = arr.length;for (var i = 0, arr2 = new Array(len); i < len; i++) {arr2[i] = arr[i];}return arr2;}function _iterableToArrayLimit(arr, i) {var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]);if (_i == null) return;var _arr = [];var _n = true;var _d = false;var _s, _e;try {for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {_arr.push(_s.value);if (i && _arr.length === i) break;}} catch (err) {_d = true;_e = err;} finally {try {if (!_n && _i["return"] != null) _i["return"]();} finally {if (_d) throw _e;}}return _arr;}function _arrayWithHoles(arr) {if (Array.isArray(arr)) return arr;}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}var

TagVariant;exports.TagVariant = TagVariant;(function (TagVariant) {TagVariant["Light"] = "light";TagVariant["Dark"] = "dark";})(TagVariant || (exports.TagVariant = TagVariant = {}));









































var TagIcon = (0, _styledComponents.default)(function (_ref) {var icon = _ref.icon,size = _ref.size,className = _ref.className;
  var sizes = { small: 12, medium: 16, large: 20 };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_icon.Icon, { icon: icon, className: className, size: sizes[size], color: (0, _cssVariables.default)('neutral-6') });
}).withConfig({ displayName: "Tag__TagIcon", componentId: "e9szmh-0" })(["flex:none;"]);



var Remove = (0, _styledComponents.default)(
function (_ref2) {var size = _ref2.size,className = _ref2.className,onClick = _ref2.onClick;
  var sizes = { small: 12, medium: 12, large: 16 };
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)("button", { type: "button", className: className, onClick: onClick, children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)(_icon.Icon, { icon: "close", size: sizes[size], color: (0, _cssVariables.default)('neutral-6') }) }));


}).withConfig({ displayName: "Tag__Remove", componentId: "e9szmh-1" })(["flex:none;padding:2px;outline:0;border:0;cursor:pointer;border-radius:2px;background:inherit;display:flex;align-items:center;justify-content:center;"]);













var AvatarIcon = (0, _styledComponents.default)(_avatar.Avatar).withConfig({ displayName: "Tag__AvatarIcon", componentId: "e9szmh-2" })(["flex:none;"]);



var Text = _styledComponents.default.span.withConfig({ displayName: "Tag__Text", componentId: "e9szmh-3" })(["flex:1;min-width:0;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;", ";line-height:16px;text-align:center;"],






_typography.Typography['label-02']);





var Wrapper = _styledComponents.default.div.withConfig({ displayName: "Tag__Wrapper", componentId: "e9szmh-4" })(["display:inline-flex;align-items:center;outline:none;max-width:", "px;& + &{margin-left:4px;}"],



function (_ref3) {var maxWidth = _ref3.maxWidth;return maxWidth;});






var themes = (_themes = {}, _defineProperty(_themes,
TagVariant.Light, (0, _styledComponents.css)(["border:1px solid ", ";color:", ";", " ", "{&:hover,&:focus{background-color:", ";}}"],
(0, _cssVariables.default)('border-2'),
(0, _cssVariables.default)('content-2'),

function (_ref4) {var isInteractive = _ref4.isInteractive;return (
    isInteractive && (0,
    _styledComponents.css)(["&:hover{border-color:", ";}"],

    (0, _cssVariables.default)('purple-7')));},



Remove,


(0, _cssVariables.default)('neutral-2'))), _defineProperty(_themes,



TagVariant.Dark, (0, _styledComponents.css)(["background:", ";border:1px solid ", ";color:", ";", " ", "{&:hover,&:focus{background:", ";}}"],
(0, _cssVariables.default)('neutral-2'),
(0, _cssVariables.default)('neutral-2'),
(0, _cssVariables.default)('neutral-7'),

function (_ref5) {var isInteractive = _ref5.isInteractive;return (
    isInteractive && (0,
    _styledComponents.css)(["&:hover{background:", ";border-color:", ";}"],

    (0, _cssVariables.default)('neutral-3'),
    (0, _cssVariables.default)('neutral-3')));},



Remove,


(0, _cssVariables.default)('neutral-3'))), _themes);





var TooltipContainer = (0, _styledComponents.default)(_tooltip.Tooltip).attrs({
  placement: 'top',
  popperProps: { modifiers: { offset: { offset: '0, 10' } } } }).withConfig({ displayName: "Tag__TooltipContainer", componentId: "e9szmh-5" })(["display:inline-flex;align-items:center;padding:0px 7px 1px;max-width:100%;min-height:20px;", " ", ":focus &{background-color:", ";border-color:", ";}", "{margin-right:4px;margin-left:-3px;}", "{margin-right:4px;margin-left:-6px;}", " ", "{margin-right:-6px;margin-left:4px;}"],









function (props) {return themes[props.tagVariant];},

Wrapper,
(0, _cssVariables.default)('purple-2'),
(0, _cssVariables.default)('purple-7'),


TagIcon,




AvatarIcon,




function (props) {
  if (props.rounded) {
    return (0, _styledComponents.css)(["&,", "{border-radius:20px;}"],

    Remove);



  }
  return (0, _styledComponents.css)(["border-radius:", "px;"],
  props.withRemoveButton ? 3 : 2);

},

Remove);





var ContentWrapper = _styledComponents.default.div.withConfig({ displayName: "Tag__ContentWrapper", componentId: "e9szmh-6" })(["max-width:100%;"]);



var TagTemplate = /*#__PURE__*/(0, _react.forwardRef)(
function (_ref6,















ref)
{var children = _ref6.children,size = _ref6.size,icon = _ref6.icon,variant = _ref6.variant,_ref6$maxWidth = _ref6.maxWidth,maxWidth = _ref6$maxWidth === void 0 ? 240 : _ref6$maxWidth,_ref6$tooltipMaxWidth = _ref6.tooltipMaxWidth,tooltipMaxWidth = _ref6$tooltipMaxWidth === void 0 ? maxWidth : _ref6$tooltipMaxWidth,_ref6$disableTooltip = _ref6.disableTooltip,disableTooltip = _ref6$disableTooltip === void 0 ? false : _ref6$disableTooltip,tooltipProps = _ref6.tooltipProps,rounded = _ref6.rounded,onRemove = _ref6.onRemove,className = _ref6.className,_ref6$isInteractive = _ref6.isInteractive,isInteractiveProp = _ref6$isInteractive === void 0 ? false : _ref6$isInteractive,props = _objectWithoutProperties(_ref6, ["children", "size", "icon", "variant", "maxWidth", "tooltipMaxWidth", "disableTooltip", "tooltipProps", "rounded", "onRemove", "className", "isInteractive"]);
  var withRemoveButton = !!onRemove;
  var isInteractive = isInteractiveProp || !!props.onClick;

  var _useState = (0, _react.useState)(-1),_useState2 = _slicedToArray(_useState, 2),width = _useState2[0],setWidth = _useState2[1];
  var contentRef = (0, _react.useRef)(null);

  /* track content width */
  (0, _react.useLayoutEffect)(function () {var _contentRef$current$g, _contentRef$current;
    var contentWidth = (_contentRef$current$g = (_contentRef$current = contentRef.current) === null || _contentRef$current === void 0 ? void 0 : _contentRef$current.getBoundingClientRect().width) !== null && _contentRef$current$g !== void 0 ? _contentRef$current$g : -1;
    if (width !== contentWidth) {
      setWidth(contentWidth);
    }
  }, [width, children, icon, size]);

  var isOverflow = width >= maxWidth - 16;

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(Wrapper, _objectSpread(_objectSpread({ tabIndex: isInteractive ? 0 : undefined, ref: ref, maxWidth: maxWidth }, props), {}, { children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)(ContentWrapper, { ref: contentRef, children: /*#__PURE__*/
        (0, _jsxRuntime.jsxs)(TooltipContainer, _objectSpread(_objectSpread({
          className: className,
          variant: _tooltip.TooltipVariant.Light,
          tagVariant: variant,
          size: size,
          rounded: rounded,
          withRemoveButton: withRemoveButton,
          disabled: disableTooltip || !isOverflow,
          content: children,
          isInteractive: isInteractive },
        tooltipProps), {}, {
          tooltipContentStyle: (0, _styledComponents.css)(["max-width:", "px;word-break:break-word;", ""],
          tooltipMaxWidth,




          tooltipProps === null || tooltipProps === void 0 ? void 0 : tooltipProps.tooltipContentStyle), children: [


          icon, /*#__PURE__*/
          (0, _jsxRuntime.jsx)(Text, { children: children }),
          withRemoveButton && /*#__PURE__*/(0, _jsxRuntime.jsx)(Remove, { size: size, onClick: onRemove })] })) }) })));




});


var Tag = (0, _styledComponents.default)( /*#__PURE__*/
(0, _react.forwardRef)(
function (_ref7, ref) {var children = _ref7.children,_ref7$size = _ref7.size,size = _ref7$size === void 0 ? 'small' : _ref7$size,_ref7$variant = _ref7.variant,variant = _ref7$variant === void 0 ? TagVariant.Light : _ref7$variant,icon = _ref7.icon,props = _objectWithoutProperties(_ref7, ["children", "size", "variant", "icon"]);
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(TagTemplate, _objectSpread(_objectSpread({},
    props), {}, {
      size: size,
      variant: variant,
      icon: icon && /*#__PURE__*/(0, _jsxRuntime.jsx)(TagIcon, { icon: icon, size: size }),
      ref: ref, children:

      children })));


})).withConfig({ displayName: "Tag", componentId: "e9szmh-7" })([""]);exports.Tag = Tag;



var AvatarTag = (0, _styledComponents.default)( /*#__PURE__*/
(0, _react.forwardRef)(
function (_ref8, ref) {var children = _ref8.children,_ref8$size = _ref8.size,size = _ref8$size === void 0 ? 'medium' : _ref8$size,_ref8$rounded = _ref8.rounded,rounded = _ref8$rounded === void 0 ? true : _ref8$rounded,_ref8$variant = _ref8.variant,variant = _ref8$variant === void 0 ? TagVariant.Dark : _ref8$variant,avatar = _ref8.avatar,props = _objectWithoutProperties(_ref8, ["children", "size", "rounded", "variant", "avatar"]);
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(TagTemplate, _objectSpread(_objectSpread({},
    props), {}, {
      size: size,
      variant: variant,
      rounded: rounded,
      icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(AvatarIcon, _objectSpread({}, avatar)),
      ref: ref, children:

      children })));


})).withConfig({ displayName: "Tag__AvatarTag", componentId: "e9szmh-8" })([""]);exports.AvatarTag = AvatarTag;
