// @atlaskit/smart-card/dist/esm/state/hooks/useSmartLink.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import { useEffect, useState } from 'react';
import { useSmartCardState as useSmartLinkState } from '../store';
import { useSmartCardActions as useSmartLinkActions } from '../actions';
import { useSmartLinkConfig } from '../config';
import { useSmartLinkAnalytics } from '../analytics';
export function useSmartLink(id, url, dispatchAnalytics) {
  var state = useSmartLinkState(url);
  var analytics = useSmartLinkAnalytics(dispatchAnalytics);
  var actions = useSmartLinkActions(id, url, analytics);
  var config = useSmartLinkConfig(); // NB: used to propagate errors from hooks to error boundaries.

  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      setState = _useState2[1]; // Register the current card.


  var register = function register() {
    actions.register().catch(function (err) {
      return setState(function () {
        throw err;
      });
    });
  }; // AFP-2511 TODO: Fix automatic suppressions below
  // eslint-disable-next-line react-hooks/exhaustive-deps


  useEffect(register, [url]); // Provide the state and card actions to consumers.

  return {
    state: state,
    actions: actions,
    config: config,
    analytics: analytics
  };
}
