// @postman/aether/esmLib/components/Illustration/illustrations/AddPayment.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function AddPayment() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "add-payment-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "4.3",
    y1: "36.65",
    x2: "8.37",
    y2: "36.65",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "131.3",
    y1: "36.65",
    x2: "135.37",
    y2: "36.65",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "4.3",
    y1: "88.65",
    x2: "8.37",
    y2: "88.65",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "131.3",
    y1: "88.65",
    x2: "135.37",
    y2: "88.65",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-1",
    points: "84.06 29.83 138.73 29.83 138.73 95.23 0.73 95.23 0.73 29.83 56.56 29.83"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M88.47,37.41a15.39,15.39,0,0,1,1,5.15,19.29,19.29,0,1,1-6.7-13.69S85.75,30.66,88.47,37.41Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M55.45,56.15C45,65,32.45,70.75,29.1,76.73c-1.59,2.83-1.41,5-.66,6.59a5.15,5.15,0,0,0,6.33,2.54,40.58,40.58,0,0,0,5.29-2.28c5.55-2.82,12.16-6,13.67-8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M52.06,59.28c6.21,12.51,1,17.29,1,17.29l-.59,8.32"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M53.17,77.2a17.14,17.14,0,0,0,3.92-19.44"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M46.49,63l0-5.3a2.34,2.34,0,0,1,2.22-2.45l5.85-.36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M87.35,35c-2.46.22-4.15,0-8.5.52-7.62.92-10.63,4.1-11.79,6.16a7.15,7.15,0,0,0-.33,4c.71,3.84,3.76,12.77,18.24,10.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M89.55,43.51a30.18,30.18,0,0,0,5-5.08c.08-.1,3.27-4.31,3.4-4.51a6.16,6.16,0,0,1,8.78-1.73l0,0a6.14,6.14,0,0,1,2,7.59C101.91,54.28,86,65.91,86,65.91s1.41,14.21,1.3,19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M86.7,54.46s3.2,7.95-.69,11.45c0,0-.48-6.46-3.07-7.76"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "64",
    y1: "22.94",
    x2: "64.78",
    y2: "24.92"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "62.98",
    cy: "19.7",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "63.14",
    cy: "20.1",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M4.77,79.58V45.29s10.36.77,10.36-11.51H52.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M4.78,79.8s10.35-.49,10.35,11.78"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M134.89,80.08s-10.12-.77-10.12,11.5"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "97.82",
    y1: "33.78",
    x2: "86.77",
    y2: "33.78"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M134.88,80V45.29s-10.36.77-10.36-11.51H108.74"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "15.51 91.73 46.35 91.73 94.59 91.73 124.77 91.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M31,73.21a13.85,13.85,0,0,1-3.11-8.49c0-8.65,9.24-16.16,22.76-19.87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M102.39,50.55c6,3.86,9.52,8.79,9.52,14.17,0,12.35-18.81,22.36-42,22.36a67.29,67.29,0,0,1-22.06-3.39,31.84,31.84,0,0,1-4.26-1.52"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M48.62,79.54h0a2.26,2.26,0,0,0,2.26,2.25h1.54"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "20.21 37.9 21.22 38.46 20.21 39.02 19.65 40.05 19.1 39.02 18.08 38.46 19.1 37.9 19.65 36.87 20.21 37.9"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "25.21 37.9 26.22 38.46 25.21 39.02 24.65 40.05 24.1 39.02 23.08 38.46 24.1 37.9 24.65 36.87 25.21 37.9"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "30.21 37.9 31.22 38.46 30.21 39.02 29.65 40.05 29.1 39.02 28.08 38.46 29.1 37.9 29.65 36.87 30.21 37.9"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "17.42",
    y1: "42.33",
    x2: "31.4",
    y2: "42.33"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "17.42",
    y1: "45.33",
    x2: "31.4",
    y2: "45.33"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "6.34 35.46 8.37 34.62 7.54 36.66 8.37 38.69 6.34 37.85 4.3 38.69 5.14 36.66 4.3 34.62 6.34 35.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "133.34 35.46 135.37 34.62 134.54 36.66 135.37 38.69 133.34 37.85 131.3 38.69 132.14 36.66 131.3 34.62 133.34 35.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "6.34 87.46 8.37 86.62 7.54 88.66 8.37 90.69 6.34 89.85 4.3 90.69 5.14 88.66 4.3 86.62 6.34 87.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-9",
    points: "133.34 87.46 135.37 86.62 134.54 88.66 135.37 90.69 133.34 89.85 131.3 90.69 132.14 88.66 131.3 86.62 133.34 87.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "6.34 35.46 8.37 34.62 7.54 36.66 8.37 38.69 6.34 37.85 4.3 38.69 5.14 36.66 4.3 34.62 6.34 35.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "133.34 35.46 135.37 34.62 134.54 36.66 135.37 38.69 133.34 37.85 131.3 38.69 132.14 36.66 131.3 34.62 133.34 35.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "6.34 87.46 8.37 86.62 7.54 88.66 8.37 90.69 6.34 89.85 4.3 90.69 5.14 88.66 4.3 86.62 6.34 87.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "133.34 87.46 135.37 86.62 134.54 88.66 135.37 90.69 133.34 89.85 131.3 90.69 132.14 88.66 131.3 86.62 133.34 87.46"
  }));
}
