// @postman/aether/esmLib/components/Illustration/illustrations/AutoFlex.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function AutoFlex() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "auto-flex-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "296.32",
    y1: "63.64",
    x2: "299.33",
    y2: "63.64",
    gradientTransform: "matrix(-1, 0, 0, 1, 416.42, 0)",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "290.65",
    y1: "63.95",
    x2: "294.17",
    y2: "63.95",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "33.44",
    y1: "43.79",
    x2: "50.41",
    y2: "43.79",
    gradientTransform: "matrix(1, 0, 0, 1, 0, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip-path"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M69.23,41a5.44,5.44,0,0,1-2,.36,5.62,5.62,0,1,1,5.11-3.26"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip-path-2"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "25.8",
    cy: "76.21",
    r: "5.63"
  })), /*#__PURE__*/React.createElement("clipPath", {
    id: "clip-path-3"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "20.17",
    cy: "26.7",
    r: "5.63"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "64.61",
    y1: "71.1",
    x2: "68.19",
    y2: "71.1",
    gradientTransform: "matrix(1, 0, 0, 1, 0, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-5",
    x1: "41.73",
    y1: "26.7",
    x2: "45.3",
    y2: "26.7",
    gradientTransform: "matrix(1, 0, 0, 1, 0, 0)",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M139.17,113.37c0,2.92-12.76,5.28-28.5,5.28s-28.51-2.36-28.51-5.28c0-1.45,3.86-3,9-3.85,1-.18.56,4.43,2.61,6,4.34,3.33,6.79,1.78,9.93,0,1.48-.83,2-7.33,2.68-7.37a37,37,0,0,1,4.27-.06c1.46,0,1.41.1,2.82.06.65,0,.64,5.49,1.85,6.5.86.72,3.57,3.24,5.83,3.24,4.2,0,5.65-3.9,6.6-4.55.66-.45-.31-4.44.42-4.36C136.25,109.91,139.17,111.48,139.17,113.37Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M126.36,45.61a19.29,19.29,0,1,0-25.29,1l2.17,1.5"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M99.11,44.52S89.47,46.87,78.89,39a6.37,6.37,0,0,0-9.15,1.6h0a6.33,6.33,0,0,0,.82,8A35.32,35.32,0,0,0,94.05,58.76l-2.74,50.1c0,4.36,2.64,8.8,7,8.74h0c4.29,0,6.78-3,7.74-7.82l2.77-18.39A1.51,1.51,0,0,1,110.32,90h0a1.51,1.51,0,0,1,1.51,1.42l1.77,19.1a7.35,7.35,0,0,0,7.35,7.35h0a7.36,7.36,0,0,0,7.36-7.35l-.92-29.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M97.54,20.94a12.42,12.42,0,0,1,7.76,14.57c-.87,3.63-3.88,6.68-7,8.28"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-1",
    x: "102.03",
    y: "46.12",
    width: "31.68",
    height: "34.09",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M131.3,79.07c0-.3,2.34,8.45-3.87,7.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M93.61,65.86s3.86,1.94,8.42-3.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M93.4,70.56s5.44.58,8.7-3.58"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "115.69",
    cy: "33.53",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M110.59,46.08v7.4a5,5,0,0,0,5,5H133.1"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "118.59",
    cy: "63.64",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "124",
    cy: "63.95",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "118.59",
    cy: "63.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "123.95",
    cy: "63.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M133.2,62h-4.06a1.65,1.65,0,0,0-1.39,1.83h0c0,1,.62,1.84,1.39,1.84h4.06"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M133.2,69H111.14a1.65,1.65,0,0,0-1.39,1.83h0c0,1,.62,1.84,1.39,1.84H133.2"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-1",
    x1: "126.65",
    y1: "14.21",
    x2: "125.08",
    y2: "16.18"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "128.48",
    cy: "12.28",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(35.67 101.97) rotate(-49.48)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-1",
    cx: "128.52",
    cy: "12.05",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(35.86 101.91) rotate(-49.48)"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "41.92",
    cy: "43.79",
    r: "8.48"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "41.92",
    cy: "43.79",
    r: "8.48"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M45.59,27.1a17.11,17.11,0,1,1-4-.39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M24.65,22.93a27.09,27.09,0,0,1,39.09,4.81,11.9,11.9,0,0,1,1.2,2,6.4,6.4,0,0,1,.28.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M68.89,46.37A27.09,27.09,0,1,1,17.57,31.93"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M20.49,74.12A37.14,37.14,0,1,1,78.75,39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M65.23,72.71a37.21,37.21,0,0,1-34.35,6.55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M77.65,54a37,37,0,0,1-9.59,16.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M44.51,44.35a1.93,1.93,0,0,0-.67-.66,3.92,3.92,0,0,0-1.07-.41l-.2-.05V41l.12,0a1.38,1.38,0,0,1,.51.36,1.08,1.08,0,0,1,.24.53h0a.56.56,0,0,0,.63.49c.63,0,.55-.49.55-.49a1.94,1.94,0,0,0-.38-1,2.28,2.28,0,0,0-.91-.72,2.56,2.56,0,0,0-.77-.21v-.57a.47.47,0,0,0-.48-.48.48.48,0,0,0-.48.48v.55a3.05,3.05,0,0,0-1,.24,2.25,2.25,0,0,0-.94.77A2.05,2.05,0,0,0,39.52,43a1.89,1.89,0,0,0,.63.66,3,3,0,0,0,1,.41l.49.12v2.37l-.27,0a1.62,1.62,0,0,1-.5-.22,1.3,1.3,0,0,1-.35-.35,1.32,1.32,0,0,1-.18-.44s-.14-.52-.53-.53c-.72,0-.67.53-.67.53a2.27,2.27,0,0,0,.23.83,2.06,2.06,0,0,0,.55.67,2.57,2.57,0,0,0,.84.44,4,4,0,0,0,.88.14v.62a.48.48,0,0,0,.48.48.47.47,0,0,0,.48-.48v-.66a3.52,3.52,0,0,0,.85-.23,2.28,2.28,0,0,0,1-.79,2.17,2.17,0,0,0,.34-1.2A2,2,0,0,0,44.51,44.35ZM41.56,43a1.64,1.64,0,0,1-.76-.37.79.79,0,0,1-.27-.61,1,1,0,0,1,.39-.8,1.38,1.38,0,0,1,.69-.27V43Zm1.77,3.07a1.14,1.14,0,0,1-.52.41,1.68,1.68,0,0,1-.24.06V44.44a3,3,0,0,1,.43.17,1.11,1.11,0,0,1,.39.35.9.9,0,0,1,.12.48A1.07,1.07,0,0,1,43.33,46.05Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-10"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M63.67,40.18A4,4,0,0,1,71.2,40"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M69.23,41a5.44,5.44,0,0,1-2,.36,5.62,5.62,0,1,1,5.11-3.26"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "67.3",
    cy: "34.08",
    r: "1.84"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-11"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M22.22,80.66a4,4,0,0,1,7.53-.22"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "25.8",
    cy: "76.21",
    r: "5.63"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "25.85",
    cy: "74.56",
    r: "1.84"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-12"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-1",
    d: "M16.59,31.15a4,4,0,0,1,7.53-.22"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "20.17",
    cy: "26.7",
    r: "5.63"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "20.22",
    cy: "25.05",
    r: "1.84"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-13",
    cx: "66.4",
    cy: "71.1",
    r: "1.79"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "66.4",
    cy: "71.1",
    r: "1.79"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-14",
    cx: "43.51",
    cy: "26.7",
    r: "1.79"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-1",
    cx: "43.51",
    cy: "26.7",
    r: "1.79"
  }))));
}
