// @postman/aether/esmLib/components/Illustration/illustrations/BootcampLesson.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function BootcampLesson() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "bootcamp-lesson-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "11.24",
    y1: "70.86",
    x2: "16.17",
    y2: "70.86",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "48.43",
    y1: "15.66",
    x2: "57.31",
    y2: "15.66",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M60.1,115.58c0,1.34-11.63,2.43-26,2.43s-26-1.09-26-2.43c0-1,5.91-1.79,14.5-2.19,1.58-.07,1.73,2.81,3.47,2.77.4,0,3.41-.29,3.81-.3.84,0,2.61-2.68,3.47-2.69s-.2,0,.71,0c1.7,0,2.62,0,4.23,0,1.31,0,2.24,2.78,3.47,2.82s3.4,0,4.42,0c.51,0,2.15-2.54,2.64-2.51C56.4,113.94,60.1,114.7,60.1,115.58Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M33.68,56.85a18.92,18.92,0,1,0-24.55-.43l.84.69"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "9.4",
    cy: "21.55",
    r: "2.68"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "9.4",
    cy: "21.55",
    r: "2.68"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M39.37,48c-6.06-.72-10.24-5.84-10.24-12.09A11.34,11.34,0,0,1,33,27"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.92,71.7S37.5,79.81,30.55,79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M39.38,78.6a9.41,9.41,0,0,1-8.83,5.59"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "10.9",
    y1: "23.79",
    x2: "12.15",
    y2: "25.86"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "22.44",
    cy: "45.45",
    r: "2.83"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "0.41",
    y: "60.67",
    width: "29.73",
    height: "25.96",
    rx: "3.68"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M27,60.49s4.79-1.73,7.26-4.15"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "16.24",
    y1: "70.85",
    x2: "29.75",
    y2: "70.85"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "0.59",
    y1: "70.85",
    x2: "10.94",
    y2: "70.85"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "11.24",
    y: "68.33",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "11.24",
    y: "68.33",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M8.89,60.52h0A3.07,3.07,0,0,1,12,57.45h6.14a3.07,3.07,0,0,1,3.07,3.07"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M33.25,93.37H13c-3.58,0-6.48-.9-6.48-4.48v-1.8"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "7.18",
    y: "93.35",
    width: "33.6",
    height: "9.36",
    rx: "4.68"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M37.64,102.55v7.62a6.17,6.17,0,0,0,6.17,6.17h0A6.17,6.17,0,0,0,50,110.17V87.82c0-4.72-3.82-6.54-8.53-6.54l-1.89-.06-.38-9.94"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M34,102.82v7.35a6.17,6.17,0,0,1-6.17,6.17h0a6.17,6.17,0,0,1-6.17-6.17l.17-7.21"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "87.57",
    y1: "66.76",
    x2: "44.14",
    y2: "66.76"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "25.44 22.77 25.44 6.45 107.21 6.45 107.21 18.11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M34.73,56S45.15,55.83,46.49,82"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "48.43",
    y: "13.56",
    width: "8.88",
    height: "4.22"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-10",
    x: "48.43",
    y: "13.56",
    width: "8.88",
    height: "4.22"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "48.43",
    y: "13.56",
    width: "8.88",
    height: "4.22"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "61.43",
    y: "13.56",
    width: "22.12",
    height: "4.22"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "31.5",
    cy: "13.34",
    r: "1.31"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "31.5",
    cy: "19.34",
    r: "1.31"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-11",
    points: "134.26 16.29 135.26 16.85 134.26 17.41 133.72 18.42 133.17 17.41 132.17 16.85 133.17 16.29 133.72 15.27 134.26 16.29"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-11",
    points: "78.68 85.84 79.7 86.41 78.68 86.97 78.13 88.01 77.57 86.97 76.56 86.41 77.57 85.84 78.13 84.81 78.68 85.84"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-11",
    points: "5.91 8.17 6.92 8.73 5.91 9.3 5.35 10.33 4.8 9.3 3.78 8.73 4.8 8.17 5.35 7.13 5.91 8.17"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M77.34,114.33c0,1.48,14.2,2.68,31.72,2.68s31.72-1.2,31.72-2.68c0-.74-4.17-.94-9.87-1.76-.62-.09-1.78,3.42-3.93,3.3-1.45-.07-3.15.51-6.65-.4-1.64-.42-2.83-3.9-3.61-3.92-3-.09-4.29.09-7.66.09-1.63,0-2.16.11-3.72.08-.73,0-1.52,3.2-2.86,3.71A13.16,13.16,0,0,1,97,117c-2.1-.06-4.34-2.44-5.4-2.77-.73-.23-.4-2.05-1.21-2C81.37,112.65,77.34,113.37,77.34,114.33Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M104.67,54.78a19.5,19.5,0,0,0,15.24-18.67,19.28,19.28,0,0,0-38.56,0c0,8.5,6.15,15.78,13.8,18.33A17.09,17.09,0,0,0,104.67,54.78Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M105.53,92.61,105.36,111c-.39,3.88-3.66,5.91-7.57,5.91h0c-4,0-7.24-2.1-7.57-6-1-11.45-2.76-36.31-2.27-60.39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M118.58,63.5l5,16.92,8,27.75c1.07,4.67-2.49,8.19-7.28,8.19h-.55a7.47,7.47,0,0,1-7.05-5l-7.77-17.84-4.81-1.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M91.26,52.63c-.21.06-.23,9.38-3.19,14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M119.39,47.86a18.39,18.39,0,0,0-.72,15.69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M113.82,50c-.41,2-1.79,12.87,5.37,16.24"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M125.5,51.45l.28-4.38a2.33,2.33,0,0,0-2.33-2.33l-5.48-.1"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "111.86",
    cy: "36.82",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M83.3,28.34c1.89-.47,4.06-.7,8.46-.38,7.66.55,10.82,3.57,12.08,5.57a7.12,7.12,0,0,1,.52,4c-.52,3.86-1.7,13.33-18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-8",
    x1: "100.95",
    y1: "14.27",
    x2: "100.87",
    y2: "16.78"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-12",
    cx: "100.93",
    cy: "11.41",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-8",
    cx: "100.93",
    cy: "11.41",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M117.45,46.33s14.09,6.56,21.3,28.39A5.72,5.72,0,0,1,134.44,82h0a5.17,5.17,0,0,1-6.25-3.15c-1.54-4-4.91-11.42-9.39-15.29"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "105.01",
    y: "59.11",
    width: "4.93",
    height: "5.07",
    rx: "0.43",
    transform: "translate(214.95 123.28) rotate(-180)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M122.67,68.49v2.14a1.48,1.48,0,0,1-1.48,1.48h0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M86.67,48.56a27.69,27.69,0,0,1-11.88-9.65,6.43,6.43,0,0,0-9.15-1.46h0a6.17,6.17,0,0,0-1.87,7.68C66.18,50,72.18,59.33,82.3,63.85c.29.13.58.24.87.35a20.83,20.83,0,0,0,3,.83l1.66.36"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-13",
    x1: "72.77",
    y1: "51.99",
    x2: "58.26",
    y2: "21.37"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-13",
    cx: "68.18",
    cy: "41.99",
    rx: "7.01",
    ry: "7.12"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "38.26",
    y1: "7",
    x2: "38.26",
    y2: "33.14"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-14",
    x1: "43.37",
    y1: "51.4",
    x2: "61.63",
    y2: "51.4"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-14",
    x1: "43.37",
    y1: "56.4",
    x2: "61.63",
    y2: "56.4"
  }))));
}
