// @postman/aether/esmLib/components/Illustration/illustrations/BootcampWelcome.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function BootcampWelcome() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "bootcamp-welcome-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "-0.41",
    y1: "54.19",
    x2: "2.99",
    y2: "54.19",
    gradientTransform: "matrix(-1, 0, 0, 1, 139.07, 0)",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "121.41",
    y1: "9.55",
    x2: "139.92",
    y2: "9.55",
    gradientTransform: "translate(-112.56)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "54.48",
    y1: "40.08",
    x2: "65.37",
    y2: "40.08",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "137.78 57.22 139.48 56.71 138.11 51.16 136.08 51.69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M14,115c0,1.48,14.18,4,31.69,4s31.74-2.48,31.74-4c0-.74-4.16-.94-9.86-1.76-.63-.09-1.78,3.42-3.93,3.3-1.46-.07-3.16.51-6.65-.4-1.64-.42-2.83-3.9-3.62-3.92-3-.09-4.29.09-7.65.09-1.63,0-2.16.11-3.73.08-.72,0-1.51,3.2-2.85,3.71a15.87,15.87,0,0,1-5.48.73c-2.11-.06-3.39-1.65-4.46-2-.73-.23-.25-2-1.06-2C19.14,113.34,14,114.07,14,115Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M82.36,115.26c0,2.26,13,4.09,29.05,4.09s25-1.83,25-4.09c0-1.13-2.35-1.09-4.81-1.71-1-.26-.84,2.72-2.81,2.55-1.33-.12-1.58,2.12-4.92,1.15-7.27-2.13-5.45-4.85-6.17-4.88-2.79-.13-3.1.11-6.18.11-1.49,0-1.32,0-2.75,0-.66,0-1.51,2.31-2.74,3.09a11.8,11.8,0,0,1-5.69,1.25c-1.92-.09-2.46-1.86-3.44-2.36-.67-.35-1.57-1.89-2.31-1.83C86.41,113.31,82.36,113.79,82.36,115.26Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M27,31a20.33,20.33,0,0,0-1,5.15A19.35,19.35,0,1,0,27,31Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M62.84,82.52l4.91,25.9a7.39,7.39,0,0,1,0,3.07,7.57,7.57,0,0,1-7.17,5.77h0c-4.09.1-6.29-2.69-7.72-7.16L48,94l-5.09-3.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M60,49.69c5.9,4.07,11.62,11,15.8,17a6.93,6.93,0,0,1-3.45,10.76"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M60.46,70S56.6,59,63.09,52.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M60.41,70.47s-6.88-7.2-2-19.17"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M68.25,57,69,51.2a2.34,2.34,0,0,0-2.22-2.45l-5.85-.36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M62.68,29.06c-2.47,0-3.68-.13-8.06,0-7.67.17-11,3-12.34,5a7.2,7.2,0,0,0-.73,4c.33,3.89,3.3,13.18,18,12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M29.51,62.41s-1.28,12.3-1.17,17.06l-.69,30.1a7.39,7.39,0,0,0,7.55,7.54h0a7.39,7.39,0,0,0,7.16-6.4L44,91.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M32.55,51.69c-2.59,1.3-3,10.93-3,10.93C25.61,59.13,28.79,48,28.79,48"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "42.89",
    y1: "15.71",
    x2: "43.08",
    y2: "17.83"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "42.34",
    cy: "12.35",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "42.38",
    cy: "12.79",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "35.82",
    y: "60.66",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M28.92,47.3s-6.69,6.38-20.79-1.21a4.13,4.13,0,0,0-3.56-.22,5.32,5.32,0,0,0-3.15,6.42,5.62,5.62,0,0,0,1.25,2.24,36.9,36.9,0,0,0,6.21,5.29c5.26,2.82,9.9,4.7,20.43,3.44"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "123.14",
    cy: "12.85",
    r: "2.79"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "123.14",
    cy: "12.85",
    r: "2.79"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "121.58",
    y1: "15.18",
    x2: "120.28",
    y2: "17.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M93.89,43.49c6.28-1,10.47-7,10.33-13.52-.08-4.13-2.62-7.53-6.39-9.54"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "115.56",
    cy: "33.7",
    r: "2.95"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "33.76",
    cy: "39.18",
    r: "2.95"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M110.12,91l1.88.86a10.29,10.29,0,0,1,6,9.41l-.07,9a6.7,6.7,0,0,0,6.7,6.75h.84a6.68,6.68,0,0,0,6.68-6.53l.18-9.3c.12-3.24.15-9.54.06-12.78-.28-11.12-1.09-23.87-4.28-33.78l-1.71-5.53-.74-2.37"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M132.1,75.17l3.43-.34a4.26,4.26,0,0,0,3.84-4.31c-.07-8.23-3.09-17.76-5.17-23.3a3.33,3.33,0,0,0-4.37-1.94L126,46.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M103,52.18c-1.83,1.47-2.45,10.74-2.45,10.74L99.09,76.59s-2.17,21.15-4.62,31.1a7.43,7.43,0,0,0,7.21,9.21h0a7.44,7.44,0,0,0,7.36-6.38l3-18.64"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "136.41",
    y1: "54.55",
    x2: "128.79",
    y2: "56.51"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "137.78 57.22 139.48 56.71 138.11 51.16 136.08 51.69"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "67.31",
    y1: "82.75",
    x2: "75.71",
    y2: "116.11"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "65.37",
    cy: "75.61",
    rx: "7.01",
    ry: "7.12"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "133.56 17.97 132.55 18.54 133.56 19.1 134.12 20.14 134.67 19.1 135.69 18.54 134.67 17.97 134.12 16.94 133.56 17.97"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "82.39 52.84 81.38 53.4 82.39 53.97 82.95 55 83.5 53.97 84.52 53.4 83.5 52.84 82.95 51.8 82.39 52.84"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "7.96 29.65 6.94 30.21 7.96 30.78 8.51 31.81 9.07 30.78 10.08 30.21 9.07 29.65 8.51 28.62 7.96 29.65"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M9.11,14.38H27.33s.77-6.67-4.75-5.54c0,0,1-4.59-5-4.07a2.24,2.24,0,0,0-1.4.63,2.7,2.7,0,0,0-1.07,2s-3.74-.66-3.44,3C11.71,10.37,7.84,10.41,9.11,14.38Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M8.81,14.38H27s.77-6.67-4.75-5.54c0,0,.95-4.59-5-4.07a2.24,2.24,0,0,0-1.4.63,2.7,2.7,0,0,0-1.07,2s-3.74-.66-3.44,3C11.41,10.37,7.54,10.41,8.81,14.38Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M84.44,42.92H73.73s-.46-3.93,2.79-3.26c0,0-.56-2.71,2.92-2.4a1.32,1.32,0,0,1,.82.37,1.57,1.57,0,0,1,.63,1.16s2.2-.38,2,1.77A1.61,1.61,0,0,1,84.44,42.92Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M84.75,42.92H74s-.45-3.93,2.79-3.26c0,0-.56-2.71,2.92-2.4a1.3,1.3,0,0,1,.82.37,1.57,1.57,0,0,1,.63,1.16s2.2-.38,2,1.77A1.62,1.62,0,0,1,84.75,42.92Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M65.29,68.22l-.94-3.72a1.49,1.49,0,0,0-1.86-1h0a1.49,1.49,0,0,0-1,1.86l.89,2.94"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M118.93,62.27l1.9,19.85a5.68,5.68,0,0,0,5.84,5.12h0a5.67,5.67,0,0,0,5.46-5.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M92.57,28.14a20.7,20.7,0,0,0-1,5.14,19.36,19.36,0,1,0,1-5.14Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "98.73",
    y1: "56.37",
    x2: "100.03",
    y2: "67.92"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M121.92,50.07s-4.43,2.42-3,12.2"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M116.68,52.73s-1.29,8,2.41,12.94"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "90.62",
    cy: "67.69",
    rx: "6.23",
    ry: "6.32"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M95,63s3.64-8.08,7.74-11.27"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M90.62,74s2,6.92,7.85,7.33"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "90.43",
    y1: "74.07",
    x2: "89.7",
    y2: "115.3"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M92.39,61.23l.21-2.93a1.51,1.51,0,0,0-1.39-1.6h0a1.5,1.5,0,0,0-1.6,1.39l-.22,3.06"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M.58,77.94s20,1.23,27.1,17.2"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "132.39",
    y1: "102.74",
    x2: "139.69",
    y2: "102.74"
  }))));
}
