// @postman/aether/esmLib/components/Illustration/illustrations/CallFailed.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function CallFailed() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "call-failed-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "100.87",
    y1: "49.62",
    x2: "115.41",
    y2: "49.62",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M111.84,37.76h-7.4a3.57,3.57,0,0,0-3.57,3.57V57.9a3.57,3.57,0,0,0,3.57,3.57h7.4a3.57,3.57,0,0,0,3.57-3.57V41.33A3.57,3.57,0,0,0,111.84,37.76ZM107.1,54.63h-3.44V52.46h3.44Zm5.7,0h-3.44V52.46h3.44Zm.67-3.86H102.66V41.06h10.81Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "100.87",
    y: "37.76",
    width: "14.54",
    height: "23.71",
    rx: "3.57"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "110.42",
    y: "34.57",
    width: "3.08",
    height: "3.08"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "111.94",
    y1: "34.07",
    x2: "111.94",
    y2: "28.74"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "102.66",
    y: "41.06",
    width: "10.81",
    height: "9.72"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "103.66",
    y: "52.46",
    width: "3.44",
    height: "2.17"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "109.36",
    y: "52.46",
    width: "3.44",
    height: "2.17"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "105.45",
    y1: "48.05",
    x2: "110.75",
    y2: "48.05"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "104.68",
    y1: "43.4",
    x2: "106.28",
    y2: "45"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "104.68",
    y1: "45",
    x2: "106.28",
    y2: "43.4"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "110.16",
    y1: "43.4",
    x2: "111.77",
    y2: "45"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "110.16",
    y1: "45",
    x2: "111.77",
    y2: "43.4"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "111.92",
    cy: "26.48",
    rx: "1.66",
    ry: "1.67"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "111.92",
    cy: "26.48",
    rx: "1.66",
    ry: "1.67"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M95.63,115.22c0,1.48-14.2,2.68-31.72,2.68s-31.72-1.2-31.72-2.68c0-.74,4.17-.94,9.87-1.76.62-.09,1.78,3.42,3.93,3.31,1.45-.08,3.15.5,6.64-.41,1.65-.42,2.84-3.9,3.62-3.92,3-.09,4.29.09,7.66.09,1.63,0,2.16.11,3.72.08.73,0,1.52,3.2,2.86,3.71A13.16,13.16,0,0,0,76,117.84c2.1-.06,4.34-2.44,5.4-2.77.73-.23.4-2,1.21-2C91.6,113.54,95.63,114.26,95.63,115.22Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M68.29,55.67C59.9,53.59,53.06,46,53.06,37a19.28,19.28,0,0,1,38.56,0c0,8.5-6.15,15.78-13.81,18.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M67.61,92.69v19.16c.39,3.88,3.66,5.91,7.56,5.91h0c4,0,7.25-2.1,7.58-6,1-11.45,2.64-35.43,2.16-59.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M49.88,78.88l-.49,2.43-8,27.75c-1.07,4.67,2.49,8.19,7.28,8.19h.55a7.45,7.45,0,0,0,7-5L64,94.43l4.81-1.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M81.71,53.52c.2.06.69,10.47,3.65,15.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M100.23,53.13c-10,2.69-16-1.28-16-1.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M85.5,68.34c11.63,1.39,16.69-1.3,22.47-4.6l2.66-1.87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M51.21,66.87,68.49,55.76c3-2.21,7.19-2.85,9.45.06h0a9.34,9.34,0,0,1,1,3.71c.14,2.7-.9,4.68-3.4,6.43l-15,10.83A11.93,11.93,0,0,1,55,79.07,10.07,10.07,0,0,1,46.44,76c-4.51-5.12-5.62-13.47,9.77-28.75"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M54.06,49.25S57.62,57.08,55,64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M59.57,51c.41,2,1.24,7.6.36,10"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M47.47,57.63,47.19,48a2.33,2.33,0,0,1,2.33-2.33l5.85-.1"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "61.11",
    cy: "37.71",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M89.67,29.23c-1.89-.47-4.06-.7-8.46-.38-7.66.55-10.82,3.57-12.08,5.57a7.13,7.13,0,0,0-.52,4c.52,3.87,1.7,13.33,18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "72.02",
    y1: "15.16",
    x2: "72.1",
    y2: "17.67"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "72.03",
    cy: "12.3",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "72.03",
    cy: "12.3",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M78.76,61.9h1.53a.43.43,0,0,1,.43.43v4.21a.43.43,0,0,1-.43.43H76.23a.44.44,0,0,1-.44-.43"
  }))));
}
