// @postman/aether/esmLib/components/Illustration/illustrations/CheckInternetConnection.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function CheckInternetConnection() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "check-internet-connection-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "21.31",
    y1: "20.43",
    x2: "39.82",
    y2: "20.43",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "100.52",
    y1: "36.13",
    x2: "111.41",
    y2: "36.13",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "42.43",
    cy: "115.83",
    rx: "14.12",
    ry: "1.56"
  }), /*#__PURE__*/React.createElement("g", {
    id: "HSvDYf"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M115.11,82.38l.11,1.2,1.22.2c0,1.76,0,3.54,0,5.33,0,.16-.5.42-.78.43-1.57,0-3.14,0-4.71,0-1.25,0-1.27,0-1,1.4h3.84v1.33h-5.21V95H111v2.5h-1c0-.82-.13-1.67-1.48-1.15v4.5l-1.36.11v1.84l-1.33.17-.11,1.21-1.24.14v3.41l1.23.18.07,1.1H103.2v-2l-1.24-.16-.16-1.2h-1.33c.12,1-.22,1.57-1.34,1.25L99,108.19l-1.23.18v4l1.23.15v1.13h-2.5v-8l-1.25-.14-.16-1.22-1.2-.16-.15-1.2-1.24-.16-.14-1.18-1.22-.17-.09-1.14-1.16-.13V92.45H91v2.46l1.3.23.15,1.14,1.22.13.15,1.21h2.57l.22-1.19,1.11-.17.15-1.07c.7-.45,2.38.55,2-1.42l1.89-.11.18-1.21,1.23-.17V83.81l1.24-.22.18-1.21Z"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "68.86",
    cy: "32.89",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M81.43,48.14c.48,24.13.66,35.24-.31,46.72-.32,3.86-4.41,6-8.26,6-.08,0,.76,0,.68,0l-2.15.13,0-.06c-3.22-.17-7.95-6.22-10.12-8.59"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "65.33",
    y1: "89.93",
    x2: "67.22",
    y2: "94.53"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M45.91,77.3,42.52,106c-.33,5.32,1.37,7.76,6.36,8.22h.54c3.92.43,6.12-1,7.4-3.37a9.74,9.74,0,0,0,1-2.66l3.69-16.77,3.89-1.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M50.08,66.09s5.6-5.63-1.42-18.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M52.8,66.61s6.57-7.93,2.29-20.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M44.8,53l-.25-7.16a2.33,2.33,0,0,1,2.34-2.33H52"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M48.66,65.57l13.51,3.11a5.86,5.86,0,0,1,4.94,6.78h0a5.88,5.88,0,0,1-5.87,4.88C32.14,80,36.72,60.68,52.13,43.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M81.76,59.65a16.64,16.64,0,0,0,6.14,2.71,5.73,5.73,0,0,1,3.54,8.42h0a5.7,5.7,0,0,1-5.77,2.82,26.38,26.38,0,0,1-3.73-1"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "94.23",
    cy: "115.2",
    rx: "6.52",
    ry: "1.37"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "14.28",
    y1: "114.58",
    x2: "17.51",
    y2: "114.58"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "23.89",
    y1: "114.58",
    x2: "46.25",
    y2: "114.58"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "49.44",
    y1: "114.58",
    x2: "122.9",
    y2: "114.58"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "126.09",
    y1: "114.58",
    x2: "129.32",
    y2: "114.58"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M21.56,25.26H39.78s.77-6.68-4.75-5.54c0,0,1-4.6-5-4.08a2.29,2.29,0,0,0-1.4.63,2.72,2.72,0,0,0-1.07,2s-3.73-.65-3.44,3C24.16,21.25,20.29,21.28,21.56,25.26Z"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M100.67,39h10.72s.45-3.93-2.8-3.26c0,0,.56-2.7-2.91-2.4a1.41,1.41,0,0,0-.83.37,1.61,1.61,0,0,0-.63,1.16s-2.2-.38-2,1.77A1.61,1.61,0,0,0,100.67,39Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M21.26,25.26H39.48s.77-6.68-4.75-5.54c0,0,1-4.6-5-4.08a2.29,2.29,0,0,0-1.4.63,2.72,2.72,0,0,0-1.07,2s-3.74-.65-3.44,3C23.86,21.25,20,21.28,21.26,25.26Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M100.37,39h10.72s.45-3.93-2.8-3.26c0,0,.56-2.7-2.92-2.4a1.41,1.41,0,0,0-.82.37,1.61,1.61,0,0,0-.63,1.16s-2.2-.38-2,1.77A1.61,1.61,0,0,0,100.37,39Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "37.93",
    y1: "80.27",
    x2: "21.52",
    y2: "80.27"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "37.93",
    y1: "73.81",
    x2: "27.98",
    y2: "73.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M86.77,25.4c-1.89-.46-4.49-.4-8.85-.09-7.66.55-10.82,3.57-12.08,5.57a7.12,7.12,0,0,0-.52,4c.52,3.86,3.12,12.94,17.71,11"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "56.86",
    cy: "34.31",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "68.72",
    y1: "10.72",
    x2: "68.8",
    y2: "13.23"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-9",
    cx: "68.85",
    cy: "8.06",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "68.73",
    cy: "7.86",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-10",
    x: "106.01",
    y: "84.26",
    width: "1.55",
    height: "1.59"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "72.27",
    y: "56.93",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }))));
}
