// @postman/aether/esmLib/components/Illustration/illustrations/ChooseWorkspace.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function ChooseWorkspace() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "choose-workspace-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "36.6",
    y1: "36.44",
    x2: "56.05",
    y2: "36.44",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M56.29,114.71c0,2.15,13.54,3.89,30.24,3.89s30.25-1.74,30.25-3.89c0-1.14-1.55-1.83-8.11-2.81-1-.15-10.73-1.43-11-.34s1.41.08.19,0c-1.38-.09-.57,2.46-2.46,2.65-.77.07-2.32,1.54-3,0a5.83,5.83,0,0,0-3.21-2.73c-1.63,0-.89-.66-2.6-.66-2.44,0,.33.1-1.94.17-.86,0-4,2.18-5.37,3.68A12.67,12.67,0,0,1,73,117.73c-1.73.1-3.54-1.44-4.61-1.86-.55-.22-1.6-4.69-2.16-4.63C59.49,112,56.29,113.49,56.29,114.71Z"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "81.77",
    cy: "38.17",
    rx: "19.17",
    ry: "19.54",
    transform: "translate(-1.09 2.43) rotate(-1.69)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M74.16,56.15s1.19,3.3-8.88,20.53c0,0-7.24,16.54,6.22,24.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.86,86.33l3.78,5.33c-1.38,3.16-6.53,8.35-11.2,12.23A7.52,7.52,0,0,0,67.09,114h0A7.53,7.53,0,0,0,77,116.22c31-18,29.49-52.89,29.49-52.89l17.58,7.45a6.43,6.43,0,0,0,8.44-3.42h0a6.43,6.43,0,0,0-2.14-7.69C124.28,55.2,112.13,47.32,99.66,45"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M56.08,39.71a51.53,51.53,0,0,0,7,3.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M48.27,51.22a30.12,30.12,0,0,1-4.74-4.17"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71.87,63a57.48,57.48,0,0,1-15.7-6.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71,54.48a15.9,15.9,0,0,0,.9,8.23"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M106.5,63.33a23.61,23.61,0,0,1-4.79-17.67"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M96.59,51s1.09,11.61,9.73,13.74"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M64.82,29c2.05-.65,4.12-1,9.47-.7,8.46.49,12,3.78,13.41,6a7.89,7.89,0,0,1,.64,4.42c-.52,4.27-2.2,14.84-20.23,12.53"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "91.82",
    y1: "18.03",
    x2: "90.87",
    y2: "20.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M100.94,37.6l8,1.22a6.42,6.42,0,0,1,5.38,7.38l-.51,3.15"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M110.44,65l-2.17,10a2.69,2.69,0,0,1-3.37,2h0"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "95.82",
    cy: "40.48",
    r: "2.39"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "93.23",
    cy: "14.69",
    r: "2.9"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "93.17",
    cy: "15.15",
    r: "2.9"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "12.26",
    y: "51.36",
    width: "19.45",
    height: "20.01",
    rx: "3.96"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "36.6",
    y: "51.36",
    width: "19.45",
    height: "20.01",
    rx: "3.96"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "36.6",
    y: "26.44",
    width: "19.45",
    height: "20.01",
    rx: "3.96"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "36.6",
    y: "26.44",
    width: "19.45",
    height: "20.01",
    rx: "3.96"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "12.26",
    y: "27.02",
    width: "19.45",
    height: "20.01",
    rx: "3.96"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M86.65,110.22l5.44,3.72a3.3,3.3,0,0,0,4.34-.53c1.85-2.07,3.42-6.23-.55-14"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "88.96",
    y: "59.77",
    width: "5.23",
    height: "5.38",
    rx: "0.46"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "42.3 86.36 41.28 86.92 42.3 87.49 42.85 88.52 43.41 87.49 44.43 86.92 43.41 86.36 42.85 85.33 42.3 86.36"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "62.3 16.36 61.28 16.92 62.3 17.49 62.85 18.52 63.41 17.49 64.43 16.92 63.41 16.36 62.85 15.32 62.3 16.36"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "112.3 26.36 111.28 26.92 112.3 27.49 112.86 28.52 113.41 27.49 114.43 26.92 113.41 26.36 112.86 25.32 112.3 26.36"
  }))));
}
