// @postman/aether/esmLib/components/Illustration/illustrations/CommentOnDiff.js
import React from 'react';
export default function CommentOnDiff() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "comment-on-diff-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "613.15",
    y1: "63.64",
    x2: "616.17",
    y2: "63.64",
    gradientTransform: "matrix(-1, 0, 0, 1, 719.25, 0)",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "607.49",
    y1: "63.95",
    x2: "611.01",
    y2: "63.95",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "50.62",
    y1: "30.74",
    x2: "66.82",
    y2: "30.74",
    gradientTransform: "matrix(1, 0, 0, 1, 0, 0)",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "58.71 50.72 36.14 50.72 36.14 48.21 56.24 48.21"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M125.17,113.37c0,2.92-12.76,5.28-28.5,5.28s-28.51-2.36-28.51-5.28c0-1.45,3.86-3,9-3.85,1-.18.56,4.43,2.61,6,4.34,3.33,6.79,1.78,9.93,0,1.48-.83,2-7.33,2.68-7.37a37,37,0,0,1,4.27-.06c1.46,0,1.41.1,2.82.06.65,0,.64,5.49,1.85,6.5.86.72,3.57,3.24,5.83,3.24,4.2,0,5.65-3.9,6.6-4.55.66-.45-.31-4.44.42-4.36C122.25,109.91,125.17,111.48,125.17,113.37Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M112.36,45.61a19.29,19.29,0,1,0-25.29,1l2.17,1.5"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M85.11,44.52S75.47,46.87,64.89,39a6.37,6.37,0,0,0-9.15,1.6h0a6.33,6.33,0,0,0,.82,8A35.32,35.32,0,0,0,80.05,58.76l-2.74,50.1c0,4.36,2.64,8.8,7,8.74h0c4.29,0,6.78-3,7.74-7.82l2.77-18.39A1.51,1.51,0,0,1,96.32,90h0a1.51,1.51,0,0,1,1.51,1.42l1.77,19.1a7.35,7.35,0,0,0,7.35,7.35h0a7.36,7.36,0,0,0,7.36-7.35l-.92-29.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M83.54,20.94A12.42,12.42,0,0,1,91.3,35.51c-.87,3.63-3.88,6.68-7,8.28"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "88.03",
    y: "46.12",
    width: "31.68",
    height: "34.09",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M117.3,79.07c0-.3,2.34,8.45-3.87,7.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M79.61,65.86S83.47,67.8,88,62.14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M79.4,70.56s5.44.58,8.7-3.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M81.41,44.79S88,47.6,88.12,52.31"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "101.69",
    cy: "33.53",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M96.59,46.08v7.4a5,5,0,0,0,5,5H119.1"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "104.59",
    cy: "63.64",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "110",
    cy: "63.95",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "104.59",
    cy: "63.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "109.95",
    cy: "63.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M119.2,62h-4.06a1.65,1.65,0,0,0-1.39,1.83h0c0,1,.62,1.84,1.39,1.84h4.06"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M119.2,69H97.14a1.65,1.65,0,0,0-1.39,1.83h0c0,1,.62,1.84,1.39,1.84H119.2"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "112.65",
    y1: "14.21",
    x2: "111.08",
    y2: "16.18"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "114.48",
    cy: "12.28",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(30.77 91.32) rotate(-49.48)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "114.52",
    cy: "12.05",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(30.95 91.27) rotate(-49.48)"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "50.21 33.76 34.14 33.72 34.14 31.21 50.62 31.26"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "16.83",
    y: "31.21",
    width: "14.39",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "34.14",
    y: "71.21",
    width: "27.08",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "16.83",
    y: "71.21",
    width: "14.39",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "54.95 41.72 30.14 41.72 30.14 39.21 56.66 39.21"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "16.83",
    y: "39.21",
    width: "10.39",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "36.14",
    y: "55.21",
    width: "25.08",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M61.13,65.72h-25V63.21h25S61.13,65.61,61.13,65.72Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "16.83",
    y: "63.21",
    width: "16.39",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "66.82 22.66 50.62 22.66 50.62 34.95 54.15 34.95 56.89 38.82 59.71 34.95 66.82 34.95 66.82 22.66"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "66.82 22.66 50.62 22.66 50.62 34.95 54.15 34.95 56.89 38.82 59.71 34.95 66.82 34.95 66.82 22.66"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "54.15",
    y1: "26.98",
    x2: "63.51",
    y2: "26.98"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "54.15",
    y1: "29.98",
    x2: "63.51",
    y2: "29.98"
  }))));
}
