// @postman/aether/esmLib/components/Illustration/illustrations/ConnectToServer.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function ConnectToServer() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "connect-to-server-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "122.24",
    y1: "43.3",
    x2: "125.61",
    y2: "43.3",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "117.02",
    y1: "43.25",
    x2: "120.39",
    y2: "43.25",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "112.02",
    y1: "43.25",
    x2: "115.39",
    y2: "43.25",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "129.4 50.07 122.75 50.07 90.87 50.07 90.87 36.38 129.42 36.38 129.42 50.08"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "123.92",
    cy: "43.3",
    r: "1.69"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "118.7",
    cy: "43.25",
    r: "1.69"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "94.18",
    y: "41.68",
    width: "14.06",
    height: "3.33",
    rx: "1.52"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "123.92",
    cy: "43.3",
    r: "1.69"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "118.7",
    cy: "43.25",
    r: "1.69"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "113.7",
    cy: "43.25",
    r: "1.69"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "113.7",
    cy: "43.25",
    r: "1.69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M36.36,44.94S41,45,41.89,58.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M41.18,43.78s4.21,3.46.71,14.65"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M57.88,89.06l.14,12a7.19,7.19,0,0,0,7.19,7.19h0a7.19,7.19,0,0,0,7.19-7.19V82.62a7,7,0,0,0-7-7h0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M72.21,105.3h0a7.19,7.19,0,0,0,7.19-7.19V79.62a7,7,0,0,0-7-7h0l-2.62-.2"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M36.53,39.89H32.4a4.9,4.9,0,0,0-4.9,4.9V48.1c0,1,.11,3.42.15,9.95"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "52.06",
    cy: "27.85",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "62.62 44.71 65.26 55.22 69.63 72.31"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M65.7,42.15s2.39,7.42.17,13.44"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "56.67",
    y: "49.9",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M65.87,55.42c6.94.13,13.4-2.35,21.54-6.37a5.84,5.84,0,0,0,3-6.81h0a6.79,6.79,0,0,0-8.78-4.55c-10,3.55-15,2.87-15,2.87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M70,21.73a32.29,32.29,0,0,0-9.84-.79c-7.66.56-10.81,3.58-12.08,5.58a7.26,7.26,0,0,0-.52,4c.52,3.87,1.7,13.34,18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "87.27",
    y1: "34.74",
    x2: "82.43",
    y2: "29.74"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "90.03",
    y1: "32.83",
    x2: "88.05",
    y2: "26.16"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "87.93",
    y1: "52.37",
    x2: "84.33",
    y2: "57.17"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "95.86 30.06 96.87 30.63 95.86 31.19 95.3 32.23 94.75 31.19 93.73 30.63 94.75 30.06 95.3 29.03 95.86 30.06"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "96.43 55.15 97.45 55.72 96.43 56.28 95.88 57.31 95.32 56.28 94.31 55.72 95.32 55.15 95.88 54.12 96.43 55.15"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "44.48",
    y1: "7.47",
    x2: "45.14",
    y2: "9.9"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-7"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-8",
    cx: "44",
    cy: "4.85",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(0.06 10.21) rotate(-13.25)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "43.84",
    cy: "4.68",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(0.09 10.17) rotate(-13.25)"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "39.18",
    cy: "29.31",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M38.78,41.83S28.64,50,22.68,73.26a6.27,6.27,0,0,0,5.82,7.82h0a6.26,6.26,0,0,0,6.26-4.45L41.9,58.44"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M22.75,77.79c-2.81,3.17-3.9,6.79-4.21,11.27-5.5.54-9.8,3.18-9.8,8.83H57.86"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M58,90.44c0-10.91-9.83-19.6-20.74-19.6"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-9",
    cx: "69.14",
    cy: "115.58",
    rx: "25.96",
    ry: "2.44"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-10",
    d: "M68.68,66.94a18.24,18.24,0,0,1,7.5-1.58c7.48,0,13.73,4.36,15.45,10.23a11.69,11.69,0,0,1,8.71,11.3H79.66"
  }))));
}
