// @postman/aether/esmLib/components/Illustration/illustrations/CreateFlow.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function CreateFlow() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "create-flow-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "82.36",
    y1: "23",
    x2: "99.51",
    y2: "23",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "61.03",
    y1: "65.04",
    x2: "64.04",
    y2: "65.04",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "55.37",
    y1: "65.34",
    x2: "58.88",
    y2: "65.34",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "30.02",
    y1: "25.74",
    x2: "46.64",
    y2: "25.74",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-5",
    x1: "15.59",
    y1: "34.78",
    x2: "22.52",
    y2: "34.78",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-6",
    x1: "115.36",
    y1: "24.49",
    x2: "120.31",
    y2: "24.49",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M100.82,113c-2.42,3.52-15.45,5.7-31.6,5.7S42.1,117,40.19,114c-.79-1.23,0-3.64,7.34-4.81.84-.13.17,5,1.92,6,5.28,2.89,7.74,1.59,11-.19,1.51-.84,1.74-7.11,2.46-7.16,2.81-.16,6.59-.09,9.7-.09,1.5,0,.79.09,2.23,0,.67,0,.75,5.65,2,6.65.89.72,4.34,3,6.66,2.87,1.93-.11,4.23-3.42,5.22-4.07.67-.45.68-4.7,1.43-4.62C98.37,109.54,101.89,111.45,100.82,113Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M98.22,21.71a1.3,1.3,0,0,0-1.19.78H82.36v1H97a1.29,1.29,0,1,0,1.2-1.78Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M53.48,46.89l-.86-.7a19.29,19.29,0,1,1,23.53,1.55"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "47.69",
    y: "47.11",
    width: "31.68",
    height: "34.12",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "52.9",
    cy: "10.67",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "52.9",
    cy: "10.67",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M83.43,37.64C77.25,36.91,73,31.7,73,25.32a11.55,11.55,0,0,1,4-9.08"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.05,60.74S84.92,67.09,79.7,69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.37,68.32S85,73.38,79.59,74.56"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "54.43",
    y1: "12.95",
    x2: "55.7",
    y2: "15.06"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M70.54,47.47v7.41a5,5,0,0,1-5,5H48"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "62.53",
    cy: "65.04",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "57.13",
    cy: "65.34",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "62.54",
    cy: "65.13",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "57.18",
    cy: "65.13",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M47.93,63.43H52c.77,0,1.39.82,1.39,1.84h0c0,1-.62,1.84-1.39,1.84H47.93"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M47.93,70.43H70c.77,0,1.39.82,1.39,1.84h0c0,1-.62,1.84-1.39,1.84H47.93"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "66.18",
    cy: "35.01",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M47.65,58.62a183.19,183.19,0,0,1-7.73,18.52,6.12,6.12,0,0,0,2.28,8h0a6.12,6.12,0,0,0,8.15-1.67L52,81.23"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M52.85,81.7l-5,28c-.84,5,3.29,7.36,7.35,7.36h0c5.19-.23,6.53-4.16,7.35-7.39l4.06-17.29a2.91,2.91,0,0,1,4.63-1.69,2.42,2.42,0,0,1,.32.32h0a3.12,3.12,0,0,1,.75,1.71L75.11,109c.71,6,3.45,7.78,7.74,7.83h0c4.36,0,7.46-2.5,7-8.75L87.11,60.65a252.29,252.29,0,0,0,18.56-20.58,6.86,6.86,0,0,0-.84-9.5h0a6.89,6.89,0,0,0-9.74.75A87.61,87.61,0,0,1,78.68,45.74"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M102.38,28.87V18.81a1.5,1.5,0,0,1,1.5-1.5h25.94a1.5,1.5,0,0,1,1.5,1.5V30.7a1.5,1.5,0,0,1-1.5,1.5H106.61"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M45.68,32.2H26.88a1.5,1.5,0,0,1-1.5-1.5V18.81a1.5,1.5,0,0,1,1.5-1.5H51.82a1.53,1.53,0,0,1,.49.08"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-10",
    points: "46.41 26.95 30.02 26.95 30.02 24.53 46.64 24.53"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "46.41 26.95 30.02 26.95 30.02 24.53 46.64 24.53"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "30.22",
    y1: "21.5",
    x2: "36.41",
    y2: "21.5"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "8.92",
    y: "47.37",
    width: "28.94",
    height: "14.89",
    rx: "1.5"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "26.02 53.04 27.64 54.66 25.94 56.36"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "19.99 56.36 18.37 54.74 20.07 53.04"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "23.74",
    y1: "52.66",
    x2: "22.09",
    y2: "56.68"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-11",
    d: "M21.23,24.8A1.3,1.3,0,0,0,19.94,26c-2.21.6-3.58,2.7-3.58,5.73V42.29a1.3,1.3,0,1,0,1,0V31.77c0-1,.24-4.17,2.92-4.81a1.27,1.27,0,0,0,.95.43,1.3,1.3,0,0,0,0-2.59Z"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-12",
    points: "115.36 21.52 115.36 27.45 120.31 24.44 115.36 21.52"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-4",
    points: "115.36 21.52 115.36 27.45 120.31 24.44 115.36 21.52"
  }))));
}
