// @postman/aether/esmLib/components/Illustration/illustrations/CriticalDanger.js
import React from 'react';
export default function CriticalDanger() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "critical-danger-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "21.98",
    y1: "103.4",
    x2: "44.07",
    y2: "103.4",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "100.11",
    y1: "107.97",
    x2: "117.32",
    y2: "107.97",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "91.47",
    y1: "94",
    x2: "91.63",
    y2: "94",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "79.04",
    y1: "104.09",
    x2: "96.35",
    y2: "104.09",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M44.07,114.44v-16a19.37,19.37,0,0,0-4.48.43c-1.86.68-2.63,4.62-2.94,7.12.88-13.73-7.74-14.49-7.74-14.49,3.59,5,4.23,17.5-.28,18.7s-1.75-8.85-1.75-8.85c-9.4,7.19-2.39,14-2.39,14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M100.68,114.35l16.64.07c-6.75-4.68-3.11-9.09-3.11-9.09-4,.32-8.16,6.74-8.16,6.74,1.06-7.81-2.62-10.06-5.94-10.55Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M91.47,94h.16A.15.15,0,0,0,91.47,94Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M96.35,101.55c-.6.09-1,.18-1,.18,1.17-7.2-2.88-7.7-3.75-7.72,1.15.5,3.26,11.41.05,11.21C88.31,105,89.62,99,89.62,99s-4.57,1.75-3.16,8.06-3.59,1.63-3.59,1.63L79,114.16,96,114Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M82,114.1c.68,0,1.45,0,2.37,0,1.63,0,2.32,0,3.88,0l7.69-.06-.06,1.74h4.78l.11-1.4,2.37.07c9,.47,13,1.2,13,2.16,0,1.48-14.2,2.68-31.72,2.68s-31.72-1.2-31.72-2.68c0-.74,4.17-.94,9.87-1.76l2-.19s3.16,2.65,8.79-.34h0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M61.84,98.57H40.47a4.09,4.09,0,0,1-4.09-4.09h0a4.09,4.09,0,0,1,4.09-4.09"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M40,54.31s4.4.92,6.23,14.5"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M45,51.49a43.32,43.32,0,0,1,1.63,17"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M41,79.63C41,84.37,38.34,95,48.24,95.7l13.68-.31V108.7a7.19,7.19,0,0,0,7.19,7.19h0a7.19,7.19,0,0,0,7.19-7.19V90.2a7,7,0,0,0-7-7h0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M74.54,114.07c5.32,1.42,8.13-2.76,8.13-6.73V88.84c0-3.88-2.93-8.19-6.81-8.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M40.77,47.82H36.64a4.9,4.9,0,0,0-4.9,4.9V56c0,1,.11,10.36.16,16.9a6.42,6.42,0,0,0,6.42,6.37H41"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "57.07",
    cy: "11.09",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "57.07",
    cy: "11.09",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "57",
    y1: "13.87",
    x2: "56.97",
    y2: "16.32"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "56.3",
    cy: "35.78",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "73.95",
    y: "76.37",
    width: "45.6",
    height: "4.26",
    rx: "2.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M42.46,50.37c-.34.42-.7.84-1.07,1.24a20.73,20.73,0,0,0-5.41,15h0A20.72,20.72,0,0,0,51.34,85.62L61,88.2a6.21,6.21,0,0,0,7.27-3.44h0a6.22,6.22,0,0,0-3.59-8.42L46.34,69.06"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "73.95 78.83 69.5 63.15 66.87 52.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M68.43,50.88a72.2,72.2,0,0,0,5,10.29c2.53,4.28,7,5.77,12.09,6.24a5.37,5.37,0,0,1,4.92,5.33h0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M70,54.54a16,16,0,0,1,.32,11.39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M79.9,75.66h0a2.52,2.52,0,0,1,2.52-2.52l16.4,0,.58-1.06,7.73-16.72a1.49,1.49,0,0,1,1.49-.86h0a1.5,1.5,0,0,1,1.25,2.1L101.4,75.81"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-9",
    points: "48.23 99.1 48.64 114.79 43.86 114.79 43.81 99.02"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-9",
    points: "100.23 81.1 100.64 115.79 95.86 115.79 95.81 81.02"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M24.27,115.31s-6.53-6.89,2.83-14.3c0,0-2.39,9.59,1.09,9.37s5.23-9.15.66-18.95c0,0,8.93.65,7.84,16.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M86.89,109s-2.74-4.43,2.48-10.45c0,0,.11,7.84,2.5,6.68s2.28-7.47-.55-11.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M101.81,102.25s5-.76,4.49,9.68c0,0,.4-3.85,7.46-6.6,0,0-1.51,7.52,3.49,8.82"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M71.76,38C72,44,65.44,50.83,57.16,51.09S42,45,41.78,38.92s0-11,14.64-11.42C70.41,27.07,71.57,31.94,71.76,38Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M75.26,32.78l2.71,1a1.46,1.46,0,0,1,.81,1.34v2.68a1.43,1.43,0,0,1-1,1.39L75,40"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M36.77,33.33l-2.64,1.21a1.45,1.45,0,0,0-.81,1.33v2.69a1.43,1.43,0,0,0,1,1.39l2.8.84"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "60.92",
    y: "57.83",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }))));
}
