// @postman/aether/esmLib/components/Illustration/illustrations/EmptyTrash.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function EmptyTrash() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "empty-trash-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "80.12",
    y1: "30.65",
    x2: "89.52",
    y2: "30.65",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.17",
    stopColor: "#fb800d"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.72",
    stopColor: "#efae2e"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "72.44",
    y1: "55.18",
    x2: "89.95",
    y2: "55.18",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M89.52,28.32a14.56,14.56,0,0,0-9.4,1.83l.76,3s4.72-4.88,8.64-4.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M72.44,51.57C74.67,57.14,81.1,62.31,90,62.37c0,0-12.57-1.73-13.37-14.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M80,30.1c2.94-2.18,6.29-2.23,9.61-2.23a17.29,17.29,0,0,1,17.29,17.29c0,9.55-6.28,16.63-15.81,17.15-6.31.35-14.61-2.27-18-10.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M80.88,33.42C83,30,87.75,28.24,91.68,28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M91.51,62.33c-6.63,0-13.74-5.11-15.33-13.87"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "91.92",
    cy: "44.69",
    rx: "9.28",
    ry: "10.14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M119.42,115.86c1.44,4.25-52.44,2.69-67.79,2.69s-27.79-1.2-27.79-2.69c0-.74,6.1-1.63,11.13-2.07,1-.09,1.24,2.78,3.12,2.67,1.28-.08.35.16,3.54.13,3.4,0,5.17-3.38,5.85-3.4,2.67-.08,1.2,0,4.15,0,1.43,0,2.43,0,3.8,0,.63,0,1.35,2.27,2.52,2.79a7.9,7.9,0,0,0,4,.7,14.65,14.65,0,0,0,4.61-1.31c.64-.23.57-1.8,1.28-1.75C75.67,114.1,118.55,113.29,119.42,115.86Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "61.43",
    cy: "35.92",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M53.46,90.85l.22,19.53a7.61,7.61,0,0,0,7.67,6.73h0A7.6,7.6,0,0,0,68.82,110c.8-11.49-1.5-31.47-2.35-55.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M34.13,60.65l-1,19.41L33,107.89c-.14,5.7,2.61,9.1,7.41,9h.55a7.47,7.47,0,0,0,7-5.1l2-19.83,4.78-1.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M34.12,64.62c0-.38,6.88-8.35-.34-21.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M34,66.68s12.78-11.1,5.44-27"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M69.82,53.48a32.59,32.59,0,0,1-2.21,13.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.73,62.68c-5.22,2.87-10.08,5.26-20.45,4.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M31.81,44.94l-.11-4C31.7,38.73,32.75,37,34,37h8.15"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M42.38,38.37S25.88,40.78,26,65.14c0,5,.76,10.94,4.26,14.46a15.8,15.8,0,0,0,3,2.4"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M110.59,75.1,108,111.41a2.77,2.77,0,0,1-2.76,2.58H77.84a2.78,2.78,0,0,1-2.77-2.58L72.49,76l-.14-1.21"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "79.04",
    y1: "83.65",
    x2: "80.18",
    y2: "113.8"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "96.04",
    y1: "82.37",
    x2: "94.79",
    y2: "113.8"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "104.19",
    y1: "83.85",
    x2: "102.33",
    y2: "113.8"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "87.27",
    y1: "91.32",
    x2: "87.72",
    y2: "113.8"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "91.47",
    cy: "74.94",
    rx: "19.21",
    ry: "2.43"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M105.29,73.25c3.34.44,5.39,1,5.39,1.69,0,1.34-8.6,2.43-19.21,2.43s-19.21-1.09-19.21-2.43,8.6-2.43,19.21-2.43c1.69,0,3.32,0,4.88.08Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "57.01",
    y: "58.2",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "55.91",
    cy: "38.01",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "60.22",
    y1: "13.9",
    x2: "60.3",
    y2: "16.41"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-8"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-9",
    cx: "60.35",
    cy: "11.24",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "60.23",
    cy: "11.04",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M67.52,53.83c-3.86-5-2.71-12.31,1.88-16.92,2.92-2.93,7.11-3.63,11.21-2.48"
  }))));
}
