// @postman/aether/esmLib/components/Illustration/illustrations/Enterprise.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function Enterprise() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "enterprise-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "88.39",
    y1: "18.02",
    x2: "108.45",
    y2: "18.02",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.17",
    stopColor: "#fb800d"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.72",
    stopColor: "#efae2e"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M94.64,116c0,2-12.76,3.66-28.5,3.66S37.65,118,37.65,116c0-1.07,3-2,8.67-2.65.94-.11.57,1,.85,2s1.51,2.68,2.66,2.6c1.3-.08,1.11,1.93,7.8,0,.71-.21,1.88-1.77,2.55-3.23.56-1.21.6-2.37,1.25-2.39,1.53,0,3.11,0,4.71,0,2.3,0,3.33-.12,5.47,0,.81,0,.15,2.27,1.42,3.68a11.85,11.85,0,0,0,5.83,2.88c1.63.1,4.36-1.36,5.37-1.75.52-.21,1.51-4.41,2-4.36C92.65,113.4,94.64,114.85,94.64,116Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M87,32.76a17.81,17.81,0,0,1,1,5.14,19.29,19.29,0,1,1-6.7-13.69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M54,51.49C45,57.7,34.4,72.57,31,78.55c-3.81,6.8,2.56,9.84,2.56,9.84s4.25,2.31,8.4-3c3.44-4.39,7.87-9.27,9.38-11.27l-.47,11.32-4.64,24.77a7.43,7.43,0,0,0,0,3.08,7.58,7.58,0,0,0,7.17,5.77h0c4.09.09,6.29-2.69,7.72-7.17L66,95.84l5.09-3.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M51.32,74.32s5.27-7.58-.94-20.09"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M51.3,76.09s9.19-11,4.32-23"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M45.77,58.83,45.06,53a2.33,2.33,0,0,1,2.22-2.44l5.85-.36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M81.23,30.37c-2.46.21.49-.06-3.85.46-7.63.93-10.63,4.1-11.8,6.16a7.18,7.18,0,0,0-.32,4C66,44.86,69,53.8,83.49,51.15"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M103.37,33.45a5.9,5.9,0,0,1,0,4.1c-6.33,17.6-18.84,23.7-18.84,23.7s1.25,15.26,1.14,20l.69,30.09a7.39,7.39,0,0,1-7.55,7.54h0a7.39,7.39,0,0,1-7.16-6.4L70.07,93.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M88.07,38.85c1.19-.67,3.47-4.25,4.16-5.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M85.23,49.8s3.2,8-.69,11.45c0,0-.48-6.45-3.07-7.76"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-4",
    points: "98.73 8.35 101.57 14.76 108.45 16.05 103.24 20.73 104.13 27.69 98.07 24.17 91.73 27.17 93.2 20.32 88.39 15.22 95.37 14.5 98.73 8.35"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "98.73 8.35 101.57 14.76 108.45 16.05 103.24 20.73 104.13 27.69 98.07 24.17 91.73 27.17 93.2 20.32 88.39 15.22 95.37 14.5 98.73 8.35"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "62.53",
    y1: "18.29",
    x2: "63.31",
    y2: "20.27"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "61.51",
    cy: "15.04",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "61.67",
    cy: "15.45",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M48.16,77.87l-.11,2.67a2.45,2.45,0,0,0,2.34,2.55l.71,0"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "81.37",
    y: "4.17",
    width: "34.96",
    height: "28.63"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "73.27",
    y: "62.46",
    width: "4.93",
    height: "5.07",
    rx: "0.43",
    transform: "translate(151.46 129.99) rotate(-180)"
  }))));
}
