// @postman/aether/esmLib/components/Illustration/illustrations/ForkOnline.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function ForkOnline() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "fork-online-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "45.08",
    y1: "36.29",
    x2: "56.23",
    y2: "36.29",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "293.04",
    y1: "59.64",
    x2: "296.05",
    y2: "59.64",
    gradientTransform: "matrix(-1, 0, 0, 1, 395.13, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "287.37",
    y1: "59.94",
    x2: "290.89",
    y2: "59.94",
    gradientTransform: "matrix(-1, 0, 0, 1, 395.13, 0)",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M121.17,109.37c0,2.91-12.76,5.28-28.51,5.28s-28.5-2.37-28.5-5.28c0-1.46,3.86-3,9-3.86,1-.17.57,4.43,2.62,6,4.34,3.33,6.79,1.78,9.93,0,1.47-.84,2-7.33,2.67-7.37a34.78,34.78,0,0,1,4.27-.07c1.47,0,1.42.11,2.82.06.66,0,.65,5.5,1.86,6.5.86.72,3.57,3.24,5.82,3.24,4.21,0,5.65-3.89,6.61-4.54.66-.45-.31-4.45.41-4.37C118.25,105.91,121.17,107.48,121.17,109.37Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M49.29,30.24c-4.65,0-8.42,2.71-8.42,6.05s3.77,6.05,8.42,6.05c0,0-5.49-3.69-1.7-9.92,2.39-3.93-2.88.79-2.88.79Z"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M51.13,42.34a6.05,6.05,0,0,1,0-12.1c2.79,0,4.4.9,5.1,3.46,0,0-5.78,0-5.1,8.64"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M52.14,44.4H26.23a7.4,7.4,0,0,1-7.4-7.4v-.78a7.4,7.4,0,0,1,7.4-7.39H51.31a7.41,7.41,0,0,1,6.94,4.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M108.35,41.6A19.27,19.27,0,0,0,99.64,8.8,19.28,19.28,0,0,0,83.07,42.57l2.16,1.49"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M81.11,40.52S71.47,42.86,60.89,35a6.38,6.38,0,0,0-9.16,1.59h0a6.35,6.35,0,0,0,.83,8A35.27,35.27,0,0,0,76.05,54.75L73.3,104.86c0,4.35,2.64,8.79,7,8.74h0c4.29,0,6.78-3,7.74-7.82l2.77-18.4a1.52,1.52,0,0,1,1.51-1.44h0a1.5,1.5,0,0,1,1.5,1.43l1.77,19.1a7.36,7.36,0,0,0,7.36,7.35h0a7.36,7.36,0,0,0,7.36-7.35l-.93-29.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M79.54,16.94A12.42,12.42,0,0,1,87.3,31.51c-.87,3.62-3.89,6.67-7,8.27"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "84.03",
    y: "42.12",
    width: "31.68",
    height: "34.09",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M113.3,75.06c0-.3,2.33,8.46-3.87,7.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M75.61,61.85S79.47,63.8,84,58.14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M75.4,66.55s5.43.58,8.69-3.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M51.06,42.34a6.05,6.05,0,1,1,5.38-8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M77.4,40.79S84,43.6,84.12,48.3"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M44.41,24.54H59.7s.65-5.59-4-4.64c0,0,.8-3.85-4.16-3.42a1.92,1.92,0,0,0-1.18.53,2.25,2.25,0,0,0-.9,1.65s-3.14-.55-2.88,2.52C46.6,21.18,43.34,21.21,44.41,24.54Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "97.68",
    cy: "29.53",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M92.58,42.07v7.41a5,5,0,0,0,5,5H115.1"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "100.59",
    cy: "59.64",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-9",
    cx: "106",
    cy: "59.94",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "100.59",
    cy: "59.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "105.95",
    cy: "59.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M115.19,58h-4c-.77,0-1.39.82-1.39,1.84h0a1.65,1.65,0,0,0,1.39,1.83h4"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M115.19,65h-22c-.77,0-1.39.82-1.39,1.84h0a1.65,1.65,0,0,0,1.39,1.83h22"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "108.65",
    y1: "10.21",
    x2: "107.07",
    y2: "12.17"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-10",
    cx: "110.48",
    cy: "8.27",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(32.41 86.88) rotate(-49.48)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "110.51",
    cy: "8.04",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(32.6 86.83) rotate(-49.48)"
  }))));
}
