// @postman/aether/esmLib/components/Illustration/illustrations/HitSend.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function HitSend() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "hit-send-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "-686.42",
    y1: "-530.5",
    x2: "-677.93",
    y2: "-530.5",
    gradientTransform: "matrix(-0.33, -0.95, -0.95, 0.33, -599.87, -446.27)",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "489.28",
    y1: "-1746.16",
    x2: "497.07",
    y2: "-1746.16",
    gradientTransform: "matrix(0.82, 0.57, -0.57, 0.82, -1336.17, 1230.06)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "112.61",
    y1: "16.82",
    x2: "129.81",
    y2: "16.82",
    gradientTransform: "matrix(1, 0, 0, 1, 0, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "107.91",
    y1: "17.44",
    x2: "116.4",
    y2: "17.44",
    gradientTransform: "matrix(1, 0, 0, 1, 0, 0)",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M125.47,20.93S128,26.3,122.2,31l-1-4.23"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M125.24,20.6s3.82,4.55-3.4,10.81l-1.22-4.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M54.34,73.05l-.46,1.4c.36.65.6.65,1.33.75a10,10,0,0,0,9.12-3.94c.34-.42.42-.53.45-.92a5.51,5.51,0,0,0-.61-2.92l-.49-1s1.73,4.3-4.44,6.08S54.34,73.05,54.34,73.05Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M80.88,111.8c0,2-12.7,3.67-28.36,3.67s-28.35-1.64-28.35-3.67c0-1.07,3.55-2,9.23-2.71.93-.11.75,1.16,1,2.18A2.09,2.09,0,0,0,36.6,113c1.29-.08.79,2.71,7.45.79a7,7,0,0,0,3-3.13c.55-1.21.14-2.47.78-2.49,1.52,0,3.09-.05,4.69-.05,2.29,0,1,0,3.1.11.81,0,.54,2.33,1.8,3.74a9.89,9.89,0,0,0,5.43,2.55c1.62.1,4-1.41,5-1.8.52-.21,2.11-3.82,2.63-3.77C76.83,109.63,80.88,110.66,80.88,111.8Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M29.06,51.74l.16-8.05a2.29,2.29,0,0,1,2.29-2.3l4.45-.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "51.72",
    y1: "10.09",
    x2: "51.8",
    y2: "12.24"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-6",
    cx: "51.85",
    cy: "7.43",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "51.73",
    cy: "7.23",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M68.08,75.87,69,89.75l.31,3.8.55,13.39a6.8,6.8,0,0,1-6.78,7h0a6.81,6.81,0,0,1-6.79-6.53L54.14,90.6a2.62,2.62,0,0,0-2.53-2.24h0A2.61,2.61,0,0,0,49,90.69l-1.54,16.49a6.82,6.82,0,0,1-6.8,6.83h0a6.82,6.82,0,0,1-6.8-6.83l.91-30.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M49.79,78.45c1.05,2.56,2.72,2.81,5.32,2.81a17.71,17.71,0,0,0,4.1-.93c13.87-5.72,17.55-10.89,18.18-17.91.83-9.15-5.68-16-9.79-19.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M68,56.21c-1.25,3.49-4.59,7.34-8.21,9.71"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M53.67,69.22c.22-1.41,1.37-2.09,3.58-2.93,2.75-1.05,5.28-1,5.89.57s-.89,3.93-3.64,5c-1.89.72-3.41,1.44-4.6.87"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M54.35,72.89a8,8,0,0,0,4.86-.57c3-1.25,5-3.52,4.39-5.09s-3.62-1.82-6.68-.58a8.9,8.9,0,0,0-3.47,2.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M37.28,43,34.6,44.72a28.7,28.7,0,0,0-8.84,20.11c-.06,2.76,2.12,12,19.7,14,.39,0,1.19,0,1.33,0,3.69,0,7.5-1.91,7.52-6.2a6.48,6.48,0,0,0-.23-2.21c-.78-2.35-2.75-3.69-6.95-4.11-5.46-.55-8.77-4.2-10.55-4.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M58.44,72.69s3.06,3.88,3.35,6.42"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M61.54,71.26s2.92,3.67,3.21,6.21"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M33.4,45.83a24,24,0,0,1,4,15.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M36.72,43.26s5.42,7.21,4.16,20.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M67,57.84S65.33,49,68.38,43.61"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M64.61,61.52a20.35,20.35,0,0,1-1.42-14.41"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "52.3",
    cy: "31.41",
    rx: "18.93",
    ry: "19.02"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72.1,46.8C83.26,41,83.58,49.47,83,54.23a.91.91,0,0,0,1.55.74l29.59-29.83"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M115.88,26.7,103.53,47s12.4-2.9,1,13.6L96.46,75.31a.91.91,0,0,0,1.05,1.38c5.53-2,17.08-4.68,13.12,9.07a.91.91,0,0,0,1.28,1.07c6.18-3.13,24.29-8.11,13.44,15a.91.91,0,0,0,1,1.29H139"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M34.34,84.78S5.22,81,9.05,103l-8,.13"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "20.16",
    cy: "76.72",
    rx: "5.36",
    ry: "3.11",
    transform: "translate(-54.5 49.93) rotate(-55.53)"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "128.94 46.83 129.94 47.38 128.94 47.94 128.4 48.96 127.85 47.94 126.86 47.38 127.85 46.83 128.4 45.81 128.94 46.83"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "83.69 34.26 84.69 34.82 83.69 35.37 83.14 36.39 82.6 35.37 81.6 34.82 82.6 34.26 83.14 33.24 83.69 34.26"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "101.81 25.15 102.81 25.71 101.81 26.27 101.27 27.29 100.72 26.27 99.73 25.71 100.72 25.15 101.27 24.13 101.81 25.15"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "13.43 58.36 14.43 58.92 13.43 59.48 12.88 60.5 12.34 59.48 11.34 58.92 12.34 58.36 12.88 57.34 13.43 58.36"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "19.17",
    cy: "24.72",
    rx: "2.27",
    ry: "2.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M124,22s0,0,0,.09.07,0,.08,0,0-.06,0-.1S123.94,21.87,124,22Zm2-1.78c.12-.1.26-.15.14-.05s-.06.05-.13.07S125.86,20.25,126,20.17Zm-4.39,6.31s-.06.05-.13.09l0,0,.06-.05Zm.27-1.76s-.06,0,0-.06,0,0,.07,0S121.94,24.76,121.87,24.72Zm.31-.33s-.08,0-.08,0,0-.06,0-.08,0,0,.07,0S122.19,24.38,122.18,24.39Zm.56-.56-.06,0s.1,0,.08,0S122.75,23.81,122.74,23.83Zm1.33-2c-.09,0-.15-.05-.14-.07s.1,0,.17,0Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M129,7c-.06-.39-.91,0-1-.43h0a5.49,5.49,0,0,0-.58,0,11.83,11.83,0,0,0-8,4.9l-2.65,3.61-4.1,6.27,3,2.26,5,3.45L125.78,20l3.07-4.45C130.71,13,129.37,9.62,129,7Zm-4.19,8.9c-.64.93-1.52,1.56-2.45.91a2.73,2.73,0,0,1-.81-3.29c.64-.93,2.2-.72,3.12-.07S125.44,15,124.8,15.89Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M128.76,6.86l-.71-.06a10,10,0,0,0-8.7,4.27l-2.57,3.59-4.46,6.21L116.49,24l4.14,3.15,4.69-6.54,2.44-3.39a10.23,10.23,0,0,0,1.53-8.91c-.14-.45-.53-1.41-.53-1.41Z"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "123.18",
    cy: "14.6",
    rx: "2.04",
    ry: "2.05"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-10",
    d: "M116.4,14.26s-5.9-.66-8.49,6.38l4.34-.39"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M118.18,25.42l-.61,1.38a.43.43,0,0,1-.65.24l-3.2-2.51a.44.44,0,0,1,0-.7l1.2-1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M116.88,14.3s-5.63-2.06-9.2,6.79l4.84-.36"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "123.18",
    cy: "14.6",
    rx: "2.04",
    ry: "2.05"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "55.48",
    y1: "70.63",
    x2: "59.24",
    y2: "68.77"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-4",
    points: "58.63 67.41 60.08 69.86 61.56 67.48 58.63 67.41"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "49.63",
    y1: "61.87",
    x2: "52.35",
    y2: "65.65"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "53.6",
    y1: "61.15",
    x2: "54.41",
    y2: "64.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M69.81,23.76a45.14,45.14,0,0,0-8.1-.17c-7.66.56-10.82,3.57-12.08,5.58a7.17,7.17,0,0,0-.53,4C49.58,36.7,53.21,45,65.92,44.64"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "54.89",
    y: "53.46",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "119.45",
    cy: "70.12",
    rx: "2.27",
    ry: "2.28"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "40.94",
    cy: "32.63",
    r: "2.88"
  }))));
}
