// @postman/aether/esmLib/components/Illustration/illustrations/InlineComment.js
import React from 'react';
export default function InlineComment() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "inline-comment-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "34.24",
    y1: "36.73",
    x2: "57.77",
    y2: "36.73",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "921.91",
    y1: "62.91",
    x2: "924.92",
    y2: "62.91",
    gradientTransform: "matrix(-1, 0, 0, 1, 1028.61, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "916.25",
    y1: "63.21",
    x2: "919.77",
    y2: "63.21",
    gradientTransform: "matrix(-1, 0, 0, 1, 1028.61, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "49.99",
    y1: "29.53",
    x2: "58.58",
    y2: "29.53",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-5",
    x1: "42.47",
    y1: "54.7",
    x2: "60.67",
    y2: "54.7",
    gradientTransform: "matrix(-1, 0, 0, 1, 77.67, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-6",
    x1: "68.63",
    y1: "59.39",
    x2: "76.2",
    y2: "59.39",
    gradientTransform: "translate(-29.78)",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M57.77,38.48H34.24V35H57.67Zm-22.53-1h21.5l0-1.51H35.24Z"
  })), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "17.58",
    y: "35.47",
    width: "14.24",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "17.58",
    y: "43.99",
    width: "10.24",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "55.9 46.49 31.58 46.49 31.58 43.99 54.82 43.99"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "102.29",
    cy: "32.79",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M97.19,45.34v7.41a5,5,0,0,0,5,5H119.7"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "105.19",
    cy: "62.91",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "110.6",
    cy: "63.21",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "105.19",
    cy: "63",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "110.55",
    cy: "63",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M119.8,61.3h-4.06c-.77,0-1.39.82-1.39,1.84h0A1.65,1.65,0,0,0,115.74,65h4.06"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M119.8,68.3H97.74c-.77,0-1.39.82-1.39,1.84h0A1.65,1.65,0,0,0,97.74,72H119.8"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "113.25",
    y1: "13.48",
    x2: "111.68",
    y2: "15.44"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "115.08",
    cy: "11.54",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(31.54 91.52) rotate(-49.48)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "115.12",
    cy: "11.31",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(31.72 91.47) rotate(-49.48)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M56.19,31.77H54.8l-1.49,2.05-1.45-2.05H50.65A.66.66,0,0,1,50,31.1V25.91a.65.65,0,0,1,.66-.66h7.26a.66.66,0,0,1,.67.66v5.84Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M58.58,28.58V25.91a.66.66,0,0,0-.67-.66H50.65a.65.65,0,0,0-.66.66V31.1a.66.66,0,0,0,.66.67h1.21l1.45,2.05,1.49-2.05h1.5"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "51.86",
    y1: "27.15",
    x2: "56.82",
    y2: "27.15"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "51.86",
    y1: "29.13",
    x2: "55.72",
    y2: "29.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M60,30.43h-.46V30a.87.87,0,1,0-1.74,0v.46h-.46a.87.87,0,0,0,0,1.74h.46v.46a.87.87,0,0,0,1.74,0v-.46H60a.87.87,0,1,0,0-1.74Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "58.69",
    y1: "30.09",
    x2: "58.69",
    y2: "32.48"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "57.5",
    y1: "31.28",
    x2: "59.89",
    y2: "31.28"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-10",
    d: "M35.19,54.7h0a1.81,1.81,0,0,1-1.74,1.88H18.53A1.73,1.73,0,0,1,17,54.7h0a1.73,1.73,0,0,1,1.54-1.88H33.45A1.81,1.81,0,0,1,35.19,54.7Z"
  })), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "33.26",
    y1: "54.7",
    x2: "18.91",
    y2: "54.7"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "37.9 51.23 36.42 51.83 35.18 51.23"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "35.05 58.1 36.54 57.51 37.77 58.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "36.51",
    y1: "57.55",
    x2: "36.51",
    y2: "52.06"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-11",
    points: "44.32 61.36 43.09 61.36 41.78 63.16 40.5 61.36 38.85 61.36 38.85 55.62 46.41 55.62 46.41 58.56 46.41 61.36 44.32 61.36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M44.32,61.36H43.09l-1.31,1.8-1.28-1.8h-.85a.8.8,0,0,1-.8-.8V56.42a.8.8,0,0,1,.8-.8h6a.8.8,0,0,1,.8.8v4.14a.8.8,0,0,1-.81.8Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "40.67",
    y1: "57.67",
    x2: "44.4",
    y2: "57.67"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "40.57",
    y1: "59.39",
    x2: "42.96",
    y2: "59.39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-12",
    d: "M125.77,112.64c0,2.91-12.76,5.28-28.5,5.28s-28.5-2.37-28.5-5.28c0-1.46,3.85-3,9-3.86,1-.17.56,4.43,2.61,6,4.34,3.33,6.79,1.78,9.93,0,1.48-.84,2-7.33,2.68-7.37a34.7,34.7,0,0,1,4.27-.07c1.46,0,1.41.11,2.82.06.65,0,.64,5.5,1.85,6.5.86.72,3.57,3.24,5.83,3.24,4.2,0,5.65-3.89,6.6-4.55.66-.44-.31-4.44.42-4.36C122.85,109.18,125.77,110.74,125.77,112.64Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M113,44.87a19.28,19.28,0,1,0-25.29,1l2.17,1.49"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M85.71,43.79s-9.64,2.34-20.22-5.57a6.37,6.37,0,0,0-9.15,1.59h0a6.35,6.35,0,0,0,.82,8A35.32,35.32,0,0,0,80.65,58l-2.74,50.11c0,4.35,2.64,8.79,7,8.74h0c4.28-.05,6.77-3,7.74-7.83l2.76-18.39a1.51,1.51,0,0,1,1.51-1.44h0a1.51,1.51,0,0,1,1.51,1.42l1.77,19.11a7.35,7.35,0,0,0,7.35,7.35h0a7.35,7.35,0,0,0,7.36-7.35L114,79.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M84.15,20.21A12.41,12.41,0,0,1,91.9,34.77C91,38.4,88,41.45,85,43.05"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "88.63",
    y: "45.38",
    width: "31.68",
    height: "34.09",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M117.9,78.33c0-.3,2.34,8.46-3.86,7.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M80.21,65.12s3.86,1.95,8.42-3.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M80,69.82s5.44.58,8.7-3.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M82,44.06s6.58,2.81,6.71,7.51"
  }))));
}
