// @postman/aether/esmLib/components/Illustration/illustrations/InternalServerError.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function InternalServerError() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "internal-server-error-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "92.35",
    y1: "22.89",
    x2: "103.45",
    y2: "22.89",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "94.56",
    y1: "49.59",
    x2: "99.69",
    y2: "49.59",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "105.09",
    y1: "43.2",
    x2: "107.77",
    y2: "43.2",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "86.68",
    y1: "11.96",
    x2: "89.36",
    y2: "11.96",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-5",
    x1: "102.95",
    y1: "106.82",
    x2: "123.01",
    y2: "106.82",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-6",
    x1: "35.03",
    y1: "66.04",
    x2: "38.57",
    y2: "66.04",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-7",
    x1: "29.67",
    y1: "66.04",
    x2: "33.21",
    y2: "66.04",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "97.9",
    cy: "22.89",
    rx: "5.55",
    ry: "5.45"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M76.31,115.2c0,2-7.78,3.58-23.73,3.58s-28.89-1.63-28.89-3.65c0-1.07,3.63-2,9.4-2.69,1-.11,1.28,1.1,1.57,2.12s.59,1.51,1.75,1.43c1.32-.08.75,3.05,7.54,1.13A7.07,7.07,0,0,0,47,114c.56-1.21.14-2.46.79-2.48,1.56,0,3.16,0,4.79,0,2.32,0,1,0,3.15.1.82,0,.65,2,1.93,3.44a10.84,10.84,0,0,0,5.44,2.83c1.65.1,4.06-1.4,5.08-1.79.53-.21.83-3.82,1.37-3.77C76,113,76.31,114.06,76.31,115.2Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M53.49,53.15l.73-.22A18.89,18.89,0,0,0,47,16.45,19.12,19.12,0,0,0,27.67,35.38a18.66,18.66,0,0,0,4.41,12l1,1.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M65.75,63.55l4.11,46.23c0,4.28-2.16,7.76-6.52,7.71h0c-4.29,0-7.74-4.39-7.74-8.6L53.76,93.53c0-.79-3.47-1.42-4.28-1.42h0c-.8,0-2.38.62-2.43,1.4l1,17c0,4-2.37,7.22-6.43,7.22h0c-4.06,0-8.28-3.23-8.28-7.22l-.45-4.65c-.81-8.28-1.35-15.7-1.37-24"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "22.07",
    y: "48.39",
    width: "31.68",
    height: "33.37",
    rx: "6.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M66.36,70s-3.54,6-12.62,4"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M58.33,50.62S70.53,49.55,78,38.79a6.43,6.43,0,0,1,9.15-1.46h0A6.19,6.19,0,0,1,89,45,35.06,35.06,0,0,1,70.09,62.06l-4.5,1.4A9.51,9.51,0,0,1,54,69.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M61.8,23.23s1.57-3.08,4.39-2.76,9.31,4.74-.16,18.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "75.89",
    y1: "35.01",
    x2: "68.52",
    y2: "35.01"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "69.45 22.2 69.45 8.25 120.21 8.25 120.21 35.01 87.86 35.01"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M74.81,42s-4.34-3-1.12-5.59L88,27.27l3,4.45-4.69,4.44"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "87.99 27.27 89.9 26.13 92.9 30.02 91.01 31.72"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "92.27",
    y1: "27.83",
    x2: "97.66",
    y2: "23.76"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M81.13,35.65a1.52,1.52,0,0,1-.28-.89,1.67,1.67,0,0,1,3.33,0,1.6,1.6,0,0,1-.13.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M86.18,28.58l2.07,2.52a1,1,0,0,1-.16,1.46h0a1.08,1.08,0,0,1-1.48-.16l-2.27-2.73"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "106",
    y1: "7.72",
    x2: "106",
    y2: "35.73"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "106.03 29.7 102.96 29.7 102.96 12.49 105.98 12.49"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M102.81,18.55c-1.91,0-3.45-1.14-3.45-2.55s1.54-2.55,3.45-2.55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M102.08,27.26h-.92a.57.57,0,0,1-.58-.56V21.76a.57.57,0,0,1,.58-.57h.92"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "81.75",
    y1: "8.42",
    x2: "81.75",
    y2: "30.93"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "96.06",
    y1: "27.99",
    x2: "96.06",
    y2: "43.29"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "95.33",
    y1: "3.34",
    x2: "94.99",
    y2: "15.05"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "98.68",
    y1: "29.75",
    x2: "103.77",
    y2: "39.07"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "97.12",
    cy: "49.59",
    rx: "2.57",
    ry: "4.99"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-8",
    cx: "106.43",
    cy: "43.2",
    rx: "1.34",
    ry: "1.29"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-9",
    cx: "88.02",
    cy: "11.96",
    rx: "1.34",
    ry: "1.29"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "97.05",
    cy: "49.95",
    rx: "2.57",
    ry: "4.99"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "106.57",
    cy: "43.27",
    rx: "1.34",
    ry: "1.29"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "88.09",
    cy: "12.03",
    rx: "1.34",
    ry: "1.29"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "94.14",
    y1: "17.8",
    x2: "88.98",
    y2: "13.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M65.54,41.07c-6.17-.72-10.23-5.73-10.23-12,0-3.67,1-6.43,3.56-8.67"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "110.94 97.01 110.94 91.78 76.6 91.78 76.6 117.3 111.86 117.3"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "110.94 100.19 110.94 107.21 110.94 113.57"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "79.83 91.7 79.83 84.63 86.05 84.63 90.73 84.63 90.73 91.7"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "85.42 84.46 85.42 80.03 104.17 80.33 104.35 85.41"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "101 91.61 106.75 81.64 108.66 82.57 103.67 91.7"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-10",
    d: "M113,97a9.86,9.86,0,1,0,10,9.85A9.94,9.94,0,0,0,113,97Zm0,17.43a7.65,7.65,0,0,1-7.72-7.58c0-4.18,3.36-6.9,7.63-6.9a6.83,6.83,0,0,1,7.22,7.08C120.11,111.19,117.24,114.4,113,114.4Z"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "112.68",
    cy: "107.05",
    rx: "10.03",
    ry: "9.85"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "112.68",
    cy: "107.05",
    rx: "7.12",
    ry: "6.99"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "94.2",
    y1: "83.06",
    x2: "101.31",
    y2: "83.06"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "94.2",
    y1: "85.78",
    x2: "101.31",
    y2: "85.78"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "85.55",
    cy: "88.18",
    rx: "1.77",
    ry: "1.74"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "43.69",
    y1: "14.21",
    x2: "43.77",
    y2: "16.72"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-11",
    cx: "43.7",
    cy: "11.35",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "43.7",
    cy: "11.35",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M44.81,48.38v7.41a5,5,0,0,1-5,5H22.29"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-12",
    cx: "36.8",
    cy: "66.04",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-13",
    cx: "31.44",
    cy: "66.04",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "36.8",
    cy: "66.04",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "31.44",
    cy: "66.04",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M22.19,64.34h4.06c.77,0,1.39.82,1.39,1.84h0A1.65,1.65,0,0,1,26.25,68H22.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M22.19,71.34H44.25c.77,0,1.39.82,1.39,1.84h0A1.65,1.65,0,0,1,44.25,75H22.19"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "45.33",
    cy: "36.44",
    r: "2.88"
  }))));
}
