// @postman/aether/esmLib/components/Illustration/illustrations/JoinTeam.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function JoinTeam() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "join-team-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "1.09",
    y1: "81.42",
    x2: "25.73",
    y2: "81.42",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M15.05,72.6l-3.12,6.57L8,80.76,4.07,78.85l-1-2.79.2-4-1.57-.19-.4.23L1.09,90.92H25.73V72.62ZM13.36,85.17A2.37,2.37,0,1,1,15.8,82.8,2.4,2.4,0,0,1,13.36,85.17Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M105.72,109.52c2.55-.12,2,0,4.78,0,1.36,0-.94,0,.37-.06.6,0,.27,4.36,3.4,5.24a24.21,24.21,0,0,0,5.82.42,8.07,8.07,0,0,0,3.85-1.51c.62-.33.82-3.6,1.5-3.54,7.51.71,11.61,2,11.61,3.44,0,2.2-11.89,4-26.55,4S84,115.68,84,113.48c0-1.1,2.51-2.13,7.31-2.78"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M68.73,113.61c0,2.22-12.74,4-28.44,4s-28.44-1.79-28.44-4c0-1.1,2.49-2.09,7.64-2.75,1.05-.13,1.66,4.18,3.58,4,1.31-.11,1.71,1.64,5,.68,7.13-2.08,5.49-6,6.19-6.07,2.73-.12,3,.12,6.06.12,1.46,0,1.28,0,2.68,0,.65,0,1.34,3.55,2.54,4.31a11.63,11.63,0,0,0,5.57,1.22c1.89-.08,2.41-1.82,3.37-2.31.66-.34,1.69-3.14,2.41-3.08C64.91,110.42,68.73,112.18,68.73,113.61Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M104.15,91.87l6.18,1.31.59,16.91c.92,5.14,5.38,5.61,8.57,5.54h0a6.33,6.33,0,0,0,6.1-6.81c-1.8-26.55-.72-42.58-.72-42.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M95.64,47.23c-4.08-1.93-4.08-1.78-8.37-6.81C80,31.83,78.38,31.7,76.73,32.58c0,0-5.66,2.05-3.13,8.22C76,46.56,79.26,54.29,93.5,62.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M127.35,71.22h-2.66"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M122.31,45.2l4,.11a3.4,3.4,0,0,1,3.31,2.38l-.05,4.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M123.33,47.93c-1.26,12.24,1.41,18,1.41,18v0c2,2.21,2.9,6.51,4.33,10.15,1.3,3.29,4.35,5.36,7.21,3.57l.42-.16c2.34-1.47,2.86-5.28,2.26-8.19-1.33-6.48-5.73-19.4-17.73-24.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M117.74,49s-2,14.74,7.21,18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M110.32,97.85s-19.41,7.36-18.47-16c0,0,2.81-27.1,3.91-35.15"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M124.56,91.86a5.15,5.15,0,0,0,1.21-.68,5.77,5.77,0,0,0-1.25-10.05l-.1,0"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "108.67",
    cy: "32.55",
    rx: "18.44",
    ry: "18.8"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "35.91",
    cy: "32.55",
    rx: "18.44",
    ry: "18.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M49.91,45.41a22.25,22.25,0,0,0,11.74-8.56c1.76-2.48,5.92-5.21,9.07-1.78h0c1.93,1.56,3,5.56,2,7.84-2.33,5.1-7.45,12.83-21.26,18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M22.11,48C24.82,60.64,21,63.85,21,63.85l-3.55,4.81c-2,2.22-2.62,3.86-4,7.5C12.1,79.45,9,82.27,6.18,80.48l-1.09-.65c-2.34-1.47-2.18-5.54-1.58-8.45,1.33-6.48,8.31-19.06,20.31-24.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M27.91,49.36s1.82,14.4-7.4,17.68"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M20.82,72.31c-.06-4.81-.11-8-.11-8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M49.57,45c.87,6.31,4.93,46.26,6.59,62.47a6.53,6.53,0,0,1-6.42,7.25h0a6.49,6.49,0,0,1-6.4-5.78l-2-19.83-6.18,1.31-1.5,19.67c-.41,3.22-4.47,5.61-7.66,5.54h0a6.34,6.34,0,0,1-6.1-6.83c.49-7.22.38-11.78.54-17.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M45.75,48.34s1.39,11.79,6,16.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M98.2,48c.31.31,0,11.62-4.5,15.26"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "68.22",
    y1: "27.44",
    x2: "65.3",
    y2: "20.35"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "72.37",
    y1: "26.23",
    x2: "72.78",
    y2: "19.64"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "75.37",
    y1: "27.02",
    x2: "79.75",
    y2: "21.19"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "15.01 72.62 26.31 72.62 26.31 90.92 0.67 90.92 0.88 71.9 3.33 71.92"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "13.36",
    cy: "82.8",
    rx: "2.44",
    ry: "2.37"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M52.28,24a25.47,25.47,0,0,0-7.78-.25c-7.36.59-10.41,3.62-11.63,5.63a7.39,7.39,0,0,0-.52,4c.48,3.86,2.15,13.27,17.83,10.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M91.59,25.18c1.78-.53,3.85-.75,8.47-.42,7.28.53,10.29,3.4,11.49,5.31a6.86,6.86,0,0,1,.5,3.82c-.5,3.67-2.08,12.77-17.6,10.56"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "25.39",
    cy: "35.08",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "118.38",
    cy: "35.08",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-5",
    x: "39.73",
    y: "56.61",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-5",
    x: "108.78",
    y: "55.69",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "33.68",
    y1: "11.27",
    x2: "33.76",
    y2: "13.78"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-6",
    cx: "33.7",
    cy: "8.41",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "33.7",
    cy: "8.41",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "33.68",
    y1: "11.27",
    x2: "33.76",
    y2: "13.78"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "109.68",
    y1: "11.27",
    x2: "109.76",
    y2: "13.78"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-6",
    cx: "109.7",
    cy: "8.41",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "109.7",
    cy: "8.41",
    rx: "2.68",
    ry: "2.7"
  }))));
}
