// @postman/aether/esmLib/components/Illustration/illustrations/MergeConditions.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function MergeConditions() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "merge-conditions-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "71.56",
    y1: "24.73",
    x2: "76.77",
    y2: "24.73",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.27",
    stopColor: "#ff7105"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.46",
    stopColor: "#ff760e"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.62",
    stopColor: "#ff7d1d"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.77",
    stopColor: "#ff8733"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.91",
    stopColor: "#ff944f"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.99",
    stopColor: "#ff9d63"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "81.26",
    y1: "24.71",
    x2: "98.28",
    y2: "24.71",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "71.56",
    y1: "9.28",
    x2: "98.36",
    y2: "9.28",
    xlinkHref: "#linear-gradient-2"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "71.56",
    y1: "16.55",
    x2: "98.4",
    y2: "16.55",
    xlinkHref: "#linear-gradient-2"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-5",
    x1: "37.57",
    y1: "66.01",
    x2: "41.11",
    y2: "66.01",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-6",
    x1: "32.21",
    y1: "66.01",
    x2: "35.75",
    y2: "66.01",
    xlinkHref: "#linear-gradient-2"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "76.77 23.48 71.56 23.48 71.56 25.99 76.55 25.99 76.64 23.63"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "81.26 25.98 98.28 25.99 98.28 23.48 81.38 23.44"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M84.23,114.11c0,2.88-13.1,5.21-29.25,5.21S25.73,117,25.73,114.11c0-1.43,3.86-3,9.15-3.89,1.08-.17.63,4.65,2.17,5.89,5.3,4.29,7.87,1.8,11.09,0,1.52-.82,1.29-7.16,2-7.2a47.76,47.76,0,0,1,4.83,0c1.5,0,.71,0,2.15,0,.67,0,.68,5.39,1.92,6.38.88.71,3.92,3.22,6.23,3.09,1.94-.11,4.61-2.92,5.59-3.56.68-.45.39-5.12,1.14-5C80.28,110.7,84.23,112.25,84.23,114.11Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "71.56",
    y: "8.03",
    width: "26.81",
    height: "2.51"
  })), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "71.38",
    y: "8.03",
    width: "27.08",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "71.56",
    y: "15.3",
    width: "26.84",
    height: "2.51"
  })), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "71.38",
    y: "15.3",
    width: "27.08",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "81.05 23.5 98.46 23.48 98.46 25.99 80.96 26.01"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "76.37 25.99 71.38 25.99 71.38 23.48 76.93 23.48"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M35.78,48.42l-.84-.68a18.84,18.84,0,0,1-6.51-14.23A19.29,19.29,0,1,1,56,50.67l-.44.22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M65.92,39.33A12,12,0,0,1,55.32,27.2a11.32,11.32,0,0,1,3.92-8.95"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "99.41",
    y1: "36.88",
    x2: "87.29",
    y2: "36.89"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "55.55 15.76 55.55 0.64 109.8 0.64 109.8 26.04"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-9",
    x: "89.51",
    y: "13.32",
    width: "2.11",
    height: "6.01",
    rx: "0.99"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-9",
    x: "74.78",
    y: "5.94",
    width: "2.11",
    height: "6.01",
    rx: "0.99"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M79,22.11c0-.54.14-1,.69-1l.5,0a1,1,0,0,1,1,1v4.05"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M35.17,82.71v28.55a7.29,7.29,0,0,0,7.35,7.24h0a7.3,7.3,0,0,0,7.36-7.24l1.77-19.83A1.49,1.49,0,0,1,53.16,90h0a1.49,1.49,0,0,1,1.5,1.42l2.77,19.12a7.77,7.77,0,0,0,7.74,7.71h0c4.36,0,7-4.32,7-8.61L69.43,62.11c12-8.53,15.84-16.4,17.46-23.46s-6-9-6-9c.41-9.73-3.65-7.53-3.65-7.53C73.62,47.78,63.21,45,63.21,45"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "73.28",
    y1: "37",
    x2: "66.85",
    y2: "37"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "25.01",
    y: "48.65",
    width: "31.68",
    height: "33.6",
    rx: "6.3"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M29.69,82.27s.57,6.34,5.53,6.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M69.39,63.46s-3.6,8.68-12.68,6.66"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M69.64,69.61S66,76.43,56.89,74.41"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M47.34,48.86v6.9a5,5,0,0,1-5,5H25.14"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-10",
    cx: "39.34",
    cy: "66.01",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "39.34",
    cy: "66.01",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-11",
    cx: "33.98",
    cy: "66.01",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "33.98",
    cy: "66.01",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M25,64.31h3.75c.77,0,1.39.82,1.39,1.84h0A1.65,1.65,0,0,1,28.79,68H25"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M25,71.31H46.79c.77,0,1.39.82,1.39,1.84h0A1.65,1.65,0,0,1,46.79,75H25"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "43.87",
    cy: "36.89",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "38.23",
    y1: "14.01",
    x2: "39.25",
    y2: "16.31"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-12",
    cx: "37.34",
    cy: "11.5",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(-1.58 14.95) rotate(-22.19)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "37.16",
    cy: "11.36",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(-1.54 14.87) rotate(-22.19)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "112.63",
    cy: "40.07",
    rx: "1.4",
    ry: "1.4"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "106.75",
    cy: "31.07",
    rx: "1.4",
    ry: "1.4"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-13",
    cx: "106.47",
    cy: "40.48",
    rx: "1.25",
    ry: "1.24"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-14",
    x1: "106.54",
    y1: "40.68",
    x2: "106.71",
    y2: "32.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M106.86,32.57s-.41,3.87,1.63,4c.73,0,1.49.07,2.15.09,1.13,0,2,.51,2,1.63"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "110.05 37.94 109.28 36.65 110.25 35.37"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "108.92",
    cy: "35.82",
    rx: "9.49",
    ry: "9.45"
  }))));
}
