// @postman/aether/esmLib/components/Illustration/illustrations/MonitorRun.js
import React from 'react';
export default function MonitorRun() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "monitor-run-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "112.99",
    y1: "20.22",
    x2: "123.88",
    y2: "20.22",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "29.39",
    y1: "57.46",
    x2: "47.9",
    y2: "57.46",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "87.57",
    cy: "34.03",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M100.14,49.27c.48,24.14.86,34.56-.1,46a7.68,7.68,0,0,1-1.52,4.27c-2.37,3-7,4.06-10.22,2l-8.45-7"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "84.04",
    y1: "91.07",
    x2: "86.04",
    y2: "94.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M64.62,78.44l-3.39,28.73c-.33,5.32,1.37,7.76,6.35,8.22h.55c3.91.43,6.11-1,7.4-3.37a9.69,9.69,0,0,0,1-2.67l3.69-16.76,3.88-1.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M69.15,66.77s5.24-5.17-1.79-18.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M71.5,67.75s6.58-7.93,2.3-20.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M63.5,54.17,63.26,47a2.33,2.33,0,0,1,2.33-2.33h5.07"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M67.36,66.7l13.52,3.12a5.86,5.86,0,0,1,4.94,6.78h0A5.88,5.88,0,0,1,80,81.48c-29.1-.31-24.52-19.66-9.11-37.06"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "56.63",
    y1: "81.41",
    x2: "40.22",
    y2: "81.41"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "56.63",
    y1: "74.94",
    x2: "46.68",
    y2: "74.94"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "79.77",
    y1: "13.5",
    x2: "80.61",
    y2: "15.87"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "79.09",
    cy: "10.92",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(0.43 24.63) rotate(-17.75)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "78.91",
    cy: "10.77",
    rx: "2.68",
    ry: "2.7",
    transform: "matrix(0.95, -0.3, 0.3, 0.95, 0.47, 24.57)"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "90.98",
    y: "58.07",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "30.02",
    y1: "58.69",
    x2: "16.42",
    y2: "58.69"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "60.89",
    y1: "58.69",
    x2: "47.29",
    y2: "58.69"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "16.06 58.35 16.06 16.43 70.31 16.43 70.31 25.57"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "16.24 36.8 22.53 36.8 22.53 36.8 26.92 31.3 31.48 36.8 39.99 36.8 43.47 32.08 48.79 40.85 53.52 31.51 56.86 36.8 56.86 36.8 63.12 36.8 64.34 36.8 68.09 36.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M100.39,61.71c6.71.12,13.91-2.26,21.66-6,2.85-1.39,4.72-4.16,3.86-7.21h0c-.94-3.33-4.52-6.12-7.76-4.93-10.73,4-16,3.25-16,3.25"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M103.81,47.66s0,10.76-2.91,14.17"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M104.74,25.61c-2.45.21-4.15,0-8.49.51-7.63.93-10.63,4.1-11.8,6.16a7.27,7.27,0,0,0-.32,4c.71,3.83,3.75,12.77,18.24,10.12"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M113.14,23.06h10.72s.45-3.92-2.79-3.25c0,0,.56-2.71-2.92-2.4a1.3,1.3,0,0,0-.83.37,1.59,1.59,0,0,0-.63,1.16s-2.2-.39-2,1.76A1.62,1.62,0,0,0,113.14,23.06Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M112.84,23.06h10.72s.45-3.92-2.79-3.25c0,0,.55-2.71-2.92-2.4a1.3,1.3,0,0,0-.83.37,1.59,1.59,0,0,0-.63,1.16s-2.2-.39-2,1.76A1.62,1.62,0,0,0,112.84,23.06Z"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "43.68 29.29 45.68 26.81 41.73 26.81 43.68 29.29"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "48.86 44.31 46.86 46.79 50.81 46.79 48.86 44.31"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M29.65,62.29H47.87s.77-6.67-4.75-5.53c0,0,1-4.6-5-4.08a2.24,2.24,0,0,0-1.4.63,2.7,2.7,0,0,0-1.07,2s-3.74-.66-3.44,3C32.25,58.28,28.38,58.32,29.65,62.29Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M29.35,62.29H47.57s.76-6.67-4.75-5.53c0,0,1-4.6-5-4.08a2.24,2.24,0,0,0-1.4.63,2.7,2.7,0,0,0-1.08,2s-3.73-.66-3.43,3C32,58.28,28.08,58.32,29.35,62.29Z"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-8",
    cx: "60.86",
    cy: "116.51",
    rx: "14.12",
    ry: "1.56"
  }))));
}
