// @postman/aether/esmLib/components/Illustration/illustrations/NoActivity.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoActivity() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-activity-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, "x", /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "44.12",
    y1: "75.97",
    x2: "83.18",
    y2: "75.97",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.17",
    stopColor: "#fb800d"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.72",
    stopColor: "#efae2e"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M58.27,61.33,44.12,90.61H69L83.18,61.33ZM75,73.29H56.37l1-2.29H76Zm-16.67-5,1-2.29H78l-1,2.29Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M57.65,113.88c0,2-12.76,3.67-28.5,3.67S.65,115.91.65,113.88c0-1.07,3.58-2,9.28-2.7.94-.11.76,1.15,1,2.18a2.13,2.13,0,0,0,2.18,1.74c1.3-.09.79,2.71,7.49.78a7,7,0,0,0,3-3.13c.55-1.21.14-2.47.78-2.48,1.54,0,3.11,0,4.72,0,2.3,0,1,0,3.12.1.8,0,.53,2.34,1.8,3.74a10,10,0,0,0,5.46,2.56c1.63.09,4-1.41,5-1.81.52-.2,2.11-3.82,2.64-3.76C53.57,111.72,57.65,112.74,57.65,113.88Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M138.65,112.88c0,2-12.76,3.67-28.5,3.67s-28.5-1.64-28.5-3.67c0-1.07,3.58-2,9.28-2.7.94-.11.76,1.15,1,2.18a2.13,2.13,0,0,0,2.18,1.74c1.3-.09.79,2.71,7.49.78a7,7,0,0,0,3-3.13c.55-1.21.14-2.47.78-2.48,1.54,0,3.11,0,4.72,0,2.3,0,3,0,5.12.1.8,0,.53,2.34,1.8,3.74a10,10,0,0,0,5.46,2.56c1.63.09,4-1.41,5-1.81.52-.2,1.11-3.82,1.64-3.76C135.57,110.72,138.65,111.74,138.65,112.88Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M105.16,89.09l-.42,19.7c-.39,3.89-2.66,6.84-6.57,6.84h0a7.6,7.6,0,0,1-7.58-7c-1-11.48.14-28.8.62-52.94"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M118,43.44c8.74,6.58,17.45,22.41,20.3,28.65,3.14,6.86-1.89,9.48-1.89,9.48s-4.13,2.23-8.06-3.24c-3.25-4.52-5.61-8.64-7-10.7a.09.09,0,0,0-.16.09L123,78.18,129,106c1.07,4.68-.48,9.14-5.28,9.14h-.54a7.48,7.48,0,0,1-7-5l-6.77-18.8-4.79-2.61"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M121.78,68.25s-7-8.21,0-21.2"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M121.81,68.85s-12.18-10.4-6.19-22.78"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "104.32",
    y: "59.1",
    width: "4.93",
    height: "5.07",
    rx: "0.43",
    transform: "translate(213.56 123.27) rotate(-180)"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "32.32",
    y: "59.1",
    width: "4.93",
    height: "5.07",
    rx: "0.43",
    transform: "translate(69.56 123.27) rotate(-180)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M125.56,50.91V44.08a2,2,0,0,0-2-2h-4.91"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M125.18,73.77a2.35,2.35,0,0,1-1.28.38l-1,0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M46.62,27.15a38.29,38.29,0,0,0-9.43-.46c-7.66.55-10.81,3.57-12.08,5.57a7.3,7.3,0,0,0-.52,4c.48,3.54,1.08,12,16.8,11.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M77.89,61.11c-3.52-9,7.77-17.24,7.77-17.24l2-1.19a5.85,5.85,0,0,1,7.92,1.79h0a5.85,5.85,0,0,1-1.1,7.68L86,61.26"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M90.46,71.44C87,69.81,83.24,67.73,81.39,66"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M5.73,52.74V47.47a2,2,0,0,1,2-2h6.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M46.1,90.65,46.54,109a6.83,6.83,0,0,1-6.82,7h0a6.82,6.82,0,0,1-6.82-6.53L30.78,91.68a2.64,2.64,0,0,0-2.55-2.24h0a2.64,2.64,0,0,0-2.66,2.34L24,109.26a6.83,6.83,0,0,1-6.83,6.83h0a6.83,6.83,0,0,1-6.83-6.83l1.57-30.55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M13,45.68,11.14,46.8c-2.07,1.72-13,11.86-9.83,21.69.86,2.62,2.27,9.22,20.74,12.37.14,0,1.2,0,1.34,0,3.71,0,8-2.42,8.05-6.71a2.42,2.42,0,0,0-.22-1.28c-.79-2.34-2.89-4-7-5-4.79-1.22-9.31-3.71-11.1-4.32"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M9.31,48.59a23,23,0,0,1,4.12,15.24"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M14,45.64s4.73,6.83,3.47,19.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M46.08,66.39a33.49,33.49,0,0,1-.82-16.48"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M45.93,70.93s-7.35-15-5.51-22.51"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "28.93",
    cy: "33.49",
    r: "19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M42.36,47.15c4.64,5,13.36,11,18.74,14.28"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "69.03 90.61 44.12 90.61 58.27 61.33 83.18 61.33 69.03 90.61"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "53.12 72.19 45.86 65.94 46.23 86.06"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M85.13,44.21A19,19,0,1,1,99.44,52.7S95.75,52.45,95,52"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "103.84",
    y1: "11.97",
    x2: "103.61",
    y2: "14.57"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-8"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-9",
    cx: "104.07",
    cy: "8.57",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "104.07",
    cy: "9.01",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M82.76,28.92a22,22,0,0,1,7-2.05c7.58-1.24,11.35,1,13,2.62a7.1,7.1,0,0,1,1.44,3.78c.27,2.64,1.1,7.95-4.8,11.7l-2.92,1.4"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "113.06",
    cy: "33.33",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "18.06",
    cy: "33.33",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M78.14,34,76,28.17l-.06,0a8.36,8.36,0,1,0-3.71,3.05Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "64.25",
    y1: "21.95",
    x2: "73.4",
    y2: "21.95"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "64.25",
    y1: "24.95",
    x2: "73.4",
    y2: "24.95"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "30.73",
    y1: "11.89",
    x2: "30.5",
    y2: "14.49"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-8"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-9",
    cx: "30.97",
    cy: "8.5",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "30.97",
    cy: "8.93",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "77.04 68.28 58.37 68.28 59.37 66 78.04 66 77.04 68.28"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "74.69 73.28 56.37 73.28 57.37 71 75.69 71 74.69 73.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M87.9,59.76a102.38,102.38,0,0,0,2.77,12.11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M6.83,76.16v3.42a2,2,0,0,0,2,2h3.1"
  }))));
}
