// @postman/aether/esmLib/components/Illustration/illustrations/NoAPIKeys.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NOAPIKeys() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-api-keys-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "57",
    y1: "71.45",
    x2: "111.09",
    y2: "71.45",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.17",
    stopColor: "#fb800d"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.72",
    stopColor: "#efae2e"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M110.07,69.22H81.54a12.4,12.4,0,0,0-24.16-1,6.89,6.89,0,0,0-.28,1.29s-.26-1,.08,1.25c.18,1.25,2,2.91,2.76,4.27.29.53,1.35,3.53,1.75,4,.62.76-.85,1.55.1,2.18a13.68,13.68,0,0,0,7.56,2.57,12.4,12.4,0,0,0,12-9.41h6.83s2-2.49,4.17-2.48c2.39,0,5,2.53,5,2.48h2.72v4.15h4.52V74.42h1.9v6.49h4.53V69.22ZM69.24,79.59A7.57,7.57,0,1,1,76.81,72,7.57,7.57,0,0,1,69.24,79.59Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M98.34,115.2c0,2.92-14.2,4.28-31.72,4.28S34.9,118.12,34.9,115.2c0-1.46,3.09-2,9.27-2.74.18,0,2.07,4.93,4.22,4.71,1.45-.14,4,1.42,7.45-.36,1.64-.83,3.55-5.81,4.33-5.85,3-.16,3.08,0,6.45,0,1.63,0,2.16.23,3.72.18.73,0,.46,3.6,1.8,4.61a10.16,10.16,0,0,0,6,2.15c2.11-.11,5-1.93,6-2.58.74-.45.56-3.47,1.37-3.39C94.48,112.82,98.34,113.3,98.34,115.2Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "75.05",
    cy: "36.51",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.83,69.31c.25-8,0-9-.21-17.55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M70.33,92.34v19.2a7.61,7.61,0,0,0,7.57,6.84h0a7.61,7.61,0,0,0,7.58-7c.59-7,1.59-23,2-36.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M51.41,83.52,44.1,108.74a7.47,7.47,0,0,0,7.28,9.13h.54a7.47,7.47,0,0,0,7-5l7.77-18.8,4.79-2.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M56.21,69s5.68-9.15,0-21.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M61.89,61.64a16,16,0,0,0-.56-11.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M88.33,67.87c3.59-6.46,2.14-19.51,2.14-19.51"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "74.98",
    y1: "14.28",
    x2: "74.93",
    y2: "17.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M80.63,66.92h1.68a.43.43,0,0,0,.43-.43V62.28a.43.43,0,0,0-.43-.43H78.24a.43.43,0,0,0-.43.43v.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M50.5,51.93V47.16a2.33,2.33,0,0,1,2.33-2.33h4.58"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "75.05",
    cy: "11.16",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "75.01",
    cy: "11.6",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "69.24",
    cy: "72.02",
    r: "7.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M61.49,81.59A12.34,12.34,0,0,0,81,74.76h6.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M97,74.76H99.7V78.9h4.51V74.76h1.89v6.48h4.51V69.57H81.17a12.35,12.35,0,0,0-24.07-1,6.78,6.78,0,0,0-.28,1.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M90.39,48.85S96,59.46,95.78,69"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "87.61",
    y: "71.98",
    width: "9.74",
    height: "10.26",
    rx: "4.87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M58.06,46.19S29.68,62.5,51.5,83.6a6,6,0,0,0,8.7-.42h0a5.94,5.94,0,0,0,.19-7.44L53.73,65.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M90.23,38.42c.19,6-6.32,12.85-14.6,13.1S60.43,45.4,60.24,39.36s0-11,14.65-11.42C88.88,27.51,90,32.38,90.23,38.42Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M94.52,32.5l2.71,1A1.46,1.46,0,0,1,98,34.82V37.5a1.43,1.43,0,0,1-1,1.39l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M55.57,33.52l-2.64,1.21a1.45,1.45,0,0,0-.81,1.33v2.68a1.43,1.43,0,0,0,1,1.39l2.79.85"
  }))));
}
