// @postman/aether/esmLib/components/Illustration/illustrations/NoAPI.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoAPI() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-api-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "30.05",
    y1: "89.4",
    x2: "52.09",
    y2: "89.4",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "57.53",
    y1: "69.08",
    x2: "78.74",
    y2: "69.08",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M100.15,115.55c0,2.21-13.73,4-29.44,4s-34.44-1.8-34.44-4c0-1.1,5.5-2.09,10.64-2.75,1.05-.14,1.76,3.78,3.69,3.61,1.3-.11,1.6,2,4.87,1.09,7.12-2.09,5.48-6,6.18-6.07,2.73-.12,6,.11,9.06.11,1.46,0,1.29,0,2.69,0,.65,0,1.33,3.55,2.53,4.32A16.7,16.7,0,0,0,82.5,117c1.89-.09,2.42-1.82,3.37-2.31.66-.34,1.69-3.15,2.42-3.08C96.33,112.36,100.15,114.12,100.15,115.55Z"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "50.16 79.46 52.09 79.46 52.09 99.33 30.05 99.33 30.05 79.46 38.9 79.47 39.22 84.11 41.38 86.91 44.93 87.77 48.27 86.15 49.53 82.64 50.16 79.46"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "68",
    cy: "36.24",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M50.57,46s-10.74,5.85-11.51,36.1a5.26,5.26,0,0,0,4.87,5.38h0a5.25,5.25,0,0,0,5.6-4.82l.72-3.82.86-12.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M85.45,84.54l2.64,25.63a7.38,7.38,0,0,1-7.38,7.39h0a7.38,7.38,0,0,1-7.38-7.39L71.09,96.92a3.9,3.9,0,0,0-3.85-3.3h0a3.91,3.91,0,0,0-3.89,3.51l-1.4,13a7.38,7.38,0,0,1-7.38,7.39h0a7.38,7.38,0,0,1-7.38-7.39l1-10.89"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "68.16",
    y1: "13.92",
    x2: "68.18",
    y2: "16.9"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "68.2",
    cy: "11.27",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "68.2",
    cy: "11.27",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.36,33.72l2.35,1.08a1.45,1.45,0,0,1,.81,1.33v2.68a1.45,1.45,0,0,1-1,1.4l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.36,33.72l2.35,1.08a1.45,1.45,0,0,1,.81,1.33v2.68a1.45,1.45,0,0,1-1,1.4l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M48.85,33.59,46.21,34.8a1.45,1.45,0,0,0-.81,1.33v2.68a1.44,1.44,0,0,0,1,1.4l2.79.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M55.05,50.49a10.35,10.35,0,0,1-1.73-5.17c-.18-6,0-11,14.65-11.42,14-.42,15.15,4.44,15.34,10.48a10.7,10.7,0,0,1-.86,4.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M75.76,59.13V57.28a.43.43,0,0,1,.43-.43h4.06a.44.44,0,0,1,.44.43v4.2a.44.44,0,0,1-.44.43h-1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M48.76,47.55s2.51,5,2.25,19.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M45.19,52.58s4.76,5,5.55,17.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M48.76,47.55s2.51,5,2.25,19.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M45.19,52.58s4.76,5,5.55,17.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M84.59,47.12S82.68,52,83.5,66.61"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M88.26,50a130.5,130.5,0,0,0-4.93,19.31"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "66.92 73.94 70.73 78.59 57.53 78.69 57.53 59.48 78.74 59.48 78.74 66.96 74.06 65.25 68.55 67.81 66.92 73.94"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "69.77 78.69 57.53 78.69 57.53 59.48 78.74 59.48 78.74 66.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M84.44,46.73S95.79,50,94.78,74.2c-.72,17.23-15.87,9.8-24.55,4.09a6.92,6.92,0,0,1-1.7-10h0a6.93,6.93,0,0,1,8.78-1.93l9.91,5.28"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "50.16 79.46 52.09 79.46 52.09 99.33 30.05 99.33 30.05 79.46 38.56 79.46"
  }))));
}
