// @postman/aether/esmLib/components/Illustration/illustrations/NoCallYet.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoCallYet() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-call-yet-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "94.79",
    y1: "49.62",
    x2: "109.34",
    y2: "49.62",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M105.77,37.76H98.36a3.57,3.57,0,0,0-3.57,3.57V57.9a3.57,3.57,0,0,0,3.57,3.57h7.41a3.57,3.57,0,0,0,3.57-3.57V41.33A3.57,3.57,0,0,0,105.77,37.76ZM101,54.63H97.59V52.46H101Zm5.69,0h-3.43V52.46h3.43Zm.67-3.86H96.59V41.06h10.8Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "94.79",
    y: "37.76",
    width: "14.54",
    height: "23.71",
    rx: "3.57"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "104.35",
    y: "34.57",
    width: "3.08",
    height: "3.08"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "105.87",
    y1: "34.07",
    x2: "105.87",
    y2: "28.74"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "96.59",
    y: "41.06",
    width: "10.81",
    height: "9.72"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "97.59",
    y: "52.46",
    width: "3.44",
    height: "2.17"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "103.29",
    y: "52.46",
    width: "3.44",
    height: "2.17"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "105.85",
    cy: "26.48",
    rx: "1.66",
    ry: "1.67"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "105.85",
    cy: "26.48",
    rx: "1.66",
    ry: "1.67"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M89.55,115.22c0,1.48-14.2,2.68-31.71,2.68s-31.72-1.2-31.72-2.68c0-.74,4.17-.94,9.87-1.76.62-.09,1.77,3.42,3.92,3.31,1.46-.08,3.16.5,6.65-.41,1.65-.42,2.84-3.9,3.62-3.92,3-.09,4.29.09,7.66.09,1.62,0,2.16.11,3.72.08.72,0,1.51,3.2,2.85,3.71A13.22,13.22,0,0,0,70,117.84c2.1-.06,4.33-2.44,5.4-2.77.73-.23.39-2,1.2-2C85.53,113.54,89.55,114.26,89.55,115.22Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M62.22,55.67A19.5,19.5,0,0,1,47,37a19.29,19.29,0,0,1,38.57,0c0,8.5-6.15,15.78-13.81,18.33C71.74,55.33,68,56.91,62.22,55.67Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M61.36,93.5l.18,18.35c.39,3.88,3.66,5.91,7.56,5.91h0c4,0,7.25-2.1,7.58-6,1-11.45,2.64-35.43,2.16-59.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M48.32,64.39l-5,16.92-8,27.75c-1.07,4.67,2.48,8.19,7.28,8.19h.55a7.45,7.45,0,0,0,7-5L58,94.43l4.8-1.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.64,53.52c.2.06.69,10.47,3.64,15.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M94.16,53.13c-10,2.69-15.58-1.53-15.58-1.53"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M79.43,68.34c11.62,1.39,16.69-1.3,22.46-4.6l2.67-1.87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M47.5,48.75a18.37,18.37,0,0,1,.73,15.69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M53.08,50.9c.41,2,1.79,12.87-5.37,16.24"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M41.4,52.34,41.12,48a2.33,2.33,0,0,1,2.33-2.33l5.48-.1"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "55.04",
    cy: "37.71",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M83.6,29.23c-1.89-.47-4.06-.7-8.46-.38-7.66.55-10.82,3.57-12.08,5.57a7.22,7.22,0,0,0-.53,4c.53,3.87,1.71,13.33,18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "65.95",
    y1: "15.16",
    x2: "66.03",
    y2: "17.67"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "65.96",
    cy: "12.3",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "65.96",
    cy: "12.3",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M49.45,47.22s-14.1,6.56-21.31,28.39a5.73,5.73,0,0,0,4.32,7.31h0a5.16,5.16,0,0,0,6.25-3.15c1.54-4,4.91-11.42,9.39-15.29"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "68.31",
    y: "59.51",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M44.23,69.38v2.15A1.47,1.47,0,0,0,45.7,73h0"
  }))));
}
