// @postman/aether/esmLib/components/Illustration/illustrations/NoCollection.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoCollection() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-collection-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "113.42",
    y1: "60.44",
    x2: "120.65",
    y2: "60.44",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "84.18",
    y1: "42.94",
    x2: "120.65",
    y2: "42.94",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "72.34",
    y1: "69.85",
    x2: "98.18",
    y2: "69.85",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "120.58 43.41 113.42 49.73 113.42 77.46 120.65 68.13"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "94.81 39.9 84.18 45.98 112.8 45.98 120.65 39.9"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M93.48,114.5c0,2.26-13,4.09-29,4.09s-29-1.83-29-4.09c0-1.13,2.55-2.14,7.81-2.81,1.06-.14,1.68,4.27,3.65,4.1,1.34-.12,1.74,1.67,5.08.7,7.27-2.13,5.6-6.17,6.32-6.2,2.79-.12,3.1.12,6.18.12,1.49,0,1.32,0,2.75,0,.66,0,1.36,3.62,2.59,4.4A11.8,11.8,0,0,0,75.46,116c1.92-.09,2.46-1.86,3.44-2.36.67-.35,1.72-3.21,2.46-3.15C89.58,111.23,93.48,113,93.48,114.5Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M53.23,49.15l-.87-.72A19.7,19.7,0,1,1,85.09,33.67a16.59,16.59,0,0,1-.29,3A20.3,20.3,0,0,1,66.25,52.93c-.25,0-1-.14-1.3-.12"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "52.64",
    cy: "12.16",
    r: "2.79"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "52.64",
    cy: "12.16",
    r: "2.79"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "54.2",
    y1: "14.48",
    x2: "55.5",
    y2: "16.64"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-8",
    points: "72.46 73.83 83.92 77.92 88.2 77.42 90.89 73.03 90.69 68.65 85.81 65.46 75.45 60.88 72.47 58.92 72.34 55.22 98.18 55.22 98.18 84.47 72.34 84.47 72.34 74.17"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "72.34 58.98 72.34 55.22 98.18 55.22 98.18 84.47 72.34 84.47 72.34 74.17"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "113.24",
    y1: "68.34",
    x2: "109.37",
    y2: "68.35"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "94.81 39.9 120.65 39.9 120.65 68.13"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "94.81",
    y1: "49.13",
    x2: "94.81",
    y2: "45.95"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M74.22,50.5c-4.52-4.46-4.38-11.82-.47-17,2.49-3.3,6.54-4.57,10.77-4"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "60.22",
    cy: "33.01",
    r: "2.95"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "94.81 39.9 84.18 45.98 112.8 45.98 120.65 39.9"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "120.58 43.41 113.42 49.73 113.42 77.46 120.65 68.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M65.65,88.32l-1.87.86a10.29,10.29,0,0,0-6,9.41l.07,11a6.7,6.7,0,0,1-6.7,6.75H50.3a6.7,6.7,0,0,1-6.68-6.54l-.17-9.3c-.13-3.24-.15-9.54-.07-12.78.29-11.12,1.09-23.87,4.28-33.77l2.45-7.91"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M61.79,68.53S51.5,79.71,44.58,70.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M58.66,65s-8.22,8.19-13.43.65"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M43.67,74.48l-3.42-.35a4.28,4.28,0,0,1-3.85-4.31c.08-8.23,3.1-17.76,5.17-23.3A3.36,3.36,0,0,1,46,44.58l3.81,1.51"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-5",
    points: "83.79 54.91 83.79 49.53 108.88 49.53 108.88 78.48 98.28 78.48"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M84.2,39.3s10-6.9,19.45.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M77.62,84.5A188.6,188.6,0,0,0,81.31,107a7.43,7.43,0,0,1-7.22,9.2h0a7.43,7.43,0,0,1-7.35-6.38l-3-20.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M65,52.82A47.59,47.59,0,0,0,86.41,65.6a6.2,6.2,0,0,1,4.52,7.19h0c-.68,3.38-4.09,5.9-7.36,4.81-9-3-20.65-6.33-24.68-12.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M47.66,54s1.73-5.74,9.33-4.25"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "88.84",
    y1: "52.78",
    x2: "105.49",
    y2: "52.78"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "98.24",
    y1: "56.66",
    x2: "105.49",
    y2: "56.66"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "98.96",
    y1: "60.54",
    x2: "105.49",
    y2: "60.54"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "99.02",
    y1: "64.43",
    x2: "105.49",
    y2: "64.43"
  }))));
}
