// @postman/aether/esmLib/components/Illustration/illustrations/NoCookies.js
import React from 'react';
export default function NoCookies() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-cookies-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "80.95",
    y1: "37.68",
    x2: "87.74",
    y2: "37.68",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "75.94",
    y1: "53.35",
    x2: "90.97",
    y2: "53.35",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M81.48,36.75s3.83-2.29,6.26-2.25c0,0-6,3.22-6.79,6.35Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.94,48.55l3.55-3.87S75.89,57.18,91,62C91,62,78.19,63.22,75.94,48.55Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M117.78,114.54c0,1.74-20.35,3.14-45.46,3.14s-45.45-1.4-45.45-3.14c0-.72,3.86-2.1,8.38-2.18.7,0,1.52,3.17,3.72,3,1-.06,4.18.48,6.24-.4,1.8-.78,2.43-3,3.36-3,3.28-.15,3.63-.13,7.26-.23,1.21,0,.15,2.93,5.08,4a14.9,14.9,0,0,0,6.7-1.62c.9-.5.7-2.76,1.77-2.77,1.29,0,1.6,0,2.94,0,4,0,1.49,0,5.22.11,2.65,0,3.65,2.93,6.08,3,1.7,0,11.36,1.56,12.94,1.46,10.47-.69,12.33-3.34,14.4-3.24C113.88,112.9,117.78,113.72,117.78,114.54Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "62.68",
    cy: "34.84",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M54.72,89.77l.22,19.53A7.6,7.6,0,0,0,62.6,116h0A7.6,7.6,0,0,0,70.08,109c.79-11.49-1.51-31.47-2.35-55.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M35.38,59.57,34.41,79l-.2,27.83c-.14,5.7,2.62,9.1,7.42,9h.54a7.46,7.46,0,0,0,7-5.1l2-19.83,4.78-1.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M35.38,63.54c0-.38,6.88-8.35-.35-21.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M35.27,65.6S48,54.5,40.7,38.63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M68.87,65.78a73.6,73.6,0,0,0,4-14.58,4.66,4.66,0,0,0,3.8-.32"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M89,61.6c-5.21,2.87-10.07,5.26-20.44,4.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M33.06,43.86l-.1-4c0-2.16,1-3.91,2.33-3.91h8.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M43.63,37.29S27.14,39.7,27.24,64.06c0,5,.76,10.94,4.25,14.46a16.2,16.2,0,0,0,3,2.4"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-5",
    x: "58.27",
    y: "57.12",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "57.16",
    cy: "36.93",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "61.48",
    y1: "12.82",
    x2: "61.56",
    y2: "15.33"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "61.61",
    cy: "10.16",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "61.49",
    cy: "9.96",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M68.78,52.74c-3.87-5-2.72-12.31,1.88-16.91,2.91-2.93,7.1-3.63,11.21-2.48"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M106.16,114.88h0a48.31,48.31,0,0,1-22.68-.31h0c-4.48-1.68-7.11-3.18-7.11-7.11V82.39a9.79,9.79,0,0,1,5-8.52h0l0-2.88,0,0c1.25,3.17,27.35,3.86,28.24,0h0l.08,2.88.12.07a9.77,9.77,0,0,1,4.8,8.42v25.1c0,3.2-2.81,5.11-5.41,6.36A14.38,14.38,0,0,1,106.16,114.88Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M81.37,71a65.21,65.21,0,0,1,28.28,0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M102.88,95.37a7,7,0,1,1-7-7,6.88,6.88,0,0,1,3.11.73,7.07,7.07,0,0,1,2.72,2.38A7,7,0,0,1,102.88,95.37Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "95.12",
    cy: "98.44",
    r: "1.2"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "90.79",
    cy: "98.13",
    r: "1.2"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "98.51",
    cy: "92.7",
    r: "1.2"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "100.38",
    cy: "96.94",
    r: "0.84"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "94.15",
    cy: "89.91",
    r: "1.34"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "89.98",
    cy: "93.19",
    r: "0.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M109,106.42H82a1,1,0,0,0-.94-1V85.15a.94.94,0,0,0,.94-1h27a.94.94,0,0,0,.94,1v20.32A1,1,0,0,0,109,106.42Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M90,61.59a13.69,13.69,0,0,1-13.67-13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M82.42,36.49A13.68,13.68,0,0,1,90,34.2"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-10",
    cx: "93",
    cy: "47.94",
    rx: "13.69",
    ry: "13.96"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "93.17",
    cy: "48.12",
    r: "10.43"
  }))));
}
