// @postman/aether/esmLib/components/Illustration/illustrations/NoCustomDomain.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoCustomDomain() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-custom-domains-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "31.51",
    y1: "37.85",
    x2: "40.69",
    y2: "37.85",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "90.51",
    y1: "69.85",
    x2: "99.69",
    y2: "69.85",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M91,108.52c5,.46,7.77,2.29,7.77,4.18,0,2.92-12.76,5.29-28.5,5.29s-28.5-2.37-28.5-5.29c0-1.45,4.41-3.21,9.59-3.9.71-.1,2.4,6,4.36,5.79,1.32-.12,2.64.76,4.75.64,1.15-.07,4.79-1.65,5.33-6.74.12-1.14-.39-.65.91-.68s7.45,0,9,0c.92,0,1.57,6.16,4,6.69a30.74,30.74,0,0,0,5.25.91A8.72,8.72,0,0,0,90,112.29c1-1.47.68-3.72.91-3.7C92.39,108.7,89.42,108.38,91,108.52Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "73.55",
    y1: "15.49",
    x2: "73.06",
    y2: "17.96"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "74.27",
    cy: "12.93",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(44.99 82.4) rotate(-77.02)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "74.2",
    cy: "12.71",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.15 82.16) rotate(-77.02)"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "70.33",
    y: "59.89",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M50.35,47.61,48.48,50a23,23,0,0,0,4.41,32.07l1.08.82a6.36,6.36,0,0,0,3.87,1.31h0a6.39,6.39,0,0,0,4.7-10.71c-3-3.29-6-7.62-9.37-10.21"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "65.92",
    cy: "35.84",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M50.38,80.49s-7.49,16.41,2.74,32a6.71,6.71,0,0,0,5.63,3h0a6.83,6.83,0,0,0,6.34-9.34c-1.57-4-2.91-11.82-.19-14.85h5.26l1.45,4.32,4.68,13.93c1.35,3.81,3.24,6,6.64,6h1a6.79,6.79,0,0,0,6.82-7.63L79.92,53.42l-.66-2.9"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71,54.36c3.25-2.33,5-5.26,5.14-8.49.19-6,0-11-14.65-11.42-10.1-.31-13.15,2-14.33,5.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M81.71,47.91s1.8,4.57-.24,11.68"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M48,51.06s0,8.67,5.3,12.21c0,0-2.85-8.24-1-13.47"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "37.45 36.49 40.7 37.84 37.46 39.2 36.11 42.44 34.75 39.2 31.51 37.85 34.75 36.5 36.09 33.25 37.45 36.49"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "37.45 36.49 40.7 37.84 37.46 39.2 36.11 42.44 34.75 39.2 31.51 37.85 34.75 36.5 36.09 33.25 37.45 36.49"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "96.45 68.49 99.69 69.84 96.46 71.2 95.11 74.44 93.75 71.2 90.51 69.85 93.75 68.5 95.09 65.25 96.45 68.49"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "96.45 68.49 99.69 69.84 96.46 71.2 95.11 74.44 93.75 71.2 90.51 69.85 93.75 68.5 95.09 65.25 96.45 68.49"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "30.35 57.03 31.35 57.58 30.35 58.14 29.8 59.16 29.26 58.14 28.26 57.58 29.26 57.03 29.8 56.01 30.35 57.03"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "105.09 79.1 106.09 79.66 105.09 80.21 104.55 81.23 104 80.21 103 79.66 104 79.1 104.55 78.08 105.09 79.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M85.34,35.49c1-.8,3.14-2.63,4.24-3.4,2.43-1.7,9-5.78,11.89-2.75h0c2.52,2.35,3.56,6.6,0,9.69C94,45.59,86.3,53.14,86.3,53.14a45,45,0,0,0-4.86,8.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M46.78,32.85l-2.16,1a1.45,1.45,0,0,0-.81,1.33v2.68a1.45,1.45,0,0,0,1,1.4l2.06.55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M84.33,42.1l2.29-.65a1.46,1.46,0,0,0,1-1.19L88,37.6a1.43,1.43,0,0,0-.75-1.52l-2-.86"
  }))));
}
