// @postman/aether/esmLib/components/Illustration/illustrations/NoEnvironment.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoEnvironment() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-environment-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "85.15",
    y1: "13.79",
    x2: "92.06",
    y2: "13.79",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "39.4",
    y1: "66.05",
    x2: "42.41",
    y2: "66.05",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "33.73",
    y1: "66.36",
    x2: "37.25",
    y2: "66.36",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M87.89,48.75a8.21,8.21,0,0,0,7.86-12.68A8.2,8.2,0,0,1,90,46.57"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "88.61",
    cy: "13.79",
    r: "3.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M85.28,113.61c0,2.92-13.1,5.29-29.25,5.29s-29.25-2.37-29.25-5.29c0-1.46,3.94-2.77,9.2-3.82.84-.17,1,4.86,2.52,6.12,5.3,4.35,7,1.61,10.24-.17,1.51-.84,1.87-7.28,2.59-7.32a44.22,44.22,0,0,1,4.7-.1c1.5,0,.79.1,2.23.05.67,0,.75,5.65,2,6.66.88.72,4.34,3,6.65,2.86,1.94-.11,4.24-3.41,5.22-4.06.68-.45.69-4.71,1.44-4.62C81.83,110.14,85.28,111.72,85.28,113.61Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M36.21,82.72v28a7.37,7.37,0,0,0,7.36,7.36h0a7.37,7.37,0,0,0,7.36-7.36l.77-19.12c0-1.54,1.7-1.43,2.5-1.43h0a1.52,1.52,0,0,1,1.51,1.45L58.48,110c.57,4.73,3.45,7.78,7.74,7.83h0c4.36.05,7-4.39,7-8.75L70.47,61.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M36.85,47.91,36,47.2a19.28,19.28,0,1,1,21,3l-.61.31"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "26.06",
    y: "48.12",
    width: "31.68",
    height: "34.12",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "36.27",
    cy: "11.68",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "36.27",
    cy: "11.68",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M66.79,38.65c-6.17-.73-10.43-5.94-10.43-12.31a11.58,11.58,0,0,1,4-9.08"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M70.42,61.76s-3.53,10.5-12.61,8.45"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M70.74,69.33s-3.85,8.21-12.93,6.16"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "37.79",
    y1: "13.96",
    x2: "39.06",
    y2: "16.07"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M48.91,48.49v7.4a5,5,0,0,1-5,5H26.39"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "40.9",
    cy: "66.05",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "35.49",
    cy: "66.36",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "40.9",
    cy: "66.14",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "35.55",
    cy: "66.14",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M26.3,64.45h4.06a1.66,1.66,0,0,1,1.39,1.83h0a1.67,1.67,0,0,1-1.39,1.84H26.3Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M26.3,75.12H48.36a1.67,1.67,0,0,0,1.39-1.84h0a1.66,1.66,0,0,0-1.39-1.83H26.3"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "49.55",
    cy: "36.03",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M62.11,47.4s12.2-1.08,19.62-11.84a6.43,6.43,0,0,1,9.15-1.45h0a6.19,6.19,0,0,1,1.87,7.68C90.33,46.62,84.34,56,74.22,60.51a21.65,21.65,0,0,1-2.67.93l-1.07.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72.82,43.2a15.77,15.77,0,0,1-.56-4.2A16.06,16.06,0,0,1,85.67,23.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M90.76,23.13a16.05,16.05,0,1,1-7.27,31.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M79.83,37.36A8.21,8.21,0,1,1,90,46.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M86,30.26l1.14-8.54a.6.6,0,0,1,.61-.51l1,0a.59.59,0,0,1,.59.52l.88,8.56"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "88.61",
    cy: "13.79",
    r: "3.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M110.44,35.69a2.76,2.76,0,0,1,5,.28,1.9,1.9,0,0,1,.63-.11,2,2,0,0,1,0,3.91h-5.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.19,64.14h5.27a2,2,0,0,0,2-1.63,2,2,0,0,0-1.93-2.28,1.9,1.9,0,0,0-.63.11,2.76,2.76,0,0,0-5-.28,2,2,0,0,0-2.27,2.41,1.67,1.67,0,0,0,1.59,1.66Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M87.19,64.14h5.27a2,2,0,0,0,2-1.63,2,2,0,0,0-1.93-2.28,1.9,1.9,0,0,0-.63.11,2.76,2.76,0,0,0-5-.28,2,2,0,0,0-2.27,2.41,1.67,1.67,0,0,0,1.59,1.66Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "109.39",
    cy: "37.91",
    r: "2.12"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "86.83",
    y1: "65.9",
    x2: "86.83",
    y2: "67.07"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "89.12",
    y1: "65.9",
    x2: "89.12",
    y2: "67.07"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "91.83",
    y1: "65.9",
    x2: "91.83",
    y2: "67.07"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "92.44",
    y1: "13.53",
    x2: "93.9",
    y2: "13.53"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "83.69",
    y1: "13.53",
    x2: "85.16",
    y2: "13.53"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "88.51",
    y1: "8.38",
    x2: "88.51",
    y2: "9.84"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "88.2",
    y1: "17.56",
    x2: "88.2",
    y2: "19.02"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "84.48",
    y1: "17.35",
    x2: "85.51",
    y2: "16.32"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "91.19",
    y1: "11.08",
    x2: "92.23",
    y2: "10.05"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "84.7",
    y1: "9.83",
    x2: "85.74",
    y2: "10.86"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "90.97",
    y1: "16.54",
    x2: "92.01",
    y2: "17.58"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "98.29",
    y1: "51.78",
    x2: "99.32",
    y2: "53.4"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "102.76",
    y1: "46.48",
    x2: "104.63",
    y2: "47.39"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "79.75",
    y1: "24.64",
    x2: "78.89",
    y2: "22.92"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "74.75",
    y1: "29.44",
    x2: "72.99",
    y2: "28.35"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "101.75",
    y1: "29.85",
    x2: "103.46",
    y2: "28.97"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "96.88",
    y1: "24.92",
    x2: "97.95",
    y2: "23.14"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "67.95 35.08 69.71 35.08 69.71 39.75 66.88 39.9"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M29.73,81.86s1.07,7.68,6,8"
  }))));
}
