// @postman/aether/esmLib/components/Illustration/illustrations/NoExportRequest.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoExportRequest() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-export-request-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "91.32",
    y1: "77.68",
    x2: "111.56",
    y2: "77.68",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M91.32,69.12V86.24h20.24V69.12Zm18,8h-7.47V72.22h7.47Z"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "102.78",
    cy: "114.53",
    rx: "18.49",
    ry: "1.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M78.1,113.17c0,1.93-14.65,3.5-32.71,3.5s-32.71-1.57-32.71-3.5c0-1.08,6.3-2.14,13.45-2.78,2.85-.26.81,1.43,4.35,1.29,1.63-.06,12.51-.69,14.22-.35.58.12,1.59,3.76,3.76,3.76,1.52,0,2.76,0,4.22,0,1.16,0,4.59-4.94,4.82-4.94.94,0,.88,3.65,3.06,3.76,1.79.1,2.41.93,4,1.06s3.94-4.29,5.33-4.12C75,111.47,78.1,112.28,78.1,113.17Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M21,91A49.53,49.53,0,0,1,30.74,79.8c5.67-6.34,10.55-3.11,11.49-1.07a4.43,4.43,0,0,1-.61,4.5,138.84,138.84,0,0,0-8.13,12.26,14.86,14.86,0,0,1-7.15,6.25c-4.33,1.75-6.61.69-8.65-.8-3.11-2-12-13.89.36-33,1.6-2.48,4.6-4.76,6.88-7.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M49.65,62.38c2.16,1.31,6.42,2.67,8.36,9.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M55.89,80.07c3.55-1.67,10.61-4.87,12.74,3.35,1.69,6.54,1.41,18.38.75,26a6,6,0,0,1-6,5.5h0a6,6,0,0,1-6-6v-2.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M42.59,80.83s11.51-7.76,14.19,2.59c1.78,6.83,1.12,18.77.39,26.08a6.56,6.56,0,0,1-6.53,5.9h0a6.57,6.57,0,0,1-6.57-6.51L43.91,89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M21.54,64s5.67,3.09,7.51,16.67"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M27.43,61.48a25.9,25.9,0,0,1,4.29,16.78"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M48,63.82S50.07,67,49.24,77.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M52.9,77.92a28.29,28.29,0,0,0-.54-13.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M22.13,103.11h0a8.57,8.57,0,0,0,8.57,8.57l13.68-.31"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M14.4,74.28s-.13-6.68-.2-8.26V62.71a4.91,4.91,0,0,1,4.9-4.91h4.14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M53.94,78.82s-1.55-6.59,6.79-7.17S67.32,80,67.32,80"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M48.46,62.33c-8.06-.33-14.33-7-14.09-12.89s.77-10.64,15-10.07a57.18,57.18,0,0,1,7.71.86"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "27.27",
    cy: "23.68",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "27.27",
    cy: "23.68",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "28.46",
    y1: "26.2",
    x2: "29.59",
    y2: "28.7"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "75.88",
    y: "101.3",
    width: "54.05",
    height: "4.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M48.72,74.38H44.66a.44.44,0,0,1-.44-.43V69.74a.44.44,0,0,1,.44-.43h4.06"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "85.6",
    cy: "110.49",
    r: "4"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "120.6",
    cy: "110.49",
    r: "4"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "120.6",
    cy: "110.49",
    r: "1.99"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "85.6",
    cy: "110.49",
    r: "1.99"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "84.35 101.24 84.35 86.67 101.56 86.67 101.56 101.24"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "104.35 101.24 104.35 86.67 121.56 86.67 121.56 101.24"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "91.32 86.24 91.32 69.11 111.56 69.11 111.56 86.24"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "101.56",
    y1: "86.67",
    x2: "104.35",
    y2: "86.67"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "91.84",
    y: "89.22",
    width: "7.47",
    height: "4.89"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "111.84",
    y: "89.22",
    width: "7.47",
    height: "4.89"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "101.84",
    y: "72.22",
    width: "7.47",
    height: "4.89"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "38.76",
    cy: "45.76",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "27.1",
    cy: "44.47",
    r: "2.88"
  }))));
}
