// @postman/aether/esmLib/components/Illustration/illustrations/NoFork.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoFork() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-fork-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "31.13",
    y1: "45.35",
    x2: "34.65",
    y2: "45.35",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "29.02",
    y1: "22.94",
    x2: "36.75",
    y2: "22.94",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "61.59",
    y1: "62.25",
    x2: "84.53",
    y2: "62.25",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M34.1,47.61c3.87,4,9.42,7.82,18.75,11.22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M29.91,42.53a9.47,9.47,0,0,0,1.25,1.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M56.69,44.85c-4.24-1.46-11-4.32-15.52-9.94a36.84,36.84,0,0,0-2.76-2.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M103.17,111.52c0,2.92-13.1,5.29-29.25,5.29s-29.25-2.37-29.25-5.29c0-1.46,2.76-2.82,8.06-3.68,1.07-.18.89,3.8,2.43,5.07,5.31,4.35,8.22,1.51,11.44-.27,1.52-.84,1.34-6.33,2.06-6.37,2.81-.16,2.15,0,5.26,0,1.5,0-.41.13,1,.08.67,0,0,5.38,3.76,6.49,1.09.32,3.66,1.86,6,1.73,1.94-.11,3.21-2.42,4.19-3.08.67-.44,1.32-4.41,2.07-4.33C99.22,108.06,103.17,109.63,103.17,111.52Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "70.98",
    cy: "31.48",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M52.89,58.69l.18,25,.32,23.74a7.35,7.35,0,0,0,7.36,7.35h0a7.35,7.35,0,0,0,7.35-7.35l1.18-15.83c.07-1.62,1-3.29,2.58-3.28h0c1.61,0,2.54,1.61,2.62,3.22l1.17,15.19a7.82,7.82,0,0,0,7.74,7.83h0c4.36,0,7.14-2.35,7-8.74l-.63-28.32"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M53,58.68S57,55.8,57.18,45.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M61.61,48.5s1,8.38-8.37,12.69"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M82.1,47.5S83.34,59.42,89,64.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M92.23,49.57l0-6.19a2.32,2.32,0,0,0-2.36-2.3L87.81,41"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M81.78,31.77c.19,6.05-6.32,12.85-14.6,13.1S52,38.75,51.79,32.71s0-11,14.65-11.42C80.43,20.86,81.59,25.73,81.78,31.77Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M75,60.15V55.94a.43.43,0,0,1,.43-.43h4.07a.43.43,0,0,1,.43.43v4.21"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "79.56",
    y1: "11.31",
    x2: "78.8",
    y2: "13.71"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "80.53",
    cy: "8.62",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.54 81.66) rotate(-70.52)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "80.53",
    cy: "8.62",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.54 81.66) rotate(-70.52)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M90.21,28.41l2.71,1a1.47,1.47,0,0,1,.81,1.34v2.68a1.43,1.43,0,0,1-1,1.39l-2.8.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M51.58,27.4l-2.64,1.21a1.45,1.45,0,0,0-.81,1.33v2.68a1.43,1.43,0,0,0,1,1.39l2.8.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M31.24,42.82l-.11,3.31a1.76,1.76,0,0,0,1.7,1.76h0a1.77,1.77,0,0,0,1.82-1.75l-.18-3.09"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M31.24,42.82l-.11,3.31a1.76,1.76,0,0,0,1.7,1.76h0a1.77,1.77,0,0,0,1.82-1.75l-.18-3.09"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M34.12,29.12l.05-1.26h0c1.42,0,2.67-.34,2.57-2.57l-1.12-8.83-.85,7.13H34l-1.12-7-1.21,6.94h-.8l-.83-7-1,8.37c-.11,2.49,1.3,2.92,2.92,2.92h0l-.05,1.56"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M34.12,29.12l.05-1.26h0c1.42,0,2.67-.34,2.57-2.57l-1.12-8.83-.85,7.13H34l-1.12-7-1.21,6.94h-.8l-.83-7-1,8.37c-.11,2.49,1.3,2.92,2.92,2.92h0l-.05,1.56"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M84.66,45.61s8.93,1.52,16.35,16.6C106.79,74,95.39,77.72,86.19,76.8c-3.57-.36-8-3-7.59-6.62h0a6.53,6.53,0,0,1,6.49-5.82h9.46"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M90.7,64.26a29.15,29.15,0,0,1-3.24-17.42"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "32.98",
    cy: "36.29",
    r: "6.77"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "73.2",
    cy: "61.82",
    rx: "13.58",
    ry: "1.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M80.05,73.82A13.68,13.68,0,0,1,59.55,62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M86.91,62a9.14,9.14,0,0,1-.16,2.17"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "70.17 67.51 67.29 69.09 70.09 70.43"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "74.29 70.43 77.17 68.84 74.38 67.51"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "72.67",
    y1: "66.81",
    x2: "71.65",
    y2: "71.22"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-8",
    cx: "73.06",
    cy: "62.25",
    rx: "11.47",
    ry: "0.73"
  }))));
}
