// @postman/aether/esmLib/components/Illustration/illustrations/NoHistory.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoHistory() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-history-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "109.67",
    y1: "100.08",
    x2: "116.19",
    y2: "100.08",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "87.16",
    y1: "100.17",
    x2: "93.67",
    y2: "100.17",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "722.95",
    y1: "-1532.9",
    x2: "730.74",
    y2: "-1532.9",
    gradientTransform: "matrix(0.82, 0.57, -0.57, 0.82, -1410.27, 925.43)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-4",
    x1: "109.11",
    y1: "96.45",
    x2: "110.6",
    y2: "96.45",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-5",
    x1: "93.41",
    y1: "87.65",
    x2: "110",
    y2: "87.65",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "90.7 88.96 82.2 88.96 82.2 54.97 120.72 54.97 120.72 88.96 112.77 88.96"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M109.67,91.71s9,2.15,5.83,16.74l-5.64-4.51"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M93.3,91.61s-9,2.68-5.21,17.12l5.58-4.67"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M109.67,91.71s9,2.15,5.83,16.74l-5.64-4.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M52.31,76.45l-.46,1.4c.36.65.59.65,1.32.75a10,10,0,0,0,9.13-3.94c.33-.42.42-.53.45-.91a5.64,5.64,0,0,0-.61-2.93l-.49-1s1.73,4.3-4.44,6.08S52.31,76.45,52.31,76.45Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M78.84,115.2c0,2-12.69,3.67-28.35,3.67s-28.36-1.64-28.36-3.67c0-1.07,3.56-2,9.23-2.71.94-.11.82.57,1.1,1.6s1.25,2.32,2.4,2.24.2,2.59,6.86.67A6.84,6.84,0,0,0,45,114.07c.55-1.21.14-2.47.78-2.48,1.53,0,3.1,0,4.7,0,2.28,0,1,0,3.1.1.8,0,.53,2.34,1.79,3.74a9.92,9.92,0,0,0,5.43,2.55c1.63.1,4-1.4,5-1.8.52-.21,2.1-3.82,2.63-3.77C74.79,113,78.84,114.06,78.84,115.2Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M27,55.14l.16-8a2.3,2.3,0,0,1,2.29-2.3l4.46-.09"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "49.68",
    y1: "13.49",
    x2: "49.76",
    y2: "15.64"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "49.81",
    cy: "10.83",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "49.69",
    cy: "10.63",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M66,79.27l.88,13.88.32,3.8.55,13.39a6.81,6.81,0,0,1-6.79,7h0a6.8,6.8,0,0,1-6.78-6.53L52.11,94a2.63,2.63,0,0,0-2.54-2.24h0a2.62,2.62,0,0,0-2.64,2.33l-1.55,16.49a6.81,6.81,0,0,1-6.8,6.83h0a6.81,6.81,0,0,1-6.79-6.83L32.7,80"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M47.76,81.85c1,2.56,2.72,2.81,5.31,2.81a17.6,17.6,0,0,0,4.1-.93C71,78,74.72,72.84,75.35,65.82c.83-9.14-5.68-16-9.78-19.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M66,59.62c-1.26,3.48-4.59,7.33-8.21,9.7"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M51.64,72.62c.22-1.41,1.37-2.09,3.57-2.93,2.76-1,5.29-1,5.9.57s-.9,3.93-3.65,5c-1.88.72-3.41,1.44-4.6.87"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M52.31,76.3a8.12,8.12,0,0,0,4.86-.58c3.06-1.24,5-3.52,4.4-5.09s-3.63-1.82-6.68-.58a8.82,8.82,0,0,0-3.47,2.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M35.24,46.39l-2.68,1.73a28.66,28.66,0,0,0-8.83,20.11c-.06,2.76,2.11,12,19.69,14,.39,0,1.19,0,1.33,0,3.7,0,7.5-1.91,7.53-6.2A6.77,6.77,0,0,0,52,73.78c-.77-2.34-2.75-3.69-6.95-4.11-5.46-.55-8.76-4.2-10.54-4.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M56.4,76.09s3.07,3.88,3.36,6.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M59.5,74.66s2.93,3.67,3.22,6.21"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M31.36,49.24a24,24,0,0,1,4,15.53"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M34.69,46.66s5.41,7.21,4.16,20.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M65,61.24S63.29,52.42,66.34,47"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M62.58,64.92a20.29,20.29,0,0,1-1.42-14.41"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "50.27",
    cy: "34.81",
    rx: "18.93",
    ry: "19.02"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-9",
    d: "M109.35,94.59s0,.07.12.07.12,0,.12-.07,0-.11-.13-.11S109.23,94.52,109.35,94.59Zm.86-4.11c.06-.24.19-.44.12-.2,0,0,0,.12-.09.22S110.12,90.7,110.21,90.48Zm.38,12s0,.12-.08.24h-.05l0-.11Zm-1.3-2.46s-.12,0-.12,0,0-.07.12,0S109.41,100,109.29,100Zm.08-.69s-.11,0-.11,0-.12,0-.13-.07,0,0,.12,0S109.37,99.3,109.37,99.34Zm.17-1.24h-.12s.11-.11.12-.07S109.54,98.06,109.54,98.1Zm-.2-3.76c-.11.07-.23.08-.23,0s.11-.11.23-.11Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-10",
    d: "M109.78,91.56,109.67,85a16,16,0,0,0-6.4-12.62c-.6-.43-2-1.27-2-1.27h0c-.33.19-.65.39-1,.58a15.72,15.72,0,0,0-6.93,13.5l.12,6.89.21,12,8.13,0,8.13.08Zm-8,0A5.55,5.55,0,0,1,96,86.26a5.5,5.5,0,1,1,5.72,5.32Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M109.35,94.59s0,.07.12.07.12,0,.12-.07,0-.11-.13-.11S109.23,94.52,109.35,94.59Zm.86-4.11c.06-.24.19-.44.12-.2,0,0,0,.12-.09.22S110.12,90.7,110.21,90.48Zm.38,12s0,.12-.08.24h-.05l0-.11Zm-1.3-2.46s-.12,0-.12,0,0-.07.12,0S109.41,100,109.29,100Zm.08-.69s-.11,0-.11,0-.12,0-.13-.07,0,0,.12,0S109.37,99.3,109.37,99.34Zm.17-1.24h-.12s.11-.11.12-.07S109.54,98.06,109.54,98.1Zm-.2-3.76c-.11.07-.23.08-.23,0s.11-.11.23-.11Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M101.26,71.19c-.34.19-.65.38-1,.58a15.68,15.68,0,0,0-6.92,13.49l.12,6.9.2,12h8.13l8.13.07-.22-12.58-.11-6.53a15.91,15.91,0,0,0-6.4-12.61c-.6-.44-2-1.28-2-1.27Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M93.3,91.61s-9,2.68-5.21,17.12l5.58-4.67"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "101.48",
    cy: "86.11",
    rx: "5.49",
    ry: "5.52",
    transform: "translate(-31.34 77.6) rotate(-36.65)"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "53.44",
    y1: "74.03",
    x2: "57.46",
    y2: "72.05"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "52.86",
    y: "56.87",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "106.25 104.11 107.06 107.76 96.56 107.76 97.37 104.11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M65.53,37.16c.19,6-6.32,12.84-14.6,13.1s-15.2-6.12-15.39-12.17,0-11,14.65-11.42C64.18,26.25,65.34,31.11,65.53,37.16Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M69.65,31.59l2.71,1a1.46,1.46,0,0,1,.81,1.34v2.68a1.43,1.43,0,0,1-1,1.39l-2.79.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M31.17,32.14l-2.64,1.21a1.46,1.46,0,0,0-.81,1.33v2.69a1.43,1.43,0,0,0,1,1.39l2.79.84"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-6",
    cx: "101.2",
    cy: "115.73",
    rx: "11.88",
    ry: "2.25"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "115.8",
    y1: "59.91",
    x2: "89.91",
    y2: "59.91"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "108.69",
    y1: "78.91",
    x2: "115.8",
    y2: "78.91"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "89.91",
    y1: "78.91",
    x2: "94.47",
    y2: "78.91"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "106.19",
    y1: "75.11",
    x2: "115.8",
    y2: "75.11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "95.2",
    y1: "75.11",
    x2: "96.59",
    y2: "75.11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "95.2",
    y1: "63.71",
    x2: "115.8",
    y2: "63.71"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "105.65",
    y1: "67.51",
    x2: "115.8",
    y2: "67.51"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "105.74",
    y1: "71.31",
    x2: "115.8",
    y2: "71.31"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "90.93",
    y: "87.22",
    width: "2.5",
    height: "3.43"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "109.93",
    y: "87.22",
    width: "2.5",
    height: "3.43"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-11",
    x1: "90",
    y1: "61.46",
    x2: "90",
    y2: "77.32"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "99.65 85.19 97.57 86.33 99.59 87.3"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "103.34 87.3 105.42 86.16 103.4 85.19"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "102",
    y1: "84.17",
    x2: "100.95",
    y2: "88.18"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "56.84 70.91 58.3 73.36 59.78 70.98 56.84 70.91"
  }))));
}
