// @postman/aether/esmLib/components/Illustration/illustrations/NoIntegrations.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoIntegrations() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-integrations-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "93.02",
    y1: "47.99",
    x2: "101.47",
    y2: "47.99",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M83.08,114.06c0,1.48-14.2,2.68-31.72,2.68s-31.71-1.2-31.71-2.68c0-.74,4.16-.94,9.86-1.76.63-.09,1.78,3.41,3.93,3.3,1.46-.07,3.16.51,6.65-.4,1.64-.42,2.83-3.9,3.61-3.93,3-.08,4.29.1,7.66.1,1.63,0,2.16.1,3.72.08.73,0,1.52,3.2,2.86,3.71,1,.36,3,1.58,5.53,1.51,2.11,0,4.34-2.43,5.4-2.76.74-.23.4-2,1.21-2C79.05,112.37,83.08,113.1,83.08,114.06Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M55.75,54.51c-8.4-2.08-15.24-9.65-15.24-18.67a19.29,19.29,0,0,1,38.57,0c0,8.5-6.16,15.78-13.81,18.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M55.6,91.9l-.54,18.79c.39,3.88,3.66,5.91,7.57,5.91h0c3.95,0,7.25-2.1,7.58-6,1-11.45,2.63-35.44,2.15-59.52"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M37.33,77.72l-.49,2.43-8,27.75c-1.07,4.66,2.48,8.19,7.28,8.19h.54a7.48,7.48,0,0,0,7.05-5l7.77-17.84,4.81-1.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M69.17,52.36c.2.06.15,10.51,3.11,15.14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M100.38,50.71a7,7,0,0,1,.88,1.93h0a7.88,7.88,0,0,1-1.51,6.45,11,11,0,0,1-1.67,1.62l-2.66,1.87c-5.77,3.29-11.22,6.35-22.84,5"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71.73,50.69s8.42,5.8,21.88-.94"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.67,65.71,55.94,54.6c3-2.21,7.19-2.86,9.46.06h0a9.46,9.46,0,0,1,1,3.71c.13,2.7-.91,4.68-3.4,6.43L48,75.62a11.75,11.75,0,0,1-5.54,2.29,10.1,10.1,0,0,1-8.59-3.11c-4.51-5.12-5.61-13.48,9.78-28.75"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M42,48.44s3.12,7.48.51,14.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M47,49.88c.41,2,1.25,7.6.36,9.95"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M97.25,51.69h0a4.3,4.3,0,0,1-1.65-.33A4.21,4.21,0,0,1,93,47.47V44.28h8.45v3.19A4.2,4.2,0,0,1,99.4,51.1,4.25,4.25,0,0,1,97.25,51.69Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M97.18,51.77h0A4.23,4.23,0,0,1,93,47.55V44.36h8.46v3.19a4.21,4.21,0,0,1-2.07,3.63A4.29,4.29,0,0,1,97.18,51.77Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M97,51.78V98A12.08,12.08,0,0,0,109.07,110h8.81"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "100.25 44 100.25 39.23 99.98 39.23 99.36 39.23 98.83 39.23 98.36 39.23 97.92 39.23 97.92 44"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "96.25 44 96.25 39.23 95.98 39.23 95.36 39.23 94.83 39.23 94.36 39.23 93.92 39.23 93.92 44"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M34.92,56.47l-.28-9.68A2.33,2.33,0,0,1,37,44.47l5.85-.1"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "48.56",
    cy: "36.55",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M77.16,28c-1.9-.47-4.1-.6-8.5-.28-7.66.55-10.81,3.57-12.08,5.57a7.3,7.3,0,0,0-.52,4c.52,3.86,1.7,13.33,18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "59.48",
    y1: "14",
    x2: "59.56",
    y2: "16.51"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-6",
    cx: "59.49",
    cy: "11.14",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "59.49",
    cy: "11.14",
    rx: "2.68",
    ry: "2.7"
  }))));
}
