// @postman/aether/esmLib/components/Illustration/illustrations/NoMonitor.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoMonitor() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-monitor-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "95.26",
    y1: "28.28",
    x2: "107.08",
    y2: "28.28",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "40.94",
    y1: "66.25",
    x2: "43.96",
    y2: "66.25",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "35.28",
    y1: "66.55",
    x2: "38.8",
    y2: "66.55",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M86.82,113.8c0,2.92-13.09,5.29-29.25,5.29s-29.24-2.37-29.24-5.29c0-1.45,3.94-2.77,9.2-3.82.84-.17,1,4.86,2.52,6.12,5.3,4.36,7,1.62,10.23-.16,1.52-.84,1.88-7.28,2.6-7.33a44.25,44.25,0,0,1,4.69-.09c1.51,0,.8.09,2.24,0,.67,0,.75,5.64,2,6.65.88.72,4.34,3,6.65,2.87,1.94-.12,4.24-3.42,5.22-4.07.68-.45.69-4.71,1.43-4.62C83.38,110.33,86.82,111.91,86.82,113.8Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "95.26",
    y: "17.93",
    width: "11.82",
    height: "20.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M37.76,82.91v28a7.36,7.36,0,0,0,7.36,7.36h0a7.35,7.35,0,0,0,7.35-7.36l.77-19.12c0-1.54,1.71-1.42,2.51-1.42h0a1.51,1.51,0,0,1,1.51,1.44L60,110.21c.57,4.73,3.45,7.78,7.74,7.83h0c4.35,0,7-4.39,7-8.75L72,62c12-8.66,15.84-16.65,17.47-23.83s-6-9.14-6-9.14c.4-9.88-3.66-7.65-3.66-7.65-3.65,26.07-14.2,23.36-14.2,23.36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M38.39,48.1l-.85-.71a19.28,19.28,0,1,1,21,3l-.62.32"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "27.61",
    y: "48.32",
    width: "31.68",
    height: "34.12",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M32,82.21s1.07,7.67,6,8"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "37.81",
    cy: "11.88",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "37.81",
    cy: "11.88",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M68.34,38.85c-6.18-.73-10.43-6-10.43-12.32a11.58,11.58,0,0,1,4-9.08"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "58.14 15.08 58.14 1.55 112.39 1.55 112.39 42.81 88.34 42.81"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "95.24",
    y: "17.99",
    width: "12.21",
    height: "20.38"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-4",
    points: "58.32 12.92 64.61 12.92 65.83 12.92 69 9.92 72.73 12.92 82.07 12.92 85.55 8.21 90 12.92 93.87 9.88 97.08 12.92 98.3 12.92 105.2 12.92 106.42 12.92 111.98 12.92"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "81.45 14.69 79.45 17.17 83.4 17.17 81.45 14.69"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "93.94 6.58 95.94 4.1 91.99 4.1 93.94 6.58"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "97.28",
    y1: "21.99",
    x2: "105.31",
    y2: "21.99"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "97.28",
    y1: "25.42",
    x2: "105.31",
    y2: "25.42"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "97.28",
    y1: "28.85",
    x2: "105.31",
    y2: "28.85"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "97.28",
    y1: "32.28",
    x2: "105.31",
    y2: "32.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72,62s-3.52,10.5-12.6,8.45"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72.29,69.52s-3.85,8.21-12.93,6.17"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "39.34",
    y1: "14.15",
    x2: "40.61",
    y2: "16.27"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M50.46,48.68v7.41a5,5,0,0,1-5,5H27.94"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "42.45",
    cy: "66.25",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-9",
    cx: "37.04",
    cy: "66.55",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "42.45",
    cy: "66.34",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "37.09",
    cy: "66.34",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M27.85,64.64h4.06c.76,0,1.38.82,1.38,1.84h0c0,1-.62,1.83-1.38,1.83H27.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M27.85,71.64H49.91c.76,0,1.38.82,1.38,1.84h0c0,1-.62,1.83-1.38,1.83H27.85"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "51.1",
    cy: "36.22",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "71.86",
    y1: "42.17",
    x2: "67.54",
    y2: "42.17"
  }))));
}
