// @postman/aether/esmLib/components/Illustration/illustrations/NoMockServer.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoMockServer() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-mock-server-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "92.06",
    y1: "57.65",
    x2: "123.19",
    y2: "57.65",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M92.06,48.15l.24,19h30.89v-19Zm7.33,13.77a2.44,2.44,0,1,1,0-3.45A2.44,2.44,0,0,1,99.39,61.92Zm11,.5h-2.85V52.87h2.85Zm4.87,0h-2.85V52.87h2.85Zm4.86,0h-2.85V52.87h2.85Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "51.44",
    y: "60.79",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "100.28",
    cy: "114.28",
    rx: "19.6",
    ry: "2.19"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "123.19 67.15 92.3 67.15 92.06 48.15 123.19 48.15 123.19 67.15"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "107.56",
    y: "52.87",
    width: "2.85",
    height: "9.55"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "112.43",
    y: "52.87",
    width: "2.85",
    height: "9.55"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "117.29",
    y: "52.87",
    width: "2.85",
    height: "9.55"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "97.67",
    cy: "60.19",
    r: "2.44"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M73.65,115.91c0,1.49-14.2,2.69-31.72,2.69s-31.71-1.2-31.71-2.69c0-.74,4.16-.94,9.86-1.75.63-.09,1.78,3.41,3.93,3.3,1.46-.08,3.16.5,6.65-.4,1.64-.43,2.84-3.91,3.62-3.93,3-.08,4.29.1,7.65.1,1.63,0,2.16.1,3.73.08.72,0,1.51,3.19,2.85,3.7a12.92,12.92,0,0,0,5.54,1.52c2.1-.06,4.33-2.43,5.4-2.77.73-.22.39-2.05,1.2-2C69.62,114.23,73.65,115,73.65,115.91Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M55.84,56c7.54-2.87,13.81-9.84,13.81-18.33a19.29,19.29,0,0,0-38.57,0,18.5,18.5,0,0,0,4.47,11.93c4.43,5.26,10.49,8,17.32,7.13A18.44,18.44,0,0,0,55.84,56Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M45.64,93.39v19.16c.4,3.87,3.67,5.9,7.57,5.9h0c3.95,0,7.25-2.09,7.58-6,1-11.46,2.63-35.44,2.15-59.52"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M32.55,64.84,27.41,82l-8,27.75c-1.07,4.66,2.48,8.19,7.28,8.19h.54a7.47,7.47,0,0,0,7.05-5l7.77-17.84,4.81-1.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M59.74,54.21c.2.07.69,10.48,3.64,15.11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M62.3,52.55s8.42,5.8,21.88-.94a5.28,5.28,0,0,1,6.77,1,4.89,4.89,0,0,1,.88,1.93h0A7.93,7.93,0,0,1,90.33,61a11.54,11.54,0,0,1-1.68,1.62L86,64.43c-5.77,3.3-11.14,6.1-22.76,4.71"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "106.19 92.15 75.3 92.15 75.06 73.15 106.19 73.15 106.19 92.15"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "90.56",
    y: "77.87",
    width: "2.85",
    height: "9.55"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "95.43",
    y: "77.87",
    width: "2.85",
    height: "9.55"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "100.29",
    y: "77.87",
    width: "2.85",
    height: "9.55"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "80.67",
    cy: "85.19",
    r: "2.44"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M27.5,79.38C23,74.25,14.77,66,34.24,47.91"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M32.21,50.5s2.93,7.33.32,14.23"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M37.59,51.74c.41,2,.49,10.77-5.32,14.62"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M25.5,57.73l-.29-9.08a2.34,2.34,0,0,1,2.34-2.33l5.85-.09"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "39.13",
    cy: "38.41",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M67.89,29.89c-1.89-.47-4.26-.66-8.66-.34-7.66.55-10.81,3.57-12.08,5.57a7.27,7.27,0,0,0-.52,4c.53,3.87,1.7,13.34,18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "50.05",
    y1: "15.86",
    x2: "50.13",
    y2: "18.37"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "50.06",
    cy: "13",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "50.06",
    cy: "13",
    rx: "2.68",
    ry: "2.7"
  }))));
}
