// @postman/aether/esmLib/components/Illustration/illustrations/NoPermission.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoPermission() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-permission-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "23.95",
    y1: "66.87",
    x2: "26.97",
    y2: "66.87",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "29.06",
    y1: "67.17",
    x2: "32.58",
    y2: "67.17",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "64.86",
    y1: "32.1",
    x2: "79.99",
    y2: "32.1",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.17",
    stopColor: "#fb800d"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.72",
    stopColor: "#efae2e"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M80.18,117.81c0,1.09-13.52,2-30.2,2s-30.21-.88-30.21-2c0-.56,2.67-.92,8.46-1.37.89-.06,1.5,2.15,3.2,2.08.69,0,4.08-.08,4.8-.1s2.6-2.54,3.3-2.55c3.65-.11,6.23,0,10.45,0,1.13,0,2.45,2.15,3.54,2.16,1.34,0,2.88.59,4.16.61.89,0,2.71-2.71,3.57-2.69C73.08,116.2,80.18,117,80.18,117.81Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M26.05,83.53v28a7.35,7.35,0,0,0,7.36,7.36h0a7.35,7.35,0,0,0,7.35-7.36l.77-19.12C41.56,90.86,43.24,91,44,91h0a1.5,1.5,0,0,1,1.5,1.44l2.77,18.41c.57,4.73,3.45,7.78,7.74,7.83h0c4.36,0,7-4.39,7-8.75L56.42,64.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M26.68,48.72,25.83,48a19.28,19.28,0,1,1,21,3l-.61.32"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "15.89",
    y: "48.94",
    width: "31.68",
    height: "34.12",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M20.31,82.83s.42,7,5.37,7.35"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "26.1",
    cy: "12.5",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "26.1",
    cy: "12.5",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M53.49,45.39c-5.84-2.12-8.8-8.16-7.35-14.36.85-3.64,4.37-7.49,7.42-9.11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M56.25,64.57s.48,8.5-8.6,6.45"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M57.08,70.11s-.26,7.19-9.34,5.14"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "27.63",
    y1: "14.77",
    x2: "28.9",
    y2: "16.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.75,49.3v7.41a5,5,0,0,1-5,5H16.23"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "25.46",
    cy: "66.87",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "30.82",
    cy: "67.17",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "30.74",
    cy: "66.96",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "25.38",
    cy: "66.96",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M16.13,65.26h4.06c.77,0,1.39.82,1.39,1.84h0a1.65,1.65,0,0,1-1.39,1.83H16.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M16.13,72.26H38.19c.77,0,1.39.82,1.39,1.84h0a1.65,1.65,0,0,1-1.39,1.83H16.13"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "37.79",
    cy: "34.92",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M50.07,49s14.69.11,19.6,33.52a5.48,5.48,0,0,1-4.49,6.2h0s-4.55.77-5.52-2.25"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M47.66,54.07a6.49,6.49,0,0,1,5.73-4"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "28.82 11.57 28.82 3.62 110.88 3.62 110.88 116.34 61.11 116.34"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "49.37",
    y1: "115.63",
    x2: "39.57",
    y2: "115.63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-8",
    d: "M64.86,26.08v12H80v-12Zm8.7,7.26.1,2.41h-2l0-2.45a1.74,1.74,0,1,1,1.9,0Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "64.86",
    y: "26.08",
    width: "15.13",
    height: "12.03"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M66.49,25.81V24.48a5.88,5.88,0,0,1,5.88-5.88h0a5.88,5.88,0,0,1,5.88,5.88v1.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M68.88,25.81V24.6a3.52,3.52,0,0,1,3.52-3.51h0a3.51,3.51,0,0,1,3.51,3.51v1.21"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M74.38,31.88a1.74,1.74,0,1,0-2.72,1.42l0,2.45h2l-.1-2.41A1.71,1.71,0,0,0,74.38,31.88Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-9",
    x: "65.11",
    y: "45.64",
    width: "15.19",
    height: "11.54"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "70.19",
    y1: "46.13",
    x2: "70.19",
    y2: "57.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "75.19",
    y1: "46.13",
    x2: "75.19",
    y2: "57.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "65.57",
    y1: "49.34",
    x2: "80.13",
    y2: "49.34"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "65.57",
    y1: "53.34",
    x2: "80.13",
    y2: "53.34"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "12.88",
    y1: "116.34",
    x2: "27.78",
    y2: "116.34"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "54.85 22.16 54.85 16.19 89.43 16.19 89.43 43.1 56.76 43.1 56.76 40.67"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "110.88",
    y1: "116.34",
    x2: "127.12",
    y2: "116.34"
  }))));
}
