// @postman/aether/esmLib/components/Illustration/illustrations/NoPullRequest.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoPullRequest() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-pull-request-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "24.38",
    y1: "29.13",
    x2: "29.64",
    y2: "29.13",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M52.94,47.69C48.7,46.23,42.61,44,38.38,38.19c-2-2.76-6.7-5.79-10.27-2h0a6.56,6.56,0,0,0-1.75,7.63C29,49.5,33.11,56.21,48.76,61.91"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M99.38,113.8c0,2.92-13.09,5.28-29.25,5.28s-29.25-2.36-29.25-5.28c0-1.46,2.77-2.82,8.06-3.69,1.08-.18.9,3.81,2.44,5.07,5.3,4.36,8.22,1.52,11.44-.26,1.51-.84,1.33-6.33,2.05-6.37,2.81-.16,2.16,0,5.26,0,1.5,0-.41.12,1,.08.67,0,0,5.38,3.77,6.48,1.09.32,3.66,1.87,6,1.74,1.94-.11,3.21-2.43,4.19-3.08.68-.45,1.33-4.41,2.07-4.33C95.44,110.33,99.38,111.9,99.38,113.8Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "67.2",
    cy: "33.76",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M84.61,62.51l2,45.61c0,4.36-2.64,8.8-7,8.75h0A7.83,7.83,0,0,1,71.87,109L71,90.63c0-.81-2.55-1.44-3.36-1.44h0c-.8,0-2.38.62-2.43,1.42l-.84,19.12A7.35,7.35,0,0,1,57,117.09h0a7.35,7.35,0,0,1-7.35-7.36L50,85.24l-.85-23.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M49.06,61.61s4.2-3.54,4.33-14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M56.68,50.11s2,9.38-7.32,16.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M79.53,48.47s-1.34,13.59,5.08,18.22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M88.19,49.46l-.07-6a2.33,2.33,0,0,0-2.37-2.3l-.45.07"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M102.42,70.07c-1.36-6.49-8.29-18.68-20.54-24.17"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M83.71,46.37c-1.28,12.27.84,16.15.84,16.15l3.63,4.83A31.34,31.34,0,0,1,92.12,74"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M97.25,71.26a15.79,15.79,0,0,1,6.26-1.55c8.65,0,15.67,7.47,15.67,16.68s-7,16.67-15.67,16.67S87.83,95.59,87.83,86.39a19,19,0,0,1,3.35-10.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M100.77,72.34a12.44,12.44,0,0,1,2.74-.3C111,72,117,78.46,117,86.39s-6,14.34-13.48,14.34S90,94.31,90,86.39a14.93,14.93,0,0,1,1.35-6.26"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M102.45,74c.24,0,.8-.14,1.06-.14,6,0,10.88,7.59,10.88,17s-4.87,17-10.88,17-10.89-7.59-10.89-17a24.74,24.74,0,0,1,1.32-8.1"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "97.22",
    cy: "77.62",
    r: "6.21"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M91.41,75.21S38.82,105.14,28,31.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M26.82,27.3c-.44-3.21-.89-4.62-1.17-8.21A1.93,1.93,0,0,1,27.73,17h0a1.93,1.93,0,0,1,1.61,2.7"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M24.6,27.62l4.74-.83a.25.25,0,0,1,.3.23v3.46a.26.26,0,0,1-.22.24l-3.82.75a.25.25,0,0,1-.29-.17l-.92-3.39A.24.24,0,0,1,24.6,27.62Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M24.75,27.76,29.17,27a.21.21,0,0,1,.24.21l.3,3.24a.22.22,0,0,1-.18.21l-3.94.62a.21.21,0,0,1-.24-.16L24.57,28A.21.21,0,0,1,24.75,27.76Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M82.17,36.63c.19,6-6.32,12.84-14.6,13.1S52.38,43.61,52.19,37.56s0-11,14.64-11.42C80.82,25.72,82,30.58,82.17,36.63Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "71.18",
    y: "57.98",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "66.48",
    y1: "11.93",
    x2: "66.56",
    y2: "14.44"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "66.49",
    cy: "9.07",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "66.49",
    cy: "9.07",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M86.32,30,89,31a1.45,1.45,0,0,1,.81,1.33V35a1.43,1.43,0,0,1-1,1.39l-2.79.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M47.84,30.52,45.2,31.73a1.45,1.45,0,0,0-.81,1.33v2.68a1.45,1.45,0,0,0,1,1.4l2.79.84"
  }))));
}
