// @postman/aether/esmLib/components/Illustration/illustrations/NoReport.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function NoReport() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-report-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "67.93",
    y1: "63.91",
    x2: "90.05",
    y2: "63.91",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "67.93",
    y1: "71.48",
    x2: "86.11",
    y2: "71.48",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M97.91,113.06l-14.19,1.3-13.15.15-2.19-.33c-2.3-.14-4.43-.79-4.6-3.09l-.41-2.52c-1,0-.6,0-2,.07-.74,0-1.39,3.45-2.94,4.24-3.3,1.68-4.49,2.86-9-.23-1.69-1.14-1.41-2.76-2.73-2.69-5.48.28-8.75,2.15-8.75,3.53,0,2.76,13.41,5,30,5a124.56,124.56,0,0,0,23.38-1.87,15.15,15.15,0,0,0,5.5-2.37Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "64.46",
    cy: "31.22",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M81,55c11.5-8.27,14.55-17.19,10.36-26.17a8.58,8.58,0,0,0-2.81-3.49,5.3,5.3,0,0,0-5.4-.61"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M46.64,69.83v37.32A7.38,7.38,0,0,0,54,114.53h0a7.38,7.38,0,0,0,7.38-7.38l.87-16.86a1,1,0,0,1,1-.92l.26.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M64.3,82.32l-11.89.17,0-10.78,5.44-.45L60.89,68l-.13-3.69-2-2.48L53,59.41c0-3.11,2.52-4.63,5.63-4.63h0c3.11,0,5.63,1.52,5.63,4.63Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M48.1,42S31.29,51,38,62.94a13.44,13.44,0,0,0,6,5.4c2.72,1.32,6.12,3.32,11,3.35a6.07,6.07,0,0,0,4.72-2.13,5.83,5.83,0,0,0,1.33-2.89,4.47,4.47,0,0,0-2.49-4.6l-11.88-5.9"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M53.55,59h0a4,4,0,0,1,4-4h2.19a4,4,0,0,1,4,4.1c-.19,9.71.83,34.28-.35,47.31-1.31,14.55,19.9,6.41,14.24-5.53h24.48s5.72,10.77-6,12.41L69,114.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M59.76,55H89.91s7.27,2.64,3.37,37.71c0,0-.69,7,2.58,8.18"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M49.08,57.33a11.93,11.93,0,0,0-3-13.65"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M53.85,47.68c1,2.73,1.35,7.2.39,9"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M74.82,48s-.42,6,1.05,6.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M79.13,43.76s-.95,10,1.25,11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M41.28,47.17V43.5a2.33,2.33,0,0,1,2.33-2.33h4.2"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M46.6,75.38h-3A2.33,2.33,0,0,1,41.28,73V66.83"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "52.17 71.84 52.17 82.93 63.96 82.93"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("rect", {
    className: "cls-5",
    x: "67.93",
    y: "62.23",
    width: "22.13",
    height: "3.35"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "67.93",
    y: "69.8",
    width: "18.18",
    height: "3.37"
  })), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "67.82",
    y: "62.32",
    width: "22.16",
    height: "3.55"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "67.82",
    y: "69.62",
    width: "18.13",
    height: "3.55"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "64.62",
    y1: "8.89",
    x2: "64.64",
    y2: "11.88"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "64.65",
    cy: "6.25",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "64.65",
    cy: "6.25",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M83.82,28.7l2.34,1.07A1.46,1.46,0,0,1,87,31.11v2.68a1.42,1.42,0,0,1-1,1.39L83.21,36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M83.82,28.7l2.34,1.07A1.46,1.46,0,0,1,87,31.11v2.68a1.42,1.42,0,0,1-1,1.39L83.21,36"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M45.3,28.57l-2.64,1.2a1.46,1.46,0,0,0-.81,1.34v2.68a1.43,1.43,0,0,0,1,1.39l2.8.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M79.77,33.92C80,40,73.45,46.77,65.17,47S50,40.9,49.78,34.86s0-11,14.65-11.42C78.41,23,79.58,27.88,79.77,33.92Z"
  }))));
}
