// @postman/aether/esmLib/components/Illustration/illustrations/NoRequestInCollection.js
import React from 'react';
export default function NoRequestInCollection() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "no-request-in-collection-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "84.85",
    y1: "67.28",
    x2: "91.91",
    y2: "67.28",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "60.75",
    y: "54.86",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M76.7,108.46c8.1.52,11,2.5,11,4.4,0,2.91-12.76,5.28-28.5,5.28s-28.5-2.37-28.5-5.28c0-1.46,4.41-3.22,9.6-3.91.7-.1,2.39,6,4.35,5.79,1.32-.12,2.64.76,4.76.64,1.14-.07,5.45-2.36,6-7.45.12-1.14-1.06.06.24,0s3,0,4.53,0c.92,0,1.55,4.87,4,5.4a50.73,50.73,0,0,0,6.61.3c1.3,0,4.12-5.21,4.65-5.18C76.91,108.54,75.09,108.36,76.7,108.46Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M59,51.17c-7.17-.88-12.4-6.23-12.57-11.68-.19-6,0-11,14.64-11.41,10.11-.31,13.52,2.14,14.7,5.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M55,51.3c.56.05,1.14.07,1.72.07A19.28,19.28,0,0,0,61,50.89s6-.53,11.66-8A17.7,17.7,0,0,0,76,32.09a19.29,19.29,0,1,0-25.6,18.23"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "57.37",
    y1: "10.31",
    x2: "57.23",
    y2: "12.82"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "57.64",
    cy: "7.46",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.07 64.21) rotate(-84.89)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "57.64",
    cy: "7.46",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.07 64.21) rotate(-84.89)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M76.11,32.05l2.09.56A1.45,1.45,0,0,1,79,33.94v2.68a1.44,1.44,0,0,1-1,1.4l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M37.3,31.32l-2.16,1a1.46,1.46,0,0,0-.81,1.34v2.68a1.42,1.42,0,0,0,1,1.39l2.79.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M40.69,72.81l-.09,6.24-.21,26.84c-.13,5.69,2.62,9.09,7.42,9h.54a7.31,7.31,0,0,0,3.19-.76,7.42,7.42,0,0,0,4-6L57.29,89l4.78-1.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M38.82,46.68S41.88,47.8,41.31,57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M44.42,53.78c.67-4.4.3-5.09-.42-7.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M73.13,59.17c.94,15.51,2.92,39.12,2.32,47.79A7.61,7.61,0,0,1,68,114.07h0a7.61,7.61,0,0,1-7.67-6.73l-.22-18.53"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "107.48",
    y1: "50.6",
    x2: "107.41",
    y2: "70.66"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M107.41,50.57l-7,7s.32,20.41,0,20.09l7-7"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "81.26 54.95 99.2 54.95 105.2 49.09 99.17 49.09"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "78.45",
    y: "57.85",
    width: "19.34",
    height: "20.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M73.22,62.18c-.54-2-3.36-11.89-3.07-16.47"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M71.87,44.23A51.07,51.07,0,0,1,94.18,42a5.69,5.69,0,0,1,4.9,6.31h0a5.73,5.73,0,0,1-5.69,5.06c-5.34,0-16.9,1.09-20.19,5.93"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M66.61,48.84S66.68,60.29,73,63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M41.3,44.19S23,59.06,33.41,70.15a8,8,0,0,0,7.8,2.34l.13,0a10.19,10.19,0,0,0,4.24-2.4L51.14,65a26.72,26.72,0,0,0,4.11-4.71c1.36-2,1.66-3.18,1.24-5.41a9,9,0,0,0-.8-2.29l-.1-.2h0c-2.26-2.91-6.5-2.27-9.46-.06L37.86,61.4"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M87,68.78v-.21a5,5,0,0,1,.15-1.36,1.93,1.93,0,0,1,.42-.82,3.15,3.15,0,0,1,.67-.57,5.22,5.22,0,0,0,.57-.41,1.8,1.8,0,0,0,.39-.45,1.06,1.06,0,0,0,0-1.07.94.94,0,0,0-.39-.35,1.19,1.19,0,0,0-.54-.12,1.21,1.21,0,0,0-.56.13,1.11,1.11,0,0,0-.59,1H84.85a3,3,0,0,1,.49-1.73,2.72,2.72,0,0,1,1.26-1,4.41,4.41,0,0,1,1.73-.33,5.09,5.09,0,0,1,1.85.32,2.89,2.89,0,0,1,1.27,1,2.51,2.51,0,0,1,.46,1.54,2.46,2.46,0,0,1-.2,1,2.4,2.4,0,0,1-.56.81,4.23,4.23,0,0,1-.84.64,3.31,3.31,0,0,0-.63.46,1.74,1.74,0,0,0-.38.56,2,2,0,0,0-.12.75v.21Zm1.12,4.3a1.23,1.23,0,0,1-.93-.38,1.19,1.19,0,0,1-.38-.93,1.18,1.18,0,0,1,.38-.91,1.27,1.27,0,0,1,.93-.38,1.29,1.29,0,0,1,.91.38,1.23,1.23,0,0,1,.39.91,1.21,1.21,0,0,1-.18.66,1.57,1.57,0,0,1-.48.48A1.3,1.3,0,0,1,88.15,73.08Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M87,68.78v-.21a5,5,0,0,1,.15-1.36,1.93,1.93,0,0,1,.42-.82,3.15,3.15,0,0,1,.67-.57,5.22,5.22,0,0,0,.57-.41,1.8,1.8,0,0,0,.39-.45,1.06,1.06,0,0,0,0-1.07.94.94,0,0,0-.39-.35,1.19,1.19,0,0,0-.54-.12,1.21,1.21,0,0,0-.56.13,1.11,1.11,0,0,0-.59,1H84.85a3,3,0,0,1,.49-1.73,2.72,2.72,0,0,1,1.26-1,4.41,4.41,0,0,1,1.73-.33,5.09,5.09,0,0,1,1.85.32,2.89,2.89,0,0,1,1.27,1,2.51,2.51,0,0,1,.46,1.54,2.46,2.46,0,0,1-.2,1,2.4,2.4,0,0,1-.56.81,4.23,4.23,0,0,1-.84.64,3.31,3.31,0,0,0-.63.46,1.74,1.74,0,0,0-.38.56,2,2,0,0,0-.12.75v.21Zm1.12,4.3a1.23,1.23,0,0,1-.93-.38,1.19,1.19,0,0,1-.38-.93,1.18,1.18,0,0,1,.38-.91,1.27,1.27,0,0,1,.93-.38,1.29,1.29,0,0,1,.91.38,1.23,1.23,0,0,1,.39.91,1.21,1.21,0,0,1-.18.66,1.57,1.57,0,0,1-.48.48A1.3,1.3,0,0,1,88.15,73.08Z"
  }))));
}
