// @postman/aether/esmLib/components/Illustration/illustrations/PageNotFound.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function PageNotFound() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "page-not-found-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "110.24",
    y1: "70.42",
    x2: "133.48",
    y2: "70.42",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "29.66",
    y1: "57.44",
    x2: "34.2",
    y2: "57.44",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M92.93,114.27c0,1.49-14.2,2.69-31.72,2.69s-31.72-1.2-31.72-2.69c0-.74,4.17-.94,9.87-1.75.62-.09,1.78,3.41,3.93,3.3,1.45-.08,3.16.5,6.65-.4,1.64-.43,2.83-3.91,3.61-3.93,3.05-.08,4.29.1,7.66.1,1.63,0,2.16.1,3.72.08.73,0,1.52,3.19,2.86,3.7a12.87,12.87,0,0,0,5.53,1.52c2.11-.06,4.34-2.43,5.4-2.77.74-.22.4-2.05,1.21-2C88.9,112.59,92.93,113.31,92.93,114.27Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M88.38,36a35.14,35.14,0,0,1-.45,4.89,18.22,18.22,0,0,1-3.44,7,13.33,13.33,0,0,1-1.66,1.7,19.89,19.89,0,0,1-12.53,5h0a19.32,19.32,0,0,1-16.74-8.25,18.22,18.22,0,0,1-3.2-10.25,19.3,19.3,0,0,1,27.31-17.5,18.5,18.5,0,0,1,9.8,11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M82.21,51.27c.48,24.08-1.19,48.06-2.15,59.52-.34,3.93-3.63,6-7.58,6h0c-3.91,0-7.18-2-7.57-5.9l-.2-18.52"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M52.24,64.44,38.69,108.12c-1.08,4.66,2.48,8.19,7.28,8.19h.54a7.48,7.48,0,0,0,7.05-5l7.77-17.84,4.81-1.72"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M79,52.57c.2.06,0,8.47,3,13.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "53.7",
    y1: "20.99",
    x2: "55.32",
    y2: "22.92"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "51.94",
    cy: "18.74",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(-0.45 36.19) rotate(-38.26)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "51.94",
    cy: "18.74",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(-0.45 36.19) rotate(-38.26)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M97.75,33.72s8,23.69-15.16,27.26"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M85,24.67c-2.35.75-5,1-9.16,2.44-7.23,2.58-9.46,6.33-10.15,8.6a7.26,7.26,0,0,0,.58,4c1.54,3.59,5.8,11.7,19.35,5.93"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "124.23 24.49 85.88 36.45 84 31.89 119.17 12.56"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "128.33 25.63 125.37 26.84 118.64 9.74 121.6 8.52 128.33 25.63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M52.24,64.64c-14-9.82-14.7-18.22-14.7-18.22C35.84,36,51,29.64,51,29.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M57.45,50.68s-.84,12-5,13.87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M53.56,46.31s0,9.5-2.89,17.16"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "50.69",
    y1: "40.77",
    x2: "48.97",
    y2: "43.35"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "121.86",
    cy: "70.42",
    r: "11.62"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "121.86",
    cy: "70.42",
    r: "11.62"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "30.02",
    cy: "20.59",
    r: "5.95"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M36.2,18.64h3.21c1.31,0,2.38.79,2.38,1.78h0c0,1-1.07,1.77-2.38,1.77h-18c-1.32,0-2.39-.79-2.39-1.77h0c0-1,1.07-1.78,2.39-1.78h2.66"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M121.6,8.52s8.89,4.5,6.73,17.11"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "28.84 74.47 29.85 75.04 28.84 75.6 28.28 76.63 27.73 75.6 26.71 75.04 27.73 74.47 28.28 73.44 28.84 74.47"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "108.84 39.47 109.85 40.04 108.84 40.6 108.28 41.63 107.72 40.6 106.71 40.04 107.72 39.47 108.28 38.45 108.84 39.47"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "132.09 98.71 133.1 99.27 132.09 99.84 131.53 100.86 130.97 99.84 129.96 99.27 130.97 98.71 131.53 97.68 132.09 98.71"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-8",
    points: "78.84 9.47 79.85 10.04 78.84 10.6 78.28 11.63 77.72 10.6 76.71 10.04 77.72 9.47 78.28 8.45 78.84 9.47"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-9",
    cx: "31.93",
    cy: "57.44",
    rx: "2.27",
    ry: "2.28"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "31.93",
    cy: "57.44",
    rx: "2.27",
    ry: "2.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M3.35,87.52s19.53-.3,41.05,1.78"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M81.36,96.18c16.27,4.68,34.56,12,47.93,22.44"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "108.6",
    y1: "18.56",
    x2: "112.43",
    y2: "28.21"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "97.1",
    y1: "24.73",
    x2: "99.67",
    y2: "31.97"
  }))));
}
