// @postman/aether/esmLib/components/Illustration/illustrations/RunAborted.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function RunAborted() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "run-aborted-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "102.67",
    y1: "41.27",
    x2: "121.18",
    y2: "41.27",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "27.5",
    y1: "78.74",
    x2: "38.39",
    y2: "78.74",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M91,35.22A23.29,23.29,0,0,1,93,40a19.28,19.28,0,1,1-9.55-11.88S88.7,30.94,91,35.22Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M68.81,92.49a9.85,9.85,0,0,0,5,5,10.13,10.13,0,0,0,9-.52l5.74-2.57,6.88,13.87a7.3,7.3,0,0,0,8.72,3h0a7.3,7.3,0,0,0,4.06-10l-7.42-14.51a9.15,9.15,0,0,0-9-4.77l-4,1.08"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M63.36,76c-1-6.23-1.51-7.68-4.91-11.86"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M67.52,76.46s1.14-7-5.08-15.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M56.09,68.51l-1.58-4.3a2.33,2.33,0,0,1,1.62-2.87l5.64-1.65"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M89.44,33.19c-2.35.75-4.17.8-8.29,2.27C73.92,38.05,71.69,41.8,71,44.07a7.27,7.27,0,0,0,.58,4c1.54,3.58,6.47,11.62,20,5.86"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M97.67,60a11.13,11.13,0,0,0-5.91-6.39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M90.14,56.66c2.81.71,4.58,6.64,4.58,6.64"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "63.9",
    y1: "26.5",
    x2: "65.1",
    y2: "28.25"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "62.2",
    cy: "23.55",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "62.44",
    cy: "23.92",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M62.88,60.77s-11,2.44-6.39,25A27.39,27.39,0,0,0,58,90.57h0a7.24,7.24,0,0,0,2.54,3.26,4.87,4.87,0,0,0,2.7.9c2.8,0,5-1.38,5.73-4.38l1.69-7A5.77,5.77,0,0,0,69,77.74h0a5.77,5.77,0,0,0-9.56,3"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M107.9,100.76c1.48,3,4.26,5,7.39,3.83h0a6.56,6.56,0,0,0,3.8-8.73l-5.44-12.55c-1.67-3.77-6.49-9.55-12.8-9.37h-.94l-3.24-.8L94.75,66"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "99.78",
    cy: "64.86",
    r: "5.19"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M100.91,73.94a10.29,10.29,0,0,1,1.21-1.24"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "69.25 89.04 70.73 92.58 73.9 90.2 70.2 84.92"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-2",
    points: "104.07 68.02 107.13 70.98 103.6 73.43 100.24 70.03"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M65.7,14a55.4,55.4,0,0,0-10.34,4.36,60.46,60.46,0,0,0-10.31,6.8C23.48,42.86,18.28,59.12,18.28,59.12c-1,.61-12.62-21.43,7.79-39.5C43.83,3.9,65.7,14,65.7,14Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M65.7,14a55.4,55.4,0,0,0-10.34,4.36,60.46,60.46,0,0,0-10.31,6.8C23.48,42.86,18.28,59.12,18.28,59.12c-1,.61-12.62-21.43,7.79-39.5C43.83,3.9,65.7,14,65.7,14Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M19.69,57.53l2.93,1.95A10.91,10.91,0,0,0,31,61l3.05-.68a10.94,10.94,0,0,1,7.72,1.15h0a10.86,10.86,0,0,0,7.49,1.2l6.86-1.37"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "63.07",
    y1: "15.18",
    x2: "63.2",
    y2: "20.98"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M25.07,46.45h0a5.33,5.33,0,0,0,3.24,1,19.12,19.12,0,0,1,12.34,4.19h0a19.1,19.1,0,0,0,11,4.16l6.63.32"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-2",
    x: "85.74",
    y: "64.13",
    width: "4.93",
    height: "5.07",
    rx: "0.43",
    transform: "translate(195.89 91.73) rotate(156.02)"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M102.93,46.1h18.22s.76-6.67-4.75-5.53c0,0,.95-4.6-5-4.08a2.26,2.26,0,0,0-1.4.63,2.7,2.7,0,0,0-1.08,2s-3.73-.66-3.43,3C105.53,42.09,101.66,42.13,102.93,46.1Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M102.63,46.1h18.21s.77-6.67-4.74-5.53c0,0,1-4.6-5-4.08a2.32,2.32,0,0,0-1.41.63,2.7,2.7,0,0,0-1.07,2s-3.73-.66-3.43,3C105.23,42.09,101.36,42.13,102.63,46.1Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M27.65,81.58H38.37s.45-3.93-2.79-3.26c0,0,.56-2.7-2.92-2.39a1.25,1.25,0,0,0-.82.37,1.56,1.56,0,0,0-.64,1.15s-2.19-.38-2,1.77A1.62,1.62,0,0,0,27.65,81.58Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M27.35,81.58H38.07s.45-3.93-2.79-3.26c0,0,.56-2.7-2.92-2.39a1.3,1.3,0,0,0-.83.37,1.59,1.59,0,0,0-.63,1.15s-2.19-.38-2,1.77A1.61,1.61,0,0,0,27.35,81.58Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "11.18",
    y1: "7.37",
    x2: "18.87",
    y2: "15.4"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "11.7",
    y1: "16.28",
    x2: "16.42",
    y2: "21.34"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-2",
    x1: "18.69",
    y1: "7.54",
    x2: "23.41",
    y2: "12.61"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "74.7",
    cy: "113.19",
    rx: "12.41",
    ry: "3.1"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-2",
    cx: "63.36",
    cy: "48.07",
    r: "2.88"
  }))));
}
