// @postman/aether/esmLib/components/Illustration/illustrations/Search.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function Search() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "search-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "80.52",
    y1: "66.46",
    x2: "116.4",
    y2: "66.46",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M46.73,110.17c2.69-.13,2.06,0,5,0,1.44,0,1,0,2.37-.07.64,0,.13,4,3.3,5.3,1.65.66,1.32,1.14,4.24.91,1.86-.15,3.34-1.56,4.28-2,.65-.34,1.87-3.61,2.58-3.54,7.93.7,11.26,2,11.26,3.43,0,2.21-12.55,4-28,4s-28-1.78-28-4c0-1.1,1.93-1.73,5.36-2.18,0,0,1.34,5.28,9.29,3.25h0a7.87,7.87,0,0,0,3.94-5h0Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M52.8,54.42c-8.4-2.08-15.24-9.65-15.24-18.66a19.31,19.31,0,0,1,27-17.65"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M36.67,74.92l-8.4,32.81a6.53,6.53,0,0,0,6.25,8.15h0c2.24.46,5.41-.26,7.24-4.18L47,94.15l8.43-2.08"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M67.46,79.58l.73,29.11a7.41,7.41,0,0,1-1.68,4.93,7.09,7.09,0,0,1-10.6.3,6.67,6.67,0,0,1-1.71-4.74l-.36-16.42"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.44,42.29l-5.23.09a2.33,2.33,0,0,0-2.33,2.33l.29,26.47A1.86,1.86,0,0,0,33,73L35.11,73"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "46.61",
    cy: "33.46",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "56.53",
    y1: "13.91",
    x2: "56.61",
    y2: "16.43"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "56.54",
    cy: "11.05",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "56.54",
    cy: "11.05",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M85.17,64.59c-5.68,3-12.07,3.5-17.73.93C56.57,60.57,52.57,46,58.51,32.94S78.06,13.31,88.93,18.26s14.87,19.53,8.94,32.58A32.92,32.92,0,0,1,95.6,55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M100.19,54.78c.42-.76.81-1.55,1.17-2.35,7-15.37,2.28-32.55-10.52-38.37S62,16,55,31.35c-6.38,14-3,29.56,7.37,36.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M67.46,70.47a22.9,22.9,0,0,0,16.13-.53"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M63.5,79.36l-1.25,2.83a2.84,2.84,0,0,1-3.7,1.54h0a2.79,2.79,0,0,1-.84-.53s-1.47-1-.39-3.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M67.46,70.47s-4-3.55-7.52-2.43c-2.87.9-5.75,8.2,3.76,10.91A10.72,10.72,0,0,0,77,71.63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.45,48.68A13.48,13.48,0,0,1,41,60.74l15,12.69a5.38,5.38,0,0,1,1.27,6.65h0c-1.47,2.73-6,4.49-8.64,2.86-3.81-2.37-8.84-5.37-10.72-7-14.4-12.69,2.7-29.8,2.7-29.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M43.65,49.43c0,.31,4.49,5,1.24,14.23"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M74.67,18.36A22.73,22.73,0,0,1,80.87,34a21.38,21.38,0,0,1-2.23,9.46A23,23,0,0,1,58.74,56"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M79.2,66.41c1.25-11.72-5.7-16.26-5.7-16.26"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M73.77,50.72s3.1,9.33.63,16.08"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M68.75,53.47a23.76,23.76,0,0,1,.73,12.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M56.29,45.36c2.83,3.07,8.88,4.82,18.64,3.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M78.61,24.78c-2.25-.56-4.88-.84-10.11-.46a32.52,32.52,0,0,0-4.39.61"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "105.48 66.08 92.4 66.02 88.25 77.73 80.52 77.54 88.61 55.2 116.4 55.2 112.51 66.02 105.48 66.08"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "87.72 77.64 80.84 77.64 88.61 55.2 116.4 55.2 112.37 66.1"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "113.66 85.01 112.39 88.7 84.6 88.7 92.38 66.26 120.16 66.26 116.86 76.03"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "120.39 84.96 106.6 84.96 109.94 76.26 123.73 76.26 120.39 84.96"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "92.05",
    y1: "60.31",
    x2: "110.52",
    y2: "60.31"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "90.97",
    y1: "63.73",
    x2: "109",
    y2: "63.73"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "89.51",
    y1: "67.16",
    x2: "91.37",
    y2: "67.16"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "88.05",
    y1: "70.31",
    x2: "90.13",
    y2: "70.31"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "94.63",
    y1: "31.45",
    x2: "81.88",
    y2: "47.72"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "94.63",
    y1: "37.45",
    x2: "81.88",
    y2: "53.72"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "96.09",
    cy: "113.52",
    rx: "8.97",
    ry: "1.53"
  }))));
}
