// @postman/aether/esmLib/components/Illustration/illustrations/SignIn.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function SignIn() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "sign-in-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "96.51",
    y1: "49.22",
    x2: "110.27",
    y2: "49.22",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M82.31,110.55c0,2.92-14.2,5.29-31.72,5.29s-31.72-2.37-31.72-5.29c0-1.45,3.63-2.35,9.27-3.73.18,0,2.06,5.93,4.21,5.71,1.46-.15,4,1.42,7.46-.36,1.64-.84,3.54-6.81,4.32-6.85,3-.16,3.09,0,6.46,0,1.63,0,2.16.22,3.72.18.73,0,.46,4.6,1.8,5.61a10.16,10.16,0,0,0,6,2.15c2.1-.12,5-1.93,6-2.58.73-.45.56-4.47,1.37-4.39C78.45,107.17,82.31,108.66,82.31,110.55Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M110.27,49.22c0,10.19-3.11,18.45-7,18.45-2.72,0-5-4.47-6.11-10.5,0-.26,2.51-2.26,2.68-2.46,1.18-1.42,1.07-2.43,1-4,0-.47.11-3.46-1.56-4.32s-2.77-.64-2.77-1.16c0-1.75-.08.56.1-1,.83-7.73,3.51-13.39,6.69-13.39C107.16,30.78,110.27,39,110.27,49.22Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "59.02",
    cy: "31.87",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M54.29,87.69v19.2a7.59,7.59,0,0,0,7.56,6.84h0a7.59,7.59,0,0,0,7.58-7c1-11.48,2.64-35.52,2.16-59.65"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M42,41.54c-8.74,6.58-20.52,24.63-23.62,30.74-3.53,7,3,9.73,3,9.73s4.34,2.13,8.27-3.34c3.25-4.53,7.65-10.87,9.08-12.94a.09.09,0,0,1,.16.09L36.07,76.28l-8,27.81a7.47,7.47,0,0,0,7.28,9.14h.55a7.48,7.48,0,0,0,7-5l7.77-18.81,4.79-2.61"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M38.25,66.35s7-8.21,0-21.2"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M38.21,67S51.29,57.9,45.3,45.53"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M74.86,48.9s1,7.87-2.56,14.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M96.82,45.46c.62-8.59,3-14.63,6.47-14.63,3.9,0,7,8.3,7,18.53s-3.15,18.52-7,18.52c-2.68,0-5-3.95-6.21-9.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M94,30.85c2.09-8.59,5.6-14.22,9.58-14.22,6.39,0,11.58,14.56,11.58,32.52S110,81.67,103.6,81.67c-3.87,0-7.3-5.34-9.41-13.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M86.71,44.49c1.05-21,8.67-37.25,17.9-37.25,10,0,18,18.9,18,42.21s-8.07,42.21-18,42.21c-7.64,0-14.17-11.13-16.79-26.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72,47.36s7.68,6.38,21.78-1.21c2.8-1.5,6.16-.48,6.73,2.55l.12.59a15.84,15.84,0,0,1,.28,3.23,3.31,3.31,0,0,1-.67,2.07,7.73,7.73,0,0,1-2.06,2L94.63,59c-5.78,3.31-11.44,5.74-23.06,4.35"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "56.19",
    y1: "9.77",
    x2: "56.42",
    y2: "12.67"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-4",
    x: "61.78",
    y: "57.2",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M34.46,49V42.52a2.34,2.34,0,0,1,2.34-2.34h4.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M34.85,71.88a2.31,2.31,0,0,0,1.27.38h1"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-6",
    cx: "55.95",
    cy: "6.89",
    r: "2.73"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "55.95",
    cy: "7.33",
    r: "2.73"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "85.28",
    y1: "21.19",
    x2: "77.41",
    y2: "12.25"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "92.54",
    y1: "90.13",
    x2: "89.34",
    y2: "94.11"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "81.23 69.1 82.25 69.66 81.23 70.23 80.68 71.26 80.12 70.23 79.11 69.66 80.12 69.1 80.68 68.06 81.23 69.1"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "89.65 9.11 90.67 9.67 89.65 10.24 89.1 11.27 88.54 10.24 87.53 9.67 88.54 9.11 89.1 8.08 89.65 9.11"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-7",
    points: "104.59 23.85 105.61 24.41 104.59 24.98 104.03 26.01 103.48 24.98 102.46 24.41 103.48 23.85 104.03 22.81 104.59 23.85"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "78.53",
    y1: "87.75",
    x2: "88.37",
    y2: "82.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M77,24.83a32.57,32.57,0,0,0-8.55-.29c-7.66.55-10.82,3.57-12.08,5.58a7.08,7.08,0,0,0-.52,4c.48,3.55,1.07,12,16.79,11.29"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "47.19",
    cy: "34.24",
    r: "2.88"
  }))));
}
