// @postman/aether/esmLib/components/Illustration/illustrations/SwitchToWorkspace.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function SwitchToWorkspace() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "switch-to-workspace-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "45.08",
    y1: "39.29",
    x2: "56.23",
    y2: "39.29",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "322.01",
    y1: "62.64",
    x2: "325.02",
    y2: "62.64",
    gradientTransform: "matrix(-1, 0, 0, 1, 424.11, 0)",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "316.35",
    y1: "62.94",
    x2: "319.87",
    y2: "62.94",
    gradientTransform: "matrix(-1, 0, 0, 1, 424.11, 0)",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M121.17,112.37c0,2.91-12.76,5.28-28.51,5.28s-28.5-2.37-28.5-5.28c0-1.46,3.86-3,9-3.86,1-.17.57,4.43,2.62,6,4.34,3.33,6.79,1.78,9.93,0,1.47-.84,2-7.33,2.67-7.37a34.78,34.78,0,0,1,4.27-.07c1.47,0,1.42.11,2.82.06.66,0,.65,5.5,1.86,6.5.86.72,3.57,3.24,5.82,3.24,4.21,0,5.65-3.89,6.61-4.54.66-.45-.31-4.45.41-4.37C118.25,108.91,121.17,110.48,121.17,112.37Z"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-3"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M49.29,33.24c-4.65,0-8.42,2.71-8.42,6s3.77,6.05,8.42,6.05c0,0-5.49-3.69-1.7-9.92,2.39-3.93-2.88.79-2.88.79Z"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M51.13,45.34a6,6,0,1,1,0-12.1c2.79,0,4.4.9,5.1,3.46,0,0-5.78,0-5.1,8.64"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M52.14,47.4H26.23a7.4,7.4,0,0,1-7.4-7.4v-.78a7.4,7.4,0,0,1,7.4-7.39H51.31a7.41,7.41,0,0,1,6.94,4.81"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M108.35,44.6a19.27,19.27,0,0,0-8.71-32.8A19.28,19.28,0,0,0,83.07,45.57l2.16,1.49"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M81.11,43.52S71.47,45.86,60.89,38a6.38,6.38,0,0,0-9.16,1.59h0a6.35,6.35,0,0,0,.83,8A35.27,35.27,0,0,0,76.05,57.75L73.3,107.86c0,4.35,2.64,8.79,7,8.74h0c4.29,0,6.78-3,7.74-7.82l2.77-18.4a1.52,1.52,0,0,1,1.51-1.44h0a1.5,1.5,0,0,1,1.5,1.43l1.77,19.1a7.36,7.36,0,0,0,7.36,7.35h0a7.36,7.36,0,0,0,7.36-7.35l-.93-29.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M79.54,19.94A12.42,12.42,0,0,1,87.3,34.51c-.87,3.62-3.89,6.67-7,8.27"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "84.03",
    y: "45.12",
    width: "31.68",
    height: "34.09",
    rx: "6.35"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M113.3,78.06c0-.3,2.33,8.46-3.87,7.89"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M75.61,64.85s3.86,2,8.41-3.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M75.4,69.55s5.43.58,8.69-3.57"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M51.06,45.34a6.05,6.05,0,1,1,5.38-8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M77.4,43.79S84,46.6,84.12,51.3"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "97.68",
    cy: "32.53",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M92.58,45.07v7.41a5,5,0,0,0,5,5H115.1"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-8",
    cx: "100.59",
    cy: "62.64",
    r: "1.51"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("circle", {
    className: "cls-9",
    cx: "106",
    cy: "62.94",
    r: "1.76"
  })), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "100.59",
    cy: "62.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-7",
    cx: "105.95",
    cy: "62.73",
    r: "1.77"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M115.19,61h-4c-.77,0-1.39.82-1.39,1.84h0a1.65,1.65,0,0,0,1.39,1.83h4"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M115.19,68h-22c-.77,0-1.39.82-1.39,1.84h0a1.65,1.65,0,0,0,1.39,1.83h22"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "108.65",
    y1: "13.21",
    x2: "107.07",
    y2: "15.17"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-5"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-10",
    cx: "110.48",
    cy: "11.27",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(30.13 87.93) rotate(-49.48)"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "110.51",
    cy: "11.04",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(30.32 87.88) rotate(-49.48)"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "44.29",
    y: "17.94",
    width: "3.93",
    height: "3.93"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "50.05",
    y: "17.94",
    width: "3.93",
    height: "3.93"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "44.29",
    y: "23.67",
    width: "3.93",
    height: "3.93"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "50.05",
    y: "23.67",
    width: "3.93",
    height: "3.93"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "44.29",
    y: "23.67",
    width: "3.93",
    height: "3.93"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "50.05",
    y: "23.67",
    width: "3.93",
    height: "3.93"
  }))));
}
