// @postman/aether/esmLib/components/Illustration/illustrations/TestFail.js
import React from 'react';
export default function TestFail() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "test-fail-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "80.2",
    y1: "14.57",
    x2: "89.51",
    y2: "14.57",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "106.27",
    y1: "32.58",
    x2: "118.09",
    y2: "32.58",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M76.62,108.28c8.1.52,11,2.5,11,4.4,0,2.91-12.76,5.28-28.5,5.28s-28.51-2.37-28.51-5.28c0-1.46,4.42-3.22,9.6-3.91.71-.1,2.4,6,4.36,5.79,1.31-.12,2.64.76,4.75.64,1.15-.07,5.46-2.36,6-7.45.12-1.14-1,.06.25,0s2.94,0,4.53,0c.91,0,1.54,4.87,4,5.4a50.73,50.73,0,0,0,6.61.3c1.29,0,4.12-5.21,4.65-5.18C76.82,108.36,75,108.18,76.62,108.28Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.32,32.38c.19,6-6.32,12.84-14.6,13.1s-15.2-6.12-15.39-12.17,0-11,14.65-11.41C74,21.47,75.13,26.33,75.32,32.38Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.9,31.91A19.29,19.29,0,1,1,56.62,12.62,19.19,19.19,0,0,1,75.9,31.91Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "57.29",
    y1: "10.13",
    x2: "57.15",
    y2: "12.64"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "57.56",
    cy: "7.28",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.18 63.96) rotate(-84.89)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "57.56",
    cy: "7.28",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.18 63.96) rotate(-84.89)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.85,28.84l2.71,1a1.45,1.45,0,0,1,.81,1.33v2.68a1.45,1.45,0,0,1-1,1.4l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M37.22,27.82,34.58,29a1.45,1.45,0,0,0-.81,1.33v2.69a1.43,1.43,0,0,0,1,1.39l2.8.84"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-5",
    x: "93.69",
    y: "9.75",
    width: "23.76",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-5",
    x: "93.69",
    y: "16.75",
    width: "23.76",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M41.22,44s-19.56,22.25-.71,34.86"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "89.51 12.01 87.59 10.12 84.95 12.66 82.29 9.91 80.41 11.83 83 14.53 80.25 17.16 80.2 17.19 82.27 19.02 84.88 16.49 87.45 19.17 87.48 19.22 89.3 17.15 86.83 14.6 89.51 12.01"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "89.51 12.01 87.59 10.12 84.95 12.66 82.29 9.91 80.41 11.83 83 14.53 80.25 17.16 80.2 17.19 82.27 19.02 84.88 16.49 87.45 19.17 87.48 19.22 89.3 17.15 86.83 14.6 89.51 12.01"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "69.82 17.63 69.82 3.56 124.07 3.56 124.07 28.65 124.07 45.67 100.48 45.83"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "106.27",
    y: "23.23",
    width: "11.82",
    height: "18.71"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "106.25",
    y: "23.29",
    width: "12.21",
    height: "18.38"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "108.3",
    y1: "27.3",
    x2: "116.33",
    y2: "27.3"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "108.3",
    y1: "30.73",
    x2: "116.33",
    y2: "30.73"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "108.3",
    y1: "34.15",
    x2: "116.33",
    y2: "34.15"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "108.3",
    y1: "37.58",
    x2: "116.33",
    y2: "37.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71.45,44.39s8.86,1.58,19.66-3.1a6.51,6.51,0,0,1,8.74,3.93h0a6.52,6.52,0,0,1-3.4,8,70.71,70.71,0,0,1-22.61,5.94"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M39.08,47.18s2.92,3,2.36,12.27l-.93,19.42-.2,26.84c-.14,5.69,2.62,9.09,7.42,9h.54A7.18,7.18,0,0,0,55.41,108l1.8-19.22L62,87"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M41.33,63.5a18.93,18.93,0,0,0,2.25-17.22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M70.8,45l.56,1.77a39.51,39.51,0,0,1,1.87,10.17c0,.27,0,.54,0,.81.91,22.44,2.85,38.05,2.1,49.05a7.62,7.62,0,0,1-7.47,7.11h0a7.61,7.61,0,0,1-7.67-6.73L60,88.63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M67.89,47.86s3.27,4.5,5.41,16"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "62.39",
    y: "54.67",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }))));
}
