// @postman/aether/esmLib/components/Illustration/illustrations/TestPass.js
import React from 'react';
export default function TestPass() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "test-pass-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "80.17",
    y1: "14.05",
    x2: "90.81",
    y2: "14.05",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "105.73",
    y1: "32.85",
    x2: "117.55",
    y2: "32.85",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M75.62,108.28c8.1.52,11,2.5,11,4.4,0,2.91-12.76,5.28-28.5,5.28s-28.51-2.37-28.51-5.28c0-1.46,3.86-3,9-3.86,0,0,3,5.93,4.94,5.74,1.31-.12,2.64.76,4.75.64,1.15-.07,5-2.28,5.49-7.37.12-1.14-.87,0,.44,0C60,107.68,67.23,107.75,75.62,108.28Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "56.62",
    cy: "31.91",
    r: "19.29"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71.14,44.75l1.09,14.37L73,79.48l1,12.77c.65,6-2.14,9.46-5.86,10.36s-8,2-13.35-7.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M39.19,59.73l-.48,24.42L39,107.88a7.36,7.36,0,0,0,7.35,7.36h0a7.36,7.36,0,0,0,7.36-7.36l1.17-15.82c.08-1.63,1-3.29,2.58-3.29h0c1.62,0,2.55,1.61,2.63,3.22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M42.81,45.72c-.13,10.5-3.74,14-3.74,14"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M46.18,48.29S48.63,55.67,39.3,60"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "57.29",
    y1: "10.13",
    x2: "57.15",
    y2: "12.64"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "57.56",
    cy: "7.28",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.18 63.96) rotate(-84.89)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "57.56",
    cy: "7.28",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(45.18 63.96) rotate(-84.89)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.85,28.84l2.71,1a1.45,1.45,0,0,1,.81,1.33v2.68a1.45,1.45,0,0,1-1,1.4l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M37.22,27.82,34.58,29a1.45,1.45,0,0,0-.81,1.33v2.69a1.43,1.43,0,0,0,1,1.39l2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M72.27,60.14c-4-2-4-12.79-4-12.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71.3,44.81c5.47-.35,10.66-5,12.42-8.33a6.73,6.73,0,0,1,8-3.25h0a6.69,6.69,0,0,1,4,9.12C88.75,57.72,72.31,60,72.31,60"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M71.3,44.81c5.47-.35,10.66-5,12.42-8.33a6.73,6.73,0,0,1,8-3.25h0a6.69,6.69,0,0,1,4,9.12C88.75,57.72,72.31,60,72.31,60"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M42.37,44.81c-5.47-.35-10.65-5-12.42-8.33a6.72,6.72,0,0,0-8-3.25h0a6.7,6.7,0,0,0-4.05,9.12c7,15.37,21.18,17.23,21.18,17.23"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.71,84.15s-3.87-3.55.1-6.41"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "95.01 45.83 124.13 45.83 124.07 28.65 124.07 3.56 69.82 3.56 69.82 17.63"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M75.59,32.38c.19,6-6.32,12.84-14.6,13.1S45.8,39.36,45.61,33.31s0-11,14.64-11.41C74.24,21.47,75.4,26.33,75.59,32.38Z"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "60.26",
    y: "55.61",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "82.17 12.25 84.39 14.54 88.57 9.97 90.81 12.17 84.24 18.12 80.17 14.73 82.17 12.25"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-6",
    points: "82.17 12.25 84.39 14.54 88.57 9.97 90.91 12.17 84.41 18.32 79.76 14.39 82.17 12.25"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "94.69",
    y: "9.75",
    width: "23.76",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-6",
    x: "94.69",
    y: "16.75",
    width: "23.76",
    height: "2.51"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-7",
    x: "105.73",
    y: "23.5",
    width: "11.82",
    height: "18.71"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "105.71",
    y: "23.56",
    width: "12.21",
    height: "18.38"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "107.75",
    y1: "27.57",
    x2: "115.78",
    y2: "27.57"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "107.75",
    y1: "30.99",
    x2: "115.78",
    y2: "30.99"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "107.75",
    y1: "34.42",
    x2: "115.78",
    y2: "34.42"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "107.75",
    y1: "37.85",
    x2: "115.78",
    y2: "37.85"
  }))));
}
