// @postman/aether/esmLib/components/Illustration/illustrations/UnableToLoad.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function UnableToLoad() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "unable-to-load-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "43.55",
    y1: "20.52",
    x2: "48.21",
    y2: "20.52",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M97,116c0,1.45-8.46,2.62-24.33,2.62S44,117.49,44,116c0-.72,5.36-2,10.56-2.46,1-.08,3.16,3.85,5.11,3.75,1.31-.08-.15.49,3.15.46,3.51,0,5.23-4.22,5.93-4.24,2.76-.08.93-.12,4-.12,1.48,0,.34,0,1.75,0,.66,0,2.27,2.94,3.49,3.44a11.63,11.63,0,0,0,5,1.18c1.91,0,3.93-2.15,4.89-2.48.67-.22.12-2,.85-1.95C96.81,114.08,97,115.11,97,116Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M80.62,54.39C86.93,51.3,92.13,44,92,36.5a18.75,18.75,0,1,0-18.47,19l2.74-.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M85.25,57.46a29,29,0,0,0,.17-6.39"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M74.25,92.06l.26,19.5A7.4,7.4,0,0,0,82,118.1h0a7.39,7.39,0,0,0,7.27-6.89c.5-7.21-.29-18.88-1.13-33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M55.43,67.06l-.54,15.83-.54,26.24c.14,6.41,2.55,8.85,7.21,8.78h.54a7.23,7.23,0,0,0,6.77-5l1.92-19.13L75.37,91"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M55.69,68.64a18.24,18.24,0,0,0,2.16-20.53"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M88.27,48.72S90,53.83,86.63,60.16"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "35.45",
    cy: "75.76",
    r: "1.2"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "113.43",
    cy: "55.69",
    r: "1.2"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M88.19,48.47s16.14,23.94,3.89,29c-9.5,3.88-13.25-6.75-17.32-16.79a4.49,4.49,0,0,1,1.42-5h0a5.93,5.93,0,0,1,8.54,1.65,102.9,102.9,0,0,0,7.76,10.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M55.57,67a21.82,21.82,0,0,0-.92-16.92"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M50.2,54.47v-5a2.27,2.27,0,0,1,2.27-2.27h5"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M58,47.62a21.37,21.37,0,0,0-7.67,7.23c-5.86,8.74-4.72,21.67,4.18,27.29"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "45.88",
    cy: "20.52",
    r: "2.33"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "45.88",
    cy: "20.52",
    r: "2.33"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M44.09,25.47c-7.88,2.17-14.44,10.86-17.19,15a.47.47,0,0,1-.84-.39C29.26,28.4,41.4,19.4,41.4,19.4"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "42.46 49.82 43.47 50.38 42.46 50.95 41.9 51.97 41.35 50.95 40.33 50.38 41.35 49.82 41.9 48.79 42.46 49.82"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "105.62 39.08 106.63 39.64 105.62 40.2 105.06 41.23 104.51 40.2 103.49 39.64 104.51 39.08 105.06 38.05 105.62 39.08"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-5",
    points: "106.62 89.08 107.63 89.64 106.62 90.2 106.06 91.23 105.51 90.2 104.49 89.64 105.51 89.08 106.06 88.05 106.62 89.08"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-6",
    x1: "73.2",
    y1: "15.25",
    x2: "73.32",
    y2: "17.76"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "73.16",
    cy: "12.39",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(-0.2 1.27) rotate(-0.99)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "73.16",
    cy: "12.39",
    rx: "2.68",
    ry: "2.7",
    transform: "translate(-0.2 1.27) rotate(-0.99)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M91.75,33.22l2.71,1a1.46,1.46,0,0,1,.81,1.33v2.69a1.43,1.43,0,0,1-1,1.39l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M54.8,33.24l-2.64,1.2a1.46,1.46,0,0,0-.81,1.34v2.68a1.43,1.43,0,0,0,1,1.39l2.38.55"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M88.45,40.59c.19,6-6.32,12.84-14.6,13.1s-15.2-6.12-15.39-12.17,0-11,14.65-11.41C87.1,29.68,88.26,34.54,88.45,40.59Z"
  }))));
}
