// @postman/aether/esmLib/components/Illustration/illustrations/UnableToPullRequest.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function UnableToPullRequest() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "unable-to-pull-request-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "102.24",
    y1: "94.63",
    x2: "109.56",
    y2: "94.63",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "15.8",
    y1: "56.9",
    x2: "97.29",
    y2: "56.9",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("polygon", {
    className: "cls-2",
    points: "102.23 83.58 106.76 105.68 109.56 104.39 104.63 83.7 102.23 83.58"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M15.8,49.37s74.13-2.61,81.49,13.71l-1.7,1.51s-3.08-15-79.79-11.17Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M106.16,115.22c0,1.49-10.65,2.7-30.61,2.7s-36.14-1.21-36.14-2.7c0-.74,3.62-1.67,10.16-2.11,1.33-.09,1.61,2.78,4.06,2.67,1.66-.08.44.17,4.59.13,4.42,0,4.59-3.47,5.48-3.49,3.47-.09,8,.11,11.85.11,1.86,0,3.16.06,4.94,0,.83,0,1.8,3.4,3.33,3.92a16.4,16.4,0,0,0,5.4.42c2.39-.05,3-1.66,4.19-2,.83-.23.71-1.89,1.63-1.84C105.27,113.54,106.16,114.26,106.16,115.22Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M97.83,82.22c6.26,7.69-.14,23.28-1.6,28-1.17,3.77-4.34,7-8.29,7,0,0-6.93,1.31-6.63-5.43a27.24,27.24,0,0,1,.45-3.6L85,95,72.47,94l-8,17.51a7.3,7.3,0,0,1-6.69,5h-.51c-4.51.06-10.26-2.63-8.35-9l.05-.15C56.21,88.47,64.3,70.28,75.17,55.07"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "79.09",
    y1: "49.8",
    x2: "76.68",
    y2: "52.59"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "94.12",
    cy: "37.55",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M89.13,72.84l7.24-8.38a5.05,5.05,0,0,1,6.85-1.26h0c2.33,1.53,3.6,5.44,2.08,7.78l-6,9.24a10.22,10.22,0,0,1-10,4c-12-2.65-11.15-28.71-11.15-28.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M103.78,54.6l6.73,6.56a13.69,13.69,0,0,1-.4,15.87l-4.19,5.56a3.1,3.1,0,0,1-.84.78,3.16,3.16,0,0,1-1.23.48,3.24,3.24,0,0,1-2-.27,3.41,3.41,0,0,1-.57-.36h0c-.79-.6-2.11-1.83-2.16-2.76"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M66.9,68.53c.75,0,5.79,3.8,11.84-1.64"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M64.79,72.21c0,.75,8.44,4.44,15.11-.6"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "90.6",
    y1: "60.39",
    x2: "92.57",
    y2: "68.74"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M102.5,83.79c1.3,6.21,2,10.34,4.11,22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M15.8,52.88s54.38-2.25,69.25,4.68c5.29,2.47,8.3,4,10.64,7.22"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M104.77,83.42c1.46,6.39,2.11,8.28,4.44,21.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M89.18,56.23c2.73,1.53,6.91,4.71,8.49,6.83"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M15.64,49.91s48.6-1.62,69.88,4.9"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M99.87,56.18s1.94,4.8-.06,6.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M102.72,55s2,5.11.45,8.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M64.71,71.94a2.2,2.2,0,0,1-3,.59l-4.48-2.76A2,2,0,0,1,56.61,67l8.28-13.31"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M66.06,51.2l3.16-4.66A2,2,0,0,1,72,45.87l4.48,2.77a2.2,2.2,0,0,1,1,2.85"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "113.85",
    y1: "22.26",
    x2: "117.89",
    y2: "15.7"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "115",
    y1: "25.94",
    x2: "120.62",
    y2: "22.45"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "116.13",
    y1: "28.9",
    x2: "123.49",
    y2: "29.15"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M79.09,32.16a.37.37,0,0,0-.66,0c-.73,1.38-2.25,4.79-.11,5.69C79.88,38.49,81.17,36.26,79.09,32.16Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M111.09,36.87a14,14,0,0,1-.84,7.57c-2.11,5.16-6.46,8.37-13.13,8.57-8.28.26-15.19-6.12-15.38-12.16s0-11,14.64-11.42C107.57,29.09,110.33,32.42,111.09,36.87Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-5",
    x1: "94.13",
    y1: "15.7",
    x2: "94.21",
    y2: "18.22"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-6"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-7",
    cx: "94.26",
    cy: "13.04",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "94.15",
    cy: "12.84",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M113.19,34.23l2.34,1.07a1.46,1.46,0,0,1,.81,1.34v2.68a1.42,1.42,0,0,1-1,1.39l-2.19.6"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M74.67,34.1,72,35.3a1.46,1.46,0,0,0-.81,1.34v2.68a1.42,1.42,0,0,0,1,1.39l2.8.85"
  }))));
}
