// @postman/aether/esmLib/components/Illustration/illustrations/UnableToVisualize.js
import React from 'react';
export default function UnableToVisualize() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "unable-to-visualize-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "43.61",
    y1: "71.18",
    x2: "52.81",
    y2: "71.18",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  }))), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M110.94,114c5,.38,7.82,1,7.82,1.71,0,1.15-8.55,2.09-19.1,2.09s-19.09-.94-19.09-2.09c0-.67,2.18-1.19,6.58-1.57L86.7,116,91,116l.5-2.22c2.64-.15,5-.22,8.14-.22,2.64,0,4.66.11,6.94.22l.54,2,4.3.13Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M74.55,116.22c0,1.48-14.2,2.68-31.71,2.68s-31.72-1.2-31.72-2.68c0-.74,4.17-.94,9.87-1.76.62-.09,1.77,3.42,3.92,3.31,1.46-.08,3.16.5,6.65-.41,1.65-.42,2.84-3.9,3.62-3.92,3-.09,4.29.09,7.66.09,1.62,0,2.16.11,3.72.08.72,0,1.51,3.2,2.85,3.71A13.22,13.22,0,0,0,55,118.84c2.1-.06,4.33-2.44,5.4-2.77.73-.23.39-2,1.2-2C70.53,114.54,74.55,115.26,74.55,116.22Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M56.74,55.33C64.4,52.8,70.55,45.5,70.55,37A19.29,19.29,0,0,0,32,37c0,9,6.85,16.59,15.24,18.67A18.74,18.74,0,0,0,56.74,55.33Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M46.54,94.46v18.39c.39,3.88,3.66,5.91,7.56,5.91h0c4,0,7.25-2.1,7.58-6,1-11.45,2.67-37,2.19-61.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M28.81,79.88l-.49,2.43-8,27.75c-1.07,4.67,2.48,8.19,7.28,8.19h.55a7.45,7.45,0,0,0,7-5L43,95.43l4.8-1.71"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M60.15,54c.2.06.52,8.49,3.48,13.12"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M35.14,48.21C22.5,60.76,21,68.63,23.35,73.89a11.07,11.07,0,0,0,3.84,4.47l.19.12A22.41,22.41,0,0,0,39,82.27c.71,0,1.46,0,2.25,0a5.52,5.52,0,0,0,5.49-5.66h0"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M33,50.45s3.55,7.19.94,14.1"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.18,50.83c.41,2,2.14,9.69-.15,13.75"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M26.4,58.63,26.12,49a2.33,2.33,0,0,1,2.33-2.33l5.85-.1"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "51.24",
    y1: "14.84",
    x2: "51.24",
    y2: "17.36"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "51.35",
    cy: "11.98",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(37.66 62.91) rotate(-88.09)"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "51.35",
    cy: "11.98",
    rx: "2.7",
    ry: "2.68",
    transform: "translate(37.66 62.91) rotate(-88.09)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M31,69.09h0A5.72,5.72,0,0,1,33.32,72h0c.34,1.21,1.69,2.34,3.66,3a35.17,35.17,0,0,0,12.83,1.66c10.54-.42,9-5.35,4.2-8.06s-13.74-4.36-20.63-3.41S31,69.09,31,69.09Z"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "37.68",
    cy: "70.95",
    rx: "1.51",
    ry: "2.84",
    transform: "translate(-41.88 82.06) rotate(-69.98)"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M66.56,84a25,25,0,0,1-2.89-4.4"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M64.41,51.89s6,16.4,10.65,23.91"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-5",
    cx: "72.52",
    cy: "81.81",
    r: "6.21"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-3",
    points: "75.5 76.27 80.2 70.64 81.16 71.5 76.7 77.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M69.14,87.06l-2,2.47a1.11,1.11,0,0,1-1.66.11h0a1.12,1.12,0,0,1-.07-1.49l2.23-2.51"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M86.07,64.16a6.56,6.56,0,0,1-4.91,7.34"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M80.2,70.64S78.91,66,85.88,64.11"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M47.77,68a4.11,4.11,0,0,0-3,.14,1.75,1.75,0,0,0-1,2.47,3.29,3.29,0,0,0,2.07,2,4.55,4.55,0,0,1,2,1.29c.32.38,1,.69,2.54.66C53.77,74.51,54.2,70,47.77,68Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M47.77,68a4.11,4.11,0,0,0-3,.14,1.75,1.75,0,0,0-1,2.47,3.29,3.29,0,0,0,2.07,2,4.55,4.55,0,0,1,2,1.29c.32.38,1,.69,2.54.66C53.77,74.51,54.2,70,47.77,68Z"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-7",
    x1: "87.43",
    y1: "82.47",
    x2: "79.61",
    y2: "82.47"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "76.83 74.8 76.83 23.8 122.61 23.8 122.61 82.47 110.61 82.47"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "93.73 84.64 86.4 116.16 91.18 116.16 98.21 84.64"
  }), /*#__PURE__*/React.createElement("polyline", {
    className: "cls-7",
    points: "104.32 84.47 111.73 116.16 106.95 116.16 99.89 84.39"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-8",
    x: "88.27",
    y: "80.46",
    width: "21.69",
    height: "3.95"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M38.28,51.05a10.16,10.16,0,0,1-1.73-5.17c-.19-6.05,0-11,14.64-11.42,14-.43,15.15,4.44,15.34,10.48a10.62,10.62,0,0,1-.86,4.45"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M70.74,33.72l2.35,1.08a1.45,1.45,0,0,1,.81,1.33v2.68a1.45,1.45,0,0,1-1,1.4l-2.8.84"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M31.85,33.59,29.21,34.8a1.45,1.45,0,0,0-.81,1.33v2.68a1.44,1.44,0,0,0,1,1.4l2.79.84"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "52.42",
    y: "59.86",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }))));
}
