// @postman/aether/esmLib/components/Illustration/illustrations/UnauthorizedError.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function UnauthorizedError() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "unauthorized-error-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "92.47",
    y1: "42.32",
    x2: "105.65",
    y2: "42.32",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "94.45",
    y1: "109.58",
    x2: "113.33",
    y2: "109.58",
    xlinkHref: "#linear-gradient"
  }), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-3",
    x1: "88.05",
    y1: "112.93",
    x2: "91.82",
    y2: "112.93",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-2",
    cx: "102.37",
    cy: "114.37",
    rx: "13.27",
    ry: "1.52"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-2",
    d: "M87.73,114.27c2,2.48-12.4,2.52-27.74,2.52s-27.79-1.2-27.79-2.69c0-.74,6.1-1.63,11.13-2.08,1-.09,1.24,2.79,3.12,2.67,1.28-.07.34.17,3.53.14,3.4,0,5.18-3.38,5.86-3.4,2.67-.08,1.2,0,4.15,0,1.42,0,2.43,0,3.8,0,.63,0,1.34,2.27,2.52,2.78a7.76,7.76,0,0,0,4,.71,14.92,14.92,0,0,0,4.61-1.31c.64-.23.57-1.8,1.28-1.76C84,112.34,86.26,112.47,87.73,114.27Z"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "70.15",
    cy: "36.92",
    r: "19.28"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M62.18,89.85l.22,18.53a7.61,7.61,0,0,0,7.67,6.72h0A7.6,7.6,0,0,0,77.54,108c.8-11.49-1.37-28-2.21-52.16"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M42.85,61.65l-1,19.4-.2,24.84c-.14,5.69,2.62,9.09,7.41,9h.55a7.46,7.46,0,0,0,7-5.1l2-18.83,4.78-1.79"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M42.84,65.62c0-.39,6.88-8.35-.34-21.13"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M42.73,67.68s12.78-11.1,5.44-27"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M78.54,54.47a32.59,32.59,0,0,1-2.21,13.38"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M40.53,45.93l-.11-4c0-2.16,1-3.91,2.34-3.91h8.15"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M51.1,39.36S34.6,41.77,34.7,66.13c0,5,.76,11,4.26,14.47A16.5,16.5,0,0,0,42,83"
  }), /*#__PURE__*/React.createElement("rect", {
    className: "cls-3",
    x: "65.73",
    y: "59.19",
    width: "4.93",
    height: "5.07",
    rx: "0.43"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "64.63",
    cy: "39.01",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-3",
    x1: "68.94",
    y1: "14.89",
    x2: "69.02",
    y2: "17.41"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-4"
  }, /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-5",
    cx: "69.07",
    cy: "12.23",
    rx: "2.68",
    ry: "2.7"
  })), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "68.95",
    cy: "12.03",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M76.24,54.82c-3.86-5-2.71-12.31,1.88-16.92C81,35,85.23,34.27,89.33,35.42"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M97.8,63a14.88,14.88,0,0,1-1.35.73C91,66.17,86.37,68.93,76,67.85"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M79.72,53.76s6.83,1.58,14.09-4.25"
  }), /*#__PURE__*/React.createElement("polygon", {
    className: "cls-3",
    points: "92.53 45.7 99.3 65.47 105.61 45.67 92.53 45.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-6",
    d: "M92.47,45.25c0-3.76,6.3-3,7-3.6,5.24-5.25,6.22-.21,6.22,3.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M92.47,45.42c0-3.76,6.3-3,7-3.6,5.24-5.25,6.22-.21,6.22,3.54"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M94.46,111.74A5.39,5.39,0,0,1,98.88,107s1.48-2.29,5.65-1.08a5,5,0,0,1,3.91,5,4.56,4.56,0,0,1,4.85,1.76c.78,1-10.1.8-10.1.8l-7.09.11A1.62,1.62,0,0,1,94.46,111.74Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M94.46,111.74A5.39,5.39,0,0,1,98.88,107s1.48-2.29,5.65-1.08a5,5,0,0,1,3.91,5s2.72-1.43,5.13,2.1c.72,1-10.38.46-10.38.46l-7.09.11A1.62,1.62,0,0,1,94.46,111.74Z"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-8",
    cx: "89.94",
    cy: "112.93",
    rx: "1.89",
    ry: "1.03"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-3",
    cx: "89.94",
    cy: "112.93",
    rx: "1.89",
    ry: "1.03"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-3",
    cx: "104.44",
    cy: "103.45",
    r: "2.15"
  }))));
}
