// @postman/aether/esmLib/components/Illustration/illustrations/Wondering.js
/* eslint-disable react/prop-types */
import React from 'react';
export default function Wondering() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    xmlnsXlink: "http://www.w3.org/1999/xlink",
    viewBox: "0 0 140 120",
    height: "120px",
    width: "140px",
    className: "wondering-illustration"
  }, /*#__PURE__*/React.createElement("defs", null, /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient",
    x1: "96.88",
    y1: "25.94",
    x2: "107.77",
    y2: "25.94",
    gradientUnits: "userSpaceOnUse"
  }, /*#__PURE__*/React.createElement("stop", {
    offset: "0",
    stopColor: "#ff7002"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.3",
    stopColor: "#fe7304"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.5",
    stopColor: "#fc7c0a"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.68",
    stopColor: "#f88b15"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.84",
    stopColor: "#f3a125"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "0.98",
    stopColor: "#ecbd39"
  }), /*#__PURE__*/React.createElement("stop", {
    offset: "1",
    stopColor: "#ebc03b"
  })), /*#__PURE__*/React.createElement("linearGradient", {
    id: "linear-gradient-2",
    x1: "18.85",
    y1: "55.21",
    x2: "37.36",
    y2: "55.21",
    xlinkHref: "#linear-gradient"
  })), /*#__PURE__*/React.createElement("g", {
    className: "cls-1"
  }, /*#__PURE__*/React.createElement("g", {
    id: "Layer_1",
    "data-name": "Layer 1"
  }, /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-3",
    d: "M97,28.78h10.72s.45-3.93-2.79-3.26c0,0,.56-2.71-2.92-2.4a1.35,1.35,0,0,0-.83.37,1.61,1.61,0,0,0-.63,1.16s-2.19-.39-2,1.77A1.61,1.61,0,0,0,97,28.78Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M96.73,28.78h10.72s.45-3.93-2.79-3.26c0,0,.56-2.71-2.92-2.4a1.35,1.35,0,0,0-.83.37,1.61,1.61,0,0,0-.63,1.16s-2.19-.39-2,1.77A1.61,1.61,0,0,0,96.73,28.78Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-5",
    d: "M44.07,114.91c0,1.48,14.33,3.81,31.85,3.81s31.58-2.33,31.58-3.81c0-.75-4.16-1-9.86-1.76-.63-.09-1.78,3.41-3.93,3.3-1.46-.07-3.16.51-6.65-.4-1.64-.42-2.84-3.91-3.62-3.93-3-.08-4.29.1-7.65.1-1.63,0-2.16.1-3.73.08-.72,0-1.51,3.2-2.85,3.71a17.61,17.61,0,0,1-6,1c-2.11-.06-3.9-2-5-2.28-.73-.23-.39-2.05-1.2-2C48.1,113.22,44.07,113.94,44.07,114.91Z"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72,55.12A19,19,0,0,0,86.64,36.69a19.29,19.29,0,0,0-38.57,0c0,8.29,5.86,15.43,13.26,18.14.37.13.74.24,1.11.34l1.46.4"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M72.08,92.38v19.16c-.4,3.88-3.67,5.91-7.57,5.91h0c-3.95,0-7.25-2.1-7.58-6C56,100,54.3,75.29,54.78,51.21"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M89.82,78.57,90.31,81l8,27.75c1.07,4.66-2.48,8.19-7.28,8.19h-.54a7.47,7.47,0,0,1-7-5L75.67,94.12,70.93,91.8"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M58,53.21c-.2.06,0,8.61-3,13.24"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M85.32,64.44,73.06,55.92c-3-2.22-7-3.16-9.26-.25l-.19.3a9.34,9.34,0,0,0-1,3.72c-.14,2.69.9,4.68,3.4,6.42L79.13,76.47a11.76,11.76,0,0,0,5.53,2.29,10.12,10.12,0,0,0,8.6-3.11c4.5-5.13,5.92-13.32-9.47-28.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M85.64,48.93s-3.57,7.84-1,14.74"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M80.14,51.08c-.41,2-1.26,7.25-.37,9.59"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M92.25,56.31l.26-8.67a2.33,2.33,0,0,0-2.34-2.32l-5.27-.27"
  }), /*#__PURE__*/React.createElement("circle", {
    className: "cls-4",
    cx: "78.59",
    cy: "37.4",
    r: "2.88"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M50,28.92c1.9-.47,4.07-.7,8.47-.38,7.66.55,10.81,3.57,12.08,5.57a7.3,7.3,0,0,1,.52,4c-.53,3.86-1.7,13.33-18,11"
  }), /*#__PURE__*/React.createElement("line", {
    className: "cls-4",
    x1: "67.67",
    y1: "14.85",
    x2: "67.59",
    y2: "17.36"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-6",
    cx: "67.66",
    cy: "11.99",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("ellipse", {
    className: "cls-4",
    cx: "67.66",
    cy: "11.99",
    rx: "2.68",
    ry: "2.7"
  }), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M53.73,50s-8.2,11.73-7.17,24.32C47,79.83,50.33,84.46,54.72,85l.67.06"
  }), /*#__PURE__*/React.createElement("g", {
    className: "cls-2"
  }, /*#__PURE__*/React.createElement("path", {
    className: "cls-7",
    d: "M19.11,60H37.33s.77-6.67-4.75-5.53c0,0,1-4.6-5-4.08a2.29,2.29,0,0,0-1.4.63,2.7,2.7,0,0,0-1.07,2s-3.74-.65-3.44,3C21.71,56,17.84,56.06,19.11,60Z"
  })), /*#__PURE__*/React.createElement("path", {
    className: "cls-4",
    d: "M18.81,60H37s.76-6.67-4.75-5.53c0,0,.95-4.6-5-4.08a2.29,2.29,0,0,0-1.4.63,2.7,2.7,0,0,0-1.08,2s-3.73-.65-3.43,3C21.41,56,17.54,56.06,18.81,60Z"
  }))));
}
