// feather/lib/components/searchSelectionDropdown/index.js
"use strict";require("core-js/modules/es.symbol.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.symbol.iterator.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.array.slice.js");require("core-js/modules/es.function.name.js");require("core-js/modules/es.array.from.js");require("core-js/modules/es.object.keys.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");Object.defineProperty(exports, "__esModule", { value: true });exports.SearchSelectionDropdown = void 0;require("core-js/modules/es.string.starts-with.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.array.concat.js");require("core-js/modules/es.array.includes.js");require("core-js/modules/es.string.includes.js");require("core-js/modules/es.number.constructor.js");require("core-js/modules/es.array.map.js");var _react = require("react");










var _reactPopper = require("react-popper");
var _reactTransitionGroup = require("react-transition-group");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _downshift = require("downshift");
var _identity = _interopRequireDefault(require("lodash/identity"));

var _animation = require("../../../lib/animation");
var _dropdownMenuItem = require("../../../lib/components/dropdown/dropdownMenuItem");
var _DropdownMenuItemList = require("../../../lib/components/dropdown/DropdownMenuItemList");
var _styleGenerators = require("../../../lib/components/dropdown/styleGenerators");
var _icon = require("../../../lib/components/icon");
var _ScrollBar = require("../../../lib/components/ScrollBar");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _typography = require("../../../lib/typography");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}function _toConsumableArray(arr) {return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread();}function _nonIterableSpread() {throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");}function _iterableToArray(iter) {if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);}function _arrayWithoutHoles(arr) {if (Array.isArray(arr)) return _arrayLikeToArray(arr);}function _slicedToArray(arr, i) {return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();}function _nonIterableRest() {throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");}function _unsupportedIterableToArray(o, minLen) {if (!o) return;if (typeof o === "string") return _arrayLikeToArray(o, minLen);var n = Object.prototype.toString.call(o).slice(8, -1);if (n === "Object" && o.constructor) n = o.constructor.name;if (n === "Map" || n === "Set") return Array.from(o);if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);}function _arrayLikeToArray(arr, len) {if (len == null || len > arr.length) len = arr.length;for (var i = 0, arr2 = new Array(len); i < len; i++) {arr2[i] = arr[i];}return arr2;}function _iterableToArrayLimit(arr, i) {var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]);if (_i == null) return;var _arr = [];var _n = true;var _d = false;var _s, _e;try {for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {_arr.push(_s.value);if (i && _arr.length === i) break;}} catch (err) {_d = true;_e = err;} finally {try {if (!_n && _i["return"] != null) _i["return"]();} finally {if (_d) throw _e;}}return _arr;}function _arrayWithHoles(arr) {if (Array.isArray(arr)) return arr;}



















var Container = _styledComponents.default.div.withConfig({ displayName: "searchSelectionDropdown__Container", componentId: "z9094r-0" })(["position:relative;"]);



var Label = _styledComponents.default.label.withConfig({ displayName: "searchSelectionDropdown__Label", componentId: "z9094r-1" })(["display:block;height:12px;font-size:12px;font-weight:500;line-height:12px;color:", ";margin-bottom:6px;"],





(0, _cssVariables.default)('neutral-10'));



var ToggleButton = _styledComponents.default.button.withConfig({ displayName: "searchSelectionDropdown__ToggleButton", componentId: "z9094r-2" })(["position:absolute;display:flex;justify-content:center;align-items:center;right:0;top:0;width:36px;height:100%;border:0;padding:0;outline:0;background-color:transparent;cursor:pointer;"]);















var Input = _styledComponents.default.input.withConfig({ displayName: "searchSelectionDropdown__Input", componentId: "z9094r-3" })(["flex:1;color:", ";height:100%;border:0;outline:0;padding:6px 0;margin:0;text-overflow:ellipsis;", ";&::placeholder{color:", ";}"],

(0, _cssVariables.default)('neutral-10'),






_typography.Body['body-short-01'],


(0, _cssVariables.default)('neutral-6'));



var InputWrapper = _styledComponents.default.div.withConfig({ displayName: "searchSelectionDropdown__InputWrapper", componentId: "z9094r-4" })(["display:flex;flex-direction:row;height:", "px;position:relative;padding:0 36px 0 16px;border-radius:4px;border:1px solid ", ";&:hover{border-color:", ";", "{color:", ";}}", " &:focus{box-shadow:0 0 0 2px ", ";}&:focus-within{border-color:", ";", "{color:", ";}}", ""],


function (_ref) {var size = _ref.size;return _styleGenerators.defaultDropdownSizeStyleMap[size].height;},



(0, _cssVariables.default)('neutral-4'),


(0, _cssVariables.default)('purple-7'),

Input,
(0, _cssVariables.default)('purple-7'),



function (_ref2) {var isOpen = _ref2.isOpen;return (
    isOpen && "\n  ".concat(

    ToggleButton, " {\n    transform: rotate(180deg);\n  }\n  "));},





(0, _cssVariables.default)('purple-7'),



(0, _cssVariables.default)('purple-7'),

Input,
(0, _cssVariables.default)('neutral-10'),



function (_ref3) {var hasError = _ref3.hasError;return hasError && "border-color: ".concat((0, _cssVariables.default)('red-5'), ";");});


var ItemList = (0, _styledComponents.default)(_DropdownMenuItemList.DropdownMenuItemList).withConfig({ displayName: "searchSelectionDropdown__ItemList", componentId: "z9094r-5" })(["min-width:100%;transition:opacity 0.1s ", ";opacity:", ";display:", ";"],

_animation.transitionDefault,
function (_ref4) {var state = _ref4.state;return state === 'entered' ? 1 : 0;},
function (_ref5) {var state = _ref5.state;return state === 'exited' ? 'none' : 'block';});


var NoResults = _styledComponents.default.div.withConfig({ displayName: "searchSelectionDropdown__NoResults", componentId: "z9094r-6" })(["display:flex;align-items:center;justify-content:center;height:88px;color:", ";text-align:center;", ";"],




(0, _cssVariables.default)('neutral-5'),

_typography.Body['body-short-01']);


var Error = _styledComponents.default.div.withConfig({ displayName: "searchSelectionDropdown__Error", componentId: "z9094r-7" })(["font-size:12px;line-height:16px;color:", ";margin-top:4px;"],


(0, _cssVariables.default)('red-5'));



var isItemStartsWithInputValue = function isItemStartsWithInputValue(inputValue) {return function (item) {return (
      item.toLowerCase().startsWith(inputValue.toLowerCase()));};};

var defaultNewItemToElement = function defaultNewItemToElement(item) {return "".concat(item, " (new)");};

var SearchSelectionDropdown = function SearchSelectionDropdown(_ref6)














{var label = _ref6.label,placeholder = _ref6.placeholder,initialSelectedItem = _ref6.initialSelectedItem,items = _ref6.items,error = _ref6.error,_ref6$noResultsText = _ref6.noResultsText,noResultsText = _ref6$noResultsText === void 0 ? 'No results' : _ref6$noResultsText,onChange = _ref6.onChange,itemToElementProp = _ref6.itemToElement,_ref6$newItemToElemen = _ref6.newItemToElement,newItemToElement = _ref6$newItemToElemen === void 0 ? defaultNewItemToElement : _ref6$newItemToElemen,_ref6$size = _ref6.size,size = _ref6$size === void 0 ? 'medium' : _ref6$size,_ref6$isCreatable = _ref6.isCreatable,isCreatable = _ref6$isCreatable === void 0 ? false : _ref6$isCreatable,className = _ref6.className,popperProps = _ref6.popperProps,_ref6$disableNoResult = _ref6.disableNoResultsView,disableNoResultsView = _ref6$disableNoResult === void 0 ? false : _ref6$disableNoResult;
  var inputRef = (0, _react.useRef)(null);
  var _useState = (0, _react.useState)(items),_useState2 = _slicedToArray(_useState, 2),inputItems = _useState2[0],setInputItems = _useState2[1];
  var _useCombobox =












  (0, _downshift.useCombobox)({
    items: inputItems,
    initialSelectedItem: initialSelectedItem,
    onInputValueChange: function onInputValueChange(_ref7) {var _ref7$inputValue = _ref7.inputValue,inputValue = _ref7$inputValue === void 0 ? '' : _ref7$inputValue;
      var filteredItems = items.filter(isItemStartsWithInputValue(inputValue));
      setInputItems(
      isCreatable ? [].concat(_toConsumableArray(
      filteredItems), _toConsumableArray(!inputValue || filteredItems.includes(inputValue) ? [] : [inputValue])) :
      filteredItems);

      setHighlightedIndex(0);
    },
    onSelectedItemChange: function onSelectedItemChange(changes) {var _changes$selectedItem;
      onChange === null || onChange === void 0 ? void 0 : onChange((_changes$selectedItem = changes.selectedItem) !== null && _changes$selectedItem !== void 0 ? _changes$selectedItem : null);
    },
    stateReducer: function stateReducer(state, _ref8) {var type = _ref8.type,changes = _ref8.changes;
      if (!state.isOpen && changes.isOpen) {
        // If it's being opened, show all items. If it's creatable, include selected item in the list.
        setInputItems([].concat(_toConsumableArray(
        items), _toConsumableArray(
        state.selectedItem && !items.includes(state.selectedItem) ? [state.selectedItem] : [])));

      }
      if (
      type === _downshift.useCombobox.stateChangeTypes.FunctionOpenMenu ||
      type === _downshift.useCombobox.stateChangeTypes.FunctionToggleMenu)
      {
        // prevent calling openMenu or toggleMenu from updating highlightedIndex
        return _objectSpread(_objectSpread({}, changes), {}, { highlightedIndex: state.highlightedIndex });
      }
      if (type === _downshift.useCombobox.stateChangeTypes.InputBlur) {
        // when input is blurred, reset the input value to selected item if possible.
        return _objectSpread(_objectSpread({}, changes), {}, { inputValue: state.selectedItem || '' });
      }
      if (type === _downshift.useCombobox.stateChangeTypes.InputKeyDownEnter) {var _inputRef$current;
        // when user selects an item by enter key, blur the input.
        (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.blur();
      }
      return changes;
    } }),isOpen = _useCombobox.isOpen,selectedItem = _useCombobox.selectedItem,highlightedIndex = _useCombobox.highlightedIndex,getToggleButtonProps = _useCombobox.getToggleButtonProps,getLabelProps = _useCombobox.getLabelProps,getMenuProps = _useCombobox.getMenuProps,getInputProps = _useCombobox.getInputProps,getComboboxProps = _useCombobox.getComboboxProps,getItemProps = _useCombobox.getItemProps,setHighlightedIndex = _useCombobox.setHighlightedIndex,openMenu = _useCombobox.openMenu,toggleMenu = _useCombobox.toggleMenu;


  (0, _react.useEffect)(function () {
    // if items prop is updated, update the inputItems state.
    if (!isCreatable || inputRef.current == null || items.includes(inputRef.current.value)) {
      setInputItems(items);
      return;
    }
    var inputValue = inputRef.current.value;
    setInputItems([].concat(_toConsumableArray(items), [inputValue]));
  }, [isCreatable, items]);

  var isItemNew = (0, _react.useCallback)(
  function (item) {
    return !items.includes(item);
  },
  [items]);


  var itemToElement = (0, _react.useMemo)(function () {
    if (itemToElementProp == null && newItemToElement == null) {
      return undefined;
    }

    return function (item) {
      var renderer = isItemNew(item) && selectedItem !== item ? newItemToElement : itemToElementProp;
      return (renderer !== null && renderer !== void 0 ? renderer : _identity.default)(item);
    };
  }, [isItemNew, itemToElementProp, newItemToElement, selectedItem]);

  var handleContainerKeyEvent = function handleContainerKeyEvent(event) {var _inputRef$current2, _inputRef$current3;
    (_inputRef$current2 = inputRef.current) === null || _inputRef$current2 === void 0 ? void 0 : _inputRef$current2.focus();
    var inputKeyboardEvent = new KeyboardEvent(event.type, event);

    // prevent the event being propagated to the container back
    inputKeyboardEvent.stopPropagation();

    (_inputRef$current3 = inputRef.current) === null || _inputRef$current3 === void 0 ? void 0 : _inputRef$current3.dispatchEvent(inputKeyboardEvent);
  };

  var handleInputFocus = function handleInputFocus() {
    setHighlightedIndex(selectedItem ? -1 : 0);
    openMenu();
  };

  var handleToggleButtonClick = function handleToggleButtonClick() {
    setHighlightedIndex(-1);
    toggleMenu();
  };

  var _getComboboxProps = getComboboxProps(),comboboxPropsRef = _getComboboxProps.ref,comboboxProps = _objectWithoutProperties(_getComboboxProps, ["ref"]);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(_reactPopper.Manager, { children: /*#__PURE__*/
      (0, _jsxRuntime.jsxs)(Container, { className: className, children: [
        label && /*#__PURE__*/(0, _jsxRuntime.jsx)(Label, _objectSpread(_objectSpread({}, getLabelProps()), {}, { children: label })), /*#__PURE__*/
        (0, _jsxRuntime.jsx)(_reactPopper.Reference, { innerRef: comboboxPropsRef, children:
          function children(_ref9) {var ref = _ref9.ref;
            var inputRefCallback = function inputRefCallback(node) {
              getInputProps().ref(node);
              inputRef.current = node;
            };

            return /*#__PURE__*/(
              (0, _jsxRuntime.jsxs)(InputWrapper, _objectSpread(_objectSpread({},
              comboboxProps), {}, {
                size: size,
                hasError: !!error,
                isOpen: isOpen,
                ref: ref,
                tabIndex: 0,
                onFocus: function onFocus() {return openMenu();},
                onKeyPress: handleContainerKeyEvent,
                onKeyDown: handleContainerKeyEvent,
                onMouseLeave: function onMouseLeave() {
                  setHighlightedIndex(-1);
                },
                "data-test-id": "InputWrapper", children: [/*#__PURE__*/

                (0, _jsxRuntime.jsx)(Input, _objectSpread(_objectSpread({},
                getInputProps()), {}, {
                  ref: inputRefCallback,
                  placeholder: placeholder,
                  "data-test-id": "Input",
                  onFocus: handleInputFocus })), /*#__PURE__*/

                (0, _jsxRuntime.jsx)(ToggleButton, _objectSpread(_objectSpread({},
                getToggleButtonProps()), {}, {
                  type: "button",
                  onClick: handleToggleButtonClick,
                  "aria-label": "toggle menu",
                  "data-test-id": "ToggleButton", children: /*#__PURE__*/

                  (0, _jsxRuntime.jsx)(_icon.Icon, { icon: "input-arrow-down", size: 20, color: (0, _cssVariables.default)('neutral-9') }) }))] })));



          } }), /*#__PURE__*/

        (0, _jsxRuntime.jsx)(_reactTransitionGroup.SwitchTransition, { children: /*#__PURE__*/
          (0, _jsxRuntime.jsx)(_reactTransitionGroup.Transition, { in: isOpen, timeout: 100, unmountOnExit: true, mountOnEnter: true, children:
            function children(state) {
              if (!isOpen || inputItems.length === 0 && disableNoResultsView) {
                return null;
              }

              /**
               * { suppressRefError: true } is added here because I can confirm the component has been working without
               * problems, and the warnings generated by `refKey` check make debugging harder. However, [the Downshift
               * documentation](https://github.com/downshift-js/downshift#getmenuprops) does not recommend to use this
               * option, and it's best to make the implementation follow the original pattern.
               */
              var _getMenuProps = getMenuProps(undefined, { suppressRefError: true }),menuPropsRef = _getMenuProps.ref,menuProps = _objectWithoutProperties(_getMenuProps, ["ref"]);

              return /*#__PURE__*/(
                (0, _jsxRuntime.jsx)(_reactPopper.Popper, _objectSpread(_objectSpread({ placement: "bottom-start" }, popperProps), {}, { innerRef: menuPropsRef, children:
                  function children(_ref10) {var ref = _ref10.ref,style = _ref10.style,placement = _ref10.placement;
                    return /*#__PURE__*/(
                      (0, _jsxRuntime.jsx)(ItemList, _objectSpread(_objectSpread({},
                      menuProps), {}, {
                        isOpen: isOpen,
                        ref: ref,
                        style: style,
                        state: state,
                        "data-placement": placement,
                        "data-test-id": "ItemList", children: /*#__PURE__*/

                        (0, _jsxRuntime.jsx)(_ScrollBar.ScrollBar, { style: { maxHeight: 240, padding: '8px 0' }, children:
                          inputItems.length > 0 ?
                          inputItems.map(function (item, index) {return /*#__PURE__*/(
                              (0, _jsxRuntime.jsx)(_dropdownMenuItem.DropdownMenuItem, {

                                item: item,
                                index: index,
                                selectedItem: selectedItem,
                                highlightedIndex: highlightedIndex,
                                itemToElement: itemToElement,
                                itemProps: getItemProps({
                                  item: item,
                                  index: index,
                                  'aria-selected': selectedItem === item,
                                  'data-test-id': 'Item',
                                  'data-is-highlighted': "".concat(index === highlightedIndex) }) }, "".concat(item).concat(index)));}) : /*#__PURE__*/




                          (0, _jsxRuntime.jsx)(NoResults, { children: noResultsText }) }) })));




                  } })));


            } }, Number(isOpen)) }), /*#__PURE__*/


        (0, _jsxRuntime.jsx)(Error, { "data-test-id": "Error", children: error })] }) }));



};exports.SearchSelectionDropdown = SearchSelectionDropdown;
