// @udacity/ureact-glyph/build/ureact-glyph.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function() {
    function defineProperties(target, props) {
        for (var i = 0; i < props.length; i++) {
            var descriptor = props[i];
            descriptor.enumerable = descriptor.enumerable || false;
            descriptor.configurable = true;
            if ("value" in descriptor) descriptor.writable = true;
            Object.defineProperty(target, descriptor.key, descriptor);
        }
    }
    return function(Constructor, protoProps, staticProps) {
        if (protoProps) defineProperties(Constructor.prototype, protoProps);
        if (staticProps) defineProperties(Constructor, staticProps);
        return Constructor;
    };
}();

var _dec, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactCssModules = require('react-css-modules');

var _reactCssModules2 = _interopRequireDefault(_reactCssModules);

var _ureactGlyphModule = require('./ureact-glyph.module.scss');

var _ureactGlyphModule2 = _interopRequireDefault(_ureactGlyphModule);

var _whitelist = require('./whitelist.json');

var _whitelist2 = _interopRequireDefault(_whitelist);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _possibleConstructorReturn(self, call) {
    if (!self) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return call && (typeof call === "object" || typeof call === "function") ? call : self;
}

function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
}

var UreactGlyph = (_dec = (0, _reactCssModules2.default)(_ureactGlyphModule2.default, {
    allowMultiple: true
}), _dec(_class = (_temp = _class2 = function(_React$Component) {
    _inherits(UreactGlyph, _React$Component);

    function UreactGlyph() {
        _classCallCheck(this, UreactGlyph);

        return _possibleConstructorReturn(this, (UreactGlyph.__proto__ || Object.getPrototypeOf(UreactGlyph)).apply(this, arguments));
    }

    _createClass(UreactGlyph, [{
        key: 'render',
        value: function render() {
            var _props = this.props,
                glyph = _props.glyph,
                text = _props.text,
                className = _props.className,
                style = _props.style;


            if (_ureactGlyphModule2.default['icon-' + glyph]) {
                return _react2.default.createElement(
                    'span', {
                        styleName: 'icon-' + glyph + ' icon',
                        className: className,
                        style: style,
                        title: text
                    },
                    _react2.default.createElement(
                        'i',
                        null,
                        text
                    )
                );
            } else {
                return null;
            }
        }
    }]);

    return UreactGlyph;
}(_react2.default.Component), _class2.propTypes = {
    glyph: _propTypes2.default.oneOf(_whitelist2.default).isRequired,
    text: _propTypes2.default.string,
    className: _propTypes2.default.string,
    style: _propTypes2.default.object
}, _class2.defaultProps = {
    text: '',
    className: null,
    style: null
}, _temp)) || _class);
exports.default = UreactGlyph;
