// @pipedrive/leads-sidebar/dist/components/SidebarFieldWrapper/SidebarFieldWrapper.styles.ts
import styled, { css } from 'styled-components';
import { colors } from '@pipedrive/convention-ui-css/dist/js/variables';

import { EditFieldProps } from '../EditField';

interface StyleProps {
	isActive: boolean;
	hasHover?: boolean;
	isDisabled?: boolean;
}

export const paddingFieldValue = '2px 16px 0 56px';

export const paddingField = css`
	padding: ${paddingFieldValue};
`;

export const IconWrapper = styled.div<{ isHighlight: boolean }>`
	width: 30px;
	height: 30px;
	border-radius: 50%;
	margin-left: 12px;
	position: absolute;
	top: 5px;
	border: 1px solid transparent;
	box-sizing: border-box;
	display: flex;
	align-items: center;
	justify-content: center;

	${({ isHighlight }) =>
		isHighlight &&
		css`
			border-color: ${colors.black8};
		`}

	.cui4-icon {
		width: 16px;
		height: 16px;
		position: relative;
		top: -1px;
	}
`;

export const FieldWrapper = styled.div<StyleProps>`
	box-sizing: border-box;
	display: flex;
	position: relative;

	${({ isActive }) =>
		isActive &&
		css`
			padding-bottom: 10px;
		`}

	${({ isDisabled }) =>
		isDisabled &&
		css`
			cursor: initial;
		`}

	${({ hasHover }) =>
		hasHover &&
		css`
			&:hover {
				background-color: ${colors.black4};
			}
		`}
`;

type ContentWrapperProps = Pick<EditFieldProps, 'fieldType' | 'placement'>;

export const ContentWrapper = styled.div<ContentWrapperProps>`
	position: relative;
	flex: 1;
	display: flex;
	flex-direction: column;
	justify-content: center;
	min-height: 32px;
	padding: 2px 0;
	max-width: 100%;

	.ComponentState_cuiText {
		display: flex;
		width: 100%;
		align-items: center;
	}

	.cui4-text {
		margin-bottom: 0;
		margin-right: 4px;
		overflow: hidden;
	}

	.editMode {
		${paddingField}
		display: flex;
		flex-direction: column;
	}

	.editMode,
	.editMode * {
		box-sizing: border-box;
	}

	.ff__edit {
		${paddingField}
		box-sizing: border-box;
		align-items: flex-start;
	}

	.ff__addValue,
	.ff__placeholder {
		${paddingField}
		display: block;
		flex: 1;
		width: 100%;
		box-sizing: border-box;
		white-space: nowrap;
		overflow: hidden;
		text-overflow: ellipsis;
	}

	.cui4-input {
		flex: 1;
	}

	${({ fieldType, placement }) =>
		fieldType === 'monetary' &&
		placement === 'bottom-end' &&
		css`
			.cui4-select__popup {
				margin-left: 73px;
			}
		`}
`;
