// @atlaskit/smart-card/dist/esm/providers/editor/transformer.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";

var isJiraRoadMap = function isJiraRoadMap(url) {
  return url.match(/^https:\/\/.*?\/jira\/software\/(c\/)?projects\/[^\/]+?\/boards\/.*?\/roadmap\/?/);
};

export var Transformer = /*#__PURE__*/function () {
  function Transformer() {
    _classCallCheck(this, Transformer);
  }

  _createClass(Transformer, [{
    key: "buildInlineAdf",
    value: function buildInlineAdf(url) {
      return {
        type: 'inlineCard',
        attrs: {
          url: url
        }
      };
    }
  }, {
    key: "buildBlockAdf",
    value: function buildBlockAdf(url) {
      return {
        type: 'blockCard',
        attrs: {
          url: url
        }
      };
    }
  }, {
    key: "buildEmbedAdf",
    value: function buildEmbedAdf(url) {
      return {
        type: 'embedCard',
        attrs: {
          url: url,
          layout: 'wide'
        }
      };
    }
  }, {
    key: "toAdf",
    value: function toAdf(url, appearance) {
      if (isJiraRoadMap(url)) {
        return this.buildEmbedAdf(url);
      } else {
        switch (appearance) {
          case 'inline':
            return this.buildInlineAdf(url);

          case 'block':
            return this.buildBlockAdf(url);

          case 'embed':
            return this.buildEmbedAdf(url);
        }
      }
    }
  }]);

  return Transformer;
}();
