// @udacity/ureact-rate-nanodegree/build/services/reviews-service.js
'use strict';

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

var endpoint = '/api/ratings/v1/reviews';

module.exports = {
    get: function get(ndKey, userKey) {
        var opts = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

        return (0, _axios2.default)({
            url: endpoint + '?userKey=' + userKey + '&nodeKey=' + ndKey,
            method: 'get'
        });
    },
    post: function post(_ref) {
        var ndKey = _ref.ndKey,
            userKey = _ref.userKey,
            userName = _ref.userName,
            rating = _ref.rating,
            review = _ref.review,
            locale = _ref.locale,
            jwt = _ref.jwt;
        var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

        var data = {
            nd_key: ndKey,
            rated_by: {
                name: userName
            },
            rating: rating,
            user_key: userKey,
            student_comment: review,
            locale: locale,
            created_at: new Date(),
            updated_at: new Date()
        };

        return (0, _axios2.default)({
            url: endpoint,
            method: 'post',
            data: data,
            headers: {
                Authorization: "Bearer " + jwt
            }
        });
    }
};
