// @pipedrive/pdw-sesheta/src/eventcollection.js
'use strict';

function EventCollection() {
	this.events_ = [];
	this.size_ = 0;
}

EventCollection.prototype.addEvent = function(event) {
	this.events_.push(event);
	this.size_ += event.getSize();
};

EventCollection.prototype.getEvents = function() {
	return this.events_;
};

EventCollection.prototype.getCount = function() {
	return this.events_.length;
};

EventCollection.prototype.getSize = function() {
	return this.size_;
};

EventCollection.prototype.getSizeWith = function(event) {
	return this.getSize() + event.getSize();
};

EventCollection.prototype.getContent = function() {
	var content = '';

	this.events_.forEach(function(event) {
		content += event.getContent();
	});

	return content;
};

EventCollection.prototype.move = function() {
	var collection = new EventCollection();

	collection.events_ = this.events_;
	collection.size_ = this.size_;

	this.reset();

	return collection;
};

EventCollection.prototype.reset = function() {
	this.events_ = [];
	this.size_ = 0;
};

module.exports = EventCollection;
