// @elastic/charts/dist/chart_types/xy_chart/state/selectors/get_specs.js
"use strict";
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSmallMultiplesIndexOrderSelector = exports.getAnnotationSpecsSelector = exports.getSeriesSpecsSelector = exports.getAxisSpecsSelector = void 0;
var __1 = require("../../..");
var constants_1 = require("../../../../specs/constants");
var create_selector_1 = require("../../../../state/create_selector");
var get_settings_specs_1 = require("../../../../state/selectors/get_settings_specs");
var utils_1 = require("../../../../state/utils");
exports.getAxisSpecsSelector = create_selector_1.createCustomCachedSelector([get_settings_specs_1.getSpecs], function (specs) {
    return utils_1.getSpecsFromStore(specs, __1.ChartType.XYAxis, constants_1.SpecType.Axis);
});
exports.getSeriesSpecsSelector = create_selector_1.createCustomCachedSelector([get_settings_specs_1.getSpecs], function (specs) {
    return utils_1.getSpecsFromStore(specs, __1.ChartType.XYAxis, constants_1.SpecType.Series);
});
exports.getAnnotationSpecsSelector = create_selector_1.createCustomCachedSelector([get_settings_specs_1.getSpecs], function (specs) {
    return utils_1.getSpecsFromStore(specs, __1.ChartType.XYAxis, constants_1.SpecType.Annotation);
});
exports.getSmallMultiplesIndexOrderSelector = create_selector_1.createCustomCachedSelector([get_settings_specs_1.getSpecs], function (specs) {
    var _a = __read(utils_1.getSpecsFromStore(specs, __1.ChartType.Global, constants_1.SpecType.SmallMultiples), 1), smallMultiples = _a[0];
    var groupBySpecs = utils_1.getSpecsFromStore(specs, __1.ChartType.Global, constants_1.SpecType.IndexOrder);
    return {
        horizontal: groupBySpecs.find(function (s) { return s.id === (smallMultiples === null || smallMultiples === void 0 ? void 0 : smallMultiples.splitHorizontally); }),
        vertical: groupBySpecs.find(function (s) { return s.id === (smallMultiples === null || smallMultiples === void 0 ? void 0 : smallMultiples.splitVertically); }),
    };
});
