// @toptal/social-urls/dist/builders.js
"use strict";
var __extends = (this && this.__extends) || (function() {
    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({
                    __proto__: []
                }
                instanceof Array && function(d, b) {
                    d.__proto__ = b;
                }) ||
            function(d, b) {
                for (var p in b)
                    if (b.hasOwnProperty(p)) d[p] = b[p];
            };
        return extendStatics(d, b);
    };
    return function(d, b) {
        extendStatics(d, b);

        function __() {
            this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function() {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
/**
 * This is a base URL builder class to inherit.
 * There are three implementations Facebook, LinkedIn and Twitter.
 * Each implementation defines its own host and query properties which
 * are being used in the base class to constuct the result.
 * URL construction logic is common and lives inside the base class.
 */
var URLBuilder = /** @class */ (function() {
    function URLBuilder(props) {
        this.host = "";
        this.query = {};
        this.props = props;
        this.buildQuery();
    }
    URLBuilder.prototype.getUrl = function() {
        return this.host + "?" + this.getQueryString();
    };
    URLBuilder.prototype.buildQuery = function() {
        return;
    };
    URLBuilder.prototype.getQueryString = function() {
        var _this = this;
        return Object.keys(this.query)
            .map(function(key) {
                return encodeURIComponent(key) + "=" + encodeURIComponent(_this.query[key]);
            })
            .join("&");
    };
    return URLBuilder;
}());
/**
 * Facebook URL Builder.
 * ```
 * new FacebookURLBuilder({ url: 'https://toptal.com' }).getUrl()
 * ```
 */
var FacebookURLBuilder = /** @class */ (function(_super) {
    __extends(FacebookURLBuilder, _super);

    function FacebookURLBuilder(props) {
        var _this = _super.call(this, props) || this;
        _this.host = "https://www.facebook.com/sharer.php";
        return _this;
    }
    FacebookURLBuilder.prototype.buildQuery = function() {
        this.query = __assign({
            u: this.props.url
        }, (this.props.title && {
            t: this.props.title
        }), (this.props.description && {
            description: this.props.description
        }), (this.props.picture && {
            description: this.props.picture
        }));
    };
    return FacebookURLBuilder;
}(URLBuilder));
exports.FacebookURLBuilder = FacebookURLBuilder;
/**
 * Linkedin URL Builder.
 * ```
 * new LinkedInURLBuilder({ url: 'https://toptal.com' }).getUrl()
 * ```
 */
var LinkedInURLBuilder = /** @class */ (function(_super) {
    __extends(LinkedInURLBuilder, _super);

    function LinkedInURLBuilder(props) {
        var _this = _super.call(this, props) || this;
        _this.host = "https://www.linkedin.com/shareArticle";
        return _this;
    }
    LinkedInURLBuilder.prototype.buildQuery = function() {
        this.query = __assign({
            mini: "true",
            url: this.props.url
        }, (this.props.title && {
            title: this.props.title
        }), (this.props.description && {
            summary: this.props.description
        }));
    };
    return LinkedInURLBuilder;
}(URLBuilder));
exports.LinkedInURLBuilder = LinkedInURLBuilder;
/**
 * Twitter URL Builder.
 * ```
 * new TwitterURLBuilder({ url: 'https://toptal.com' }).getUrl()
 * ```
 */
var TwitterURLBuilder = /** @class */ (function(_super) {
    __extends(TwitterURLBuilder, _super);

    function TwitterURLBuilder(props) {
        var _this = _super.call(this, props) || this;
        _this.host = "https://twitter.com/intent/tweet";
        return _this;
    }
    TwitterURLBuilder.prototype.buildQuery = function() {
        this.query = __assign({
            url: this.props.url
        }, (this.props.title && {
            text: this.props.title
        }), (this.props.hashtags && {
            hashtags: this.props.hashtags
        }));
    };
    return TwitterURLBuilder;
}(URLBuilder));
exports.TwitterURLBuilder = TwitterURLBuilder;
/**
 * Pocket URL Builder.
 * ```
 * new PocketURLBuilder({ url: 'https://toptal.com' }).getUrl()
 * ```
 */
var PocketURLBuilder = /** @class */ (function(_super) {
    __extends(PocketURLBuilder, _super);

    function PocketURLBuilder(props) {
        var _this = _super.call(this, props) || this;
        _this.host = "https://getpocket.com/save";
        return _this;
    }
    PocketURLBuilder.prototype.buildQuery = function() {
        this.query = {
            url: this.props.url
        };
    };
    return PocketURLBuilder;
}(URLBuilder));
exports.PocketURLBuilder = PocketURLBuilder;
