// @splitsoftware/splitio/lib/storage/ImpressionsCache/InMemory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
Copyright 2016 Split Software

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
var ImpressionsCacheInMemory = /*#__PURE__*/function () {
  function ImpressionsCacheInMemory() {
    (0, _classCallCheck2.default)(this, ImpressionsCacheInMemory);
    this.queue = [];
  }
  /**
   * Get collected data
   */


  (0, _createClass2.default)(ImpressionsCacheInMemory, [{
    key: "state",
    value: function state() {
      return this.queue;
    }
    /**
     * Store objects in sequential order
     */

  }, {
    key: "track",
    value: function track(data) {
      var _this$queue;

      (_this$queue = this.queue).push.apply(_this$queue, (0, _toConsumableArray2.default)(data));

      return this;
    }
    /**
     * Recycle the collector queue
     */

  }, {
    key: "clear",
    value: function clear() {
      this.queue.length = 0;
      return this;
    }
    /**
     * Hook JSON.stringify to expose the state of the counters
     */

  }, {
    key: "toJSON",
    value: function toJSON() {
      return this.queue;
    }
    /**
     * Check if the is data changed from the defaults
     */

  }, {
    key: "isEmpty",
    value: function isEmpty() {
      return this.queue.length === 0;
    }
  }]);
  return ImpressionsCacheInMemory;
}();

var _default = ImpressionsCacheInMemory;
exports.default = _default;
