// @atlaskit/motion/dist/esm/entering/slide-in.js
import { __assign, __rest } from "tslib";
import React from 'react';
import { easeIn, easeOut } from '../utils/curves';
import { mediumDurationMs } from '../utils/durations';
import KeyframesMotion from './keyframes-motion';
export var slideInAnimation = function (from, state, fade) {
    var _a;
    var initial = state === 'entering' ? '0%' : '100%';
    var end = state === 'entering' ? '100%' : '0%';
    var fromMap = {
        top: 'translate3d(0, -100%, 0)',
        right: 'translate3d(100%, 0, 0)',
        bottom: 'translate3d(0, 100%, 0)',
        left: 'translate3d(-100%, 0, 0)',
    };
    var animateOpacity = fade === 'inout' ||
        (fade === 'in' && state === 'entering') ||
        (fade === 'out' && state === 'exiting');
    return _a = {},
        _a[initial] = {
            transform: fromMap[from],
            opacity: animateOpacity ? 0 : 1,
        },
        _a[end] = {
            transform: 'none',
            opacity: 1,
        },
        _a;
};
var SlideIn = function (_a) {
    var children = _a.children, enterFrom = _a.enterFrom, exitTo = _a.exitTo, _b = _a.fade, fade = _b === void 0 ? 'none' : _b, _c = _a.duration, duration = _c === void 0 ? mediumDurationMs : _c, _d = _a.animationTimingFunction, animationTimingFunction = _d === void 0 ? function (state) { return (state === 'entering' ? easeOut : easeIn); } : _d, props = __rest(_a, ["children", "enterFrom", "exitTo", "fade", "duration", "animationTimingFunction"]);
    return (React.createElement(KeyframesMotion, __assign({ duration: duration, enteringAnimation: slideInAnimation(enterFrom, 'entering', fade), exitingAnimation: slideInAnimation(exitTo || enterFrom, 'exiting', fade), animationTimingFunction: animationTimingFunction }, props), children));
};
export default SlideIn;
