// feather/lib/components/splitButton/index.js
"use strict";require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");Object.defineProperty(exports, "__esModule", { value: true });exports.SplitButton = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));

var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));


var _animation = require("../../animation");
var _button = require("../button");
var _dropdown = require("../dropdown");
var _icon = require("../icon");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}

var StyledSplitButton = _styledComponents.default.div.withConfig({ displayName: "splitButton__StyledSplitButton", componentId: "sc-1dtj5w4-0" })(["display:flex;align-items:center;& > button:first-child{border-top-right-radius:0;border-bottom-right-radius:0;}.splitButtonRight,.splitButtonDropdown{width:", ";border-top-left-radius:0;border-bottom-left-radius:0;}.splitButtonDropdown{height:", ";svg{transition:fill 0.2s ", ";}}"],









function (_ref) {var size = _ref.size;return size === 'small' ? '20px' : '16px';},




function (_ref2) {var size = _ref2.size;return size === 'small' ? '32px' : '24px';},

_animation.transitionDefault);




var SplitButton = function SplitButton(_ref3)





{var _ref3$buttonType = _ref3.buttonType,buttonType = _ref3$buttonType === void 0 ? 'secondary' : _ref3$buttonType,_ref3$size = _ref3.size,size = _ref3$size === void 0 ? 'small' : _ref3$size,left = _ref3.left,right = _ref3.right,_ref3$dropdown = _ref3.dropdown,dropdown = _ref3$dropdown === void 0 ? null : _ref3$dropdown;
  var renderToggle = function renderToggle(_ref4) {var isOpen = _ref4.isOpen;
    var iconProps = isOpen ? { color: (0, _cssVariables.default)('purple-7') } : {};
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_icon.Icon, _objectSpread({ icon: right.icon, size: size === 'small' ? 20 : 16 }, iconProps));
  };
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(StyledSplitButton, { size: size, children: [/*#__PURE__*/
      (0, _jsxRuntime.jsx)(_button.IconButton, _objectSpread({ buttonType: buttonType, size: size }, left)),
      dropdown ? /*#__PURE__*/
      (0, _jsxRuntime.jsx)(_dropdown.Dropdown, _objectSpread(_objectSpread({
        toggleRenderer: renderToggle,
        variant: "inline",
        size: "small",
        showArrow: false },
      dropdown), {}, {
        className: "splitButtonDropdown".concat(dropdown.className ? " ".concat(dropdown.className) : '') })) : /*#__PURE__*/


      (0, _jsxRuntime.jsx)(_button.IconButton, _objectSpread(_objectSpread({
        buttonType: buttonType,
        size: size },
      right), {}, {
        className: "splitButtonRight".concat(right.className ? " ".concat(right.className) : '') }))] }));




};exports.SplitButton = SplitButton;
