// @udacity/web-perf-metrics/dist/index.js
! function(t, e) {
    if ("object" == typeof exports && "object" == typeof module) module.exports = e();
    else if ("function" == typeof define && define.amd) define([], e);
    else {
        var n = e();
        for (var i in n)("object" == typeof exports ? exports : t)[i] = n[i]
    }
}(this, (function() {
    return function(t) {
        var e = {};

        function n(i) {
            if (e[i]) return e[i].exports;
            var r = e[i] = {
                i: i,
                l: !1,
                exports: {}
            };
            return t[i].call(r.exports, r, r.exports, n), r.l = !0, r.exports
        }
        return n.m = t, n.c = e, n.d = function(t, e, i) {
            n.o(t, e) || Object.defineProperty(t, e, {
                enumerable: !0,
                get: i
            })
        }, n.r = function(t) {
            "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(t, Symbol.toStringTag, {
                value: "Module"
            }), Object.defineProperty(t, "__esModule", {
                value: !0
            })
        }, n.t = function(t, e) {
            if (1 & e && (t = n(t)), 8 & e) return t;
            if (4 & e && "object" == typeof t && t && t.__esModule) return t;
            var i = Object.create(null);
            if (n.r(i), Object.defineProperty(i, "default", {
                    enumerable: !0,
                    value: t
                }), 2 & e && "string" != typeof t)
                for (var r in t) n.d(i, r, function(e) {
                    return t[e]
                }.bind(null, r));
            return i
        }, n.n = function(t) {
            var e = t && t.__esModule ? function() {
                return t.default
            } : function() {
                return t
            };
            return n.d(e, "a", e), e
        }, n.o = function(t, e) {
            return Object.prototype.hasOwnProperty.call(t, e)
        }, n.p = "", n(n.s = 0)
    }([function(t, e, n) {
        "use strict";
        n.r(e);
        var i, r, o = function() {
                return "".concat(Date.now(), "-").concat(Math.floor(8999999999999 * Math.random()) + 1e12)
            },
            a = function(t) {
                var e = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : -1;
                return {
                    name: t,
                    value: e,
                    delta: 0,
                    entries: [],
                    id: o(),
                    isFinal: !1
                }
            },
            u = function(t, e) {
                try {
                    if (PerformanceObserver.supportedEntryTypes.includes(t)) {
                        var n = new PerformanceObserver((function(t) {
                            return t.getEntries().map(e)
                        }));
                        return n.observe({
                            type: t,
                            buffered: !0
                        }), n
                    }
                } catch (t) {}
            },
            c = !1,
            l = !1,
            f = function(t) {
                c = !t.persisted
            },
            s = function() {
                addEventListener("pagehide", f), addEventListener("beforeunload", (function() {}))
            },
            d = function(t) {
                var e = arguments.length > 1 && void 0 !== arguments[1] && arguments[1];
                l || (s(), l = !0), addEventListener("visibilitychange", (function(e) {
                    var n = e.timeStamp;
                    "hidden" === document.visibilityState && t({
                        timeStamp: n,
                        isUnloading: c
                    })
                }), {
                    capture: !0,
                    once: e
                })
            },
            p = function(t, e, n, i) {
                var r;
                return function() {
                    n && e.isFinal && n.disconnect(), e.value >= 0 && (i || e.isFinal || "hidden" === document.visibilityState) && (e.delta = e.value - (r || 0), (e.delta || e.isFinal || void 0 === r) && (t(e), r = e.value))
                }
            },
            v = function(t) {
                var e, n = arguments.length > 1 && void 0 !== arguments[1] && arguments[1],
                    i = a("CLS", 0),
                    r = function(t) {
                        t.hadRecentInput || (i.value += t.value, i.entries.push(t), e())
                    },
                    o = u("layout-shift", r);
                o && (e = p(t, i, o, n), d((function(t) {
                    var n = t.isUnloading;
                    o.takeRecords().map(r), n && (i.isFinal = !0), e()
                })))
            },
            m = function() {
                return void 0 === i && (i = "hidden" === document.visibilityState ? 0 : 1 / 0, d((function(t) {
                    var e = t.timeStamp;
                    return i = e
                }), !0)), {
                    get timeStamp() {
                        return i
                    }
                }
            },
            y = function(t) {
                var e, n = a("FCP"),
                    i = m(),
                    r = u("paint", (function(t) {
                        "first-contentful-paint" === t.name && t.startTime < i.timeStamp && (n.value = t.startTime, n.isFinal = !0, n.entries.push(t), e())
                    }));
                r && (e = p(t, n, r))
            },
            h = function(t) {
                var e = a("FID"),
                    n = m(),
                    i = function(t) {
                        t.startTime < n.timeStamp && (e.value = t.processingStart - t.startTime, e.entries.push(t), e.isFinal = !0, o())
                    },
                    r = u("first-input", i),
                    o = p(t, e, r);
                r ? d((function() {
                    r.takeRecords().map(i), r.disconnect()
                }), !0) : window.perfMetrics && window.perfMetrics.onFirstInputDelay && window.perfMetrics.onFirstInputDelay((function(t, i) {
                    i.timeStamp < n.timeStamp && (e.value = t, e.isFinal = !0, e.entries = [{
                        entryType: "first-input",
                        name: i.type,
                        target: i.target,
                        cancelable: i.cancelable,
                        startTime: i.timeStamp,
                        processingStart: i.timeStamp + t
                    }], o())
                }))
            },
            b = function() {
                return r || (r = new Promise((function(t) {
                    return ["scroll", "keydown", "pointerdown"].map((function(e) {
                        addEventListener(e, t, {
                            once: !0,
                            passive: !0,
                            capture: !0
                        })
                    }))
                }))), r
            },
            w = function(t) {
                var e, n = arguments.length > 1 && void 0 !== arguments[1] && arguments[1],
                    i = a("LCP"),
                    r = m(),
                    o = function(t) {
                        var n = t.startTime;
                        n < r.timeStamp ? (i.value = n, i.entries.push(t)) : i.isFinal = !0, e()
                    },
                    c = u("largest-contentful-paint", o);
                if (c) {
                    e = p(t, i, c, n);
                    var l = function() {
                        i.isFinal || (c.takeRecords().map(o), i.isFinal = !0, e())
                    };
                    b().then(l), d(l, !0)
                }
            },
            g = function(t, e, n, i) {
                return new(n || (n = Promise))((function(r, o) {
                    function a(t) {
                        try {
                            c(i.next(t))
                        } catch (t) {
                            o(t)
                        }
                    }

                    function u(t) {
                        try {
                            c(i.throw(t))
                        } catch (t) {
                            o(t)
                        }
                    }

                    function c(t) {
                        var e;
                        t.done ? r(t.value) : (e = t.value, e instanceof n ? e : new n((function(t) {
                            t(e)
                        }))).then(a, u)
                    }
                    c((i = i.apply(t, e || [])).next())
                }))
            },
            S = function(t, e) {
                var n, i, r, o, a = {
                    label: 0,
                    sent: function() {
                        if (1 & r[0]) throw r[1];
                        return r[1]
                    },
                    trys: [],
                    ops: []
                };
                return o = {
                    next: u(0),
                    throw: u(1),
                    return: u(2)
                }, "function" == typeof Symbol && (o[Symbol.iterator] = function() {
                    return this
                }), o;

                function u(o) {
                    return function(u) {
                        return function(o) {
                            if (n) throw new TypeError("Generator is already executing.");
                            for (; a;) try {
                                if (n = 1, i && (r = 2 & o[0] ? i.return : o[0] ? i.throw || ((r = i.return) && r.call(i), 0) : i.next) && !(r = r.call(i, o[1])).done) return r;
                                switch (i = 0, r && (o = [2 & o[0], r.value]), o[0]) {
                                    case 0:
                                    case 1:
                                        r = o;
                                        break;
                                    case 4:
                                        return a.label++, {
                                            value: o[1],
                                            done: !1
                                        };
                                    case 5:
                                        a.label++, i = o[1], o = [0];
                                        continue;
                                    case 7:
                                        o = a.ops.pop(), a.trys.pop();
                                        continue;
                                    default:
                                        if (!(r = a.trys, (r = r.length > 0 && r[r.length - 1]) || 6 !== o[0] && 2 !== o[0])) {
                                            a = 0;
                                            continue
                                        }
                                        if (3 === o[0] && (!r || o[1] > r[0] && o[1] < r[3])) {
                                            a.label = o[1];
                                            break
                                        }
                                        if (6 === o[0] && a.label < r[1]) {
                                            a.label = r[1], r = o;
                                            break
                                        }
                                        if (r && a.label < r[2]) {
                                            a.label = r[2], a.ops.push(o);
                                            break
                                        }
                                        r[2] && a.ops.pop(), a.trys.pop();
                                        continue
                                }
                                o = e.call(t, a)
                            } catch (t) {
                                o = [6, t], i = 0
                            } finally {
                                n = r = 0
                            }
                            if (5 & o[0]) throw o[1];
                            return {
                                value: o[0] ? o[1] : void 0,
                                done: !0
                            }
                        }([o, u])
                    }
                }
            };
        e.default = {
            track: function(t) {
                return g(void 0, void 0, void 0, (function() {
                    var e, n, i, r, o, a, u, c, l, f;
                    return S(this, (function(s) {
                        switch (s.label) {
                            case 0:
                                return window.analytics ? (e = !1, n = !1, i = document.cookie.indexOf("_jwt") >= 0, r = {
                                    fullPageUrl: window.location.href,
                                    siteIdentifier: t,
                                    userStatus: i ? "logged_in" : "logged_out"
                                }, a = (o = function(t, e) {
                                    return g(void 0, void 0, void 0, (function() {
                                        return S(this, (function(n) {
                                            return [2, new Promise((function(n) {
                                                t((function(t) {
                                                    var i = Math.round(100 * t.value) / 100;
                                                    r[e] = i, n(i)
                                                }))
                                            }))]
                                        }))
                                    }))
                                })(y, "firstContentfulPaint"), u = o(w, "largestContentfulPaint"), c = o(h, "firstInputDelay"), l = function() {
                                    e || (window.analytics.track("performance", Object.assign({}, r)), e = !0, r.cumulativeLayoutShift && (n = !0))
                                }, f = Promise.all([a, u, c]).then(l), o(v, "cumulativeLayoutShift").then((function(r) {
                                    e && !n ? window.analytics.track("performance", {
                                        cumulativeLayoutShift: r,
                                        fullPageUrl: window.location.href,
                                        siteIdentifier: t,
                                        userStatus: i ? "logged_in" : "logged_out"
                                    }) : e || l(), n = !0
                                })), window.addEventListener("beforeunload", l), [4, f]) : (console.error("Could not track performance metrics: window.analytics is not defined"), [2]);
                            case 1:
                                return s.sent(), [2]
                        }
                    }))
                }))
            }
        }
    }])
}));
