// @pipedrive/leads-sidebar/dist/components/CustomField/types/CustomFieldLargeText.tsx
import React from 'react';
import { createFragmentContainer, graphql } from '@pipedrive/relay';

import { CustomField } from '../CustomField';
import { FormValue } from '../../../types/types';

import type { CustomFieldLargeText_data } from './__generated__/CustomFieldLargeText_data.graphql';

type CustomFieldLargeTextInput = {
	id: string;
	text?: string | null;
};

type Props = {
	readonly data: CustomFieldLargeText_data;
	readonly onUpdate: (input: { CustomFieldLargeText: CustomFieldLargeTextInput }) => Promise<boolean>;
};

const CustomFieldLargeTextWithoutData: React.FC<Props> = ({ data, onUpdate }) => {
	function handleUpdate(rawFormFieldData: { [customFieldId: string]: FormValue }) {
		const [id, text] = Object.entries(rawFormFieldData)[0];

		if (typeof text !== 'string') {
			return Promise.resolve(false);
		}

		return onUpdate({
			CustomFieldLargeText: {
				id,
				text,
			},
		});
	}

	return <CustomField icon="largetext" type="text" value={data.text} data={data} onUpdate={handleUpdate} />;
};

export const CustomFieldLargeText = createFragmentContainer(CustomFieldLargeTextWithoutData, {
	data: graphql`
		fragment CustomFieldLargeText_data on CustomFieldLargeText {
			...CustomField_data
			text
		}
	`,
});
