// @postman/aether/esmLib/components/Button/SplitButton.js
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React, { useRef } from 'react';
import styled from 'styled-components';
import propTypes from 'prop-types';
import Popover from '../Popover/Popover';
import Button from './Button';
import { LIST_OF_ICONS } from '../../design-tokens/Icons';
/**
 * Function to get the CSS for split-button's visual appeareance based on the type prop
 * @param {string} type - defines the variant of the button. Eg: primary, secondary, tertiary
 * @param {object} theme - theme object containing theme specific design tokens
 */

function getSplitButtonStyles(type, theme) {
  switch (type) {
    case 'primary':
      return "\n        background-color: ".concat(theme['button-primary-background-color'], ";\n\n        & .aether-split-button__splitter {\n          border-left-color: ").concat(theme['splitbutton-primary-splitter-color'], ";\n        }\n      ");

    case 'secondary':
      return "\n        background-color: ".concat(theme['button-secondary-background-color'], ";\n\n        & .aether-split-button__splitter {\n          border-left-color: ").concat(theme['splitbutton-secondary-splitter-color'], ";\n        }\n      ");

    case 'tertiary':
      return "\n        background-color: ".concat(theme['button-tertiary-background-color'], ";\n\n        & .aether-split-button__splitter {\n          border-left-color: ").concat(theme['splitbutton-tertiary-splitter-color'], ";\n        }\n      ");

    default:
      return null;
  }
}
/**
 * Function to render the secondary button of the SplitButton component
 *
 * Popover component needs the ref of the dropdown button to define its position
 * So, here we are chcecking if the consumer has passed a custom secondary button then give
 * them control to define and pass ref to the Popover component or, if they haven't pass then
 * render a default dropdown button as passed its ref as Popover trigger.
 *
 * @param {node} secondaryButton - defines the custom Button component
 * @param {string} type - defines the variant of the button. Eg: primary, secondary, tertiary
 * @param {object} dropdownButtonRef - the ref of the default dropdown button
 */


function renderSecondaryButton(secondaryButton, type, dropdownButtonRef) {
  if (secondaryButton && secondaryButton.type === Button) {
    var icon = secondaryButton.props.icon;
    return /*#__PURE__*/React.cloneElement(secondaryButton, {
      type: type,
      icon: icon || 'icon-direction-down',
      className: 'aether-split-button__secondary-btn'
    });
  }

  return /*#__PURE__*/React.createElement(Button, {
    type: type,
    icon: "icon-direction-down",
    ref: dropdownButtonRef,
    className: "aether-split-button__secondary-btn"
  });
}

var StyledSplitButton = styled.div.withConfig({
  displayName: "SplitButton__StyledSplitButton",
  componentId: "sc-iap9jj-0"
})(["width:max-content;box-sizing:border-box;cursor:pointer;user-select:none;display:inline-flex;align-items:center;border-radius:", ";height:", ";", ";& .aether-split-button__primary-btn:disabled{border-top-right-radius:0;border-bottom-right-radius:0;}& .aether-split-button__secondary-btn:disabled{border-top-left-radius:0;border-bottom-left-radius:0;}&:hover{& .aether-split-button__splitter{border-left-color:transparent;}}&:active{& .aether-split-button__splitter{border-left-color:transparent;}}"], function (props) {
  return props.theme['border-radius-default'];
}, function (props) {
  return props.theme['size-m'];
}, function (props) {
  return getSplitButtonStyles(props.type, props.theme);
});
var StyledButtonSplitter = styled.div.withConfig({
  displayName: "SplitButton__StyledButtonSplitter",
  componentId: "sc-iap9jj-1"
})(["border-width:", ";border-left-style:", ";height:20px;"], function (props) {
  return "".concat(props.theme['border-width-default']);
}, function (props) {
  return "".concat(props.theme['border-style-solid']);
});

function SplitButton(props) {
  var dropdownButtonRef = useRef();
  var text = props.text,
      onClick = props.onClick,
      children = props.children,
      type = props.type,
      icon = props.icon,
      tooltip = props.tooltip,
      minWidth = props.minWidth,
      className = props.className,
      isLoading = props.isLoading,
      isDisabled = props.isDisabled,
      secondaryButton = props.secondaryButton;
  var clonedChildren = React.Children.map(children, function (child) {
    if ( /*#__PURE__*/React.isValidElement(child) && child.type === Popover) {
      return /*#__PURE__*/React.cloneElement(child, _objectSpread({
        placement: 'bottom-end'
      }, !secondaryButton && {
        triggerRef: dropdownButtonRef
      }));
    }

    return child;
  });
  return /*#__PURE__*/React.createElement(StyledSplitButton, {
    type: type,
    className: className
  }, /*#__PURE__*/React.createElement(Button, {
    type: type,
    text: text,
    icon: icon,
    isLoading: isLoading,
    isDisabled: isDisabled,
    minWidth: minWidth,
    onClick: onClick,
    tooltip: tooltip,
    className: "aether-split-button__primary-btn"
  }), !isDisabled && !isLoading && /*#__PURE__*/React.createElement(StyledButtonSplitter, {
    className: "aether-split-button__splitter"
  }), renderSecondaryButton(secondaryButton, type, dropdownButtonRef), clonedChildren);
}

SplitButton.displayName = 'SplitButton';
export default SplitButton;
SplitButton.defaultProps = {
  text: '',
  type: 'secondary',
  icon: null,
  tooltip: '',
  minWidth: '',
  className: '',
  isLoading: false,
  isDisabled: false,
  secondaryButton: null
};
SplitButton.propTypes = {
  text: propTypes.string,
  onClick: propTypes.func.isRequired,
  children: propTypes.node.isRequired,
  type: propTypes.oneOf(['primary', 'secondary', 'tertiary']),
  icon: propTypes.oneOf(LIST_OF_ICONS),
  tooltip: propTypes.oneOfType([propTypes.string, propTypes.object]),
  minWidth: propTypes.string,
  className: propTypes.string,
  isLoading: propTypes.bool,
  isDisabled: propTypes.bool,
  secondaryButton: propTypes.node
};
