// libphonenumber-js/es6/findPhoneNumbersInText.js
function _typeof(obj) {
    if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
        _typeof = function _typeof(obj) {
            return typeof obj;
        };
    } else {
        _typeof = function _typeof(obj) {
            return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
        };
    }
    return _typeof(obj);
}

function _objectSpread(target) {
    for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === 'function') {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _defineProperty(target, key, source[key]);
        });
    }
    return target;
}

function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}

import findNumbers from './findNumbers';
export default function findPhoneNumbersInText(text, defaultCountry, options, metadata) {
    var args = getArguments(defaultCountry, options, metadata);
    return findNumbers(text, args.options, args.metadata);
}
export function getArguments(defaultCountry, options, metadata) {
    if (metadata) {
        if (defaultCountry) {
            options = _objectSpread({}, options, {
                defaultCountry: defaultCountry
            });
        }
    } else {
        if (options) {
            metadata = options;

            if (defaultCountry) {
                if (is_object(defaultCountry)) {
                    options = defaultCountry;
                } else {
                    options = {
                        defaultCountry: defaultCountry
                    };
                }
            } else {
                options = undefined;
            }
        } else {
            metadata = defaultCountry;
            options = undefined;
        }
    }

    return {
        options: _objectSpread({}, options, {
            v2: true
        }),
        metadata: metadata
    };
} // Babel transforms `typeof` into some "branches"
// so istanbul will show this as "branch not covered".

/* istanbul ignore next */

var is_object = function is_object(_) {
    return _typeof(_) === 'object';
};
