// libphonenumber-js/es6/PhoneNumber.js
function _objectSpread(target) {
    for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === 'function') {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _defineProperty(target, key, source[key]);
        });
    }
    return target;
}

function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}

import Metadata from './metadata';
import isPossibleNumber from './isPossibleNumber_';
import isValidNumber from './validate_';
import isValidNumberForRegion from './isValidNumberForRegion_';
import getNumberType from './getNumberType_';
import formatNumber from './format_';
var USE_NON_GEOGRAPHIC_COUNTRY_CODE = false;

var PhoneNumber =
    /*#__PURE__*/
    function() {
        function PhoneNumber(countryCallingCode, nationalNumber, metadata) {
            _classCallCheck(this, PhoneNumber);

            if (!countryCallingCode) {
                throw new TypeError('`country` or `countryCallingCode` not passed');
            }

            if (!nationalNumber) {
                throw new TypeError('`nationalNumber` not passed');
            }

            var _metadata = new Metadata(metadata); // If country code is passed then derive `countryCallingCode` from it.
            // Also store the country code as `.country`.


            if (isCountryCode(countryCallingCode)) {
                this.country = countryCallingCode;

                _metadata.country(countryCallingCode);

                countryCallingCode = _metadata.countryCallingCode();
            } else {
                /* istanbul ignore if */
                if (USE_NON_GEOGRAPHIC_COUNTRY_CODE) {
                    if (_metadata.isNonGeographicCallingCode(countryCallingCode)) {
                        this.country = '001';
                    }
                }
            }

            this.countryCallingCode = countryCallingCode;
            this.nationalNumber = nationalNumber;
            this.number = '+' + this.countryCallingCode + this.nationalNumber;
            this.metadata = metadata;
        }

        _createClass(PhoneNumber, [{
            key: "isPossible",
            value: function isPossible() {
                return isPossibleNumber(this, {
                    v2: true
                }, this.metadata);
            }
        }, {
            key: "isValid",
            value: function isValid() {
                return isValidNumber(this, {
                    v2: true
                }, this.metadata);
            }
        }, {
            key: "isNonGeographic",
            value: function isNonGeographic() {
                var metadata = new Metadata(this.metadata);
                return metadata.isNonGeographicCallingCode(this.countryCallingCode);
            }
        }, {
            key: "isEqual",
            value: function isEqual(phoneNumber) {
                return this.number === phoneNumber.number && this.ext === phoneNumber.ext;
            } // // Is just an alias for `this.isValid() && this.country === country`.
            // // https://github.com/googlei18n/libphonenumber/blob/master/FAQ.md#when-should-i-use-isvalidnumberforregion
            // isValidForRegion(country) {
            // 	return isValidNumberForRegion(this, country, { v2: true }, this.metadata)
            // }

        }, {
            key: "getType",
            value: function getType() {
                return getNumberType(this, {
                    v2: true
                }, this.metadata);
            }
        }, {
            key: "format",
            value: function format(_format, options) {
                return formatNumber(this, _format, options ? _objectSpread({}, options, {
                    v2: true
                }) : {
                    v2: true
                }, this.metadata);
            }
        }, {
            key: "formatNational",
            value: function formatNational(options) {
                return this.format('NATIONAL', options);
            }
        }, {
            key: "formatInternational",
            value: function formatInternational(options) {
                return this.format('INTERNATIONAL', options);
            }
        }, {
            key: "getURI",
            value: function getURI(options) {
                return this.format('RFC3966', options);
            }
        }]);

        return PhoneNumber;
    }();

export {
    PhoneNumber as
    default
};

var isCountryCode = function isCountryCode(value) {
    return /^[A-Z]{2}$/.test(value);
};
