// @postman/data-editor/lib/components/base/NumberInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactShortcuts = require('react-shortcuts');

var _utils = require('../../utilities/utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NumberInput = function (_Component) {
  _inherits(NumberInput, _Component);

  function NumberInput(props) {
    _classCallCheck(this, NumberInput);

    var _this = _possibleConstructorReturn(this, (NumberInput.__proto__ || Object.getPrototypeOf(NumberInput)).call(this, props));

    _this.handleFocus = _this.handleFocus.bind(_this);
    _this.handleChange = _this.handleChange.bind(_this);
    _this.handleShortcut = _this.handleShortcut.bind(_this);
    return _this;
  }

  _createClass(NumberInput, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.subscription = this.context.events.addListener('focusCell', this.handleFocus);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.subscription.remove();
    }
  }, {
    key: 'handleFocus',
    value: function handleFocus(_ref) {
      var node = _ref.node,
          column = _ref.column;

      if (node !== this.props.cellInfo.node || column !== this.props.cellInfo.column) {
        // This is not meant for me
        return;
      }

      this.refs.input && this.refs.input.focus();
    }
  }, {
    key: 'handleChange',
    value: function handleChange(e) {
      this.props.onAction && this.props.onAction('edit', { value: parseInt(e.target.value, 10) });
    }
  }, {
    key: 'handleShortcut',
    value: function handleShortcut(action, e) {
      var value = _utils2.default.getImmutablePathValue(this.props.value, ['value']);

      switch (action) {
        case 'MOVE_LEFT':
          if (this.refs.input.selectionStart === 0) {
            e && e.preventDefault();
            e && e.stopPropagation();

            this.context.events.emit('moveToCell', {
              node: this.props.cellInfo.node,
              column: this.props.cellInfo.column,
              direction: 'left'
            });
          }
          break;
        case 'MOVE_RIGHT':
          if (this.refs.input.selectionStart === value.length) {
            e && e.preventDefault();
            e && e.stopPropagation();

            this.context.events.emit('moveToCell', {
              node: this.props.cellInfo.node,
              column: this.props.cellInfo.column,
              direction: 'right'
            });
          }
          break;
        case 'MOVE_UP':
          e && e.preventDefault();
          e && e.stopPropagation();
          this.context.events.emit('moveToCell', {
            node: this.props.cellInfo.node,
            column: this.props.cellInfo.column,
            direction: 'up'
          });
          break;
        case 'MOVE_DOWN':
          e && e.preventDefault();
          e && e.stopPropagation();
          this.context.events.emit('moveToCell', {
            node: this.props.cellInfo.node,
            column: this.props.cellInfo.column,
            direction: 'down'
          });
          break;
        default:
          break;
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var value = _utils2.default.getImmutablePathValue(this.props.value, ['value']);

      return _react2.default.createElement(
        _reactShortcuts.Shortcuts,
        {
          name: 'navigation',
          handler: this.handleShortcut,
          className: 'data-editor__cell__number'
        },
        _react2.default.createElement('input', {
          ref: 'input',
          onChange: this.handleChange,
          value: value
        })
      );
    }
  }]);

  return NumberInput;
}(_react.Component);

exports.default = NumberInput;


NumberInput.contextTypes = { events: _propTypes2.default.object };
