// @postman/data-editor/lib/components/base/ObjectNameInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactShortcuts = require('react-shortcuts');

var _TypeSelector = require('./TypeSelector');

var _TypeSelector2 = _interopRequireDefault(_TypeSelector);

var _utils = require('../../utilities/utils');

var _utils2 = _interopRequireDefault(_utils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ObjectNameInput = function (_Component) {
  _inherits(ObjectNameInput, _Component);

  function ObjectNameInput(props) {
    _classCallCheck(this, ObjectNameInput);

    var _this = _possibleConstructorReturn(this, (ObjectNameInput.__proto__ || Object.getPrototypeOf(ObjectNameInput)).call(this, props));

    _this.handleFocus = _this.handleFocus.bind(_this);
    _this.handleChangeName = _this.handleChangeName.bind(_this);
    _this.handleChangeType = _this.handleChangeType.bind(_this);
    _this.handleCollapse = _this.handleCollapse.bind(_this);
    _this.handleShortcut = _this.handleShortcut.bind(_this);
    return _this;
  }

  _createClass(ObjectNameInput, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.subscription = this.context.events.addListener('focusCell', this.handleFocus);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.subscription.remove();
    }
  }, {
    key: 'handleFocus',
    value: function handleFocus(_ref) {
      var node = _ref.node,
          column = _ref.column;

      if (node !== this.props.cellInfo.node || column !== this.props.cellInfo.column) {
        // This is not meant for me
        return;
      }

      this.refs.input && this.refs.input.focus();
    }
  }, {
    key: 'handleChangeName',
    value: function handleChangeName(e) {
      this.props.onAction && this.props.onAction('edit', { name: e.target.value });
    }
  }, {
    key: 'handleChangeType',
    value: function handleChangeType(e) {
      this.props.onAction && this.props.onAction('edit', { type: e.target.value });
    }
  }, {
    key: 'handleShortcut',
    value: function handleShortcut(action, e) {
      e && e.preventDefault();
      e && e.stopPropagation();

      switch (action) {
        case 'MOVE_LEFT':
          this.context.events.emit('moveToCell', {
            node: this.props.cellInfo.node,
            column: this.props.cellInfo.column,
            direction: 'left'
          });
          break;
        case 'MOVE_RIGHT':
          this.context.events.emit('moveToCell', {
            node: this.props.cellInfo.node,
            column: this.props.cellInfo.column,
            direction: 'right'
          });
          break;
        case 'MOVE_UP':
          this.context.events.emit('moveToCell', {
            node: this.props.cellInfo.node,
            column: this.props.cellInfo.column,
            direction: 'up'
          });
          break;
        case 'MOVE_DOWN':
          this.context.events.emit('moveToCell', {
            node: this.props.cellInfo.node,
            column: this.props.cellInfo.column,
            direction: 'down'
          });
          break;
        default:
          break;
      }
    }
  }, {
    key: 'handleCollapse',
    value: function handleCollapse() {
      if (this.props.isCollapsed) {
        this.props.onAction && this.props.onAction('expand');
      } else {
        this.props.onAction && this.props.onAction('collapse');
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var name = _utils2.default.getImmutablePathValue(this.props.value, ['name']),
          type = _utils2.default.getImmutablePathValue(this.props.value, ['type']),
          isCollapsed = this.props.isCollapsed;

      return _react2.default.createElement(
        _reactShortcuts.Shortcuts,
        {
          name: 'navigation',
          handler: this.handleShortcut,
          className: 'data-editor__object-name'
        },
        _react2.default.createElement(
          'div',
          { className: 'data-editor__object-name__content', style: this.props.style },
          _react2.default.createElement(
            'div',
            {
              className: 'data-editor__collapse',
              onClick: this.handleCollapse
            },
            _react2.default.createElement('div', { className: 'data-editor__collapse__icon ' + (isCollapsed ? 'is-collapsed' : '') })
          ),
          _react2.default.createElement('input', {
            ref: 'input',
            value: name,
            onChange: this.handleChangeName
          }),
          _react2.default.createElement(_TypeSelector2.default, {
            value: type,
            onChange: this.handleChangeType
          })
        )
      );
    }
  }]);

  return ObjectNameInput;
}(_react.Component);

exports.default = ObjectNameInput;


ObjectNameInput.contextTypes = { events: _propTypes2.default.object };
