// @pipedrive/insights-core/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ReportType;
(function (ReportType) {
    ReportType["STATS"] = "stats";
    ReportType["LIST"] = "list";
    ReportType["RECURRING_REVENUE"] = "recurring_revenue";
    ReportType["RECURRING_REVENUE_MOVEMENT"] = "revenue_movement";
    ReportType["CASH_FLOW"] = "cash_flow";
    ReportType["CONVERSION_OVERALL"] = "conversion_overall";
    ReportType["FUNNEL_CONVERSION"] = "funnel_conversion";
    ReportType["DURATION"] = "duration";
    ReportType["PROGRESS"] = "progress";
    ReportType["ACTIVITY"] = "activity";
    ReportType["EMAIL"] = "email";
})(ReportType = exports.ReportType || (exports.ReportType = {}));
var ChartType;
(function (ChartType) {
    ChartType["COLUMN"] = "column";
    ChartType["BAR"] = "bar";
    ChartType["PIE"] = "pie";
    ChartType["TABLE"] = "table";
    ChartType["FUNNEL"] = "funnel";
    ChartType["STACKED_BAR_CHART"] = "stackedBarChart";
    ChartType["SCORECARD"] = "scorecard";
})(ChartType = exports.ChartType || (exports.ChartType = {}));
var MeasureByType;
(function (MeasureByType) {
    MeasureByType["COUNT"] = "count";
    MeasureByType["VALUE"] = "sum";
    MeasureByType["AVERAGE_VALUE"] = "avg";
    MeasureByType["WEIGHTED_VALUE"] = "weightedValue";
    MeasureByType["PAYMENTS_AMOUNT"] = "paymentsAmount";
    MeasureByType["PAYMENTS_CHANGE_AMOUNT"] = "paymentsChangeAmount";
})(MeasureByType = exports.MeasureByType || (exports.MeasureByType = {}));
exports.DEFAULT_PIE_CHART_SEGMENT_BY = 'status';
exports.DEFAULT_PROGRESS_GROUPING_HANDLED_BY_API = 'stageId';
exports.OperandType = {
    EQ: 'isExactly',
    IN: 'isAnyOf',
    NOT: 'isNotAnyOf',
    FROM: 'isMoreThan',
    TO: 'isLessThan',
    FORMAT: 'format',
};
exports.REPORTS_REQUIRING_CONVERTTO = [
    ReportType.STATS,
    ReportType.CONVERSION_OVERALL,
    ReportType.FUNNEL_CONVERSION,
    ReportType.PROGRESS,
];
exports.CHART_TYPES_WITHOUT_GROUP_BY = [ChartType.PIE, ChartType.TABLE, ChartType.SCORECARD];
exports.CHART_TYPES_WITHOUT_SEGMENT_BY = [ChartType.SCORECARD];
exports.WITH_LABELS = ['personId', 'orgId', 'person', 'org', 'teamId'];
exports.STATS_FIELDS = ['avg', 'count', 'sum', 'isOtherSegment'];
exports.METADATA_FIELDS = ['hasNextPage', 'lastPage', 'totalItems'];
exports.SUMMARY_TOTAL_FIELD = 'total { sum count }';
exports.SUMMARY_WIN_RATE_FIELD = 'winRate { sum count }';
exports.PAYMENTS_DUE_AT_FIELD = 'paymentsDueAt';
exports.FLAT_RESPONSE_REPORT_TYPES = [ReportType.ACTIVITY];
var QueryFilterField;
(function (QueryFilterField) {
    QueryFilterField["GROUP_BY"] = "groupBy";
    QueryFilterField["FILTER"] = "filter";
})(QueryFilterField = exports.QueryFilterField || (exports.QueryFilterField = {}));
var QueryReportTypeField;
(function (QueryReportTypeField) {
    QueryReportTypeField["LIST"] = "list";
    QueryReportTypeField["STATS"] = "stats";
    QueryReportTypeField["CONVERSION_OVERALL"] = "overallConversion";
    QueryReportTypeField["FUNNEL_CONVERSION"] = "funnelConversion";
    QueryReportTypeField["REVENUE_MOVEMENT"] = "revenueMovements";
    QueryReportTypeField["DURATION"] = "duration";
    QueryReportTypeField["PROGRESS"] = "progress";
    QueryReportTypeField["ACTIVITY"] = "activities";
})(QueryReportTypeField = exports.QueryReportTypeField || (exports.QueryReportTypeField = {}));
