// @pipedrive/insights-core/lib/constants/dealsConstants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PAYMENTS_DUE_AT_FIELD = exports.SUMMARY_WIN_RATE_FIELD = exports.SUMMARY_TOTAL_FIELD = exports.METADATA_FIELDS = exports.STATS_FIELDS = exports.CHART_TYPES_WITHOUT_SEGMENT_BY = exports.CHART_TYPES_WITHOUT_GROUP_BY = exports.REPORTS_REQUIRING_IS_CUMULATIVE_ARG = exports.REPORTS_REQUIRING_PAGE_AND_SORT_ARGS = exports.REPORTS_REQUIRING_CONVERTTO = exports.DEFAULT_PROGRESS_GROUPING_HANDLED_BY_API = exports.DEFAULT_PIE_CHART_SEGMENT_BY = exports.FIELDS_WITH_LABELS = void 0;
const types_1 = require("../types");
exports.FIELDS_WITH_LABELS = ['personId', 'orgId', 'person', 'org', 'teamId', 'dealProductsProductId'];
exports.DEFAULT_PIE_CHART_SEGMENT_BY = 'status';
exports.DEFAULT_PROGRESS_GROUPING_HANDLED_BY_API = 'stageId';
exports.REPORTS_REQUIRING_CONVERTTO = [
    types_1.ReportType.DEALS_STATS,
    types_1.ReportType.DEALS_CONVERSION_OVERALL,
    types_1.ReportType.DEALS_FUNNEL_CONVERSION,
    types_1.ReportType.DEALS_PROGRESS,
    types_1.ReportType.DEALS_REVENUE_FORECAST,
    types_1.ReportType.DEALS_RECURRING_REVENUE,
    types_1.ReportType.DEALS_RECURRING_REVENUE_MOVEMENT,
];
exports.REPORTS_REQUIRING_PAGE_AND_SORT_ARGS = [
    types_1.ReportType.DEALS_STATS,
    types_1.ReportType.DEALS_CONVERSION_OVERALL,
    types_1.ReportType.DEALS_FUNNEL_CONVERSION,
    types_1.ReportType.DEALS_PROGRESS,
    types_1.ReportType.DEALS_RECURRING_REVENUE,
    types_1.ReportType.DEALS_RECURRING_REVENUE_MOVEMENT,
    types_1.ReportType.DEALS_DURATION,
];
exports.REPORTS_REQUIRING_IS_CUMULATIVE_ARG = [
    types_1.ReportType.DEALS_REVENUE_FORECAST,
];
exports.CHART_TYPES_WITHOUT_GROUP_BY = [types_1.ChartType.PIE, types_1.ChartType.TABLE, types_1.ChartType.SCORECARD];
exports.CHART_TYPES_WITHOUT_SEGMENT_BY = [types_1.ChartType.TABLE, types_1.ChartType.SCORECARD];
exports.STATS_FIELDS = ['avg', 'count', 'sum', 'isOtherSegment'];
exports.METADATA_FIELDS = ['hasNextPage', 'lastPage', 'totalItems'];
exports.SUMMARY_TOTAL_FIELD = 'total { sum count }';
exports.SUMMARY_WIN_RATE_FIELD = 'winRate { sum count }';
exports.PAYMENTS_DUE_AT_FIELD = 'paymentsDueAt';
