// feather/lib/types/components/alerts.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.AlertMessageAlign = exports.AlertSlideFrom = exports.AlertStatus = void 0;var

AlertStatus;exports.AlertStatus = AlertStatus;(function (AlertStatus) {AlertStatus["PRIMARY"] = "PRIMARY";AlertStatus["SECONDARY"] = "SECONDARY";AlertStatus["SUCCESS"] = "SUCCESS";AlertStatus["DANGER"] = "DANGER";AlertStatus["WARNING"] = "WARNING";AlertStatus["INFO"] = "INFO";AlertStatus["LIGHT"] = "LIGHT";AlertStatus["DARK"] = "DARK";})(AlertStatus || (exports.AlertStatus = AlertStatus = {}));var










AlertSlideFrom;exports.AlertSlideFrom = AlertSlideFrom;(function (AlertSlideFrom) {AlertSlideFrom["TOP"] = "TOP";AlertSlideFrom["RIGHT"] = "RIGHT";AlertSlideFrom["BOTTOM"] = "BOTTOM";AlertSlideFrom["LEFT"] = "LEFT";})(AlertSlideFrom || (exports.AlertSlideFrom = AlertSlideFrom = {}));var






AlertMessageAlign;exports.AlertMessageAlign = AlertMessageAlign;(function (AlertMessageAlign) {AlertMessageAlign["RIGHT"] = "right";AlertMessageAlign["LEFT"] = "left";AlertMessageAlign["CENTER"] = "center";})(AlertMessageAlign || (exports.AlertMessageAlign = AlertMessageAlign = {}));
