// @pipedrive/leads-sidebar/dist/components/CustomField/__generated__/CustomFieldRendererMutation.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ConcreteRequest } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type CustomFieldInput = {
    CustomFieldDate?: CustomFieldDateInput | null;
    CustomFieldLargeText?: CustomFieldLargeTextInput | null;
    CustomFieldMonetary?: CustomFieldMonetaryInput | null;
    CustomFieldMultipleOptions?: CustomFieldMultipleOptionsInput | null;
    CustomFieldNumeric?: CustomFieldNumericInput | null;
    CustomFieldPhone?: CustomFieldPhoneInput | null;
    CustomFieldSingleOption?: CustomFieldSingleOptionInput | null;
    CustomFieldText?: CustomFieldTextInput | null;
};
export type CustomFieldDateInput = {
    date?: string | null;
    id: string;
};
export type CustomFieldLargeTextInput = {
    id: string;
    text?: string | null;
};
export type CustomFieldMonetaryInput = {
    id: string;
    value?: MoneyInput | null;
};
export type MoneyInput = {
    amount: string;
    currency: string;
};
export type CustomFieldMultipleOptionsInput = {
    id: string;
    options?: Array<string> | null;
};
export type CustomFieldNumericInput = {
    id: string;
    number?: number | null;
};
export type CustomFieldPhoneInput = {
    id: string;
    phone?: string | null;
};
export type CustomFieldSingleOptionInput = {
    id: string;
    option?: string | null;
};
export type CustomFieldTextInput = {
    id: string;
    text?: string | null;
};
export type CustomFieldRendererMutationVariables = {
    customFieldInput: CustomFieldInput;
};
export type CustomFieldRendererMutationResponse = {
    readonly updateCustomFieldValue: {
        readonly __isErrorFIXME?: string;
        readonly " $fragmentRefs": FragmentRefs<"CustomFieldRenderer_customField" | "CustomFieldRenderer_customFieldInfo">;
    } | null;
};
export type CustomFieldRendererMutation = {
    readonly response: CustomFieldRendererMutationResponse;
    readonly variables: CustomFieldRendererMutationVariables;
};



/*
mutation CustomFieldRendererMutation(
  $customFieldInput: CustomFieldInput!
) {
  updateCustomFieldValue(input: $customFieldInput) {
    __typename
    ... on Error {
      __isError: __typename
      __isErrorFIXME: message
    }
    ...CustomFieldRenderer_customField
    ...CustomFieldRenderer_customFieldInfo
    ... on CustomFieldDate {
      id
    }
    ... on CustomFieldLargeText {
      id
    }
    ... on CustomFieldMonetary {
      id
    }
    ... on CustomFieldMultipleOptions {
      id
    }
    ... on CustomFieldNumeric {
      id
    }
    ... on CustomFieldPhone {
      id
    }
    ... on CustomFieldSingleOption {
      id
    }
    ... on CustomFieldText {
      id
    }
  }
}

fragment CustomFieldAddress_data on CustomFieldAddress {
  ...CustomField_data
  address
}

fragment CustomFieldAutocomplete_data on CustomFieldAutocomplete {
  ...CustomField_data
  value
}

fragment CustomFieldDateRange_data on CustomFieldDateRange {
  ...CustomField_data
  dateRangeStart: start
  dateRangeEnd: end
}

fragment CustomFieldDate_data on CustomFieldDate {
  ...CustomField_data
  date
}

fragment CustomFieldLargeText_data on CustomFieldLargeText {
  ...CustomField_data
  text
}

fragment CustomFieldMonetary_data on CustomFieldMonetary {
  ...CustomField_data
  moneyValue: value {
    amount
    currency {
      code
      name
      symbol
      id
    }
  }
}

fragment CustomFieldMultipleOptions_data on CustomFieldMultipleOptions {
  ...CustomField_data
  selectedOptions: options {
    id
    label
  }
  availableOptions {
    id
    label
  }
}

fragment CustomFieldNumeric_data on CustomFieldNumeric {
  ...CustomField_data
  number
}

fragment CustomFieldOrganization_data on CustomFieldOrganization {
  ...CustomField_data
  name
  organization {
    name
    id
  }
}

fragment CustomFieldPerson_data on CustomFieldPerson {
  ...CustomField_data
  name
  person {
    name
    id
  }
}

fragment CustomFieldPhone_data on CustomFieldPhone {
  ...CustomField_data
  phone
}

fragment CustomFieldRenderer_customField on CustomField {
  __isCustomField: __typename
  __typename
  ... on CustomFieldAutocomplete {
    ...CustomFieldAutocomplete_data
  }
  ... on CustomFieldDate {
    ...CustomFieldDate_data
  }
  ... on CustomFieldDateRange {
    ...CustomFieldDateRange_data
  }
  ... on CustomFieldLargeText {
    ...CustomFieldLargeText_data
  }
  ... on CustomFieldMonetary {
    ...CustomFieldMonetary_data
  }
  ... on CustomFieldMultipleOptions {
    ...CustomFieldMultipleOptions_data
  }
  ... on CustomFieldNumeric {
    ...CustomFieldNumeric_data
  }
  ... on CustomFieldPhone {
    ...CustomFieldPhone_data
  }
  ... on CustomFieldSingleOption {
    ...CustomFieldSingleOption_data
  }
  ... on CustomFieldText {
    ...CustomFieldText_data
  }
  ... on CustomFieldTime {
    ...CustomFieldTime_data
  }
  ... on CustomFieldTimeRange {
    ...CustomFieldTimeRange_data
  }
  ... on CustomFieldAddress {
    ...CustomFieldAddress_data
  }
  ... on CustomFieldOrganization {
    ...CustomFieldOrganization_data
  }
  ... on CustomFieldPerson {
    ...CustomFieldPerson_data
  }
}

fragment CustomFieldRenderer_customFieldInfo on CustomField {
  __isCustomField: __typename
  isVisible
  valueHash
}

fragment CustomFieldSingleOption_data on CustomFieldSingleOption {
  ...CustomField_data
  selectedOption: option {
    id
    label
  }
  availableOptions {
    id
    label
  }
}

fragment CustomFieldText_data on CustomFieldText {
  ...CustomField_data
  text
}

fragment CustomFieldTimeRange_data on CustomFieldTimeRange {
  ...CustomField_data
  timeRangeStart: start
  timeRangeEnd: end
}

fragment CustomFieldTime_data on CustomFieldTime {
  ...CustomField_data
  time
}

fragment CustomField_data on CustomField {
  __isCustomField: __typename
  id
  name
}
*/

const node: ConcreteRequest = (function () {
    var v0 = [
        {
            "defaultValue": null,
            "kind": "LocalArgument",
            "name": "customFieldInput"
        } as any
    ], v1 = [
        {
            "kind": "Variable",
            "name": "input",
            "variableName": "customFieldInput"
        } as any
    ], v2 = {
        "kind": "InlineFragment",
        "selections": [
            {
                "alias": "__isErrorFIXME",
                "args": null,
                "kind": "ScalarField",
                "name": "message",
                "storageKey": null
            }
        ],
        "type": "Error",
        "abstractKey": "__isError"
    } as any, v3 = {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "id",
        "storageKey": null
    } as any, v4 = {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "name",
        "storageKey": null
    } as any, v5 = {
        "kind": "InlineFragment",
        "selections": [
            (v3 /*: any*/),
            (v4 /*: any*/)
        ],
        "type": "CustomField",
        "abstractKey": "__isCustomField"
    } as any, v6 = [
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "text",
            "storageKey": null
        } as any,
        (v5 /*: any*/)
    ], v7 = [
        (v3 /*: any*/),
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "label",
            "storageKey": null
        } as any
    ], v8 = {
        "alias": null,
        "args": null,
        "concreteType": "CustomFieldOption",
        "kind": "LinkedField",
        "name": "availableOptions",
        "plural": true,
        "selections": (v7 /*: any*/),
        "storageKey": null
    } as any, v9 = [
        (v4 /*: any*/),
        (v3 /*: any*/)
    ], v10 = [
        (v3 /*: any*/)
    ], v11 = {
        "kind": "InlineFragment",
        "selections": (v10 /*: any*/),
        "type": "CustomField",
        "abstractKey": "__isCustomField"
    } as any;
    return {
        "fragment": {
            "argumentDefinitions": (v0 /*: any*/),
            "kind": "Fragment",
            "metadata": null,
            "name": "CustomFieldRendererMutation",
            "selections": [
                {
                    "alias": null,
                    "args": (v1 /*: any*/),
                    "concreteType": null,
                    "kind": "LinkedField",
                    "name": "updateCustomFieldValue",
                    "plural": false,
                    "selections": [
                        (v2 /*: any*/),
                        {
                            "args": null,
                            "kind": "FragmentSpread",
                            "name": "CustomFieldRenderer_customField"
                        },
                        {
                            "args": null,
                            "kind": "FragmentSpread",
                            "name": "CustomFieldRenderer_customFieldInfo"
                        }
                    ],
                    "storageKey": null
                }
            ],
            "type": "RootMutation",
            "abstractKey": null
        },
        "kind": "Request",
        "operation": {
            "argumentDefinitions": (v0 /*: any*/),
            "kind": "Operation",
            "name": "CustomFieldRendererMutation",
            "selections": [
                {
                    "alias": null,
                    "args": (v1 /*: any*/),
                    "concreteType": null,
                    "kind": "LinkedField",
                    "name": "updateCustomFieldValue",
                    "plural": false,
                    "selections": [
                        {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "__typename",
                            "storageKey": null
                        },
                        (v2 /*: any*/),
                        {
                            "kind": "InlineFragment",
                            "selections": [
                                {
                                    "alias": null,
                                    "args": null,
                                    "kind": "ScalarField",
                                    "name": "isVisible",
                                    "storageKey": null
                                },
                                {
                                    "alias": null,
                                    "args": null,
                                    "kind": "ScalarField",
                                    "name": "valueHash",
                                    "storageKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "value",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldAutocomplete",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "date",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldDate",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": "dateRangeStart",
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "start",
                                            "storageKey": null
                                        },
                                        {
                                            "alias": "dateRangeEnd",
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "end",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldDateRange",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": (v6 /*: any*/),
                                    "type": "CustomFieldLargeText",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": "moneyValue",
                                            "args": null,
                                            "concreteType": "Money",
                                            "kind": "LinkedField",
                                            "name": "value",
                                            "plural": false,
                                            "selections": [
                                                {
                                                    "alias": null,
                                                    "args": null,
                                                    "kind": "ScalarField",
                                                    "name": "amount",
                                                    "storageKey": null
                                                },
                                                {
                                                    "alias": null,
                                                    "args": null,
                                                    "concreteType": "MoneyCurrency",
                                                    "kind": "LinkedField",
                                                    "name": "currency",
                                                    "plural": false,
                                                    "selections": [
                                                        {
                                                            "alias": null,
                                                            "args": null,
                                                            "kind": "ScalarField",
                                                            "name": "code",
                                                            "storageKey": null
                                                        },
                                                        (v4 /*: any*/),
                                                        {
                                                            "alias": null,
                                                            "args": null,
                                                            "kind": "ScalarField",
                                                            "name": "symbol",
                                                            "storageKey": null
                                                        },
                                                        (v3 /*: any*/)
                                                    ],
                                                    "storageKey": null
                                                }
                                            ],
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldMonetary",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": "selectedOptions",
                                            "args": null,
                                            "concreteType": "CustomFieldOption",
                                            "kind": "LinkedField",
                                            "name": "options",
                                            "plural": true,
                                            "selections": (v7 /*: any*/),
                                            "storageKey": null
                                        },
                                        (v8 /*: any*/),
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldMultipleOptions",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "number",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldNumeric",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "phone",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldPhone",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": "selectedOption",
                                            "args": null,
                                            "concreteType": "CustomFieldOption",
                                            "kind": "LinkedField",
                                            "name": "option",
                                            "plural": false,
                                            "selections": (v7 /*: any*/),
                                            "storageKey": null
                                        },
                                        (v8 /*: any*/),
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldSingleOption",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": (v6 /*: any*/),
                                    "type": "CustomFieldText",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "time",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldTime",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": "timeRangeStart",
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "start",
                                            "storageKey": null
                                        },
                                        {
                                            "alias": "timeRangeEnd",
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "end",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldTimeRange",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        {
                                            "alias": null,
                                            "args": null,
                                            "kind": "ScalarField",
                                            "name": "address",
                                            "storageKey": null
                                        },
                                        (v5 /*: any*/)
                                    ],
                                    "type": "CustomFieldAddress",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        (v4 /*: any*/),
                                        {
                                            "alias": null,
                                            "args": null,
                                            "concreteType": "Organization",
                                            "kind": "LinkedField",
                                            "name": "organization",
                                            "plural": false,
                                            "selections": (v9 /*: any*/),
                                            "storageKey": null
                                        },
                                        (v11 /*: any*/)
                                    ],
                                    "type": "CustomFieldOrganization",
                                    "abstractKey": null
                                },
                                {
                                    "kind": "InlineFragment",
                                    "selections": [
                                        (v4 /*: any*/),
                                        {
                                            "alias": null,
                                            "args": null,
                                            "concreteType": "Person",
                                            "kind": "LinkedField",
                                            "name": "person",
                                            "plural": false,
                                            "selections": (v9 /*: any*/),
                                            "storageKey": null
                                        },
                                        (v11 /*: any*/)
                                    ],
                                    "type": "CustomFieldPerson",
                                    "abstractKey": null
                                }
                            ],
                            "type": "CustomField",
                            "abstractKey": "__isCustomField"
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldDate",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldLargeText",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldMonetary",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldMultipleOptions",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldNumeric",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldPhone",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldSingleOption",
                            "abstractKey": null
                        },
                        {
                            "kind": "InlineFragment",
                            "selections": (v10 /*: any*/),
                            "type": "CustomFieldText",
                            "abstractKey": null
                        }
                    ],
                    "storageKey": null
                }
            ]
        },
        "params": {
            "cacheID": "659e57c32e79101dd6b315f923323e27",
            "id": null,
            "metadata": {},
            "name": "CustomFieldRendererMutation",
            "operationKind": "mutation",
            "text": "mutation CustomFieldRendererMutation(\n  $customFieldInput: CustomFieldInput!\n) {\n  updateCustomFieldValue(input: $customFieldInput) {\n    __typename\n    ... on Error {\n      __isError: __typename\n      __isErrorFIXME: message\n    }\n    ...CustomFieldRenderer_customField\n    ...CustomFieldRenderer_customFieldInfo\n    ... on CustomFieldDate {\n      id\n    }\n    ... on CustomFieldLargeText {\n      id\n    }\n    ... on CustomFieldMonetary {\n      id\n    }\n    ... on CustomFieldMultipleOptions {\n      id\n    }\n    ... on CustomFieldNumeric {\n      id\n    }\n    ... on CustomFieldPhone {\n      id\n    }\n    ... on CustomFieldSingleOption {\n      id\n    }\n    ... on CustomFieldText {\n      id\n    }\n  }\n}\n\nfragment CustomFieldAddress_data on CustomFieldAddress {\n  ...CustomField_data\n  address\n}\n\nfragment CustomFieldAutocomplete_data on CustomFieldAutocomplete {\n  ...CustomField_data\n  value\n}\n\nfragment CustomFieldDateRange_data on CustomFieldDateRange {\n  ...CustomField_data\n  dateRangeStart: start\n  dateRangeEnd: end\n}\n\nfragment CustomFieldDate_data on CustomFieldDate {\n  ...CustomField_data\n  date\n}\n\nfragment CustomFieldLargeText_data on CustomFieldLargeText {\n  ...CustomField_data\n  text\n}\n\nfragment CustomFieldMonetary_data on CustomFieldMonetary {\n  ...CustomField_data\n  moneyValue: value {\n    amount\n    currency {\n      code\n      name\n      symbol\n      id\n    }\n  }\n}\n\nfragment CustomFieldMultipleOptions_data on CustomFieldMultipleOptions {\n  ...CustomField_data\n  selectedOptions: options {\n    id\n    label\n  }\n  availableOptions {\n    id\n    label\n  }\n}\n\nfragment CustomFieldNumeric_data on CustomFieldNumeric {\n  ...CustomField_data\n  number\n}\n\nfragment CustomFieldOrganization_data on CustomFieldOrganization {\n  ...CustomField_data\n  name\n  organization {\n    name\n    id\n  }\n}\n\nfragment CustomFieldPerson_data on CustomFieldPerson {\n  ...CustomField_data\n  name\n  person {\n    name\n    id\n  }\n}\n\nfragment CustomFieldPhone_data on CustomFieldPhone {\n  ...CustomField_data\n  phone\n}\n\nfragment CustomFieldRenderer_customField on CustomField {\n  __isCustomField: __typename\n  __typename\n  ... on CustomFieldAutocomplete {\n    ...CustomFieldAutocomplete_data\n  }\n  ... on CustomFieldDate {\n    ...CustomFieldDate_data\n  }\n  ... on CustomFieldDateRange {\n    ...CustomFieldDateRange_data\n  }\n  ... on CustomFieldLargeText {\n    ...CustomFieldLargeText_data\n  }\n  ... on CustomFieldMonetary {\n    ...CustomFieldMonetary_data\n  }\n  ... on CustomFieldMultipleOptions {\n    ...CustomFieldMultipleOptions_data\n  }\n  ... on CustomFieldNumeric {\n    ...CustomFieldNumeric_data\n  }\n  ... on CustomFieldPhone {\n    ...CustomFieldPhone_data\n  }\n  ... on CustomFieldSingleOption {\n    ...CustomFieldSingleOption_data\n  }\n  ... on CustomFieldText {\n    ...CustomFieldText_data\n  }\n  ... on CustomFieldTime {\n    ...CustomFieldTime_data\n  }\n  ... on CustomFieldTimeRange {\n    ...CustomFieldTimeRange_data\n  }\n  ... on CustomFieldAddress {\n    ...CustomFieldAddress_data\n  }\n  ... on CustomFieldOrganization {\n    ...CustomFieldOrganization_data\n  }\n  ... on CustomFieldPerson {\n    ...CustomFieldPerson_data\n  }\n}\n\nfragment CustomFieldRenderer_customFieldInfo on CustomField {\n  __isCustomField: __typename\n  isVisible\n  valueHash\n}\n\nfragment CustomFieldSingleOption_data on CustomFieldSingleOption {\n  ...CustomField_data\n  selectedOption: option {\n    id\n    label\n  }\n  availableOptions {\n    id\n    label\n  }\n}\n\nfragment CustomFieldText_data on CustomFieldText {\n  ...CustomField_data\n  text\n}\n\nfragment CustomFieldTimeRange_data on CustomFieldTimeRange {\n  ...CustomField_data\n  timeRangeStart: start\n  timeRangeEnd: end\n}\n\nfragment CustomFieldTime_data on CustomFieldTime {\n  ...CustomField_data\n  time\n}\n\nfragment CustomField_data on CustomField {\n  __isCustomField: __typename\n  id\n  name\n}\n"
        }
    } as any;
})();
(node as any).hash = '476fa80cf1a08ff942b6d6ff07c53f1c';
export default node;
