// @pipedrive/leads-sidebar/dist/types/types.ts
import { Phone, Email } from '@pipedrive/form-fields';

export type CompoundField = {
	value: string;
	label: string;
};

export type AddressGmaps = {
	description: string;
};

type Address = AddressGmaps | string;

/*
	CompoundField = regular phone field with Home, Work, etc. dropdown (label)
	string = phone custom field (without dropdown)
 */
export type Phone = CompoundField | CompoundField[] | string;

export type Email = CompoundField | CompoundField[];

export type Monetary = CompoundField;

// SingleOption and MultipleOptions
export type FormFieldOption = {
	id: number | string;
	label: string;
	value: string;
};

export type FormValue =
	| string // Text, LargeText, Date
	| Address
	| Email
	| Phone
	| Monetary
	| number
	| number[]
	| FormFieldOption // single option
	| FormFieldOption[] // multiple option
	| undefined
	| null;

export type Locale = {
	countryCode: string;
	languageCode: string;
	languageTag: string;
};

export type Currency = {
	active_flag: boolean;
	code: string;
	decimal_points: number;
	id: number;
	is_custom_flag: boolean;
	name: string;
	symbol: string;
};

type PhoneSettings = React.ComponentProps<typeof Phone>;
export type Settings = PhoneSettings['settings'];

export type PopoverPlacement =
	| 'left'
	| 'right'
	| 'top'
	| 'bottom'
	| 'top-start'
	| 'top-end'
	| 'bottom-start'
	| 'bottom-end'
	| 'left-start'
	| 'left-end'
	| 'right-start'
	| 'right-end';

export function isCompoundFieldFormValue(formValue: FormValue): formValue is CompoundField {
	return (formValue as CompoundField).value !== undefined && (formValue as CompoundField).label !== undefined;
}
