// react-bootstrap/lib/Overlay.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = require("react-dom");

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Overlay = _interopRequireDefault(require("react-overlays/Overlay"));

var _elementType = _interopRequireDefault(require("prop-types-extra/lib/elementType"));

var _Fade = _interopRequireDefault(require("./Fade"));

var propTypes = (0, _extends2.default)({}, _Overlay.default.propTypes, {
    /**
     * Set the visibility of the Overlay
     */
    show: _propTypes.default.bool,

    /**
     * Specify whether the overlay should trigger onHide when the user clicks outside the overlay
     */
    rootClose: _propTypes.default.bool,

    /**
     * A callback invoked by the overlay when it wishes to be hidden. Required if
     * `rootClose` is specified.
     */
    onHide: _propTypes.default.func,

    /**
     * Use animation
     */
    transition: _propTypes.default.oneOfType([_propTypes.default.bool, _elementType.default]),

    /**
     * Callback fired before the Overlay transitions in
     */
    onEnter: _propTypes.default.func,

    /**
     * Callback fired as the Overlay begins to transition in
     */
    onEntering: _propTypes.default.func,

    /**
     * Callback fired after the Overlay finishes transitioning in
     */
    onEntered: _propTypes.default.func,

    /**
     * Callback fired right before the Overlay transitions out
     */
    onExit: _propTypes.default.func,

    /**
     * Callback fired as the Overlay begins to transition out
     */
    onExiting: _propTypes.default.func,

    /**
     * Callback fired after the Overlay finishes transitioning out
     */
    onExited: _propTypes.default.func,

    /**
     * Sets the direction of the Overlay.
     */
    placement: _propTypes.default.oneOf(['top', 'right', 'bottom', 'left'])
});
var defaultProps = {
    transition: _Fade.default,
    rootClose: false,
    show: false,
    placement: 'top'
};

function wrapRefs(props, arrowProps) {
    var ref = props.ref;
    var aRef = arrowProps.ref;

    props.ref = ref.__wrapped || (ref.__wrapped = function(r) {
        return ref((0, _reactDom.findDOMNode)(r));
    });

    arrowProps.ref = aRef.__wrapped || (aRef.__wrapped = function(r) {
        return aRef((0, _reactDom.findDOMNode)(r));
    });
}

function Overlay(_ref) {
    var overlay = _ref.children,
        transition = _ref.transition,
        outerProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["children", "transition"]);
    transition = transition === true ? _Fade.default : transition || null;
    return _react.default.createElement(_Overlay.default, (0, _extends2.default)({}, outerProps, {
        transition: transition
    }), function(_ref2) {
        var overlayProps = _ref2.props,
            arrowProps = _ref2.arrowProps,
            show = _ref2.show,
            props = (0, _objectWithoutPropertiesLoose2.default)(_ref2, ["props", "arrowProps", "show"]);
        wrapRefs(overlayProps, arrowProps);
        if (typeof overlay === 'function') return overlay((0, _extends2.default)({}, props, overlayProps, {
            show: show,
            arrowProps: arrowProps
        }));
        return _react.default.cloneElement(overlay, (0, _extends2.default)({}, props, overlayProps, {
            arrowProps: arrowProps,
            className: (0, _classnames.default)(overlay.props.className, !transition && show && 'show'),
            style: (0, _extends2.default)({}, overlay.props.style, overlayProps.style)
        }));
    });
}

Overlay.propTypes = propTypes;
Overlay.defaultProps = defaultProps;
var _default = Overlay;
exports.default = _default;
module.exports = exports["default"];
