// react-bootstrap/lib/Popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _isRequiredForA11y = _interopRequireDefault(require("prop-types-extra/lib/isRequiredForA11y"));

var _ThemeProvider = require("./ThemeProvider");

var propTypes = {
    /**
     * @default 'popover'
     */
    bsPrefix: _propTypes.default.string,

    /**
     * An html id attribute, necessary for accessibility
     * @type {string|number}
     * @required
     */
    id: (0, _isRequiredForA11y.default)(_propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.number])),

    /**
     * Sets the direction the Popover is positioned towards.
     *
     * > This is generally provided by the `Overlay` component positioning the popover
     */
    placement: _propTypes.default.oneOf(['auto-start', 'auto', 'auto-end', 'top-start', 'top', 'top-end', 'right-start', 'right', 'right-end', 'bottom-end', 'bottom', 'bottom-start', 'left-end', 'left', 'left-start']),

    /**
     * An Overlay injected set of props for positioning the popover arrow.
     *
     * > This is generally provided by the `Overlay` component positioning the popover
     */
    arrowProps: _propTypes.default.shape({
        ref: _propTypes.default.any,
        style: _propTypes.default.object
    }),

    /** @private */
    innerRef: _propTypes.default.any,

    /** @private */
    scheduleUpdate: _propTypes.default.func,

    /** @private */
    outOfBoundaries: _propTypes.default.func,

    /**
     * Title content
     */
    title: _propTypes.default.node
};
var defaultProps = {
    placement: 'right'
};

function Popover(_ref) {
    var bsPrefix = _ref.bsPrefix,
        innerRef = _ref.innerRef,
        placement = _ref.placement,
        className = _ref.className,
        style = _ref.style,
        title = _ref.title,
        children = _ref.children,
        arrowProps = _ref.arrowProps,
        _ = _ref.scheduleUpdate,
        _1 = _ref.outOfBoundaries,
        props = (0, _objectWithoutPropertiesLoose2.default)(_ref, ["bsPrefix", "innerRef", "placement", "className", "style", "title", "children", "arrowProps", "scheduleUpdate", "outOfBoundaries"]);
    return _react.default.createElement("div", (0, _extends2.default)({
        role: "tooltip",
        ref: innerRef,
        style: style,
        "x-placement": placement,
        className: (0, _classnames.default)(className, bsPrefix, "bs-popover-" + placement)
    }, props), _react.default.createElement("div", (0, _extends2.default)({
        className: "arrow"
    }, arrowProps)), title && _react.default.createElement("div", {
        className: bsPrefix + "-header h3"
    }, title), _react.default.createElement("div", {
        className: bsPrefix + "-body"
    }, children));
}

Popover.propTypes = propTypes;
Popover.defaultProps = defaultProps;

var _default = (0, _ThemeProvider.createBootstrapComponent)(Popover, 'popover');

exports.default = _default;
module.exports = exports["default"];
