// @pipedrive/leads-sidebar/dist/DealInfo/__generated__/DealInfo_dealInfo.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ReaderFragment } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type DealInfo_dealInfo = {
    readonly value: {
        readonly currency: {
            readonly code: string;
        };
        readonly amount: string;
    } | null;
    readonly customFields: ReadonlyArray<{
        readonly id: string;
        readonly " $fragmentRefs": FragmentRefs<"CustomFieldRenderer_customField" | "CustomFieldRenderer_customFieldInfo">;
    } | null> | null;
    readonly " $refType": "DealInfo_dealInfo";
};
export type DealInfo_dealInfo$data = DealInfo_dealInfo;
export type DealInfo_dealInfo$key = {
    readonly " $data"?: DealInfo_dealInfo$data;
    readonly " $fragmentRefs": FragmentRefs<"DealInfo_dealInfo">;
};



const node: ReaderFragment = {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "DealInfo_dealInfo",
    "selections": [
        {
            "alias": null,
            "args": null,
            "concreteType": "Money",
            "kind": "LinkedField",
            "name": "value",
            "plural": false,
            "selections": [
                {
                    "alias": null,
                    "args": null,
                    "concreteType": "MoneyCurrency",
                    "kind": "LinkedField",
                    "name": "currency",
                    "plural": false,
                    "selections": [
                        {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "code",
                            "storageKey": null
                        }
                    ],
                    "storageKey": null
                },
                {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "amount",
                    "storageKey": null
                }
            ],
            "storageKey": null
        },
        {
            "alias": null,
            "args": null,
            "concreteType": null,
            "kind": "LinkedField",
            "name": "customFields",
            "plural": true,
            "selections": [
                {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "id",
                    "storageKey": null
                },
                {
                    "args": null,
                    "kind": "FragmentSpread",
                    "name": "CustomFieldRenderer_customField"
                },
                {
                    "args": null,
                    "kind": "FragmentSpread",
                    "name": "CustomFieldRenderer_customFieldInfo"
                }
            ],
            "storageKey": null
        }
    ],
    "type": "DealInfo",
    "abstractKey": null
} as any;
(node as any).hash = '2d5a04b16027d220e78161939fcb31b6';
export default node;
