// feather/lib/components/checkbox/index.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/es.weak-map.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.symbol.iterator.js");Object.defineProperty(exports, "__esModule", { value: true });exports.Checkbox = void 0;require("core-js/modules/es.function.name.js");require("core-js/modules/es.object.keys.js");var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _animation = require("../../../lib/animation");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}








var Input = _styledComponents.default.input.withConfig({ displayName: "checkbox__Input", componentId: "pidi1f-0" })(["-webkit-appearance:none;width:18px;height:18px;border:2px solid ", ";border-radius:2px;background-color:white;outline:none;box-shadow:none;transition:", ";&:disabled{border:2px solid ", ";&:checked{background-color:", ";border-color:", ";}}&:focus:not(:active){box-shadow:", " 0px 0px 0px 2px;}&:not(:disabled){cursor:pointer;}&:checked{border-color:", ";background-color:", ";}&:indeterminate{border-color:", ";background-color:", ";&:disabled{background-color:", ";border-color:", ";}&::after{content:'';display:block;position:absolute;top:9px;left:5px;width:10px;height:2px;border-radius:1px;background-color:white;}}"],



(0, _cssVariables.default)('neutral-6'),




(0, _animation.transitions)({ duration: 0.2, properties: ['background-color', 'border-color', 'box-shadow'] }),


(0, _cssVariables.default)('neutral-5'),

(0, _cssVariables.default)('purple-3'),
(0, _cssVariables.default)('purple-3'),




(0, _cssVariables.default)('purple-7'),







(0, _cssVariables.default)('purple-7'),
(0, _cssVariables.default)('purple-7'),



(0, _cssVariables.default)('purple-7'),
(0, _cssVariables.default)('purple-7'),


(0, _cssVariables.default)('purple-3'),
(0, _cssVariables.default)('purple-3'));
















var CheckmarkContainer = _styledComponents.default.div.withConfig({ displayName: "checkbox__CheckmarkContainer", componentId: "pidi1f-1" })(["opacity:0;position:absolute;display:flex;align-items:center;top:6px;left:5px;transition:", ";pointer-events:none;svg{fill:white;}"],






(0, _animation.transitions)({ duration: 0.2, properties: ['opacity'] }));






var Container = _styledComponents.default.div.withConfig({ displayName: "checkbox__Container", componentId: "pidi1f-2" })(["display:flex;align-items:center;justify-content:center;position:relative;width:20px;height:20px;input:checked:not(:indeterminate) + ", "{opacity:1;}"],







CheckmarkContainer);




var Label = _styledComponents.default.label.withConfig({ displayName: "checkbox__Label", componentId: "pidi1f-3" })(["flex:1;display:inline-block;margin-left:", "px;font-size:14px;font-weight:500;line-height:1.43;color:", ";cursor:pointer;transition:", ";"],


function (props) {return props.hasLabel ? 8 : 0;},



(0, _cssVariables.default)('neutral-10'),

(0, _animation.transitions)({ duration: 0.2, properties: ['color'] }));


var Wrapper = _styledComponents.default.div.withConfig({ displayName: "checkbox__Wrapper", componentId: "pidi1f-4" })(["display:flex;", " ", ""],


function (props) {return (
    props.hasError && (0,
    _styledComponents.css)(["input{border-color:", ";}", "{color:", ";}"],

    (0, _cssVariables.default)('red-5'),


    Label,
    (0, _cssVariables.default)('red-5')));},



function (props) {return (
    props.disabled && (0,
    _styledComponents.css)(["", "{color:", ";}"],
    Label,
    (0, _cssVariables.default)('neutral-5')));});




var Checkmark = function Checkmark() {return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(CheckmarkContainer, { children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)("svg", { xmlns: "http://www.w3.org/2000/svg", width: "10", height: "9", viewBox: "0 0 10 9", children: /*#__PURE__*/
        (0, _jsxRuntime.jsx)("path", {
          fillRule: "evenodd",
          d: "M8.327 1.26a1 1 0 1 1 1.346 1.48l-5.5 5a1 1 0 0 1-1.346 0l-2.5-2.273a1 1 0 0 1 1.346-1.48L3.5 5.65 8.327 1.26z" }) }) }));};





var Checkbox = /*#__PURE__*/_react.default.forwardRef(
function (_ref, ref) {var id = _ref.id,name = _ref.name,label = _ref.label,className = _ref.className,_ref$indeterminate = _ref.indeterminate,indeterminate = _ref$indeterminate === void 0 ? false : _ref$indeterminate,_ref$hasError = _ref.hasError,hasError = _ref$hasError === void 0 ? false : _ref$hasError,disabled = _ref.disabled,rest = _objectWithoutProperties(_ref, ["id", "name", "label", "className", "indeterminate", "hasError", "disabled"]);
  var inputRef = (0, _react.useRef)(null);

  (0, _react.useEffect)(function () {
    if (inputRef.current) {
      inputRef.current.indeterminate = indeterminate;
    }

    if (ref) {
      typeof ref === 'function' ?
      ref(inputRef.current) :
      ref.current = inputRef.current;
    }
  });

  var handleMouseUp = (0, _react.useCallback)(function () {
    if (inputRef.current) {
      inputRef.current.blur();
    }
  }, []);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(Wrapper, { hasError: hasError, disabled: disabled, className: className, children: [/*#__PURE__*/
      (0, _jsxRuntime.jsxs)(Container, { children: [/*#__PURE__*/
        (0, _jsxRuntime.jsx)(Input, _objectSpread({
          ref: inputRef,
          id: id || name,
          name: name,
          type: "checkbox",
          disabled: disabled,
          onMouseUp: handleMouseUp },
        rest)), /*#__PURE__*/

        (0, _jsxRuntime.jsx)(Checkmark, {})] }), /*#__PURE__*/

      (0, _jsxRuntime.jsx)(Label, { htmlFor: id || name, hasLabel: !!label, children:
        label })] }));



});exports.Checkbox = Checkbox;
