// feather/lib/components/ContextualHelp/ContextualHelp.js
"use strict";require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");Object.defineProperty(exports, "__esModule", { value: true });exports.ContextualHelp = void 0;require("core-js/modules/es.object.keys.js");var _react = require("react");
var _styledComponents = _interopRequireDefault(require("styled-components"));

var _TooltipWrapper = require("../../../lib/components/tooltip/TooltipWrapper");

var _elevation = require("../../../lib/elevation");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _themes = require("../../../lib/themes");
var _typography = require("../../../lib/typography");
var _zIndexes = require("../../../lib/zIndexes");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}

var TooltipContent = _styledComponents.default.div.withConfig({ displayName: "ContextualHelp__TooltipContent", componentId: "sc-1dodlgv-0" })(["padding:16px 20px;border-radius:4px;background-color:", ";border:1px solid ", ";z-index:", ";font-family:", ";font-size:14px;line-height:1.43;color:", ";", " ", ""],


(0, _cssVariables.default)('bg-1'),
(0, _cssVariables.default)('neutral-4'),
_zIndexes.ZIndexes.tooltip,
_typography.typeface.system,


(0, _cssVariables.default)('neutral-10'),
_elevation.elevation.popover,

function (props) {return props.$tooltipContentStyle;});


// `&::before` is for 1px border and `&::after` is for background.
var BorderArrow = _styledComponents.default.div.withConfig({ displayName: "ContextualHelp__BorderArrow", componentId: "sc-1dodlgv-1" })(["position:absolute;pointer-events:none;&[data-placement*='bottom']{top:0;&::before{position:absolute;left:-8px;margin-top:-8px;width:16px;height:8x;border-top-width:0;border-bottom-color:", ";}&::after{left:-7px;margin-top:-7px;width:14px;height:7px;border-top-width:0;border-bottom-color:white;}}&[data-placement*='top']{bottom:0;&::before{position:absolute;left:-8px;margin-bottom:-8px;width:16px;height:8px;border-bottom-width:0;border-top-color:", ";}&::after{left:-7px;margin-bottom:-7px;width:14px;height:7px;border-bottom-width:0;border-top-color:white;}}&[data-placement*='left']{right:0;&::before{position:absolute;top:-8px;margin-right:-8px;width:8px;height:16x;border-right-width:0;border-left-color:", ";}&::after{top:-7px;margin-right:-7px;width:7px;height:14px;border-right-width:0;border-left-color:white;}}&[data-placement*='right']{left:0;&::before{position:absolute;top:-8px;margin-left:-8px;width:8px;height:16x;border-left-width:0;border-right-color:", ";}&::after{top:-7px;margin-left:-7px;width:7px;height:14px;border-left-width:0;border-right-color:white;}}&::before,&::after{position:absolute;content:'';margin:auto;display:block;width:0;height:0;border-style:solid;pointer-events:none;border-color:transparent;}&::before{border-width:8px;}&::after{border-width:7px;}"],













(0, _cssVariables.default)('neutral-3'),






















(0, _cssVariables.default)('neutral-3'),






















(0, _cssVariables.default)('neutral-3'),






















(0, _cssVariables.default)('neutral-3'));

































var ContextualHelp = /*#__PURE__*/(0, _react.forwardRef)(
function (_ref, ref) {var children = _ref.children,tooltipContentStyle = _ref.tooltipContentStyle,props = _objectWithoutProperties(_ref, ["children", "tooltipContentStyle"]);
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(_TooltipWrapper.TooltipWrapper, _objectSpread(_objectSpread({}, props), {}, { tooltipRef: ref, target: children, children:
      function children(_ref2) {var ref = _ref2.ref,style = _ref2.style,placement = _ref2.placement,arrowProps = _ref2.arrowProps,onTooltipReferenceMouseLeave = _ref2.onTooltipReferenceMouseLeave;
        return /*#__PURE__*/(
          (0, _jsxRuntime.jsx)(_themes.FeatherThemeContext.Provider, { value: _themes.FeatherThemeEnum.Default, children: /*#__PURE__*/
            (0, _jsxRuntime.jsxs)(TooltipContent, {
              ref: ref,
              role: "tooltip",
              onMouseLeave: onTooltipReferenceMouseLeave,
              style: style,
              $tooltipContentStyle: tooltipContentStyle, children: [/*#__PURE__*/

              (0, _jsxRuntime.jsx)(BorderArrow, {
                ref: arrowProps.ref,
                style: arrowProps.style,
                "data-placement": placement,
                "data-is-arrow": "true" }),

              props.content] }) }));



      } })));


});exports.ContextualHelp = ContextualHelp;
