// @postman/aether/esmLib/components/Spinner/SpinnerSmall.js
import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../Text/Text';
import Flex from '../Flex/Flex';

function SpinnerSmallSVG() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "16",
    height: "16",
    x: "0",
    y: "0",
    enableBackground: "new 0 0 50 50",
    version: "1.1",
    viewBox: "0 0 16 16",
    xmlSpace: "preserve"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "4",
    height: "10",
    fill: "#333",
    opacity: "0.2",
    rx: "2",
    ry: "2",
    className: "aether-spinner-no-flicker-style"
  }, /*#__PURE__*/React.createElement("animate", {
    attributeName: "opacity",
    attributeType: "XML",
    begin: "0s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "0.2; 1; 0.2"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "height",
    attributeType: "XML",
    begin: "0s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "8; 16; 8"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "y",
    attributeType: "XML",
    begin: "0s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "5; 0; 5"
  })), /*#__PURE__*/React.createElement("rect", {
    width: "4",
    height: "10",
    x: "6",
    y: "10",
    fill: "#333",
    opacity: "0.2",
    rx: "2",
    ry: "2",
    className: "aether-spinner-no-flicker-style"
  }, /*#__PURE__*/React.createElement("animate", {
    attributeName: "opacity",
    attributeType: "XML",
    begin: "0.15s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "0.2; 1; 0.2"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "height",
    attributeType: "XML",
    begin: "0.15s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "8; 16; 8"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "y",
    attributeType: "XML",
    begin: "0.15s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "5; 0; 5"
  })), /*#__PURE__*/React.createElement("rect", {
    width: "4",
    height: "10",
    x: "12",
    y: "10",
    fill: "#333",
    opacity: "0.2",
    rx: "2",
    ry: "2",
    className: "aether-spinner-no-flicker-style"
  }, /*#__PURE__*/React.createElement("animate", {
    attributeName: "opacity",
    attributeType: "XML",
    begin: "0.3s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "0.2; 1; 0.2"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "height",
    attributeType: "XML",
    begin: "0.3s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "8; 16; 8"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "y",
    attributeType: "XML",
    begin: "0.3s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "5; 0; 5"
  })));
}

var StyledSmallSpinnerContainer = styled(Flex).withConfig({
  displayName: "SpinnerSmall__StyledSmallSpinnerContainer",
  componentId: "sc-1m7cak8-0"
})(["svg path,svg rect{fill:", ";}"], function (props) {
  return props.theme['content-color-tertiary'];
});
var StyledSmallSpinnerDescription = styled(Text).withConfig({
  displayName: "SpinnerSmall__StyledSmallSpinnerDescription",
  componentId: "sc-1m7cak8-1"
})(["display:inline-block;max-width:80ch;margin-left:", ";word-break:break-word;"], function (props) {
  return props.theme['spacing-s'];
});
export default function SpinnerSmall(props) {
  var description = props.description,
      className = props.className;
  return /*#__PURE__*/React.createElement(StyledSmallSpinnerContainer, {
    justifyContent: "flex-start",
    alignItems: "center",
    className: className
  }, /*#__PURE__*/React.createElement(SpinnerSmallSVG, null), description && /*#__PURE__*/React.createElement(StyledSmallSpinnerDescription, {
    color: "content-color-secondary"
  }, description));
}
SpinnerSmall.defaultProps = {
  description: '',
  className: ''
};
SpinnerSmall.propTypes = {
  description: PropTypes.string,
  className: PropTypes.string
};
