// @postman/aether/esmLib/components/Spinner/SpinnerLarge.js
import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../Text/Text';
var StyledSpinnerLargeContainer = styled.div.withConfig({
  displayName: "SpinnerLarge__StyledSpinnerLargeContainer",
  componentId: "sc-m1dnkq-0"
})(["height:100%;width:100%;display:flex;flex-direction:column;justify-content:center;align-items:center;background-color:transparent;svg path,svg rect{fill:", ";}"], function (props) {
  return props.theme['content-color-tertiary'];
});
var StyledSpinnerLarge = styled.div.withConfig({
  displayName: "SpinnerLarge__StyledSpinnerLarge",
  componentId: "sc-m1dnkq-1"
})(["width:36px;height:36px;margin-bottom:", ";"], function (props) {
  return props.theme['spacing-l'];
});
var StyledSpinnerDescription = styled(Text).withConfig({
  displayName: "SpinnerLarge__StyledSpinnerDescription",
  componentId: "sc-m1dnkq-2"
})(["display:inline-block;max-width:80ch;word-break:break-word;text-align:center;font-size:", ";"], function (props) {
  return props.theme['text-size-l'];
});

function SpinnerLargeSVG() {
  return /*#__PURE__*/React.createElement("svg", {
    xmlns: "http://www.w3.org/2000/svg",
    width: "36",
    height: "36",
    x: "0",
    y: "0",
    enableBackground: "new 0 0 50 50",
    version: "1.1",
    viewBox: "0 0 16 16",
    xmlSpace: "preserve"
  }, /*#__PURE__*/React.createElement("rect", {
    width: "4",
    height: "10",
    fill: "#333",
    opacity: "0.2",
    rx: "2",
    ry: "2",
    className: "aether-spinner-no-flicker-style"
  }, /*#__PURE__*/React.createElement("animate", {
    attributeName: "opacity",
    attributeType: "XML",
    begin: "0s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "0.2; 1; 0.2"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "height",
    attributeType: "XML",
    begin: "0s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "8; 16; 8"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "y",
    attributeType: "XML",
    begin: "0s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "5; 0; 5"
  })), /*#__PURE__*/React.createElement("rect", {
    width: "4",
    height: "10",
    x: "6",
    y: "10",
    fill: "#333",
    opacity: "0.2",
    rx: "2",
    ry: "2",
    className: "aether-spinner-no-flicker-style"
  }, /*#__PURE__*/React.createElement("animate", {
    attributeName: "opacity",
    attributeType: "XML",
    begin: "0.15s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "0.2; 1; 0.2"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "height",
    attributeType: "XML",
    begin: "0.15s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "8; 16; 8"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "y",
    attributeType: "XML",
    begin: "0.15s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "5; 0; 5"
  })), /*#__PURE__*/React.createElement("rect", {
    width: "4",
    height: "10",
    x: "12",
    y: "10",
    fill: "#333",
    opacity: "0.2",
    rx: "2",
    ry: "2",
    className: "aether-spinner-no-flicker-style"
  }, /*#__PURE__*/React.createElement("animate", {
    attributeName: "opacity",
    attributeType: "XML",
    begin: "0.3s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "0.2; 1; 0.2"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "height",
    attributeType: "XML",
    begin: "0.3s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "8; 16; 8"
  }), /*#__PURE__*/React.createElement("animate", {
    attributeName: "y",
    attributeType: "XML",
    begin: "0.3s",
    dur: "0.6s",
    repeatCount: "indefinite",
    values: "5; 0; 5"
  })));
}

export default function SpinnerLarge(props) {
  var className = props.className,
      description = props.description;
  return /*#__PURE__*/React.createElement(StyledSpinnerLargeContainer, {
    className: className
  }, /*#__PURE__*/React.createElement(StyledSpinnerLarge, null, /*#__PURE__*/React.createElement(SpinnerLargeSVG, null)), /*#__PURE__*/React.createElement(StyledSpinnerDescription, {
    color: "content-color-secondary"
  }, description));
}
SpinnerLarge.defaultProps = {
  className: '',
  description: ''
};
SpinnerLarge.propTypes = {
  className: PropTypes.string,
  description: PropTypes.string
};
