// @postman/aether/esmLib/components/Tab/TabPanel.js
import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
var TabPanelElement = styled.div.withConfig({
  displayName: "TabPanel__TabPanelElement",
  componentId: "sc-g8hg64-0"
})(["display:", ";color:", ";"], function (props) {
  var _isActive = props._isActive;
  return _isActive ? 'block' : 'none';
}, function (props) {
  return props.theme['content-color-primary'];
});
var TabPanel = /*#__PURE__*/React.forwardRef(function TabPanel(props, ref) {
  var _isActive = props._isActive,
      _index = props._index,
      className = props.className,
      children = props.children;
  return /*#__PURE__*/React.createElement(TabPanelElement, {
    _isActive: _isActive,
    ref: ref,
    className: className,
    key: "TAB_CONTENT ".concat(_index)
  }, children);
});
TabPanel.displayName = 'TabPanel';
export default TabPanel;
TabPanel.defaultProps = {
  _isActive: false,
  className: '',
  _index: null,
  children: null
};
TabPanel.propTypes = {
  _isActive: PropTypes.bool,
  _index: PropTypes.number,
  className: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
};
