// @splitsoftware/splitio/lib/utils/backoff/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var Backoff = /*#__PURE__*/function () {
  /**
   * Schedule function calls with exponential backoff
   *
   * @param {function} cb
   * @param {number} baseMillis
   * @param {number} maxMillis
   */
  function Backoff(cb, baseMillis, maxMillis) {
    (0, _classCallCheck2.default)(this, Backoff);
    this.baseMillis = baseMillis || Backoff.DEFAULT_BASE_MILLIS;
    this.maxMillis = maxMillis || Backoff.DEFAULT_MAX_MILLIS;
    this.attempts = 0;
    this.cb = cb;
  }
  /**
   * Schedule a next call to `cb`
   * @returns scheduled delay in milliseconds
   */


  (0, _createClass2.default)(Backoff, [{
    key: "scheduleCall",
    value: function scheduleCall() {
      var _this = this;

      var delayInMillis = Math.min(this.baseMillis * Math.pow(2, this.attempts), this.maxMillis);
      if (this.timeoutID) clearTimeout(this.timeoutID);
      this.timeoutID = setTimeout(function () {
        _this.cb();
      }, delayInMillis);
      this.attempts++;
      return delayInMillis;
    }
  }, {
    key: "reset",
    value: function reset() {
      this.attempts = 0;

      if (this.timeoutID) {
        clearTimeout(this.timeoutID);
        this.timeoutID = 0;
      }
    }
  }]);
  return Backoff;
}();

Backoff.DEFAULT_BASE_MILLIS = 1000; // 1 second

Backoff.DEFAULT_MAX_MILLIS = 1800000; // 30 minutes

var _default = Backoff;
exports.default = _default;
