// @atlaskit/onboarding/dist/esm/components/Card.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _taggedTemplateLiteral from "@babel/runtime/helpers/taggedTemplateLiteral";

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  padding-top: ", "px;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  justify-content: space-between;\n  align-items: baseline;\n  padding-bottom: ", "px;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  ", ";\n  color: inherit;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n  flex-direction: column;\n  padding: ", "px ", "px;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  ", ";\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

import React from 'react';
import styled from 'styled-components';
import Button from '@atlaskit/button/custom-theme-button';
import { createTheme } from '@atlaskit/theme/components';
import { borderRadius, gridSize, layers } from '@atlaskit/theme/constants';
import { multiply } from '@atlaskit/theme/math';
import { h600 } from '@atlaskit/theme/typography';
import { ActionItem, ActionItems } from '../styled/Dialog';
var Container = styled.div(_templateObject(), function (_ref) {
  var theme = _ref.theme;
  return theme;
});
var Body = styled.div(_templateObject2(), multiply(gridSize, 2), multiply(gridSize, 2.5));
var Heading = styled.h4(_templateObject3(), h600);
var DefaultHeader = styled.div(_templateObject4(), gridSize);
var DefaultFooter = styled.div(_templateObject5(), gridSize);
var Theme = createTheme(function () {
  return {
    container: {
      overflow: 'auto',
      borderRadius: "".concat(borderRadius(), "px"),
      height: 'fit-content',
      zIndex: "".concat(layers.spotlight() + 1)
    }
  };
});

var Card = function Card(_ref2) {
  var _ref2$actions = _ref2.actions,
      actions = _ref2$actions === void 0 ? [] : _ref2$actions,
      actionsBeforeElement = _ref2.actionsBeforeElement,
      children = _ref2.children,
      _ref2$components = _ref2.components,
      components = _ref2$components === void 0 ? {} : _ref2$components,
      image = _ref2.image,
      heading = _ref2.heading,
      headingAfterElement = _ref2.headingAfterElement,
      theme = _ref2.theme,
      innerRef = _ref2.innerRef,
      testId = _ref2.testId;
  var _components$Header = components.Header,
      Header = _components$Header === void 0 ? DefaultHeader : _components$Header,
      _components$Footer = components.Footer,
      Footer = _components$Footer === void 0 ? DefaultFooter : _components$Footer;
  return /*#__PURE__*/React.createElement(Theme.Provider, {
    value: theme
  }, /*#__PURE__*/React.createElement(Theme.Consumer, null, function (_ref3) {
    var container = _ref3.container;
    return /*#__PURE__*/React.createElement(Container, {
      theme: container,
      innerRef: innerRef,
      "data-testid": testId
    }, typeof image === 'string' ? /*#__PURE__*/React.createElement("img", {
      src: image,
      alt: ""
    }) : image, /*#__PURE__*/React.createElement(Body, null, heading || headingAfterElement ? /*#__PURE__*/React.createElement(Header, null, /*#__PURE__*/React.createElement(Heading, null, heading), headingAfterElement || /*#__PURE__*/React.createElement("span", null)) : null, children, actions.length > 0 || actionsBeforeElement ? /*#__PURE__*/React.createElement(Footer, null, actionsBeforeElement || /*#__PURE__*/React.createElement("span", null), /*#__PURE__*/React.createElement(ActionItems, null, actions.map(function (_ref4, idx) {
      var text = _ref4.text,
          key = _ref4.key,
          rest = _objectWithoutProperties(_ref4, ["text", "key"]);

      return /*#__PURE__*/React.createElement(ActionItem, {
        key: key || (typeof text === 'string' ? text : "".concat(idx))
      }, /*#__PURE__*/React.createElement(Button, rest, text));
    }))) : null));
  }));
};

export default /*#__PURE__*/React.forwardRef(function (props, ref) {
  return /*#__PURE__*/React.createElement(Card, _extends({}, props, {
    innerRef: ref
  }));
});
