// input-format/modules/input control.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.onCut = onCut;
exports.onPaste = onPaste;
exports.onChange = onChange;
exports.onKeyDown = onKeyDown;

var _edit = require('./edit');

var _edit2 = _interopRequireDefault(_edit);

var _parse3 = require('./parse');

var _parse4 = _interopRequireDefault(_parse3);

var _format2 = require('./format');

var _format3 = _interopRequireDefault(_format2);

var _dom = require('./dom');

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}

function onCut(event, input, _parse, _format, on_change) {
    // The actual cut hasn't happened just yet hence the timeout.
    setTimeout(function() {
        return format_input_text(input, _parse, _format, undefined, on_change);
    }, 0);
}

function onPaste(event, input, _parse, _format, on_change) {
    var selection = (0, _dom.getSelection)(input);

    // If selection is made,
    // just erase the selected text
    // prior to pasting
    if (selection) {
        erase_selection(input, selection);
    }

    format_input_text(input, _parse, _format, undefined, on_change);
}

function onChange(event, input, _parse, _format, on_change) {
    format_input_text(input, _parse, _format, undefined, on_change);
}

// Intercepts "Delete" and "Backspace" keys.
// (hitting "Delete" or "Backspace" at any caret
//  position should always result in rasing a digit)
function onKeyDown(event, input, _parse, _format, on_change) {
    var operation = (0, _dom.getOperation)(event);

    switch (operation) {
        case 'Delete':
        case 'Backspace':
            // Intercept this operation and perform it manually.
            event.preventDefault();

            var selection = (0, _dom.getSelection)(input);

            // If selection is made,
            // just erase the selected text,
            // and don't apply any more operations to it.
            if (selection) {
                erase_selection(input, selection);
                return format_input_text(input, _parse, _format, undefined, on_change);
            }

            // Else, perform the (character erasing) operation manually
            return format_input_text(input, _parse, _format, operation, on_change);

        default:
            // Will be handled when `onChange` fires.
    }
}

/**
 * Erases the selected text inside an `<input/>`.
 * @param  {DOMElement} input
 * @param  {Selection} selection
 */
function erase_selection(input, selection) {
    var text = input.value;
    text = text.slice(0, selection.start) + text.slice(selection.end);

    input.value = text;
    (0, _dom.setCaretPosition)(input, selection.start);
}

/**
 * Parses and re-formats `<input/>` textual value.
 * E.g. when a user enters something into the `<input/>`
 * that raw input must first be parsed and the re-formatted properly.
 * Is called either after some user input (e.g. entered a character, pasted something)
 * or after the user performed an `operation` (e.g. "Backspace", "Delete").
 * @param  {DOMElement} input
 * @param  {Function} parse
 * @param  {Function} format
 * @param  {string} [operation] - The operation that triggered `<input/>` textual value change. E.g. "Backspace", "Delete".
 * @param  {Function} onChange
 */
function format_input_text(input, _parse, _format, operation, on_change) {
    // Parse `<input/>` textual value.
    // Get `value` and `caret` position.
    var _parse2 = (0, _parse4.default)(input.value, (0, _dom.getCaretPosition)(input), _parse),
        value = _parse2.value,
        caret = _parse2.caret;

    // If a user performed an operation (e.g. "Backspace", "Delete")
    // then apply that operation and get new `value` and `caret` position.


    if (operation) {
        var operation_applied = (0, _edit2.default)(value, caret, operation);

        value = operation_applied.value;
        caret = operation_applied.caret;
    }

    // Format the `value`.
    // (and reposition the caret accordingly)
    var formatted = (0, _format3.default)(value, caret, _format);

    var text = formatted.text;
    caret = formatted.caret;

    // Set `<input/>` textual value manually
    // to prevent React from resetting the caret position
    // later inside subsequent `render()`.
    // Doesn't work for custom `inputComponent`s for some reason.
    input.value = text;
    // Position the caret properly.
    (0, _dom.setCaretPosition)(input, caret);

    // `<input/>` textual value may have changed,
    // so the parsed `value` may have changed too.
    // The `value` didn't neccessarily change
    // but it might have.
    on_change(value);
}
