// @pipedrive/insights-core/lib/builders/unifiedQuery/unifiedQueryFragments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GROUPABLE_FIELD_FRAGMENT = exports.FILTERABLE_FIELD_FRAGMENT = exports.FIELD_FRAGMENT = exports.CUSTOM_FIELDS_FRAGMENT = void 0;
exports.CUSTOM_FIELDS_FRAGMENT = `
	fragment customFieldFields on CustomField {
		id
		type
		... on CustomFieldDateRange {
			beginning_DATE_RANGE: beginning
			end_DATE_RANGE: end
		}
		... on CustomFieldDate {
			value_DATE: value
		}
		... on CustomFieldMonetary {
			value_MONETARY: value
			currency: currency
		}
		... on CustomFieldMultipleOptions {
			value_MULTIPLE_OPTIONS: value
		}
		... on CustomFieldNumerical {
			value_NUMERICAL: value
		}
		... on CustomFieldOrganization {
			value_ORGANIZATION: value
		}
		... on CustomFieldPerson {
			value_PERSON: value
		}
		... on CustomFieldSingleOption {
			value_SINGLE_OPTION: value
		}
		... on CustomFieldTimeRange {
			beginning_TIME_RANGE: beginning
			end_TIME_RANGE: end
		}
		... on CustomFieldTime {
			value_TIME: value
		}
		... on CustomFieldUser {
			value_USER: value
		}
	}
`;
exports.FIELD_FRAGMENT = `
	fragment fieldFields on InsightsField {
		id
		name
		type
		isAllowed
	}
`;
exports.FILTERABLE_FIELD_FRAGMENT = `
	fragment filterFieldFields on InsightsFieldFilterable {
		operators
  	}
`;
exports.GROUPABLE_FIELD_FRAGMENT = `
	fragment groupByFieldFields on InsightsFieldGroupable {
		settingsType
	}
`;
