// @udacity/ureact-experiments/dist/index.js
! function(t, e) {
    if ("object" == typeof exports && "object" == typeof module) module.exports = e();
    else if ("function" == typeof define && define.amd) define([], e);
    else {
        var n = e();
        for (var r in n)("object" == typeof exports ? exports : t)[r] = n[r]
    }
}(global, (function() {
    return function(t) {
        var e = {};

        function n(r) {
            if (e[r]) return e[r].exports;
            var o = e[r] = {
                i: r,
                l: !1,
                exports: {}
            };
            return t[r].call(o.exports, o, o.exports, n), o.l = !0, o.exports
        }
        return n.m = t, n.c = e, n.d = function(t, e, r) {
            n.o(t, e) || Object.defineProperty(t, e, {
                enumerable: !0,
                get: r
            })
        }, n.r = function(t) {
            "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(t, Symbol.toStringTag, {
                value: "Module"
            }), Object.defineProperty(t, "__esModule", {
                value: !0
            })
        }, n.t = function(t, e) {
            if (1 & e && (t = n(t)), 8 & e) return t;
            if (4 & e && "object" == typeof t && t && t.__esModule) return t;
            var r = Object.create(null);
            if (n.r(r), Object.defineProperty(r, "default", {
                    enumerable: !0,
                    value: t
                }), 2 & e && "string" != typeof t)
                for (var o in t) n.d(r, o, function(e) {
                    return t[e]
                }.bind(null, o));
            return r
        }, n.n = function(t) {
            var e = t && t.__esModule ? function() {
                return t.default
            } : function() {
                return t
            };
            return n.d(e, "a", e), e
        }, n.o = function(t, e) {
            return Object.prototype.hasOwnProperty.call(t, e)
        }, n.p = "", n(n.s = 2)
    }([function(t, e) {
        t.exports = require("react")
    }, function(t, e) {
        t.exports = require("lodash/isEqual")
    }, function(t, e, n) {
        "use strict";
        n.r(e);
        var r = n(0),
            o = n.n(r),
            i = n(1),
            u = n.n(i);

        function a(t, e) {
            var n, r, o, i, u;
            void 0 === t && (t = "get");
            var a = {
                method: t
            };
            return (null === (r = e) || void 0 === r ? void 0 : r.projectId) && (a.headers = ((n = {})["X-Optimizely-Project-ID"] = e.projectId, n)), "post" === t && ((null === (o = e) || void 0 === o ? void 0 : o.attributes) && (a.body = JSON.stringify({
                attributes: e.attributes
            })), (null === (i = e) || void 0 === i ? void 0 : i.eventTags) && (a.body = JSON.stringify(e.eventTags))), {
                fetchParams: a,
                queryParams: "get" === t && (null === (u = e) || void 0 === u ? void 0 : u.attributes) ? "?attributes=" + JSON.stringify(e.attributes) : ""
            }
        }
        var c = function(t, e, n) {
                var r = "/api/experiments/v1/users/" + t + "/enabled-features/" + e,
                    o = a("get", n),
                    i = o.fetchParams,
                    u = o.queryParams;
                return fetch("" + r + u, i).then((function(t) {
                    return t.json()
                })).then((function(t) {
                    return t.enabled || !1
                })).catch((function() {
                    return !1
                }))
            },
            s = function(t, e, n) {
                var r, o, i = null != (o = null === (r = n) || void 0 === r ? void 0 : r.endpoint) ? o : "activation",
                    u = "/api/experiments/v1/users/" + t + "/experiments/" + e + "/" + i,
                    c = a("activation" === i ? "post" : "get", n),
                    s = c.fetchParams,
                    f = c.queryParams;
                return fetch("" + u + f, s).then((function(t) {
                    return t.json()
                })).then((function(t) {
                    return "variation/check" === i ? t.variation ? t.variation.key : null : t.variationKey || null
                })).catch((function() {
                    return null
                }))
            },
            f = function(t, e, n) {
                var r = "/api/experiments/v1/users/" + t + "/events/" + e + "/tracking",
                    o = a("post", n),
                    i = o.fetchParams,
                    u = o.queryParams;
                return fetch("" + r + u, i).then((function(t) {
                    if (!t.ok) return new Error(t.statusText)
                }))
            };

        function l(t, e) {
            void 0 === e && (e = {});
            var n = Object.keys(e).map((function(t) {
                var n = e[t];
                return "string" == typeof n && (n = n.split(" ").join("-")), t + "_" + n
            })).join("_");
            return n ? t + "__" + n : t
        }
        var p, h = (p = function(t, e) {
                return (p = Object.setPrototypeOf || {
                        __proto__: []
                    }
                    instanceof Array && function(t, e) {
                        t.__proto__ = e
                    } || function(t, e) {
                        for (var n in e) e.hasOwnProperty(n) && (t[n] = e[n])
                    })(t, e)
            }, function(t, e) {
                function n() {
                    this.constructor = t
                }
                p(t, e), t.prototype = null === e ? Object.create(e) : (n.prototype = e.prototype, new n)
            }),
            d = function() {
                return (d = Object.assign || function(t) {
                    for (var e, n = 1, r = arguments.length; n < r; n++)
                        for (var o in e = arguments[n]) Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
                    return t
                }).apply(this, arguments)
            },
            y = function(t, e, n, r) {
                return new(n || (n = Promise))((function(o, i) {
                    function u(t) {
                        try {
                            c(r.next(t))
                        } catch (t) {
                            i(t)
                        }
                    }

                    function a(t) {
                        try {
                            c(r.throw(t))
                        } catch (t) {
                            i(t)
                        }
                    }

                    function c(t) {
                        var e;
                        t.done ? o(t.value) : (e = t.value, e instanceof n ? e : new n((function(t) {
                            t(e)
                        }))).then(u, a)
                    }
                    c((r = r.apply(t, e || [])).next())
                }))
            },
            v = function(t, e) {
                var n, r, o, i, u = {
                    label: 0,
                    sent: function() {
                        if (1 & o[0]) throw o[1];
                        return o[1]
                    },
                    trys: [],
                    ops: []
                };
                return i = {
                    next: a(0),
                    throw: a(1),
                    return: a(2)
                }, "function" == typeof Symbol && (i[Symbol.iterator] = function() {
                    return this
                }), i;

                function a(i) {
                    return function(a) {
                        return function(i) {
                            if (n) throw new TypeError("Generator is already executing.");
                            for (; u;) try {
                                if (n = 1, r && (o = 2 & i[0] ? r.return : i[0] ? r.throw || ((o = r.return) && o.call(r), 0) : r.next) && !(o = o.call(r, i[1])).done) return o;
                                switch (r = 0, o && (i = [2 & i[0], o.value]), i[0]) {
                                    case 0:
                                    case 1:
                                        o = i;
                                        break;
                                    case 4:
                                        return u.label++, {
                                            value: i[1],
                                            done: !1
                                        };
                                    case 5:
                                        u.label++, r = i[1], i = [0];
                                        continue;
                                    case 7:
                                        i = u.ops.pop(), u.trys.pop();
                                        continue;
                                    default:
                                        if (!(o = (o = u.trys).length > 0 && o[o.length - 1]) && (6 === i[0] || 2 === i[0])) {
                                            u = 0;
                                            continue
                                        }
                                        if (3 === i[0] && (!o || i[1] > o[0] && i[1] < o[3])) {
                                            u.label = i[1];
                                            break
                                        }
                                        if (6 === i[0] && u.label < o[1]) {
                                            u.label = o[1], o = i;
                                            break
                                        }
                                        if (o && u.label < o[2]) {
                                            u.label = o[2], u.ops.push(i);
                                            break
                                        }
                                        o[2] && u.ops.pop(), u.trys.pop();
                                        continue
                                }
                                i = e.call(t, u)
                            } catch (t) {
                                i = [6, t], r = 0
                            } finally {
                                n = o = 0
                            }
                            if (5 & i[0]) throw i[1];
                            return {
                                value: i[0] ? i[1] : void 0,
                                done: !0
                            }
                        }([i, a])
                    }
                }
            },
            b = o.a.createContext(null),
            m = b.Provider,
            x = b.Consumer,
            g = function(t) {
                function e() {
                    var e = null !== t && t.apply(this, arguments) || this;
                    return e.state = {
                        experiments: {},
                        features: {}
                    }, e.fetchExperimentVariation = function(t, n, r) {
                        return y(e, void 0, void 0, (function() {
                            var e, o, i, u, a, c, f;
                            return v(this, (function(p) {
                                switch (p.label) {
                                    case 0:
                                        return e = this.props, o = e.user, i = e.projectId, u = e.experimentsEnvironment, o ? (a = d(d({}, o.attributes), n), [4, s(o.id, t, {
                                            attributes: a,
                                            projectId: i,
                                            environment: u,
                                            endpoint: r
                                        })]) : [2, Promise.resolve(null)];
                                    case 1:
                                        return c = p.sent(), f = l(t, n), this.setState((function(t) {
                                            var e;
                                            return {
                                                experiments: d(d({}, t.experiments), (e = {}, e[f] = c, e))
                                            }
                                        })), [2, c]
                                }
                            }))
                        }))
                    }, e.fetchFeatureEnabled = function(t, n) {
                        return y(e, void 0, void 0, (function() {
                            var e, r, o, i, u, a, s;
                            return v(this, (function(f) {
                                switch (f.label) {
                                    case 0:
                                        return e = this.props, r = e.user, o = e.projectId, i = e.experimentsEnvironment, r ? (u = d(d({}, r.attributes), n), [4, c(r.id, t, {
                                            attributes: u,
                                            projectId: o,
                                            environment: i
                                        })]) : [2, Promise.resolve(!1)];
                                    case 1:
                                        return a = f.sent(), s = l(t, n), this.setState((function(t) {
                                            var e;
                                            return {
                                                features: d(d({}, t.features), (e = {}, e[s] = a, e))
                                            }
                                        })), [2, a]
                                }
                            }))
                        }))
                    }, e.track = function(t, n) {
                        return y(e, void 0, void 0, (function() {
                            var e, r, o, i;
                            return v(this, (function(u) {
                                return e = this.props, r = e.user, o = e.projectId, i = e.experimentsEnvironment, r ? [2, f(r.id, t, {
                                    eventTags: n,
                                    projectId: o,
                                    environment: i
                                })] : [2, Promise.resolve(new Error("Cannot track without a user"))]
                            }))
                        }))
                    }, e
                }
                return h(e, t), e.prototype.render = function() {
                    var t = d(d({}, this.state), {
                        user: this.props.user,
                        projectId: this.props.projectId,
                        fetchExperimentVariation: this.fetchExperimentVariation,
                        fetchFeatureEnabled: this.fetchFeatureEnabled,
                        track: this.track
                    });
                    return o.a.createElement(m, {
                        value: t
                    }, this.props.children)
                }, e
            }(r.Component),
            w = function() {
                return (w = Object.assign || function(t) {
                    for (var e, n = 1, r = arguments.length; n < r; n++)
                        for (var o in e = arguments[n]) Object.prototype.hasOwnProperty.call(e, o) && (t[o] = e[o]);
                    return t
                }).apply(this, arguments)
            };

        function j() {
            return Object(r.useContext)(b)
        }

        function E(t) {
            return function(e) {
                return o.a.createElement(x, null, (function(n) {
                    return n && o.a.createElement(t, w({}, e, {
                        experimentContext: n
                    }))
                }))
            }
        }
        var O = function(t) {
                return o.a.createElement(o.a.Fragment, null, t.children)
            },
            _ = function() {
                var t = function(e, n) {
                    return (t = Object.setPrototypeOf || {
                            __proto__: []
                        }
                        instanceof Array && function(t, e) {
                            t.__proto__ = e
                        } || function(t, e) {
                            for (var n in e) e.hasOwnProperty(n) && (t[n] = e[n])
                        })(e, n)
                };
                return function(e, n) {
                    function r() {
                        this.constructor = e
                    }
                    t(e, n), e.prototype = null === n ? Object.create(n) : (r.prototype = n.prototype, new r)
                }
            }(),
            P = function(t, e, n, r) {
                return new(n || (n = Promise))((function(o, i) {
                    function u(t) {
                        try {
                            c(r.next(t))
                        } catch (t) {
                            i(t)
                        }
                    }

                    function a(t) {
                        try {
                            c(r.throw(t))
                        } catch (t) {
                            i(t)
                        }
                    }

                    function c(t) {
                        var e;
                        t.done ? o(t.value) : (e = t.value, e instanceof n ? e : new n((function(t) {
                            t(e)
                        }))).then(u, a)
                    }
                    c((r = r.apply(t, e || [])).next())
                }))
            },
            k = function(t, e) {
                var n, r, o, i, u = {
                    label: 0,
                    sent: function() {
                        if (1 & o[0]) throw o[1];
                        return o[1]
                    },
                    trys: [],
                    ops: []
                };
                return i = {
                    next: a(0),
                    throw: a(1),
                    return: a(2)
                }, "function" == typeof Symbol && (i[Symbol.iterator] = function() {
                    return this
                }), i;

                function a(i) {
                    return function(a) {
                        return function(i) {
                            if (n) throw new TypeError("Generator is already executing.");
                            for (; u;) try {
                                if (n = 1, r && (o = 2 & i[0] ? r.return : i[0] ? r.throw || ((o = r.return) && o.call(r), 0) : r.next) && !(o = o.call(r, i[1])).done) return o;
                                switch (r = 0, o && (i = [2 & i[0], o.value]), i[0]) {
                                    case 0:
                                    case 1:
                                        o = i;
                                        break;
                                    case 4:
                                        return u.label++, {
                                            value: i[1],
                                            done: !1
                                        };
                                    case 5:
                                        u.label++, r = i[1], i = [0];
                                        continue;
                                    case 7:
                                        i = u.ops.pop(), u.trys.pop();
                                        continue;
                                    default:
                                        if (!(o = (o = u.trys).length > 0 && o[o.length - 1]) && (6 === i[0] || 2 === i[0])) {
                                            u = 0;
                                            continue
                                        }
                                        if (3 === i[0] && (!o || i[1] > o[0] && i[1] < o[3])) {
                                            u.label = i[1];
                                            break
                                        }
                                        if (6 === i[0] && u.label < o[1]) {
                                            u.label = o[1], o = i;
                                            break
                                        }
                                        if (o && u.label < o[2]) {
                                            u.label = o[2], u.ops.push(i);
                                            break
                                        }
                                        o[2] && u.ops.pop(), u.trys.pop();
                                        continue
                                }
                                i = e.call(t, u)
                            } catch (t) {
                                i = [6, t], r = 0
                            } finally {
                                n = o = 0
                            }
                            if (5 & i[0]) throw i[1];
                            return {
                                value: i[0] ? i[1] : void 0,
                                done: !0
                            }
                        }([i, a])
                    }
                }
            },
            C = function(t) {
                function e() {
                    var e = null !== t && t.apply(this, arguments) || this;
                    return e.state = {
                        loading: !1
                    }, e
                }
                return _(e, t), e.prototype.componentDidMount = function() {
                    this.fetchExperimentVariation()
                }, e.prototype.componentDidUpdate = function(t) {
                    u()(t, this.props) || this.state.loading || this.fetchExperimentVariation()
                }, e.prototype.fetchExperimentVariation = function() {
                    return P(this, void 0, void 0, (function() {
                        var t, e, n, r, o, i, u, a;
                        return k(this, (function(c) {
                            switch (c.label) {
                                case 0:
                                    return t = this.props, e = t.attributes, n = t.experimentContext, r = t.experimentKey, o = t.variationFetchEndpoint, i = n.experiments, u = n.fetchExperimentVariation, a = l(r, e), Object.keys(i).includes(a) ? [3, 2] : (this.setState({
                                        loading: !0
                                    }), [4, u(r, e, o)]);
                                case 1:
                                    return c.sent(), this.setState({
                                        loading: !1
                                    }), [3, 3];
                                case 2:
                                    this.state.loading && this.setState({
                                        loading: !1
                                    }), c.label = 3;
                                case 3:
                                    return [2]
                            }
                        }))
                    }))
                }, e.prototype.render = function() {
                    var t = this.props,
                        e = t.attributes,
                        n = t.children,
                        r = t.defaultComponent,
                        i = t.experimentKey,
                        u = t.experimentContext,
                        a = t.loadingComponent;
                    if (this.state.loading) return a || null;
                    var c, s, f = l(i, e),
                        p = u.experiments[f];
                    return "function" == typeof n ? n(p) : (o.a.Children.toArray(n).find((function(t) {
                        t.type === O && (t.props.variationKey === p ? c = t : t.props.default && (s = t))
                    })), c || s || r || null)
                }, e
            }(r.Component),
            S = E(C),
            F = function() {
                var t = function(e, n) {
                    return (t = Object.setPrototypeOf || {
                            __proto__: []
                        }
                        instanceof Array && function(t, e) {
                            t.__proto__ = e
                        } || function(t, e) {
                            for (var n in e) e.hasOwnProperty(n) && (t[n] = e[n])
                        })(e, n)
                };
                return function(e, n) {
                    function r() {
                        this.constructor = e
                    }
                    t(e, n), e.prototype = null === n ? Object.create(n) : (r.prototype = n.prototype, new r)
                }
            }(),
            I = function(t, e, n, r) {
                return new(n || (n = Promise))((function(o, i) {
                    function u(t) {
                        try {
                            c(r.next(t))
                        } catch (t) {
                            i(t)
                        }
                    }

                    function a(t) {
                        try {
                            c(r.throw(t))
                        } catch (t) {
                            i(t)
                        }
                    }

                    function c(t) {
                        var e;
                        t.done ? o(t.value) : (e = t.value, e instanceof n ? e : new n((function(t) {
                            t(e)
                        }))).then(u, a)
                    }
                    c((r = r.apply(t, e || [])).next())
                }))
            },
            T = function(t, e) {
                var n, r, o, i, u = {
                    label: 0,
                    sent: function() {
                        if (1 & o[0]) throw o[1];
                        return o[1]
                    },
                    trys: [],
                    ops: []
                };
                return i = {
                    next: a(0),
                    throw: a(1),
                    return: a(2)
                }, "function" == typeof Symbol && (i[Symbol.iterator] = function() {
                    return this
                }), i;

                function a(i) {
                    return function(a) {
                        return function(i) {
                            if (n) throw new TypeError("Generator is already executing.");
                            for (; u;) try {
                                if (n = 1, r && (o = 2 & i[0] ? r.return : i[0] ? r.throw || ((o = r.return) && o.call(r), 0) : r.next) && !(o = o.call(r, i[1])).done) return o;
                                switch (r = 0, o && (i = [2 & i[0], o.value]), i[0]) {
                                    case 0:
                                    case 1:
                                        o = i;
                                        break;
                                    case 4:
                                        return u.label++, {
                                            value: i[1],
                                            done: !1
                                        };
                                    case 5:
                                        u.label++, r = i[1], i = [0];
                                        continue;
                                    case 7:
                                        i = u.ops.pop(), u.trys.pop();
                                        continue;
                                    default:
                                        if (!(o = (o = u.trys).length > 0 && o[o.length - 1]) && (6 === i[0] || 2 === i[0])) {
                                            u = 0;
                                            continue
                                        }
                                        if (3 === i[0] && (!o || i[1] > o[0] && i[1] < o[3])) {
                                            u.label = i[1];
                                            break
                                        }
                                        if (6 === i[0] && u.label < o[1]) {
                                            u.label = o[1], o = i;
                                            break
                                        }
                                        if (o && u.label < o[2]) {
                                            u.label = o[2], u.ops.push(i);
                                            break
                                        }
                                        o[2] && u.ops.pop(), u.trys.pop();
                                        continue
                                }
                                i = e.call(t, u)
                            } catch (t) {
                                i = [6, t], r = 0
                            } finally {
                                n = o = 0
                            }
                            if (5 & i[0]) throw i[1];
                            return {
                                value: i[0] ? i[1] : void 0,
                                done: !0
                            }
                        }([i, a])
                    }
                }
            },
            V = function(t) {
                function e() {
                    var e = null !== t && t.apply(this, arguments) || this;
                    return e.state = {
                        loading: !1
                    }, e
                }
                return F(e, t), e.prototype.componentDidMount = function() {
                    this.fetchFeatureEnabled()
                }, e.prototype.componentDidUpdate = function(t) {
                    u()(t, this.props) || this.state.loading || this.fetchFeatureEnabled()
                }, e.prototype.fetchFeatureEnabled = function() {
                    return I(this, void 0, void 0, (function() {
                        var t, e, n, r, o, i, u;
                        return T(this, (function(a) {
                            switch (a.label) {
                                case 0:
                                    return t = this.props, e = t.attributes, n = t.experimentContext, r = t.featureKey, o = n.features, i = n.fetchFeatureEnabled, u = l(r, e), Object.keys(o).includes(u) ? [3, 2] : (this.setState({
                                        loading: !0
                                    }), [4, i(r, e)]);
                                case 1:
                                    return a.sent(), this.setState({
                                        loading: !1
                                    }), [3, 3];
                                case 2:
                                    this.state.loading && this.setState({
                                        loading: !1
                                    }), a.label = 3;
                                case 3:
                                    return [2]
                            }
                        }))
                    }))
                }, e.prototype.render = function() {
                    var t = this.props,
                        e = t.attributes,
                        n = t.children,
                        r = t.defaultComponent,
                        o = t.experimentContext,
                        i = t.featureKey,
                        u = t.loadingComponent;
                    if (this.state.loading) return u || null;
                    var a = l(i, e),
                        c = o.features[a];
                    if (null != n) {
                        if ("function" == typeof n) return n(c);
                        if (c) return n
                    }
                    return r || null
                }, e
            }(r.Component),
            q = E(V);
        n.d(e, "ExperimentComponent", (function() {
            return C
        })), n.d(e, "Experiment", (function() {
            return S
        })), n.d(e, "FeatureFlagComponent", (function() {
            return V
        })), n.d(e, "FeatureFlag", (function() {
            return q
        })), n.d(e, "ExperimentContext", (function() {
            return b
        })), n.d(e, "ExperimentProvider", (function() {
            return m
        })), n.d(e, "ExperimentConsumer", (function() {
            return x
        })), n.d(e, "OptimizelyProvider", (function() {
            return g
        })), n.d(e, "Variation", (function() {
            return O
        })), n.d(e, "useExperimentContext", (function() {
            return j
        })), n.d(e, "withExperimentContext", (function() {
            return E
        }))
    }])
}));
