// anchor-server-common/utilities/sharedRequestHeaders.js
/**
 * Shared HTTP headers for cross-cutting concerns.
 */
module.exports = {
    /**
     * Optional header for 2-letter country code. Passed in by Anchor app clients to denote the user's location.
     * Values are ISO 3166-1 alpha-2 codes. Example expected values include: 'US', 'PT', 'JP'
     *
     * @see https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2
     */
    ANCHOR_CLIENT_COUNTRY: 'Anchor-Client-Country',

    /**
     * Optional header to pass in an optimizely user id.  Passed in by Anchor clients when a user is not logged in.
     */
    ANCHOR_CLIENT_OPTIMIZELY_USER_ID: 'Anchor-Client-Optimizely-User-Id',

    /**
     * Optional header for time zone. Passed in by Anchor app clients to help add specificity to the user's location.
     * Values are names from the tz database. Example expected values include: 'America/New_York', 'Europe/Lisbon', 'Asia/Tokyo'
     *
     @see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
     */
    ANCHOR_CLIENT_TIMEZONE: 'Anchor-Client-Timezone',

    /**
     * Optional header for the end client IP passed down by our CDN provider, Fastly. When inferring location from IP (e.g. via Maxmind),
     * this header should be preferred because it correctly represents the end client IP after the request has passed through proxy layers
     * in our infrastructure (load balancers, public-web proxy etc.). The behavior of `req.ip` is less clear-cut.
     */
    FASTLY_CLIENT_IP: 'fastly-client-ip',

    /**
     * Optional API version header. Passed in by Anchor app clients to request a specific API version for a given endpoint.
     */
    ANCHOR_API_VERSION: 'Anchor-API-Version',

    /**
     * Optional header added by the Anchor public-website proxy layer. Value: 'true'.
     */
    ANCHOR_API_PROXY: 'Anchor-API-Proxy',
};
