// @postman/monaco-yaml/lib/esm/monaco.contribution.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
import { setupMode } from './yamlMode';
var Emitter = monaco.Emitter;
// --- YAML configuration and defaults ---------
var LanguageServiceDefaultsImpl = /** @class */ (function () {
    function LanguageServiceDefaultsImpl(languageId, diagnosticsOptions, instanceSettings) {
        this._onDidChange = new Emitter();
        this._languageId = languageId;
        this.setDiagnosticsOptions(diagnosticsOptions);
        this._instanceSettings = instanceSettings;
    }
    Object.defineProperty(LanguageServiceDefaultsImpl.prototype, "onDidChange", {
        get: function () {
            return this._onDidChange.event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LanguageServiceDefaultsImpl.prototype, "languageId", {
        get: function () {
            return this._languageId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LanguageServiceDefaultsImpl.prototype, "diagnosticsOptions", {
        get: function () {
            return this._diagnosticsOptions;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LanguageServiceDefaultsImpl.prototype, "instanceSettings", {
        get: function () {
            return this._instanceSettings;
        },
        enumerable: true,
        configurable: true
    });
    LanguageServiceDefaultsImpl.prototype.setDiagnosticsOptions = function (options) {
        this._diagnosticsOptions = options || Object.create(null);
        this._onDidChange.fire(this);
    };
    LanguageServiceDefaultsImpl.prototype.setInstanceSettings = function (uri, setting) {
        var _this = this;
        this._instanceSettings.settings[uri] = setting;
        this._onDidChange.fire(this);
        return {
            dispose: function () {
                delete _this._instanceSettings.settings[uri];
                _this._onDidChange.fire(_this);
            },
        };
    };
    return LanguageServiceDefaultsImpl;
}());
export { LanguageServiceDefaultsImpl };
var diagnosticDefault = {
    validate: true,
    schemas: [],
    enableSchemaRequest: false,
    completion: true,
};
var instanceSettingsYamlDefault = {
    settings: {},
};
var yamlDefaults = new LanguageServiceDefaultsImpl('yaml', diagnosticDefault, instanceSettingsYamlDefault);
// Export API
function createAPI() {
    return {
        yamlDefaults: yamlDefaults,
    };
}
monaco.languages.yaml = createAPI();
// --- Registration to monaco editor ---
monaco.languages.register({
    id: 'yaml',
    extensions: ['.yaml', '.yml'],
    aliases: ['YAML', 'yaml', 'YML', 'yml'],
    mimetypes: ['application/x-yaml'],
});
monaco.languages.onLanguage('yaml', function () {
    setupMode(yamlDefaults);
});
