// feather/lib/animation.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.transitions = exports.transitionDefault = exports.rotateRightDynamic = exports.animationLoading = exports.animationFadeOut = exports.animationFadeIn = exports.animationFadeInOut = exports.animationOpacity = exports.animationScaleY = exports.animationBounceDelay = void 0;require("core-js/modules/es.array.join.js");require("core-js/modules/es.array.map.js");require("core-js/modules/es.array.concat.js");var _styledComponents = require("styled-components");


var animationBounceDelay = function animationBounceDelay() {return (0, _styledComponents.keyframes)(["0%,80%,100%{transform:scale(0);}40%{transform:scale(1.0);}"]);};exports.animationBounceDelay = animationBounceDelay;







var animationScaleY = function animationScaleY() {return (0, _styledComponents.keyframes)(["0%{transform:scaleY(0);opacity:0;}100%{transform:scaleY(1.0);opacity:1;}"]);};exports.animationScaleY = animationScaleY;









var animationOpacity = function animationOpacity() {return (0, _styledComponents.keyframes)(["from{opacity:0;}to{opacity:1;}"]);};exports.animationOpacity = animationOpacity;







var animationFadeInOut = function animationFadeInOut() {return (0, _styledComponents.keyframes)(["0%{opacity:0;}50%{opacity:1;}100%{opacity:0;}"]);};exports.animationFadeInOut = animationFadeInOut;









var animationFadeIn = function animationFadeIn() {return (0, _styledComponents.keyframes)(["0%{opacity:0;display:none;}100%{opacity:1;display:block;}"]);};exports.animationFadeIn = animationFadeIn;









var animationFadeOut = function animationFadeOut() {return (0, _styledComponents.keyframes)(["0%{opacity:1;}100%{opacity:0;display:none;}"]);};exports.animationFadeOut = animationFadeOut;









var animationLoading = function animationLoading() {return (0, _styledComponents.keyframes)(["100%{transform:rotate(360deg);}"]);};exports.animationLoading = animationLoading;





var rotateRightDynamic = function rotateRightDynamic() {return (0, _styledComponents.keyframes)(["0%{transform:rotate(0);}10%{transform:rotate(20deg);}20%{transform:rotate(55deg);}50%{transform:rotate(170deg);}60%{transform:rotate(220deg);}80%{transform:rotate(280deg);}90%{transform:rotate(320deg);}100%{transform:rotate(360deg);}"]);};exports.rotateRightDynamic = rotateRightDynamic;

































var transitionDefault = 'cubic-bezier(.4, 0, .2, 1)';exports.transitionDefault = transitionDefault;

var transitions = function transitions(_ref) {var duration = _ref.duration,properties = _ref.properties;return (
    properties.map(function (property) {return "".concat(property, " ").concat(duration, "s ").concat(transitionDefault);}).join(', '));};exports.transitions = transitions;
