// feather/lib/colors.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.Colors = void 0; /**
 * @deprecated
 * The new color will replace the Color object below.
 */


/**
 * @deprecated Colors
 * The new color will replace the Color object below.
 */
var Colors = {
  purple: {
    50: '#ededff',
    100: '#e2dfff',
    200: '#9e8cf5',
    300: '#825eeb',
    400: '#6440c4',
    500: '#4d29a6' },

  red: {
    50: '#ffebee',
    100: '#ffccd4',
    200: '#ff949e',
    300: '#f24d6b',
    400: '#d92148',
    500: '#8d3149' },

  green: {
    50: '#e3fcf4',
    100: '#aef2dc',
    200: '#6fd6be',
    300: '#2fba9f',
    400: '#189a8d',
    500: '#007a7a' },

  indigo: {
    50: '#deefff',
    100: '#baddff',
    200: '#82baff',
    300: '#5292fa',
    400: '#2876e8',
    500: '#0050bf' },

  yellow: {
    50: '#fff8de',
    100: '#fff2b6',
    200: '#ffd873',
    300: '#ffc233',
    400: '#e47f2c',
    500: '#d26b2e' },

  orange: {
    50: '#ffeadd',
    100: '#ffdaba',
    200: '#ffb581',
    300: '#ff8d4b',
    400: '#ff6d44',
    500: '#e64d2e' },

  pink: {
    50: '#ffedfa',
    100: '#ffccf2',
    200: '#ffa6e8',
    300: '#fb70cd',
    400: '#d941b3',
    500: '#8d3a7f' },

  violet: {
    50: '#f6ecff',
    100: '#e3c7ff',
    200: '#c58efa',
    300: '#ad5ce6',
    400: '#8e3ebd',
    500: '#7c29a6' },

  navy: {
    10: '#fafbfd',
    50: '#f6f8fc',
    80: '#eef2fa',
    100: '#dee2f2',
    200: '#c9d0e6',
    300: '#b6bdd7',
    400: '#8a92ba',
    500: '#747aa5',
    600: '#595e8a',
    700: '#444974',
    800: '#353761',
    900: '#212242' } };exports.Colors = Colors;
