// feather/lib/elevation.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.elevation = exports.shadow = void 0;var _styledComponents = require("styled-components");
var _cssVariables = _interopRequireDefault(require("../lib/theme/cssVariables"));function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

var shadow = {
  0: (0, _styledComponents.css)(["box-shadow:none;"]),


  1: (0, _styledComponents.css)(["box-shadow:0 0 2px 0 rgba(13,13,13,0.32);"]),


  2: (0, _styledComponents.css)(["box-shadow:0 1px 5px 0 rgba(13,13,13,0.12),0 0 1px 0 ", ",0 2px 2px 0 ", ";"],
  (0, _cssVariables.default)('bg-overlay-3'),
  (0, _cssVariables.default)('border-overlay')),

  8: (0, _styledComponents.css)(["box-shadow:0 3px 5px -3px ", ",0 3px 14px 2px ", ",0 8px 10px 1px rgba(13,13,13,0.12);;"],
  (0, _cssVariables.default)('bg-overlay-1'), (0, _cssVariables.default)('border-overlay')),


  12: (0, _styledComponents.css)(["box-shadow:0 5px 8px -4px ", ",0 5px 22px 4px ", ",0 12px 17px 2px rgba(13,13,13,0.12);"],
  (0, _cssVariables.default)('bg-overlay-1'), (0, _cssVariables.default)('border-overlay')),


  16: (0, _styledComponents.css)(["box-shadow:0 6px 10px -5px ", ",0 6px 30px 5px ", ",0 16px 24px 2px rgba(13,13,13,0.12);"],
  (0, _cssVariables.default)('bg-overlay-1'), (0, _cssVariables.default)('border-overlay')),


  24: (0, _styledComponents.css)(["box-shadow:0 9px 15px -7px ", ",0 9px 46px 8px ", ",0 24px 38px 3px rgba(13,13,13,0.12);"],
  (0, _cssVariables.default)('bg-overlay-1'), (0, _cssVariables.default)('border-overlay')) };exports.shadow = shadow;




var elevation = {
  modal: shadow[24],
  navi: shadow[16],
  popover: shadow[8],
  card: shadow[2],
  flat: shadow[0] };exports.elevation = elevation;
