// feather/lib/typography.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.Body = exports.Subtitles = exports.Headings = exports.Typography = exports.typeface = void 0;var _styledComponents = require("styled-components");

// https://css-tricks.com/snippets/css/system-font-stack/

var typeface = {
  system:
  "-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen-Sans, Ubuntu, Cantarell, 'Helvetica Neue', Arial, sans-serif",
  avenir: "'Avenir Next', 'Helvetica Neue', helvetica, Arial, sans-serif",
  robotoMono: "'Roboto Mono', 'Andale Mono', Consolas, Monaco, 'Ubuntu Mono', monospace" };exports.typeface = typeface;


var Typography = {
  'label-01': (0, _styledComponents.css)(["font-family:", ";text-transform:uppercase;font-size:11px;font-weight:600;line-height:1.09;"],
  typeface.system),





  'label-02': (0, _styledComponents.css)(["font-family:", ";font-size:12px;font-weight:500;line-height:1;"],
  typeface.system),




  'label-03': (0, _styledComponents.css)(["font-family:", ";font-size:13px;font-weight:600;line-height:1.23;"],
  typeface.system),




  'caption-01': (0, _styledComponents.css)(["font-family:", ";font-size:12px;font-weight:400;line-height:1.33;"],
  typeface.system),




  'code-01': (0, _styledComponents.css)(["font-family:", ";font-size:13px;font-weight:400;line-height:1.54;letter-spacing:-0.3px;"],
  typeface.robotoMono),





  'unit-01': (0, _styledComponents.css)(["font-family:", ";font-size:10px;font-weight:400;line-height:1.2;"],
  typeface.system) };exports.Typography = Typography;






var Headings = {
  'heading-01': (0, _styledComponents.css)(["font-family:", ";font-size:14px;font-weight:600;line-height:1.43;letter-spacing:-0.1px;"],
  typeface.system),





  'heading-02': (0, _styledComponents.css)(["font-family:", ";font-size:16px;font-weight:600;line-height:1.25;letter-spacing:-0.15px;"],
  typeface.system),





  'heading-03': (0, _styledComponents.css)(["font-family:Gellix;font-size:18px;font-weight:600;line-height:1.33;letter-spacing:-0.25px;&:before{content:'';display:table;margin-bottom:-2px;}&:after{content:'';display:table;margin-top:2px;}"]),
















  'heading-04': (0, _styledComponents.css)(["font-family:Gellix;font-size:20px;font-weight:600;line-height:1.4;letter-spacing:-0.25px;&:before{content:'';display:table;margin-bottom:-2px;}&:after{content:'';display:table;margin-top:2px;}"]),
















  'heading-05': (0, _styledComponents.css)(["font-family:Gellix;font-size:24px;font-weight:600;line-height:1.33;letter-spacing:-0.25px;&:before{content:'';display:table;margin-bottom:-1px;}&:after{content:'';display:table;margin-top:1px;}"]),
















  'heading-06': (0, _styledComponents.css)(["font-family:Gellix;font-size:36px;font-weight:600;line-height:1.22;letter-spacing:-0.5px;&:before{content:'';display:table;margin-bottom:-2px;}&:after{content:'';display:table;margin-top:2px;}"]),
















  'heading-07': (0, _styledComponents.css)(["font-family:Gellix;font-size:44px;font-weight:600;line-height:1.27;letter-spacing:-1px;&:before{content:'';display:table;margin-bottom:-3px;}&:after{content:'';display:table;margin-top:3px;}"]),
















  'heading-08': (0, _styledComponents.css)(["font-family:Gellix;font-size:56px;font-weight:600;line-height:1.08;letter-spacing:-1.5px;&:before{content:'';display:table;margin-bottom:-4px;}&:after{content:'';display:table;margin-top:4px;}"]) };exports.Headings = Headings;


















var Subtitles = {
  'subtitle-01': (0, _styledComponents.css)(["font-family:", ";font-size:14px;font-weight:500;line-height:1.43;letter-spacing:-0.1px;"],
  typeface.system),





  'subtitle-02': (0, _styledComponents.css)(["font-family:", ";font-size:16px;font-weight:500;line-height:1.25;letter-spacing:-0.15px;"],
  typeface.system),





  'subtitle-03': (0, _styledComponents.css)(["font-family:", ";font-size:18px;font-weight:500;line-height:1.33;letter-spacing:-0.25px;"],
  typeface.system) };exports.Subtitles = Subtitles;







var Body = {
  'body-short-01': (0, _styledComponents.css)(["font-family:", ";font-size:14px;font-weight:400;line-height:1.43;letter-spacing:-0.1px;"],
  typeface.system),





  'body-short-02': (0, _styledComponents.css)(["font-family:", ";font-size:16px;font-weight:400;line-height:1.5;letter-spacing:-0.15px;"],
  typeface.system),





  'body-long-02': (0, _styledComponents.css)(["font-family:", ";font-size:16px;font-weight:400;letter-spacing:-0.15px;line-height:28px;"],
  typeface.system),





  'body-03': (0, _styledComponents.css)(["font-family:", ";font-size:18px;font-weight:400;line-height:1.56;letter-spacing:-0.25px;"],
  typeface.system),





  'body-04': (0, _styledComponents.css)(["font-family:", ";font-size:20px;font-weight:400;line-height:1.4;letter-spacing:-0.25px;"],
  typeface.system) };exports.Body = Body;
