// @udacity/ureact-payment/src/components/address-form-fields.js
import _ from 'lodash';
import {
    __
} from '../helpers/localization';
import countries from '../constants/countries';
import regions from '../constants/regions';

export default {
    line1: {
        type: 'text',
        errorMessage: __('Please enter an address'),
        // TODO: Get autocomplete working
        // autoComplete: 'shipping address-line1',
        placeholder: 'Address',
        maxLength: '100'
    },
    line2: {
        type: 'text',
        placeholder: __('Apt, Unit, etc.'),
        // autoComplete: 'shipping address-line2',
        maxLength: '50'
    },
    city: {
        type: 'text',
        errorMessage: __('Please enter a city'),
        // autoComplete: 'shipping locality',
        placeholder: __('City'),
        maxLength: '100'
    },
    country: {
        type: 'select',
        errorMessage: __('Please select a country'),
        placeholder: __('Country'),
        searchable: true,
        clearable: false,
        options: _.map(countries, country => ({
            value: country.country,
            label: country.text
        }))
    },
    region: (country) => ({
        type: 'select',
        errorMessage: __('Please select a state'),
        placeholder: __('State'),
        searchable: true,
        clearable: false,
        options: _.map(regions[country] || [], r => ({
            value: r.name,
            label: r.name
        }))
    }),
    postalCode: {
        type: 'text',
        errorMessage: __('Please enter a postal code'),
        placeholder: __('Postal Code'),
        // autoComplete: 'shipping postal-code',
        maxLength: '50'
    },
    phoneNumber: {
        type: 'text',
        errorMessage: __('Please enter a phone number'),
        placeholder: __('Phone Number'),
        maxLength: '14'
    }
};
