// @udacity/ureact-payment/src/components/ebanx-billing-details-form-fields.js
import _ from 'lodash';
import {
    __
} from '../helpers/localization';
import countries from '../constants/countries';
import regions from '../constants/regions';

export default {
    line1: {
        type: 'text',
        errorMessage: __('Please enter an address'),
        placeholder: __('Address'),
        maxLength: '100'
    },
    line2: {
        type: 'text',
        placeholder: __('Apt, Unit, etc.'),
        maxLength: '100'
    },
    streetNumber: {
        type: 'text',
        placeholder: __('Street number'),
        errorMessage: __('Please enter a street number'),
        maxLength: '30'
    },
    city: {
        type: 'text',
        errorMessage: __('Please enter a city'),
        placeholder: __('City'),
        maxLength: '80'
    },
    country: {
        type: 'select',
        errorMessage: __('Please select a country'),
        placeholder: __('Country'),
        searchable: true,
        clearable: false,
        options: _.map(countries, country => ({
            value: country.country,
            label: country.text
        }))
    },
    region: (country) => ({
        type: 'select',
        errorMessage: __('Please select a state'),
        placeholder: __('State'),
        searchable: true,
        clearable: false,
        options: _.map(regions[country] || [], r => ({
            value: r.region,
            label: r.name
        }))
    }),
    postalCode: {
        type: 'text',
        errorMessage: __('Please enter a postal code'),
        placeholder: __('Postal Code'),
        maxLength: '9'
    },
    phoneNumber: {
        type: 'text',
        errorMessage: __('Please enter a phone number'),
        placeholder: __('Phone Number'),
        maxLength: '15'
    },
    cpf: {
        type: 'text',
        errorMessage: __('Please enter a valid CPF'),
        placeholder: __('CPF'),
        maxLength: '14'
    },
    birthDate: {
        type: 'date',
        errorMessage: __('Please enter your birth date'),
    }
};
