// @udacity/ureact-payment/src/components/ebanx-boleto-form.jsx
import _                         from 'lodash';
import React, {Component}        from 'react';
import PropTypes                 from 'prop-types';

import {formTypes}  from '../constants/forms';
import {__}                      from '../helpers/localization';
import formHelper                from '../helpers/form-helper';
import BillingDetailsForm        from './ebanx-billing-details-form';
import ebanxBillingDetailsFields from './ebanx-billing-details-form-fields';
import ebanxLogo                 from '../images/icons/ebanx.svg';

import styles                    from './ebanx-boleto-form.module.scss';

export default class EbanxBoletoForm extends Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    validateAllFields: PropTypes.bool,
    country: PropTypes.string,
    // Accept as props for autofill
    region: PropTypes.string,
    city: PropTypes.string,
    line1: PropTypes.string,
  }

  state = {
    boleto: {
      ..._.mapValues(ebanxBillingDetailsFields, (value, key) => _.get(this.props, key, '')),
    },
    validation: {
      address: true,
    },
    valid: false
  }

  handleAddressChange = (addressProps) => {
    this.setState({
      boleto: {
        ...this.state.boleto,
        ...addressProps.fields
      },
      validation: {
        ...this.state.validation,
        address: addressProps.valid
      }
    }, this.handleOnChange);
  }

  handleOnChange = () => {
    const {boleto} = this.state;
    const valid = formHelper.allFieldsValid(formTypes.EBANX_BOLETO, boleto, boleto.country);
    this.setState({valid}, () => this.props.onChange({boleto, valid}));
  }

  validateAllFields() {
    this.handleOnChange();
  }

  render() {
    return (
      <div>
        <div className={styles.inputs_container}>
          <div className={styles.subheader}>{__('Billing Details')}</div>

          <BillingDetailsForm
            country={this.props.country}
            onChange={this.handleAddressChange}
            validateAllFields={this.props.validateAllFields}
          />

          <div className={styles.input_row}>
            <p className={styles.ebanx_boleto}>{__("Boleto generated by")}</p>
            <img src={ebanxLogo} />
          </div>
        </div>
      </div>
    );
  }
}
