// @udacity/ureact-payment/src/components/sepa-form-fields.js
import {
    __
} from '../helpers/localization';

export default {
    name: {
        type: 'text',
        errorMessage: __('Please enter your name'),
        placeholder: __('Cardholder Name'),
        // autoComplete: 'name',
        maxLength: '50',
        required: true
    },
    accountNumber: {
        type: 'text',
        errorMessage: __('Please enter a valid IBAN number'),
        placeholder: __('IBAN'),
        maxLength: '34',
        required: true
    },
    line1: {
        type: 'text',
        errorMessage: __('Please enter your address'),
        placeholder: __('Address Line 1'),
        // autoComplete: 'shipping address-line1',
        maxLength: '50',
        required: true
    },
    line2: {
        type: 'text',
        placeholder: __('Address Line 2'),
        // autoComplete: 'shipping address-line2',
        maxLength: '50'
    },
    city: {
        type: 'text',
        errorMessage: __('Please enter your city'),
        placeholder: __('City'),
        // autoComplete: 'shipping locality',
        maxLength: '50',
        required: true
    },
    postalCode: {
        type: 'text',
        errorMessage: __('Please enter your postal code'),
        placeholder: __('Postal Code'),
        // autoComplete: 'shipping postal-code',
        maxLength: '50',
        required: true
    }
};
