// @pipedrive/leads-sidebar/dist/components/CustomField/types/__generated__/CustomFieldMultipleOptions_data.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ReaderFragment } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type CustomFieldMultipleOptions_data = {
    readonly selectedOptions: ReadonlyArray<{
        readonly id: string;
        readonly label: string | null;
    }> | null;
    readonly availableOptions: ReadonlyArray<{
        readonly id: string;
        readonly label: string | null;
    }> | null;
    readonly " $fragmentRefs": FragmentRefs<"CustomField_data">;
    readonly " $refType": "CustomFieldMultipleOptions_data";
};
export type CustomFieldMultipleOptions_data$data = CustomFieldMultipleOptions_data;
export type CustomFieldMultipleOptions_data$key = {
    readonly " $data"?: CustomFieldMultipleOptions_data$data;
    readonly " $fragmentRefs": FragmentRefs<"CustomFieldMultipleOptions_data">;
};



const node: ReaderFragment = (function () {
    var v0 = [
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "id",
            "storageKey": null
        } as any,
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "label",
            "storageKey": null
        } as any
    ];
    return {
        "argumentDefinitions": [],
        "kind": "Fragment",
        "metadata": null,
        "name": "CustomFieldMultipleOptions_data",
        "selections": [
            {
                "alias": "selectedOptions",
                "args": null,
                "concreteType": "CustomFieldOption",
                "kind": "LinkedField",
                "name": "options",
                "plural": true,
                "selections": (v0 /*: any*/),
                "storageKey": null
            },
            {
                "alias": null,
                "args": null,
                "concreteType": "CustomFieldOption",
                "kind": "LinkedField",
                "name": "availableOptions",
                "plural": true,
                "selections": (v0 /*: any*/),
                "storageKey": null
            },
            {
                "args": null,
                "kind": "FragmentSpread",
                "name": "CustomField_data"
            }
        ],
        "type": "CustomFieldMultipleOptions",
        "abstractKey": null
    } as any;
})();
(node as any).hash = 'e0350fe7cb70ccb62b7deacd37dc65a2';
export default node;
