// @pipedrive/leads-sidebar/dist/components/CustomField/types/__generated__/CustomFieldMonetary_data.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ReaderFragment } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type CustomFieldMonetary_data = {
    readonly moneyValue: {
        readonly amount: string;
        readonly currency: {
            readonly code: string;
            readonly name: string;
            readonly symbol: string | null;
        };
    } | null;
    readonly " $fragmentRefs": FragmentRefs<"CustomField_data">;
    readonly " $refType": "CustomFieldMonetary_data";
};
export type CustomFieldMonetary_data$data = CustomFieldMonetary_data;
export type CustomFieldMonetary_data$key = {
    readonly " $data"?: CustomFieldMonetary_data$data;
    readonly " $fragmentRefs": FragmentRefs<"CustomFieldMonetary_data">;
};



const node: ReaderFragment = {
    "argumentDefinitions": [],
    "kind": "Fragment",
    "metadata": null,
    "name": "CustomFieldMonetary_data",
    "selections": [
        {
            "alias": "moneyValue",
            "args": null,
            "concreteType": "Money",
            "kind": "LinkedField",
            "name": "value",
            "plural": false,
            "selections": [
                {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "amount",
                    "storageKey": null
                },
                {
                    "alias": null,
                    "args": null,
                    "concreteType": "MoneyCurrency",
                    "kind": "LinkedField",
                    "name": "currency",
                    "plural": false,
                    "selections": [
                        {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "code",
                            "storageKey": null
                        },
                        {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "name",
                            "storageKey": null
                        },
                        {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "symbol",
                            "storageKey": null
                        }
                    ],
                    "storageKey": null
                }
            ],
            "storageKey": null
        },
        {
            "args": null,
            "kind": "FragmentSpread",
            "name": "CustomField_data"
        }
    ],
    "type": "CustomFieldMonetary",
    "abstractKey": null
} as any;
(node as any).hash = 'e442e239691d0f931ac5836277c3442f';
export default node;
