// @pipedrive/leads-sidebar/dist/components/CustomField/types/__generated__/CustomFieldSingleOption_data.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ReaderFragment } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type CustomFieldSingleOption_data = {
    readonly selectedOption: {
        readonly id: string;
        readonly label: string | null;
    } | null;
    readonly availableOptions: ReadonlyArray<{
        readonly id: string;
        readonly label: string | null;
    }> | null;
    readonly " $fragmentRefs": FragmentRefs<"CustomField_data">;
    readonly " $refType": "CustomFieldSingleOption_data";
};
export type CustomFieldSingleOption_data$data = CustomFieldSingleOption_data;
export type CustomFieldSingleOption_data$key = {
    readonly " $data"?: CustomFieldSingleOption_data$data;
    readonly " $fragmentRefs": FragmentRefs<"CustomFieldSingleOption_data">;
};



const node: ReaderFragment = (function () {
    var v0 = [
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "id",
            "storageKey": null
        } as any,
        {
            "alias": null,
            "args": null,
            "kind": "ScalarField",
            "name": "label",
            "storageKey": null
        } as any
    ];
    return {
        "argumentDefinitions": [],
        "kind": "Fragment",
        "metadata": null,
        "name": "CustomFieldSingleOption_data",
        "selections": [
            {
                "alias": "selectedOption",
                "args": null,
                "concreteType": "CustomFieldOption",
                "kind": "LinkedField",
                "name": "option",
                "plural": false,
                "selections": (v0 /*: any*/),
                "storageKey": null
            },
            {
                "alias": null,
                "args": null,
                "concreteType": "CustomFieldOption",
                "kind": "LinkedField",
                "name": "availableOptions",
                "plural": true,
                "selections": (v0 /*: any*/),
                "storageKey": null
            },
            {
                "args": null,
                "kind": "FragmentSpread",
                "name": "CustomField_data"
            }
        ],
        "type": "CustomFieldSingleOption",
        "abstractKey": null
    } as any;
})();
(node as any).hash = 'b45fb806b1e6b0dd27aa9138f8555cbc';
export default node;
