// @postman/aether/esmLib/components/Menu/MenuItemGroup.js
import React from 'react';
import styled from 'styled-components';
import propTypes from 'prop-types';
import MenuItem from './MenuItem';
var StyledMenuItemGroupTitle = styled.p.withConfig({
  displayName: "MenuItemGroup__StyledMenuItemGroupTitle",
  componentId: "sc-1oza99q-0"
})(["font-size:", ";line-height:", ";font-weight:", ";text-align:left;margin:", ";margin-bottom:", ";padding:", ";color:", ";"], function (props) {
  return props.theme['text-size-s'];
}, function (props) {
  return props.theme['line-height-s'];
}, function (props) {
  return props.theme['text-weight-medium'];
}, function (props) {
  return props.theme['spacing-zero'];
}, function (props) {
  return props.theme['spacing-s'];
}, function (props) {
  return "".concat(props.theme['spacing-zero'], " ").concat(props.theme['spacing-m'], ";");
}, function (props) {
  return props.theme['content-color-secondary'];
});

function MenuItemGroup(props) {
  var title = props.title,
      children = props.children,
      closeOnSelect = props.closeOnSelect,
      closeMenu = props.closeMenu;
  var clonedChildren = React.Children.map(children, function (child) {
    if ( /*#__PURE__*/React.isValidElement(child) && child.type === MenuItem) {
      return /*#__PURE__*/React.cloneElement(child, {
        closeOnSelect: closeOnSelect,
        closeMenu: closeMenu
      });
    }

    return child;
  });
  return /*#__PURE__*/React.createElement("div", {
    className: "aether-menu__item-group"
  }, /*#__PURE__*/React.createElement(StyledMenuItemGroupTitle, null, title), clonedChildren);
}

MenuItemGroup.displayName = 'MenuItemGroup';
export default MenuItemGroup;
MenuItemGroup.propTypes = {
  children: propTypes.arrayOf(propTypes.instanceOf(MenuItem)).isRequired,
  title: propTypes.string.isRequired,
  closeOnSelect: propTypes.bool.isRequired,
  closeMenu: propTypes.func.isRequired
};
