// feather/lib/components/globalNavigationBar/globalNavigationBar.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/es.weak-map.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.symbol.iterator.js");Object.defineProperty(exports, "__esModule", { value: true });exports.GlobalNavigationBar = void 0;var _react = _interopRequireWildcard(require("react"));
var _reactRouterDom = require("react-router-dom");
var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));


var _button = require("../button");
var hideOutlineEventListeners = _interopRequireWildcard(require("../button/hideOutlineEventListeners"));
var _currentUserDropdown = require("./currentUserDropdown");
var _gnbIconButton = require("./gnbIconButton");
var _gnbOverflowMenu = require("./gnbOverflowMenu");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}

var Container = _styledComponents.default.div.withConfig({ displayName: "globalNavigationBar__Container", componentId: "vupwjq-0" })(["display:flex;flex-direction:row;align-items:center;font-size:14px;height:48px;background-color:", ";padding-left:8px;padding-right:4px;", " + ", ",", " + ", ",", " + ", ",", " + ", "{margin-left:4px;}"],





(0, _cssVariables.default)('purple-7'),



_gnbIconButton.GNBIconButton, _gnbIconButton.GNBIconButton,
_gnbIconButton.GNBIconButton, _gnbOverflowMenu.GNBOverflowMenu,
_gnbOverflowMenu.GNBOverflowMenu, _gnbIconButton.GNBIconButton,
_gnbOverflowMenu.GNBOverflowMenu, _gnbOverflowMenu.GNBOverflowMenu);




var StartElementContainer = _styledComponents.default.div.withConfig({ displayName: "globalNavigationBar__StartElementContainer", componentId: "vupwjq-1" })(["flex:1;display:inherit;flex-direction:inherit;min-width:0;"]);






var logoButtonCSS = (0, _styledComponents.css)(["", " flex:none;width:127px;height:40px;"],
(0, _button.generateButtonContainerStyle)({
  contentColor: 'white',
  disabledContentColor: 'white',
  bgColor: 'transparent',
  activeBgColor: 'transparent',
  disabledBgColor: 'transparent',
  borderColor: 'transparent',
  disabledBorderColor: 'transparent',
  focusOutlineColor: 'white' }));






var LogoButton = _styledComponents.default.button.withConfig({ displayName: "globalNavigationBar__LogoButton", componentId: "vupwjq-2" })(["", ""],
logoButtonCSS);


var LogoLink = _styledComponents.default.a.withConfig({ displayName: "globalNavigationBar__LogoLink", componentId: "vupwjq-3" })(["", ""],
logoButtonCSS);


var LogoReactRouterLink = (0, _styledComponents.default)(_reactRouterDom.Link).withConfig({ displayName: "globalNavigationBar__LogoReactRouterLink", componentId: "vupwjq-4" })(["", ""],
logoButtonCSS);


var Divider = _styledComponents.default.div.withConfig({ displayName: "globalNavigationBar__Divider", componentId: "vupwjq-5" })(["flex:none;width:1px;height:20px;background-color:", ";margin:0 8px;"],



(0, _cssVariables.default)('purple-8'));



var LogoIcon = function LogoIcon() {return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)("svg", { xmlns: "http://www.w3.org/2000/svg", width: "108", height: "24", viewBox: "0 0 108 24", children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)("path", {
        fill: "#FFF",
        fillRule: "evenodd",
        d: "M10.813 3.603c4.867 0 8.813 3.946 8.813 8.813 0 4.78-3.806 8.671-8.553 8.809l-.26.004c-4.78-.005-8.672-3.81-8.81-8.558L2 12.411h4.317c-2.322-.047-4.203-1.885-4.312-4.19L2 8.008V3.603h8.813zM27.38 7.96c2.218 0 4.213 1.113 4.574 3.321l.03.225h-3.07c-.043-.541-.626-1.017-1.6-1.017-.691 0-1.21.238-1.21.735 0 1.644 5.989-.086 5.989 4.346 0 1.946-1.838 3.438-4.238 3.438-2.683 0-4.574-1.321-4.883-3.59l-.025-.215h2.919c.065.713.822 1.189 1.838 1.189.908 0 1.427-.303 1.427-.8 0-1.751-5.968-.195-5.968-4.151 0-1.968 1.6-3.481 4.217-3.481zm11.333 0c2.876 0 5.233 2.248 5.233 5.167l-.006.258c-.007.201-.02.45-.042.667l-.017.156H35.99c.216 1.384 1.276 2.27 2.703 2.27 1.119 0 1.855-.46 2.082-1.126l.037-.128h2.983c-.346 2.292-2.356 3.784-5.016 3.784-3.135 0-5.578-2.443-5.578-5.578 0-3.092 2.422-5.47 5.513-5.47zm29.097-3.828v14.66h-3.049v-1.514c-.735 1.103-1.88 1.73-3.373 1.73-2.745 0-4.93-2.486-4.93-5.535s2.185-5.513 4.93-5.513c1.404 0 2.502.555 3.24 1.54l.133.19V4.131h3.049zm4.714 0v5.536c.735-1.103 1.88-1.708 3.35-1.708 2.747 0 4.93 2.464 4.93 5.513 0 3.048-2.183 5.535-4.93 5.535-1.383 0-2.48-.536-3.216-1.519l-.134-.189v1.492h-3.049V4.132h3.049zm32.475 0v14.66h-3.049v-1.514c-.735 1.103-1.88 1.73-3.373 1.73-2.745 0-4.93-2.486-4.93-5.535s2.185-5.513 4.93-5.513c1.405 0 2.502.555 3.24 1.54l.133.19V4.131H105zM50.976 7.96c2.648 0 4.275 1.716 4.364 4.573l.003.248v6.01h-3.048v-5.62c0-1.384-.843-2.293-2.098-2.293-1.22 0-2.024.831-2.092 2.115l-.005.177v5.622h-3.048V8.176H48.1V9.56c.67-1.017 1.665-1.6 2.876-1.6zm41.437 0c.243 0 .498.036.757.082l.259.047v2.768c-.368-.087-.735-.151-1.103-.151-1.418 0-2.574.824-2.655 2.647l-.004.207v5.232h-3.049V8.176h3.049v1.21c.627-.907 1.556-1.426 2.746-1.426zm-7.458.216v10.616h-3.049V8.176h3.049zm-22.852 2.68c-1.47 0-2.66 1.169-2.66 2.617 0 1.47 1.19 2.616 2.66 2.616 1.492 0 2.66-1.124 2.66-2.616 0-1.47-1.168-2.616-2.66-2.616zm13.081 0c-1.492 0-2.66 1.147-2.66 2.617 0 1.492 1.168 2.616 2.66 2.616 1.47 0 2.638-1.146 2.638-2.616 0-1.448-1.168-2.616-2.638-2.616zm24.108 0c-1.47 0-2.66 1.169-2.66 2.617 0 1.47 1.19 2.616 2.66 2.616 1.492 0 2.66-1.124 2.66-2.616 0-1.47-1.168-2.616-2.66-2.616zm-60.707-.432c-1.106 0-1.98.524-2.381 1.646l-.063.192h4.627c-.086-1.08-1.06-1.838-2.183-1.838zM83.42 3.603c1.016 0 1.859.843 1.859 1.838 0 1.016-.843 1.86-1.86 1.86-1.016 0-1.859-.844-1.859-1.86 0-.995.843-1.838 1.86-1.838z" }) }));};




/**
 * If you want to add buttons on both sides of `GlobalNavigationBar`, just use `GNBIconButton`, `GNBOverflowMenu`, and
 * `GNBDropdown` for the consistent style.
 */
var GlobalNavigationBar = function GlobalNavigationBar(_ref)











{var className = _ref.className,onLogoClick = _ref.onLogoClick,logoHref = _ref.logoHref,organization = _ref.organization,user = _ref.user,organizationMenus = _ref.organizationMenus,userMenus = _ref.userMenus,startButtonNode = _ref.startButtonNode,endButtonNode = _ref.endButtonNode,userDropdownDownshiftProps = _ref.userDropdownDownshiftProps,userDropdownPopperProps = _ref.userDropdownPopperProps;
  var logoButton = (0, _react.useMemo)(function () {
    if (logoHref) {
      if (typeof logoHref === 'string' || !logoHref.useReactRouterLink) {
        var linkURL = typeof logoHref === 'string' ? logoHref : logoHref.url;
        return /*#__PURE__*/(
          (0, _jsxRuntime.jsx)(LogoLink, _objectSpread(_objectSpread({}, hideOutlineEventListeners), {}, { onClick: onLogoClick, href: linkURL, children: /*#__PURE__*/
            (0, _jsxRuntime.jsx)(LogoIcon, {}) })));


      }
      return /*#__PURE__*/(
        (0, _jsxRuntime.jsx)(LogoReactRouterLink, _objectSpread(_objectSpread({}, hideOutlineEventListeners), {}, { onClick: onLogoClick, to: logoHref.url, children: /*#__PURE__*/
          (0, _jsxRuntime.jsx)(LogoIcon, {}) })));


    }
    return /*#__PURE__*/(
      (0, _jsxRuntime.jsx)(LogoButton, _objectSpread(_objectSpread({}, hideOutlineEventListeners), {}, { type: "button", onClick: onLogoClick, children: /*#__PURE__*/
        (0, _jsxRuntime.jsx)(LogoIcon, {}) })));


  }, [logoHref, onLogoClick]);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(Container, { className: className, children: [/*#__PURE__*/
      (0, _jsxRuntime.jsxs)(StartElementContainer, { children: [
        logoButton,
        startButtonNode] }),

      endButtonNode, /*#__PURE__*/
      (0, _jsxRuntime.jsx)(Divider, {}), /*#__PURE__*/
      (0, _jsxRuntime.jsx)(_currentUserDropdown.CurrentUserDropdown, {
        user: user,
        organization: organization,
        userMenus: userMenus,
        organizationMenus: organizationMenus,
        downshiftProps: userDropdownDownshiftProps,
        popperProps: userDropdownPopperProps })] }));



};exports.GlobalNavigationBar = GlobalNavigationBar;
