// feather/lib/components/globalNavigationBar/gnbDropdown.js
"use strict";require("core-js/modules/es.object.keys.js");require("core-js/modules/es.symbol.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");Object.defineProperty(exports, "__esModule", { value: true });exports.GNBDropdown = void 0;var _styledComponents = _interopRequireDefault(require("styled-components"));




var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));


var _dropdown = require("../dropdown");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}

var DropdownWrapper = _styledComponents.default.div.withConfig({ displayName: "gnbDropdown__DropdownWrapper", componentId: "sc-1c70bic-0" })(["flex-shrink:1;min-width:0;> div > button{max-width:100%;overflow-x:hidden;text-overflow:ellipsis;}"]);var _StyledDropdown = (0, _styledComponents.default)(_dropdown.Dropdown).withConfig({ displayName: "gnbDropdown___StyledDropdown", componentId: "sc-1c70bic-1" })(["height:40px;font-weight:500;color:white;padding-left:8px;white-space:nowrap;text-overflow:ellipsis;"]);










var GNBDropdown = function GNBDropdown(props) {
  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(DropdownWrapper, { children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)(_StyledDropdown, _objectSpread(_objectSpread({},
      props), {}, {
        variant: "inline",
        size: "small",
        toggleTheme: {
          contentColor: 'white',
          hoverContentColor: 'white',
          pressedContentColor: 'white',
          activeContentColor: 'white',
          hoverBgColor: (0, _cssVariables.default)('purple-8'),
          activeBgColor: (0, _cssVariables.default)('purple-9'),
          pressedBgColor: (0, _cssVariables.default)('purple-8'),
          focusOutlineColor: 'white' } })) }));












};exports.GNBDropdown = GNBDropdown;
