// @pipedrive/leads-sidebar/dist/Person/__generated__/Person_person.graphql.ts
/* tslint:disable */
/* eslint-disable */
// @ts-nocheck

import { ReaderFragment } from "relay-runtime";

import { FragmentRefs } from "relay-runtime";
export type EmailContactLabel = "HOME" | "OTHER" | "WORK" | "%future added value";
export type PhoneContactLabel = "HOME" | "MOBILE" | "OTHER" | "WORK" | "%future added value";
export type Person_person = {
    readonly relatedPersonId: string;
    readonly isLinkedToLead: boolean;
    readonly isDeleted: boolean | null;
    readonly name: string | null;
    readonly emails: ReadonlyArray<{
        readonly email: string | null;
        readonly label: EmailContactLabel | null;
    } | null> | null;
    readonly phones: ReadonlyArray<{
        readonly phone: string | null;
        readonly label: PhoneContactLabel | null;
    } | null> | null;
    readonly customFields: ReadonlyArray<{
        readonly id: string;
        readonly " $fragmentRefs": FragmentRefs<"CustomFieldRenderer_customField" | "CustomFieldRenderer_customFieldInfo">;
    } | null> | null;
    readonly " $refType": "Person_person";
};
export type Person_person$data = Person_person;
export type Person_person$key = {
    readonly " $data"?: Person_person$data;
    readonly " $fragmentRefs": FragmentRefs<"Person_person">;
};



const node: ReaderFragment = (function () {
    var v0 = [
        {
            "kind": "Literal",
            "name": "first",
            "value": 1
        } as any
    ], v1 = {
        "alias": null,
        "args": null,
        "kind": "ScalarField",
        "name": "label",
        "storageKey": null
    } as any;
    return {
        "argumentDefinitions": [],
        "kind": "Fragment",
        "metadata": null,
        "name": "Person_person",
        "selections": [
            {
                "alias": "relatedPersonId",
                "args": [
                    {
                        "kind": "Literal",
                        "name": "opaque",
                        "value": false
                    }
                ],
                "kind": "ScalarField",
                "name": "id",
                "storageKey": "id(opaque:false)"
            },
            {
                "alias": null,
                "args": null,
                "kind": "ScalarField",
                "name": "isLinkedToLead",
                "storageKey": null
            },
            {
                "alias": null,
                "args": null,
                "kind": "ScalarField",
                "name": "isDeleted",
                "storageKey": null
            },
            {
                "alias": null,
                "args": null,
                "kind": "ScalarField",
                "name": "name",
                "storageKey": null
            },
            {
                "alias": null,
                "args": (v0 /*: any*/),
                "concreteType": "EmailContact",
                "kind": "LinkedField",
                "name": "emails",
                "plural": true,
                "selections": [
                    {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "email",
                        "storageKey": null
                    },
                    (v1 /*: any*/)
                ],
                "storageKey": "emails(first:1)"
            },
            {
                "alias": null,
                "args": (v0 /*: any*/),
                "concreteType": "PhoneContact",
                "kind": "LinkedField",
                "name": "phones",
                "plural": true,
                "selections": [
                    {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "phone",
                        "storageKey": null
                    },
                    (v1 /*: any*/)
                ],
                "storageKey": "phones(first:1)"
            },
            {
                "alias": null,
                "args": null,
                "concreteType": null,
                "kind": "LinkedField",
                "name": "customFields",
                "plural": true,
                "selections": [
                    {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "id",
                        "storageKey": null
                    },
                    {
                        "args": null,
                        "kind": "FragmentSpread",
                        "name": "CustomFieldRenderer_customField"
                    },
                    {
                        "args": null,
                        "kind": "FragmentSpread",
                        "name": "CustomFieldRenderer_customFieldInfo"
                    }
                ],
                "storageKey": null
            }
        ],
        "type": "Person",
        "abstractKey": null
    } as any;
})();
(node as any).hash = 'e7405e47313be058698fe009c49508e1';
export default node;
