// @bbyca/account-components/dist/providers/LinkOrderAccountProvider/ApiLinkOrderAccountProvider.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as url from "url";
import { StatusCode } from "../../errors";
import { addCsrfToken } from "../../utils/CSRFToken";
import ApiBaseProvider from "../BaseProvider/ApiBaseProvider";
export class ApiLinkOrderAccountProvider extends ApiBaseProvider {
    constructor(cieServiceUrl) {
        super(cieServiceUrl);
        this.cieServiceUrl = cieServiceUrl;
    }
    linkOrderToAccount(props) {
        const _super = Object.create(null, {
            handleError: { get: () => super.handleError }
        });
        return __awaiter(this, void 0, void 0, function* () {
            const linkOrderToAccountUrl = url.parse(`${this.cieServiceUrl}/link-order`);
            const formattedUrl = url.format(linkOrderToAccountUrl);
            const headers = {
                "Content-Type": "application/json",
            };
            yield addCsrfToken(headers, this.cieServiceUrl);
            const response = yield fetch(formattedUrl, {
                body: JSON.stringify({
                    captchaResponse: props.captchaResponse,
                    customerId: props.customerId,
                    email: props.email,
                    orderNumber: props.orderNumber || null,
                    tid: props.tid,
                }),
                credentials: "include",
                headers,
                method: "post",
            });
            if (response.status !== StatusCode.OK) {
                return _super.handleError.call(this, response.status, formattedUrl);
            }
        });
    }
}
export default ApiLinkOrderAccountProvider;
