// @splitsoftware/splitio/lib/utils/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OPTIMIZED = exports.DEBUG = exports.FILTERS_METADATA = exports.SPLIT_EVENT = exports.SPLIT_IMPRESSION = exports.SPLIT_TO_GOOGLE_ANALYTICS = exports.GOOGLE_ANALYTICS_TO_SPLIT = exports.NA = exports.UNKNOWN = exports.CONTROL_WITH_CONFIG = exports.CONTROL = exports.STORAGE_LOCALSTORAGE = exports.STORAGE_REDIS = exports.STORAGE_MEMORY = exports.CONSUMER_MODE = exports.PRODUCER_MODE = exports.STANDALONE_MODE = exports.LOCALHOST_MODE = void 0;
// SDK Modes
var LOCALHOST_MODE = 'localhost';
exports.LOCALHOST_MODE = LOCALHOST_MODE;
var STANDALONE_MODE = 'standalone';
exports.STANDALONE_MODE = STANDALONE_MODE;
var PRODUCER_MODE = 'producer';
exports.PRODUCER_MODE = PRODUCER_MODE;
var CONSUMER_MODE = 'consumer'; // Storage types

exports.CONSUMER_MODE = CONSUMER_MODE;
var STORAGE_MEMORY = 'MEMORY';
exports.STORAGE_MEMORY = STORAGE_MEMORY;
var STORAGE_REDIS = 'REDIS';
exports.STORAGE_REDIS = STORAGE_REDIS;
var STORAGE_LOCALSTORAGE = 'LOCALSTORAGE'; // Special treatments

exports.STORAGE_LOCALSTORAGE = STORAGE_LOCALSTORAGE;
var CONTROL = 'control';
exports.CONTROL = CONTROL;
var CONTROL_WITH_CONFIG = {
  treatment: CONTROL,
  config: null
}; // Constants for unknown and not-applicable values

exports.CONTROL_WITH_CONFIG = CONTROL_WITH_CONFIG;
var UNKNOWN = 'unknown';
exports.UNKNOWN = UNKNOWN;
var NA = 'NA'; // Integration types

exports.NA = NA;
var GOOGLE_ANALYTICS_TO_SPLIT = 'GOOGLE_ANALYTICS_TO_SPLIT';
exports.GOOGLE_ANALYTICS_TO_SPLIT = GOOGLE_ANALYTICS_TO_SPLIT;
var SPLIT_TO_GOOGLE_ANALYTICS = 'SPLIT_TO_GOOGLE_ANALYTICS'; // Integration data types

exports.SPLIT_TO_GOOGLE_ANALYTICS = SPLIT_TO_GOOGLE_ANALYTICS;
var SPLIT_IMPRESSION = 'IMPRESSION';
exports.SPLIT_IMPRESSION = SPLIT_IMPRESSION;
var SPLIT_EVENT = 'EVENT'; // Split filters metadata.
// Ordered according to their precedency when forming the filter query string: `&names=<values>&prefixes=<values>`

exports.SPLIT_EVENT = SPLIT_EVENT;
var FILTERS_METADATA = [{
  type: 'byName',
  maxLength: 400,
  queryParam: 'names='
}, {
  type: 'byPrefix',
  maxLength: 50,
  queryParam: 'prefixes='
}]; // Impression collection modes

exports.FILTERS_METADATA = FILTERS_METADATA;
var DEBUG = 'DEBUG';
exports.DEBUG = DEBUG;
var OPTIMIZED = 'OPTIMIZED';
exports.OPTIMIZED = OPTIMIZED;
