// slate-edit-table/dist/handlers/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onKeyDown = exports.onUpDown = exports.onBackspace = exports.onTab = exports.onModEnter = exports.onEnter = undefined;

var _onEnter = require('./onEnter');

var _onEnter2 = _interopRequireDefault(_onEnter);

var _onModEnter = require('./onModEnter');

var _onModEnter2 = _interopRequireDefault(_onModEnter);

var _onTab = require('./onTab');

var _onTab2 = _interopRequireDefault(_onTab);

var _onBackspace = require('./onBackspace');

var _onBackspace2 = _interopRequireDefault(_onBackspace);

var _onUpDown = require('./onUpDown');

var _onUpDown2 = _interopRequireDefault(_onUpDown);

var _onKeyDown = require('./onKeyDown');

var _onKeyDown2 = _interopRequireDefault(_onKeyDown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.onEnter = _onEnter2.default;
exports.onModEnter = _onModEnter2.default;
exports.onTab = _onTab2.default;
exports.onBackspace = _onBackspace2.default;
exports.onUpDown = _onUpDown2.default;
exports.onKeyDown = _onKeyDown2.default;
