// feather/lib/components/input/index.js
"use strict";function _typeof(obj) {"@babel/helpers - typeof";if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {_typeof = function _typeof(obj) {return typeof obj;};} else {_typeof = function _typeof(obj) {return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;};}return _typeof(obj);}require("core-js/modules/es.symbol.js");require("core-js/modules/es.symbol.description.js");require("core-js/modules/es.object.to-string.js");require("core-js/modules/es.symbol.iterator.js");require("core-js/modules/es.array.iterator.js");require("core-js/modules/es.string.iterator.js");require("core-js/modules/web.dom-collections.iterator.js");require("core-js/modules/es.array.slice.js");require("core-js/modules/es.array.from.js");require("core-js/modules/es.object.keys.js");require("core-js/modules/es.array.filter.js");require("core-js/modules/es.object.get-own-property-descriptor.js");require("core-js/modules/web.dom-collections.for-each.js");require("core-js/modules/es.object.get-own-property-descriptors.js");require("core-js/modules/es.weak-map.js");Object.defineProperty(exports, "__esModule", { value: true });exports.InputSelect = exports.InputTextarea = exports.InputText = exports.FormGroup = exports.ErrorMessage = exports.StyledInputTextarea = exports.StyledInputText = exports.inputSizeMap = void 0;require("core-js/modules/es.array.map.js");require("core-js/modules/es.function.name.js");var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));

var _nonSecure = require("nanoid/non-secure");

var _animation = require("../../../lib/animation");
var _tooltip = require("../../../lib/components/tooltip");
var _mixins = require("../../../lib/mixins");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));


var _button = require("../button");
var _dropdown = require("../dropdown");
var _useAnimatedErrorMessage = require("./useAnimatedErrorMessage");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _getRequireWildcardCache(nodeInterop) {if (typeof WeakMap !== "function") return null;var cacheBabelInterop = new WeakMap();var cacheNodeInterop = new WeakMap();return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {return nodeInterop ? cacheNodeInterop : cacheBabelInterop;})(nodeInterop);}function _interopRequireWildcard(obj, nodeInterop) {if (!nodeInterop && obj && obj.__esModule) {return obj;}if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {return { default: obj };}var cache = _getRequireWildcardCache(nodeInterop);if (cache && cache.has(obj)) {return cache.get(obj);}var newObj = {};var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;for (var key in obj) {if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;if (desc && (desc.get || desc.set)) {Object.defineProperty(newObj, key, desc);} else {newObj[key] = obj[key];}}}newObj.default = obj;if (cache) {cache.set(obj, newObj);}return newObj;}function _objectWithoutProperties(source, excluded) {if (source == null) return {};var target = _objectWithoutPropertiesLoose(source, excluded);var key, i;if (Object.getOwnPropertySymbols) {var sourceSymbolKeys = Object.getOwnPropertySymbols(source);for (i = 0; i < sourceSymbolKeys.length; i++) {key = sourceSymbolKeys[i];if (excluded.indexOf(key) >= 0) continue;if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;target[key] = source[key];}}return target;}function _objectWithoutPropertiesLoose(source, excluded) {if (source == null) return {};var target = {};var sourceKeys = Object.keys(source);var key, i;for (i = 0; i < sourceKeys.length; i++) {key = sourceKeys[i];if (excluded.indexOf(key) >= 0) continue;target[key] = source[key];}return target;}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) {symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});}keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(Object(source), true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(Object(source)).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}function _slicedToArray(arr, i) {return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();}function _nonIterableRest() {throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");}function _unsupportedIterableToArray(o, minLen) {if (!o) return;if (typeof o === "string") return _arrayLikeToArray(o, minLen);var n = Object.prototype.toString.call(o).slice(8, -1);if (n === "Object" && o.constructor) n = o.constructor.name;if (n === "Map" || n === "Set") return Array.from(o);if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);}function _arrayLikeToArray(arr, len) {if (len == null || len > arr.length) len = arr.length;for (var i = 0, arr2 = new Array(len); i < len; i++) {arr2[i] = arr[i];}return arr2;}function _iterableToArrayLimit(arr, i) {var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]);if (_i == null) return;var _arr = [];var _n = true;var _d = false;var _s, _e;try {for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) {_arr.push(_s.value);if (i && _arr.length === i) break;}} catch (err) {_d = true;_e = err;} finally {try {if (!_n && _i["return"] != null) _i["return"]();} finally {if (_d) throw _e;}}return _arr;}function _arrayWithHoles(arr) {if (Array.isArray(arr)) return arr;}





































































































var inputSizeMap = {
  large: { height: 48, fontSize: 16 },
  medium: { height: 40, fontSize: 14 },
  small: { height: 32, fontSize: 14 } };exports.inputSizeMap = inputSizeMap;


var defaultInputStyle = (0, _styledComponents.css)(["display:block;padding:0 16px;width:100%;color:", ";border:1px solid ", ";border-radius:4px;outline:none;background:white;transition:0.2s ", ";transition-property:color,background,border,box-shadow;will-change:color,background,border,box-shadow;-webkit-appearance:none;", " &:focus{border-color:", ";box-shadow:0 0 0 1px ", ";}&:active{border-color:", ";box-shadow:none;}&:read-only{color:", ";background:", ";border-color:", ";}&:disabled{color:", ";border-color:", ";}"],



(0, _cssVariables.default)('neutral-10'),
(0, _cssVariables.default)('neutral-4'),



_animation.transitionDefault,




_mixins.placeholder,


(0, _cssVariables.default)('purple-7'),
(0, _cssVariables.default)('purple-7'),



(0, _cssVariables.default)('purple-7'),




(0, _cssVariables.default)('content-2'),
(0, _cssVariables.default)('neutral-3'),
(0, _cssVariables.default)('neutral-3'),



(0, _cssVariables.default)('content-disabled'),
(0, _cssVariables.default)('border-disabled'));



var InputBoxSuffix = _styledComponents.default.div.withConfig({ displayName: "input__InputBoxSuffix", componentId: "mltx1b-0" })(["position:absolute;right:8px;"]);




var StyledInputText = _styledComponents.default.input.withConfig({ displayName: "input__StyledInputText", componentId: "mltx1b-1" })(["", ";height:", ";font-size:", ";cursor:", ";transition:0.2s ", ";transition-property:height;will-change:height;", " ", " ", ""],
defaultInputStyle,
function (props) {return "".concat(inputSizeMap[props.inputSize || 'medium'].height, "px");},
function (props) {return "".concat(inputSizeMap[props.inputSize || 'medium'].fontSize, "px");},
function (props) {return props.readOnly ? 'text' : 'auto';},
_animation.transitionDefault,



function (props) {return (
    props.isPrefixButtonExist ? (0,
    _styledComponents.css)(["flex:1;border-top-left-radius:0;border-bottom-left-radius:0;border-left:none;"]) :





    null);},

function (props) {return (
    props.isSuffixButtonExist ? (0,
    _styledComponents.css)(["flex:1;border-top-right-radius:0;border-bottom-right-radius:0;border-right:none;"]) :





    null);},

function (props) {return props.styles;});exports.StyledInputText = StyledInputText;


var StyledInputTextarea = _styledComponents.default.textarea.withConfig({ displayName: "input__StyledInputTextarea", componentId: "mltx1b-2" })(["", ";padding-top:12px;padding-bottom:12px;height:auto;min-height:88px;font-size:", ";resize:", ";", ";"],
defaultInputStyle,




function (props) {return "".concat(inputSizeMap[props.inputSize || 'medium'].fontSize, "px");},
function (props) {return props.resize || 'vertical';},

function (props) {return props.styles;});


// prefix / suffix button
exports.StyledInputTextarea = StyledInputTextarea;var PrefixButtonContainer = _styledComponents.default.div.withConfig({ displayName: "input__PrefixButtonContainer", componentId: "mltx1b-3" })(["display:flex;button{border-top-right-radius:0;border-bottom-right-radius:0;}button + button{border-top-left-radius:0;border-bottom-left-radius:0;border-left:0;&:hover,&:active,&:focus{border-left:0 !important;}}"]);




















var SuffixButtonContainer = _styledComponents.default.div.withConfig({ displayName: "input__SuffixButtonContainer", componentId: "mltx1b-4" })(["display:flex;button{border-top-right-radius:0;border-bottom-right-radius:0;}button + button{border-top-left-radius:0;border-bottom-left-radius:0;border-left:0;&:hover,&:active,&:focus{border-left:0 !important;}}button:first-child{border-top-left-radius:0;border-bottom-left-radius:0;}button:last-child{border-top-right-radius:4px;border-bottom-right-radius:4px;}"]);






























// form layout division
var FormLabel = _styledComponents.default.div.withConfig({ displayName: "input__FormLabel", componentId: "mltx1b-5" })(["display:flex;align-items:center;position:relative;"]);





var FormInput = _styledComponents.default.div.withConfig({ displayName: "input__FormInput", componentId: "mltx1b-6" })(["display:flex;position:relative;"]);




var FormError = _styledComponents.default.div.withConfig({ displayName: "input__FormError", componentId: "mltx1b-7" })(["position:relative;padding-top:", ";"],

function (props) {return props.hasError ? '4px' : 0;});


var FormHelper = _styledComponents.default.div.withConfig({ displayName: "input__FormHelper", componentId: "mltx1b-8" })(["position:relative;"]);



var Label = _styledComponents.default.label.withConfig({ displayName: "input__Label", componentId: "mltx1b-9" })(["display:", ";flex:1;align-items:center;height:", ";margin-bottom:6px;font-size:12px;font-weight:500;color:", ";transition:0.2s ", ";transition-property:height,color;"],
function (props) {return props.visible ? 'flex' : 'none';},


function (props) {return props.visible ? '12px' : 0;},



(0, _cssVariables.default)('neutral-10'),
_animation.transitionDefault);



var LabelExtra = _styledComponents.default.div.withConfig({ displayName: "input__LabelExtra", componentId: "mltx1b-10" })(["flex:none;"]);



// input
var InputBox = _styledComponents.default.div.withConfig({ displayName: "input__InputBox", componentId: "mltx1b-11" })(["display:flex;flex:1;align-items:center;position:relative;"]);






var HelperText = _styledComponents.default.div.withConfig({ displayName: "input__HelperText", componentId: "mltx1b-12" })(["display:block;font-size:12px;line-height:16px;color:", ";margin-top:4px;"],



(0, _cssVariables.default)('neutral-7'));



var ErrorMessageContainer = _styledComponents.default.div.withConfig({ displayName: "input__ErrorMessageContainer", componentId: "mltx1b-13" })(["height:", "px;text-align:left;transition:0.2s ", ";transition-property:height,opacity,transform;will-change:height,opacity,transform;&[aria-hidden='true']{height:0;opacity:0;transform:translateY(-12px);}"],
function (_ref) {var errorHeight = _ref.errorHeight;return errorHeight;},


_animation.transitionDefault);










var ErrorMessageText = _styledComponents.default.span.withConfig({ displayName: "input__ErrorMessageText", componentId: "mltx1b-14" })(["display:block;font-size:12px;line-height:16px;color:", ";"],



(0, _cssVariables.default)('red-5'));


var FormGroupContainer = _styledComponents.default.div.withConfig({ displayName: "input__FormGroupContainer", componentId: "mltx1b-15" })(["& + &{margin-top:16px;}", ",", "{", ";}", ";", ";"],




StyledInputText,
StyledInputTextarea,
function (props) {return (
    props.hasError ? (0,
    _styledComponents.css)(["border-color:", ";&:focus{box-shadow:none;border-color:", ";}"],
    (0, _cssVariables.default)('red-5'),



    (0, _cssVariables.default)('red-5')) :


    null);},


function (_ref2) {var iconLength = _ref2.iconLength;return (
    iconLength && (0,
    _styledComponents.css)(["", "{padding-right:", "px;}button + button{margin-left:4px;}"],
    StyledInputText,
    48 + (iconLength - 1) * 36));},






function (_ref3) {var disabled = _ref3.disabled;return (
    disabled && (0,
    _styledComponents.css)(["", "{color:", ";}svg{fill:", ";}"],
    Label,
    (0, _cssVariables.default)('neutral-5'),



    (0, _cssVariables.default)('neutral-5')));});




var ErrorMessage = /*#__PURE__*/_react.default.memo(function (_ref4) {var _errorRef$current$scr, _errorRef$current, _error$hasError, _error$message;var error = _ref4.error,fallbackErrorMessage = _ref4.fallbackErrorMessage;
  var errorRef = (0, _react.useRef)(null);
  var _useState = (0, _react.useState)((_errorRef$current$scr = (_errorRef$current = errorRef.current) === null || _errorRef$current === void 0 ? void 0 : _errorRef$current.scrollHeight) !== null && _errorRef$current$scr !== void 0 ? _errorRef$current$scr : 0),_useState2 = _slicedToArray(_useState, 2),errorHeight = _useState2[0],setErrorHeight = _useState2[1];

  var _useAnimatedErrorMess = (0, _useAnimatedErrorMessage.useAnimatedErrorMessage)((_error$hasError =
  error === null || error === void 0 ? void 0 : error.hasError) !== null && _error$hasError !== void 0 ? _error$hasError : false, (_error$message =
  error === null || error === void 0 ? void 0 : error.message) !== null && _error$message !== void 0 ? _error$message : fallbackErrorMessage),isHidden = _useAnimatedErrorMess.hasNegativeTranslateY,visibleMessage = _useAnimatedErrorMess.visibleMessage;


  (0, _react.useEffect)(function () {
    if (visibleMessage) {var _errorRef$current$scr2, _errorRef$current2;
      setErrorHeight((_errorRef$current$scr2 = (_errorRef$current2 = errorRef.current) === null || _errorRef$current2 === void 0 ? void 0 : _errorRef$current2.scrollHeight) !== null && _errorRef$current$scr2 !== void 0 ? _errorRef$current$scr2 : 0);
    }
  }, [visibleMessage]);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(ErrorMessageContainer, { "data-test-id": "ErrorMessageContainer", errorHeight: errorHeight, "aria-hidden": isHidden, children: /*#__PURE__*/
      (0, _jsxRuntime.jsx)(ErrorMessageText, { ref: errorRef, children: visibleMessage }) }));


});exports.ErrorMessage = ErrorMessage;

var FormGroup = /*#__PURE__*/_react.default.memo(
function (_ref5)














{var inputId = _ref5.inputId,inputName = _ref5.inputName,className = _ref5.className,label = _ref5.label,labelExtra = _ref5.labelExtra,labelHelperText = _ref5.labelHelperText,_ref5$icons = _ref5.icons,icons = _ref5$icons === void 0 ? [] : _ref5$icons,prefixNode = _ref5.prefixNode,suffixNode = _ref5.suffixNode,helperText = _ref5.helperText,readOnly = _ref5.readOnly,disabled = _ref5.disabled,error = _ref5.error,children = _ref5.children;
  var hasError = !!(error && error.hasError);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(FormGroupContainer, {
      className: className,
      iconLength: icons.length,
      readOnly: readOnly,
      disabled: disabled,
      hasError: hasError, children: [

      (0, _react.useMemo)(function () {
        return (
          label && /*#__PURE__*/
          (0, _jsxRuntime.jsxs)(FormLabel, { children: [/*#__PURE__*/
            (0, _jsxRuntime.jsxs)(Label, { htmlFor: inputId, visible: !!label, children: [
              label,
              labelHelperText ? /*#__PURE__*/
              (0, _jsxRuntime.jsx)(_tooltip.Tooltip, { content: labelHelperText, placement: "top", children: /*#__PURE__*/
                (0, _jsxRuntime.jsx)(_tooltip.TooltipTargetIcon, { icon: "info", size: 16 }) }) :

              null] }), /*#__PURE__*/

            (0, _jsxRuntime.jsx)(LabelExtra, { children: labelExtra })] }));



      }, [inputId, label, labelExtra, labelHelperText]), /*#__PURE__*/
      (0, _jsxRuntime.jsxs)(FormInput, { children: [
        prefixNode, /*#__PURE__*/
        (0, _jsxRuntime.jsxs)(InputBox, { children: [
          children,
          icons.length > 0 ? /*#__PURE__*/
          (0, _jsxRuntime.jsx)(InputBoxSuffix, { children:
            icons.map(
            function (iconButtonProps) {return (
                iconButtonProps.icon && /*#__PURE__*/
                (0, _jsxRuntime.jsx)(_button.IconButton, _objectSpread({

                  type: "button",
                  buttonType: "tertiary",
                  size: "small" },
                iconButtonProps), iconButtonProps.key || iconButtonProps.icon));}) }) :




          null] }),

        suffixNode] }),

      (0, _react.useMemo)(function () {
        var fallbackErrorMessage = "".concat((label !== null && label !== void 0 ? label : inputName) || 'This field', " is required.");
        return /*#__PURE__*/(
          (0, _jsxRuntime.jsx)(FormError, { hasError: hasError, children: /*#__PURE__*/
            (0, _jsxRuntime.jsx)(ErrorMessage, { error: error, fallbackErrorMessage: fallbackErrorMessage }) }));


      }, [label, inputName, hasError, error]),

      (0, _react.useMemo)(
      function () {return (
          helperText ? /*#__PURE__*/
          (0, _jsxRuntime.jsx)(FormHelper, { children: /*#__PURE__*/
            (0, _jsxRuntime.jsx)(HelperText, { children: helperText }) }) :

          null);},
      [helperText])] }));



});exports.FormGroup = FormGroup;


var useInputId = function useInputId(idPropValue) {
  var _useState3 = (0, _react.useState)((0, _nonSecure.nanoid)()),_useState4 = _slicedToArray(_useState3, 1),id = _useState4[0];
  return idPropValue || id;
};

var InputText = /*#__PURE__*/_react.default.forwardRef(
function (_ref6,





















ref)
{var id = _ref6.id,className = _ref6.className,name = _ref6.name,_ref6$type = _ref6.type,type = _ref6$type === void 0 ? 'text' : _ref6$type,icons = _ref6.icons,size = _ref6.size,label = _ref6.label,labelHelperText = _ref6.labelHelperText,labelExtra = _ref6.labelExtra,helperText = _ref6.helperText,_ref6$value = _ref6.value,value = _ref6$value === void 0 ? undefined : _ref6$value,_ref6$readOnly = _ref6.readOnly,readOnly = _ref6$readOnly === void 0 ? false : _ref6$readOnly,_ref6$disabled = _ref6.disabled,disabled = _ref6$disabled === void 0 ? false : _ref6$disabled,error = _ref6.error,prefixNode = _ref6.prefixNode,suffixNode = _ref6.suffixNode,prefixButtons = _ref6.prefixButtons,suffixButtons = _ref6.suffixButtons,inputProps = _objectWithoutProperties(_ref6, ["id", "className", "name", "type", "icons", "size", "label", "labelHelperText", "labelExtra", "helperText", "value", "readOnly", "disabled", "error", "prefixNode", "suffixNode", "prefixButtons", "suffixButtons"]);
  var inputId = useInputId(id);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsxs)(FormGroup, {
      inputId: inputId,
      className: className,
      inputName: name,
      label: label,
      labelHelperText: labelHelperText,
      labelExtra: labelExtra,
      icons: icons,
      helperText: helperText,
      readOnly: readOnly,
      disabled: disabled,
      error: error,
      prefixNode: prefixNode,
      suffixNode: suffixNode, children: [

      prefixButtons ? /*#__PURE__*/
      (0, _jsxRuntime.jsx)(PrefixButtonContainer, { children:
        prefixButtons.map(function (_ref7) {var key = _ref7.key,label = _ref7.label,rest = _objectWithoutProperties(_ref7, ["key", "label"]);return /*#__PURE__*/(
            (0, _jsxRuntime.jsx)(_button.Button, _objectSpread(_objectSpread({ disabled: disabled }, rest), {}, { children:
              label }), key));}) }) :



      null, /*#__PURE__*/
      (0, _jsxRuntime.jsx)(StyledInputText, _objectSpread({
        ref: ref,
        id: inputId,
        name: name,
        type: type,
        inputSize: size,
        value: value,
        readOnly: readOnly,
        disabled: disabled,
        isPrefixButtonExist: !!prefixButtons && prefixButtons.length > 0,
        isSuffixButtonExist: !!suffixButtons && suffixButtons.length > 0 },
      inputProps)),

      suffixButtons ? /*#__PURE__*/
      (0, _jsxRuntime.jsx)(SuffixButtonContainer, { children:
        suffixButtons.map(function (_ref8) {var key = _ref8.key,label = _ref8.label,rest = _objectWithoutProperties(_ref8, ["key", "label"]);return /*#__PURE__*/(
            (0, _jsxRuntime.jsx)(_button.Button, _objectSpread(_objectSpread({ disabled: disabled }, rest), {}, { children:
              label }), key));}) }) :



      null] }));


});exports.InputText = InputText;


var InputTextarea = /*#__PURE__*/_react.default.forwardRef(
function (_ref9,

















ref)
{var id = _ref9.id,className = _ref9.className,name = _ref9.name,icons = _ref9.icons,size = _ref9.size,label = _ref9.label,labelHelperText = _ref9.labelHelperText,helperText = _ref9.helperText,labelExtra = _ref9.labelExtra,resize = _ref9.resize,_ref9$value = _ref9.value,value = _ref9$value === void 0 ? undefined : _ref9$value,_ref9$readOnly = _ref9.readOnly,readOnly = _ref9$readOnly === void 0 ? false : _ref9$readOnly,_ref9$disabled = _ref9.disabled,disabled = _ref9$disabled === void 0 ? false : _ref9$disabled,error = _ref9.error,textareaProps = _objectWithoutProperties(_ref9, ["id", "className", "name", "icons", "size", "label", "labelHelperText", "helperText", "labelExtra", "resize", "value", "readOnly", "disabled", "error"]);
  var inputId = useInputId(id);

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(FormGroup, {
      inputId: inputId,
      className: className,
      inputName: name,
      label: label,
      labelHelperText: labelHelperText,
      labelExtra: labelExtra,
      icons: icons,
      helperText: helperText,
      readOnly: readOnly,
      disabled: disabled,
      error: error, children: /*#__PURE__*/

      (0, _jsxRuntime.jsx)(StyledInputTextarea, _objectSpread({
        ref: ref,
        inputSize: size,
        id: inputId,
        name: name,
        value: value,
        readOnly: readOnly,
        disabled: disabled,
        resize: resize },
      textareaProps)) }));



});exports.InputTextarea = InputTextarea;


var InputSelect = /*#__PURE__*/_react.default.forwardRef(
function (_ref10,
















ref)
{var id = _ref10.id,className = _ref10.className,name = _ref10.name,icons = _ref10.icons,size = _ref10.size,width = _ref10.width,label = _ref10.label,labelHelperText = _ref10.labelHelperText,labelExtra = _ref10.labelExtra,_ref10$helperText = _ref10.helperText,helperText = _ref10$helperText === void 0 ? '' : _ref10$helperText,_ref10$readOnly = _ref10.readOnly,readOnly = _ref10$readOnly === void 0 ? false : _ref10$readOnly,_ref10$disabled = _ref10.disabled,disabled = _ref10$disabled === void 0 ? false : _ref10$disabled,error = _ref10.error,dropdownProps = _objectWithoutProperties(_ref10, ["id", "className", "name", "icons", "size", "width", "label", "labelHelperText", "labelExtra", "helperText", "readOnly", "disabled", "error"]);
  var inputId = useInputId(id);

  var defaultItemToString = function defaultItemToString(item) {return "".concat(item.label);};
  var defaultItemToElement = function defaultItemToElement(item) {return item.node ? item.node : "".concat(item.label);};

  return /*#__PURE__*/(
    (0, _jsxRuntime.jsx)(FormGroup, {
      inputId: inputId,
      className: className,
      inputName: name,
      label: label,
      labelHelperText: labelHelperText,
      labelExtra: labelExtra,
      icons: icons,
      helperText: helperText,
      readOnly: readOnly,
      disabled: disabled,
      error: error, children: /*#__PURE__*/

      (0, _jsxRuntime.jsx)(_dropdown.Dropdown, _objectSpread({
        ref: ref,
        size: size,
        width: width,
        hasError: error ? error.hasError : false,
        itemToString: defaultItemToString,
        itemToElement: defaultItemToElement,
        disabled: disabled },
      dropdownProps)) }));



});exports.InputSelect = InputSelect;
