// @atlassiansox/feature-flag-web-client/dist/esm/fetcher/Fetcher.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import { EnvironmentType } from '../index';
import { ResponseError } from './errors';
var REQUEST_TIMEOUT = 2000;
var PROD_BASE_URL = 'https://api.atlassian.com/flags';
var STAGING_BASE_URL = 'https://api.stg.atlassian.com/flags';
var DEV_BASE_URL = 'https://api.dev.atlassian.com/flags';

var Fetcher = /*#__PURE__*/function () {
  function Fetcher(apiKey) {
    var serviceEnv = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : EnvironmentType.PROD;
    var baseUrl = arguments.length > 2 ? arguments[2] : undefined;

    _classCallCheck(this, Fetcher);

    this.apiKey = apiKey;
    this.baseUrl = Fetcher.buildBaseUrl(serviceEnv, baseUrl);
  }

  _createClass(Fetcher, [{
    key: "fetchFeatureFlags",
    value: function fetchFeatureFlags(user, metadata, version) {
      var requestBody = {
        apiKey: this.apiKey,
        identifier: user.identifier,
        isAnonymous: user.isAnonymous,
        additionalIdentifiers: user.additionalIdentifiers,
        customAttributes: user.custom,
        versionData: version,
        metadata: metadata
      }; // in case there are still IE users, check AbortController before init,
      // IE has no timeout applied on fetch since AbortController isn't available

      var controller = typeof AbortController !== 'undefined' ? new AbortController() : null;
      var timeoutId = setTimeout(function () {
        return controller === null || controller === void 0 ? void 0 : controller.abort();
      }, REQUEST_TIMEOUT);
      return fetch("".concat(this.baseUrl, "/api/v1/frontend/featureFlagValues"), {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-Client-Name': metadata.client.name,
          'X-Client-Version': metadata.client.version,
          'X-Api-Key': this.apiKey
        },
        body: JSON.stringify(requestBody),
        signal: controller === null || controller === void 0 ? void 0 : controller.signal
      }).then(function (response) {
        return Fetcher.handleResponseError(response);
      }).then(function (response) {
        return Fetcher.extractBody(response);
      }).catch(function (error) {
        return Promise.reject(error);
      }).finally(function () {
        return clearTimeout(timeoutId);
      });
    }
  }], [{
    key: "handleResponseError",
    value: function () {
      var _handleResponseError = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(response) {
        var body;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (response.ok) {
                  _context.next = 5;
                  break;
                }

                _context.next = 3;
                return response.text();

              case 3:
                body = _context.sent;
                throw new ResponseError(response.status, body);

              case 5:
                return _context.abrupt("return", response);

              case 6:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      function handleResponseError(_x) {
        return _handleResponseError.apply(this, arguments);
      }

      return handleResponseError;
    }()
  }, {
    key: "extractBody",
    value: function () {
      var _extractBody = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(response) {
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (!(response.status === 204)) {
                  _context2.next = 2;
                  break;
                }

                return _context2.abrupt("return", {});

              case 2:
                return _context2.abrupt("return", response.text().then(function (text) {
                  return JSON.parse(text || '{}');
                }));

              case 3:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2);
      }));

      function extractBody(_x2) {
        return _extractBody.apply(this, arguments);
      }

      return extractBody;
    }()
  }, {
    key: "buildBaseUrl",
    value: function buildBaseUrl(serviceEnv, baseUrl) {
      if (baseUrl) {
        return baseUrl;
      }

      if (serviceEnv === EnvironmentType.DEV || serviceEnv === EnvironmentType.LOCAL) {
        return DEV_BASE_URL;
      }

      if (serviceEnv === EnvironmentType.STAGING) {
        return STAGING_BASE_URL;
      }

      return PROD_BASE_URL;
    }
  }]);

  return Fetcher;
}();

export { Fetcher as default };
