// @atlassiansox/feature-flag-web-client/dist/esm/fetcher/types.js
export var ClientVisibilityState;

(function (ClientVisibilityState) {
  ClientVisibilityState["VISIBLE"] = "visible";
  ClientVisibilityState["HIDDEN"] = "hidden";
  ClientVisibilityState["PRERENDER"] = "prerender";
})(ClientVisibilityState || (ClientVisibilityState = {}));

export var ClientCauseReason;

(function (ClientCauseReason) {
  ClientCauseReason["INITIALIZATION"] = "initialization";
  ClientCauseReason["POLLING"] = "polling";
  ClientCauseReason["RETRY"] = "retry";
})(ClientCauseReason || (ClientCauseReason = {}));

export var ClientStorageState;

(function (ClientStorageState) {
  ClientStorageState["AVAILABLE"] = "available";
  ClientStorageState["NOT_AVAILABLE"] = "notAvailable";
  ClientStorageState["FULL"] = "full";
  ClientStorageState["ERRORED"] = "errored";
})(ClientStorageState || (ClientStorageState = {}));

export var ClientUserState;

(function (ClientUserState) {
  ClientUserState["NEW"] = "new";
  ClientUserState["SAME"] = "same";
  ClientUserState["SWITCHED"] = "switched";
})(ClientUserState || (ClientUserState = {}));
