// @pipedrive/insights-core/lib/builders/nestedQuery/reports/deals/funnelConversion/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../../../../types");
const constants_1 = require("../../../../../constants");
const dealsHelpers_1 = require("../../../../../helpers/dealsHelpers");
const getFunnelConversionFields = () => {
    const statsFieldsForFunnel = constants_1.deals.STATS_FIELDS.filter(item => item !== 'isOtherSegment');
    const fieldElements = [
        'conversionFrom',
        'sumConversionFrom',
        'stageId',
        'conversionToWon',
        'conversionToLost',
        ...statsFieldsForFunnel,
    ];
    fieldElements.sort();
    return dealsHelpers_1.prepareForQueryBuilder(fieldElements);
};
const getFunnelConversionQuery = (args, multiSelectFilter) => {
    return {
        field: types_1.Deals.QueryReportTypeField.FUNNEL_CONVERSION,
        args: Object.assign(Object.assign({}, args), { stages: `[${multiSelectFilter}]` }),
        fields: {
            data: getFunnelConversionFields(),
            meta: dealsHelpers_1.prepareForQueryBuilder(constants_1.deals.METADATA_FIELDS),
            summary: dealsHelpers_1.prepareForQueryBuilder([constants_1.deals.SUMMARY_TOTAL_FIELD, constants_1.deals.SUMMARY_WIN_RATE_FIELD]),
        },
    };
};
exports.default = getFunnelConversionQuery;
