// feather/lib/components/dropdown/constants.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.defaultItemToString = exports.defaultEmptyView = exports.menuTransition = exports.defaultTransition = exports.MENU_SCROLLBAR_VERTICAL_PADDING = void 0;require("core-js/modules/es.array.concat.js");var _styledComponents = _interopRequireDefault(require("styled-components"));var _animation = require("../../../lib/animation");
var _cssVariables = _interopRequireDefault(require("../../../lib/theme/cssVariables"));
var _typography = require("../../../lib/typography");var _jsxRuntime = require("react/jsx-runtime");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}

var MENU_SCROLLBAR_VERTICAL_PADDING = 8;exports.MENU_SCROLLBAR_VERTICAL_PADDING = MENU_SCROLLBAR_VERTICAL_PADDING;

var getTransitionConstantObject = function getTransitionConstantObject(durationMs) {
  var durationCss = "".concat(durationMs, "ms");
  return {
    duration: { css: durationCss, milliseconds: durationMs },
    css: "".concat(durationCss, " ").concat(_animation.transitionDefault) };

};

/** constants for the transition of dropdown elements like toggle buttons, menu items, ... */
var defaultTransition = getTransitionConstantObject(200);

/** constants for the fade transition of a dropdown menu */exports.defaultTransition = defaultTransition;
var menuTransition = getTransitionConstantObject(100);exports.menuTransition = menuTransition;var _StyledDiv = (0, _styledComponents.default)("div").withConfig({ displayName: "constants___StyledDiv", componentId: "sc-1riq5yj-0" })(["display:flex;align-items:center;justify-content:center;height:88px;color:", ";text-align:center;", ";"], function (p) {return p._css;}, function (p) {return p._css2;});

var defaultEmptyView = /*#__PURE__*/
(0, _jsxRuntime.jsx)(_StyledDiv, { _css:





  (0, _cssVariables.default)('neutral-5'), _css2:

  _typography.Body['body-short-01'], children: "No results" });exports.defaultEmptyView = defaultEmptyView;






var defaultItemToString = function defaultItemToString(item) {return "".concat(item);};exports.defaultItemToString = defaultItemToString;
