// @pipedrive/insights-core/lib/types/commonTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Interval = exports.OperandType = exports.ChartType = exports.ReportType = exports.DataType = void 0;
var DataType;
(function (DataType) {
    DataType["ACTIVITIES"] = "activities";
    DataType["DEALS"] = "deals";
    DataType["MAILS"] = "mails";
})(DataType = exports.DataType || (exports.DataType = {}));
var ReportType;
(function (ReportType) {
    ReportType["ACTIVITIES_STATS"] = "activity";
    ReportType["ACTIVITIES_LIST"] = "activity_list";
    ReportType["DEALS_STATS"] = "stats";
    ReportType["DEALS_LIST"] = "list";
    ReportType["DEALS_RECURRING_REVENUE"] = "recurring_revenue";
    ReportType["DEALS_RECURRING_REVENUE_MOVEMENT"] = "revenue_movement";
    ReportType["DEALS_REVENUE_FORECAST"] = "revenue_forecast";
    ReportType["DEALS_CONVERSION_OVERALL"] = "conversion_overall";
    ReportType["DEALS_FUNNEL_CONVERSION"] = "funnel_conversion";
    ReportType["DEALS_DURATION"] = "duration";
    ReportType["DEALS_PROGRESS"] = "progress";
    ReportType["MAILS_STATS"] = "email";
    ReportType["MAILS_LIST"] = "mails_list";
})(ReportType = exports.ReportType || (exports.ReportType = {}));
var ChartType;
(function (ChartType) {
    ChartType["COLUMN"] = "column";
    ChartType["BAR"] = "bar";
    ChartType["PIE"] = "pie";
    ChartType["TABLE"] = "table";
    ChartType["FUNNEL"] = "funnel";
    ChartType["STACKED_BAR_CHART"] = "stackedBarChart";
    ChartType["SCORECARD"] = "scorecard";
})(ChartType = exports.ChartType || (exports.ChartType = {}));
var OperandType;
(function (OperandType) {
    OperandType["EQ"] = "isExactly";
    OperandType["IN"] = "isAnyOf";
    OperandType["NOT"] = "isNotAnyOf";
    OperandType["FROM"] = "isMoreThan";
    OperandType["TO"] = "isLessThan";
    OperandType["FORMAT"] = "format";
})(OperandType = exports.OperandType || (exports.OperandType = {}));
var Interval;
(function (Interval) {
    Interval["YEAR"] = "year";
    Interval["QUARTER"] = "quarter";
    Interval["MONTH"] = "month";
    Interval["WEEK"] = "week";
    Interval["DAY"] = "day";
})(Interval = exports.Interval || (exports.Interval = {}));
