// @pipedrive/insights-core/lib/types/dealsTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryReportTypeField = exports.QueryFilterField = exports.MeasureByType = void 0;
var MeasureByType;
(function (MeasureByType) {
    MeasureByType["COUNT"] = "count";
    MeasureByType["VALUE"] = "sum";
    MeasureByType["AVERAGE_VALUE"] = "avg";
    MeasureByType["WEIGHTED_VALUE"] = "weightedValue";
    MeasureByType["PAYMENTS_AMOUNT"] = "paymentsAmount";
    MeasureByType["PAYMENTS_CHANGE_AMOUNT"] = "paymentsChangeAmount";
    MeasureByType["PRODUCTS_SUM"] = "dealProductsSum";
    MeasureByType["PRODUCTS_AMOUNT"] = "dealProductsAmount";
})(MeasureByType = exports.MeasureByType || (exports.MeasureByType = {}));
var QueryFilterField;
(function (QueryFilterField) {
    QueryFilterField["GROUP_BY"] = "groupBy";
    QueryFilterField["FILTER"] = "filter";
})(QueryFilterField = exports.QueryFilterField || (exports.QueryFilterField = {}));
var QueryReportTypeField;
(function (QueryReportTypeField) {
    QueryReportTypeField["EXPORT_LIST"] = "exportList";
    QueryReportTypeField["LIST"] = "list";
    QueryReportTypeField["STATS"] = "stats";
    QueryReportTypeField["REVENUE_FORECAST"] = "revenueForecast";
    QueryReportTypeField["CONVERSION_OVERALL"] = "overallConversion";
    QueryReportTypeField["FUNNEL_CONVERSION"] = "funnelConversion";
    QueryReportTypeField["REVENUE_MOVEMENT"] = "revenueMovements";
    QueryReportTypeField["DURATION"] = "duration";
    QueryReportTypeField["PROGRESS"] = "progress";
    QueryReportTypeField["ACTIVITY"] = "activities";
})(QueryReportTypeField = exports.QueryReportTypeField || (exports.QueryReportTypeField = {}));
