// @pipedrive/insights-core/lib/types/unifiedTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnifiedApiExportListFormat = exports.UnifiedApiQueryField = exports.UnifiedApiFieldsField = void 0;
var UnifiedApiFieldsField;
(function (UnifiedApiFieldsField) {
    UnifiedApiFieldsField["FILTER"] = "filter";
    UnifiedApiFieldsField["MEASURE_BY"] = "measureBy";
    UnifiedApiFieldsField["GROUP_BY"] = "groupBy";
    UnifiedApiFieldsField["SEGMENT_BY"] = "segmentBy";
})(UnifiedApiFieldsField = exports.UnifiedApiFieldsField || (exports.UnifiedApiFieldsField = {}));
var UnifiedApiQueryField;
(function (UnifiedApiQueryField) {
    UnifiedApiQueryField["REPORT"] = "report";
    UnifiedApiQueryField["LIST"] = "list";
    UnifiedApiQueryField["EXPORT_LIST"] = "exportList";
    UnifiedApiQueryField["FIELDS"] = "fields";
    UnifiedApiQueryField["SUMMARY"] = "summary";
})(UnifiedApiQueryField = exports.UnifiedApiQueryField || (exports.UnifiedApiQueryField = {}));
var UnifiedApiExportListFormat;
(function (UnifiedApiExportListFormat) {
    UnifiedApiExportListFormat["CSV"] = "csv";
    UnifiedApiExportListFormat["XLS"] = "xls";
})(UnifiedApiExportListFormat = exports.UnifiedApiExportListFormat || (exports.UnifiedApiExportListFormat = {}));
