// @udacity/ureact-i18n/build/index.js
! function(e, t) {
    if ("object" == typeof exports && "object" == typeof module) module.exports = t();
    else if ("function" == typeof define && define.amd) define([], t);
    else {
        var r = t();
        for (var n in r)("object" == typeof exports ? exports : e)[n] = r[n]
    }
}(global, (function() {
    return function(e) {
        var t = {};

        function r(n) {
            if (t[n]) return t[n].exports;
            var o = t[n] = {
                i: n,
                l: !1,
                exports: {}
            };
            return e[n].call(o.exports, o, o.exports, r), o.l = !0, o.exports
        }
        return r.m = e, r.c = t, r.d = function(e, t, n) {
            r.o(e, t) || Object.defineProperty(e, t, {
                enumerable: !0,
                get: n
            })
        }, r.r = function(e) {
            "undefined" != typeof Symbol && Symbol.toStringTag && Object.defineProperty(e, Symbol.toStringTag, {
                value: "Module"
            }), Object.defineProperty(e, "__esModule", {
                value: !0
            })
        }, r.t = function(e, t) {
            if (1 & t && (e = r(e)), 8 & t) return e;
            if (4 & t && "object" == typeof e && e && e.__esModule) return e;
            var n = Object.create(null);
            if (r.r(n), Object.defineProperty(n, "default", {
                    enumerable: !0,
                    value: e
                }), 2 & t && "string" != typeof e)
                for (var o in e) r.d(n, o, function(t) {
                    return e[t]
                }.bind(null, o));
            return n
        }, r.n = function(e) {
            var t = e && e.__esModule ? function() {
                return e.default
            } : function() {
                return e
            };
            return r.d(t, "a", t), t
        }, r.o = function(e, t) {
            return Object.prototype.hasOwnProperty.call(e, t)
        }, r.p = "", r(r.s = 1)
    }([, function(e, t, r) {
        "use strict";
        Object.defineProperty(t, "__esModule", {
            value: !0
        }), Object.defineProperty(t, "I18n", {
            enumerable: !0,
            get: function() {
                return n.I18n
            }
        }), Object.defineProperty(t, "I18N_CONSTANTS", {
            enumerable: !0,
            get: function() {
                return n.I18N_CONSTANTS
            }
        });
        var n = r(2)
    }, function(e, t, r) {
        "use strict";

        function n(e) {
            return (n = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function(e) {
                return typeof e
            } : function(e) {
                return e && "function" == typeof Symbol && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e
            })(e)
        }
        Object.defineProperty(t, "__esModule", {
            value: !0
        }), t.I18n = t.I18N_CONSTANTS = void 0;
        var o = l(r(3)),
            a = l(r(4)),
            i = l(r(5)),
            u = l(r(6)),
            c = l(r(7)),
            f = l(r(8));

        function l(e) {
            return e && e.__esModule ? e : {
                default: e
            }
        }

        function s(e, t) {
            return (s = Object.setPrototypeOf || function(e, t) {
                return e.__proto__ = t, e
            })(e, t)
        }

        function d(e) {
            var t = function() {
                if ("undefined" == typeof Reflect || !Reflect.construct) return !1;
                if (Reflect.construct.sham) return !1;
                if ("function" == typeof Proxy) return !0;
                try {
                    return Date.prototype.toString.call(Reflect.construct(Date, [], (function() {}))), !0
                } catch (e) {
                    return !1
                }
            }();
            return function() {
                var r, n = h(e);
                if (t) {
                    var o = h(this).constructor;
                    r = Reflect.construct(n, arguments, o)
                } else r = n.apply(this, arguments);
                return g(this, r)
            }
        }

        function g(e, t) {
            return !t || "object" !== n(t) && "function" != typeof t ? function(e) {
                if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
                return e
            }(e) : t
        }

        function h(e) {
            return (h = Object.setPrototypeOf ? Object.getPrototypeOf : function(e) {
                return e.__proto__ || Object.getPrototypeOf(e)
            })(e)
        }

        function p(e, t) {
            var r = Object.keys(e);
            if (Object.getOwnPropertySymbols) {
                var n = Object.getOwnPropertySymbols(e);
                t && (n = n.filter((function(t) {
                    return Object.getOwnPropertyDescriptor(e, t).enumerable
                }))), r.push.apply(r, n)
            }
            return r
        }

        function C(e) {
            for (var t = 1; t < arguments.length; t++) {
                var r = null != arguments[t] ? arguments[t] : {};
                t % 2 ? p(Object(r), !0).forEach((function(t) {
                    v(e, t, r[t])
                })) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(r)) : p(Object(r)).forEach((function(t) {
                    Object.defineProperty(e, t, Object.getOwnPropertyDescriptor(r, t))
                }))
            }
            return e
        }

        function m(e, t) {
            if (!(e instanceof t)) throw new TypeError("Cannot call a class as a function")
        }

        function _(e, t) {
            for (var r = 0; r < t.length; r++) {
                var n = t[r];
                n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(e, n.key, n)
            }
        }

        function y(e, t, r) {
            return t && _(e.prototype, t), r && _(e, r), e
        }

        function v(e, t, r) {
            return t in e ? Object.defineProperty(e, t, {
                value: r,
                enumerable: !0,
                configurable: !0,
                writable: !0
            }) : e[t] = r, e
        }
        var S = "@udacity/ureact-i18n__locale",
            b = "udacity__locale",
            O = "@udacity/ureact-i18n__locale_map",
            L = "_jwt",
            E = "_jwt_token",
            P = {
                QUERY_PARAM: "_setLocaleFromQueryParam",
                BROWSER_COOKIE: "_setLocaleFromCookie",
                USER_PREFERENCE_API: "_setLocaleFromUserPreferenceApi",
                BROWSER_PREFERENCE: "_setLocaleFromBrowserPreference",
                GEO_API: "_setLocaleFromGeoAPI",
                DEFAULT: "_setDefaultLocale"
            },
            w = "__test__",
            R = "/api/classroom-content/v1/graphql",
            j = {
                DEFAULT_LOCALE: "en-us",
                DEFAULT_LOCATION: "US",
                GEO_API: "/api/geode",
                LOCALE_MAP: {
                    ar: "العربية",
                    de: "Deutsch",
                    "en-us": "English",
                    es: "Español",
                    id: "Bahasa Indonesia",
                    ja: "日本語",
                    ko: "한국어",
                    "pt-br": "Português",
                    uz: "O'zbekcha",
                    "zh-cn": "中文（简体）"
                }
            };
        t.I18N_CONSTANTS = j;
        var A = [P.QUERY_PARAM, P.BROWSER_COOKIE, P.BROWSER_PREFERENCE, P.DEFAULT],
            k = [P.QUERY_PARAM, P.USER_PREFERENCE_API, P.BROWSER_COOKIE, P.BROWSER_PREFERENCE, P.DEFAULT],
            T = function() {
                function e(t, r) {
                    var n = this;
                    m(this, e), v(this, "__", (function(e) {
                        for (var t = n._lookup(e, n.getLocale()), r = arguments.length, o = new Array(r > 1 ? r - 1 : 0), a = 1; a < r; a++) o[a - 1] = arguments[a];
                        return N(o) ? I(t(o[0])) : t(o[0])
                    })), v(this, "_lookup", a.default.memoize((function(e, t) {
                        var r, n = this._getLocaleMap(t);
                        return -1 !== (r = this.options.debug && t === w ? (0, f.default)(e) : e ? n && n[e] || e : "").indexOf("%s") ? r : a.default.template(r)
                    }), (function(e, t) {
                        return "".concat(t, "-").concat(e)
                    }))), v(this, "_tryNextLocaleMethod", (function(e) {
                        return function(t) {
                            return t || n[e]()
                        }
                    }));
                    this.options = C(C({}, {
                        allowPseudo: !1
                    }), r), this._setLocaleMap(t)
                }
                return y(e, [{
                    key: "activate",
                    value: function() {
                        var e = this;
                        return (this.options.refresh ? k : A).reduce((function(t, r) {
                            return t.then(e._tryNextLocaleMethod(r)).catch((function() {
                                return !1
                            }))
                        }), Promise.resolve(!1))
                    }
                }]), y(e, [{
                    key: "getLocale",
                    value: function() {
                        return window[S] || F("UREACT-I18N - MISUSE WARNING: Component might not have been activated properly.\n         Make sure you run i18n.activate() before your DOM renders (check the docs).\n         Default language is ".concat(j.DEFAULT_LOCALE)), (window[S] || j.DEFAULT_LOCALE).toLowerCase()
                    }
                }, {
                    key: "getLanguage",
                    value: function() {
                        return j.LOCALE_MAP[this.getLocale()]
                    }
                }, {
                    key: "saveLocaleToCookie",
                    value: function(e, t) {
                        if (!t) {
                            var r = new Date;
                            r.setFullYear(r.getFullYear() + 5), t = r.toUTCString()
                        }
                        document.cookie = "".concat(b, "=").concat(e, "; domain=.udacity.com;expires=").concat(t, "; path=/;secure;")
                    }
                }, {
                    key: "_setDefaultLocale",
                    value: function() {
                        return this._setLocale(j.DEFAULT_LOCALE)
                    }
                }, {
                    key: "_setLocaleFromCookie",
                    value: function() {
                        return this._setLocale(this._getCookieLocale())
                    }
                }, {
                    key: "_setLocaleFromGeoAPI",
                    value: function() {
                        var e = this;
                        return this.fetchGeoLocation().then((function(t) {
                            return e._setLocale(t.countryCode)
                        }))
                    }
                }, {
                    key: "_setLocaleFromBrowserPreference",
                    value: function() {
                        return this._setLocale(this._getNavigatorLocale())
                    }
                }, {
                    key: "_setLocaleFromQueryParam",
                    value: function() {
                        return this._setLocale(function(e, t) {
                            t || (t = window.location.href);
                            e = e.replace(/[[\]]/g, "\\$&");
                            var r = new RegExp("[?&]" + e + "(=([^&#]*)|&|#|$)").exec(t);
                            if (!r) return null;
                            if (!r[2]) return "";
                            return decodeURIComponent(r[2].replace(/\+/g, " "))
                        }("locale", this._getCurrentURLParams()))
                    }
                }, {
                    key: "_setLocaleFromUserPreference",
                    value: function() {
                        var e = this._getJWT();
                        return this._setLocale(e ? (0, u.default)(e).l : null)
                    }
                }, {
                    key: "_setLocaleFromUserPreferenceApi",
                    value: function() {
                        var e = this,
                            t = this._getJWT();
                        return t ? (0, i.default)({
                            method: "POST",
                            url: R,
                            headers: {
                                Authorization: "Bearer ".concat(t)
                            },
                            data: {
                                query: "\n    query UserSettingsQuery {\n      user {\n        preferred_language\n      }\n    }"
                            }
                        }).then((function(t) {
                            t.error && t.error.length && console.error(t.error[0]);
                            var r = a.default.get(t, "data.user.preferred_language") || a.default.get(t, "data.data.user.preferred_language");
                            return e._setLocale(r || null)
                        })) : this._setLocale(null)
                    }
                }, {
                    key: "fetchGeoLocation",
                    value: function() {
                        return i.default.get(j.GEO_API).then((function(e) {
                            return e.data
                        }))
                    }
                }, {
                    key: "_setLocale",
                    value: function(e) {
                        var t = this;
                        return new Promise((function(r) {
                            return a.default.isString(e) && "" !== e ? (e = e.toLowerCase(), window[S] = e, t.options && t.options.moment && t.options.moment.locale(e), r(t.getLocale())) : r(!1)
                        }))
                    }
                }, {
                    key: "_setLocaleMap",
                    value: function(e) {
                        window[O] = function(e, t) {
                            var r;
                            a.default.isFunction(e) ? (F("\n      UREACT-I18N - DEPRECATION WARNING: use of 'require.context' when initializing a new I18n instance is deprecated. See updated usage documentation at: https://github.com/udacity/ureact-i18n#step-1-instantiate-a-new-i18n-object\n    "), r = a.default.chain(e.keys()).filter((function(e) {
                                return /\.js$/.test(e)
                            })).map((function(t) {
                                var r;
                                return [(r = t, a.default.last(r.split("/")).split(".")[0]).toLowerCase(), e(t)]
                            })).fromPairs().value()) : r = e;
                            return a.default.reduce(r, (function(e, r, n) {
                                return e[n] = a.default.isUndefined(t) || a.default.isUndefined(t[n]) ? r : C(C({}, t[n]), r), e
                            }), t || {})
                        }(e, window[O])
                    }
                }, {
                    key: "_getLocaleMap",
                    value: function(e) {
                        return window[O][e] || window[O][e.split("-")[0]]
                    }
                }, {
                    key: "_getCurrentURLParams",
                    value: function() {
                        return window.location.search
                    }
                }, {
                    key: "_getNavigatorLocale",
                    value: function() {
                        return navigator.language || navigator.userLanguage || navigator.browserLanguage || navigator.userLanguage
                    }
                }, {
                    key: "_getCookieLocale",
                    value: function() {
                        var e = (document.cookie ? document.cookie.split(";") : []).filter((function(e) {
                            return e.indexOf(b) > -1
                        }))[0];
                        if (e) {
                            var t = e.split("=")[1];
                            if (t && "" !== t.trim()) return t.trim()
                        }
                    }
                }, {
                    key: "_getJWT",
                    value: function() {
                        return o.default.get(L) || o.default.get(E)
                    }
                }]), e
            }();

        function I(e) {
            var t = function(t) {
                ! function(e, t) {
                    if ("function" != typeof t && null !== t) throw new TypeError("Super expression must either be null or a function");
                    e.prototype = Object.create(t && t.prototype, {
                        constructor: {
                            value: e,
                            writable: !0,
                            configurable: !0
                        }
                    }), t && s(e, t)
                }(n, t);
                var r = d(n);

                function n() {
                    return m(this, n), r.apply(this, arguments)
                }
                return y(n, [{
                    key: "render",
                    value: function() {
                        return c.default.createElement("span", {
                            dangerouslySetInnerHTML: {
                                __html: e
                            }
                        })
                    }
                }]), n
            }(c.default.Component);
            return c.default.createElement(t, null)
        }

        function N(e) {
            var t = e[e.length - 1];
            return t && t.renderHTML
        }

        function F(e) {
            0
        }
        t.I18n = T
    }, function(e, t) {
        e.exports = require("cookies-js")
    }, function(e, t) {
        e.exports = require("lodash")
    }, function(e, t) {
        e.exports = require("axios")
    }, function(e, t) {
        e.exports = require("jwt-decode")
    }, function(e, t) {
        e.exports = require("react")
    }, function(e, t, r) {
        "use strict";
        Object.defineProperty(t, "__esModule", {
            value: !0
        }), t.default = function() {
            var e = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "",
                t = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : .35,
                r = /(<%=?.+?%>)/g;
            return e.split(r).map((function(e) {
                return r.test(e) ? e : e.split(/\b/).map((function(e) {
                    return o(a(e), t)
                })).join("")
            })).join("")
        };
        var n = {
            A: String.fromCharCode(192, 193, 194, 195, 196, 197, 256, 258, 260, 506, 512, 514),
            a: String.fromCharCode(224, 225, 226, 227, 228, 229, 257, 259, 261, 507, 513, 515),
            B: String.fromCharCode(223, 385, 579, 665),
            b: String.fromCharCode(384, 386, 387, 388, 389, 595),
            C: String.fromCharCode(262, 264, 266, 268),
            c: String.fromCharCode(263, 265, 267, 269),
            D: String.fromCharCode(270, 272, 393, 394),
            d: String.fromCharCode(271, 273),
            E: String.fromCharCode(274, 276, 278, 280, 282, 516, 518),
            e: String.fromCharCode(275, 277, 279, 281, 283, 517, 519),
            F: String.fromCharCode(401),
            f: String.fromCharCode(402),
            G: String.fromCharCode(284, 286, 288, 290),
            g: String.fromCharCode(285, 287, 289, 291),
            H: String.fromCharCode(292, 294),
            h: String.fromCharCode(293, 295),
            I: String.fromCharCode(296, 298, 300, 302, 304),
            i: String.fromCharCode(297, 299, 301, 303, 305),
            J: String.fromCharCode(309),
            j: String.fromCharCode(308),
            K: String.fromCharCode(310, 408),
            k: String.fromCharCode(311, 312, 409),
            L: String.fromCharCode(313, 315, 317, 319, 321),
            l: String.fromCharCode(314, 316, 318, 320, 322),
            N: String.fromCharCode(323, 325, 327, 330, 413),
            n: String.fromCharCode(324, 326, 328, 329, 331, 414),
            O: String.fromCharCode(332, 334, 336, 416),
            o: String.fromCharCode(333, 335, 337, 417),
            P: String.fromCharCode(420),
            p: String.fromCharCode(421, 447),
            Q: String.fromCharCode(490, 492),
            q: String.fromCharCode(491, 493, 587),
            R: String.fromCharCode(340, 342, 344, 422),
            r: String.fromCharCode(341, 343, 345),
            S: String.fromCharCode(346, 348, 350, 352),
            s: String.fromCharCode(347, 349, 351, 353),
            T: String.fromCharCode(354, 356, 358),
            t: String.fromCharCode(355, 357, 359),
            U: String.fromCharCode(360, 362, 364, 366, 368, 370),
            u: String.fromCharCode(361, 363, 365, 367, 369, 371),
            W: String.fromCharCode(372),
            w: String.fromCharCode(373),
            Y: String.fromCharCode(374, 376, 435, 562, 590),
            y: String.fromCharCode(375, 436, 563, 591),
            Z: String.fromCharCode(377, 379, 381, 437),
            z: String.fromCharCode(378, 380, 382, 438)
        };

        function o() {
            var e = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "",
                t = arguments.length > 1 ? arguments[1] : void 0;
            return e + "~".repeat(Math.floor(e.trim().length * t))
        }

        function a() {
            var e = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : "";
            return e.split("").map((function(e) {
                var t = n[e] || e;
                return t[Math.floor(Math.random() * t.length)]
            })).join("")
        }
        e.exports = t.default
    }])
}));
