// m3u8-parser/dist/m3u8-parser.es.js
/*! @name m3u8-parser @version 4.5.0 @license Apache-2.0 */
import _inheritsLoose from '@babel/runtime/helpers/inheritsLoose';
import Stream from '@videojs/vhs-utils/dist/stream.js';
import _extends from '@babel/runtime/helpers/extends';
import _assertThisInitialized from '@babel/runtime/helpers/assertThisInitialized';
import decodeB64ToUint8Array from '@videojs/vhs-utils/dist/decode-b64-to-uint8-array.js';

/**
 * A stream that buffers string input and generates a `data` event for each
 * line.
 *
 * @class LineStream
 * @extends Stream
 */

var LineStream =
    /*#__PURE__*/
    function(_Stream) {
        _inheritsLoose(LineStream, _Stream);

        function LineStream() {
            var _this;

            _this = _Stream.call(this) || this;
            _this.buffer = '';
            return _this;
        }
        /**
         * Add new data to be parsed.
         *
         * @param {string} data the text to process
         */


        var _proto = LineStream.prototype;

        _proto.push = function push(data) {
            var nextNewline;
            this.buffer += data;
            nextNewline = this.buffer.indexOf('\n');

            for (; nextNewline > -1; nextNewline = this.buffer.indexOf('\n')) {
                this.trigger('data', this.buffer.substring(0, nextNewline));
                this.buffer = this.buffer.substring(nextNewline + 1);
            }
        };

        return LineStream;
    }(Stream);

/**
 * "forgiving" attribute list psuedo-grammar:
 * attributes -> keyvalue (',' keyvalue)*
 * keyvalue   -> key '=' value
 * key        -> [^=]*
 * value      -> '"' [^"]* '"' | [^,]*
 */

var attributeSeparator = function attributeSeparator() {
    var key = '[^=]*';
    var value = '"[^"]*"|[^,]*';
    var keyvalue = '(?:' + key + ')=(?:' + value + ')';
    return new RegExp('(?:^|,)(' + keyvalue + ')');
};
/**
 * Parse attributes from a line given the separator
 *
 * @param {string} attributes the attribute line to parse
 */


var parseAttributes = function parseAttributes(attributes) {
    // split the string using attributes as the separator
    var attrs = attributes.split(attributeSeparator());
    var result = {};
    var i = attrs.length;
    var attr;

    while (i--) {
        // filter out unmatched portions of the string
        if (attrs[i] === '') {
            continue;
        } // split the key and value


        attr = /([^=]*)=(.*)/.exec(attrs[i]).slice(1); // trim whitespace and remove optional quotes around the value

        attr[0] = attr[0].replace(/^\s+|\s+$/g, '');
        attr[1] = attr[1].replace(/^\s+|\s+$/g, '');
        attr[1] = attr[1].replace(/^['"](.*)['"]$/g, '$1');
        result[attr[0]] = attr[1];
    }

    return result;
};
/**
 * A line-level M3U8 parser event stream. It expects to receive input one
 * line at a time and performs a context-free parse of its contents. A stream
 * interpretation of a manifest can be useful if the manifest is expected to
 * be too large to fit comfortably into memory or the entirety of the input
 * is not immediately available. Otherwise, it's probably much easier to work
 * with a regular `Parser` object.
 *
 * Produces `data` events with an object that captures the parser's
 * interpretation of the input. That object has a property `tag` that is one
 * of `uri`, `comment`, or `tag`. URIs only have a single additional
 * property, `line`, which captures the entirety of the input without
 * interpretation. Comments similarly have a single additional property
 * `text` which is the input without the leading `#`.
 *
 * Tags always have a property `tagType` which is the lower-cased version of
 * the M3U8 directive without the `#EXT` or `#EXT-X-` prefix. For instance,
 * `#EXT-X-MEDIA-SEQUENCE` becomes `media-sequence` when parsed. Unrecognized
 * tags are given the tag type `unknown` and a single additional property
 * `data` with the remainder of the input.
 *
 * @class ParseStream
 * @extends Stream
 */


var ParseStream =
    /*#__PURE__*/
    function(_Stream) {
        _inheritsLoose(ParseStream, _Stream);

        function ParseStream() {
            var _this;

            _this = _Stream.call(this) || this;
            _this.customParsers = [];
            _this.tagMappers = [];
            return _this;
        }
        /**
         * Parses an additional line of input.
         *
         * @param {string} line a single line of an M3U8 file to parse
         */


        var _proto = ParseStream.prototype;

        _proto.push = function push(line) {
            var _this2 = this;

            var match;
            var event; // strip whitespace

            line = line.trim();

            if (line.length === 0) {
                // ignore empty lines
                return;
            } // URIs


            if (line[0] !== '#') {
                this.trigger('data', {
                    type: 'uri',
                    uri: line
                });
                return;
            } // map tags


            var newLines = this.tagMappers.reduce(function(acc, mapper) {
                var mappedLine = mapper(line); // skip if unchanged

                if (mappedLine === line) {
                    return acc;
                }

                return acc.concat([mappedLine]);
            }, [line]);
            newLines.forEach(function(newLine) {
                for (var i = 0; i < _this2.customParsers.length; i++) {
                    if (_this2.customParsers[i].call(_this2, newLine)) {
                        return;
                    }
                } // Comments


                if (newLine.indexOf('#EXT') !== 0) {
                    _this2.trigger('data', {
                        type: 'comment',
                        text: newLine.slice(1)
                    });

                    return;
                } // strip off any carriage returns here so the regex matching
                // doesn't have to account for them.


                newLine = newLine.replace('\r', ''); // Tags

                match = /^#EXTM3U/.exec(newLine);

                if (match) {
                    _this2.trigger('data', {
                        type: 'tag',
                        tagType: 'm3u'
                    });

                    return;
                }

                match = /^#EXTINF:?([0-9\.]*)?,?(.*)?$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'inf'
                    };

                    if (match[1]) {
                        event.duration = parseFloat(match[1]);
                    }

                    if (match[2]) {
                        event.title = match[2];
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-TARGETDURATION:?([0-9.]*)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'targetduration'
                    };

                    if (match[1]) {
                        event.duration = parseInt(match[1], 10);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#ZEN-TOTAL-DURATION:?([0-9.]*)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'totalduration'
                    };

                    if (match[1]) {
                        event.duration = parseInt(match[1], 10);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-VERSION:?([0-9.]*)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'version'
                    };

                    if (match[1]) {
                        event.version = parseInt(match[1], 10);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-MEDIA-SEQUENCE:?(\-?[0-9.]*)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'media-sequence'
                    };

                    if (match[1]) {
                        event.number = parseInt(match[1], 10);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-DISCONTINUITY-SEQUENCE:?(\-?[0-9.]*)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'discontinuity-sequence'
                    };

                    if (match[1]) {
                        event.number = parseInt(match[1], 10);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-PLAYLIST-TYPE:?(.*)?$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'playlist-type'
                    };

                    if (match[1]) {
                        event.playlistType = match[1];
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-BYTERANGE:?([0-9.]*)?@?([0-9.]*)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'byterange'
                    };

                    if (match[1]) {
                        event.length = parseInt(match[1], 10);
                    }

                    if (match[2]) {
                        event.offset = parseInt(match[2], 10);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-ALLOW-CACHE:?(YES|NO)?/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'allow-cache'
                    };

                    if (match[1]) {
                        event.allowed = !/NO/.test(match[1]);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-MAP:?(.*)$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'map'
                    };

                    if (match[1]) {
                        var attributes = parseAttributes(match[1]);

                        if (attributes.URI) {
                            event.uri = attributes.URI;
                        }

                        if (attributes.BYTERANGE) {
                            var _attributes$BYTERANGE = attributes.BYTERANGE.split('@'),
                                length = _attributes$BYTERANGE[0],
                                offset = _attributes$BYTERANGE[1];

                            event.byterange = {};

                            if (length) {
                                event.byterange.length = parseInt(length, 10);
                            }

                            if (offset) {
                                event.byterange.offset = parseInt(offset, 10);
                            }
                        }
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-STREAM-INF:?(.*)$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'stream-inf'
                    };

                    if (match[1]) {
                        event.attributes = parseAttributes(match[1]);

                        if (event.attributes.RESOLUTION) {
                            var split = event.attributes.RESOLUTION.split('x');
                            var resolution = {};

                            if (split[0]) {
                                resolution.width = parseInt(split[0], 10);
                            }

                            if (split[1]) {
                                resolution.height = parseInt(split[1], 10);
                            }

                            event.attributes.RESOLUTION = resolution;
                        }

                        if (event.attributes.BANDWIDTH) {
                            event.attributes.BANDWIDTH = parseInt(event.attributes.BANDWIDTH, 10);
                        }

                        if (event.attributes['PROGRAM-ID']) {
                            event.attributes['PROGRAM-ID'] = parseInt(event.attributes['PROGRAM-ID'], 10);
                        }
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-MEDIA:?(.*)$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'media'
                    };

                    if (match[1]) {
                        event.attributes = parseAttributes(match[1]);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-ENDLIST/.exec(newLine);

                if (match) {
                    _this2.trigger('data', {
                        type: 'tag',
                        tagType: 'endlist'
                    });

                    return;
                }

                match = /^#EXT-X-DISCONTINUITY/.exec(newLine);

                if (match) {
                    _this2.trigger('data', {
                        type: 'tag',
                        tagType: 'discontinuity'
                    });

                    return;
                }

                match = /^#EXT-X-PROGRAM-DATE-TIME:?(.*)$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'program-date-time'
                    };

                    if (match[1]) {
                        event.dateTimeString = match[1];
                        event.dateTimeObject = new Date(match[1]);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-KEY:?(.*)$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'key'
                    };

                    if (match[1]) {
                        event.attributes = parseAttributes(match[1]); // parse the IV string into a Uint32Array

                        if (event.attributes.IV) {
                            if (event.attributes.IV.substring(0, 2).toLowerCase() === '0x') {
                                event.attributes.IV = event.attributes.IV.substring(2);
                            }

                            event.attributes.IV = event.attributes.IV.match(/.{8}/g);
                            event.attributes.IV[0] = parseInt(event.attributes.IV[0], 16);
                            event.attributes.IV[1] = parseInt(event.attributes.IV[1], 16);
                            event.attributes.IV[2] = parseInt(event.attributes.IV[2], 16);
                            event.attributes.IV[3] = parseInt(event.attributes.IV[3], 16);
                            event.attributes.IV = new Uint32Array(event.attributes.IV);
                        }
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-START:?(.*)$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'start'
                    };

                    if (match[1]) {
                        event.attributes = parseAttributes(match[1]);
                        event.attributes['TIME-OFFSET'] = parseFloat(event.attributes['TIME-OFFSET']);
                        event.attributes.PRECISE = /YES/.test(event.attributes.PRECISE);
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-CUE-OUT-CONT:?(.*)?$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'cue-out-cont'
                    };

                    if (match[1]) {
                        event.data = match[1];
                    } else {
                        event.data = '';
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-CUE-OUT:?(.*)?$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'cue-out'
                    };

                    if (match[1]) {
                        event.data = match[1];
                    } else {
                        event.data = '';
                    }

                    _this2.trigger('data', event);

                    return;
                }

                match = /^#EXT-X-CUE-IN:?(.*)?$/.exec(newLine);

                if (match) {
                    event = {
                        type: 'tag',
                        tagType: 'cue-in'
                    };

                    if (match[1]) {
                        event.data = match[1];
                    } else {
                        event.data = '';
                    }

                    _this2.trigger('data', event);

                    return;
                } // unknown tag type


                _this2.trigger('data', {
                    type: 'tag',
                    data: newLine.slice(4)
                });
            });
        }
        /**
         * Add a parser for custom headers
         *
         * @param {Object}   options              a map of options for the added parser
         * @param {RegExp}   options.expression   a regular expression to match the custom header
         * @param {string}   options.customType   the custom type to register to the output
         * @param {Function} [options.dataParser] function to parse the line into an object
         * @param {boolean}  [options.segment]    should tag data be attached to the segment object
         */
        ;

        _proto.addParser = function addParser(_ref) {
            var _this3 = this;

            var expression = _ref.expression,
                customType = _ref.customType,
                dataParser = _ref.dataParser,
                segment = _ref.segment;

            if (typeof dataParser !== 'function') {
                dataParser = function dataParser(line) {
                    return line;
                };
            }

            this.customParsers.push(function(line) {
                var match = expression.exec(line);

                if (match) {
                    _this3.trigger('data', {
                        type: 'custom',
                        data: dataParser(line),
                        customType: customType,
                        segment: segment
                    });

                    return true;
                }
            });
        }
        /**
         * Add a custom header mapper
         *
         * @param {Object}   options
         * @param {RegExp}   options.expression   a regular expression to match the custom header
         * @param {Function} options.map          function to translate tag into a different tag
         */
        ;

        _proto.addTagMapper = function addTagMapper(_ref2) {
            var expression = _ref2.expression,
                map = _ref2.map;

            var mapFn = function mapFn(line) {
                if (expression.test(line)) {
                    return map(line);
                }

                return line;
            };

            this.tagMappers.push(mapFn);
        };

        return ParseStream;
    }(Stream);

/**
 * A parser for M3U8 files. The current interpretation of the input is
 * exposed as a property `manifest` on parser objects. It's just two lines to
 * create and parse a manifest once you have the contents available as a string:
 *
 * ```js
 * var parser = new m3u8.Parser();
 * parser.push(xhr.responseText);
 * ```
 *
 * New input can later be applied to update the manifest object by calling
 * `push` again.
 *
 * The parser attempts to create a usable manifest object even if the
 * underlying input is somewhat nonsensical. It emits `info` and `warning`
 * events during the parse if it encounters input that seems invalid or
 * requires some property of the manifest object to be defaulted.
 *
 * @class Parser
 * @extends Stream
 */

var Parser =
    /*#__PURE__*/
    function(_Stream) {
        _inheritsLoose(Parser, _Stream);

        function Parser() {
            var _this;

            _this = _Stream.call(this) || this;
            _this.lineStream = new LineStream();
            _this.parseStream = new ParseStream();

            _this.lineStream.pipe(_this.parseStream);
            /* eslint-disable consistent-this */


            var self = _assertThisInitialized(_this);
            /* eslint-enable consistent-this */


            var uris = [];
            var currentUri = {}; // if specified, the active EXT-X-MAP definition

            var currentMap; // if specified, the active decryption key

            var _key;

            var noop = function noop() {};

            var defaultMediaGroups = {
                'AUDIO': {},
                'VIDEO': {},
                'CLOSED-CAPTIONS': {},
                'SUBTITLES': {}
            }; // This is the Widevine UUID from DASH IF IOP. The same exact string is
            // used in MPDs with Widevine encrypted streams.

            var widevineUuid = 'urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed'; // group segments into numbered timelines delineated by discontinuities

            var currentTimeline = 0; // the manifest is empty until the parse stream begins delivering data

            _this.manifest = {
                allowCache: true,
                discontinuityStarts: [],
                segments: []
            }; // keep track of the last seen segment's byte range end, as segments are not required
            // to provide the offset, in which case it defaults to the next byte after the
            // previous segment

            var lastByterangeEnd = 0; // update the manifest with the m3u8 entry from the parse stream

            _this.parseStream.on('data', function(entry) {
                var mediaGroup;
                var rendition;
                ({
                    tag: function tag() {
                        // switch based on the tag type
                        (({
                            'allow-cache': function allowCache() {
                                this.manifest.allowCache = entry.allowed;

                                if (!('allowed' in entry)) {
                                    this.trigger('info', {
                                        message: 'defaulting allowCache to YES'
                                    });
                                    this.manifest.allowCache = true;
                                }
                            },
                            byterange: function byterange() {
                                var byterange = {};

                                if ('length' in entry) {
                                    currentUri.byterange = byterange;
                                    byterange.length = entry.length;

                                    if (!('offset' in entry)) {
                                        /*
                                         * From the latest spec (as of this writing):
                                         * https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-4.3.2.2
                                         *
                                         * Same text since EXT-X-BYTERANGE's introduction in draft 7:
                                         * https://tools.ietf.org/html/draft-pantos-http-live-streaming-07#section-3.3.1)
                                         *
                                         * "If o [offset] is not present, the sub-range begins at the next byte
                                         * following the sub-range of the previous media segment."
                                         */
                                        entry.offset = lastByterangeEnd;
                                    }
                                }

                                if ('offset' in entry) {
                                    currentUri.byterange = byterange;
                                    byterange.offset = entry.offset;
                                }

                                lastByterangeEnd = byterange.offset + byterange.length;
                            },
                            endlist: function endlist() {
                                this.manifest.endList = true;
                            },
                            inf: function inf() {
                                if (!('mediaSequence' in this.manifest)) {
                                    this.manifest.mediaSequence = 0;
                                    this.trigger('info', {
                                        message: 'defaulting media sequence to zero'
                                    });
                                }

                                if (!('discontinuitySequence' in this.manifest)) {
                                    this.manifest.discontinuitySequence = 0;
                                    this.trigger('info', {
                                        message: 'defaulting discontinuity sequence to zero'
                                    });
                                }

                                if (entry.duration > 0) {
                                    currentUri.duration = entry.duration;
                                }

                                if (entry.duration === 0) {
                                    currentUri.duration = 0.01;
                                    this.trigger('info', {
                                        message: 'updating zero segment duration to a small value'
                                    });
                                }

                                this.manifest.segments = uris;
                            },
                            key: function key() {
                                if (!entry.attributes) {
                                    this.trigger('warn', {
                                        message: 'ignoring key declaration without attribute list'
                                    });
                                    return;
                                } // clear the active encryption key


                                if (entry.attributes.METHOD === 'NONE') {
                                    _key = null;
                                    return;
                                }

                                if (!entry.attributes.URI) {
                                    this.trigger('warn', {
                                        message: 'ignoring key declaration without URI'
                                    });
                                    return;
                                } // check if the content is encrypted for Widevine
                                // Widevine/HLS spec: https://storage.googleapis.com/wvdocs/Widevine_DRM_HLS.pdf


                                if (entry.attributes.KEYFORMAT === widevineUuid) {
                                    var VALID_METHODS = ['SAMPLE-AES', 'SAMPLE-AES-CTR', 'SAMPLE-AES-CENC'];

                                    if (VALID_METHODS.indexOf(entry.attributes.METHOD) === -1) {
                                        this.trigger('warn', {
                                            message: 'invalid key method provided for Widevine'
                                        });
                                        return;
                                    }

                                    if (entry.attributes.METHOD === 'SAMPLE-AES-CENC') {
                                        this.trigger('warn', {
                                            message: 'SAMPLE-AES-CENC is deprecated, please use SAMPLE-AES-CTR instead'
                                        });
                                    }

                                    if (entry.attributes.URI.substring(0, 23) !== 'data:text/plain;base64,') {
                                        this.trigger('warn', {
                                            message: 'invalid key URI provided for Widevine'
                                        });
                                        return;
                                    }

                                    if (!(entry.attributes.KEYID && entry.attributes.KEYID.substring(0, 2) === '0x')) {
                                        this.trigger('warn', {
                                            message: 'invalid key ID provided for Widevine'
                                        });
                                        return;
                                    } // if Widevine key attributes are valid, store them as `contentProtection`
                                    // on the manifest to emulate Widevine tag structure in a DASH mpd


                                    this.manifest.contentProtection = {
                                        'com.widevine.alpha': {
                                            attributes: {
                                                schemeIdUri: entry.attributes.KEYFORMAT,
                                                // remove '0x' from the key id string
                                                keyId: entry.attributes.KEYID.substring(2)
                                            },
                                            // decode the base64-encoded PSSH box
                                            pssh: decodeB64ToUint8Array(entry.attributes.URI.split(',')[1])
                                        }
                                    };
                                    return;
                                }

                                if (!entry.attributes.METHOD) {
                                    this.trigger('warn', {
                                        message: 'defaulting key method to AES-128'
                                    });
                                } // setup an encryption key for upcoming segments


                                _key = {
                                    method: entry.attributes.METHOD || 'AES-128',
                                    uri: entry.attributes.URI
                                };

                                if (typeof entry.attributes.IV !== 'undefined') {
                                    _key.iv = entry.attributes.IV;
                                }
                            },
                            'media-sequence': function mediaSequence() {
                                if (!isFinite(entry.number)) {
                                    this.trigger('warn', {
                                        message: 'ignoring invalid media sequence: ' + entry.number
                                    });
                                    return;
                                }

                                this.manifest.mediaSequence = entry.number;
                            },
                            'discontinuity-sequence': function discontinuitySequence() {
                                if (!isFinite(entry.number)) {
                                    this.trigger('warn', {
                                        message: 'ignoring invalid discontinuity sequence: ' + entry.number
                                    });
                                    return;
                                }

                                this.manifest.discontinuitySequence = entry.number;
                                currentTimeline = entry.number;
                            },
                            'playlist-type': function playlistType() {
                                if (!/VOD|EVENT/.test(entry.playlistType)) {
                                    this.trigger('warn', {
                                        message: 'ignoring unknown playlist type: ' + entry.playlist
                                    });
                                    return;
                                }

                                this.manifest.playlistType = entry.playlistType;
                            },
                            map: function map() {
                                currentMap = {};

                                if (entry.uri) {
                                    currentMap.uri = entry.uri;
                                }

                                if (entry.byterange) {
                                    currentMap.byterange = entry.byterange;
                                }
                            },
                            'stream-inf': function streamInf() {
                                this.manifest.playlists = uris;
                                this.manifest.mediaGroups = this.manifest.mediaGroups || defaultMediaGroups;

                                if (!entry.attributes) {
                                    this.trigger('warn', {
                                        message: 'ignoring empty stream-inf attributes'
                                    });
                                    return;
                                }

                                if (!currentUri.attributes) {
                                    currentUri.attributes = {};
                                }

                                _extends(currentUri.attributes, entry.attributes);
                            },
                            media: function media() {
                                this.manifest.mediaGroups = this.manifest.mediaGroups || defaultMediaGroups;

                                if (!(entry.attributes && entry.attributes.TYPE && entry.attributes['GROUP-ID'] && entry.attributes.NAME)) {
                                    this.trigger('warn', {
                                        message: 'ignoring incomplete or missing media group'
                                    });
                                    return;
                                } // find the media group, creating defaults as necessary


                                var mediaGroupType = this.manifest.mediaGroups[entry.attributes.TYPE];
                                mediaGroupType[entry.attributes['GROUP-ID']] = mediaGroupType[entry.attributes['GROUP-ID']] || {};
                                mediaGroup = mediaGroupType[entry.attributes['GROUP-ID']]; // collect the rendition metadata

                                rendition = {
                                    default: /yes/i.test(entry.attributes.DEFAULT)
                                };

                                if (rendition.default) {
                                    rendition.autoselect = true;
                                } else {
                                    rendition.autoselect = /yes/i.test(entry.attributes.AUTOSELECT);
                                }

                                if (entry.attributes.LANGUAGE) {
                                    rendition.language = entry.attributes.LANGUAGE;
                                }

                                if (entry.attributes.URI) {
                                    rendition.uri = entry.attributes.URI;
                                }

                                if (entry.attributes['INSTREAM-ID']) {
                                    rendition.instreamId = entry.attributes['INSTREAM-ID'];
                                }

                                if (entry.attributes.CHARACTERISTICS) {
                                    rendition.characteristics = entry.attributes.CHARACTERISTICS;
                                }

                                if (entry.attributes.FORCED) {
                                    rendition.forced = /yes/i.test(entry.attributes.FORCED);
                                } // insert the new rendition


                                mediaGroup[entry.attributes.NAME] = rendition;
                            },
                            discontinuity: function discontinuity() {
                                currentTimeline += 1;
                                currentUri.discontinuity = true;
                                this.manifest.discontinuityStarts.push(uris.length);
                            },
                            'program-date-time': function programDateTime() {
                                if (typeof this.manifest.dateTimeString === 'undefined') {
                                    // PROGRAM-DATE-TIME is a media-segment tag, but for backwards
                                    // compatibility, we add the first occurence of the PROGRAM-DATE-TIME tag
                                    // to the manifest object
                                    // TODO: Consider removing this in future major version
                                    this.manifest.dateTimeString = entry.dateTimeString;
                                    this.manifest.dateTimeObject = entry.dateTimeObject;
                                }

                                currentUri.dateTimeString = entry.dateTimeString;
                                currentUri.dateTimeObject = entry.dateTimeObject;
                            },
                            targetduration: function targetduration() {
                                if (!isFinite(entry.duration) || entry.duration < 0) {
                                    this.trigger('warn', {
                                        message: 'ignoring invalid target duration: ' + entry.duration
                                    });
                                    return;
                                }

                                this.manifest.targetDuration = entry.duration;
                            },
                            totalduration: function totalduration() {
                                if (!isFinite(entry.duration) || entry.duration < 0) {
                                    this.trigger('warn', {
                                        message: 'ignoring invalid total duration: ' + entry.duration
                                    });
                                    return;
                                }

                                this.manifest.totalDuration = entry.duration;
                            },
                            start: function start() {
                                if (!entry.attributes || isNaN(entry.attributes['TIME-OFFSET'])) {
                                    this.trigger('warn', {
                                        message: 'ignoring start declaration without appropriate attribute list'
                                    });
                                    return;
                                }

                                this.manifest.start = {
                                    timeOffset: entry.attributes['TIME-OFFSET'],
                                    precise: entry.attributes.PRECISE
                                };
                            },
                            'cue-out': function cueOut() {
                                currentUri.cueOut = entry.data;
                            },
                            'cue-out-cont': function cueOutCont() {
                                currentUri.cueOutCont = entry.data;
                            },
                            'cue-in': function cueIn() {
                                currentUri.cueIn = entry.data;
                            }
                        })[entry.tagType] || noop).call(self);
                    },
                    uri: function uri() {
                        currentUri.uri = entry.uri;
                        uris.push(currentUri); // if no explicit duration was declared, use the target duration

                        if (this.manifest.targetDuration && !('duration' in currentUri)) {
                            this.trigger('warn', {
                                message: 'defaulting segment duration to the target duration'
                            });
                            currentUri.duration = this.manifest.targetDuration;
                        } // annotate with encryption information, if necessary


                        if (_key) {
                            currentUri.key = _key;
                        }

                        currentUri.timeline = currentTimeline; // annotate with initialization segment information, if necessary

                        if (currentMap) {
                            currentUri.map = currentMap;
                        } // prepare for the next URI


                        currentUri = {};
                    },
                    comment: function comment() { // comments are not important for playback
                    },
                    custom: function custom() {
                        // if this is segment-level data attach the output to the segment
                        if (entry.segment) {
                            currentUri.custom = currentUri.custom || {};
                            currentUri.custom[entry.customType] = entry.data; // if this is manifest-level data attach to the top level manifest object
                        } else {
                            this.manifest.custom = this.manifest.custom || {};
                            this.manifest.custom[entry.customType] = entry.data;
                        }
                    }
                })[entry.type].call(self);
            });

            return _this;
        }
        /**
         * Parse the input string and update the manifest object.
         *
         * @param {string} chunk a potentially incomplete portion of the manifest
         */


        var _proto = Parser.prototype;

        _proto.push = function push(chunk) {
            this.lineStream.push(chunk);
        }
        /**
         * Flush any remaining input. This can be handy if the last line of an M3U8
         * manifest did not contain a trailing newline but the file has been
         * completely received.
         */
        ;

        _proto.end = function end() {
            // flush any buffered input
            this.lineStream.push('\n');
        }
        /**
         * Add an additional parser for non-standard tags
         *
         * @param {Object}   options              a map of options for the added parser
         * @param {RegExp}   options.expression   a regular expression to match the custom header
         * @param {string}   options.type         the type to register to the output
         * @param {Function} [options.dataParser] function to parse the line into an object
         * @param {boolean}  [options.segment]    should tag data be attached to the segment object
         */
        ;

        _proto.addParser = function addParser(options) {
            this.parseStream.addParser(options);
        }
        /**
         * Add a custom header mapper
         *
         * @param {Object}   options
         * @param {RegExp}   options.expression   a regular expression to match the custom header
         * @param {Function} options.map          function to translate tag into a different tag
         */
        ;

        _proto.addTagMapper = function addTagMapper(options) {
            this.parseStream.addTagMapper(options);
        };

        return Parser;
    }(Stream);

export {
    LineStream,
    ParseStream,
    Parser
};
